{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Exploring all data from a tissue in the Cell Census\n",
    "\n",
    "The Cell Census is a versioned container for the single-cell data hosted at [CELLxGENE Discover](https://cellxgene.cziscience.com/). The Cell Census utilizes [SOMA](https://github.com/single-cell-data/SOMA/blob/main/abstract_specification.md) powered by [TileDB](https://tiledb.com/products/tiledb-embedded) for storing, accessing, and efficiently filtering data.\n",
    "\n",
    "This notebook shows you how to learn about the Cell Census contents and how to query it.\n",
    "\n",
    "This notebook provides a demonstration for:\n",
    "\n",
    "1. Learning about the lung data.\n",
    "   1. Learning about cells of lung data.\n",
    "   1. Larning about genes of lung data.\n",
    "1. Fetching all human lung data from the Cell Census.\n",
    "1. Obtaining QC metrics for this data slice.\n",
    "\n",
    "## Learning about the lung data in the Cell Census\n",
    "\n",
    "First we open the Cell Census, if you are not familiar with the basics of Cell Census API you should take a look at notebook \"Learning about the CELLxGENE Cell Census\" at `comp_bio_census_info.ipynb`.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import cell_census\n",
    "import scanpy as sc\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from scipy.sparse import coo_matrix\n",
    "\n",
    "census = cell_census.open_soma(census_version=\"latest\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's first take a look at the number of cells from human lung:\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>soma_joinid</th>\n",
       "      <th>organism</th>\n",
       "      <th>category</th>\n",
       "      <th>ontology_term_id</th>\n",
       "      <th>unique_cell_count</th>\n",
       "      <th>total_cell_count</th>\n",
       "      <th>label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>881</th>\n",
       "      <td>881</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>tissue_general</td>\n",
       "      <td>UBERON:0002048</td>\n",
       "      <td>1194348</td>\n",
       "      <td>4351243</td>\n",
       "      <td>lung</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     soma_joinid      organism        category ontology_term_id  \\\n",
       "881          881  Homo sapiens  tissue_general   UBERON:0002048   \n",
       "\n",
       "     unique_cell_count  total_cell_count label  \n",
       "881            1194348           4351243  lung  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "summary_table = census[\"census_info\"][\"summary_cell_counts\"].read().concat().to_pandas()\n",
    "summary_table.query(\"organism == 'Homo sapiens' & category == 'tissue_general' & label =='lung'\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "There you can see the total of cells of under `total_cell_count` and the unique number cells under `unique_cell_count` (i.e. after removing cells that were included in multiple datasets).\n",
    "\n",
    "Let's now take a look at the cell and gene information of this slice of the Cell Census.\n",
    "\n",
    "### Learning about cells of lung data\n",
    "\n",
    "Let's load the cell metadata for all lung cells and select only the unique cells using `is_primary_data`.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>soma_joinid</th>\n",
       "      <th>dataset_id</th>\n",
       "      <th>assay</th>\n",
       "      <th>assay_ontology_term_id</th>\n",
       "      <th>cell_type</th>\n",
       "      <th>cell_type_ontology_term_id</th>\n",
       "      <th>development_stage</th>\n",
       "      <th>development_stage_ontology_term_id</th>\n",
       "      <th>disease</th>\n",
       "      <th>disease_ontology_term_id</th>\n",
       "      <th>...</th>\n",
       "      <th>is_primary_data</th>\n",
       "      <th>self_reported_ethnicity</th>\n",
       "      <th>self_reported_ethnicity_ontology_term_id</th>\n",
       "      <th>sex</th>\n",
       "      <th>sex_ontology_term_id</th>\n",
       "      <th>suspension_type</th>\n",
       "      <th>tissue</th>\n",
       "      <th>tissue_ontology_term_id</th>\n",
       "      <th>tissue_general</th>\n",
       "      <th>tissue_general_ontology_term_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>5997601</td>\n",
       "      <td>2adb1f8a-a6b1-4909-8ee8-484814e2d4bf</td>\n",
       "      <td>microwell-seq</td>\n",
       "      <td>EFO:0030002</td>\n",
       "      <td>type II pneumocyte</td>\n",
       "      <td>CL:0002063</td>\n",
       "      <td>21-year-old human stage</td>\n",
       "      <td>HsapDv:0000115</td>\n",
       "      <td>normal</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>Han Chinese</td>\n",
       "      <td>HANCESTRO:0027</td>\n",
       "      <td>male</td>\n",
       "      <td>PATO:0000384</td>\n",
       "      <td>cell</td>\n",
       "      <td>lung</td>\n",
       "      <td>UBERON:0002048</td>\n",
       "      <td>lung</td>\n",
       "      <td>UBERON:0002048</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>5997602</td>\n",
       "      <td>2adb1f8a-a6b1-4909-8ee8-484814e2d4bf</td>\n",
       "      <td>microwell-seq</td>\n",
       "      <td>EFO:0030002</td>\n",
       "      <td>type II pneumocyte</td>\n",
       "      <td>CL:0002063</td>\n",
       "      <td>21-year-old human stage</td>\n",
       "      <td>HsapDv:0000115</td>\n",
       "      <td>normal</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>Han Chinese</td>\n",
       "      <td>HANCESTRO:0027</td>\n",
       "      <td>male</td>\n",
       "      <td>PATO:0000384</td>\n",
       "      <td>cell</td>\n",
       "      <td>lung</td>\n",
       "      <td>UBERON:0002048</td>\n",
       "      <td>lung</td>\n",
       "      <td>UBERON:0002048</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>5997603</td>\n",
       "      <td>2adb1f8a-a6b1-4909-8ee8-484814e2d4bf</td>\n",
       "      <td>microwell-seq</td>\n",
       "      <td>EFO:0030002</td>\n",
       "      <td>type II pneumocyte</td>\n",
       "      <td>CL:0002063</td>\n",
       "      <td>21-year-old human stage</td>\n",
       "      <td>HsapDv:0000115</td>\n",
       "      <td>normal</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>Han Chinese</td>\n",
       "      <td>HANCESTRO:0027</td>\n",
       "      <td>male</td>\n",
       "      <td>PATO:0000384</td>\n",
       "      <td>cell</td>\n",
       "      <td>lung</td>\n",
       "      <td>UBERON:0002048</td>\n",
       "      <td>lung</td>\n",
       "      <td>UBERON:0002048</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>5997604</td>\n",
       "      <td>2adb1f8a-a6b1-4909-8ee8-484814e2d4bf</td>\n",
       "      <td>microwell-seq</td>\n",
       "      <td>EFO:0030002</td>\n",
       "      <td>type II pneumocyte</td>\n",
       "      <td>CL:0002063</td>\n",
       "      <td>21-year-old human stage</td>\n",
       "      <td>HsapDv:0000115</td>\n",
       "      <td>normal</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>Han Chinese</td>\n",
       "      <td>HANCESTRO:0027</td>\n",
       "      <td>male</td>\n",
       "      <td>PATO:0000384</td>\n",
       "      <td>cell</td>\n",
       "      <td>lung</td>\n",
       "      <td>UBERON:0002048</td>\n",
       "      <td>lung</td>\n",
       "      <td>UBERON:0002048</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5997605</td>\n",
       "      <td>2adb1f8a-a6b1-4909-8ee8-484814e2d4bf</td>\n",
       "      <td>microwell-seq</td>\n",
       "      <td>EFO:0030002</td>\n",
       "      <td>type II pneumocyte</td>\n",
       "      <td>CL:0002063</td>\n",
       "      <td>21-year-old human stage</td>\n",
       "      <td>HsapDv:0000115</td>\n",
       "      <td>normal</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>Han Chinese</td>\n",
       "      <td>HANCESTRO:0027</td>\n",
       "      <td>male</td>\n",
       "      <td>PATO:0000384</td>\n",
       "      <td>cell</td>\n",
       "      <td>lung</td>\n",
       "      <td>UBERON:0002048</td>\n",
       "      <td>lung</td>\n",
       "      <td>UBERON:0002048</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1194343</th>\n",
       "      <td>43087668</td>\n",
       "      <td>8c42cfd0-0b0a-46d5-910c-fc833d83c45e</td>\n",
       "      <td>10x 3' v2</td>\n",
       "      <td>EFO:0009899</td>\n",
       "      <td>pericyte</td>\n",
       "      <td>CL:0000669</td>\n",
       "      <td>51-year-old human stage</td>\n",
       "      <td>HsapDv:0000145</td>\n",
       "      <td>normal</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>female</td>\n",
       "      <td>PATO:0000383</td>\n",
       "      <td>cell</td>\n",
       "      <td>lung</td>\n",
       "      <td>UBERON:0002048</td>\n",
       "      <td>lung</td>\n",
       "      <td>UBERON:0002048</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1194344</th>\n",
       "      <td>43087669</td>\n",
       "      <td>8c42cfd0-0b0a-46d5-910c-fc833d83c45e</td>\n",
       "      <td>10x 3' v2</td>\n",
       "      <td>EFO:0009899</td>\n",
       "      <td>pericyte</td>\n",
       "      <td>CL:0000669</td>\n",
       "      <td>51-year-old human stage</td>\n",
       "      <td>HsapDv:0000145</td>\n",
       "      <td>normal</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>female</td>\n",
       "      <td>PATO:0000383</td>\n",
       "      <td>cell</td>\n",
       "      <td>lung</td>\n",
       "      <td>UBERON:0002048</td>\n",
       "      <td>lung</td>\n",
       "      <td>UBERON:0002048</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1194345</th>\n",
       "      <td>43087670</td>\n",
       "      <td>8c42cfd0-0b0a-46d5-910c-fc833d83c45e</td>\n",
       "      <td>10x 3' v2</td>\n",
       "      <td>EFO:0009899</td>\n",
       "      <td>pericyte</td>\n",
       "      <td>CL:0000669</td>\n",
       "      <td>51-year-old human stage</td>\n",
       "      <td>HsapDv:0000145</td>\n",
       "      <td>normal</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>female</td>\n",
       "      <td>PATO:0000383</td>\n",
       "      <td>cell</td>\n",
       "      <td>lung</td>\n",
       "      <td>UBERON:0002048</td>\n",
       "      <td>lung</td>\n",
       "      <td>UBERON:0002048</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1194346</th>\n",
       "      <td>43087671</td>\n",
       "      <td>8c42cfd0-0b0a-46d5-910c-fc833d83c45e</td>\n",
       "      <td>10x 3' v2</td>\n",
       "      <td>EFO:0009899</td>\n",
       "      <td>pericyte</td>\n",
       "      <td>CL:0000669</td>\n",
       "      <td>51-year-old human stage</td>\n",
       "      <td>HsapDv:0000145</td>\n",
       "      <td>normal</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>female</td>\n",
       "      <td>PATO:0000383</td>\n",
       "      <td>cell</td>\n",
       "      <td>lung</td>\n",
       "      <td>UBERON:0002048</td>\n",
       "      <td>lung</td>\n",
       "      <td>UBERON:0002048</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1194347</th>\n",
       "      <td>43087672</td>\n",
       "      <td>8c42cfd0-0b0a-46d5-910c-fc833d83c45e</td>\n",
       "      <td>10x 3' v2</td>\n",
       "      <td>EFO:0009899</td>\n",
       "      <td>pericyte</td>\n",
       "      <td>CL:0000669</td>\n",
       "      <td>51-year-old human stage</td>\n",
       "      <td>HsapDv:0000145</td>\n",
       "      <td>normal</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>female</td>\n",
       "      <td>PATO:0000383</td>\n",
       "      <td>cell</td>\n",
       "      <td>lung</td>\n",
       "      <td>UBERON:0002048</td>\n",
       "      <td>lung</td>\n",
       "      <td>UBERON:0002048</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1194348 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         soma_joinid                            dataset_id          assay  \\\n",
       "0            5997601  2adb1f8a-a6b1-4909-8ee8-484814e2d4bf  microwell-seq   \n",
       "1            5997602  2adb1f8a-a6b1-4909-8ee8-484814e2d4bf  microwell-seq   \n",
       "2            5997603  2adb1f8a-a6b1-4909-8ee8-484814e2d4bf  microwell-seq   \n",
       "3            5997604  2adb1f8a-a6b1-4909-8ee8-484814e2d4bf  microwell-seq   \n",
       "4            5997605  2adb1f8a-a6b1-4909-8ee8-484814e2d4bf  microwell-seq   \n",
       "...              ...                                   ...            ...   \n",
       "1194343     43087668  8c42cfd0-0b0a-46d5-910c-fc833d83c45e      10x 3' v2   \n",
       "1194344     43087669  8c42cfd0-0b0a-46d5-910c-fc833d83c45e      10x 3' v2   \n",
       "1194345     43087670  8c42cfd0-0b0a-46d5-910c-fc833d83c45e      10x 3' v2   \n",
       "1194346     43087671  8c42cfd0-0b0a-46d5-910c-fc833d83c45e      10x 3' v2   \n",
       "1194347     43087672  8c42cfd0-0b0a-46d5-910c-fc833d83c45e      10x 3' v2   \n",
       "\n",
       "        assay_ontology_term_id           cell_type cell_type_ontology_term_id  \\\n",
       "0                  EFO:0030002  type II pneumocyte                 CL:0002063   \n",
       "1                  EFO:0030002  type II pneumocyte                 CL:0002063   \n",
       "2                  EFO:0030002  type II pneumocyte                 CL:0002063   \n",
       "3                  EFO:0030002  type II pneumocyte                 CL:0002063   \n",
       "4                  EFO:0030002  type II pneumocyte                 CL:0002063   \n",
       "...                        ...                 ...                        ...   \n",
       "1194343            EFO:0009899            pericyte                 CL:0000669   \n",
       "1194344            EFO:0009899            pericyte                 CL:0000669   \n",
       "1194345            EFO:0009899            pericyte                 CL:0000669   \n",
       "1194346            EFO:0009899            pericyte                 CL:0000669   \n",
       "1194347            EFO:0009899            pericyte                 CL:0000669   \n",
       "\n",
       "               development_stage development_stage_ontology_term_id disease  \\\n",
       "0        21-year-old human stage                     HsapDv:0000115  normal   \n",
       "1        21-year-old human stage                     HsapDv:0000115  normal   \n",
       "2        21-year-old human stage                     HsapDv:0000115  normal   \n",
       "3        21-year-old human stage                     HsapDv:0000115  normal   \n",
       "4        21-year-old human stage                     HsapDv:0000115  normal   \n",
       "...                          ...                                ...     ...   \n",
       "1194343  51-year-old human stage                     HsapDv:0000145  normal   \n",
       "1194344  51-year-old human stage                     HsapDv:0000145  normal   \n",
       "1194345  51-year-old human stage                     HsapDv:0000145  normal   \n",
       "1194346  51-year-old human stage                     HsapDv:0000145  normal   \n",
       "1194347  51-year-old human stage                     HsapDv:0000145  normal   \n",
       "\n",
       "        disease_ontology_term_id  ... is_primary_data  \\\n",
       "0                   PATO:0000461  ...            True   \n",
       "1                   PATO:0000461  ...            True   \n",
       "2                   PATO:0000461  ...            True   \n",
       "3                   PATO:0000461  ...            True   \n",
       "4                   PATO:0000461  ...            True   \n",
       "...                          ...  ...             ...   \n",
       "1194343             PATO:0000461  ...            True   \n",
       "1194344             PATO:0000461  ...            True   \n",
       "1194345             PATO:0000461  ...            True   \n",
       "1194346             PATO:0000461  ...            True   \n",
       "1194347             PATO:0000461  ...            True   \n",
       "\n",
       "         self_reported_ethnicity self_reported_ethnicity_ontology_term_id  \\\n",
       "0                    Han Chinese                           HANCESTRO:0027   \n",
       "1                    Han Chinese                           HANCESTRO:0027   \n",
       "2                    Han Chinese                           HANCESTRO:0027   \n",
       "3                    Han Chinese                           HANCESTRO:0027   \n",
       "4                    Han Chinese                           HANCESTRO:0027   \n",
       "...                          ...                                      ...   \n",
       "1194343                  unknown                                  unknown   \n",
       "1194344                  unknown                                  unknown   \n",
       "1194345                  unknown                                  unknown   \n",
       "1194346                  unknown                                  unknown   \n",
       "1194347                  unknown                                  unknown   \n",
       "\n",
       "            sex sex_ontology_term_id suspension_type tissue  \\\n",
       "0          male         PATO:0000384            cell   lung   \n",
       "1          male         PATO:0000384            cell   lung   \n",
       "2          male         PATO:0000384            cell   lung   \n",
       "3          male         PATO:0000384            cell   lung   \n",
       "4          male         PATO:0000384            cell   lung   \n",
       "...         ...                  ...             ...    ...   \n",
       "1194343  female         PATO:0000383            cell   lung   \n",
       "1194344  female         PATO:0000383            cell   lung   \n",
       "1194345  female         PATO:0000383            cell   lung   \n",
       "1194346  female         PATO:0000383            cell   lung   \n",
       "1194347  female         PATO:0000383            cell   lung   \n",
       "\n",
       "        tissue_ontology_term_id tissue_general tissue_general_ontology_term_id  \n",
       "0                UBERON:0002048           lung                  UBERON:0002048  \n",
       "1                UBERON:0002048           lung                  UBERON:0002048  \n",
       "2                UBERON:0002048           lung                  UBERON:0002048  \n",
       "3                UBERON:0002048           lung                  UBERON:0002048  \n",
       "4                UBERON:0002048           lung                  UBERON:0002048  \n",
       "...                         ...            ...                             ...  \n",
       "1194343          UBERON:0002048           lung                  UBERON:0002048  \n",
       "1194344          UBERON:0002048           lung                  UBERON:0002048  \n",
       "1194345          UBERON:0002048           lung                  UBERON:0002048  \n",
       "1194346          UBERON:0002048           lung                  UBERON:0002048  \n",
       "1194347          UBERON:0002048           lung                  UBERON:0002048  \n",
       "\n",
       "[1194348 rows x 21 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lung_obs = (\n",
    "    census[\"census_data\"][\"homo_sapiens\"]\n",
    "    .obs.read(value_filter=\"tissue_general == 'lung' and is_primary_data == True\")\n",
    "    .concat()\n",
    "    .to_pandas()\n",
    ")\n",
    "lung_obs\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "You can see that the number or rows represents the total number of unique lung cells in the Cell Census. Now let's take a deeper dive into the characteristics of these cells.\n",
    "\n",
    "#### Datasets\n",
    "\n",
    "First let's start by looking at what are the datasets and collections from [CELLxGENE Discover](https://cellxgene.cziscience.com/collections) contributing to lung. For this we will use the dataset table at `census[\"census-info\"][\"datasets\"]` that contains metadata of all datasets used to build this Cell Census.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>collection_name</th>\n",
       "      <th>dataset_title</th>\n",
       "      <th>soma_joinid</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dataset_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>edc8d3fe-153c-4e3d-8be0-2108d30f8d70</th>\n",
       "      <td>Local and systemic responses to SARS-CoV-2 inf...</td>\n",
       "      <td>Airway</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2a498ace-872a-4935-984b-1afa70fd9886</th>\n",
       "      <td>Local and systemic responses to SARS-CoV-2 inf...</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>f512b8b6-369d-4a85-a695-116e0806857f</th>\n",
       "      <td>Transcriptional Programming of Normal and Infl...</td>\n",
       "      <td>Skin</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>fa8605cf-f27e-44af-ac2a-476bee4410d3</th>\n",
       "      <td>Acute COVID-19 cohort across a range of WHO ca...</td>\n",
       "      <td>PBMCs</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5bc42b88-bb76-4954-927b-8bb7369adc64</th>\n",
       "      <td>Single-cell reconstruction of the early matern...</td>\n",
       "      <td>Pregnant Uterus (All)</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>e04daea4-4412-45b5-989e-76a9be070a89</th>\n",
       "      <td>A molecular cell atlas of the human lung from ...</td>\n",
       "      <td>Krasnow Lab Human Lung Cell Atlas, Smart-seq2</td>\n",
       "      <td>466</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8c42cfd0-0b0a-46d5-910c-fc833d83c45e</th>\n",
       "      <td>A molecular cell atlas of the human lung from ...</td>\n",
       "      <td>Krasnow Lab Human Lung Cell Atlas, 10X</td>\n",
       "      <td>467</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4dd00779-7f73-4f50-89bb-e2d3c6b71b18</th>\n",
       "      <td>Intra- and Inter-cellular Rewiring of the Huma...</td>\n",
       "      <td>Intra- and Inter-cellular Rewiring of the Huma...</td>\n",
       "      <td>468</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8e47ed12-c658-4252-b126-381df8d52a3d</th>\n",
       "      <td>Single-Cell Sequencing of Developing Human Gut...</td>\n",
       "      <td>Paediatric Human Gut (4-14y)</td>\n",
       "      <td>469</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b46237d1-19c6-4af2-9335-9854634bad16</th>\n",
       "      <td>Single-Cell Sequencing of Developing Human Gut...</td>\n",
       "      <td>Fetal Human Gut (6-11 PCW)</td>\n",
       "      <td>470</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>471 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                        collection_name  \\\n",
       "dataset_id                                                                                \n",
       "edc8d3fe-153c-4e3d-8be0-2108d30f8d70  Local and systemic responses to SARS-CoV-2 inf...   \n",
       "2a498ace-872a-4935-984b-1afa70fd9886  Local and systemic responses to SARS-CoV-2 inf...   \n",
       "f512b8b6-369d-4a85-a695-116e0806857f  Transcriptional Programming of Normal and Infl...   \n",
       "fa8605cf-f27e-44af-ac2a-476bee4410d3  Acute COVID-19 cohort across a range of WHO ca...   \n",
       "5bc42b88-bb76-4954-927b-8bb7369adc64  Single-cell reconstruction of the early matern...   \n",
       "...                                                                                 ...   \n",
       "e04daea4-4412-45b5-989e-76a9be070a89  A molecular cell atlas of the human lung from ...   \n",
       "8c42cfd0-0b0a-46d5-910c-fc833d83c45e  A molecular cell atlas of the human lung from ...   \n",
       "4dd00779-7f73-4f50-89bb-e2d3c6b71b18  Intra- and Inter-cellular Rewiring of the Huma...   \n",
       "8e47ed12-c658-4252-b126-381df8d52a3d  Single-Cell Sequencing of Developing Human Gut...   \n",
       "b46237d1-19c6-4af2-9335-9854634bad16  Single-Cell Sequencing of Developing Human Gut...   \n",
       "\n",
       "                                                                          dataset_title  \\\n",
       "dataset_id                                                                                \n",
       "edc8d3fe-153c-4e3d-8be0-2108d30f8d70                                             Airway   \n",
       "2a498ace-872a-4935-984b-1afa70fd9886                                               PBMC   \n",
       "f512b8b6-369d-4a85-a695-116e0806857f                                               Skin   \n",
       "fa8605cf-f27e-44af-ac2a-476bee4410d3                                              PBMCs   \n",
       "5bc42b88-bb76-4954-927b-8bb7369adc64                              Pregnant Uterus (All)   \n",
       "...                                                                                 ...   \n",
       "e04daea4-4412-45b5-989e-76a9be070a89      Krasnow Lab Human Lung Cell Atlas, Smart-seq2   \n",
       "8c42cfd0-0b0a-46d5-910c-fc833d83c45e             Krasnow Lab Human Lung Cell Atlas, 10X   \n",
       "4dd00779-7f73-4f50-89bb-e2d3c6b71b18  Intra- and Inter-cellular Rewiring of the Huma...   \n",
       "8e47ed12-c658-4252-b126-381df8d52a3d                       Paediatric Human Gut (4-14y)   \n",
       "b46237d1-19c6-4af2-9335-9854634bad16                         Fetal Human Gut (6-11 PCW)   \n",
       "\n",
       "                                      soma_joinid  \n",
       "dataset_id                                         \n",
       "edc8d3fe-153c-4e3d-8be0-2108d30f8d70            0  \n",
       "2a498ace-872a-4935-984b-1afa70fd9886            1  \n",
       "f512b8b6-369d-4a85-a695-116e0806857f            2  \n",
       "fa8605cf-f27e-44af-ac2a-476bee4410d3            3  \n",
       "5bc42b88-bb76-4954-927b-8bb7369adc64            4  \n",
       "...                                           ...  \n",
       "e04daea4-4412-45b5-989e-76a9be070a89          466  \n",
       "8c42cfd0-0b0a-46d5-910c-fc833d83c45e          467  \n",
       "4dd00779-7f73-4f50-89bb-e2d3c6b71b18          468  \n",
       "8e47ed12-c658-4252-b126-381df8d52a3d          469  \n",
       "b46237d1-19c6-4af2-9335-9854634bad16          470  \n",
       "\n",
       "[471 rows x 3 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "census_datasets = (\n",
    "    census[\"census_info\"][\"datasets\"]\n",
    "    .read(column_names=[\"collection_name\", \"dataset_title\", \"dataset_id\", \"soma_joinid\"])\n",
    "    .concat()\n",
    "    .to_pandas()\n",
    ")\n",
    "census_datasets = census_datasets.set_index(\"dataset_id\")\n",
    "census_datasets\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The `obs` cell metadata `pandas.DataFrame` contains a column `dataset_id` that can be used for joining to the `census_dataset` `pandas.DataFrame` we just created.\n",
    "\n",
    "So let's take a look at the cell counts per `dataset_id` of the lung slice and then join to the dataset table to append the human-readable labels.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cell_counts</th>\n",
       "      <th>collection_name</th>\n",
       "      <th>dataset_title</th>\n",
       "      <th>soma_joinid</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dataset_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>066943a2-fdac-4b29-b348-40cede398e4e</th>\n",
       "      <td>278224</td>\n",
       "      <td>The integrated Human Lung Cell Atlas</td>\n",
       "      <td>The integrated Human Lung Cell Atlas</td>\n",
       "      <td>450</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>f7c1c579-2dc0-47e2-ba19-8165c5a0e353</th>\n",
       "      <td>217738</td>\n",
       "      <td>A human cell atlas of fetal gene expression</td>\n",
       "      <td>Survey of human embryonic development</td>\n",
       "      <td>268</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1e6a6ef9-7ec9-4c90-bbfb-2ad3c3165fd1</th>\n",
       "      <td>102115</td>\n",
       "      <td>High-resolution single-cell atlas reveals dive...</td>\n",
       "      <td>The single-cell lung cancer atlas (LuCA) -- ex...</td>\n",
       "      <td>271</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>576f193c-75d0-4a11-bd25-8676587e6dc2</th>\n",
       "      <td>90384</td>\n",
       "      <td>HTAN MSK - Single cell profiling reveals novel...</td>\n",
       "      <td>Combined samples</td>\n",
       "      <td>239</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3dc61ca1-ce40-46b6-8337-f27260fd9a03</th>\n",
       "      <td>71752</td>\n",
       "      <td>A human fetal lung cell atlas uncovers proxima...</td>\n",
       "      <td>All cells</td>\n",
       "      <td>257</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8c42cfd0-0b0a-46d5-910c-fc833d83c45e</th>\n",
       "      <td>60993</td>\n",
       "      <td>A molecular cell atlas of the human lung from ...</td>\n",
       "      <td>Krasnow Lab Human Lung Cell Atlas, 10X</td>\n",
       "      <td>467</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2672b679-8048-4f5e-9786-f1b196ccfd08</th>\n",
       "      <td>57019</td>\n",
       "      <td>scRNA-seq assessment of the human lung, spleen...</td>\n",
       "      <td>Lung Parenchyma</td>\n",
       "      <td>184</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9dbab10c-118d-496b-966a-67f1763a6b7d</th>\n",
       "      <td>49014</td>\n",
       "      <td>COVID-19 immune features revealed by a large-s...</td>\n",
       "      <td>Large-scale single-cell analysis reveals criti...</td>\n",
       "      <td>129</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3de0ad6d-4378-4f62-b37b-ec0b75a50d94</th>\n",
       "      <td>46500</td>\n",
       "      <td>LungMAP  Human data from a broad age healthy d...</td>\n",
       "      <td>Single-cell multiomic profiling of human lungs...</td>\n",
       "      <td>288</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2f132ec9-24b5-422f-9be0-ccef03b4fe28</th>\n",
       "      <td>39778</td>\n",
       "      <td>SARS-CoV-2 receptor ACE2 and TMPRSS2 are prima...</td>\n",
       "      <td>Lung</td>\n",
       "      <td>80</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53d208b0-2cfd-4366-9866-c3c6114081bc</th>\n",
       "      <td>35682</td>\n",
       "      <td>Tabula Sapiens</td>\n",
       "      <td>Tabula Sapiens - All Cells</td>\n",
       "      <td>287</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1b9d8702-5af8-4142-85ed-020eb06ec4f6</th>\n",
       "      <td>35419</td>\n",
       "      <td>Cross-tissue immune cell analysis reveals tiss...</td>\n",
       "      <td>Global</td>\n",
       "      <td>310</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4ed927e9-c099-49af-b8ce-a2652d069333</th>\n",
       "      <td>35284</td>\n",
       "      <td>Single-nucleus cross-tissue molecular referenc...</td>\n",
       "      <td>Single-nucleus cross-tissue molecular referenc...</td>\n",
       "      <td>246</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2adb1f8a-a6b1-4909-8ee8-484814e2d4bf</th>\n",
       "      <td>33698</td>\n",
       "      <td>Construction of a human cell landscape at sing...</td>\n",
       "      <td>Construction of a human cell landscape at sing...</td>\n",
       "      <td>46</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>232f6a5a-a04c-4758-a6e8-88ab2e3a6e69</th>\n",
       "      <td>20787</td>\n",
       "      <td>High-resolution single-cell atlas reveals dive...</td>\n",
       "      <td>The single-cell lung cancer atlas (LuCA) -- co...</td>\n",
       "      <td>270</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>01209dce-3575-4bed-b1df-129f57fbc031</th>\n",
       "      <td>11059</td>\n",
       "      <td>Single-cell transcriptomics of human T cells r...</td>\n",
       "      <td>Single-cell transcriptomics of human T cells r...</td>\n",
       "      <td>79</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>e04daea4-4412-45b5-989e-76a9be070a89</th>\n",
       "      <td>8657</td>\n",
       "      <td>A molecular cell atlas of the human lung from ...</td>\n",
       "      <td>Krasnow Lab Human Lung Cell Atlas, Smart-seq2</td>\n",
       "      <td>466</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>f9846bb4-784d-4582-92c1-3f279e4c6f0c</th>\n",
       "      <td>176</td>\n",
       "      <td>A human fetal lung cell atlas uncovers proxima...</td>\n",
       "      <td>Fibroblast and smooth muscle</td>\n",
       "      <td>253</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>f64e1be1-de15-4d27-8da4-82225cd4c035</th>\n",
       "      <td>55</td>\n",
       "      <td>HTAN MSK - Single cell profiling reveals novel...</td>\n",
       "      <td>Immune cells</td>\n",
       "      <td>240</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>810ac45f-8969-4698-b42c-652f802f75c2</th>\n",
       "      <td>10</td>\n",
       "      <td>A human fetal lung cell atlas uncovers proxima...</td>\n",
       "      <td>Endothelium</td>\n",
       "      <td>259</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0ba16f4b-cb87-4fa3-9363-19fc51eec6e7</th>\n",
       "      <td>4</td>\n",
       "      <td>A human fetal lung cell atlas uncovers proxima...</td>\n",
       "      <td>Myeloid</td>\n",
       "      <td>256</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                      cell_counts  \\\n",
       "dataset_id                                          \n",
       "066943a2-fdac-4b29-b348-40cede398e4e       278224   \n",
       "f7c1c579-2dc0-47e2-ba19-8165c5a0e353       217738   \n",
       "1e6a6ef9-7ec9-4c90-bbfb-2ad3c3165fd1       102115   \n",
       "576f193c-75d0-4a11-bd25-8676587e6dc2        90384   \n",
       "3dc61ca1-ce40-46b6-8337-f27260fd9a03        71752   \n",
       "8c42cfd0-0b0a-46d5-910c-fc833d83c45e        60993   \n",
       "2672b679-8048-4f5e-9786-f1b196ccfd08        57019   \n",
       "9dbab10c-118d-496b-966a-67f1763a6b7d        49014   \n",
       "3de0ad6d-4378-4f62-b37b-ec0b75a50d94        46500   \n",
       "2f132ec9-24b5-422f-9be0-ccef03b4fe28        39778   \n",
       "53d208b0-2cfd-4366-9866-c3c6114081bc        35682   \n",
       "1b9d8702-5af8-4142-85ed-020eb06ec4f6        35419   \n",
       "4ed927e9-c099-49af-b8ce-a2652d069333        35284   \n",
       "2adb1f8a-a6b1-4909-8ee8-484814e2d4bf        33698   \n",
       "232f6a5a-a04c-4758-a6e8-88ab2e3a6e69        20787   \n",
       "01209dce-3575-4bed-b1df-129f57fbc031        11059   \n",
       "e04daea4-4412-45b5-989e-76a9be070a89         8657   \n",
       "f9846bb4-784d-4582-92c1-3f279e4c6f0c          176   \n",
       "f64e1be1-de15-4d27-8da4-82225cd4c035           55   \n",
       "810ac45f-8969-4698-b42c-652f802f75c2           10   \n",
       "0ba16f4b-cb87-4fa3-9363-19fc51eec6e7            4   \n",
       "\n",
       "                                                                        collection_name  \\\n",
       "dataset_id                                                                                \n",
       "066943a2-fdac-4b29-b348-40cede398e4e               The integrated Human Lung Cell Atlas   \n",
       "f7c1c579-2dc0-47e2-ba19-8165c5a0e353        A human cell atlas of fetal gene expression   \n",
       "1e6a6ef9-7ec9-4c90-bbfb-2ad3c3165fd1  High-resolution single-cell atlas reveals dive...   \n",
       "576f193c-75d0-4a11-bd25-8676587e6dc2  HTAN MSK - Single cell profiling reveals novel...   \n",
       "3dc61ca1-ce40-46b6-8337-f27260fd9a03  A human fetal lung cell atlas uncovers proxima...   \n",
       "8c42cfd0-0b0a-46d5-910c-fc833d83c45e  A molecular cell atlas of the human lung from ...   \n",
       "2672b679-8048-4f5e-9786-f1b196ccfd08  scRNA-seq assessment of the human lung, spleen...   \n",
       "9dbab10c-118d-496b-966a-67f1763a6b7d  COVID-19 immune features revealed by a large-s...   \n",
       "3de0ad6d-4378-4f62-b37b-ec0b75a50d94  LungMAP  Human data from a broad age healthy d...   \n",
       "2f132ec9-24b5-422f-9be0-ccef03b4fe28  SARS-CoV-2 receptor ACE2 and TMPRSS2 are prima...   \n",
       "53d208b0-2cfd-4366-9866-c3c6114081bc                                     Tabula Sapiens   \n",
       "1b9d8702-5af8-4142-85ed-020eb06ec4f6  Cross-tissue immune cell analysis reveals tiss...   \n",
       "4ed927e9-c099-49af-b8ce-a2652d069333  Single-nucleus cross-tissue molecular referenc...   \n",
       "2adb1f8a-a6b1-4909-8ee8-484814e2d4bf  Construction of a human cell landscape at sing...   \n",
       "232f6a5a-a04c-4758-a6e8-88ab2e3a6e69  High-resolution single-cell atlas reveals dive...   \n",
       "01209dce-3575-4bed-b1df-129f57fbc031  Single-cell transcriptomics of human T cells r...   \n",
       "e04daea4-4412-45b5-989e-76a9be070a89  A molecular cell atlas of the human lung from ...   \n",
       "f9846bb4-784d-4582-92c1-3f279e4c6f0c  A human fetal lung cell atlas uncovers proxima...   \n",
       "f64e1be1-de15-4d27-8da4-82225cd4c035  HTAN MSK - Single cell profiling reveals novel...   \n",
       "810ac45f-8969-4698-b42c-652f802f75c2  A human fetal lung cell atlas uncovers proxima...   \n",
       "0ba16f4b-cb87-4fa3-9363-19fc51eec6e7  A human fetal lung cell atlas uncovers proxima...   \n",
       "\n",
       "                                                                          dataset_title  \\\n",
       "dataset_id                                                                                \n",
       "066943a2-fdac-4b29-b348-40cede398e4e               The integrated Human Lung Cell Atlas   \n",
       "f7c1c579-2dc0-47e2-ba19-8165c5a0e353              Survey of human embryonic development   \n",
       "1e6a6ef9-7ec9-4c90-bbfb-2ad3c3165fd1  The single-cell lung cancer atlas (LuCA) -- ex...   \n",
       "576f193c-75d0-4a11-bd25-8676587e6dc2                                   Combined samples   \n",
       "3dc61ca1-ce40-46b6-8337-f27260fd9a03                                          All cells   \n",
       "8c42cfd0-0b0a-46d5-910c-fc833d83c45e             Krasnow Lab Human Lung Cell Atlas, 10X   \n",
       "2672b679-8048-4f5e-9786-f1b196ccfd08                                    Lung Parenchyma   \n",
       "9dbab10c-118d-496b-966a-67f1763a6b7d  Large-scale single-cell analysis reveals criti...   \n",
       "3de0ad6d-4378-4f62-b37b-ec0b75a50d94  Single-cell multiomic profiling of human lungs...   \n",
       "2f132ec9-24b5-422f-9be0-ccef03b4fe28                                               Lung   \n",
       "53d208b0-2cfd-4366-9866-c3c6114081bc                         Tabula Sapiens - All Cells   \n",
       "1b9d8702-5af8-4142-85ed-020eb06ec4f6                                             Global   \n",
       "4ed927e9-c099-49af-b8ce-a2652d069333  Single-nucleus cross-tissue molecular referenc...   \n",
       "2adb1f8a-a6b1-4909-8ee8-484814e2d4bf  Construction of a human cell landscape at sing...   \n",
       "232f6a5a-a04c-4758-a6e8-88ab2e3a6e69  The single-cell lung cancer atlas (LuCA) -- co...   \n",
       "01209dce-3575-4bed-b1df-129f57fbc031  Single-cell transcriptomics of human T cells r...   \n",
       "e04daea4-4412-45b5-989e-76a9be070a89      Krasnow Lab Human Lung Cell Atlas, Smart-seq2   \n",
       "f9846bb4-784d-4582-92c1-3f279e4c6f0c                       Fibroblast and smooth muscle   \n",
       "f64e1be1-de15-4d27-8da4-82225cd4c035                                       Immune cells   \n",
       "810ac45f-8969-4698-b42c-652f802f75c2                                        Endothelium   \n",
       "0ba16f4b-cb87-4fa3-9363-19fc51eec6e7                                            Myeloid   \n",
       "\n",
       "                                      soma_joinid  \n",
       "dataset_id                                         \n",
       "066943a2-fdac-4b29-b348-40cede398e4e          450  \n",
       "f7c1c579-2dc0-47e2-ba19-8165c5a0e353          268  \n",
       "1e6a6ef9-7ec9-4c90-bbfb-2ad3c3165fd1          271  \n",
       "576f193c-75d0-4a11-bd25-8676587e6dc2          239  \n",
       "3dc61ca1-ce40-46b6-8337-f27260fd9a03          257  \n",
       "8c42cfd0-0b0a-46d5-910c-fc833d83c45e          467  \n",
       "2672b679-8048-4f5e-9786-f1b196ccfd08          184  \n",
       "9dbab10c-118d-496b-966a-67f1763a6b7d          129  \n",
       "3de0ad6d-4378-4f62-b37b-ec0b75a50d94          288  \n",
       "2f132ec9-24b5-422f-9be0-ccef03b4fe28           80  \n",
       "53d208b0-2cfd-4366-9866-c3c6114081bc          287  \n",
       "1b9d8702-5af8-4142-85ed-020eb06ec4f6          310  \n",
       "4ed927e9-c099-49af-b8ce-a2652d069333          246  \n",
       "2adb1f8a-a6b1-4909-8ee8-484814e2d4bf           46  \n",
       "232f6a5a-a04c-4758-a6e8-88ab2e3a6e69          270  \n",
       "01209dce-3575-4bed-b1df-129f57fbc031           79  \n",
       "e04daea4-4412-45b5-989e-76a9be070a89          466  \n",
       "f9846bb4-784d-4582-92c1-3f279e4c6f0c          253  \n",
       "f64e1be1-de15-4d27-8da4-82225cd4c035          240  \n",
       "810ac45f-8969-4698-b42c-652f802f75c2          259  \n",
       "0ba16f4b-cb87-4fa3-9363-19fc51eec6e7          256  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataset_cell_counts = pd.DataFrame(lung_obs[[\"dataset_id\"]].value_counts())\n",
    "dataset_cell_counts = dataset_cell_counts.rename(columns={0: \"cell_counts\"})\n",
    "dataset_cell_counts = dataset_cell_counts.merge(census_datasets, on=\"dataset_id\")\n",
    "\n",
    "dataset_cell_counts\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "These are all the datasets lung cells whose counts are reprensented in the column `cell_counts`. The top collections with lung data are:\n",
    "\n",
    "1. [The integrated Human Lung Cell Atlas](https://cellxgene.cziscience.com/collections/6f6d381a-7701-4781-935c-db10d30de293).\n",
    "2. [A human cell atlas of fetal gene expression](https://cellxgene.cziscience.com/collections/c114c20f-1ef4-49a5-9c2e-d965787fb90c).\n",
    "3. [High-resolution single-cell atlas reveals diversity and plasticity of tumor-associated neutrophils in non-small cell lung cancer](https://cellxgene.cziscience.com/collections/edb893ee-4066-4128-9aec-5eb2b03f8287).\n",
    "4. [HTAN MSK - Single cell profiling reveals novel tumor and myeloid subpopulations in small cell lung cancer](https://cellxgene.cziscience.com/collections/62e8f058-9c37-48bc-9200-e767f318a8ec).\n",
    "5. [A human fetal lung cell atlas uncovers proximal-distal gradients of differentiation and key regulators of epithelial fates.](https://cellxgene.cziscience.com/collections/2d2e2acd-dade-489f-a2da-6c11aa654028).\n",
    "\n",
    "#### Assays\n",
    "\n",
    "Let's use similar logic to take a look at all the assays available for human lung data. This tells us that most assays are from 10x technologies and sci-RNA-seq.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "assay                                   \n",
       "10x 3' v2                                   371991\n",
       "sci-RNA-seq                                 217738\n",
       "10x 5' v1                                   197647\n",
       "10x 3' v3                                   187550\n",
       "BD Rhapsody Whole Transcriptome Analysis    122902\n",
       "10x 5' v2                                    45798\n",
       "microwell-seq                                33698\n",
       "Smart-seq2                                   10247\n",
       "10x 3' transcription profiling                3857\n",
       "10x 3' v1                                     2920\n",
       "dtype: int64"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lung_obs[[\"assay\"]].value_counts()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Disease\n",
    "\n",
    "And now let's take a look at diseased cell counts, with `normal` indicating non-diseased cells.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "disease                     \n",
       "normal                          944311\n",
       "lung adenocarcinoma             132967\n",
       "COVID-19                         49014\n",
       "squamous cell lung carcinoma     36516\n",
       "small cell lung carcinoma        31540\n",
       "dtype: int64"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lung_obs[[\"disease\"]].value_counts()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Sex\n",
    "\n",
    "There doesn't seem to be strong biases for sex.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "sex    \n",
       "male       645694\n",
       "female     548464\n",
       "unknown       190\n",
       "dtype: int64"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lung_obs[[\"sex\"]].value_counts()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Cell vs nucleus\n",
    "\n",
    "The majority of data are from cells and not nucleus.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "suspension_type\n",
       "cell               894826\n",
       "nucleus            299522\n",
       "dtype: int64"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lung_obs[[\"suspension_type\"]].value_counts()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Cell types\n",
    "\n",
    "Let's take a look at the counts of the top 20 cell types.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "cell_type                                 \n",
       "alveolar macrophage                           118184\n",
       "epithelial cell of lower respiratory tract    101249\n",
       "type II pneumocyte                             99444\n",
       "stromal cell                                   87301\n",
       "macrophage                                     56132\n",
       "CD4-positive, alpha-beta T cell                43949\n",
       "natural killer cell                            43290\n",
       "T cell                                         40789\n",
       "CD8-positive, alpha-beta T cell                40310\n",
       "fibroblast of lung                             36967\n",
       "epithelial cell                                35636\n",
       "classical monocyte                             25267\n",
       "elicited macrophage                            23525\n",
       "endothelial cell                               22528\n",
       "epithelial cell of alveolus of lung            22305\n",
       "fibroblast                                     20537\n",
       "type I pneumocyte                              19291\n",
       "neutrophil                                     16971\n",
       "monocyte                                       16952\n",
       "endothelial cell of vascular tree              15725\n",
       "dtype: int64"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lung_obs[[\"cell_type\"]].value_counts().head(20)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Sub-tissues\n",
    "\n",
    "We can look at the original tissue annotations that were mapped to \"lung\".\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "tissue                 \n",
       "lung                       812762\n",
       "lung parenchyma            335243\n",
       "lingula of left lung        35284\n",
       "lower lobe of left lung     11059\n",
       "dtype: int64"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lung_obs[[\"tissue\"]].value_counts()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Learning about genes of lung data\n",
    "\n",
    "Let's load the gene metadata of the Cell Census.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>soma_joinid</th>\n",
       "      <th>feature_id</th>\n",
       "      <th>feature_name</th>\n",
       "      <th>feature_length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000121410</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>3999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000268895</td>\n",
       "      <td>A1BG-AS1</td>\n",
       "      <td>3374</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000148584</td>\n",
       "      <td>A1CF</td>\n",
       "      <td>9603</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>ENSG00000175899</td>\n",
       "      <td>A2M</td>\n",
       "      <td>6318</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>ENSG00000245105</td>\n",
       "      <td>A2M-AS1</td>\n",
       "      <td>2948</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60659</th>\n",
       "      <td>60659</td>\n",
       "      <td>ENSG00000288699</td>\n",
       "      <td>RP11-182N22.10</td>\n",
       "      <td>654</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60660</th>\n",
       "      <td>60660</td>\n",
       "      <td>ENSG00000288700</td>\n",
       "      <td>RP11-22E12.2</td>\n",
       "      <td>6888</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60661</th>\n",
       "      <td>60661</td>\n",
       "      <td>ENSG00000288710</td>\n",
       "      <td>RP11-386G11.12</td>\n",
       "      <td>2968</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60662</th>\n",
       "      <td>60662</td>\n",
       "      <td>ENSG00000288711</td>\n",
       "      <td>AP000326.5</td>\n",
       "      <td>1307</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60663</th>\n",
       "      <td>60663</td>\n",
       "      <td>ENSG00000288721</td>\n",
       "      <td>RP5-973N23.5</td>\n",
       "      <td>7765</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>60664 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       soma_joinid       feature_id    feature_name  feature_length\n",
       "0                0  ENSG00000121410            A1BG            3999\n",
       "1                1  ENSG00000268895        A1BG-AS1            3374\n",
       "2                2  ENSG00000148584            A1CF            9603\n",
       "3                3  ENSG00000175899             A2M            6318\n",
       "4                4  ENSG00000245105         A2M-AS1            2948\n",
       "...            ...              ...             ...             ...\n",
       "60659        60659  ENSG00000288699  RP11-182N22.10             654\n",
       "60660        60660  ENSG00000288700    RP11-22E12.2            6888\n",
       "60661        60661  ENSG00000288710  RP11-386G11.12            2968\n",
       "60662        60662  ENSG00000288711      AP000326.5            1307\n",
       "60663        60663  ENSG00000288721    RP5-973N23.5            7765\n",
       "\n",
       "[60664 rows x 4 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lung_var = census[\"census_data\"][\"homo_sapiens\"].ms[\"RNA\"].var.read().concat().to_pandas()\n",
    "lung_var\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "You can see the total number of genes represented by the number of rows. This number is actually misleading becausue it is the join of all genes in the Cell Census. However we know that the lung data comes from a subset of datasets.\n",
    "\n",
    "So let's take a look at the number of genes that were measured in each of those datasets.\n",
    "\n",
    "To accomplish this we can use the \"dataset presence matrix\" at `census[\"census_data\"][\"homo_sapiens\"].ms[\"RNA\"][\"feature_dataset_presence_matrix\"]`. This is a boolean matrix `N x M` where `N` is the number of datasets and `M` is the number of genes in the Cell Census.\n",
    "\n",
    "So we can select the rows corresponding to the lung datasets and perform a row-wise sum.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "presence_matrix = cell_census.get_presence_matrix(census, \"Homo sapiens\", \"RNA\")\n",
    "presence_matrix = presence_matrix[dataset_cell_counts.soma_joinid, :]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([28024, 44150, 17811, 22447, 26347, 20921, 24672, 27705, 26323,\n",
       "       27181, 57030, 32610, 29620, 26454, 17811, 38676, 47307, 23740,\n",
       "       22552, 20594, 19952], dtype=uint64)"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "presence_matrix.sum(axis=1).A1\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cell_counts</th>\n",
       "      <th>collection_name</th>\n",
       "      <th>dataset_title</th>\n",
       "      <th>soma_joinid</th>\n",
       "      <th>genes_measured</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dataset_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>066943a2-fdac-4b29-b348-40cede398e4e</th>\n",
       "      <td>278224</td>\n",
       "      <td>The integrated Human Lung Cell Atlas</td>\n",
       "      <td>The integrated Human Lung Cell Atlas</td>\n",
       "      <td>450</td>\n",
       "      <td>28024</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>f7c1c579-2dc0-47e2-ba19-8165c5a0e353</th>\n",
       "      <td>217738</td>\n",
       "      <td>A human cell atlas of fetal gene expression</td>\n",
       "      <td>Survey of human embryonic development</td>\n",
       "      <td>268</td>\n",
       "      <td>44150</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1e6a6ef9-7ec9-4c90-bbfb-2ad3c3165fd1</th>\n",
       "      <td>102115</td>\n",
       "      <td>High-resolution single-cell atlas reveals dive...</td>\n",
       "      <td>The single-cell lung cancer atlas (LuCA) -- ex...</td>\n",
       "      <td>271</td>\n",
       "      <td>17811</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>576f193c-75d0-4a11-bd25-8676587e6dc2</th>\n",
       "      <td>90384</td>\n",
       "      <td>HTAN MSK - Single cell profiling reveals novel...</td>\n",
       "      <td>Combined samples</td>\n",
       "      <td>239</td>\n",
       "      <td>22447</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3dc61ca1-ce40-46b6-8337-f27260fd9a03</th>\n",
       "      <td>71752</td>\n",
       "      <td>A human fetal lung cell atlas uncovers proxima...</td>\n",
       "      <td>All cells</td>\n",
       "      <td>257</td>\n",
       "      <td>26347</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8c42cfd0-0b0a-46d5-910c-fc833d83c45e</th>\n",
       "      <td>60993</td>\n",
       "      <td>A molecular cell atlas of the human lung from ...</td>\n",
       "      <td>Krasnow Lab Human Lung Cell Atlas, 10X</td>\n",
       "      <td>467</td>\n",
       "      <td>20921</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2672b679-8048-4f5e-9786-f1b196ccfd08</th>\n",
       "      <td>57019</td>\n",
       "      <td>scRNA-seq assessment of the human lung, spleen...</td>\n",
       "      <td>Lung Parenchyma</td>\n",
       "      <td>184</td>\n",
       "      <td>24672</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9dbab10c-118d-496b-966a-67f1763a6b7d</th>\n",
       "      <td>49014</td>\n",
       "      <td>COVID-19 immune features revealed by a large-s...</td>\n",
       "      <td>Large-scale single-cell analysis reveals criti...</td>\n",
       "      <td>129</td>\n",
       "      <td>27705</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3de0ad6d-4378-4f62-b37b-ec0b75a50d94</th>\n",
       "      <td>46500</td>\n",
       "      <td>LungMAP  Human data from a broad age healthy d...</td>\n",
       "      <td>Single-cell multiomic profiling of human lungs...</td>\n",
       "      <td>288</td>\n",
       "      <td>26323</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2f132ec9-24b5-422f-9be0-ccef03b4fe28</th>\n",
       "      <td>39778</td>\n",
       "      <td>SARS-CoV-2 receptor ACE2 and TMPRSS2 are prima...</td>\n",
       "      <td>Lung</td>\n",
       "      <td>80</td>\n",
       "      <td>27181</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53d208b0-2cfd-4366-9866-c3c6114081bc</th>\n",
       "      <td>35682</td>\n",
       "      <td>Tabula Sapiens</td>\n",
       "      <td>Tabula Sapiens - All Cells</td>\n",
       "      <td>287</td>\n",
       "      <td>57030</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1b9d8702-5af8-4142-85ed-020eb06ec4f6</th>\n",
       "      <td>35419</td>\n",
       "      <td>Cross-tissue immune cell analysis reveals tiss...</td>\n",
       "      <td>Global</td>\n",
       "      <td>310</td>\n",
       "      <td>32610</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4ed927e9-c099-49af-b8ce-a2652d069333</th>\n",
       "      <td>35284</td>\n",
       "      <td>Single-nucleus cross-tissue molecular referenc...</td>\n",
       "      <td>Single-nucleus cross-tissue molecular referenc...</td>\n",
       "      <td>246</td>\n",
       "      <td>29620</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2adb1f8a-a6b1-4909-8ee8-484814e2d4bf</th>\n",
       "      <td>33698</td>\n",
       "      <td>Construction of a human cell landscape at sing...</td>\n",
       "      <td>Construction of a human cell landscape at sing...</td>\n",
       "      <td>46</td>\n",
       "      <td>26454</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>232f6a5a-a04c-4758-a6e8-88ab2e3a6e69</th>\n",
       "      <td>20787</td>\n",
       "      <td>High-resolution single-cell atlas reveals dive...</td>\n",
       "      <td>The single-cell lung cancer atlas (LuCA) -- co...</td>\n",
       "      <td>270</td>\n",
       "      <td>17811</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>01209dce-3575-4bed-b1df-129f57fbc031</th>\n",
       "      <td>11059</td>\n",
       "      <td>Single-cell transcriptomics of human T cells r...</td>\n",
       "      <td>Single-cell transcriptomics of human T cells r...</td>\n",
       "      <td>79</td>\n",
       "      <td>38676</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>e04daea4-4412-45b5-989e-76a9be070a89</th>\n",
       "      <td>8657</td>\n",
       "      <td>A molecular cell atlas of the human lung from ...</td>\n",
       "      <td>Krasnow Lab Human Lung Cell Atlas, Smart-seq2</td>\n",
       "      <td>466</td>\n",
       "      <td>47307</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>f9846bb4-784d-4582-92c1-3f279e4c6f0c</th>\n",
       "      <td>176</td>\n",
       "      <td>A human fetal lung cell atlas uncovers proxima...</td>\n",
       "      <td>Fibroblast and smooth muscle</td>\n",
       "      <td>253</td>\n",
       "      <td>23740</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>f64e1be1-de15-4d27-8da4-82225cd4c035</th>\n",
       "      <td>55</td>\n",
       "      <td>HTAN MSK - Single cell profiling reveals novel...</td>\n",
       "      <td>Immune cells</td>\n",
       "      <td>240</td>\n",
       "      <td>22552</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>810ac45f-8969-4698-b42c-652f802f75c2</th>\n",
       "      <td>10</td>\n",
       "      <td>A human fetal lung cell atlas uncovers proxima...</td>\n",
       "      <td>Endothelium</td>\n",
       "      <td>259</td>\n",
       "      <td>20594</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0ba16f4b-cb87-4fa3-9363-19fc51eec6e7</th>\n",
       "      <td>4</td>\n",
       "      <td>A human fetal lung cell atlas uncovers proxima...</td>\n",
       "      <td>Myeloid</td>\n",
       "      <td>256</td>\n",
       "      <td>19952</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                      cell_counts  \\\n",
       "dataset_id                                          \n",
       "066943a2-fdac-4b29-b348-40cede398e4e       278224   \n",
       "f7c1c579-2dc0-47e2-ba19-8165c5a0e353       217738   \n",
       "1e6a6ef9-7ec9-4c90-bbfb-2ad3c3165fd1       102115   \n",
       "576f193c-75d0-4a11-bd25-8676587e6dc2        90384   \n",
       "3dc61ca1-ce40-46b6-8337-f27260fd9a03        71752   \n",
       "8c42cfd0-0b0a-46d5-910c-fc833d83c45e        60993   \n",
       "2672b679-8048-4f5e-9786-f1b196ccfd08        57019   \n",
       "9dbab10c-118d-496b-966a-67f1763a6b7d        49014   \n",
       "3de0ad6d-4378-4f62-b37b-ec0b75a50d94        46500   \n",
       "2f132ec9-24b5-422f-9be0-ccef03b4fe28        39778   \n",
       "53d208b0-2cfd-4366-9866-c3c6114081bc        35682   \n",
       "1b9d8702-5af8-4142-85ed-020eb06ec4f6        35419   \n",
       "4ed927e9-c099-49af-b8ce-a2652d069333        35284   \n",
       "2adb1f8a-a6b1-4909-8ee8-484814e2d4bf        33698   \n",
       "232f6a5a-a04c-4758-a6e8-88ab2e3a6e69        20787   \n",
       "01209dce-3575-4bed-b1df-129f57fbc031        11059   \n",
       "e04daea4-4412-45b5-989e-76a9be070a89         8657   \n",
       "f9846bb4-784d-4582-92c1-3f279e4c6f0c          176   \n",
       "f64e1be1-de15-4d27-8da4-82225cd4c035           55   \n",
       "810ac45f-8969-4698-b42c-652f802f75c2           10   \n",
       "0ba16f4b-cb87-4fa3-9363-19fc51eec6e7            4   \n",
       "\n",
       "                                                                        collection_name  \\\n",
       "dataset_id                                                                                \n",
       "066943a2-fdac-4b29-b348-40cede398e4e               The integrated Human Lung Cell Atlas   \n",
       "f7c1c579-2dc0-47e2-ba19-8165c5a0e353        A human cell atlas of fetal gene expression   \n",
       "1e6a6ef9-7ec9-4c90-bbfb-2ad3c3165fd1  High-resolution single-cell atlas reveals dive...   \n",
       "576f193c-75d0-4a11-bd25-8676587e6dc2  HTAN MSK - Single cell profiling reveals novel...   \n",
       "3dc61ca1-ce40-46b6-8337-f27260fd9a03  A human fetal lung cell atlas uncovers proxima...   \n",
       "8c42cfd0-0b0a-46d5-910c-fc833d83c45e  A molecular cell atlas of the human lung from ...   \n",
       "2672b679-8048-4f5e-9786-f1b196ccfd08  scRNA-seq assessment of the human lung, spleen...   \n",
       "9dbab10c-118d-496b-966a-67f1763a6b7d  COVID-19 immune features revealed by a large-s...   \n",
       "3de0ad6d-4378-4f62-b37b-ec0b75a50d94  LungMAP  Human data from a broad age healthy d...   \n",
       "2f132ec9-24b5-422f-9be0-ccef03b4fe28  SARS-CoV-2 receptor ACE2 and TMPRSS2 are prima...   \n",
       "53d208b0-2cfd-4366-9866-c3c6114081bc                                     Tabula Sapiens   \n",
       "1b9d8702-5af8-4142-85ed-020eb06ec4f6  Cross-tissue immune cell analysis reveals tiss...   \n",
       "4ed927e9-c099-49af-b8ce-a2652d069333  Single-nucleus cross-tissue molecular referenc...   \n",
       "2adb1f8a-a6b1-4909-8ee8-484814e2d4bf  Construction of a human cell landscape at sing...   \n",
       "232f6a5a-a04c-4758-a6e8-88ab2e3a6e69  High-resolution single-cell atlas reveals dive...   \n",
       "01209dce-3575-4bed-b1df-129f57fbc031  Single-cell transcriptomics of human T cells r...   \n",
       "e04daea4-4412-45b5-989e-76a9be070a89  A molecular cell atlas of the human lung from ...   \n",
       "f9846bb4-784d-4582-92c1-3f279e4c6f0c  A human fetal lung cell atlas uncovers proxima...   \n",
       "f64e1be1-de15-4d27-8da4-82225cd4c035  HTAN MSK - Single cell profiling reveals novel...   \n",
       "810ac45f-8969-4698-b42c-652f802f75c2  A human fetal lung cell atlas uncovers proxima...   \n",
       "0ba16f4b-cb87-4fa3-9363-19fc51eec6e7  A human fetal lung cell atlas uncovers proxima...   \n",
       "\n",
       "                                                                          dataset_title  \\\n",
       "dataset_id                                                                                \n",
       "066943a2-fdac-4b29-b348-40cede398e4e               The integrated Human Lung Cell Atlas   \n",
       "f7c1c579-2dc0-47e2-ba19-8165c5a0e353              Survey of human embryonic development   \n",
       "1e6a6ef9-7ec9-4c90-bbfb-2ad3c3165fd1  The single-cell lung cancer atlas (LuCA) -- ex...   \n",
       "576f193c-75d0-4a11-bd25-8676587e6dc2                                   Combined samples   \n",
       "3dc61ca1-ce40-46b6-8337-f27260fd9a03                                          All cells   \n",
       "8c42cfd0-0b0a-46d5-910c-fc833d83c45e             Krasnow Lab Human Lung Cell Atlas, 10X   \n",
       "2672b679-8048-4f5e-9786-f1b196ccfd08                                    Lung Parenchyma   \n",
       "9dbab10c-118d-496b-966a-67f1763a6b7d  Large-scale single-cell analysis reveals criti...   \n",
       "3de0ad6d-4378-4f62-b37b-ec0b75a50d94  Single-cell multiomic profiling of human lungs...   \n",
       "2f132ec9-24b5-422f-9be0-ccef03b4fe28                                               Lung   \n",
       "53d208b0-2cfd-4366-9866-c3c6114081bc                         Tabula Sapiens - All Cells   \n",
       "1b9d8702-5af8-4142-85ed-020eb06ec4f6                                             Global   \n",
       "4ed927e9-c099-49af-b8ce-a2652d069333  Single-nucleus cross-tissue molecular referenc...   \n",
       "2adb1f8a-a6b1-4909-8ee8-484814e2d4bf  Construction of a human cell landscape at sing...   \n",
       "232f6a5a-a04c-4758-a6e8-88ab2e3a6e69  The single-cell lung cancer atlas (LuCA) -- co...   \n",
       "01209dce-3575-4bed-b1df-129f57fbc031  Single-cell transcriptomics of human T cells r...   \n",
       "e04daea4-4412-45b5-989e-76a9be070a89      Krasnow Lab Human Lung Cell Atlas, Smart-seq2   \n",
       "f9846bb4-784d-4582-92c1-3f279e4c6f0c                       Fibroblast and smooth muscle   \n",
       "f64e1be1-de15-4d27-8da4-82225cd4c035                                       Immune cells   \n",
       "810ac45f-8969-4698-b42c-652f802f75c2                                        Endothelium   \n",
       "0ba16f4b-cb87-4fa3-9363-19fc51eec6e7                                            Myeloid   \n",
       "\n",
       "                                      soma_joinid  genes_measured  \n",
       "dataset_id                                                         \n",
       "066943a2-fdac-4b29-b348-40cede398e4e          450           28024  \n",
       "f7c1c579-2dc0-47e2-ba19-8165c5a0e353          268           44150  \n",
       "1e6a6ef9-7ec9-4c90-bbfb-2ad3c3165fd1          271           17811  \n",
       "576f193c-75d0-4a11-bd25-8676587e6dc2          239           22447  \n",
       "3dc61ca1-ce40-46b6-8337-f27260fd9a03          257           26347  \n",
       "8c42cfd0-0b0a-46d5-910c-fc833d83c45e          467           20921  \n",
       "2672b679-8048-4f5e-9786-f1b196ccfd08          184           24672  \n",
       "9dbab10c-118d-496b-966a-67f1763a6b7d          129           27705  \n",
       "3de0ad6d-4378-4f62-b37b-ec0b75a50d94          288           26323  \n",
       "2f132ec9-24b5-422f-9be0-ccef03b4fe28           80           27181  \n",
       "53d208b0-2cfd-4366-9866-c3c6114081bc          287           57030  \n",
       "1b9d8702-5af8-4142-85ed-020eb06ec4f6          310           32610  \n",
       "4ed927e9-c099-49af-b8ce-a2652d069333          246           29620  \n",
       "2adb1f8a-a6b1-4909-8ee8-484814e2d4bf           46           26454  \n",
       "232f6a5a-a04c-4758-a6e8-88ab2e3a6e69          270           17811  \n",
       "01209dce-3575-4bed-b1df-129f57fbc031           79           38676  \n",
       "e04daea4-4412-45b5-989e-76a9be070a89          466           47307  \n",
       "f9846bb4-784d-4582-92c1-3f279e4c6f0c          253           23740  \n",
       "f64e1be1-de15-4d27-8da4-82225cd4c035          240           22552  \n",
       "810ac45f-8969-4698-b42c-652f802f75c2          259           20594  \n",
       "0ba16f4b-cb87-4fa3-9363-19fc51eec6e7          256           19952  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes_measured = presence_matrix.sum(axis=1).A1\n",
    "dataset_cell_counts[\"genes_measured\"] = genes_measured\n",
    "dataset_cell_counts\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "You can see the genes measured in each dataset represented in `genes_measured`. Now lets get the **genes that were measured in all datasets**.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "var_somaid = np.nonzero(presence_matrix.sum(axis=0).A1 == presence_matrix.shape[0])[0].tolist()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>soma_joinid</th>\n",
       "      <th>feature_id</th>\n",
       "      <th>feature_name</th>\n",
       "      <th>feature_length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000121410</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>3999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>ENSG00000175899</td>\n",
       "      <td>A2M</td>\n",
       "      <td>6318</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>9</td>\n",
       "      <td>ENSG00000128274</td>\n",
       "      <td>A4GALT</td>\n",
       "      <td>3358</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>11</td>\n",
       "      <td>ENSG00000094914</td>\n",
       "      <td>AAAS</td>\n",
       "      <td>4727</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>12</td>\n",
       "      <td>ENSG00000081760</td>\n",
       "      <td>AACS</td>\n",
       "      <td>16039</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33226</th>\n",
       "      <td>33226</td>\n",
       "      <td>ENSG00000198205</td>\n",
       "      <td>ZXDA</td>\n",
       "      <td>5029</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33227</th>\n",
       "      <td>33227</td>\n",
       "      <td>ENSG00000198455</td>\n",
       "      <td>ZXDB</td>\n",
       "      <td>5467</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33228</th>\n",
       "      <td>33228</td>\n",
       "      <td>ENSG00000070476</td>\n",
       "      <td>ZXDC</td>\n",
       "      <td>7329</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33230</th>\n",
       "      <td>33230</td>\n",
       "      <td>ENSG00000162378</td>\n",
       "      <td>ZYG11B</td>\n",
       "      <td>8206</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33232</th>\n",
       "      <td>33232</td>\n",
       "      <td>ENSG00000074755</td>\n",
       "      <td>ZZEF1</td>\n",
       "      <td>15888</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>11948 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       soma_joinid       feature_id feature_name  feature_length\n",
       "0                0  ENSG00000121410         A1BG            3999\n",
       "3                3  ENSG00000175899          A2M            6318\n",
       "9                9  ENSG00000128274       A4GALT            3358\n",
       "11              11  ENSG00000094914         AAAS            4727\n",
       "12              12  ENSG00000081760         AACS           16039\n",
       "...            ...              ...          ...             ...\n",
       "33226        33226  ENSG00000198205         ZXDA            5029\n",
       "33227        33227  ENSG00000198455         ZXDB            5467\n",
       "33228        33228  ENSG00000070476         ZXDC            7329\n",
       "33230        33230  ENSG00000162378       ZYG11B            8206\n",
       "33232        33232  ENSG00000074755        ZZEF1           15888\n",
       "\n",
       "[11948 rows x 4 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lung_var = lung_var.query(f\"soma_joinid in {var_somaid}\")\n",
    "lung_var\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The number of rows represents the genes that were measured in all lung datasets.\n",
    "\n",
    "### Summary of lung metadata\n",
    "\n",
    "In the previous sections, using the Cell Census we learned the following information:\n",
    "\n",
    "- There are **1,194,348** unique lung cells and those are composed of:\n",
    "  - **21** datasets.\n",
    "  - **10** sequencing technologies, most of which are 10x\n",
    "  - Mostly human data, but **4** diseases exist, primarily \"lung adenocarcinoma\" and \"COVID-19 infected\"\n",
    "  - No sex biases.\n",
    "  - Mostly data from cells (\\~80%) rather than nucleus (\\~20%)\n",
    "- A total of **11,948** genes were measured across all cells.\n",
    "\n",
    "## Fetching all human lung data from the Cell Census.\n",
    "\n",
    "Now let's try to load all data into an `anndata.AnnData` object and select only the unique cells using `is_primary_data`.\n",
    "\n",
    "We will also only select lung cells and the genes measured in all lung datasets that we obtained in the sections above.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "lung_genes = list(lung_var[\"feature_id\"])\n",
    "lung_adata = cell_census.get_anndata(\n",
    "    census,\n",
    "    organism=\"Homo sapiens\",\n",
    "    obs_value_filter=\"tissue_general=='lung' and is_primary_data==True\",\n",
    "    var_value_filter=f\"feature_id in {lung_genes}\",\n",
    ")\n",
    "\n",
    "lung_adata.var_names = lung_adata.var[\"feature_name\"]\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### QC metrics on gene expression of all Lung data\n",
    "\n",
    "Now let's take a look at some QC metrics\n",
    "\n",
    "**Top genes per cell**\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.highest_expr_genes(lung_adata, n_top=20)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Number of sequenced genes by assay**\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pp.calculate_qc_metrics(lung_adata, percent_top=None, log1p=False, inplace=True)\n",
    "sc.pl.violin(lung_adata, \"n_genes_by_counts\", groupby=\"assay\", jitter=0.4, rotation=90)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Total counts by assay**\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.violin(lung_adata, \"total_counts\", groupby=\"assay\", jitter=0.4, rotation=90)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "You can see that Smart-Seq2 is an outlier for the total counts per cell, so let's exlcude it to see how the rest of the assays look like\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "sc.pl.violin(\n",
    "    lung_adata[\n",
    "        lung_adata.obs[\"assay\"] != \"Smart-seq2\",\n",
    "    ],\n",
    "    \"total_counts\",\n",
    "    groupby=\"assay\",\n",
    "    jitter=0.4,\n",
    "    rotation=90,\n",
    ")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Quick normalization and embeddings\n",
    "\n",
    "Let's perform a bread and butter normalization and take a look at UMAP embeddings, but for all the data below we'll exclude Smart-seq2 as this requires an extra step to normalize based on gene lengths\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "lung_adata = lung_adata[\n",
    "    lung_adata.obs[\"assay\"] != \"Smart-seq2\",\n",
    "].copy()\n",
    "lung_adata.layers[\"counts\"] = lung_adata.X\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now let's do some basic normalization:\n",
    "\n",
    "- Normalize by sequencing depth\n",
    "- Transform to log-scale\n",
    "- Select 500 highly variable genes\n",
    "- Scale values across the gene axis\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pp.normalize_total(lung_adata, target_sum=1e4)\n",
    "sc.pp.log1p(lung_adata)\n",
    "sc.pp.highly_variable_genes(lung_adata, n_top_genes=500, flavor=\"seurat_v3\", layer=\"counts\")\n",
    "lung_adata = lung_adata[:, lung_adata.var.highly_variable]\n",
    "sc.pp.scale(lung_adata, max_value=10)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "And reduce dimensionality by obtaining UMAP embeddings.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.tl.pca(lung_adata)\n",
    "sc.pp.neighbors(lung_adata)\n",
    "sc.tl.umap(lung_adata)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "And plot these embeddings.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "n_cell_types = len(lung_adata.obs[\"cell_type\"].drop_duplicates())\n",
    "\n",
    "from random import randint\n",
    "\n",
    "colors = []\n",
    "\n",
    "for i in range(len(lung_adata.obs[\"cell_type\"].drop_duplicates())):\n",
    "    colors.append(\"#%06X\" % randint(0, 0xFFFFFF))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(lung_adata, color=\"cell_type\", palette=colors)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(lung_adata, color=\"cell_type\", palette=colors, legend_loc=None)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's color by assay.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(lung_adata, color=\"assay\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Given the high number of cell types it makes it hard to visualize, so let's look at the top 20 most abundant cell types.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "top_cell_types = lung_adata.obs[\"cell_type\"].value_counts()\n",
    "top_cell_types = list(top_cell_types.reset_index().head(20)[\"index\"])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "lung_adata_top_cell_types = lung_adata[[i in top_cell_types for i in lung_adata.obs[\"cell_type\"]], :]\n",
    "sc.pl.umap(lung_adata_top_cell_types, color=\"cell_type\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's color by assay of this subset of the data.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(lung_adata_top_cell_types, color=\"assay\")\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  },
  "vscode": {
   "interpreter": {
    "hash": "3da8ec1c162cd849e59e6ea2824b2e353dce799884e910aae99411be5277f953"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
