

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8">
  <meta name="generator" content="Docutils 0.19: https://docutils.sourceforge.io/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Python API &mdash; cellxgene-census  documentation</title>
  

  
  <link rel="icon" type="image/png" sizes="32x32" href="_static/img/favicon_32x32_v2.png"/>
  <link rel="icon" type="image/png" sizes="16x16" href="_static/img/favicon_16x16_v2.png"/>
  <link rel="preconnect" href="https://fonts.googleapis.com">
  <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
  <link href="https://fonts.googleapis.com/css2?family=Inter:wght@400;500;600;700&display=swap" rel="stylesheet">
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/css/custom.css" type="text/css" />

  
  
  
  

  
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/sphinx_highlight.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script data-domain="chanzuckerberg.github.io/cellxgene-census" defer="defer" src="https://plausible.io/js/script.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="cellxgene_census.open_soma" href="_autosummary/cellxgene_census.open_soma.html" />
    <link rel="prev" title="Census data releases" href="cellxgene_census_docsite_data_release_info.html" /> 
</head>

<body class="wy-body-for-nav">

  <div style="height: 100vh; overflow: hidden;">
     
    <!-- top 56px for navbar height -->
    <div class="navbar-cxg">
        <div style="padding: 0 16px; display: flex; height: inherit; align-items: center; justify-content: space-between;">
            <span style="display:flex; gap: 32px; align-items: center;">
                <a href="https://cellxgene.cziscience.com/">
                  <img src="_static/img/cellxgene-discover-logo.svg" />
                </a>
                <span class="navbar-cxg-nav-wrapper">
                  <span class="navbar-cxg-section">
                    <span class="navbar-cxg-nav-section-title">Application</span>
                    <span class="navbar-cxg-nav-item-container">
                      <span class="navbar-cxg-link">
                          <a class="navbar-cxg-anchor" href="https://cellxgene.cziscience.com/collections">Collections</a>
                      </span>
                      <span class="navbar-cxg-link">
                          <a class="navbar-cxg-anchor" href="https://cellxgene.cziscience.com/datasets">Datasets</a>
                      </span>
                      <span class="navbar-cxg-link">
                          <a class="navbar-cxg-anchor" href="https://cellxgene.cziscience.com/gene-expression">Gene Expression</a>
                      </span>
                      <span class="navbar-cxg-link">
                        <a class="navbar-cxg-anchor" href="https://cellxgene.cziscience.com/cellguide">Cell Guide</a>
                        <div style="height: 16px!important; display: flex;">
                          <span class="beta">BETA</span>
                        </div>
                      </span>
                    </span>
                  </span>
                  <hr class="navbar-divider"/>
                  <span class="navbar-cxg-section">
                    <span class="navbar-cxg-nav-section-title">Census</span>
                    <span class="navbar-cxg-nav-item-container">
                      <span class="navbar-cxg-link active-link">
                        <a class="navbar-cxg-anchor" href="/cellxgene-census/index.html">API</a>
                      </span>
                      <span class="navbar-cxg-link">
                        <a class="navbar-cxg-anchor" href="https://cellxgene.cziscience.com/census-models">Models</a>
                      </span>
                    </span>
                  </span>
                </span>
            </span>

            <span class="navbar-cxg-link">
              <a class="navbar-cxg-anchor" href="https://cellxgene.cziscience.com/docs">Help & Documentation</a>
            </span>
        </div>
    </div>
    <div style="width: 100%; height: 100%; overflow: auto; padding-top: 56px;">
        
        <!-- top 56px for navbar height -->
        <nav data-toggle="wy-nav-shift" class="wy-nav-side" style="top: 56px;" >
        <div class="wy-side-scroll">
            <div class="wy-side-nav-search" >
            

            
                <a href="index.html" class="icon icon-home"> cellxgene-census
            

            
            </a>

            
                
                
                <div class="version">
                    1.15.0
                </div>
                
            

            
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

            
            </div>

            
            <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
            
                
                
                
                
                
                <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="cellxgene_census_docsite_installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="cellxgene_census_docsite_quick_start.html">Quick start</a></li>
<li class="toctree-l1"><a class="reference internal" href="articles.html">What’s new?</a></li>
<li class="toctree-l1"><a class="reference internal" href="cellxgene_census_aws_open_data.html">Census in AWS ☁️</a></li>
<li class="toctree-l1"><a class="reference internal" href="cellxgene_census_docsite_schema.html">Census data and schema</a></li>
<li class="toctree-l1"><a class="reference internal" href="cellxgene_census_docsite_data_release_info.html">Census data releases</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Python API</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#open-retrieve-cell-census-data">Open/retrieve Cell Census data</a><ul>
<li class="toctree-l3"><a class="reference internal" href="_autosummary/cellxgene_census.open_soma.html">cellxgene_census.open_soma</a></li>
<li class="toctree-l3"><a class="reference internal" href="_autosummary/cellxgene_census.get_default_soma_context.html">cellxgene_census.get_default_soma_context</a></li>
<li class="toctree-l3"><a class="reference internal" href="_autosummary/cellxgene_census.get_source_h5ad_uri.html">cellxgene_census.get_source_h5ad_uri</a></li>
<li class="toctree-l3"><a class="reference internal" href="_autosummary/cellxgene_census.download_source_h5ad.html">cellxgene_census.download_source_h5ad</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#get-slice-as-anndata">Get slice as AnnData</a><ul>
<li class="toctree-l3"><a class="reference internal" href="_autosummary/cellxgene_census.get_anndata.html">cellxgene_census.get_anndata</a></li>
<li class="toctree-l3"><a class="reference internal" href="_autosummary/cellxgene_census.get_obs.html">cellxgene_census.get_obs</a></li>
<li class="toctree-l3"><a class="reference internal" href="_autosummary/cellxgene_census.get_var.html">cellxgene_census.get_var</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#feature-presence-matrix">Feature presence matrix</a><ul>
<li class="toctree-l3"><a class="reference internal" href="_autosummary/cellxgene_census.get_presence_matrix.html">cellxgene_census.get_presence_matrix</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#versioning-of-cell-census-builds">Versioning of Cell Census builds</a><ul>
<li class="toctree-l3"><a class="reference internal" href="_autosummary/cellxgene_census.get_census_version_description.html">cellxgene_census.get_census_version_description</a></li>
<li class="toctree-l3"><a class="reference internal" href="_autosummary/cellxgene_census.get_census_version_directory.html">cellxgene_census.get_census_version_directory</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#experimental-machine-learning">Experimental: Machine Learning</a><ul>
<li class="toctree-l3"><a class="reference internal" href="_autosummary/cellxgene_census.experimental.ml.pytorch.experiment_dataloader.html">cellxgene_census.experimental.ml.pytorch.experiment_dataloader</a></li>
<li class="toctree-l3"><a class="reference internal" href="_autosummary/cellxgene_census.experimental.ml.pytorch.ExperimentDataPipe.html">cellxgene_census.experimental.ml.pytorch.ExperimentDataPipe</a></li>
<li class="toctree-l3"><a class="reference internal" href="_autosummary/cellxgene_census.experimental.ml.pytorch.Stats.html">cellxgene_census.experimental.ml.pytorch.Stats</a></li>
<li class="toctree-l3"><a class="reference internal" href="_autosummary/cellxgene_census.experimental.ml.encoders.Encoder.html">cellxgene_census.experimental.ml.encoders.Encoder</a></li>
<li class="toctree-l3"><a class="reference internal" href="_autosummary/cellxgene_census.experimental.ml.encoders.LabelEncoder.html">cellxgene_census.experimental.ml.encoders.LabelEncoder</a></li>
<li class="toctree-l3"><a class="reference internal" href="_autosummary/cellxgene_census.experimental.ml.encoders.BatchEncoder.html">cellxgene_census.experimental.ml.encoders.BatchEncoder</a></li>
<li class="toctree-l3"><a class="reference internal" href="_autosummary/cellxgene_census.experimental.ml.huggingface.CellDatasetBuilder.html">cellxgene_census.experimental.ml.huggingface.CellDatasetBuilder</a></li>
<li class="toctree-l3"><a class="reference internal" href="_autosummary/cellxgene_census.experimental.ml.huggingface.GeneformerTokenizer.html">cellxgene_census.experimental.ml.huggingface.GeneformerTokenizer</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#experimental-processing">Experimental: Processing</a><ul>
<li class="toctree-l3"><a class="reference internal" href="_autosummary/cellxgene_census.experimental.pp.get_highly_variable_genes.html">cellxgene_census.experimental.pp.get_highly_variable_genes</a></li>
<li class="toctree-l3"><a class="reference internal" href="_autosummary/cellxgene_census.experimental.pp.highly_variable_genes.html">cellxgene_census.experimental.pp.highly_variable_genes</a></li>
<li class="toctree-l3"><a class="reference internal" href="_autosummary/cellxgene_census.experimental.pp.mean_variance.html">cellxgene_census.experimental.pp.mean_variance</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#experimental-embeddings">Experimental: Embeddings</a><ul>
<li class="toctree-l3"><a class="reference internal" href="_autosummary/cellxgene_census.experimental.get_embedding.html">cellxgene_census.experimental.get_embedding</a></li>
<li class="toctree-l3"><a class="reference internal" href="_autosummary/cellxgene_census.experimental.get_embedding_metadata.html">cellxgene_census.experimental.get_embedding_metadata</a></li>
<li class="toctree-l3"><a class="reference internal" href="_autosummary/cellxgene_census.experimental.get_embedding_metadata_by_name.html">cellxgene_census.experimental.get_embedding_metadata_by_name</a></li>
<li class="toctree-l3"><a class="reference internal" href="_autosummary/cellxgene_census.experimental.get_all_available_embeddings.html">cellxgene_census.experimental.get_all_available_embeddings</a></li>
<li class="toctree-l3"><a class="reference internal" href="_autosummary/cellxgene_census.experimental.get_all_census_versions_with_embedding.html">cellxgene_census.experimental.get_all_census_versions_with_embedding</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="examples.html">Python tutorials</a></li>
<li class="toctree-l1"><a class="reference external" href="https://chanzuckerberg.github.io/cellxgene-census/r/index.html">R API &amp; tutorials</a></li>
<li class="toctree-l1"><a class="reference internal" href="cellxgene_census_docsite_FAQ.html">FAQ</a></li>
<li class="toctree-l1"><a class="reference external" href="https://github.com/chanzuckerberg/cellxgene-census/releases">Release notes</a></li>
</ul>

                
            
            </div>
            
        </div>
        </nav>

        <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

        
        <nav class="wy-nav-top" aria-label="top navigation">
            
            <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
            <a href="index.html">cellxgene-census</a>
            
        </nav>


        <div class="wy-nav-content">
            
            <div class="rst-content">
            
            <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">Python API</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/python-api.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
            <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
            <div itemprop="articleBody">
                
  <section id="module-cellxgene_census">
<span id="python-api"></span><h1>Python API<a class="headerlink" href="#module-cellxgene_census" title="Permalink to this heading"></a></h1>
<p>An API to facilitate use of the CZI Science CELLxGENE Census. The Census is a versioned container of single-cell data hosted at <a class="reference external" href="https://cellxgene.cziscience.com/">CELLxGENE Discover</a>.</p>
<p>The API is built on the <cite>tiledbsoma</cite> SOMA API, and provides a number of helper functions including:</p>
<blockquote>
<div><ul class="simple">
<li><p>Open a named version of the Census, for use with the SOMA API</p></li>
<li><p>Get a list of available Census versions, and for each version, a description</p></li>
<li><p>Get a slice of the Census as an AnnData, for use with ScanPy</p></li>
<li><p>Get the URI for, or directly download, underlying data in H5AD format</p></li>
</ul>
</div></blockquote>
<p>For more information on the API, visit the <a class="reference external" href="https://github.com/chanzuckerberg/cellxgene-census/">cellxgene_census repo</a>. For more information on SOMA, see the <a class="reference external" href="https://github.com/single-cell-data/TileDB-SOMA">tiledbsoma repo</a>.</p>
<section id="open-retrieve-cell-census-data">
<h2>Open/retrieve Cell Census data<a class="headerlink" href="#open-retrieve-cell-census-data" title="Permalink to this heading"></a></h2>
<table class="autosummary longtable docutils align-default">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="_autosummary/cellxgene_census.open_soma.html#cellxgene_census.open_soma" title="cellxgene_census.open_soma"><code class="xref py py-obj docutils literal notranslate"><span class="pre">cellxgene_census.open_soma</span></code></a></p></td>
<td><p>Open the Census by version or URI.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="_autosummary/cellxgene_census.get_default_soma_context.html#cellxgene_census.get_default_soma_context" title="cellxgene_census.get_default_soma_context"><code class="xref py py-obj docutils literal notranslate"><span class="pre">cellxgene_census.get_default_soma_context</span></code></a></p></td>
<td><p>Return a <a class="reference external" href="https://tiledbsoma.readthedocs.io/en/latest/_autosummary/tiledbsoma.SOMATileDBContext.html#tiledbsoma.SOMATileDBContext" title="(in TileDB-SOMA-Py)"><code class="xref py py-class docutils literal notranslate"><span class="pre">tiledbsoma.SOMATileDBContext</span></code></a> with sensible defaults that can be further customized by the user.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="_autosummary/cellxgene_census.get_source_h5ad_uri.html#cellxgene_census.get_source_h5ad_uri" title="cellxgene_census.get_source_h5ad_uri"><code class="xref py py-obj docutils literal notranslate"><span class="pre">cellxgene_census.get_source_h5ad_uri</span></code></a></p></td>
<td><p>Open the named version of the census, and return the URI for the <code class="docutils literal notranslate"><span class="pre">dataset_id</span></code>.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="_autosummary/cellxgene_census.download_source_h5ad.html#cellxgene_census.download_source_h5ad" title="cellxgene_census.download_source_h5ad"><code class="xref py py-obj docutils literal notranslate"><span class="pre">cellxgene_census.download_source_h5ad</span></code></a></p></td>
<td><p>Download the source H5AD dataset, for the given <cite>dataset_id</cite>, to the user-specified file name.</p></td>
</tr>
</tbody>
</table>
</section>
<section id="get-slice-as-anndata">
<h2>Get slice as AnnData<a class="headerlink" href="#get-slice-as-anndata" title="Permalink to this heading"></a></h2>
<table class="autosummary longtable docutils align-default">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="_autosummary/cellxgene_census.get_anndata.html#cellxgene_census.get_anndata" title="cellxgene_census.get_anndata"><code class="xref py py-obj docutils literal notranslate"><span class="pre">cellxgene_census.get_anndata</span></code></a></p></td>
<td><p>Convenience wrapper around <a class="reference external" href="https://tiledbsoma.readthedocs.io/en/latest/_autosummary/tiledbsoma.Experiment.html#tiledbsoma.Experiment" title="(in TileDB-SOMA-Py)"><code class="xref py py-class docutils literal notranslate"><span class="pre">tiledbsoma.Experiment</span></code></a> query, to build and execute a query, and return it as an <a class="reference external" href="https://anndata.readthedocs.io/en/latest/generated/anndata.AnnData.html#anndata.AnnData" title="(in anndata v0.1.dev50+g9918044)"><code class="xref py py-class docutils literal notranslate"><span class="pre">anndata.AnnData</span></code></a> object.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="_autosummary/cellxgene_census.get_obs.html#cellxgene_census.get_obs" title="cellxgene_census.get_obs"><code class="xref py py-obj docutils literal notranslate"><span class="pre">cellxgene_census.get_obs</span></code></a></p></td>
<td><p>Get the observation metadata for a query on the census.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="_autosummary/cellxgene_census.get_var.html#cellxgene_census.get_var" title="cellxgene_census.get_var"><code class="xref py py-obj docutils literal notranslate"><span class="pre">cellxgene_census.get_var</span></code></a></p></td>
<td><p>Get the variable metadata for a query on the census.</p></td>
</tr>
</tbody>
</table>
</section>
<section id="feature-presence-matrix">
<h2>Feature presence matrix<a class="headerlink" href="#feature-presence-matrix" title="Permalink to this heading"></a></h2>
<table class="autosummary longtable docutils align-default">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="_autosummary/cellxgene_census.get_presence_matrix.html#cellxgene_census.get_presence_matrix" title="cellxgene_census.get_presence_matrix"><code class="xref py py-obj docutils literal notranslate"><span class="pre">cellxgene_census.get_presence_matrix</span></code></a></p></td>
<td><p>Read the feature dataset presence matrix and return as a <a class="reference external" href="https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.csr_array.html#scipy.sparse.csr_array" title="(in SciPy v1.14.0)"><code class="xref py py-class docutils literal notranslate"><span class="pre">scipy.sparse.csr_array</span></code></a>.</p></td>
</tr>
</tbody>
</table>
</section>
<section id="versioning-of-cell-census-builds">
<h2>Versioning of Cell Census builds<a class="headerlink" href="#versioning-of-cell-census-builds" title="Permalink to this heading"></a></h2>
<table class="autosummary longtable docutils align-default">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="_autosummary/cellxgene_census.get_census_version_description.html#cellxgene_census.get_census_version_description" title="cellxgene_census.get_census_version_description"><code class="xref py py-obj docutils literal notranslate"><span class="pre">cellxgene_census.get_census_version_description</span></code></a></p></td>
<td><p>Get release description for given Census version, from the Census release directory.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="_autosummary/cellxgene_census.get_census_version_directory.html#cellxgene_census.get_census_version_directory" title="cellxgene_census.get_census_version_directory"><code class="xref py py-obj docutils literal notranslate"><span class="pre">cellxgene_census.get_census_version_directory</span></code></a></p></td>
<td><p>Get the directory of Census versions currently available, optionally filtering by specified flags.</p></td>
</tr>
</tbody>
</table>
</section>
<section id="experimental-machine-learning">
<h2>Experimental: Machine Learning<a class="headerlink" href="#experimental-machine-learning" title="Permalink to this heading"></a></h2>
<table class="autosummary longtable docutils align-default">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="_autosummary/cellxgene_census.experimental.ml.pytorch.experiment_dataloader.html#cellxgene_census.experimental.ml.pytorch.experiment_dataloader" title="cellxgene_census.experimental.ml.pytorch.experiment_dataloader"><code class="xref py py-obj docutils literal notranslate"><span class="pre">cellxgene_census.experimental.ml.pytorch.experiment_dataloader</span></code></a></p></td>
<td><p>Factory method for <a class="reference external" href="https://pytorch.org/docs/stable/data.html#torch.utils.data.DataLoader" title="(in PyTorch v2.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">torch.utils.data.DataLoader</span></code></a>.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="_autosummary/cellxgene_census.experimental.ml.pytorch.ExperimentDataPipe.html#cellxgene_census.experimental.ml.pytorch.ExperimentDataPipe" title="cellxgene_census.experimental.ml.pytorch.ExperimentDataPipe"><code class="xref py py-obj docutils literal notranslate"><span class="pre">cellxgene_census.experimental.ml.pytorch.ExperimentDataPipe</span></code></a></p></td>
<td><p>An <a class="reference external" href="https://pytorch.org/data/beta/torchdata.datapipes.iter.html#torchdata.datapipes.iter.IterDataPipe" title="(in TorchData v0.7.0)"><code class="xref py py-class docutils literal notranslate"><span class="pre">torchdata.datapipes.iter.IterDataPipe</span></code></a> that reads <code class="docutils literal notranslate"><span class="pre">obs</span></code> and <code class="docutils literal notranslate"><span class="pre">X</span></code> data from a <a class="reference external" href="https://tiledbsoma.readthedocs.io/en/latest/_autosummary/tiledbsoma.Experiment.html#tiledbsoma.Experiment" title="(in TileDB-SOMA-Py)"><code class="xref py py-class docutils literal notranslate"><span class="pre">tiledbsoma.Experiment</span></code></a>, based upon the specified queries along the <code class="docutils literal notranslate"><span class="pre">obs</span></code> and <code class="docutils literal notranslate"><span class="pre">var</span></code> axes.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="_autosummary/cellxgene_census.experimental.ml.pytorch.Stats.html#cellxgene_census.experimental.ml.pytorch.Stats" title="cellxgene_census.experimental.ml.pytorch.Stats"><code class="xref py py-obj docutils literal notranslate"><span class="pre">cellxgene_census.experimental.ml.pytorch.Stats</span></code></a></p></td>
<td><p>Statistics about the data retrieved by <code class="docutils literal notranslate"><span class="pre">ExperimentDataPipe</span></code> via SOMA API.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="_autosummary/cellxgene_census.experimental.ml.encoders.Encoder.html#cellxgene_census.experimental.ml.encoders.Encoder" title="cellxgene_census.experimental.ml.encoders.Encoder"><code class="xref py py-obj docutils literal notranslate"><span class="pre">cellxgene_census.experimental.ml.encoders.Encoder</span></code></a></p></td>
<td><p>Base class for <code class="docutils literal notranslate"><span class="pre">obs</span></code> encoders.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="_autosummary/cellxgene_census.experimental.ml.encoders.LabelEncoder.html#cellxgene_census.experimental.ml.encoders.LabelEncoder" title="cellxgene_census.experimental.ml.encoders.LabelEncoder"><code class="xref py py-obj docutils literal notranslate"><span class="pre">cellxgene_census.experimental.ml.encoders.LabelEncoder</span></code></a></p></td>
<td><p>Default encoder based on <a class="reference external" href="https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.LabelEncoder.html#sklearn.preprocessing.LabelEncoder" title="(in scikit-learn v1.5)"><code class="xref py py-class docutils literal notranslate"><span class="pre">sklearn.preprocessing.LabelEncoder</span></code></a>.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="_autosummary/cellxgene_census.experimental.ml.encoders.BatchEncoder.html#cellxgene_census.experimental.ml.encoders.BatchEncoder" title="cellxgene_census.experimental.ml.encoders.BatchEncoder"><code class="xref py py-obj docutils literal notranslate"><span class="pre">cellxgene_census.experimental.ml.encoders.BatchEncoder</span></code></a></p></td>
<td><p>An encoder that concatenates and encodes several <code class="docutils literal notranslate"><span class="pre">obs</span></code> columns.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="_autosummary/cellxgene_census.experimental.ml.huggingface.CellDatasetBuilder.html#cellxgene_census.experimental.ml.huggingface.CellDatasetBuilder" title="cellxgene_census.experimental.ml.huggingface.CellDatasetBuilder"><code class="xref py py-obj docutils literal notranslate"><span class="pre">cellxgene_census.experimental.ml.huggingface.CellDatasetBuilder</span></code></a></p></td>
<td><p>Abstract base class for methods to process CELLxGENE Census ExperimentAxisQuery results into a Hugging Face Dataset in which each item represents one cell.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="_autosummary/cellxgene_census.experimental.ml.huggingface.GeneformerTokenizer.html#cellxgene_census.experimental.ml.huggingface.GeneformerTokenizer" title="cellxgene_census.experimental.ml.huggingface.GeneformerTokenizer"><code class="xref py py-obj docutils literal notranslate"><span class="pre">cellxgene_census.experimental.ml.huggingface.GeneformerTokenizer</span></code></a></p></td>
<td><p>Generate a Hugging Face <cite>Dataset</cite> containing Geneformer token sequences for each cell in CELLxGENE Census ExperimentAxisQuery results (human).</p></td>
</tr>
</tbody>
</table>
</section>
<section id="experimental-processing">
<h2>Experimental: Processing<a class="headerlink" href="#experimental-processing" title="Permalink to this heading"></a></h2>
<table class="autosummary longtable docutils align-default">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="_autosummary/cellxgene_census.experimental.pp.get_highly_variable_genes.html#cellxgene_census.experimental.pp.get_highly_variable_genes" title="cellxgene_census.experimental.pp.get_highly_variable_genes"><code class="xref py py-obj docutils literal notranslate"><span class="pre">cellxgene_census.experimental.pp.get_highly_variable_genes</span></code></a></p></td>
<td><p>Convience wrapper around <a class="reference external" href="https://tiledbsoma.readthedocs.io/en/latest/_autosummary/tiledbsoma.Experiment.html#tiledbsoma.Experiment" title="(in TileDB-SOMA-Py)"><code class="xref py py-class docutils literal notranslate"><span class="pre">tiledbsoma.Experiment</span></code></a> query and <a class="reference internal" href="_autosummary/cellxgene_census.experimental.pp.highly_variable_genes.html#cellxgene_census.experimental.pp.highly_variable_genes" title="cellxgene_census.experimental.pp.highly_variable_genes"><code class="xref py py-func docutils literal notranslate"><span class="pre">cellxgene_census.experimental.pp.highly_variable_genes()</span></code></a> function, to build and execute a query, and annotate the query result genes (<code class="docutils literal notranslate"><span class="pre">var</span></code> dataframe) based upon variability.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="_autosummary/cellxgene_census.experimental.pp.highly_variable_genes.html#cellxgene_census.experimental.pp.highly_variable_genes" title="cellxgene_census.experimental.pp.highly_variable_genes"><code class="xref py py-obj docutils literal notranslate"><span class="pre">cellxgene_census.experimental.pp.highly_variable_genes</span></code></a></p></td>
<td><p>Identify and annotate highly variable genes contained in the query results.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="_autosummary/cellxgene_census.experimental.pp.mean_variance.html#cellxgene_census.experimental.pp.mean_variance" title="cellxgene_census.experimental.pp.mean_variance"><code class="xref py py-obj docutils literal notranslate"><span class="pre">cellxgene_census.experimental.pp.mean_variance</span></code></a></p></td>
<td><p>Calculate  mean and/or variance along the <code class="docutils literal notranslate"><span class="pre">obs</span></code> axis from query results.</p></td>
</tr>
</tbody>
</table>
</section>
<section id="experimental-embeddings">
<h2>Experimental: Embeddings<a class="headerlink" href="#experimental-embeddings" title="Permalink to this heading"></a></h2>
<table class="autosummary longtable docutils align-default">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="_autosummary/cellxgene_census.experimental.get_embedding.html#cellxgene_census.experimental.get_embedding" title="cellxgene_census.experimental.get_embedding"><code class="xref py py-obj docutils literal notranslate"><span class="pre">cellxgene_census.experimental.get_embedding</span></code></a></p></td>
<td><p>Read cell (obs) embeddings and return as a dense <a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v2.0)"><code class="xref py py-class docutils literal notranslate"><span class="pre">numpy.ndarray</span></code></a>.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="_autosummary/cellxgene_census.experimental.get_embedding_metadata.html#cellxgene_census.experimental.get_embedding_metadata" title="cellxgene_census.experimental.get_embedding_metadata"><code class="xref py py-obj docutils literal notranslate"><span class="pre">cellxgene_census.experimental.get_embedding_metadata</span></code></a></p></td>
<td><p>Read embedding metadata and return as a Python dict.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="_autosummary/cellxgene_census.experimental.get_embedding_metadata_by_name.html#cellxgene_census.experimental.get_embedding_metadata_by_name" title="cellxgene_census.experimental.get_embedding_metadata_by_name"><code class="xref py py-obj docutils literal notranslate"><span class="pre">cellxgene_census.experimental.get_embedding_metadata_by_name</span></code></a></p></td>
<td><p>Return metadata for a specific embedding.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="_autosummary/cellxgene_census.experimental.get_all_available_embeddings.html#cellxgene_census.experimental.get_all_available_embeddings" title="cellxgene_census.experimental.get_all_available_embeddings"><code class="xref py py-obj docutils literal notranslate"><span class="pre">cellxgene_census.experimental.get_all_available_embeddings</span></code></a></p></td>
<td><p>Return a dictionary of all available embeddings for a given Census version.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="_autosummary/cellxgene_census.experimental.get_all_census_versions_with_embedding.html#cellxgene_census.experimental.get_all_census_versions_with_embedding" title="cellxgene_census.experimental.get_all_census_versions_with_embedding"><code class="xref py py-obj docutils literal notranslate"><span class="pre">cellxgene_census.experimental.get_all_census_versions_with_embedding</span></code></a></p></td>
<td><p>Get a list of all census versions that contain a specific embedding.</p></td>
</tr>
</tbody>
</table>
</section>
</section>


            </div>
            
            </div>
            <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="cellxgene_census_docsite_data_release_info.html" class="btn btn-neutral float-left" title="Census data releases" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="_autosummary/cellxgene_census.open_soma.html" class="btn btn-neutral float-right" title="cellxgene_census.open_soma" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2022, Chan Zuckerberg Initiative.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
            </div>
        </div>

        </section>

    </div>

  

  
  
    
   

</body>
</html>