<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="cellxgene.census">
<title>Genes measured in each cell (dataset presence matrix) • cellxgene.census</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Genes measured in each cell (dataset presence matrix)">
<meta property="og:description" content="cellxgene.census">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><script defer data-domain="chanzuckerberg.github.io/cellxgene-census" src="https://plausible.io/js/script.js"></script>
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">cellxgene.census</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.0.9000</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/census_dataset_presence.html">Genes measured in each cell (dataset presence matrix)</a>
    <a class="dropdown-item" href="../articles/census_datasets.html">Census Datasets example</a>
    <a class="dropdown-item" href="../articles/census_query_extract.html">Querying and fetching the single-cell data and cell/gene metadata</a>
    <a class="dropdown-item" href="../articles/comp_bio_census_info.html">Learning about the CZ CELLxGENE Census</a>
    <a class="dropdown-item" href="../articles/comp_bio_summarize_axis_query.html">Summarizing cell and gene metadata</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/chanzuckerberg/cellxgene-census/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Genes measured in each cell (dataset presence matrix)</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/chanzuckerberg/cellxgene-census/blob/HEAD/vignettes/census_dataset_presence.Rmd" class="external-link"><code>vignettes/census_dataset_presence.Rmd</code></a></small>
      <div class="d-none name"><code>census_dataset_presence.Rmd</code></div>
    </div>

    
    
<!--
THIS VIGNETTE IS BASED ON:
https://github.com/chanzuckerberg/cellxgene-census/blob/main/api/python/notebooks/api_demo/census_dataset_presence.ipynb
-->
<p>The Census is a compilation of cells from multiple datasets that may
differ by the sets of genes they measure. This notebook describes the
way to identify the genes measured per dataset.</p>
<p>The presence matrix is a sparse boolean array, indicating which
features (var) were present in each dataset. The array has dimensions
[n_datasets, n_var], and is stored in the SOMA Measurement
<code>varp</code> collection. The first dimension is indexed by the
<code>soma_joinid</code> in the <code>census_datasets</code> dataframe.
The second is indexed by the <code>soma_joinid</code> in the
<code>var</code> dataframe of the measurement.</p>
<p>As a reminder the <code>obs</code> data frame has a column
<code>dataset_id</code> that can be used to link any cell in the Census
to the presence matrix.</p>
<p>Contents</p>
<ol style="list-style-type: decimal">
<li>Opening the Census.</li>
<li>Fetching the IDs of the Census datasets.</li>
<li>Fetching the dataset presence matrix.</li>
<li>Identifying genes measured in a specific dataset.</li>
<li>Identifying datasets that measured specific genes.</li>
<li>Identifying all genes measured in a dataset.</li>
</ol>
<div class="section level2">
<h2 id="opening-the-census">Opening the Census<a class="anchor" aria-label="anchor" href="#opening-the-census"></a>
</h2>
<p>The <code>cellxgene.census</code> R package contains a convenient API
to open any version of the Census (by default, the newest stable
version).</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">census</span> <span class="op">&lt;-</span> <span class="fu">cellxgene.census</span><span class="fu">::</span><span class="fu"><a href="../reference/open_soma.html">open_soma</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; The stable Census release is currently 2023-05-15. Specify census_version = "2023-05-15" in future calls to open_soma() to ensure data consistency.</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="fetching-the-ids-of-the-census-datasets">Fetching the IDs of the Census datasets<a class="anchor" aria-label="anchor" href="#fetching-the-ids-of-the-census-datasets"></a>
</h2>
<p>Let’s grab a table of all the datasets included in the Census and use
this table in combination with the presence matrix below.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Grab the experiment containing human data, and the measurement therein with RNA</span></span>
<span><span class="va">human</span> <span class="op">&lt;-</span> <span class="va">census</span><span class="op">$</span><span class="fu">get</span><span class="op">(</span><span class="st">"census_data"</span><span class="op">)</span><span class="op">$</span><span class="fu">get</span><span class="op">(</span><span class="st">"homo_sapiens"</span><span class="op">)</span></span>
<span><span class="va">human_rna</span> <span class="op">&lt;-</span> <span class="va">human</span><span class="op">$</span><span class="va">ms</span><span class="op">$</span><span class="fu">get</span><span class="op">(</span><span class="st">"RNA"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># The census-wide datasets</span></span>
<span><span class="va">datasets_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="va">census</span><span class="op">$</span><span class="fu">get</span><span class="op">(</span><span class="st">"census_info"</span><span class="op">)</span><span class="op">$</span><span class="fu">get</span><span class="op">(</span><span class="st">"datasets"</span><span class="op">)</span><span class="op">$</span><span class="fu">read</span><span class="op">(</span><span class="op">)</span><span class="op">$</span><span class="fu">concat</span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">datasets_df</span><span class="op">)</span></span>
<span><span class="co">#&gt;    soma_joinid                        collection_id</span></span>
<span><span class="co">#&gt; 1            0 6b701826-37bb-4356-9792-ff41fc4c3161</span></span>
<span><span class="co">#&gt; 2            1 4195ab4c-20bd-4cd3-8b3d-65601277e731</span></span>
<span><span class="co">#&gt; 3            2 4195ab4c-20bd-4cd3-8b3d-65601277e731</span></span>
<span><span class="co">#&gt; 4            3 4195ab4c-20bd-4cd3-8b3d-65601277e731</span></span>
<span><span class="co">#&gt; 5            4 4195ab4c-20bd-4cd3-8b3d-65601277e731</span></span>
<span><span class="co">#&gt; 6            5 74e10dc4-cbb2-4605-a189-8a1cd8e44d8c</span></span>
<span><span class="co">#&gt; 7            6 74e10dc4-cbb2-4605-a189-8a1cd8e44d8c</span></span>
<span><span class="co">#&gt; 8            7 74e10dc4-cbb2-4605-a189-8a1cd8e44d8c</span></span>
<span><span class="co">#&gt; 9            8 74e10dc4-cbb2-4605-a189-8a1cd8e44d8c</span></span>
<span><span class="co">#&gt; 10           9 6d203948-a779-4b69-9b3f-1ee1dadc3980</span></span>
<span><span class="co">#&gt; 11          10 6d203948-a779-4b69-9b3f-1ee1dadc3980</span></span>
<span><span class="co">#&gt; 12          11 6d203948-a779-4b69-9b3f-1ee1dadc3980</span></span>
<span><span class="co">#&gt;                                                                                        collection_name</span></span>
<span><span class="co">#&gt; 1                                                                       Abdominal White Adipose Tissue</span></span>
<span><span class="co">#&gt; 2                             A spatially resolved single cell genomic atlas of the adult human breast</span></span>
<span><span class="co">#&gt; 3                             A spatially resolved single cell genomic atlas of the adult human breast</span></span>
<span><span class="co">#&gt; 4                             A spatially resolved single cell genomic atlas of the adult human breast</span></span>
<span><span class="co">#&gt; 5                             A spatially resolved single cell genomic atlas of the adult human breast</span></span>
<span><span class="co">#&gt; 6       Spatial proteogenomics reveals distinct and evolutionarily conserved hepatic macrophage niches</span></span>
<span><span class="co">#&gt; 7       Spatial proteogenomics reveals distinct and evolutionarily conserved hepatic macrophage niches</span></span>
<span><span class="co">#&gt; 8       Spatial proteogenomics reveals distinct and evolutionarily conserved hepatic macrophage niches</span></span>
<span><span class="co">#&gt; 9       Spatial proteogenomics reveals distinct and evolutionarily conserved hepatic macrophage niches</span></span>
<span><span class="co">#&gt; 10 Differential cell composition and split epidermal differentiation in human palm, sole, and hip skin</span></span>
<span><span class="co">#&gt; 11 Differential cell composition and split epidermal differentiation in human palm, sole, and hip skin</span></span>
<span><span class="co">#&gt; 12 Differential cell composition and split epidermal differentiation in human palm, sole, and hip skin</span></span>
<span><span class="co">#&gt;                  collection_doi                           dataset_id</span></span>
<span><span class="co">#&gt; 1                               9d8e5dca-03a3-457d-b7fb-844c75735c83</span></span>
<span><span class="co">#&gt; 2                               a6388a6f-6076-401b-9b30-7d4306a20035</span></span>
<span><span class="co">#&gt; 3                               842c6f5d-4a94-4eef-8510-8c792d1124bc</span></span>
<span><span class="co">#&gt; 4                               74520626-b0ba-4ee9-86b5-714649554def</span></span>
<span><span class="co">#&gt; 5                               396a9124-fb20-4822-bf9c-e93fdf7c999a</span></span>
<span><span class="co">#&gt; 6    10.1016/j.cell.2021.12.018 e84f2780-51e8-4cfa-8aa0-13bbfef677c7</span></span>
<span><span class="co">#&gt; 7    10.1016/j.cell.2021.12.018 dfdf1ae2-d624-4004-9353-f18b902f6bca</span></span>
<span><span class="co">#&gt; 8    10.1016/j.cell.2021.12.018 d1cbed97-d88f-4954-8925-13302fe30b39</span></span>
<span><span class="co">#&gt; 9    10.1016/j.cell.2021.12.018 b03e4ef8-4e6b-47f4-84a7-e8ed033d08cd</span></span>
<span><span class="co">#&gt; 10 10.1016/j.celrep.2023.111994 f1f123cc-ca2c-460f-b7f1-88240efb1e82</span></span>
<span><span class="co">#&gt; 11 10.1016/j.celrep.2023.111994 de94c504-4b58-4f42-b68d-74a8e4892f0e</span></span>
<span><span class="co">#&gt; 12 10.1016/j.celrep.2023.111994 da684768-fb01-455b-9f0f-b63a3e2f844f</span></span>
<span><span class="co">#&gt;                                     dataset_title</span></span>
<span><span class="co">#&gt; 1                           22 integrated samples</span></span>
<span><span class="co">#&gt; 2                  scRNA-seq data - myeloid cells</span></span>
<span><span class="co">#&gt; 3                      scRNA-seq data - all cells</span></span>
<span><span class="co">#&gt; 4                        scRNA-seq data - T cells</span></span>
<span><span class="co">#&gt; 5                        scRNA-seq data - B cells</span></span>
<span><span class="co">#&gt; 6              All cells from human liver dataset</span></span>
<span><span class="co">#&gt; 7          Myeloid cells from human liver dataset</span></span>
<span><span class="co">#&gt; 8    CD45 negative cells from human liver dataset</span></span>
<span><span class="co">#&gt; 9         Lymphoid cells from human liver dataset</span></span>
<span><span class="co">#&gt; 10 UMAP visualization of keratinocyte subclusters</span></span>
<span><span class="co">#&gt; 11          UMAP visualization of all 12 datasets</span></span>
<span><span class="co">#&gt; 12   UMAP visualization of fibroblast subclusters</span></span>
<span><span class="co">#&gt;                            dataset_h5ad_path dataset_total_cell_count</span></span>
<span><span class="co">#&gt; 1  9d8e5dca-03a3-457d-b7fb-844c75735c83.h5ad                    72335</span></span>
<span><span class="co">#&gt; 2  a6388a6f-6076-401b-9b30-7d4306a20035.h5ad                    30789</span></span>
<span><span class="co">#&gt; 3  842c6f5d-4a94-4eef-8510-8c792d1124bc.h5ad                   714331</span></span>
<span><span class="co">#&gt; 4  74520626-b0ba-4ee9-86b5-714649554def.h5ad                    76567</span></span>
<span><span class="co">#&gt; 5  396a9124-fb20-4822-bf9c-e93fdf7c999a.h5ad                    12510</span></span>
<span><span class="co">#&gt; 6  e84f2780-51e8-4cfa-8aa0-13bbfef677c7.h5ad                   167598</span></span>
<span><span class="co">#&gt; 7  dfdf1ae2-d624-4004-9353-f18b902f6bca.h5ad                    40821</span></span>
<span><span class="co">#&gt; 8  d1cbed97-d88f-4954-8925-13302fe30b39.h5ad                    15481</span></span>
<span><span class="co">#&gt; 9  b03e4ef8-4e6b-47f4-84a7-e8ed033d08cd.h5ad                   105790</span></span>
<span><span class="co">#&gt; 10 f1f123cc-ca2c-460f-b7f1-88240efb1e82.h5ad                     9471</span></span>
<span><span class="co">#&gt; 11 de94c504-4b58-4f42-b68d-74a8e4892f0e.h5ad                    15243</span></span>
<span><span class="co">#&gt; 12 da684768-fb01-455b-9f0f-b63a3e2f844f.h5ad                     2303</span></span>
<span><span class="co">#&gt;  [ reached 'max' / getOption("max.print") -- omitted 550 rows ]</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="fetching-the-dataset-presence-matrix">Fetching the dataset presence matrix<a class="anchor" aria-label="anchor" href="#fetching-the-dataset-presence-matrix"></a>
</h2>
<p>Now let’s fetch the dataset presence matrix.</p>
<p>For convenience, read the entire presence matrix (for Homo sapiens)
into a sparse matrix. There is a convenience API providing this
capability:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">presence_matrix</span> <span class="op">&lt;-</span> <span class="fu">cellxgene.census</span><span class="fu">::</span><span class="fu"><a href="../reference/get_presence_matrix.html">get_presence_matrix</a></span><span class="op">(</span><span class="va">census</span>, <span class="st">"Homo sapiens"</span>, <span class="st">"RNA"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">presence_matrix</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; NULL</span></span></code></pre></div>
<p>We also need the <code>var</code> dataframe, which is read into an R
data frame for convenient manipulation:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">var_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="va">human_rna</span><span class="op">$</span><span class="va">var</span><span class="op">$</span><span class="fu">read</span><span class="op">(</span><span class="op">)</span><span class="op">$</span><span class="fu">concat</span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">var_df</span><span class="op">)</span></span>
<span><span class="co">#&gt;    soma_joinid      feature_id  feature_name feature_length</span></span>
<span><span class="co">#&gt; 1            0 ENSG00000243485   MIR1302-2HG           1021</span></span>
<span><span class="co">#&gt; 2            1 ENSG00000237613       FAM138A           1219</span></span>
<span><span class="co">#&gt; 3            2 ENSG00000186092         OR4F5           2618</span></span>
<span><span class="co">#&gt; 4            3 ENSG00000238009  RP11-34P13.7           3726</span></span>
<span><span class="co">#&gt; 5            4 ENSG00000239945  RP11-34P13.8           1319</span></span>
<span><span class="co">#&gt; 6            5 ENSG00000239906 RP11-34P13.14            323</span></span>
<span><span class="co">#&gt; 7            6 ENSG00000241860 RP11-34P13.13           7559</span></span>
<span><span class="co">#&gt; 8            7 ENSG00000241599  RP11-34P13.9            457</span></span>
<span><span class="co">#&gt; 9            8 ENSG00000286448    AP006222.3            736</span></span>
<span><span class="co">#&gt; 10           9 ENSG00000236601  RP4-669L17.2           1095</span></span>
<span><span class="co">#&gt; 11          10 ENSG00000284733        OR4F29            939</span></span>
<span><span class="co">#&gt; 12          11 ENSG00000235146  RP5-857K21.2            635</span></span>
<span><span class="co">#&gt; 13          12 ENSG00000284662        OR4F16            995</span></span>
<span><span class="co">#&gt; 14          13 ENSG00000229905 RP11-206L10.4            456</span></span>
<span><span class="co">#&gt; 15          14 ENSG00000237491     LINC01409           8413</span></span>
<span><span class="co">#&gt; 16          15 ENSG00000177757        FAM87B           1947</span></span>
<span><span class="co">#&gt; 17          16 ENSG00000228794     LINC01128          15682</span></span>
<span><span class="co">#&gt; 18          17 ENSG00000225880     LINC00115           1317</span></span>
<span><span class="co">#&gt; 19          18 ENSG00000230368        FAM41C           1971</span></span>
<span><span class="co">#&gt; 20          19 ENSG00000272438  RP11-54O7.16            351</span></span>
<span><span class="co">#&gt; 21          20 ENSG00000230699   RP11-54O7.1           3043</span></span>
<span><span class="co">#&gt; 22          21 ENSG00000241180   RP11-54O7.2            443</span></span>
<span><span class="co">#&gt; 23          22 ENSG00000223764     LINC02593           4152</span></span>
<span><span class="co">#&gt; 24          23 ENSG00000187634        SAMD11           4172</span></span>
<span><span class="co">#&gt; 25          24 ENSG00000188976         NOC2L           5540</span></span>
<span><span class="co">#&gt;  [ reached 'max' / getOption("max.print") -- omitted 60639 rows ]</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="identifying-genes-measured-in-a-specific-dataset">Identifying genes measured in a specific dataset<a class="anchor" aria-label="anchor" href="#identifying-genes-measured-in-a-specific-dataset"></a>
</h2>
<p>Now that we have the dataset table, the genes metadata table, and the
dataset presence matrix, we can check if a gene or set of genes were
measured in a specific dataset.</p>
<p><strong>Important:</strong> the presence matrix is indexed by
<code>soma_joinid</code>, and is <em>NOT</em> positionally indexed. In
other words:</p>
<ul>
<li>the first dimension of the presence matrix is the dataset’s
<code>soma_joinid</code>, as stored in the <code>census_datasets</code>
dataframe.</li>
<li>the second dimension of the presence matrix is the feature’s
<code>soma_joinid</code>, as stored in the <code>var</code>
dataframe.</li>
</ul>
<p>Let’s find out if the the gene <code>"ENSG00000286096"</code> was
measured in the dataset with id
<code>"97a17473-e2b1-4f31-a544-44a60773e2dd"</code>.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">var_joinid</span> <span class="op">&lt;-</span> <span class="va">var_df</span><span class="op">$</span><span class="va">soma_joinid</span><span class="op">[</span><span class="va">var_df</span><span class="op">$</span><span class="va">feature_id</span> <span class="op">==</span> <span class="st">"ENSG00000286096"</span><span class="op">]</span></span>
<span><span class="va">dataset_joinid</span> <span class="op">&lt;-</span> <span class="va">datasets_df</span><span class="op">$</span><span class="va">soma_joinid</span><span class="op">[</span><span class="va">datasets_df</span><span class="op">$</span><span class="va">dataset_id</span> <span class="op">==</span> <span class="st">"97a17473-e2b1-4f31-a544-44a60773e2dd"</span><span class="op">]</span></span>
<span><span class="va">is_present</span> <span class="op">&lt;-</span> <span class="va">presence_matrix</span><span class="op">$</span><span class="fu">take</span><span class="op">(</span><span class="va">dataset_joinid</span>, <span class="va">var_joinid</span><span class="op">)</span><span class="op">$</span><span class="fu">get_one_based_matrix</span><span class="op">(</span><span class="op">)</span><span class="op">[</span>, , drop <span class="op">=</span> <span class="cn">TRUE</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"Feature is"</span>, <span class="kw">if</span> <span class="op">(</span><span class="va">is_present</span><span class="op">)</span> <span class="st">"present."</span> <span class="kw">else</span> <span class="st">"not present."</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; Feature is present.</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="identifying-datasets-that-measured-specific-genes">Identifying datasets that measured specific genes<a class="anchor" aria-label="anchor" href="#identifying-datasets-that-measured-specific-genes"></a>
</h2>
<p>Similarly, we can determine the datasets that measured a specific
gene or set of genes.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Grab the feature's soma_joinid from the var dataframe</span></span>
<span><span class="va">var_joinid</span> <span class="op">&lt;-</span> <span class="va">var_df</span><span class="op">$</span><span class="va">soma_joinid</span><span class="op">[</span><span class="va">var_df</span><span class="op">$</span><span class="va">feature_id</span> <span class="op">==</span> <span class="st">"ENSG00000286096"</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># The presence matrix is indexed by the joinids of the dataset and var dataframes,</span></span>
<span><span class="co"># so slice out the feature of interest by its joinid.</span></span>
<span><span class="va">dataset_joinids</span> <span class="op">&lt;-</span> <span class="va">datasets_df</span><span class="op">$</span><span class="va">soma_joinid</span><span class="op">[</span></span>
<span>  <span class="va">presence_matrix</span><span class="op">$</span><span class="fu">take</span><span class="op">(</span>j <span class="op">=</span> <span class="va">var_joinid</span><span class="op">)</span><span class="op">$</span><span class="fu">get_one_based_matrix</span><span class="op">(</span><span class="op">)</span><span class="op">[</span>, , drop <span class="op">=</span> <span class="cn">TRUE</span><span class="op">]</span> <span class="op">!=</span> <span class="fl">0</span></span>
<span><span class="op">]</span></span>
<span></span>
<span><span class="co"># From the datasets dataframe, slice out the datasets which have a joinid in the list</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">datasets_df</span><span class="op">[</span><span class="va">dataset_joinids</span>, <span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt;     soma_joinid                        collection_id</span></span>
<span><span class="co">#&gt; 105         104 283d65eb-dd53-496d-adb7-7570c7caa443</span></span>
<span><span class="co">#&gt; 109         108 283d65eb-dd53-496d-adb7-7570c7caa443</span></span>
<span><span class="co">#&gt; 126         125 283d65eb-dd53-496d-adb7-7570c7caa443</span></span>
<span><span class="co">#&gt; 131         130 283d65eb-dd53-496d-adb7-7570c7caa443</span></span>
<span><span class="co">#&gt; 141         140 283d65eb-dd53-496d-adb7-7570c7caa443</span></span>
<span><span class="co">#&gt; 142         141 283d65eb-dd53-496d-adb7-7570c7caa443</span></span>
<span><span class="co">#&gt; 143         142 283d65eb-dd53-496d-adb7-7570c7caa443</span></span>
<span><span class="co">#&gt; 144         143 283d65eb-dd53-496d-adb7-7570c7caa443</span></span>
<span><span class="co">#&gt; 149         148 283d65eb-dd53-496d-adb7-7570c7caa443</span></span>
<span><span class="co">#&gt; 162         161 283d65eb-dd53-496d-adb7-7570c7caa443</span></span>
<span><span class="co">#&gt; 163         162 283d65eb-dd53-496d-adb7-7570c7caa443</span></span>
<span><span class="co">#&gt; 172         171 283d65eb-dd53-496d-adb7-7570c7caa443</span></span>
<span><span class="co">#&gt;                                                         collection_name</span></span>
<span><span class="co">#&gt; 105 Transcriptomic diversity of cell types across the adult human brain</span></span>
<span><span class="co">#&gt; 109 Transcriptomic diversity of cell types across the adult human brain</span></span>
<span><span class="co">#&gt; 126 Transcriptomic diversity of cell types across the adult human brain</span></span>
<span><span class="co">#&gt; 131 Transcriptomic diversity of cell types across the adult human brain</span></span>
<span><span class="co">#&gt; 141 Transcriptomic diversity of cell types across the adult human brain</span></span>
<span><span class="co">#&gt; 142 Transcriptomic diversity of cell types across the adult human brain</span></span>
<span><span class="co">#&gt; 143 Transcriptomic diversity of cell types across the adult human brain</span></span>
<span><span class="co">#&gt; 144 Transcriptomic diversity of cell types across the adult human brain</span></span>
<span><span class="co">#&gt; 149 Transcriptomic diversity of cell types across the adult human brain</span></span>
<span><span class="co">#&gt; 162 Transcriptomic diversity of cell types across the adult human brain</span></span>
<span><span class="co">#&gt; 163 Transcriptomic diversity of cell types across the adult human brain</span></span>
<span><span class="co">#&gt; 172 Transcriptomic diversity of cell types across the adult human brain</span></span>
<span><span class="co">#&gt;                collection_doi                           dataset_id</span></span>
<span><span class="co">#&gt; 105 10.1101/2022.10.12.511898 ff7d15fa-f4b6-4a0e-992e-fd0c9d088ded</span></span>
<span><span class="co">#&gt; 109 10.1101/2022.10.12.511898 f9034091-2e8f-4ac6-9874-e7b7eb566824</span></span>
<span><span class="co">#&gt; 126 10.1101/2022.10.12.511898 e1f595f6-ba2c-495e-9bee-7056f116b1e4</span></span>
<span><span class="co">#&gt; 131 10.1101/2022.10.12.511898 d5452b83-7c3d-4d7c-ab7a-c7fece7196c5</span></span>
<span><span class="co">#&gt; 141 10.1101/2022.10.12.511898 c4dd26a8-d956-4bee-a233-44b573f2ce27</span></span>
<span><span class="co">#&gt; 142 10.1101/2022.10.12.511898 c4b03352-af8d-492a-8d6b-40f304e0a122</span></span>
<span><span class="co">#&gt; 143 10.1101/2022.10.12.511898 c2aad8fc-b63b-4f9b-9cfd-baf7bc9c1771</span></span>
<span><span class="co">#&gt; 144 10.1101/2022.10.12.511898 c202b243-1aa1-4b16-bc9a-b36241f3b1e3</span></span>
<span><span class="co">#&gt; 149 10.1101/2022.10.12.511898 adda0684-f8ea-4403-b393-2a25607077c4</span></span>
<span><span class="co">#&gt; 162 10.1101/2022.10.12.511898 949ec7fc-ac54-47c9-bb6f-ee9e67688cce</span></span>
<span><span class="co">#&gt; 163 10.1101/2022.10.12.511898 9372df2d-13d6-4fac-980b-919a5b7eb483</span></span>
<span><span class="co">#&gt; 172 10.1101/2022.10.12.511898 7cd3c321-e487-4b51-a8af-9bf671b59fe5</span></span>
<span><span class="co">#&gt;                                                                                                                                                                  dataset_title</span></span>
<span><span class="co">#&gt; 105                                                                                            Dissection: Cerebral cortex (Cx) - Cuneus, rostral part - Area Prostriata - Pro</span></span>
<span><span class="co">#&gt; 109 Dissection: Myelencephalon (medulla oblongata) (Mo) - medullary reticular formation (MoRF) &amp; efferent nuclei of cranial nerves in the medulla oblongata (MoEN) - MoRF-MoEN</span></span>
<span><span class="co">#&gt; 126                                                                                                             Dissection: Cerebral cortex (Cx) - Middle Temporal Gyrus - MTG</span></span>
<span><span class="co">#&gt; 131                                                                                                                        Dissection: Body of hippocampus (HiB) - Rostral CA3</span></span>
<span><span class="co">#&gt; 141                                                                                                                      Dissection: Body of hippocampus (HiB) - Rostral CA1-2</span></span>
<span><span class="co">#&gt; 142                                                                                                                                          Supercluster: Medium spiny neuron</span></span>
<span><span class="co">#&gt; 143                                                                                            Dissection: Cerebral cortex (Cx) - Temporal pole (TP) - Temporopolar area - A38</span></span>
<span><span class="co">#&gt; 144                                                                                                                                          Supercluster: Amygdala excitatory</span></span>
<span><span class="co">#&gt; 149                                                                   Dissection: Thalamus (THM) - posterior nuclear complex of thalamus (PoN) - medial geniculate nuclei (MG)</span></span>
<span><span class="co">#&gt; 162                                                                                       Dissection: Cerebral cortex (Cx) - Short insular gyri - Granular insular cortex - Ig</span></span>
<span><span class="co">#&gt; 163                                                                                                     Dissection: Midbrain (M) - Periaqueductal gray and nearby nuclei - PAG</span></span>
<span><span class="co">#&gt; 172                                                                         Dissection: Pons (Pn) - other nuclei in pontine tegmentum (XPnTg) - dorsal tegmental nucleus - DTg</span></span>
<span><span class="co">#&gt;                             dataset_h5ad_path dataset_total_cell_count</span></span>
<span><span class="co">#&gt; 105 ff7d15fa-f4b6-4a0e-992e-fd0c9d088ded.h5ad                    28051</span></span>
<span><span class="co">#&gt; 109 f9034091-2e8f-4ac6-9874-e7b7eb566824.h5ad                    23120</span></span>
<span><span class="co">#&gt; 126 e1f595f6-ba2c-495e-9bee-7056f116b1e4.h5ad                   107301</span></span>
<span><span class="co">#&gt; 131 d5452b83-7c3d-4d7c-ab7a-c7fece7196c5.h5ad                     8077</span></span>
<span><span class="co">#&gt; 141 c4dd26a8-d956-4bee-a233-44b573f2ce27.h5ad                    11265</span></span>
<span><span class="co">#&gt; 142 c4b03352-af8d-492a-8d6b-40f304e0a122.h5ad                   152189</span></span>
<span><span class="co">#&gt; 143 c2aad8fc-b63b-4f9b-9cfd-baf7bc9c1771.h5ad                    37642</span></span>
<span><span class="co">#&gt; 144 c202b243-1aa1-4b16-bc9a-b36241f3b1e3.h5ad                   109452</span></span>
<span><span class="co">#&gt; 149 adda0684-f8ea-4403-b393-2a25607077c4.h5ad                    14226</span></span>
<span><span class="co">#&gt; 162 949ec7fc-ac54-47c9-bb6f-ee9e67688cce.h5ad                    38937</span></span>
<span><span class="co">#&gt; 163 9372df2d-13d6-4fac-980b-919a5b7eb483.h5ad                    33794</span></span>
<span><span class="co">#&gt; 172 7cd3c321-e487-4b51-a8af-9bf671b59fe5.h5ad                    26095</span></span>
<span><span class="co">#&gt;  [ reached 'max' / getOption("max.print") -- omitted 12 rows ]</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="identifying-all-genes-measured-in-a-dataset">Identifying all genes measured in a dataset<a class="anchor" aria-label="anchor" href="#identifying-all-genes-measured-in-a-dataset"></a>
</h2>
<p>Finally, we can find the set of genes that were measured in the cells
of a given dataset.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Slice the dataset(s) of interest, and get the joinid(s)</span></span>
<span><span class="va">dataset_joinids</span> <span class="op">&lt;-</span> <span class="va">datasets_df</span><span class="op">$</span><span class="va">soma_joinid</span><span class="op">[</span><span class="va">datasets_df</span><span class="op">$</span><span class="va">collection_id</span> <span class="op">==</span> <span class="st">"17481d16-ee44-49e5-bcf0-28c0780d8c4a"</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># Slice the presence matrix by the first dimension, i.e., by dataset</span></span>
<span><span class="va">var_joinids</span> <span class="op">&lt;-</span> <span class="va">var_df</span><span class="op">$</span><span class="va">soma_joinid</span><span class="op">[</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="fu">Matrix</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/colSums.html" class="external-link">colSums</a></span><span class="op">(</span><span class="va">presence_matrix</span><span class="op">$</span><span class="fu">take</span><span class="op">(</span>i <span class="op">=</span> <span class="va">dataset_joinids</span><span class="op">)</span><span class="op">$</span><span class="fu">get_one_based_matrix</span><span class="op">(</span><span class="op">)</span><span class="op">[</span>, , drop <span class="op">=</span> <span class="cn">TRUE</span><span class="op">]</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span></span>
<span><span class="op">]</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">var_df</span><span class="op">[</span><span class="va">var_joinids</span>, <span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt;    soma_joinid      feature_id  feature_name feature_length</span></span>
<span><span class="co">#&gt; 3            2 ENSG00000186092         OR4F5           2618</span></span>
<span><span class="co">#&gt; 4            3 ENSG00000238009  RP11-34P13.7           3726</span></span>
<span><span class="co">#&gt; 13          12 ENSG00000284662        OR4F16            995</span></span>
<span><span class="co">#&gt; 14          13 ENSG00000229905 RP11-206L10.4            456</span></span>
<span><span class="co">#&gt; 15          14 ENSG00000237491     LINC01409           8413</span></span>
<span><span class="co">#&gt; 17          16 ENSG00000228794     LINC01128          15682</span></span>
<span><span class="co">#&gt; 18          17 ENSG00000225880     LINC00115           1317</span></span>
<span><span class="co">#&gt; 19          18 ENSG00000230368        FAM41C           1971</span></span>
<span><span class="co">#&gt; 20          19 ENSG00000272438  RP11-54O7.16            351</span></span>
<span><span class="co">#&gt; 21          20 ENSG00000230699   RP11-54O7.1           3043</span></span>
<span><span class="co">#&gt; 22          21 ENSG00000241180   RP11-54O7.2            443</span></span>
<span><span class="co">#&gt; 23          22 ENSG00000223764     LINC02593           4152</span></span>
<span><span class="co">#&gt; 24          23 ENSG00000187634        SAMD11           4172</span></span>
<span><span class="co">#&gt; 25          24 ENSG00000188976         NOC2L           5540</span></span>
<span><span class="co">#&gt; 26          25 ENSG00000187961        KLHL17           3402</span></span>
<span><span class="co">#&gt; 27          26 ENSG00000187583       PLEKHN1           3451</span></span>
<span><span class="co">#&gt; 28          27 ENSG00000187642         PERM1           3424</span></span>
<span><span class="co">#&gt; 29          28 ENSG00000272512  RP11-54O7.17           2086</span></span>
<span><span class="co">#&gt; 30          29 ENSG00000188290          HES4           1118</span></span>
<span><span class="co">#&gt; 31          30 ENSG00000187608         ISG15            867</span></span>
<span><span class="co">#&gt; 32          31 ENSG00000224969  RP11-54O7.11            471</span></span>
<span><span class="co">#&gt; 35          34 ENSG00000273443  RP11-54O7.18            829</span></span>
<span><span class="co">#&gt; 36          35 ENSG00000237330        RNF223           1900</span></span>
<span><span class="co">#&gt; 38          37 ENSG00000285812 RP11-465B22.9           1095</span></span>
<span><span class="co">#&gt; 39          38 ENSG00000223823     LINC01342           1620</span></span>
<span><span class="co">#&gt;  [ reached 'max' / getOption("max.print") -- omitted 27186 rows ]</span></span></code></pre></div>
<div class="section level3">
<h3 id="close-the-census">Close the census<a class="anchor" aria-label="anchor" href="#close-the-census"></a>
</h3>
<p>After use, the census object should be closed to release memory and
other resources.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">census</span><span class="op">$</span><span class="fu">close</span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>This also closes all SOMA objects accessed via the top-level
<code>census</code>. Closing can be automated using
<code>on.exit(census$close(), add = TRUE)</code> immediately after
<code>census &lt;- open_soma()</code>.</p>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Chan Zuckerberg Initiative.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
