<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="cellxgene.census">
<title>Computing on X using online (incremental) algorithms • cellxgene.census</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Computing on X using online (incremental) algorithms">
<meta property="og:description" content="cellxgene.census">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><script defer data-domain="chanzuckerberg.github.io/cellxgene-census" src="https://plausible.io/js/script.js"></script>
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">cellxgene.census</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.9.1</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="external-link nav-link" href="https://chanzuckerberg.github.io/cellxgene-census/">Main Census Site</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <h6 class="dropdown-header" data-toc-skip>Use Census embeddings</h6>
    <a class="dropdown-item" href="../articles/census_access_maintained_embeddings.html">Access CELLxGENE-maintained embeddings (scVI, Geneformer)</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Explore and analyze the data</h6>
    <a class="dropdown-item" href="../articles/comp_bio_census_info.html">Learning about the CZ CELLxGENE Census</a>
    <a class="dropdown-item" href="../articles/comp_bio_summarize_axis_query.html">Summarizing cell and gene metadata</a>
    <a class="dropdown-item" href="../articles/comp_bio_data_integration.html">Integrating multi-dataset slices of data with Seurat</a>
    <a class="dropdown-item" href="../articles/comp_bio_normalizing_full_gene_sequencing.html">Normalizing full-length gene sequencing data</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>cellxgene.census capabilities</h6>
    <a class="dropdown-item" href="../articles/census_query_extract.html">Querying and fetching the single-cell data and cell/gene metadata</a>
    <a class="dropdown-item" href="../articles/census_datasets.html">Census Datasets example</a>
    <a class="dropdown-item" href="../articles/census_dataset_presence.html">Genes measured in each cell (dataset presence matrix)</a>
    <a class="dropdown-item" href="../articles/census_compute_over_X.html">Computing on X using online (incremental) algorithms</a>
    <a class="dropdown-item" href="../articles/census_access_maintained_embeddings.html">Access CELLxGENE-maintained embeddings (scVI, Geneformer)</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/chanzuckerberg/cellxgene-census/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">

<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Computing on X using online (incremental) algorithms</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/chanzuckerberg/cellxgene-census/tree/main/api/r/cellxgene.census/vignettes_/census_compute_over_X.Rmd" class="external-link"><code>vignettes/census_compute_over_X.Rmd</code></a></small>
      <div class="d-none name"><code>census_compute_over_X.Rmd</code></div>
    </div>

    
    
<!--
THIS VIGNETTE IS BASED ON:
https://github.com/chanzuckerberg/cellxgene-census/blob/main/api/python/notebooks/api_demo/census_compute_over_X.ipynb
-->
<p>This tutorial showcases computing per-gene and per-cell statistics for a user-defined query using out-of-core operations, where incremental (online) algorithms are needed.</p>
<p><em>NOTE:</em> when query results are small enough to fit in memory, it’s usually easier to use <code><a href="../reference/get_seurat.html">get_seurat()</a></code> or <code><a href="../reference/get_single_cell_experiment.html">get_single_cell_experiment()</a></code>.</p>
<p><strong>Contents</strong></p>
<ul>
<li>Incremental mean calculation</li>
<li>Counting cells grouped by dataset and gene</li>
</ul>
<div class="section level2">
<h2 id="incremental-mean-calculation">Incremental mean calculation<a class="anchor" aria-label="anchor" href="#incremental-mean-calculation"></a>
</h2>
<p>Many statistics, such as marginal means, are easy to calculate incrementally. Let’s begin with a query on the <code>X$raw</code> sparse matrix of unnormalized read counts, which will return results in shards incrementally accumulate the read count for each gene, then divide by the cell count to get the mean reads per cell for each gene.</p>
<p>First define a query - in this case a slice over the obs axis for cells with a specific tissue &amp; sex value, and all genes on the var axis. The <code>query$X()</code> method returns an iterator of results, each as an Arrow Table. Each table will contain the sparse X data and obs/var coordinates, using standard SOMA names:</p>
<ul>
<li>
<code>soma_data</code> - the X values (float32)</li>
<li>
<code>soma_dim_0</code> - the obs coordinate (int64)</li>
<li>
<code>soma_dim_1</code> - the var coordinate (int64)</li>
</ul>
<p><strong>Important:</strong> the X matrices are joined to var/obs axis DataFrames by an integer join “id” (aka <code>soma_joinid</code>). They are NOT positionally indexed, and any given cell or gene may have a <code>soma_joinid</code> of any value (e.g., a large integer). In other words, for any given X value, the <code>soma_dim_0</code> corresponds to the soma_joinid in the <code>obs</code> dataframe, and the <code>soma_dim_1</code> coordinate corresponds to the <code>soma_joinid</code> in the <code>var</code> dataframe.</p>
<p>For convenience, the query class includes a utility to simplify operations on query slices. <code>query$indexer</code> is an indexer used to wrap the output of <code>query$X()</code>, converting from <code>soma_joinid</code>s to positional indexing in the query results. Positions are <code>[0, N)</code>, where N are the number of results on the query for any given axis.</p>
<p>Key points:</p>
<ul>
<li>it is expensive to query and read the results - so rather than make multiple passes over the data, read it once and perform multiple computations.</li>
<li>by default, data in the census is indexed by <code>soma_joinid</code> and not positionally.</li>
</ul>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/single-cell-data/TileDB-SOMA" class="external-link">"tiledbsoma"</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/chanzuckerberg/cellxgene-census" class="external-link">"cellxgene.census"</a></span><span class="op">)</span></span>
<span><span class="va">census</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/open_soma.html">open_soma</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="va">query</span> <span class="op">&lt;-</span> <span class="va">census</span><span class="op">$</span><span class="fu">get</span><span class="op">(</span><span class="st">"census_data"</span><span class="op">)</span><span class="op">$</span><span class="fu">get</span><span class="op">(</span><span class="st">"mus_musculus"</span><span class="op">)</span><span class="op">$</span><span class="fu">axis_query</span><span class="op">(</span></span>
<span>  measurement_name <span class="op">=</span> <span class="st">"RNA"</span>,</span>
<span>  obs_query <span class="op">=</span> <span class="va"><a href="https://rdrr.io/pkg/tiledbsoma/man/SOMAAxisQuery.html" class="external-link">SOMAAxisQuery</a></span><span class="op">$</span><span class="fu">new</span><span class="op">(</span>value_filter <span class="op">=</span> <span class="st">"tissue=='brain' &amp;&amp; sex=='male'"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">genes_df</span> <span class="op">&lt;-</span> <span class="va">query</span><span class="op">$</span><span class="fu">var</span><span class="op">(</span>column_names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"feature_id"</span>, <span class="st">"feature_name"</span><span class="op">)</span><span class="op">)</span><span class="op">$</span><span class="fu">concat</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">genes_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="va">genes_df</span><span class="op">)</span></span>
<span><span class="va">n_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">genes_df</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># accumulator vector (for each gene) for the total count over all cells in X("raw")</span></span>
<span><span class="va">raw_sum_by_gene</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">numeric</a></span><span class="op">(</span><span class="va">n_genes</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">raw_sum_by_gene</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">genes_df</span><span class="op">$</span><span class="va">feature_id</span></span>
<span></span>
<span><span class="co"># iterate through in-memory shards of query results</span></span>
<span><span class="va">tables</span> <span class="op">&lt;-</span> <span class="va">query</span><span class="op">$</span><span class="fu">X</span><span class="op">(</span><span class="st">"raw"</span><span class="op">)</span><span class="op">$</span><span class="fu">tables</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="kw">while</span> <span class="op">(</span><span class="op">!</span><span class="va">tables</span><span class="op">$</span><span class="fu">read_complete</span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">table_part</span> <span class="op">&lt;-</span> <span class="va">tables</span><span class="op">$</span><span class="fu">read_next</span><span class="op">(</span><span class="op">)</span></span>
<span>  <span class="co"># table_part is an Arrow table with the columns mentioned above. The result</span></span>
<span>  <span class="co"># order is not guaranteed!</span></span>
<span></span>
<span>  <span class="co"># table_part$soma_dim_1 is the var/gene soma_joinid. But note that these are</span></span>
<span>  <span class="co"># arbitrary int64 id's, and moreover each table_part may exhibit only a subset</span></span>
<span>  <span class="co"># of the values we'll see over all query results. query$indexer helps us map</span></span>
<span>  <span class="co"># any given soma_dim_1 values onto positions in query$var() (genes_df), that is</span></span>
<span>  <span class="co"># the union of all values we'll see.</span></span>
<span>  <span class="va">gene_indexes</span> <span class="op">&lt;-</span> <span class="va">query</span><span class="op">$</span><span class="va">indexer</span><span class="op">$</span><span class="fu">by_var</span><span class="op">(</span><span class="va">table_part</span><span class="op">$</span><span class="va">soma_dim_1</span><span class="op">)</span><span class="op">$</span><span class="fu">as_vector</span><span class="op">(</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/stopifnot.html" class="external-link">stopifnot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">gene_indexes</span> <span class="op">&gt;=</span> <span class="va">n_genes</span><span class="op">)</span> <span class="op">==</span> <span class="fl">0</span><span class="op">)</span></span>
<span>  <span class="co"># sum(table_part) group by gene, yielding a numeric vector with the gene_index</span></span>
<span>  <span class="co"># in its names</span></span>
<span>  <span class="va">sum_part</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tapply.html" class="external-link">tapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/vector.html" class="external-link">as.vector</a></span><span class="op">(</span><span class="va">table_part</span><span class="op">$</span><span class="va">soma_data</span><span class="op">)</span>, <span class="va">gene_indexes</span>, <span class="va">sum</span><span class="op">)</span></span>
<span>  <span class="co"># update the accumulator vector</span></span>
<span>  <span class="va">which_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/integer.html" class="external-link">as.integer</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">sum_part</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fl">1</span>  <span class="co"># nb: gene_indexes is zero-based</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/stopifnot.html" class="external-link">stopifnot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">which_genes</span> <span class="op">&gt;</span> <span class="va">n_genes</span><span class="op">)</span> <span class="op">==</span> <span class="fl">0</span><span class="op">)</span></span>
<span>  <span class="va">raw_sum_by_gene</span><span class="op">[</span><span class="va">which_genes</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">raw_sum_by_gene</span><span class="op">[</span><span class="va">which_genes</span><span class="op">]</span> <span class="op">+</span> <span class="va">sum_part</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="co"># Divide each sum by cell count to get mean reads per cell (for each gene),</span></span>
<span><span class="co"># implicitly averaging in all zero entries in X even though they weren't included</span></span>
<span><span class="co"># in the sparse query results.</span></span>
<span><span class="va">genes_df</span><span class="op">$</span><span class="va">raw_mean</span> <span class="op">&lt;-</span> <span class="va">raw_sum_by_gene</span> <span class="op">/</span> <span class="va">query</span><span class="op">$</span><span class="va">n_obs</span></span>
<span><span class="va">genes_df</span></span>
<span><span class="co">#&gt;            feature_id  feature_name     raw_mean</span></span>
<span><span class="co">#&gt; 1  ENSMUSG00000051951          Xkr4 1.283861e+00</span></span>
<span><span class="co">#&gt; 2  ENSMUSG00000089699        Gm1992 0.000000e+00</span></span>
<span><span class="co">#&gt; 3  ENSMUSG00000102343       Gm37381 0.000000e+00</span></span>
<span><span class="co">#&gt; 4  ENSMUSG00000025900           Rp1 2.914160e-01</span></span>
<span><span class="co">#&gt; 5  ENSMUSG00000025902         Sox17 6.074154e+01</span></span>
<span><span class="co">#&gt; 6  ENSMUSG00000104328       Gm37323 5.701742e-05</span></span>
<span><span class="co">#&gt; 7  ENSMUSG00000033845        Mrpl15 3.621738e+01</span></span>
<span><span class="co">#&gt; 8  ENSMUSG00000025903        Lypla1 1.827366e+01</span></span>
<span><span class="co">#&gt; 9  ENSMUSG00000104217       Gm37988 0.000000e+00</span></span>
<span><span class="co">#&gt; 10 ENSMUSG00000033813         Tcea1 3.960339e+01</span></span>
<span><span class="co">#&gt; 11 ENSMUSG00000002459         Rgs20 3.212989e+00</span></span>
<span><span class="co">#&gt; 12 ENSMUSG00000085623       Gm16041 5.701742e-05</span></span>
<span><span class="co">#&gt; 13 ENSMUSG00000033793       Atp6v1h 6.866793e+01</span></span>
<span><span class="co">#&gt; 14 ENSMUSG00000025905         Oprk1 4.198763e-01</span></span>
<span><span class="co">#&gt; 15 ENSMUSG00000033774        Npbwr1 1.140348e-04</span></span>
<span><span class="co">#&gt; 16 ENSMUSG00000025907        Rb1cc1 3.342227e+01</span></span>
<span><span class="co">#&gt; 17 ENSMUSG00000090031 4732440D04Rik 1.317102e-02</span></span>
<span><span class="co">#&gt; 18 ENSMUSG00000087247        Alkal1 5.701742e-05</span></span>
<span><span class="co">#&gt; 19 ENSMUSG00000033740          St18 1.525501e+01</span></span>
<span><span class="co">#&gt; 20 ENSMUSG00000051285        Pcmtd1 4.981224e+01</span></span>
<span><span class="co">#&gt; 21 ENSMUSG00000097797       Gm26901 4.276306e-04</span></span>
<span><span class="co">#&gt; 22 ENSMUSG00000103067       Gm30414 0.000000e+00</span></span>
<span><span class="co">#&gt; 23 ENSMUSG00000025909         Sntg1 1.083131e+00</span></span>
<span><span class="co">#&gt; 24 ENSMUSG00000061024          Rrs1 1.929504e+01</span></span>
<span><span class="co">#&gt; 25 ENSMUSG00000025911        Adhfe1 1.163515e+01</span></span>
<span><span class="co">#&gt; 26 ENSMUSG00000067879           Vxn 9.911623e+00</span></span>
<span><span class="co">#&gt; 27 ENSMUSG00000099827       Gm29520 0.000000e+00</span></span>
<span><span class="co">#&gt; 28 ENSMUSG00000025912         Mybl1 2.439775e-01</span></span>
<span><span class="co">#&gt; 29 ENSMUSG00000045210        Vcpip1 3.177732e+01</span></span>
<span><span class="co">#&gt; 30 ENSMUSG00000097893 1700034P13Rik 5.257291e-01</span></span>
<span><span class="co">#&gt; 31 ENSMUSG00000025915          Sgk3 1.850455e+01</span></span>
<span><span class="co">#&gt; 32 ENSMUSG00000046101        Mcmdc2 6.555578e-01</span></span>
<span><span class="co">#&gt; 33 ENSMUSG00000098234         Snhg6 6.245488e+00</span></span>
<span><span class="co">#&gt;  [ reached 'max' / getOption("max.print") -- omitted 52359 rows ]</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="counting-cells-grouped-by-dataset-and-gene">Counting cells grouped by dataset and gene<a class="anchor" aria-label="anchor" href="#counting-cells-grouped-by-dataset-and-gene"></a>
</h2>
<p>The goal of this example is to count the number of cells with nonzero reads, grouped by gene and Census <code>dataset_id</code>. The result is a data frame with dataset, gene, and the number of cells with nonzero reads for that dataset and gene.</p>
<p>For this multi-factor aggregation, we’ll take advantage of <code>dplyr</code> routines instead of the lower-level vector indexer shown above. And for presentation purposes, we’ll limit the query to four genes, but this can be expanded to all genes easily.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://dplyr.tidyverse.org" class="external-link">"dplyr"</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">query</span> <span class="op">&lt;-</span> <span class="va">census</span><span class="op">$</span><span class="fu">get</span><span class="op">(</span><span class="st">"census_data"</span><span class="op">)</span><span class="op">$</span><span class="fu">get</span><span class="op">(</span><span class="st">"mus_musculus"</span><span class="op">)</span><span class="op">$</span><span class="fu">axis_query</span><span class="op">(</span></span>
<span>  measurement_name <span class="op">=</span> <span class="st">"RNA"</span>,</span>
<span>  obs_query <span class="op">=</span> <span class="va"><a href="https://rdrr.io/pkg/tiledbsoma/man/SOMAAxisQuery.html" class="external-link">SOMAAxisQuery</a></span><span class="op">$</span><span class="fu">new</span><span class="op">(</span>value_filter <span class="op">=</span> <span class="st">"tissue=='brain'"</span><span class="op">)</span>,</span>
<span>  var_query <span class="op">=</span> <span class="va"><a href="https://rdrr.io/pkg/tiledbsoma/man/SOMAAxisQuery.html" class="external-link">SOMAAxisQuery</a></span><span class="op">$</span><span class="fu">new</span><span class="op">(</span>value_filter <span class="op">=</span> <span class="st">"feature_name %in% c('Malat1', 'Ptprd', 'Dlg2', 'Pcdh9')"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">obs_tbl</span> <span class="op">&lt;-</span> <span class="va">query</span><span class="op">$</span><span class="fu">obs</span><span class="op">(</span>column_names<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"soma_joinid"</span>, <span class="st">"dataset_id"</span><span class="op">)</span><span class="op">)</span><span class="op">$</span><span class="fu">concat</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">obs_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>  <span class="co"># materialize soma_joinid as character to avoid overflowing R 32-bit integer</span></span>
<span>  cell_id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="va">obs_tbl</span><span class="op">$</span><span class="va">soma_joinid</span><span class="op">)</span>,</span>
<span>  dataset_id <span class="op">=</span> <span class="va">obs_tbl</span><span class="op">$</span><span class="va">dataset_id</span><span class="op">$</span><span class="fu">as_vector</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">var_tbl</span> <span class="op">&lt;-</span> <span class="va">query</span><span class="op">$</span><span class="fu">var</span><span class="op">(</span>column_names<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"soma_joinid"</span>, <span class="st">"feature_name"</span><span class="op">)</span><span class="op">)</span><span class="op">$</span><span class="fu">concat</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">var_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>  gene_id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="va">var_tbl</span><span class="op">$</span><span class="va">soma_joinid</span><span class="op">)</span>,</span>
<span>  feature_name <span class="op">=</span> <span class="va">var_tbl</span><span class="op">$</span><span class="va">feature_name</span><span class="op">$</span><span class="fu">as_vector</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># accumulator for # cells by dataset &amp; gene</span></span>
<span><span class="va">n_cells_grouped</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>  <span class="st">"dataset_id"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">character</a></span><span class="op">(</span><span class="fl">0</span><span class="op">)</span>,</span>
<span>  <span class="st">"gene_id"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">character</a></span><span class="op">(</span><span class="fl">0</span><span class="op">)</span>,</span>
<span>  <span class="st">"n_cells"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">numeric</a></span><span class="op">(</span><span class="fl">0</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># iterate through in-memory shards of query results</span></span>
<span><span class="va">tables</span> <span class="op">&lt;-</span> <span class="va">query</span><span class="op">$</span><span class="fu">X</span><span class="op">(</span><span class="st">"raw"</span><span class="op">)</span><span class="op">$</span><span class="fu">tables</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="kw">while</span> <span class="op">(</span><span class="op">!</span><span class="va">tables</span><span class="op">$</span><span class="fu">read_complete</span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">table_part</span> <span class="op">&lt;-</span> <span class="va">tables</span><span class="op">$</span><span class="fu">read_next</span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span>  <span class="co"># prepare a (dataset,gene,1) tuple for each entry in table_part</span></span>
<span>  <span class="va">n_cells_part</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>    <span class="st">"cell_id"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="va">table_part</span><span class="op">$</span><span class="va">soma_dim_0</span><span class="op">)</span>,</span>
<span>    <span class="st">"gene_id"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="va">table_part</span><span class="op">$</span><span class="va">soma_dim_1</span><span class="op">)</span>,</span>
<span>    <span class="st">"n_cells"</span> <span class="op">=</span> <span class="fl">1</span></span>
<span>  <span class="op">)</span></span>
<span>  <span class="va">n_cells_part</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html" class="external-link">left_join</a></span><span class="op">(</span><span class="va">n_cells_part</span>, <span class="va">obs_df</span>, by <span class="op">=</span> <span class="st">"cell_id"</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/stopifnot.html" class="external-link">stopifnot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NULL.html" class="external-link">is.null</a></span><span class="op">(</span><span class="va">n_cells_part</span><span class="op">$</span><span class="va">dataset_id</span><span class="op">)</span><span class="op">)</span> <span class="op">==</span> <span class="fl">0</span><span class="op">)</span></span>
<span></span>
<span>  <span class="co"># fold those into n_cells_grouped</span></span>
<span>  <span class="va">n_cells_grouped</span> <span class="op">&lt;-</span> <span class="va">n_cells_part</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>                     <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="op">-</span><span class="va">cell_id</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>                     <span class="fu"><a href="https://dplyr.tidyverse.org/reference/bind_rows.html" class="external-link">bind_rows</a></span><span class="op">(</span><span class="va">n_cells_grouped</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>                      <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">dataset_id</span>, <span class="va">gene_id</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>                      <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarise</a></span><span class="op">(</span>n_cells <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">n_cells</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>                      <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">ungroup</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="co"># add gene names for display</span></span>
<span><span class="va">n_cells_grouped</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html" class="external-link">left_join</a></span><span class="op">(</span><span class="va">n_cells_grouped</span>, <span class="va">var_df</span>, by <span class="op">=</span> <span class="st">"gene_id"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/stopifnot.html" class="external-link">stopifnot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NULL.html" class="external-link">is.null</a></span><span class="op">(</span><span class="va">n_cells_grouped</span><span class="op">$</span><span class="va">feature_name</span><span class="op">)</span><span class="op">)</span> <span class="op">==</span> <span class="fl">0</span><span class="op">)</span></span>
<span><span class="va">n_cells_grouped</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"dataset_id"</span>, <span class="st">"feature_name"</span>, <span class="st">"n_cells"</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 17 × 3</span></span></span>
<span><span class="co">#&gt;    dataset_id                           feature_name n_cells</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                                <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>          <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> 3bbb6cf9-72b9-41be-b568-656de6eb18b5 Dlg2           <span style="text-decoration: underline;">79</span>513</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> 3bbb6cf9-72b9-41be-b568-656de6eb18b5 Pcdh9          <span style="text-decoration: underline;">79</span>476</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> 3bbb6cf9-72b9-41be-b568-656de6eb18b5 Malat1         <span style="text-decoration: underline;">79</span>667</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> 3bbb6cf9-72b9-41be-b568-656de6eb18b5 Ptprd          <span style="text-decoration: underline;">79</span>578</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> 58b01044-c5e5-4b0f-8a2d-6ebf951e01ff Dlg2              81</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> 58b01044-c5e5-4b0f-8a2d-6ebf951e01ff Pcdh9            125</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> 58b01044-c5e5-4b0f-8a2d-6ebf951e01ff Malat1         <span style="text-decoration: underline;">12</span>622</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> 58b01044-c5e5-4b0f-8a2d-6ebf951e01ff Ptprd            474</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> 66ff82b4-9380-469c-bc4b-cfa08eacd325 Dlg2             856</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> 66ff82b4-9380-469c-bc4b-cfa08eacd325 Pcdh9           <span style="text-decoration: underline;">2</span>910</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">11</span> 66ff82b4-9380-469c-bc4b-cfa08eacd325 Malat1          <span style="text-decoration: underline;">7</span>102</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">12</span> 98e5ea9f-16d6-47ec-a529-686e76515e39 Dlg2             908</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">13</span> 98e5ea9f-16d6-47ec-a529-686e76515e39 Pcdh9           <span style="text-decoration: underline;">3</span>027</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">14</span> 98e5ea9f-16d6-47ec-a529-686e76515e39 Malat1         <span style="text-decoration: underline;">20</span>094</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">15</span> c08f8441-4a10-4748-872a-e70c0bcccdba Dlg2              52</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">16</span> c08f8441-4a10-4748-872a-e70c0bcccdba Pcdh9            117</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">17</span> c08f8441-4a10-4748-872a-e70c0bcccdba Malat1         <span style="text-decoration: underline;">12</span>992</span></span></code></pre></div>
<p>Don’t forget to close the census.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">census</span><span class="op">$</span><span class="fu">close</span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Chan Zuckerberg Initiative Foundation.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
