<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="cellxgene.census">
<title>Normalizing full-length gene sequencing data • cellxgene.census</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Normalizing full-length gene sequencing data">
<meta property="og:description" content="cellxgene.census">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><script defer data-domain="chanzuckerberg.github.io/cellxgene-census" src="https://plausible.io/js/script.js"></script>
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">cellxgene.census</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.6.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="external-link nav-link" href="https://chanzuckerberg.github.io/cellxgene-census/">Main Census Site</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <h6 class="dropdown-header" data-toc-skip>Explore and analyze the data</h6>
    <a class="dropdown-item" href="../articles/comp_bio_census_info.html">Learning about the CZ CELLxGENE Census</a>
    <a class="dropdown-item" href="../articles/comp_bio_summarize_axis_query.html">Summarizing cell and gene metadata</a>
    <a class="dropdown-item" href="../articles/comp_bio_data_integration.html">Integrating multi-dataset slices of data with Seurat</a>
    <a class="dropdown-item" href="../articles/comp_bio_normalizing_full_gene_sequencing.html">Normalizing full-length gene sequencing data</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>cellxgene.census capabilities</h6>
    <a class="dropdown-item" href="../articles/census_query_extract.html">Querying and fetching the single-cell data and cell/gene metadata</a>
    <a class="dropdown-item" href="../articles/census_datasets.html">Census Datasets example</a>
    <a class="dropdown-item" href="../articles/census_dataset_presence.html">Genes measured in each cell (dataset presence matrix)</a>
    <a class="dropdown-item" href="../articles/census_compute_over_X.html">Computing on X using online (incremental) algorithms</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/chanzuckerberg/cellxgene-census/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Normalizing full-length gene sequencing data</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/chanzuckerberg/cellxgene-census/tree/main/api/r/cellxgene.census/vignettes_/comp_bio_normalizing_full_gene_sequencing.Rmd" class="external-link"><code>vignettes/comp_bio_normalizing_full_gene_sequencing.Rmd</code></a></small>
      <div class="d-none name"><code>comp_bio_normalizing_full_gene_sequencing.Rmd</code></div>
    </div>

    
    
<!--
THIS VIGNETTE IS BASED ON:
https://github.com/chanzuckerberg/cellxgene-census/blob/main/api/python/notebooks/analysis_demo/comp_bio_normalizing_full_gene_sequencing.ipynb
-->
<p>This tutorial shows you how to fetch full-length gene sequencing data from the Census and normalize it to account for gene length.</p>
<p><strong>Contents</strong></p>
<ol style="list-style-type: decimal">
<li>Opening the census</li>
<li>Fetching example full-length sequencing data (Smart-Seq2)</li>
<li>Normalizing expression to account for gene length</li>
<li>Validation through clustering exploration</li>
</ol>
<div class="section level2">
<h2 id="opening-the-census">Opening the census<a class="anchor" aria-label="anchor" href="#opening-the-census"></a>
</h2>
<p>First we open the Census:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://satijalab.org/seurat" class="external-link">"Seurat"</a></span><span class="op">)</span></span>
<span><span class="va">census</span> <span class="op">&lt;-</span> <span class="fu">cellxgene.census</span><span class="fu">::</span><span class="fu"><a href="../reference/open_soma.html">open_soma</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>You can learn more about all of the <code>cellxgene.census</code> methods by accessing their corresponding documentation, for example <code><a href="../reference/open_soma.html">?cellxgene.census::open_soma</a></code>.</p>
</div>
<div class="section level2">
<h2 id="fetching-full-length-example-sequencing-data-smart-seq">Fetching full-length example sequencing data (Smart-Seq)<a class="anchor" aria-label="anchor" href="#fetching-full-length-example-sequencing-data-smart-seq"></a>
</h2>
<p>Let’s get some example data, in this case we’ll fetch all cells from a relatively small dataset derived from the Smart-Seq2 technology which performs full-length gene sequencing:</p>
<ul>
<li>Collection: <a href="https://cellxgene.cziscience.com/collections/0b9d8a04-bb9d-44da-aa27-705bb65b54eb" class="external-link">Tabula Muris Senis</a>
</li>
<li>Dataset: <a href="https://cellxgene.cziscience.com/e/524179b0-b406-4723-9c46-293ffa77ca81.cxg/" class="external-link">Liver - A single-cell transcriptomic atlas characterizes ageing tissues in the mouse - Smart-seq2</a>
</li>
</ul>
<p>Let’s first find this dataset’s id by using the dataset table of the Census.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">liver_dataset</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span></span>
<span>  <span class="va">census</span><span class="op">$</span><span class="fu">get</span><span class="op">(</span><span class="st">"census_info"</span><span class="op">)</span><span class="op">$</span><span class="fu">get</span><span class="op">(</span><span class="st">"datasets"</span><span class="op">)</span></span>
<span>        <span class="op">$</span><span class="fu">read</span><span class="op">(</span>value_filter<span class="op">=</span><span class="st">"dataset_title == 'Liver - A single-cell transcriptomic atlas characterizes ageing tissues in the mouse - Smart-seq2'"</span><span class="op">)</span></span>
<span>        <span class="op">$</span><span class="fu">concat</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">liver_dataset</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 1 × 8</span></span></span>
<span><span class="co">#&gt;   soma_joinid collection_id      collection_name collection_doi dataset_id dataset_title</span></span>
<span><span class="co">#&gt;         <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>              <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>           <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>          <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>        </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span>         525 0b9d8a04-bb9d-44d… Tabula Muris S… 10.1038/s4158… 4546e757-… Liver - A si…</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 2 more variables: dataset_h5ad_path &lt;chr&gt;, dataset_total_cell_count &lt;int&gt;</span></span></span></code></pre></div>
<p>Now we can use this id to fetch the data.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">liver_dataset_id</span> <span class="op">&lt;-</span> <span class="va">liver_dataset</span><span class="op">[</span><span class="fl">1</span>,<span class="st">"dataset_id"</span><span class="op">]</span></span>
<span><span class="va">liver_seurat</span> <span class="op">&lt;-</span> <span class="fu">cellxgene.census</span><span class="fu">::</span><span class="fu"><a href="../reference/get_seurat.html">get_seurat</a></span><span class="op">(</span></span>
<span>  <span class="va">census</span>,</span>
<span>  organism <span class="op">=</span> <span class="st">"Mus musculus"</span>,</span>
<span>  obs_value_filter <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"dataset_id == '"</span>, <span class="va">liver_dataset_id</span>, <span class="st">"'"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Let’s make sure this data only contains Smart-Seq2 cells.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">liver_seurat</span><span class="op">$</span><span class="va">assay</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Smart-seq2 </span></span>
<span><span class="co">#&gt;       2859</span></span></code></pre></div>
<p>Great! As you can see this a small dataset only containing <strong>2,859</strong> cells. Now let’s proceed to normalize by gene lengths.</p>
</div>
<div class="section level2">
<h2 id="normalizing-expression-to-account-for-gene-length">Normalizing expression to account for gene length<a class="anchor" aria-label="anchor" href="#normalizing-expression-to-account-for-gene-length"></a>
</h2>
<p>By default <code>cellxgene_census::get_seurat()</code> fetches all genes in the Census. So let’s first identify the genes that were measured in this dataset and subset the Seurat obect to only include those.</p>
<p>To this goal we can use the “Dataset Presence Matrix” in <code>census$get("census_data")$get("mus_musculus")$ms$get("RNA")$get("feature_dataset_presence_matrix")</code>. This is a boolean matrix <code>N x M</code> where <code>N</code> is the number of datasets, <code>M</code> is the number of genes in the Census, and a 1 entry indicates that a gene was measured in a dataset. (Note that Seurat objects have the transposed layout <code>M x N</code>.)</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">liver_seurat</span></span>
<span><span class="co">#&gt; An object of class Seurat </span></span>
<span><span class="co">#&gt; 52392 features across 2859 samples within 1 assay </span></span>
<span><span class="co">#&gt; Active assay: RNA (52392 features, 0 variable features)</span></span></code></pre></div>
<p>Let’s get the genes measured in this dataset.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">liver_dataset_joinid</span> <span class="op">&lt;-</span> <span class="va">liver_dataset</span><span class="op">$</span><span class="va">soma_joinid</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span></span>
<span><span class="va">presence_matrix</span> <span class="op">&lt;-</span> <span class="fu">cellxgene.census</span><span class="fu">::</span><span class="fu"><a href="../reference/get_presence_matrix.html">get_presence_matrix</a></span><span class="op">(</span><span class="va">census</span>, <span class="st">"Mus musculus"</span>, <span class="st">"RNA"</span><span class="op">)</span></span>
<span><span class="va">presence_matrix</span> <span class="op">&lt;-</span> <span class="va">presence_matrix</span><span class="op">$</span><span class="fu">take</span><span class="op">(</span><span class="va">liver_dataset_joinid</span><span class="op">)</span></span>
<span><span class="va">gene_presence</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/vector.html" class="external-link">as.vector</a></span><span class="op">(</span><span class="va">presence_matrix</span><span class="op">$</span><span class="fu">get_one_based_matrix</span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">liver_seurat</span> <span class="op">&lt;-</span> <span class="va">liver_seurat</span><span class="op">[</span><span class="va">gene_presence</span>, <span class="op">]</span></span>
<span><span class="va">liver_seurat</span></span>
<span><span class="co">#&gt; An object of class Seurat </span></span>
<span><span class="co">#&gt; 17992 features across 2859 samples within 1 assay </span></span>
<span><span class="co">#&gt; Active assay: RNA (17992 features, 0 variable features)</span></span></code></pre></div>
<p>We can see that out of all genes in the Census <strong>17,992</strong> were measured in this dataset.</p>
<p>Now let’s normalize these genes by gene length. We can easily do this because the Census has gene lengths included in the gene metadata under <code>feature_length</code>.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/AssayData.html" class="external-link">GetAssayData</a></span><span class="op">(</span><span class="va">liver_seurat</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span>, slot <span class="op">=</span> <span class="st">"data"</span><span class="op">)</span></span>
<span><span class="co">#&gt; 5 x 5 sparse Matrix of class "dgCMatrix"</span></span>
<span><span class="co">#&gt;                    cell3530079 cell3530080 cell3530081 cell3530082 cell3530083</span></span>
<span><span class="co">#&gt; ENSMUSG00000025900           .           .           .           .           .</span></span>
<span><span class="co">#&gt; ENSMUSG00000025902           .           .           .           .        2250</span></span>
<span><span class="co">#&gt; ENSMUSG00000033845           .         559        1969           .           .</span></span>
<span><span class="co">#&gt; ENSMUSG00000025903           .           .           .           .           .</span></span>
<span><span class="co">#&gt; ENSMUSG00000033813           .           .         828           1          54</span></span></code></pre></div>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gene_lengths</span> <span class="op">&lt;-</span> <span class="va">liver_seurat</span><span class="op">$</span><span class="va">RNA</span><span class="op">@</span><span class="va">meta.features</span><span class="op">$</span><span class="va">feature_length</span></span>
<span><span class="va">liver_seurat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/AssayData.html" class="external-link">SetAssayData</a></span><span class="op">(</span></span>
<span>  <span class="va">liver_seurat</span>, </span>
<span>  new.data <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sweep.html" class="external-link">sweep</a></span><span class="op">(</span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/AssayData.html" class="external-link">GetAssayData</a></span><span class="op">(</span><span class="va">liver_seurat</span>, slot <span class="op">=</span> <span class="st">"data"</span><span class="op">)</span>, <span class="fl">1</span>, <span class="va">gene_lengths</span>, <span class="st">"/"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/AssayData.html" class="external-link">GetAssayData</a></span><span class="op">(</span><span class="va">liver_seurat</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span>, slot <span class="op">=</span> <span class="st">"data"</span><span class="op">)</span></span>
<span><span class="co">#&gt; 5 x 5 sparse Matrix of class "dgCMatrix"</span></span>
<span><span class="co">#&gt;                    cell3530079 cell3530080 cell3530081  cell3530082 cell3530083</span></span>
<span><span class="co">#&gt; ENSMUSG00000025900           .  .            .         .             .         </span></span>
<span><span class="co">#&gt; ENSMUSG00000025902           .  .            .         .             0.47150042</span></span>
<span><span class="co">#&gt; ENSMUSG00000033845           .  0.06586544   0.2320019 .             .         </span></span>
<span><span class="co">#&gt; ENSMUSG00000025903           .  .            .         .             .         </span></span>
<span><span class="co">#&gt; ENSMUSG00000033813           .  .            0.2744448 0.0003314551  0.01789857</span></span></code></pre></div>
<p>All done! You can now see that we have real numbers instead of integers.</p>
</div>
<div class="section level2">
<h2 id="validation-through-clustering-exploration">Validation through clustering exploration<a class="anchor" aria-label="anchor" href="#validation-through-clustering-exploration"></a>
</h2>
<p>Let’s perform some basic clustering analysis to see if cell types cluster as expected using the normalized counts.</p>
<p>First we do some basic filtering of cells and genes.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cells_per_gene</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/AssayData.html" class="external-link">GetAssayData</a></span><span class="op">(</span><span class="va">liver_seurat</span>, slot <span class="op">=</span> <span class="st">"counts"</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span></span>
<span><span class="va">genes_per_cell</span> <span class="op">&lt;-</span> <span class="fu">Matrix</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/colSums.html" class="external-link">colSums</a></span><span class="op">(</span><span class="va">liver_seurat</span><span class="op">$</span><span class="va">RNA</span><span class="op">@</span><span class="va">counts</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span></span>
<span><span class="va">liver_seurat</span> <span class="op">&lt;-</span> <span class="va">liver_seurat</span><span class="op">[</span><span class="va">cells_per_gene</span> <span class="op">&gt;=</span> <span class="fl">5</span>, <span class="va">genes_per_cell</span> <span class="op">&gt;=</span> <span class="fl">500</span><span class="op">]</span></span></code></pre></div>
<p>Then we normalize to account for sequencing depth and transform data to log scale.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">liver_seurat</span> <span class="op">&lt;-</span> <span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://satijalab.org/seurat/reference/NormalizeData.html" class="external-link">NormalizeData</a></span><span class="op">(</span></span>
<span>  <span class="va">liver_seurat</span>,</span>
<span>  normalization.method <span class="op">=</span> <span class="st">"LogNormalize"</span>,</span>
<span>  scale.factor <span class="op">=</span> <span class="fl">10000</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Then we subset to highly variable genes.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">liver_seurat</span> <span class="op">&lt;-</span> <span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://satijalab.org/seurat/reference/FindVariableFeatures.html" class="external-link">FindVariableFeatures</a></span><span class="op">(</span></span>
<span>  <span class="va">liver_seurat</span>,</span>
<span>  selection.method <span class="op">=</span> <span class="st">"vst"</span>,</span>
<span>  nfeatures <span class="op">=</span> <span class="fl">1000</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>And finally we scale values across the gene axis.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">all.genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">liver_seurat</span><span class="op">)</span></span>
<span><span class="va">liver_seurat</span> <span class="op">&lt;-</span> <span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://satijalab.org/seurat/reference/ScaleData.html" class="external-link">ScaleData</a></span><span class="op">(</span><span class="va">liver_seurat</span>, features <span class="op">=</span> <span class="va">all.genes</span><span class="op">)</span></span></code></pre></div>
<p>Now we can proceed to do clustering analysis.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">liver_seurat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunPCA.html" class="external-link">RunPCA</a></span><span class="op">(</span></span>
<span>  <span class="va">liver_seurat</span>,</span>
<span>  features <span class="op">=</span> <span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/VariableFeatures.html" class="external-link">VariableFeatures</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">liver_seurat</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">liver_seurat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindNeighbors.html" class="external-link">FindNeighbors</a></span><span class="op">(</span><span class="va">liver_seurat</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">40</span><span class="op">)</span></span>
<span><span class="va">liver_seurat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunUMAP.html" class="external-link">RunUMAP</a></span><span class="op">(</span><span class="va">liver_seurat</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">40</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">liver_seurat</span>, reduction <span class="op">=</span> <span class="st">"umap"</span>, group.by <span class="op">=</span> <span class="st">"cell_type"</span><span class="op">)</span></span></code></pre></div>
<p><img src="comp_bio_normalizing_full_gene_sequencing_files/figure-html/unnamed-chunk-15-1.png" width="672"></p>
<p>With a few exceptions we can see that all cells from the same cell type cluster near each other which serves as a sanity check for the gene-length normalization that we applied.</p>
<p>Don’t forget to close the census.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">census</span><span class="op">$</span><span class="fu">close</span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Chan Zuckerberg Initiative.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
