<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Export Census slices to SingleCellExperiment — get_single_cell_experiment • cellxgene.census</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Export Census slices to SingleCellExperiment — get_single_cell_experiment"><meta name="description" content="Convenience wrapper around SOMAExperimentAxisQuery, to build and execute a
query, and return it as a SingleCellExperiment object."><meta property="og:description" content="Convenience wrapper around SOMAExperimentAxisQuery, to build and execute a
query, and return it as a SingleCellExperiment object."><script defer data-domain="chanzuckerberg.github.io/cellxgene-census" src="https://plausible.io/js/script.js"></script></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">cellxgene.census</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.15.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="external-link nav-link" href="https://chanzuckerberg.github.io/cellxgene-census/">Main Census Site</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Use Census embeddings</h6></li>
    <li><a class="dropdown-item" href="../articles/census_access_maintained_embeddings.html">Access CELLxGENE collaboration embeddings (scVI, Geneformer)</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Explore and analyze the data</h6></li>
    <li><a class="dropdown-item" href="../articles/comp_bio_census_info.html">Learning about the CZ CELLxGENE Census</a></li>
    <li><a class="dropdown-item" href="../articles/comp_bio_summarize_axis_query.html">Summarizing cell and gene metadata</a></li>
    <li><a class="dropdown-item" href="../articles/comp_bio_data_integration.html">Integrating multi-dataset slices of data with Seurat</a></li>
    <li><a class="dropdown-item" href="../articles/comp_bio_normalizing_full_gene_sequencing.html">Normalizing full-length gene sequencing data</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>cellxgene.census capabilities</h6></li>
    <li><a class="dropdown-item" href="../articles/census_query_extract.html">Querying and fetching the single-cell data and cell/gene metadata</a></li>
    <li><a class="dropdown-item" href="../articles/census_citation_generation.html">Generating citations for Census slices</a></li>
    <li><a class="dropdown-item" href="../articles/census_datasets.html">Census Datasets example</a></li>
    <li><a class="dropdown-item" href="../articles/census_dataset_presence.html">Genes measured in each cell (dataset presence matrix)</a></li>
    <li><a class="dropdown-item" href="../articles/census_compute_over_X.html">Computing on X using online (incremental) algorithms</a></li>
    <li><a class="dropdown-item" href="../articles/census_access_maintained_embeddings.html">Access CELLxGENE collaboration embeddings (scVI, Geneformer)</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/chanzuckerberg/cellxgene-census/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Export Census slices to <code>SingleCellExperiment</code></h1>
      <small class="dont-index">Source: <a href="https://github.com/chanzuckerberg/cellxgene-census/tree/main/api/r/cellxgene.census/R/get_helpers.R" class="external-link"><code>R/get_helpers.R</code></a></small>
      <div class="d-none name"><code>get_single_cell_experiment.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Convenience wrapper around <code>SOMAExperimentAxisQuery</code>, to build and execute a
query, and return it as a <code>SingleCellExperiment</code> object.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">get_single_cell_experiment</span><span class="op">(</span></span>
<span>  <span class="va">census</span>,</span>
<span>  <span class="va">organism</span>,</span>
<span>  measurement_name <span class="op">=</span> <span class="st">"RNA"</span>,</span>
<span>  X_layers <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>counts <span class="op">=</span> <span class="st">"raw"</span><span class="op">)</span>,</span>
<span>  obs_value_filter <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  obs_coords <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  obs_column_names <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  obsm_layers <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  var_value_filter <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  var_coords <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  var_column_names <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  var_index <span class="op">=</span> <span class="st">"feature_id"</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-census">census<a class="anchor" aria-label="anchor" href="#arg-census"></a></dt>
<dd><p>The census object, usually returned by <code><a href="open_soma.html">cellxgene.census::open_soma()</a></code>.</p></dd>


<dt id="arg-organism">organism<a class="anchor" aria-label="anchor" href="#arg-organism"></a></dt>
<dd><p>The organism to query, usually one of <code>Homo sapiens</code> or <code>Mus musculus</code></p></dd>


<dt id="arg-measurement-name">measurement_name<a class="anchor" aria-label="anchor" href="#arg-measurement-name"></a></dt>
<dd><p>The measurement object to query. Defaults to <code>RNA</code>.</p></dd>


<dt id="arg-x-layers">X_layers<a class="anchor" aria-label="anchor" href="#arg-x-layers"></a></dt>
<dd><p>A character vector of X layers to add as assays in
the main experiment; may optionally be named to set the
name of the resulting assay (eg. ‘X_layers = c(counts =
"raw")’ will load in X layer “‘raw’” as assay “‘counts’”);
by default, loads in all X layers</p></dd>


<dt id="arg-obs-value-filter">obs_value_filter<a class="anchor" aria-label="anchor" href="#arg-obs-value-filter"></a></dt>
<dd><p>A SOMA <code>value_filter</code> across columns in the <code>obs</code> dataframe, expressed as string.</p></dd>


<dt id="arg-obs-coords">obs_coords<a class="anchor" aria-label="anchor" href="#arg-obs-coords"></a></dt>
<dd><p>A set of coordinates on the obs dataframe index, expressed in any type or format supported by SOMADataFrame's read() method.</p></dd>


<dt id="arg-obs-column-names">obs_column_names<a class="anchor" aria-label="anchor" href="#arg-obs-column-names"></a></dt>
<dd><p>Columns to fetch for the <code>obs</code> data frame.</p></dd>


<dt id="arg-obsm-layers">obsm_layers<a class="anchor" aria-label="anchor" href="#arg-obsm-layers"></a></dt>
<dd><p>Names of arrays in obsm to add as the cell embeddings; pass FALSE to suppress loading in any dimensional reductions.</p></dd>


<dt id="arg-var-value-filter">var_value_filter<a class="anchor" aria-label="anchor" href="#arg-var-value-filter"></a></dt>
<dd><p>Same as <code>obs_value_filter</code> but for <code>var</code>.</p></dd>


<dt id="arg-var-coords">var_coords<a class="anchor" aria-label="anchor" href="#arg-var-coords"></a></dt>
<dd><p>Same as <code>obs_coords</code> but for <code>var</code>.</p></dd>


<dt id="arg-var-column-names">var_column_names<a class="anchor" aria-label="anchor" href="#arg-var-column-names"></a></dt>
<dd><p>Columns to fetch for the <code>var</code> data frame.</p></dd>


<dt id="arg-var-index">var_index<a class="anchor" aria-label="anchor" href="#arg-var-index"></a></dt>
<dd><p>Name of column in ‘var’ to add as feature names.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A <code>SingleCellExperiment</code> object containing the sensus slice.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span> <span class="co"># \dontrun{</span></span></span>
<span class="r-in"><span><span class="va">census</span> <span class="op">&lt;-</span> <span class="fu"><a href="open_soma.html">open_soma</a></span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">sce_obj</span> <span class="op">&lt;-</span> <span class="fu">get_single_cell_experiment</span><span class="op">(</span></span></span>
<span class="r-in"><span>  <span class="va">census</span>,</span></span>
<span class="r-in"><span>  organism <span class="op">=</span> <span class="st">"Homo sapiens"</span>,</span></span>
<span class="r-in"><span>  obs_value_filter <span class="op">=</span> <span class="st">"cell_type == 'leptomeningeal cell'"</span>,</span></span>
<span class="r-in"><span>  var_value_filter <span class="op">=</span> <span class="st">"feature_id %in% c('ENSG00000107317', 'ENSG00000106034')"</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">sce_obj</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">census</span><span class="op">$</span><span class="fu">close</span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span> <span class="co"># }</span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Chan Zuckerberg Initiative Foundation.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

    </footer></div>





  </body></html>

