{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Genes measured in each cell (dataset presence matrix)\n",
    "\n",
    "The Census is a compilation of cells from multiple datasets that may differ by the sets of genes they measure. This notebook describes the way to identify the genes measured per dataset.\n",
    "\n",
    "The presence matrix is a sparse boolean array, indicating which features (var) were present in each dataset.  The array has dimensions [n_datasets, n_var], and is stored in the SOMA Measurement `varp` collection. The first dimension is indexed by the `soma_joinid` in the `census_datasets` dataframe. The second is indexed by the `soma_joinid` in the `var` dataframe of the measurement.\n",
    "\n",
    "As a reminder the `obs` data frame has a column `dataset_id` that can be used to link any cell in the Census to the presence matrix.\n",
    "\n",
    "**Contents** \n",
    "\n",
    "1. Opening the Census.\n",
    "2. Fetching the IDs of the Census datasets.\n",
    "3. Fetching the dataset presence matrix.\n",
    "4. Identifying genes measured in a specific dataset.\n",
    "5. Identifying datasets that measured specific genes.\n",
    "6. Identifying all genes measured in a dataset.\n",
    "\n",
    "\n",
    "## Opening the Census\n",
    "\n",
    "The `cellxgene_census` python package contains a convenient API to open the latest version of the Census."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-05-17T15:36:27.867928Z",
     "iopub.status.busy": "2023-05-17T15:36:27.867789Z",
     "iopub.status.idle": "2023-05-17T15:36:29.666773Z",
     "shell.execute_reply": "2023-05-17T15:36:29.665064Z"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "The \"stable\" release is currently 2023-05-15. Specify 'census_version=\"2023-05-15\"' in future calls to open_soma() to ensure data consistency.\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "from scipy import sparse\n",
    "import cellxgene_census\n",
    "\n",
    "census = cellxgene_census.open_soma()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Fetching the IDs of the Census datasets\n",
    "\n",
    "Let's grab a table of all the datasets included in the Census and use this table in combination with the presence matrix below."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-05-17T15:36:29.669589Z",
     "iopub.status.busy": "2023-05-17T15:36:29.669326Z",
     "iopub.status.idle": "2023-05-17T15:36:30.497682Z",
     "shell.execute_reply": "2023-05-17T15:36:30.497286Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>soma_joinid</th>\n",
       "      <th>collection_id</th>\n",
       "      <th>collection_name</th>\n",
       "      <th>collection_doi</th>\n",
       "      <th>dataset_id</th>\n",
       "      <th>dataset_title</th>\n",
       "      <th>dataset_h5ad_path</th>\n",
       "      <th>dataset_total_cell_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>6b701826-37bb-4356-9792-ff41fc4c3161</td>\n",
       "      <td>Abdominal White Adipose Tissue</td>\n",
       "      <td></td>\n",
       "      <td>9d8e5dca-03a3-457d-b7fb-844c75735c83</td>\n",
       "      <td>22 integrated samples</td>\n",
       "      <td>9d8e5dca-03a3-457d-b7fb-844c75735c83.h5ad</td>\n",
       "      <td>72335</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>4195ab4c-20bd-4cd3-8b3d-65601277e731</td>\n",
       "      <td>A spatially resolved single cell genomic atlas...</td>\n",
       "      <td></td>\n",
       "      <td>a6388a6f-6076-401b-9b30-7d4306a20035</td>\n",
       "      <td>scRNA-seq data - myeloid cells</td>\n",
       "      <td>a6388a6f-6076-401b-9b30-7d4306a20035.h5ad</td>\n",
       "      <td>30789</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>4195ab4c-20bd-4cd3-8b3d-65601277e731</td>\n",
       "      <td>A spatially resolved single cell genomic atlas...</td>\n",
       "      <td></td>\n",
       "      <td>842c6f5d-4a94-4eef-8510-8c792d1124bc</td>\n",
       "      <td>scRNA-seq data - all cells</td>\n",
       "      <td>842c6f5d-4a94-4eef-8510-8c792d1124bc.h5ad</td>\n",
       "      <td>714331</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>4195ab4c-20bd-4cd3-8b3d-65601277e731</td>\n",
       "      <td>A spatially resolved single cell genomic atlas...</td>\n",
       "      <td></td>\n",
       "      <td>74520626-b0ba-4ee9-86b5-714649554def</td>\n",
       "      <td>scRNA-seq data - T cells</td>\n",
       "      <td>74520626-b0ba-4ee9-86b5-714649554def.h5ad</td>\n",
       "      <td>76567</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>4195ab4c-20bd-4cd3-8b3d-65601277e731</td>\n",
       "      <td>A spatially resolved single cell genomic atlas...</td>\n",
       "      <td></td>\n",
       "      <td>396a9124-fb20-4822-bf9c-e93fdf7c999a</td>\n",
       "      <td>scRNA-seq data - B cells</td>\n",
       "      <td>396a9124-fb20-4822-bf9c-e93fdf7c999a.h5ad</td>\n",
       "      <td>12510</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>557</th>\n",
       "      <td>557</td>\n",
       "      <td>180bff9c-c8a5-4539-b13b-ddbc00d643e6</td>\n",
       "      <td>Molecular characterization of selectively vuln...</td>\n",
       "      <td>10.1038/s41593-020-00764-7</td>\n",
       "      <td>f9ad5649-f372-43e1-a3a8-423383e5a8a2</td>\n",
       "      <td>Molecular characterization of selectively vuln...</td>\n",
       "      <td>f9ad5649-f372-43e1-a3a8-423383e5a8a2.h5ad</td>\n",
       "      <td>8168</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>558</th>\n",
       "      <td>558</td>\n",
       "      <td>a72afd53-ab92-4511-88da-252fb0e26b9a</td>\n",
       "      <td>Single-cell atlas of peripheral immune respons...</td>\n",
       "      <td>10.1038/s41591-020-0944-y</td>\n",
       "      <td>456e8b9b-f872-488b-871d-94534090a865</td>\n",
       "      <td>Single-cell atlas of peripheral immune respons...</td>\n",
       "      <td>456e8b9b-f872-488b-871d-94534090a865.h5ad</td>\n",
       "      <td>44721</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>559</th>\n",
       "      <td>559</td>\n",
       "      <td>38833785-fac5-48fd-944a-0f62a4c23ed1</td>\n",
       "      <td>Construction of a human cell landscape at sing...</td>\n",
       "      <td>10.1038/s41586-020-2157-4</td>\n",
       "      <td>2adb1f8a-a6b1-4909-8ee8-484814e2d4bf</td>\n",
       "      <td>Construction of a human cell landscape at sing...</td>\n",
       "      <td>2adb1f8a-a6b1-4909-8ee8-484814e2d4bf.h5ad</td>\n",
       "      <td>598266</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>560</th>\n",
       "      <td>560</td>\n",
       "      <td>5d445965-6f1a-4b68-ba3a-b8f765155d3a</td>\n",
       "      <td>A molecular cell atlas of the human lung from ...</td>\n",
       "      <td>10.1038/s41586-020-2922-4</td>\n",
       "      <td>e04daea4-4412-45b5-989e-76a9be070a89</td>\n",
       "      <td>Krasnow Lab Human Lung Cell Atlas, Smart-seq2</td>\n",
       "      <td>e04daea4-4412-45b5-989e-76a9be070a89.h5ad</td>\n",
       "      <td>9409</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>561</th>\n",
       "      <td>561</td>\n",
       "      <td>5d445965-6f1a-4b68-ba3a-b8f765155d3a</td>\n",
       "      <td>A molecular cell atlas of the human lung from ...</td>\n",
       "      <td>10.1038/s41586-020-2922-4</td>\n",
       "      <td>8c42cfd0-0b0a-46d5-910c-fc833d83c45e</td>\n",
       "      <td>Krasnow Lab Human Lung Cell Atlas, 10X</td>\n",
       "      <td>8c42cfd0-0b0a-46d5-910c-fc833d83c45e.h5ad</td>\n",
       "      <td>65662</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>562 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     soma_joinid                         collection_id   \n",
       "0              0  6b701826-37bb-4356-9792-ff41fc4c3161  \\\n",
       "1              1  4195ab4c-20bd-4cd3-8b3d-65601277e731   \n",
       "2              2  4195ab4c-20bd-4cd3-8b3d-65601277e731   \n",
       "3              3  4195ab4c-20bd-4cd3-8b3d-65601277e731   \n",
       "4              4  4195ab4c-20bd-4cd3-8b3d-65601277e731   \n",
       "..           ...                                   ...   \n",
       "557          557  180bff9c-c8a5-4539-b13b-ddbc00d643e6   \n",
       "558          558  a72afd53-ab92-4511-88da-252fb0e26b9a   \n",
       "559          559  38833785-fac5-48fd-944a-0f62a4c23ed1   \n",
       "560          560  5d445965-6f1a-4b68-ba3a-b8f765155d3a   \n",
       "561          561  5d445965-6f1a-4b68-ba3a-b8f765155d3a   \n",
       "\n",
       "                                       collection_name   \n",
       "0                       Abdominal White Adipose Tissue  \\\n",
       "1    A spatially resolved single cell genomic atlas...   \n",
       "2    A spatially resolved single cell genomic atlas...   \n",
       "3    A spatially resolved single cell genomic atlas...   \n",
       "4    A spatially resolved single cell genomic atlas...   \n",
       "..                                                 ...   \n",
       "557  Molecular characterization of selectively vuln...   \n",
       "558  Single-cell atlas of peripheral immune respons...   \n",
       "559  Construction of a human cell landscape at sing...   \n",
       "560  A molecular cell atlas of the human lung from ...   \n",
       "561  A molecular cell atlas of the human lung from ...   \n",
       "\n",
       "                 collection_doi                            dataset_id   \n",
       "0                                9d8e5dca-03a3-457d-b7fb-844c75735c83  \\\n",
       "1                                a6388a6f-6076-401b-9b30-7d4306a20035   \n",
       "2                                842c6f5d-4a94-4eef-8510-8c792d1124bc   \n",
       "3                                74520626-b0ba-4ee9-86b5-714649554def   \n",
       "4                                396a9124-fb20-4822-bf9c-e93fdf7c999a   \n",
       "..                          ...                                   ...   \n",
       "557  10.1038/s41593-020-00764-7  f9ad5649-f372-43e1-a3a8-423383e5a8a2   \n",
       "558   10.1038/s41591-020-0944-y  456e8b9b-f872-488b-871d-94534090a865   \n",
       "559   10.1038/s41586-020-2157-4  2adb1f8a-a6b1-4909-8ee8-484814e2d4bf   \n",
       "560   10.1038/s41586-020-2922-4  e04daea4-4412-45b5-989e-76a9be070a89   \n",
       "561   10.1038/s41586-020-2922-4  8c42cfd0-0b0a-46d5-910c-fc833d83c45e   \n",
       "\n",
       "                                         dataset_title   \n",
       "0                                22 integrated samples  \\\n",
       "1                       scRNA-seq data - myeloid cells   \n",
       "2                           scRNA-seq data - all cells   \n",
       "3                             scRNA-seq data - T cells   \n",
       "4                             scRNA-seq data - B cells   \n",
       "..                                                 ...   \n",
       "557  Molecular characterization of selectively vuln...   \n",
       "558  Single-cell atlas of peripheral immune respons...   \n",
       "559  Construction of a human cell landscape at sing...   \n",
       "560      Krasnow Lab Human Lung Cell Atlas, Smart-seq2   \n",
       "561             Krasnow Lab Human Lung Cell Atlas, 10X   \n",
       "\n",
       "                             dataset_h5ad_path  dataset_total_cell_count  \n",
       "0    9d8e5dca-03a3-457d-b7fb-844c75735c83.h5ad                     72335  \n",
       "1    a6388a6f-6076-401b-9b30-7d4306a20035.h5ad                     30789  \n",
       "2    842c6f5d-4a94-4eef-8510-8c792d1124bc.h5ad                    714331  \n",
       "3    74520626-b0ba-4ee9-86b5-714649554def.h5ad                     76567  \n",
       "4    396a9124-fb20-4822-bf9c-e93fdf7c999a.h5ad                     12510  \n",
       "..                                         ...                       ...  \n",
       "557  f9ad5649-f372-43e1-a3a8-423383e5a8a2.h5ad                      8168  \n",
       "558  456e8b9b-f872-488b-871d-94534090a865.h5ad                     44721  \n",
       "559  2adb1f8a-a6b1-4909-8ee8-484814e2d4bf.h5ad                    598266  \n",
       "560  e04daea4-4412-45b5-989e-76a9be070a89.h5ad                      9409  \n",
       "561  8c42cfd0-0b0a-46d5-910c-fc833d83c45e.h5ad                     65662  \n",
       "\n",
       "[562 rows x 8 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Grab the experiment containing human data, and the measurement therein with RNA\n",
    "human = census[\"census_data\"][\"homo_sapiens\"]\n",
    "human_rna = human.ms[\"RNA\"]\n",
    "\n",
    "# The census-wide datasets\n",
    "datasets_df = census[\"census_info\"][\"datasets\"].read().concat().to_pandas()\n",
    "\n",
    "datasets_df"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Fetching the dataset presence matrix\n",
    "\n",
    "Now let's fetch the dataset presence matrix. \n",
    "\n",
    "For convenience, read the entire presence matrix (for Homo sapiens) into a SciPy array. There is a convenience API providing this capability, returning the matrix in a `scipy.sparse.array`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-05-17T15:36:30.500034Z",
     "iopub.status.busy": "2023-05-17T15:36:30.499845Z",
     "iopub.status.idle": "2023-05-17T15:36:32.051353Z",
     "shell.execute_reply": "2023-05-17T15:36:32.050525Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<562x60664 sparse matrix of type '<class 'numpy.uint8'>'\n",
       "\twith 14829450 stored elements in Compressed Sparse Row format>"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "presence_matrix = cellxgene_census.get_presence_matrix(census, organism=\"Homo sapiens\", measurement_name=\"RNA\")\n",
    "\n",
    "presence_matrix"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We also need the `var` dataframe, which is read into a Pandas DataFrame for convenient manipulation:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-05-17T15:36:32.053880Z",
     "iopub.status.busy": "2023-05-17T15:36:32.053734Z",
     "iopub.status.idle": "2023-05-17T15:36:32.463825Z",
     "shell.execute_reply": "2023-05-17T15:36:32.463335Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>soma_joinid</th>\n",
       "      <th>feature_id</th>\n",
       "      <th>feature_name</th>\n",
       "      <th>feature_length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000243485</td>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>1021</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000237613</td>\n",
       "      <td>FAM138A</td>\n",
       "      <td>1219</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000186092</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>2618</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>ENSG00000238009</td>\n",
       "      <td>RP11-34P13.7</td>\n",
       "      <td>3726</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>ENSG00000239945</td>\n",
       "      <td>RP11-34P13.8</td>\n",
       "      <td>1319</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60659</th>\n",
       "      <td>60659</td>\n",
       "      <td>ENSG00000288719</td>\n",
       "      <td>RP4-669P10.21</td>\n",
       "      <td>4252</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60660</th>\n",
       "      <td>60660</td>\n",
       "      <td>ENSG00000288720</td>\n",
       "      <td>RP11-852E15.3</td>\n",
       "      <td>7007</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60661</th>\n",
       "      <td>60661</td>\n",
       "      <td>ENSG00000288721</td>\n",
       "      <td>RP5-973N23.5</td>\n",
       "      <td>7765</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60662</th>\n",
       "      <td>60662</td>\n",
       "      <td>ENSG00000288723</td>\n",
       "      <td>RP11-553N16.6</td>\n",
       "      <td>1015</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60663</th>\n",
       "      <td>60663</td>\n",
       "      <td>ENSG00000288724</td>\n",
       "      <td>RP13-546I2.2</td>\n",
       "      <td>625</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>60664 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       soma_joinid       feature_id   feature_name  feature_length\n",
       "0                0  ENSG00000243485    MIR1302-2HG            1021\n",
       "1                1  ENSG00000237613        FAM138A            1219\n",
       "2                2  ENSG00000186092          OR4F5            2618\n",
       "3                3  ENSG00000238009   RP11-34P13.7            3726\n",
       "4                4  ENSG00000239945   RP11-34P13.8            1319\n",
       "...            ...              ...            ...             ...\n",
       "60659        60659  ENSG00000288719  RP4-669P10.21            4252\n",
       "60660        60660  ENSG00000288720  RP11-852E15.3            7007\n",
       "60661        60661  ENSG00000288721   RP5-973N23.5            7765\n",
       "60662        60662  ENSG00000288723  RP11-553N16.6            1015\n",
       "60663        60663  ENSG00000288724   RP13-546I2.2             625\n",
       "\n",
       "[60664 rows x 4 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "var_df = human_rna.var.read().concat().to_pandas()\n",
    "\n",
    "var_df"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Identifying genes measured in a specific dataset.\n",
    "\n",
    "Now that we have the dataset table, the genes metadata table, and the dataset presence matrix, we can check if a gene or set of genes were measured in a specific dataset.\n",
    "\n",
    "**Important:** the presence matrix is indexed by soma_joinid, and is *NOT* positionally indexed.  In other words:\n",
    "\n",
    "* the first dimension of the presence matrix is the dataset's `soma_joinid`, as stored in the `census_datasets` dataframe.\n",
    "* the second dimension of the presence matrix is the feature's `soma_joinid`, as stored in the `var` dataframe.\n",
    "\n",
    "Let's find out if the the gene `\"ENSG00000286096\"` was measured in the dataset with id `\"97a17473-e2b1-4f31-a544-44a60773e2dd\"`.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-05-17T15:36:32.466403Z",
     "iopub.status.busy": "2023-05-17T15:36:32.466250Z",
     "iopub.status.idle": "2023-05-17T15:36:32.473605Z",
     "shell.execute_reply": "2023-05-17T15:36:32.473023Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Feature is present.\n"
     ]
    }
   ],
   "source": [
    "var_joinid = var_df.loc[var_df.feature_id == \"ENSG00000286096\"].soma_joinid\n",
    "dataset_joinid = datasets_df.loc[datasets_df.dataset_id == \"97a17473-e2b1-4f31-a544-44a60773e2dd\"].soma_joinid\n",
    "is_present = presence_matrix[dataset_joinid, var_joinid][0, 0]\n",
    "print(f'Feature is {\"present\" if is_present else \"not present\"}.')"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Identifying datasets that measured specific genes\n",
    "\n",
    "Similarly, we can determine the datasets that measured a specific gene or set of genes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-05-17T15:36:32.475790Z",
     "iopub.status.busy": "2023-05-17T15:36:32.475653Z",
     "iopub.status.idle": "2023-05-17T15:36:32.505636Z",
     "shell.execute_reply": "2023-05-17T15:36:32.505050Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>soma_joinid</th>\n",
       "      <th>collection_id</th>\n",
       "      <th>collection_name</th>\n",
       "      <th>collection_doi</th>\n",
       "      <th>dataset_id</th>\n",
       "      <th>dataset_title</th>\n",
       "      <th>dataset_h5ad_path</th>\n",
       "      <th>dataset_total_cell_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>105</th>\n",
       "      <td>105</td>\n",
       "      <td>283d65eb-dd53-496d-adb7-7570c7caa443</td>\n",
       "      <td>Transcriptomic diversity of cell types across ...</td>\n",
       "      <td>10.1101/2022.10.12.511898</td>\n",
       "      <td>fe1a73ab-a203-45fd-84e9-0f7fd19efcbd</td>\n",
       "      <td>Dissection: Amygdaloid complex (AMY) - basolat...</td>\n",
       "      <td>fe1a73ab-a203-45fd-84e9-0f7fd19efcbd.h5ad</td>\n",
       "      <td>35285</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109</th>\n",
       "      <td>109</td>\n",
       "      <td>283d65eb-dd53-496d-adb7-7570c7caa443</td>\n",
       "      <td>Transcriptomic diversity of cell types across ...</td>\n",
       "      <td>10.1101/2022.10.12.511898</td>\n",
       "      <td>f8dda921-5fb4-4c94-a654-c6fc346bfd6d</td>\n",
       "      <td>Dissection: Cerebral cortex (Cx) - Occipitotem...</td>\n",
       "      <td>f8dda921-5fb4-4c94-a654-c6fc346bfd6d.h5ad</td>\n",
       "      <td>31899</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>126</th>\n",
       "      <td>126</td>\n",
       "      <td>283d65eb-dd53-496d-adb7-7570c7caa443</td>\n",
       "      <td>Transcriptomic diversity of cell types across ...</td>\n",
       "      <td>10.1101/2022.10.12.511898</td>\n",
       "      <td>dd03ce70-3243-4c96-9561-330cc461e4d7</td>\n",
       "      <td>Dissection: Cerebral cortex (Cx) - Perirhinal ...</td>\n",
       "      <td>dd03ce70-3243-4c96-9561-330cc461e4d7.h5ad</td>\n",
       "      <td>23732</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>131</th>\n",
       "      <td>131</td>\n",
       "      <td>283d65eb-dd53-496d-adb7-7570c7caa443</td>\n",
       "      <td>Transcriptomic diversity of cell types across ...</td>\n",
       "      <td>10.1101/2022.10.12.511898</td>\n",
       "      <td>d2b5efc1-14c6-4b5f-bd98-40f9084872d7</td>\n",
       "      <td>Dissection: Tail of Hippocampus (HiT) - Caudal...</td>\n",
       "      <td>d2b5efc1-14c6-4b5f-bd98-40f9084872d7.h5ad</td>\n",
       "      <td>36886</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>141</th>\n",
       "      <td>141</td>\n",
       "      <td>283d65eb-dd53-496d-adb7-7570c7caa443</td>\n",
       "      <td>Transcriptomic diversity of cell types across ...</td>\n",
       "      <td>10.1101/2022.10.12.511898</td>\n",
       "      <td>c4b03352-af8d-492a-8d6b-40f304e0a122</td>\n",
       "      <td>Supercluster: Medium spiny neuron</td>\n",
       "      <td>c4b03352-af8d-492a-8d6b-40f304e0a122.h5ad</td>\n",
       "      <td>152189</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>142</th>\n",
       "      <td>142</td>\n",
       "      <td>283d65eb-dd53-496d-adb7-7570c7caa443</td>\n",
       "      <td>Transcriptomic diversity of cell types across ...</td>\n",
       "      <td>10.1101/2022.10.12.511898</td>\n",
       "      <td>c2aad8fc-b63b-4f9b-9cfd-baf7bc9c1771</td>\n",
       "      <td>Dissection: Cerebral cortex (Cx) - Temporal po...</td>\n",
       "      <td>c2aad8fc-b63b-4f9b-9cfd-baf7bc9c1771.h5ad</td>\n",
       "      <td>37642</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>143</th>\n",
       "      <td>143</td>\n",
       "      <td>283d65eb-dd53-496d-adb7-7570c7caa443</td>\n",
       "      <td>Transcriptomic diversity of cell types across ...</td>\n",
       "      <td>10.1101/2022.10.12.511898</td>\n",
       "      <td>c202b243-1aa1-4b16-bc9a-b36241f3b1e3</td>\n",
       "      <td>Supercluster: Amygdala excitatory</td>\n",
       "      <td>c202b243-1aa1-4b16-bc9a-b36241f3b1e3.h5ad</td>\n",
       "      <td>109452</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>144</th>\n",
       "      <td>144</td>\n",
       "      <td>283d65eb-dd53-496d-adb7-7570c7caa443</td>\n",
       "      <td>Transcriptomic diversity of cell types across ...</td>\n",
       "      <td>10.1101/2022.10.12.511898</td>\n",
       "      <td>bdb26abd-f4ba-4ea3-8862-c2340e7a4f55</td>\n",
       "      <td>Supercluster: CGE interneuron</td>\n",
       "      <td>bdb26abd-f4ba-4ea3-8862-c2340e7a4f55.h5ad</td>\n",
       "      <td>227671</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>149</th>\n",
       "      <td>149</td>\n",
       "      <td>283d65eb-dd53-496d-adb7-7570c7caa443</td>\n",
       "      <td>Transcriptomic diversity of cell types across ...</td>\n",
       "      <td>10.1101/2022.10.12.511898</td>\n",
       "      <td>acae7679-d077-461c-b857-ee6ccfeb267f</td>\n",
       "      <td>Dissection: Head of hippocampus (HiH) - CA1</td>\n",
       "      <td>acae7679-d077-461c-b857-ee6ccfeb267f.h5ad</td>\n",
       "      <td>39147</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>162</th>\n",
       "      <td>162</td>\n",
       "      <td>283d65eb-dd53-496d-adb7-7570c7caa443</td>\n",
       "      <td>Transcriptomic diversity of cell types across ...</td>\n",
       "      <td>10.1101/2022.10.12.511898</td>\n",
       "      <td>9372df2d-13d6-4fac-980b-919a5b7eb483</td>\n",
       "      <td>Dissection: Midbrain (M) - Periaqueductal gray...</td>\n",
       "      <td>9372df2d-13d6-4fac-980b-919a5b7eb483.h5ad</td>\n",
       "      <td>33794</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163</th>\n",
       "      <td>163</td>\n",
       "      <td>283d65eb-dd53-496d-adb7-7570c7caa443</td>\n",
       "      <td>Transcriptomic diversity of cell types across ...</td>\n",
       "      <td>10.1101/2022.10.12.511898</td>\n",
       "      <td>93131426-0124-4ab4-a013-9dfbcd99d467</td>\n",
       "      <td>Dissection: Epithalamus - ETH</td>\n",
       "      <td>93131426-0124-4ab4-a013-9dfbcd99d467.h5ad</td>\n",
       "      <td>24327</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>172</th>\n",
       "      <td>172</td>\n",
       "      <td>283d65eb-dd53-496d-adb7-7570c7caa443</td>\n",
       "      <td>Transcriptomic diversity of cell types across ...</td>\n",
       "      <td>10.1101/2022.10.12.511898</td>\n",
       "      <td>7c1c3d47-3166-43e5-9a95-65ceb2d45f78</td>\n",
       "      <td>Dissection: Pons (Pn) - Pontine reticular form...</td>\n",
       "      <td>7c1c3d47-3166-43e5-9a95-65ceb2d45f78.h5ad</td>\n",
       "      <td>49512</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>174</th>\n",
       "      <td>174</td>\n",
       "      <td>283d65eb-dd53-496d-adb7-7570c7caa443</td>\n",
       "      <td>Transcriptomic diversity of cell types across ...</td>\n",
       "      <td>10.1101/2022.10.12.511898</td>\n",
       "      <td>7a0a8891-9a22-4549-a55b-c2aca23c3a2a</td>\n",
       "      <td>Supercluster: Hippocampal CA1-3</td>\n",
       "      <td>7a0a8891-9a22-4549-a55b-c2aca23c3a2a.h5ad</td>\n",
       "      <td>74979</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>186</th>\n",
       "      <td>186</td>\n",
       "      <td>283d65eb-dd53-496d-adb7-7570c7caa443</td>\n",
       "      <td>Transcriptomic diversity of cell types across ...</td>\n",
       "      <td>10.1101/2022.10.12.511898</td>\n",
       "      <td>5e5ab909-f73f-4b57-98a0-6d2c5662f6a4</td>\n",
       "      <td>Dissection: Midbrain (M) - Inferior colliculus...</td>\n",
       "      <td>5e5ab909-f73f-4b57-98a0-6d2c5662f6a4.h5ad</td>\n",
       "      <td>32306</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>209</th>\n",
       "      <td>209</td>\n",
       "      <td>283d65eb-dd53-496d-adb7-7570c7caa443</td>\n",
       "      <td>Transcriptomic diversity of cell types across ...</td>\n",
       "      <td>10.1101/2022.10.12.511898</td>\n",
       "      <td>3f56901c-dd4a-47d6-b60b-7b0c0111cfb2</td>\n",
       "      <td>Dissection: Head of hippocampus (HiH) - CA1-3</td>\n",
       "      <td>3f56901c-dd4a-47d6-b60b-7b0c0111cfb2.h5ad</td>\n",
       "      <td>37911</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>211</th>\n",
       "      <td>211</td>\n",
       "      <td>283d65eb-dd53-496d-adb7-7570c7caa443</td>\n",
       "      <td>Transcriptomic diversity of cell types across ...</td>\n",
       "      <td>10.1101/2022.10.12.511898</td>\n",
       "      <td>3a7f3ab4-a280-4b3b-b2c0-6dd05614a78c</td>\n",
       "      <td>Supercluster: Splatter</td>\n",
       "      <td>3a7f3ab4-a280-4b3b-b2c0-6dd05614a78c.h5ad</td>\n",
       "      <td>291833</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>215</th>\n",
       "      <td>215</td>\n",
       "      <td>283d65eb-dd53-496d-adb7-7570c7caa443</td>\n",
       "      <td>Transcriptomic diversity of cell types across ...</td>\n",
       "      <td>10.1101/2022.10.12.511898</td>\n",
       "      <td>35c8a04c-8639-4d15-8228-765d8d93fc96</td>\n",
       "      <td>Dissection: Hypothalamus (HTH) - supraoptic re...</td>\n",
       "      <td>35c8a04c-8639-4d15-8228-765d8d93fc96.h5ad</td>\n",
       "      <td>16753</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>236</th>\n",
       "      <td>236</td>\n",
       "      <td>283d65eb-dd53-496d-adb7-7570c7caa443</td>\n",
       "      <td>Transcriptomic diversity of cell types across ...</td>\n",
       "      <td>10.1101/2022.10.12.511898</td>\n",
       "      <td>07b1d7c8-5c2e-42f7-9246-26f746cd6013</td>\n",
       "      <td>Dissection: Myelencephalon (medulla oblongata)...</td>\n",
       "      <td>07b1d7c8-5c2e-42f7-9246-26f746cd6013.h5ad</td>\n",
       "      <td>27210</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>239</th>\n",
       "      <td>239</td>\n",
       "      <td>283d65eb-dd53-496d-adb7-7570c7caa443</td>\n",
       "      <td>Transcriptomic diversity of cell types across ...</td>\n",
       "      <td>10.1101/2022.10.12.511898</td>\n",
       "      <td>0325478a-9b52-45b5-b40a-2e2ab0d72eb1</td>\n",
       "      <td>Supercluster: Upper-layer intratelencephalic</td>\n",
       "      <td>0325478a-9b52-45b5-b40a-2e2ab0d72eb1.h5ad</td>\n",
       "      <td>455006</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>444</th>\n",
       "      <td>444</td>\n",
       "      <td>e5f58829-1a66-40b5-a624-9046778e74f5</td>\n",
       "      <td>Tabula Sapiens</td>\n",
       "      <td>10.1126/science.abl4896</td>\n",
       "      <td>53d208b0-2cfd-4366-9866-c3c6114081bc</td>\n",
       "      <td>Tabula Sapiens - All Cells</td>\n",
       "      <td>53d208b0-2cfd-4366-9866-c3c6114081bc.h5ad</td>\n",
       "      <td>483152</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>445</th>\n",
       "      <td>445</td>\n",
       "      <td>e5f58829-1a66-40b5-a624-9046778e74f5</td>\n",
       "      <td>Tabula Sapiens</td>\n",
       "      <td>10.1126/science.abl4896</td>\n",
       "      <td>a68b64d8-aee3-4947-81b7-36b8fe5a44d2</td>\n",
       "      <td>Tabula Sapiens - Stromal</td>\n",
       "      <td>a68b64d8-aee3-4947-81b7-36b8fe5a44d2.h5ad</td>\n",
       "      <td>82478</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>446</th>\n",
       "      <td>446</td>\n",
       "      <td>e5f58829-1a66-40b5-a624-9046778e74f5</td>\n",
       "      <td>Tabula Sapiens</td>\n",
       "      <td>10.1126/science.abl4896</td>\n",
       "      <td>c5d88abe-f23a-45fa-a534-788985e93dad</td>\n",
       "      <td>Tabula Sapiens - Immune</td>\n",
       "      <td>c5d88abe-f23a-45fa-a534-788985e93dad.h5ad</td>\n",
       "      <td>264824</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>447</th>\n",
       "      <td>447</td>\n",
       "      <td>e5f58829-1a66-40b5-a624-9046778e74f5</td>\n",
       "      <td>Tabula Sapiens</td>\n",
       "      <td>10.1126/science.abl4896</td>\n",
       "      <td>5a11f879-d1ef-458a-910c-9b0bdfca5ebf</td>\n",
       "      <td>Tabula Sapiens - Endothelial</td>\n",
       "      <td>5a11f879-d1ef-458a-910c-9b0bdfca5ebf.h5ad</td>\n",
       "      <td>31691</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>448</th>\n",
       "      <td>448</td>\n",
       "      <td>e5f58829-1a66-40b5-a624-9046778e74f5</td>\n",
       "      <td>Tabula Sapiens</td>\n",
       "      <td>10.1126/science.abl4896</td>\n",
       "      <td>97a17473-e2b1-4f31-a544-44a60773e2dd</td>\n",
       "      <td>Tabula Sapiens - Epithelial</td>\n",
       "      <td>97a17473-e2b1-4f31-a544-44a60773e2dd.h5ad</td>\n",
       "      <td>104148</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     soma_joinid                         collection_id   \n",
       "105          105  283d65eb-dd53-496d-adb7-7570c7caa443  \\\n",
       "109          109  283d65eb-dd53-496d-adb7-7570c7caa443   \n",
       "126          126  283d65eb-dd53-496d-adb7-7570c7caa443   \n",
       "131          131  283d65eb-dd53-496d-adb7-7570c7caa443   \n",
       "141          141  283d65eb-dd53-496d-adb7-7570c7caa443   \n",
       "142          142  283d65eb-dd53-496d-adb7-7570c7caa443   \n",
       "143          143  283d65eb-dd53-496d-adb7-7570c7caa443   \n",
       "144          144  283d65eb-dd53-496d-adb7-7570c7caa443   \n",
       "149          149  283d65eb-dd53-496d-adb7-7570c7caa443   \n",
       "162          162  283d65eb-dd53-496d-adb7-7570c7caa443   \n",
       "163          163  283d65eb-dd53-496d-adb7-7570c7caa443   \n",
       "172          172  283d65eb-dd53-496d-adb7-7570c7caa443   \n",
       "174          174  283d65eb-dd53-496d-adb7-7570c7caa443   \n",
       "186          186  283d65eb-dd53-496d-adb7-7570c7caa443   \n",
       "209          209  283d65eb-dd53-496d-adb7-7570c7caa443   \n",
       "211          211  283d65eb-dd53-496d-adb7-7570c7caa443   \n",
       "215          215  283d65eb-dd53-496d-adb7-7570c7caa443   \n",
       "236          236  283d65eb-dd53-496d-adb7-7570c7caa443   \n",
       "239          239  283d65eb-dd53-496d-adb7-7570c7caa443   \n",
       "444          444  e5f58829-1a66-40b5-a624-9046778e74f5   \n",
       "445          445  e5f58829-1a66-40b5-a624-9046778e74f5   \n",
       "446          446  e5f58829-1a66-40b5-a624-9046778e74f5   \n",
       "447          447  e5f58829-1a66-40b5-a624-9046778e74f5   \n",
       "448          448  e5f58829-1a66-40b5-a624-9046778e74f5   \n",
       "\n",
       "                                       collection_name   \n",
       "105  Transcriptomic diversity of cell types across ...  \\\n",
       "109  Transcriptomic diversity of cell types across ...   \n",
       "126  Transcriptomic diversity of cell types across ...   \n",
       "131  Transcriptomic diversity of cell types across ...   \n",
       "141  Transcriptomic diversity of cell types across ...   \n",
       "142  Transcriptomic diversity of cell types across ...   \n",
       "143  Transcriptomic diversity of cell types across ...   \n",
       "144  Transcriptomic diversity of cell types across ...   \n",
       "149  Transcriptomic diversity of cell types across ...   \n",
       "162  Transcriptomic diversity of cell types across ...   \n",
       "163  Transcriptomic diversity of cell types across ...   \n",
       "172  Transcriptomic diversity of cell types across ...   \n",
       "174  Transcriptomic diversity of cell types across ...   \n",
       "186  Transcriptomic diversity of cell types across ...   \n",
       "209  Transcriptomic diversity of cell types across ...   \n",
       "211  Transcriptomic diversity of cell types across ...   \n",
       "215  Transcriptomic diversity of cell types across ...   \n",
       "236  Transcriptomic diversity of cell types across ...   \n",
       "239  Transcriptomic diversity of cell types across ...   \n",
       "444                                     Tabula Sapiens   \n",
       "445                                     Tabula Sapiens   \n",
       "446                                     Tabula Sapiens   \n",
       "447                                     Tabula Sapiens   \n",
       "448                                     Tabula Sapiens   \n",
       "\n",
       "                collection_doi                            dataset_id   \n",
       "105  10.1101/2022.10.12.511898  fe1a73ab-a203-45fd-84e9-0f7fd19efcbd  \\\n",
       "109  10.1101/2022.10.12.511898  f8dda921-5fb4-4c94-a654-c6fc346bfd6d   \n",
       "126  10.1101/2022.10.12.511898  dd03ce70-3243-4c96-9561-330cc461e4d7   \n",
       "131  10.1101/2022.10.12.511898  d2b5efc1-14c6-4b5f-bd98-40f9084872d7   \n",
       "141  10.1101/2022.10.12.511898  c4b03352-af8d-492a-8d6b-40f304e0a122   \n",
       "142  10.1101/2022.10.12.511898  c2aad8fc-b63b-4f9b-9cfd-baf7bc9c1771   \n",
       "143  10.1101/2022.10.12.511898  c202b243-1aa1-4b16-bc9a-b36241f3b1e3   \n",
       "144  10.1101/2022.10.12.511898  bdb26abd-f4ba-4ea3-8862-c2340e7a4f55   \n",
       "149  10.1101/2022.10.12.511898  acae7679-d077-461c-b857-ee6ccfeb267f   \n",
       "162  10.1101/2022.10.12.511898  9372df2d-13d6-4fac-980b-919a5b7eb483   \n",
       "163  10.1101/2022.10.12.511898  93131426-0124-4ab4-a013-9dfbcd99d467   \n",
       "172  10.1101/2022.10.12.511898  7c1c3d47-3166-43e5-9a95-65ceb2d45f78   \n",
       "174  10.1101/2022.10.12.511898  7a0a8891-9a22-4549-a55b-c2aca23c3a2a   \n",
       "186  10.1101/2022.10.12.511898  5e5ab909-f73f-4b57-98a0-6d2c5662f6a4   \n",
       "209  10.1101/2022.10.12.511898  3f56901c-dd4a-47d6-b60b-7b0c0111cfb2   \n",
       "211  10.1101/2022.10.12.511898  3a7f3ab4-a280-4b3b-b2c0-6dd05614a78c   \n",
       "215  10.1101/2022.10.12.511898  35c8a04c-8639-4d15-8228-765d8d93fc96   \n",
       "236  10.1101/2022.10.12.511898  07b1d7c8-5c2e-42f7-9246-26f746cd6013   \n",
       "239  10.1101/2022.10.12.511898  0325478a-9b52-45b5-b40a-2e2ab0d72eb1   \n",
       "444    10.1126/science.abl4896  53d208b0-2cfd-4366-9866-c3c6114081bc   \n",
       "445    10.1126/science.abl4896  a68b64d8-aee3-4947-81b7-36b8fe5a44d2   \n",
       "446    10.1126/science.abl4896  c5d88abe-f23a-45fa-a534-788985e93dad   \n",
       "447    10.1126/science.abl4896  5a11f879-d1ef-458a-910c-9b0bdfca5ebf   \n",
       "448    10.1126/science.abl4896  97a17473-e2b1-4f31-a544-44a60773e2dd   \n",
       "\n",
       "                                         dataset_title   \n",
       "105  Dissection: Amygdaloid complex (AMY) - basolat...  \\\n",
       "109  Dissection: Cerebral cortex (Cx) - Occipitotem...   \n",
       "126  Dissection: Cerebral cortex (Cx) - Perirhinal ...   \n",
       "131  Dissection: Tail of Hippocampus (HiT) - Caudal...   \n",
       "141                  Supercluster: Medium spiny neuron   \n",
       "142  Dissection: Cerebral cortex (Cx) - Temporal po...   \n",
       "143                  Supercluster: Amygdala excitatory   \n",
       "144                      Supercluster: CGE interneuron   \n",
       "149        Dissection: Head of hippocampus (HiH) - CA1   \n",
       "162  Dissection: Midbrain (M) - Periaqueductal gray...   \n",
       "163                      Dissection: Epithalamus - ETH   \n",
       "172  Dissection: Pons (Pn) - Pontine reticular form...   \n",
       "174                    Supercluster: Hippocampal CA1-3   \n",
       "186  Dissection: Midbrain (M) - Inferior colliculus...   \n",
       "209      Dissection: Head of hippocampus (HiH) - CA1-3   \n",
       "211                             Supercluster: Splatter   \n",
       "215  Dissection: Hypothalamus (HTH) - supraoptic re...   \n",
       "236  Dissection: Myelencephalon (medulla oblongata)...   \n",
       "239       Supercluster: Upper-layer intratelencephalic   \n",
       "444                         Tabula Sapiens - All Cells   \n",
       "445                           Tabula Sapiens - Stromal   \n",
       "446                            Tabula Sapiens - Immune   \n",
       "447                       Tabula Sapiens - Endothelial   \n",
       "448                        Tabula Sapiens - Epithelial   \n",
       "\n",
       "                             dataset_h5ad_path  dataset_total_cell_count  \n",
       "105  fe1a73ab-a203-45fd-84e9-0f7fd19efcbd.h5ad                     35285  \n",
       "109  f8dda921-5fb4-4c94-a654-c6fc346bfd6d.h5ad                     31899  \n",
       "126  dd03ce70-3243-4c96-9561-330cc461e4d7.h5ad                     23732  \n",
       "131  d2b5efc1-14c6-4b5f-bd98-40f9084872d7.h5ad                     36886  \n",
       "141  c4b03352-af8d-492a-8d6b-40f304e0a122.h5ad                    152189  \n",
       "142  c2aad8fc-b63b-4f9b-9cfd-baf7bc9c1771.h5ad                     37642  \n",
       "143  c202b243-1aa1-4b16-bc9a-b36241f3b1e3.h5ad                    109452  \n",
       "144  bdb26abd-f4ba-4ea3-8862-c2340e7a4f55.h5ad                    227671  \n",
       "149  acae7679-d077-461c-b857-ee6ccfeb267f.h5ad                     39147  \n",
       "162  9372df2d-13d6-4fac-980b-919a5b7eb483.h5ad                     33794  \n",
       "163  93131426-0124-4ab4-a013-9dfbcd99d467.h5ad                     24327  \n",
       "172  7c1c3d47-3166-43e5-9a95-65ceb2d45f78.h5ad                     49512  \n",
       "174  7a0a8891-9a22-4549-a55b-c2aca23c3a2a.h5ad                     74979  \n",
       "186  5e5ab909-f73f-4b57-98a0-6d2c5662f6a4.h5ad                     32306  \n",
       "209  3f56901c-dd4a-47d6-b60b-7b0c0111cfb2.h5ad                     37911  \n",
       "211  3a7f3ab4-a280-4b3b-b2c0-6dd05614a78c.h5ad                    291833  \n",
       "215  35c8a04c-8639-4d15-8228-765d8d93fc96.h5ad                     16753  \n",
       "236  07b1d7c8-5c2e-42f7-9246-26f746cd6013.h5ad                     27210  \n",
       "239  0325478a-9b52-45b5-b40a-2e2ab0d72eb1.h5ad                    455006  \n",
       "444  53d208b0-2cfd-4366-9866-c3c6114081bc.h5ad                    483152  \n",
       "445  a68b64d8-aee3-4947-81b7-36b8fe5a44d2.h5ad                     82478  \n",
       "446  c5d88abe-f23a-45fa-a534-788985e93dad.h5ad                    264824  \n",
       "447  5a11f879-d1ef-458a-910c-9b0bdfca5ebf.h5ad                     31691  \n",
       "448  97a17473-e2b1-4f31-a544-44a60773e2dd.h5ad                    104148  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Grab the feature's soma_joinid from the var dataframe\n",
    "var_joinid = var_df.loc[var_df.feature_id == \"ENSG00000286096\"].soma_joinid\n",
    "\n",
    "# The presence matrix is indexed by the joinids of the dataset and var dataframes,\n",
    "# so slice out the feature of interest by its joinid.\n",
    "dataset_joinids = presence_matrix[:, var_joinid].tocoo().row\n",
    "\n",
    "# From the datasets dataframe, slice out the datasets which have a joinid in the list\n",
    "datasets_df.loc[datasets_df.soma_joinid.isin(dataset_joinids)]"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Identifying all genes measured in a dataset \n",
    "\n",
    "Finally, we can find the set of genes that were measured in the cells of a given dataset."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-05-17T15:36:32.508017Z",
     "iopub.status.busy": "2023-05-17T15:36:32.507876Z",
     "iopub.status.idle": "2023-05-17T15:36:32.518051Z",
     "shell.execute_reply": "2023-05-17T15:36:32.517508Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>soma_joinid</th>\n",
       "      <th>feature_id</th>\n",
       "      <th>feature_name</th>\n",
       "      <th>feature_length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>ENSG00000238009</td>\n",
       "      <td>RP11-34P13.7</td>\n",
       "      <td>3726</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>ENSG00000239945</td>\n",
       "      <td>RP11-34P13.8</td>\n",
       "      <td>1319</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>13</td>\n",
       "      <td>ENSG00000229905</td>\n",
       "      <td>RP11-206L10.4</td>\n",
       "      <td>456</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>14</td>\n",
       "      <td>ENSG00000237491</td>\n",
       "      <td>LINC01409</td>\n",
       "      <td>8413</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>15</td>\n",
       "      <td>ENSG00000177757</td>\n",
       "      <td>FAM87B</td>\n",
       "      <td>1947</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51459</th>\n",
       "      <td>51459</td>\n",
       "      <td>ENSG00000277778</td>\n",
       "      <td>PGM5P2</td>\n",
       "      <td>1980</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51983</th>\n",
       "      <td>51983</td>\n",
       "      <td>ENSG00000254893</td>\n",
       "      <td>RAP1BL</td>\n",
       "      <td>555</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58043</th>\n",
       "      <td>58043</td>\n",
       "      <td>ENSG00000261408</td>\n",
       "      <td>TEN1-CDK3</td>\n",
       "      <td>3898</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58688</th>\n",
       "      <td>58688</td>\n",
       "      <td>ENSG00000279457</td>\n",
       "      <td>WASH9P</td>\n",
       "      <td>1397</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58928</th>\n",
       "      <td>58928</td>\n",
       "      <td>ENSG00000279152</td>\n",
       "      <td>RP11-599B13.10</td>\n",
       "      <td>756</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>27211 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       soma_joinid       feature_id    feature_name  feature_length\n",
       "3                3  ENSG00000238009    RP11-34P13.7            3726\n",
       "4                4  ENSG00000239945    RP11-34P13.8            1319\n",
       "13              13  ENSG00000229905   RP11-206L10.4             456\n",
       "14              14  ENSG00000237491       LINC01409            8413\n",
       "15              15  ENSG00000177757          FAM87B            1947\n",
       "...            ...              ...             ...             ...\n",
       "51459        51459  ENSG00000277778          PGM5P2            1980\n",
       "51983        51983  ENSG00000254893          RAP1BL             555\n",
       "58043        58043  ENSG00000261408       TEN1-CDK3            3898\n",
       "58688        58688  ENSG00000279457          WASH9P            1397\n",
       "58928        58928  ENSG00000279152  RP11-599B13.10             756\n",
       "\n",
       "[27211 rows x 4 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Slice the dataset(s) of interest, and get the joinid(s)\n",
    "dataset_joinids = datasets_df.loc[datasets_df.collection_id == \"17481d16-ee44-49e5-bcf0-28c0780d8c4a\"].soma_joinid\n",
    "\n",
    "# Slice the presence matrix by the first dimension, i.e., by dataset\n",
    "var_joinids = presence_matrix[dataset_joinids, :].tocoo().col\n",
    "\n",
    "# From the feature (var) dataframe, slice out features which have a joinid in the list.\n",
    "var_df.loc[var_df.soma_joinid.isin(var_joinids)]"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.11"
  },
  "vscode": {
   "interpreter": {
    "hash": "3da8ec1c162cd849e59e6ea2824b2e353dce799884e910aae99411be5277f953"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
