

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8">
  <meta name="generator" content="Docutils 0.19: https://docutils.sourceforge.io/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Access CELLxGENE-hosted embeddings &mdash; cellxgene-census  documentation</title>
  

  
  <link rel="icon" type="image/png" sizes="32x32" href="../../_static/img/favicon_32x32_v2.png"/>
  <link rel="icon" type="image/png" sizes="16x16" href="../../_static/img/favicon_16x16_v2.png"/>
  <link rel="preconnect" href="https://fonts.googleapis.com">
  <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
  <link href="https://fonts.googleapis.com/css2?family=Inter:wght@400;500;600;700&display=swap" rel="stylesheet">
  <link rel="stylesheet" href="../../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/css/custom.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/nbsphinx-code-cells.css" type="text/css" />

  
  
  
  

  
  <!--[if lt IE 9]>
    <script src="../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../../" src="../../_static/documentation_options.js"></script>
        <script data-url_root="../../" id="documentation_options" src="../../_static/documentation_options.js"></script>
        <script src="../../_static/doctools.js"></script>
        <script src="../../_static/sphinx_highlight.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
        <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
        <script data-domain="chanzuckerberg.github.io/cellxgene-census" defer="defer" src="https://plausible.io/js/script.js"></script>
    
    <script type="text/javascript" src="../../_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="scVI for cell type prediction and data projection" href="../analysis_demo/comp_bio_scvi_model_use.html" />
    <link rel="prev" title="Access CELLxGENE collaboration embeddings (scVI, Geneformer)" href="census_access_maintained_embeddings.html" /> 
</head>

<body class="wy-body-for-nav">

  <div style="height: 100vh; overflow: hidden;">
     
    <!-- top 56px for navbar height -->
    <div class="navbar-cxg">
        <div style="padding: 0 16px; display: flex; height: inherit; align-items: center; justify-content: space-between;">
            <span style="display:flex; gap: 32px; align-items: center;">
                <a href="https://cellxgene.cziscience.com/">
                  <img src="../../_static/img/cellxgene-discover-logo.svg" />
                </a>
                <span class="navbar-cxg-nav-wrapper">
                  <span class="navbar-cxg-section">
                    <span class="navbar-cxg-nav-section-title">Application</span>
                    <span class="navbar-cxg-nav-item-container">
                      <span class="navbar-cxg-link">
                          <a class="navbar-cxg-anchor" href="https://cellxgene.cziscience.com/collections">Collections</a>
                      </span>
                      <span class="navbar-cxg-link">
                          <a class="navbar-cxg-anchor" href="https://cellxgene.cziscience.com/datasets">Datasets</a>
                      </span>
                      <span class="navbar-cxg-link">
                          <a class="navbar-cxg-anchor" href="https://cellxgene.cziscience.com/gene-expression">Gene Expression</a>
                      </span>
                      <span class="navbar-cxg-link">
                        <a class="navbar-cxg-anchor" href="https://cellxgene.cziscience.com/cellguide">Cell Guide</a>
                        <div style="height: 16px!important; display: flex;">
                          <span class="beta">BETA</span>
                        </div>
                      </span>
                    </span>
                  </span>
                  <hr class="navbar-divider"/>
                  <span class="navbar-cxg-section">
                    <span class="navbar-cxg-nav-section-title">Census</span>
                    <span class="navbar-cxg-nav-item-container">
                      <span class="navbar-cxg-link active-link">
                        <a class="navbar-cxg-anchor" href="/cellxgene-census/index.html">API</a>
                      </span>
                      <span class="navbar-cxg-link">
                        <a class="navbar-cxg-anchor" href="https://cellxgene.cziscience.com/census-models">Models</a>
                      </span>
                    </span>
                  </span>
                </span>
            </span>

            <span class="navbar-cxg-link">
              <a class="navbar-cxg-anchor" href="https://cellxgene.cziscience.com/docs">Help & Documentation</a>
            </span>
        </div>
    </div>
    <div style="width: 100%; height: 100%; overflow: auto; padding-top: 56px;">
        
        <!-- top 56px for navbar height -->
        <nav data-toggle="wy-nav-shift" class="wy-nav-side" style="top: 56px;" >
        <div class="wy-side-scroll">
            <div class="wy-side-nav-search" >
            

            
                <a href="../../index.html" class="icon icon-home"> cellxgene-census
            

            
            </a>

            
                
                
                <div class="version">
                    v1.13.1
                </div>
                
            

            

            
            </div>

            
            <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
            
                
                
                
                
                
                <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../../cellxgene_census_docsite_installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../cellxgene_census_docsite_quick_start.html">Quick start</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../articles.html">What‚Äôs new?</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../cellxgene_census_aws_open_data.html">Census in AWS ‚òÅÔ∏è</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../cellxgene_census_docsite_schema.html">Census data and schema</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../cellxgene_census_docsite_data_release_info.html">Census data releases</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../python-api.html">Python API</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../../examples.html">Python tutorials</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="../../examples.html#exporting-data">Exporting data</a></li>
<li class="toctree-l2 current"><a class="reference internal" href="../../examples.html#new-using-integrated-embeddings-and-models">[NEW! üöÄ] Using integrated embeddings and models</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="census_access_maintained_embeddings.html">Access CELLxGENE collaboration embeddings (scVI, Geneformer)</a></li>
<li class="toctree-l3 current"><a class="current reference internal" href="#">Access CELLxGENE-hosted embeddings</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#Contents">Contents</a></li>
<li class="toctree-l4"><a class="reference internal" href="#Background">Background</a></li>
<li class="toctree-l4"><a class="reference internal" href="#Quick-start">Quick start</a></li>
<li class="toctree-l4"><a class="reference internal" href="#Storage-format">Storage format</a></li>
<li class="toctree-l4"><a class="reference internal" href="#Query-cells-and-load-associated-embeddings">Query cells and load associated embeddings</a></li>
<li class="toctree-l4"><a class="reference internal" href="#Load-embeddings-and-fetch-associated-Census-data">Load embeddings and fetch associated Census data</a></li>
<li class="toctree-l4"><a class="reference internal" href="#Embedding-Metadata">Embedding Metadata</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../analysis_demo/comp_bio_scvi_model_use.html">scVI for cell type prediction and data projection</a></li>
<li class="toctree-l3"><a class="reference internal" href="../analysis_demo/comp_bio_geneformer_prediction.html">Geneformer for cell class prediction and data projection</a></li>
<li class="toctree-l3"><a class="reference internal" href="../analysis_demo/comp_bio_embedding_exploration.html">Exploring biologically relevant clusters in Census embeddings</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../examples.html#understanding-census-data">Understanding Census data</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../examples.html#analyzing-census-data">Analyzing Census data</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../examples.html#scalable-computing">Scalable computing</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../examples.html#scalable-machine-learning">Scalable machine learning</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference external" href="https://chanzuckerberg.github.io/cellxgene-census/r/index.html">R API &amp; tutorials</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../cellxgene_census_docsite_FAQ.html">FAQ</a></li>
</ul>

                
            
            </div>
            
        </div>
        </nav>

        <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

        
        <nav class="wy-nav-top" aria-label="top navigation">
            
            <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
            <a href="../../index.html">cellxgene-census</a>
            
        </nav>


        <div class="wy-nav-content">
            
            <div class="rst-content">
            
            <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../../examples.html">Python tutorials</a></li>
      <li class="breadcrumb-item active">Access CELLxGENE-hosted embeddings</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../../_sources/notebooks/api_demo/census_embedding.ipynb.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
            <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
            <div itemprop="articleBody">
                
  <section id="Access-CELLxGENE-hosted-embeddings">
<h1>Access CELLxGENE-hosted embeddings<a class="headerlink" href="#Access-CELLxGENE-hosted-embeddings" title="Permalink to this heading">¬∂</a></h1>
<p>This notebook demonstrates basic access to CELLxGENE-hosted embeddings of the Census. <strong>CELLxGENE-hosted embeddings have been contributed by the community</strong>, CELLxGENE Discover does not actively maintain or update them. Find out more about these in the <a class="reference external" href="https://cellxgene.cziscience.com/census-models">Census model page</a>.</p>
<p><strong>IMPORTANT:</strong> This tutorial requires cellxgene-census package version 1.9.1 or later.</p>
<section id="Contents">
<h2>Contents<a class="headerlink" href="#Contents" title="Permalink to this heading">¬∂</a></h2>
<ol class="arabic simple">
<li><p>Background</p></li>
<li><p>Quick start</p></li>
<li><p>Query cells and load associated embeddings</p></li>
<li><p>Load embeddings and fetch associated Census data</p></li>
<li><p>Embedding metadata</p></li>
</ol>
<blockquote>
<div><p>‚ö†Ô∏è Note that the Census RNA data includes duplicate cells present across multiple datasets. Duplicate cells can be filtered in or out using the cell metadata variable <code class="docutils literal notranslate"><span class="pre">is_primary_data</span></code> which is described in the <a class="reference external" href="https://github.com/chanzuckerberg/cellxgene-census/blob/main/docs/cellxgene_census_schema.md#repeated-data">Census schema</a>.</p>
</div></blockquote>
</section>
<section id="Background">
<h2>Background<a class="headerlink" href="#Background" title="Permalink to this heading">¬∂</a></h2>
<p>This notebook demonstrates access to CELLxGENE-hosted embeddings of the Census. The Census has multiple releases, named by a <code class="docutils literal notranslate"><span class="pre">census_version</span></code>, which normally looks like an ISO date, e.g., <code class="docutils literal notranslate"><span class="pre">2023-02-01</span></code>. A CELLxGENE-hosted embedding is a 2D sparse matrix of cell embeddings <em>for a given census version</em>, encoded as a SOMA SparseNDArray.</p>
<p>‚ö†Ô∏è Note that embeddings may be available for one or both organisms, see the <a class="reference external" href="https://cellxgene.cziscience.com/census-models">Census model page</a> for the latest availability.</p>
<p>‚ö†Ô∏è <strong>IMPORTANT:</strong> embeddings are only meaningful in the context of the Census from which they were created. Each embedding contains a metadata field indicating the source Census, suitable for confirming embedding lineage.</p>
</section>
<section id="Quick-start">
<h2>Quick start<a class="headerlink" href="#Quick-start" title="Permalink to this heading">¬∂</a></h2>
<p>The easiest way to access Census CELLxGENE-hosted embeddings is by calling the <code class="docutils literal notranslate"><span class="pre">get_anndata</span></code> function with an <code class="docutils literal notranslate"><span class="pre">obs_embeddings</span></code> or <code class="docutils literal notranslate"><span class="pre">var_embeddings</span></code> parameter.</p>
<p>Let‚Äôs start by exploring what embeddings are available:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">cellxgene_census.experimental</span> <span class="kn">import</span> <span class="n">get_all_available_embeddings</span>

<span class="n">census_version</span> <span class="o">=</span> <span class="s2">&quot;2023-12-15&quot;</span>

<span class="k">for</span> <span class="n">e</span> <span class="ow">in</span> <span class="n">get_all_available_embeddings</span><span class="p">(</span><span class="n">census_version</span><span class="p">):</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">e</span><span class="p">[</span><span class="s1">&#39;embedding_name&#39;</span><span class="p">]</span><span class="si">:</span><span class="s2">15</span><span class="si">}</span><span class="s2"> </span><span class="si">{</span><span class="n">e</span><span class="p">[</span><span class="s1">&#39;experiment_name&#39;</span><span class="p">]</span><span class="si">:</span><span class="s2">15</span><span class="si">}</span><span class="s2"> </span><span class="si">{</span><span class="n">e</span><span class="p">[</span><span class="s1">&#39;data_type&#39;</span><span class="p">]</span><span class="si">:</span><span class="s2">15</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
scvi            mus_musculus    obs_embedding
scvi            homo_sapiens    obs_embedding
geneformer      homo_sapiens    obs_embedding
scgpt           homo_sapiens    obs_embedding
uce             homo_sapiens    obs_embedding
uce             mus_musculus    obs_embedding
nmf             homo_sapiens    obs_embedding
nmf             homo_sapiens    var_embedding
</pre></div></div>
</div>
<p>These can also be viewed on the <a class="reference external" href="https://cellxgene.cziscience.com/census-models">CELLxGENE Census Models page</a>.</p>
<p>For this example, we‚Äôll use <code class="docutils literal notranslate"><span class="pre">scgpt</span></code>. We can call <code class="docutils literal notranslate"><span class="pre">get_anndata</span></code> with the <code class="docutils literal notranslate"><span class="pre">obs_embeddings</span></code> parameter:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">cellxgene_census</span>
<span class="kn">from</span> <span class="nn">cellxgene_census.experimental</span> <span class="kn">import</span> <span class="n">get_embedding</span>

<span class="n">CENSUS_VERSION</span> <span class="o">=</span> <span class="s2">&quot;2023-12-15&quot;</span>

<span class="k">with</span> <span class="n">cellxgene_census</span><span class="o">.</span><span class="n">open_soma</span><span class="p">(</span><span class="n">census_version</span><span class="o">=</span><span class="n">CENSUS_VERSION</span><span class="p">)</span> <span class="k">as</span> <span class="n">census</span><span class="p">:</span>
    <span class="n">adata</span> <span class="o">=</span> <span class="n">cellxgene_census</span><span class="o">.</span><span class="n">get_anndata</span><span class="p">(</span>
        <span class="n">census</span><span class="p">,</span>
        <span class="n">organism</span><span class="o">=</span><span class="s2">&quot;homo_sapiens&quot;</span><span class="p">,</span>
        <span class="n">measurement_name</span><span class="o">=</span><span class="s2">&quot;RNA&quot;</span><span class="p">,</span>
        <span class="n">obs_value_filter</span><span class="o">=</span><span class="s2">&quot;tissue == &#39;tongue&#39;&quot;</span><span class="p">,</span>
        <span class="n">obs_embeddings</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;scgpt&quot;</span><span class="p">],</span>
    <span class="p">)</span>
</pre></div>
</div>
</div>
<p>And now you can use these data for downstream analysis.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
AnnData object with n_obs √ó n_vars = 372 √ó 60664
    obs: &#39;soma_joinid&#39;, &#39;dataset_id&#39;, &#39;assay&#39;, &#39;assay_ontology_term_id&#39;, &#39;cell_type&#39;, &#39;cell_type_ontology_term_id&#39;, &#39;development_stage&#39;, &#39;development_stage_ontology_term_id&#39;, &#39;disease&#39;, &#39;disease_ontology_term_id&#39;, &#39;donor_id&#39;, &#39;is_primary_data&#39;, &#39;self_reported_ethnicity&#39;, &#39;self_reported_ethnicity_ontology_term_id&#39;, &#39;sex&#39;, &#39;sex_ontology_term_id&#39;, &#39;suspension_type&#39;, &#39;tissue&#39;, &#39;tissue_ontology_term_id&#39;, &#39;tissue_general&#39;, &#39;tissue_general_ontology_term_id&#39;, &#39;raw_sum&#39;, &#39;nnz&#39;, &#39;raw_mean_nnz&#39;, &#39;raw_variance_nnz&#39;, &#39;n_measured_vars&#39;
    var: &#39;soma_joinid&#39;, &#39;feature_id&#39;, &#39;feature_name&#39;, &#39;feature_length&#39;, &#39;nnz&#39;, &#39;n_measured_obs&#39;
    obsm: &#39;scgpt&#39;
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span><span class="o">.</span><span class="n">obsm</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
AxisArrays with keys: scgpt
</pre></div></div>
</div>
<p>‚ö†Ô∏è <strong>IMPORTANT</strong>: <code class="docutils literal notranslate"><span class="pre">get_embedding</span></code> will fill in missing values in the embedding matrix with <code class="docutils literal notranslate"><span class="pre">NaN</span></code>, therefore missing cells are represented with rows where all values are <code class="docutils literal notranslate"><span class="pre">NaN</span></code>. To learn more use <code class="docutils literal notranslate"><span class="pre">help(get_embedding)</span></code>.</p>
</section>
<section id="Storage-format">
<h2>Storage format<a class="headerlink" href="#Storage-format" title="Permalink to this heading">¬∂</a></h2>
<p>Each embedding is encoded as a SOMA SparseNDArray, where:</p>
<ul class="simple">
<li><p>dimension 0 (<code class="docutils literal notranslate"><span class="pre">soma_dim_0</span></code>) encodes the cell (obs) <code class="docutils literal notranslate"><span class="pre">soma_joinid</span></code> value</p></li>
<li><p>dimension 1 (<code class="docutils literal notranslate"><span class="pre">soma_dim_1</span></code>) encodes the embedding feature, and is in the range [0, N) where N is the number of featues in the embedding</p></li>
<li><p>data (<code class="docutils literal notranslate"><span class="pre">soma_data</span></code>) is float32</p></li>
</ul>
<p>‚ö†Ô∏è <strong>IMPORTANT:</strong> CELLxGENE-hosted embeddings may embed a subset of the cells in any given Census version. If a cell has an embedding, it will be explicitly stored in the sparse array, <em>even if the embedding value is zero</em>. In other words, missing array values values imply that the cell was not embedded, whereas zero valued embeddings are explicitly stored. Put another way, the <code class="docutils literal notranslate"><span class="pre">nnz</span></code> of the embedding array indicate the number of embedded cells, not the number of non-zero values.</p>
<p>The first axis of the embedding array will have the same shape as the corresponding <code class="docutils literal notranslate"><span class="pre">obs</span></code> DataFrame for the Census build and experiment. The second axis of the embedding will have a shape (0, N) where N is the number of features in the embedding.</p>
<p>Embedding values, while stored as a float32, are precision reduced. Currently they are equivalent to a bfloat16, i.e., have 8 bits of exponent and 7 bits of mantissa.</p>
</section>
<section id="Query-cells-and-load-associated-embeddings">
<h2>Query cells and load associated embeddings<a class="headerlink" href="#Query-cells-and-load-associated-embeddings" title="Permalink to this heading">¬∂</a></h2>
<p>This section demonstrates two methods to query cells from the Census by <code class="docutils literal notranslate"><span class="pre">obs</span></code> metadata, and then fetch CELLxGENE-hosted embeddings associated with each cell.</p>
<ol class="arabic simple">
<li><p>Load an embedding into an AnnData <code class="docutils literal notranslate"><span class="pre">obsm</span></code> slot</p></li>
<li><p>Load an embedding into a dense NumPy array</p></li>
</ol>
<p>Let‚Äôs first do a few imports and utility functions used throughout this notebook.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># A few imports and utility functions used throughout this notebook</span>


<span class="kn">import</span> <span class="nn">warnings</span>

<span class="kn">import</span> <span class="nn">cellxgene_census</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">scanpy</span>
<span class="kn">import</span> <span class="nn">tiledbsoma</span> <span class="k">as</span> <span class="nn">soma</span>
<span class="kn">from</span> <span class="nn">cellxgene_census.experimental</span> <span class="kn">import</span> <span class="n">get_embedding_metadata</span>

<span class="n">warnings</span><span class="o">.</span><span class="n">filterwarnings</span><span class="p">(</span><span class="s2">&quot;ignore&quot;</span><span class="p">)</span>

<span class="c1"># The Census version to utilize</span>
<span class="n">CENSUS_VERSION</span> <span class="o">=</span> <span class="s2">&quot;2023-12-15&quot;</span>
<span class="n">EXPERIMENT_NAME</span> <span class="o">=</span> <span class="s2">&quot;homo_sapiens&quot;</span>
<span class="n">MEASUREMENT_NAME</span> <span class="o">=</span> <span class="s2">&quot;RNA&quot;</span>

<span class="c1"># The location of the embedding.  See &lt;https://cellxgene.cziscience.com/census-models&gt;</span>
<span class="c1"># for available CELLxGENE-hosted embeddings.</span>
<span class="n">EMBEDDING_URI</span> <span class="o">=</span> <span class="s2">&quot;s3://cellxgene-contrib-public/contrib/cell-census/soma/2023-12-15/CxG-contrib-1/&quot;</span>
</pre></div>
</div>
</div>
<section id="Load-an-embedding-into-an-AnnData-obsm-slot">
<h3>Load an embedding into an AnnData <code class="docutils literal notranslate"><span class="pre">obsm</span></code> slot<a class="headerlink" href="#Load-an-embedding-into-an-AnnData-obsm-slot" title="Permalink to this heading">¬∂</a></h3>
<p>There are two main ways to load hosted embeddings into an AnnData.</p>
<ol class="arabic simple">
<li><p>Via <code class="docutils literal notranslate"><span class="pre">cellxgene_census.get_anndata()</span></code>, followed by merging embeddings.</p></li>
<li><p>With a lazy query via <code class="docutils literal notranslate"><span class="pre">ExperimentAxisQuery</span></code>, followed by merging embeddings.</p></li>
</ol>
<section id="AnnData-embeddings-via-cellxgene_census.get_anndata()">
<h4>AnnData embeddings via cellxgene_census.get_anndata()<a class="headerlink" href="#AnnData-embeddings-via-cellxgene_census.get_anndata()" title="Permalink to this heading">¬∂</a></h4>
<p>This is the simplest way of getting the embeddings. In this example we create an AnnData for all ‚Äúcentral nervous system‚Äù cells, and use the <code class="docutils literal notranslate"><span class="pre">obs_embeddings</span></code> parameter to add scGPT embeddings to the <code class="docutils literal notranslate"><span class="pre">obsm</span></code> slot.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">with</span> <span class="n">cellxgene_census</span><span class="o">.</span><span class="n">open_soma</span><span class="p">(</span><span class="n">census_version</span><span class="o">=</span><span class="n">CENSUS_VERSION</span><span class="p">)</span> <span class="k">as</span> <span class="n">census</span><span class="p">:</span>
    <span class="n">adata</span> <span class="o">=</span> <span class="n">cellxgene_census</span><span class="o">.</span><span class="n">get_anndata</span><span class="p">(</span>
        <span class="n">census</span><span class="p">,</span>
        <span class="n">organism</span><span class="o">=</span><span class="n">EXPERIMENT_NAME</span><span class="p">,</span>
        <span class="n">measurement_name</span><span class="o">=</span><span class="n">MEASUREMENT_NAME</span><span class="p">,</span>
        <span class="n">obs_value_filter</span><span class="o">=</span><span class="s2">&quot;tissue_general == &#39;central nervous system&#39;&quot;</span><span class="p">,</span>
        <span class="n">column_names</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;obs&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;cell_type&quot;</span><span class="p">,</span> <span class="s2">&quot;soma_joinid&quot;</span><span class="p">]},</span>
        <span class="n">obs_embeddings</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;scgpt&quot;</span><span class="p">],</span>
    <span class="p">)</span>
</pre></div>
</div>
</div>
<p>‚ö†Ô∏è <strong>IMPORTANT</strong>: <code class="docutils literal notranslate"><span class="pre">get_embedding</span></code> will fill in missing values in the embedding matrix with <code class="docutils literal notranslate"><span class="pre">NaN</span></code>, therefore missing cells are represented with rows where all values are <code class="docutils literal notranslate"><span class="pre">NaN</span></code>. To learn more use <code class="docutils literal notranslate"><span class="pre">help(get_embedding)</span></code>.</p>
<p>Then we can take a quick look at the embeddings in a 2D scatter plot via UMAP.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">scanpy</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">neighbors</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">use_rep</span><span class="o">=</span><span class="s2">&quot;scgpt&quot;</span><span class="p">)</span>
<span class="n">scanpy</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
<span class="n">scanpy</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;cell_type&quot;</span><span class="p">,</span> <span class="n">title</span><span class="o">=</span><span class="s2">&quot;scGPT&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/notebooks_api_demo_census_embedding_15_0.png" src="../../_images/notebooks_api_demo_census_embedding_15_0.png" />
</div>
</div>
</section>
<section id="AnnData-embeddings-via-ExperimentAxisQuery">
<h4>AnnData embeddings via <code class="docutils literal notranslate"><span class="pre">ExperimentAxisQuery</span></code><a class="headerlink" href="#AnnData-embeddings-via-ExperimentAxisQuery" title="Permalink to this heading">¬∂</a></h4>
<p>Using an <code class="docutils literal notranslate"><span class="pre">ExperimentAxisQuery</span></code> to get embeddings into an AnnData has the main advantage of inspecting the query in a lazy manner before loading all data into AnnData.</p>
<p>As a reminder this class offers a lazy interface to query Census based on cell and gene metadata, and provides access to the correspondong expression data, and cell/gene metadata.</p>
<p>Let‚Äôs initiate a lazy query with the same filters as the previous example.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">census</span> <span class="o">=</span> <span class="n">cellxgene_census</span><span class="o">.</span><span class="n">open_soma</span><span class="p">(</span><span class="n">census_version</span><span class="o">=</span><span class="n">CENSUS_VERSION</span><span class="p">)</span>

<span class="n">experiment</span> <span class="o">=</span> <span class="n">census</span><span class="p">[</span><span class="s2">&quot;census_data&quot;</span><span class="p">][</span><span class="n">EXPERIMENT_NAME</span><span class="p">]</span>
<span class="n">query</span> <span class="o">=</span> <span class="n">experiment</span><span class="o">.</span><span class="n">axis_query</span><span class="p">(</span>
    <span class="n">measurement_name</span><span class="o">=</span><span class="s2">&quot;RNA&quot;</span><span class="p">,</span>
    <span class="n">obs_query</span><span class="o">=</span><span class="n">soma</span><span class="o">.</span><span class="n">AxisQuery</span><span class="p">(</span><span class="n">value_filter</span><span class="o">=</span><span class="s2">&quot;tissue_general == &#39;central nervous system&#39;&quot;</span><span class="p">),</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<p>Now, before downloading all the data we can take a look at different attributes, for example the number of cells in our query.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">query</span><span class="o">.</span><span class="n">n_obs</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
31780
</pre></div></div>
</div>
<p>And for example grab all the <code class="docutils literal notranslate"><span class="pre">soma_joinid</span></code> values for the cells in this query.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">soma_joinids</span> <span class="o">=</span> <span class="n">query</span><span class="o">.</span><span class="n">obs_joinids</span><span class="p">()</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">()</span>
</pre></div>
</div>
</div>
<p>Then create an AnnData.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span> <span class="o">=</span> <span class="n">query</span><span class="o">.</span><span class="n">to_anndata</span><span class="p">(</span><span class="n">X_name</span><span class="o">=</span><span class="s2">&quot;raw&quot;</span><span class="p">,</span> <span class="n">column_names</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;obs&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;cell_type&quot;</span><span class="p">]})</span>
</pre></div>
</div>
</div>
<p>And finally add the embedding matrix via <code class="docutils literal notranslate"><span class="pre">get_embedding</span></code> for the cells of the query using the <code class="docutils literal notranslate"><span class="pre">soma_joinid</span></code> values.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s2">&quot;scgpt&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">get_embedding</span><span class="p">(</span>
    <span class="n">census_version</span><span class="o">=</span><span class="n">CENSUS_VERSION</span><span class="p">,</span>
    <span class="n">embedding_uri</span><span class="o">=</span><span class="n">EMBEDDING_URI</span><span class="p">,</span>
    <span class="n">obs_soma_joinids</span><span class="o">=</span><span class="n">soma_joinids</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<p>‚ö†Ô∏è <strong>IMPORTANT</strong>: <code class="docutils literal notranslate"><span class="pre">get_embedding</span></code> will fill in missing values in the embedding matrix with <code class="docutils literal notranslate"><span class="pre">NaN</span></code>, therefore missing cells are represented with rows where all values are <code class="docutils literal notranslate"><span class="pre">NaN</span></code>. To learn more use <code class="docutils literal notranslate"><span class="pre">help(get_embedding)</span></code>.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[13]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[13]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
AnnData object with n_obs √ó n_vars = 31780 √ó 60664
    obs: &#39;cell_type&#39;, &#39;tissue_general&#39;
    var: &#39;soma_joinid&#39;, &#39;feature_id&#39;, &#39;feature_name&#39;, &#39;feature_length&#39;, &#39;nnz&#39;, &#39;n_measured_obs&#39;
    obsm: &#39;scgpt&#39;
</pre></div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">query</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
<span class="n">census</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</pre></div>
</div>
</div>
</section>
</section>
<section id="Load-an-embedding-into-a-dense-NumPy-array">
<h3>Load an embedding into a dense NumPy array<a class="headerlink" href="#Load-an-embedding-into-a-dense-NumPy-array" title="Permalink to this heading">¬∂</a></h3>
<p>To load a embeddinng into a stand-alone numpy array you can select cells from the Census based on obs metadata, then given the resulting cells, use the <code class="docutils literal notranslate"><span class="pre">soma_joinid</span></code> values to download an embedding, and finally save as a dense NDArray.</p>
<p>Let‚Äôs first select cells based on cell metadata</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[15]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">census</span> <span class="o">=</span> <span class="n">cellxgene_census</span><span class="o">.</span><span class="n">open_soma</span><span class="p">(</span><span class="n">census_version</span><span class="o">=</span><span class="n">CENSUS_VERSION</span><span class="p">)</span>

<span class="n">obs_df</span> <span class="o">=</span> <span class="n">census</span><span class="p">[</span><span class="s2">&quot;census_data&quot;</span><span class="p">][</span><span class="n">EXPERIMENT_NAME</span><span class="p">]</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">read</span><span class="p">(</span>
    <span class="n">value_filter</span><span class="o">=</span><span class="s2">&quot;tissue_general == &#39;exocrine gland&#39;&quot;</span><span class="p">,</span>
    <span class="n">column_names</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;soma_joinid&quot;</span><span class="p">,</span> <span class="s2">&quot;cell_type&quot;</span><span class="p">],</span>
<span class="p">)</span>

<span class="n">obs_df</span> <span class="o">=</span> <span class="n">obs_df</span><span class="o">.</span><span class="n">concat</span><span class="p">()</span><span class="o">.</span><span class="n">to_pandas</span><span class="p">()</span>
</pre></div>
</div>
</div>
<p>Now you can use the <code class="docutils literal notranslate"><span class="pre">soma_joinid</span></code> values to download the corresponding rows of the embedding matrix via <code class="docutils literal notranslate"><span class="pre">get_embedding</span></code>.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[16]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">embeddings</span> <span class="o">=</span> <span class="n">get_embedding</span><span class="p">(</span><span class="n">CENSUS_VERSION</span><span class="p">,</span> <span class="n">EMBEDDING_URI</span><span class="p">,</span> <span class="n">obs_df</span><span class="o">.</span><span class="n">soma_joinid</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">())</span>
<span class="n">embeddings</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="mi">3</span><span class="p">,</span> <span class="mi">0</span><span class="p">:</span><span class="mi">4</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[16]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
array([[-0.00506592, -0.01348877, -0.03173828, -0.02331543],
       [ 0.02404785, -0.02441406, -0.00595093, -0.0065918 ],
       [ 0.00070572,  0.00187683, -0.04663086, -0.04614258]],
      dtype=float32)
</pre></div></div>
</div>
<p>‚ö†Ô∏è <strong>IMPORTANT</strong>: <code class="docutils literal notranslate"><span class="pre">get_embedding</span></code> will fill in missing values in the embedding matrix with <code class="docutils literal notranslate"><span class="pre">NaN</span></code>, therefore missing cells are represented with rows where all values are <code class="docutils literal notranslate"><span class="pre">NaN</span></code>.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[17]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">embeddings</span><span class="o">.</span><span class="n">shape</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[17]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
(115722, 512)
</pre></div></div>
</div>
</section>
</section>
<section id="Load-embeddings-and-fetch-associated-Census-data">
<h2>Load embeddings and fetch associated Census data<a class="headerlink" href="#Load-embeddings-and-fetch-associated-Census-data" title="Permalink to this heading">¬∂</a></h2>
<p>This section describes a more advanced use case. Here we showcase how to load large slices of an embeding matrix, and then append cell metadata to them.</p>
<p>The method starts with the loaded embedding, and for each embedded cell loads metadata or X data.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[18]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Load a portion of the embedding (caution: embeddings can be quite large)</span>

<span class="kn">import</span> <span class="nn">cellxgene_census</span>
<span class="kn">import</span> <span class="nn">tiledbsoma</span> <span class="k">as</span> <span class="nn">soma</span>

<span class="c1"># Fetch first 500_000 joinids from the embedding.</span>
<span class="c1"># NOTE: will fail if the there are no cells embedded within this obs joinid range</span>
<span class="n">embedding_slice</span> <span class="o">=</span> <span class="p">(</span><span class="nb">slice</span><span class="p">(</span><span class="mi">500_000</span><span class="p">),)</span>

<span class="n">emb_data</span> <span class="o">=</span> <span class="p">[]</span>
<span class="n">emb_joinids</span> <span class="o">=</span> <span class="p">[]</span>
<span class="n">ctx</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;vfs.s3.region&quot;</span><span class="p">:</span> <span class="s2">&quot;us-west-2&quot;</span><span class="p">,</span> <span class="s2">&quot;vfs.s3.no_sign_request&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">}</span>
<span class="k">with</span> <span class="n">soma</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="n">EMBEDDING_URI</span><span class="p">,</span> <span class="n">context</span><span class="o">=</span><span class="n">soma</span><span class="o">.</span><span class="n">SOMATileDBContext</span><span class="p">(</span><span class="n">tiledb_config</span><span class="o">=</span><span class="n">ctx</span><span class="p">))</span> <span class="k">as</span> <span class="n">E</span><span class="p">:</span>
    <span class="c1"># read embedding and obs joinids for each embedded cell</span>
    <span class="k">for</span> <span class="n">d</span><span class="p">,</span> <span class="p">(</span><span class="n">obs_joinids</span><span class="p">,</span> <span class="n">_</span><span class="p">)</span> <span class="ow">in</span> <span class="p">(</span>
        <span class="n">E</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">coords</span><span class="o">=</span><span class="n">embedding_slice</span><span class="p">)</span><span class="o">.</span><span class="n">blockwise</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="mi">2</span><span class="o">**</span><span class="mi">20</span><span class="p">,</span> <span class="n">reindex_disable_on_axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">scipy</span><span class="p">()</span>
    <span class="p">):</span>
        <span class="n">embedding_presence_mask</span> <span class="o">=</span> <span class="n">d</span><span class="o">.</span><span class="n">getnnz</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span> <span class="o">!=</span> <span class="mi">0</span>
        <span class="n">emb_joinids</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">obs_joinids</span><span class="p">[</span><span class="n">embedding_presence_mask</span><span class="p">])</span>
        <span class="n">emb_data</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">d</span><span class="p">[</span><span class="n">embedding_presence_mask</span><span class="p">,</span> <span class="p">:]</span><span class="o">.</span><span class="n">toarray</span><span class="p">())</span>

    <span class="c1"># concat</span>
    <span class="n">embedding_data</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">vstack</span><span class="p">(</span><span class="n">emb_data</span><span class="p">)</span>
    <span class="n">embedding_joinids</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">concatenate</span><span class="p">(</span><span class="n">emb_joinids</span><span class="p">)</span>

<span class="c1"># Load the associated metadata - in this case, obs.suspension_type</span>
<span class="k">with</span> <span class="n">cellxgene_census</span><span class="o">.</span><span class="n">open_soma</span><span class="p">(</span><span class="n">census_version</span><span class="o">=</span><span class="n">CENSUS_VERSION</span><span class="p">)</span> <span class="k">as</span> <span class="n">census</span><span class="p">:</span>
    <span class="n">experiment</span> <span class="o">=</span> <span class="n">census</span><span class="p">[</span><span class="s2">&quot;census_data&quot;</span><span class="p">][</span><span class="n">EXPERIMENT_NAME</span><span class="p">]</span>
    <span class="k">with</span> <span class="n">experiment</span><span class="o">.</span><span class="n">axis_query</span><span class="p">(</span>
        <span class="n">measurement_name</span><span class="o">=</span><span class="n">MEASUREMENT_NAME</span><span class="p">,</span>
        <span class="n">obs_query</span><span class="o">=</span><span class="n">soma</span><span class="o">.</span><span class="n">AxisQuery</span><span class="p">(</span><span class="n">coords</span><span class="o">=</span><span class="p">(</span><span class="n">embedding_joinids</span><span class="p">,)),</span>
    <span class="p">)</span> <span class="k">as</span> <span class="n">query</span><span class="p">:</span>
        <span class="n">obs_df</span> <span class="o">=</span> <span class="n">query</span><span class="o">.</span><span class="n">obs</span><span class="p">(</span><span class="n">column_names</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;soma_joinid&quot;</span><span class="p">,</span> <span class="s2">&quot;suspension_type&quot;</span><span class="p">])</span><span class="o">.</span><span class="n">concat</span><span class="p">()</span><span class="o">.</span><span class="n">to_pandas</span><span class="p">()</span>

<span class="c1"># Display the first few cells, and the first few columns of their embeddings</span>
<span class="n">display</span><span class="p">(</span><span class="n">obs_df</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="mi">3</span><span class="p">])</span>
<span class="n">display</span><span class="p">(</span><span class="n">embedding_data</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="mi">3</span><span class="p">,</span> <span class="mi">0</span><span class="p">:</span><span class="mi">4</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>soma_joinid</th>
      <th>suspension_type</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>0</td>
      <td>nucleus</td>
    </tr>
    <tr>
      <th>1</th>
      <td>1</td>
      <td>nucleus</td>
    </tr>
    <tr>
      <th>2</th>
      <td>2</td>
      <td>nucleus</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
array([[-0.00762939, -0.00076675, -0.00047874, -0.03588867],
       [ 0.00405884, -0.00239563,  0.00982666, -0.00946045],
       [ 0.00473022,  0.0135498 , -0.01049805, -0.03051758]],
      dtype=float32)
</pre></div></div>
</div>
</section>
<section id="Embedding-Metadata">
<h2>Embedding Metadata<a class="headerlink" href="#Embedding-Metadata" title="Permalink to this heading">¬∂</a></h2>
<p>Each embedding contains descriptive information stored in the SOMA <code class="docutils literal notranslate"><span class="pre">metadata</span></code> slot, encoded as a JSON string. This metadata includes:</p>
<ul class="simple">
<li><p>census_version - the Census which is embedded. It is critical to confirm this matches the Census in use, or the embeddings will be meaningless.</p></li>
<li><p>experiment_name - the Census experiment embedded, e.g., <code class="docutils literal notranslate"><span class="pre">homo_sapiens</span></code> or <code class="docutils literal notranslate"><span class="pre">mus_musculus</span></code>.</p></li>
<li><p>measurement_name - the Census measurement embedded, e.g., <code class="docutils literal notranslate"><span class="pre">RNA</span></code></p></li>
</ul>
<p>There are a variety of other metadata values <a class="reference external" href="https://github.com/chanzuckerberg/cellxgene-census/blob/main/tools/census_contrib/embedding_metadata.md">documented here</a>.</p>
<p>The following example demonstrates how to access and decode the metadata into a Python dictionary.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[19]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">embedding_metadata</span> <span class="o">=</span> <span class="n">get_embedding_metadata</span><span class="p">(</span><span class="n">EMBEDDING_URI</span><span class="p">)</span>

<span class="n">embedding_metadata</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[19]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
{&#39;id&#39;: &#39;CxG-contrib-1&#39;,
 &#39;title&#39;: &#39;scGPT: Towards Building a Foundation Model for Single-Cell Multi-omics Using Generative AI&#39;,
 &#39;description&#39;: &#39;Utilizing the burgeoning single-cell sequencing data, we have pioneered the construction of a foundation model for single-cell biology, scGPT, which is based on generative pre-trained transformer across a repository of over 33 million cells. Our findings illustrate that scGPT, a generative pre-trained transformer, effectively distills critical biological insights concerning genes and cells. Through the further adaptation of transfer learning, scGPT can be optimized to achieve superior performance across diverse downstream applications.&#39;,
 &#39;primary_contact&#39;: {&#39;name&#39;: &#39;Bo Wang&#39;,
  &#39;email&#39;: &#39;bowang@vectorinstitute.ai&#39;,
  &#39;affiliation&#39;: &#39;Bo Wang Lab, University of Toronto&#39;},
 &#39;additional_contacts&#39;: [],
 &#39;DOI&#39;: &#39;10.1101/2023.04.30.538439&#39;,
 &#39;additional_information&#39;: &#39;&#39;,
 &#39;model_link&#39;: &#39;https://github.com/bowang-lab/scGPT&#39;,
 &#39;data_type&#39;: &#39;obs_embedding&#39;,
 &#39;census_version&#39;: &#39;2023-12-15&#39;,
 &#39;experiment_name&#39;: &#39;homo_sapiens&#39;,
 &#39;measurement_name&#39;: &#39;RNA&#39;,
 &#39;n_embeddings&#39;: 62998417,
 &#39;n_features&#39;: 512,
 &#39;submission_date&#39;: &#39;2023-11-09&#39;}
</pre></div></div>
</div>
<p>A common use case for accessing embedding metadata is the prevention of nonsense results, which will occur when the embeddings were not derived from the Census or experiment in use.</p>
<p>This demonstrates an easy way to grab the embedding metadata and confirm that it matches your expectations.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[20]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">embedding_metadata</span> <span class="o">=</span> <span class="n">get_embedding_metadata</span><span class="p">(</span><span class="n">EMBEDDING_URI</span><span class="p">)</span>

<span class="k">assert</span> <span class="n">embedding_metadata</span><span class="p">[</span><span class="s2">&quot;census_version&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="n">CENSUS_VERSION</span>
<span class="k">assert</span> <span class="n">embedding_metadata</span><span class="p">[</span><span class="s2">&quot;experiment_name&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="n">EXPERIMENT_NAME</span>
<span class="k">assert</span> <span class="n">embedding_metadata</span><span class="p">[</span><span class="s2">&quot;measurement_name&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="n">MEASUREMENT_NAME</span>

<span class="n">display</span><span class="p">(</span><span class="s2">&quot;all good!&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&#39;all good!&#39;
</pre></div></div>
</div>
</section>
</section>


            </div>
            
            </div>
            <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="census_access_maintained_embeddings.html" class="btn btn-neutral float-left" title="Access CELLxGENE collaboration embeddings (scVI, Geneformer)" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="../analysis_demo/comp_bio_scvi_model_use.html" class="btn btn-neutral float-right" title="scVI for cell type prediction and data projection" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2022, Chan Zuckerberg Initiative.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
            </div>
        </div>

        </section>

    </div>
    <!-- Newsletter Banner -->
    <div role="banner" id="newsletter-banner">
      <span id="newsletter-subscribe-button" role="button">Subscribe</span>&nbsp;to our newsletter to receive updates about new features.
      <div id="newsletter-banner-close-button" role="button">X</div>
    </div>
    <!-- Newsletter Modal -->
    <dialog id="newsletter-modal">
      <div id="newsletter-header">
        <img id="newsletter-logo" src="../../_static/img/cellxGene-newsletter-logo.svg" />
        <div id="newsletter-close-button" role="button">X</div>
      </div>
      <div id="newsletter-content">
        <div id="newsletter-callout">Join Our Newsletter</div>
        <div id="newsletter-description">Get a quarterly email with the latest CELLxGENE features and data.</div>
        <!-- HubSpot Form target -->
        <div id="newsletter-form-container"></div>
      </div>

      <div id="newsletter-footnote">Unsubscribe at any time.</div>
    </dialog>
  </div>
  

  <script>
    // (thuang): 30 days
    const NEWSLETTER_BANNER_DISMISSED_TTL_MS = 30 * 24 * 60 * 60 * 1000;
    const NEWSLETTER_BANNER_DISMISSED_KEY = "newsletterBannerDismissed"

    var script = document.createElement('script');
    script.src = 'https://js.hsforms.net/forms/v2.js';
    script.defer = true;
    document.head.appendChild(script);

    // Run the code once the script is loaded
    script.onload = async function() {
      await hbspt.forms.create({
        region: "na1",
        portalId: "7272273",
        formId: "eb65b811-0451-414d-8304-7b9b6f468ce5",
        target: '#newsletter-form-container',
        onFormReady() {
          // get element by type "email"
          const emailInput = document.querySelector('#email-eb65b811-0451-414d-8304-7b9b6f468ce5');
          emailInput.setAttribute('placeholder', 'Enter email address');

          // remove the label element for emailInput
          const emailLabel = document.querySelector('#label-email-eb65b811-0451-414d-8304-7b9b6f468ce5');
          emailLabel.remove();
        },
        submitText: 'Subscribe',
      });
    };

    checkNewsletterBanner();

    document.querySelector('#newsletter-banner-close-button').addEventListener('click', () => {
      document.querySelector('#newsletter-banner').remove();
      localStorage.setItem(NEWSLETTER_BANNER_DISMISSED_KEY, Date.now());
    });

    const modal = document.querySelector('#newsletter-modal');

    document.querySelector('#newsletter-subscribe-button').addEventListener('click', () => {
      modal.showModal();
    });

    document.querySelector('#newsletter-close-button').addEventListener('click', () => {
      modal.close();
    });

    function checkNewsletterBanner() {
      /**
       * (thuang): Use LocalStorage to store dismissed state for 30 days
       * NOTE: Currently Census doc page doesn't share the same domain as the main site,
       * so dismissing the banner on the main site won't dismiss it on the Census doc page.
       * And vice versa.
       */
      const newsletterBannerDismissed = localStorage.getItem('newsletterBannerDismissed');

      if (newsletterBannerDismissed) {
        return;
      }

      if (newsletterBannerDismissed && Date.now() - newsletterBannerDismissed > NEWSLETTER_BANNER_DISMISSED_TTL_MS) {
        localStorage.removeItem(NEWSLETTER_BANNER_DISMISSED_KEY);
      }

      const newsletterBanner = document.querySelector('#newsletter-banner');

      if (!newsletterBannerDismissed) {
        newsletterBanner.style.display = 'flex';
      }
    }
  </script>

  
  
    
   

</body>
</html>