{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Genes measured in each cell (dataset presence matrix)\n",
    "\n",
    "The Census is a compilation of cells from multiple datasets that may differ by the sets of genes they measure. This notebook describes the way to identify the genes measured per dataset.\n",
    "\n",
    "The presence matrix is a sparse boolean array, indicating which features (var) were present in each dataset.  The array has dimensions [n_datasets, n_var], and is stored in the SOMA Measurement `varp` collection. The first dimension is indexed by the `soma_joinid` in the `census_datasets` dataframe. The second is indexed by the `soma_joinid` in the `var` dataframe of the measurement.\n",
    "\n",
    "As a reminder the `obs` data frame has a column `dataset_id` that can be used to link any cell in the Census to the presence matrix.\n",
    "\n",
    "**Contents** \n",
    "\n",
    "1. Opening the Census.\n",
    "2. Fetching the IDs of the Census datasets.\n",
    "3. Fetching the dataset presence matrix.\n",
    "4. Identifying genes measured in a specific dataset.\n",
    "5. Identifying datasets that measured specific genes.\n",
    "6. Identifying all genes measured in a dataset.\n",
    "\n",
    "\n",
    "## Opening the Census\n",
    "\n",
    "The `cellxgene_census` python package contains a convenient API to open the latest version of the Census."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-04-05T14:16:01.185036Z",
     "iopub.status.busy": "2023-04-05T14:16:01.184392Z",
     "iopub.status.idle": "2023-04-05T14:16:03.294047Z",
     "shell.execute_reply": "2023-04-05T14:16:03.292703Z"
    }
   },
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "from scipy import sparse\n",
    "import cellxgene_census\n",
    "\n",
    "census = cellxgene_census.open_soma()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Fetching the IDs of the Census datasets\n",
    "\n",
    "Let's grab a table of all the datasets included in the Census and use this table in combination with the presence matrix below."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>soma_joinid</th>\n",
       "      <th>collection_id</th>\n",
       "      <th>collection_name</th>\n",
       "      <th>collection_doi</th>\n",
       "      <th>dataset_id</th>\n",
       "      <th>dataset_title</th>\n",
       "      <th>dataset_h5ad_path</th>\n",
       "      <th>dataset_total_cell_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>43d4bb39-21af-4d05-b973-4c1fed7b916c</td>\n",
       "      <td>Transcriptional Programming of Normal and Infl...</td>\n",
       "      <td>10.1016/j.celrep.2018.09.006</td>\n",
       "      <td>f512b8b6-369d-4a85-a695-116e0806857f</td>\n",
       "      <td>Skin</td>\n",
       "      <td>f512b8b6-369d-4a85-a695-116e0806857f.h5ad</td>\n",
       "      <td>68036</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>2b02dff7-e427-4cdc-96fb-c0f354c099aa</td>\n",
       "      <td>Single-Cell Analysis of Crohn’s Disease Lesion...</td>\n",
       "      <td>10.1016/j.cell.2019.08.008</td>\n",
       "      <td>36c867a7-be10-4e69-9b39-5de12b0af6da</td>\n",
       "      <td>Ileum</td>\n",
       "      <td>36c867a7-be10-4e69-9b39-5de12b0af6da.h5ad</td>\n",
       "      <td>32458</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>e9eec7f5-8519-42f6-99b4-6dbd9cc5ef03</td>\n",
       "      <td>Humoral immunity at the brain borders in homeo...</td>\n",
       "      <td>10.1016/j.coi.2022.102188</td>\n",
       "      <td>58b01044-c5e5-4b0f-8a2d-6ebf951e01ff</td>\n",
       "      <td>A scRNA-seq atlas of immune cells at the CNS b...</td>\n",
       "      <td>58b01044-c5e5-4b0f-8a2d-6ebf951e01ff.h5ad</td>\n",
       "      <td>130908</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>a72afd53-ab92-4511-88da-252fb0e26b9a</td>\n",
       "      <td>Single-cell atlas of peripheral immune respons...</td>\n",
       "      <td>10.1038/s41591-020-0944-y</td>\n",
       "      <td>456e8b9b-f872-488b-871d-94534090a865</td>\n",
       "      <td>Single-cell atlas of peripheral immune respons...</td>\n",
       "      <td>456e8b9b-f872-488b-871d-94534090a865.h5ad</td>\n",
       "      <td>44721</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>e4c9ed14-e560-4900-a3bf-b0f8d2ce6a10</td>\n",
       "      <td>A molecular single-cell lung atlas of lethal C...</td>\n",
       "      <td>10.1038/s41586-021-03569-1</td>\n",
       "      <td>d8da613f-e681-4c69-b463-e94f5e66847f</td>\n",
       "      <td>A molecular single-cell lung atlas of lethal C...</td>\n",
       "      <td>d8da613f-e681-4c69-b463-e94f5e66847f.h5ad</td>\n",
       "      <td>116313</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>524</th>\n",
       "      <td>524</td>\n",
       "      <td>e3f391f6-5a75-4e96-8450-da47c3d2a939</td>\n",
       "      <td>COVID-19 mRNA vaccine elicits a potent adaptiv...</td>\n",
       "      <td></td>\n",
       "      <td>30498543-4fdd-4f86-9e1b-05c1a1454a6a</td>\n",
       "      <td>B cells -- CV19 infection, vaccination and HC</td>\n",
       "      <td>30498543-4fdd-4f86-9e1b-05c1a1454a6a.h5ad</td>\n",
       "      <td>20727</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>525</th>\n",
       "      <td>525</td>\n",
       "      <td>e3f391f6-5a75-4e96-8450-da47c3d2a939</td>\n",
       "      <td>COVID-19 mRNA vaccine elicits a potent adaptiv...</td>\n",
       "      <td></td>\n",
       "      <td>b5191f01-f67d-44b8-bc8d-511a4ecd07bb</td>\n",
       "      <td>innate T cells -- CV19 infection, vaccination ...</td>\n",
       "      <td>b5191f01-f67d-44b8-bc8d-511a4ecd07bb.h5ad</td>\n",
       "      <td>33415</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>526</th>\n",
       "      <td>526</td>\n",
       "      <td>e3f391f6-5a75-4e96-8450-da47c3d2a939</td>\n",
       "      <td>COVID-19 mRNA vaccine elicits a potent adaptiv...</td>\n",
       "      <td></td>\n",
       "      <td>e463dae9-3fc1-476d-870e-d98a04c56cd6</td>\n",
       "      <td>M cells -- CV19 infection, vaccination and HC</td>\n",
       "      <td>e463dae9-3fc1-476d-870e-d98a04c56cd6.h5ad</td>\n",
       "      <td>41130</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>527</th>\n",
       "      <td>527</td>\n",
       "      <td>e3f391f6-5a75-4e96-8450-da47c3d2a939</td>\n",
       "      <td>COVID-19 mRNA vaccine elicits a potent adaptiv...</td>\n",
       "      <td></td>\n",
       "      <td>1b699e04-1127-42ea-998b-011ace4a5b81</td>\n",
       "      <td>T cells -- CV19 infection, vaccination and HC</td>\n",
       "      <td>1b699e04-1127-42ea-998b-011ace4a5b81.h5ad</td>\n",
       "      <td>98068</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>528</th>\n",
       "      <td>528</td>\n",
       "      <td>e3f391f6-5a75-4e96-8450-da47c3d2a939</td>\n",
       "      <td>COVID-19 mRNA vaccine elicits a potent adaptiv...</td>\n",
       "      <td></td>\n",
       "      <td>87ce26ed-e5d1-44b4-81cc-cc5b709a169f</td>\n",
       "      <td>Circulating Immune cells -- CV19 infection, va...</td>\n",
       "      <td>87ce26ed-e5d1-44b4-81cc-cc5b709a169f.h5ad</td>\n",
       "      <td>195632</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>529 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     soma_joinid                         collection_id   \n",
       "0              0  43d4bb39-21af-4d05-b973-4c1fed7b916c  \\\n",
       "1              1  2b02dff7-e427-4cdc-96fb-c0f354c099aa   \n",
       "2              2  e9eec7f5-8519-42f6-99b4-6dbd9cc5ef03   \n",
       "3              3  a72afd53-ab92-4511-88da-252fb0e26b9a   \n",
       "4              4  e4c9ed14-e560-4900-a3bf-b0f8d2ce6a10   \n",
       "..           ...                                   ...   \n",
       "524          524  e3f391f6-5a75-4e96-8450-da47c3d2a939   \n",
       "525          525  e3f391f6-5a75-4e96-8450-da47c3d2a939   \n",
       "526          526  e3f391f6-5a75-4e96-8450-da47c3d2a939   \n",
       "527          527  e3f391f6-5a75-4e96-8450-da47c3d2a939   \n",
       "528          528  e3f391f6-5a75-4e96-8450-da47c3d2a939   \n",
       "\n",
       "                                       collection_name   \n",
       "0    Transcriptional Programming of Normal and Infl...  \\\n",
       "1    Single-Cell Analysis of Crohn’s Disease Lesion...   \n",
       "2    Humoral immunity at the brain borders in homeo...   \n",
       "3    Single-cell atlas of peripheral immune respons...   \n",
       "4    A molecular single-cell lung atlas of lethal C...   \n",
       "..                                                 ...   \n",
       "524  COVID-19 mRNA vaccine elicits a potent adaptiv...   \n",
       "525  COVID-19 mRNA vaccine elicits a potent adaptiv...   \n",
       "526  COVID-19 mRNA vaccine elicits a potent adaptiv...   \n",
       "527  COVID-19 mRNA vaccine elicits a potent adaptiv...   \n",
       "528  COVID-19 mRNA vaccine elicits a potent adaptiv...   \n",
       "\n",
       "                   collection_doi                            dataset_id   \n",
       "0    10.1016/j.celrep.2018.09.006  f512b8b6-369d-4a85-a695-116e0806857f  \\\n",
       "1      10.1016/j.cell.2019.08.008  36c867a7-be10-4e69-9b39-5de12b0af6da   \n",
       "2       10.1016/j.coi.2022.102188  58b01044-c5e5-4b0f-8a2d-6ebf951e01ff   \n",
       "3       10.1038/s41591-020-0944-y  456e8b9b-f872-488b-871d-94534090a865   \n",
       "4      10.1038/s41586-021-03569-1  d8da613f-e681-4c69-b463-e94f5e66847f   \n",
       "..                            ...                                   ...   \n",
       "524                                30498543-4fdd-4f86-9e1b-05c1a1454a6a   \n",
       "525                                b5191f01-f67d-44b8-bc8d-511a4ecd07bb   \n",
       "526                                e463dae9-3fc1-476d-870e-d98a04c56cd6   \n",
       "527                                1b699e04-1127-42ea-998b-011ace4a5b81   \n",
       "528                                87ce26ed-e5d1-44b4-81cc-cc5b709a169f   \n",
       "\n",
       "                                         dataset_title   \n",
       "0                                                 Skin  \\\n",
       "1                                                Ileum   \n",
       "2    A scRNA-seq atlas of immune cells at the CNS b...   \n",
       "3    Single-cell atlas of peripheral immune respons...   \n",
       "4    A molecular single-cell lung atlas of lethal C...   \n",
       "..                                                 ...   \n",
       "524      B cells -- CV19 infection, vaccination and HC   \n",
       "525  innate T cells -- CV19 infection, vaccination ...   \n",
       "526      M cells -- CV19 infection, vaccination and HC   \n",
       "527      T cells -- CV19 infection, vaccination and HC   \n",
       "528  Circulating Immune cells -- CV19 infection, va...   \n",
       "\n",
       "                             dataset_h5ad_path  dataset_total_cell_count  \n",
       "0    f512b8b6-369d-4a85-a695-116e0806857f.h5ad                     68036  \n",
       "1    36c867a7-be10-4e69-9b39-5de12b0af6da.h5ad                     32458  \n",
       "2    58b01044-c5e5-4b0f-8a2d-6ebf951e01ff.h5ad                    130908  \n",
       "3    456e8b9b-f872-488b-871d-94534090a865.h5ad                     44721  \n",
       "4    d8da613f-e681-4c69-b463-e94f5e66847f.h5ad                    116313  \n",
       "..                                         ...                       ...  \n",
       "524  30498543-4fdd-4f86-9e1b-05c1a1454a6a.h5ad                     20727  \n",
       "525  b5191f01-f67d-44b8-bc8d-511a4ecd07bb.h5ad                     33415  \n",
       "526  e463dae9-3fc1-476d-870e-d98a04c56cd6.h5ad                     41130  \n",
       "527  1b699e04-1127-42ea-998b-011ace4a5b81.h5ad                     98068  \n",
       "528  87ce26ed-e5d1-44b4-81cc-cc5b709a169f.h5ad                    195632  \n",
       "\n",
       "[529 rows x 8 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Grab the experiment containing human data, and the measurement therein with RNA\n",
    "human = census[\"census_data\"][\"homo_sapiens\"]\n",
    "human_rna = human.ms[\"RNA\"]\n",
    "\n",
    "# The census-wide datasets\n",
    "datasets_df = census[\"census_info\"][\"datasets\"].read().concat().to_pandas()\n",
    "\n",
    "datasets_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Fetching the dataset presence matrix\n",
    "\n",
    "Now let's fetch the dataset presence matrix. \n",
    "\n",
    "For convenience, read the entire presence matrix (for Homo sapiens) into a SciPy array. There is a convenience API providing this capability, returning the matrix in a `scipy.sparse.array`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-04-05T14:16:03.297696Z",
     "iopub.status.busy": "2023-04-05T14:16:03.297454Z",
     "iopub.status.idle": "2023-04-05T14:16:04.967669Z",
     "shell.execute_reply": "2023-04-05T14:16:04.966701Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<529x60664 sparse matrix of type '<class 'numpy.uint8'>'\n",
       "\twith 14006372 stored elements in Compressed Sparse Row format>"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "presence_matrix = cellxgene_census.get_presence_matrix(\n",
    "    census, organism=\"Homo sapiens\", measurement_name=\"RNA\"\n",
    ")\n",
    "\n",
    "presence_matrix"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We also need the `var` dataframe, which is read into a Pandas DataFrame for convenient manipulation:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-04-05T14:16:04.971059Z",
     "iopub.status.busy": "2023-04-05T14:16:04.970894Z",
     "iopub.status.idle": "2023-04-05T14:16:05.509255Z",
     "shell.execute_reply": "2023-04-05T14:16:05.508560Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>soma_joinid</th>\n",
       "      <th>feature_id</th>\n",
       "      <th>feature_name</th>\n",
       "      <th>feature_length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000238009</td>\n",
       "      <td>RP11-34P13.7</td>\n",
       "      <td>3726</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000279457</td>\n",
       "      <td>WASH9P</td>\n",
       "      <td>1397</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000228463</td>\n",
       "      <td>AP006222.1</td>\n",
       "      <td>8224</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>ENSG00000237094</td>\n",
       "      <td>RP4-669L17.4</td>\n",
       "      <td>6204</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>ENSG00000230021</td>\n",
       "      <td>RP11-206L10.17</td>\n",
       "      <td>5495</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60659</th>\n",
       "      <td>60659</td>\n",
       "      <td>ENSG00000288719</td>\n",
       "      <td>RP4-669P10.21</td>\n",
       "      <td>4252</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60660</th>\n",
       "      <td>60660</td>\n",
       "      <td>ENSG00000288720</td>\n",
       "      <td>RP11-852E15.3</td>\n",
       "      <td>7007</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60661</th>\n",
       "      <td>60661</td>\n",
       "      <td>ENSG00000288721</td>\n",
       "      <td>RP5-973N23.5</td>\n",
       "      <td>7765</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60662</th>\n",
       "      <td>60662</td>\n",
       "      <td>ENSG00000288723</td>\n",
       "      <td>RP11-553N16.6</td>\n",
       "      <td>1015</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60663</th>\n",
       "      <td>60663</td>\n",
       "      <td>ENSG00000288724</td>\n",
       "      <td>RP13-546I2.2</td>\n",
       "      <td>625</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>60664 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       soma_joinid       feature_id    feature_name  feature_length\n",
       "0                0  ENSG00000238009    RP11-34P13.7            3726\n",
       "1                1  ENSG00000279457          WASH9P            1397\n",
       "2                2  ENSG00000228463      AP006222.1            8224\n",
       "3                3  ENSG00000237094    RP4-669L17.4            6204\n",
       "4                4  ENSG00000230021  RP11-206L10.17            5495\n",
       "...            ...              ...             ...             ...\n",
       "60659        60659  ENSG00000288719   RP4-669P10.21            4252\n",
       "60660        60660  ENSG00000288720   RP11-852E15.3            7007\n",
       "60661        60661  ENSG00000288721    RP5-973N23.5            7765\n",
       "60662        60662  ENSG00000288723   RP11-553N16.6            1015\n",
       "60663        60663  ENSG00000288724    RP13-546I2.2             625\n",
       "\n",
       "[60664 rows x 4 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "var_df = human_rna.var.read().concat().to_pandas()\n",
    "\n",
    "var_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Identifying genes measured in a specific dataset.\n",
    "\n",
    "Now that we have the dataset table, the genes metadata table, and the dataset presence matrix, we can check if a gene or set of genes were measured in a specific dataset.\n",
    "\n",
    "**Important:** the presence matrix is indexed by soma_joinid, and is *NOT* positionally indexed.  In other words:\n",
    "* the first dimension of the presence matrix is the dataset's `soma_joinid`, as stored in the `census_datasets` dataframe.\n",
    "* the second dimension of the presence matrix is the feature's `soma_joinid`, as stored in the `var` dataframe.\n",
    "\n",
    "Let's find out if the the gene `\"ENSG00000286096\"` was measured in the dataset with id `\"97a17473-e2b1-4f31-a544-44a60773e2dd\"`.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-04-05T14:16:05.511908Z",
     "iopub.status.busy": "2023-04-05T14:16:05.511755Z",
     "iopub.status.idle": "2023-04-05T14:16:05.519610Z",
     "shell.execute_reply": "2023-04-05T14:16:05.518587Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Feature is present.\n"
     ]
    }
   ],
   "source": [
    "var_joinid = var_df.loc[var_df.feature_id == \"ENSG00000286096\"].soma_joinid\n",
    "dataset_joinid = datasets_df.loc[\n",
    "    datasets_df.dataset_id == \"97a17473-e2b1-4f31-a544-44a60773e2dd\"\n",
    "].soma_joinid\n",
    "is_present = presence_matrix[dataset_joinid, var_joinid][0, 0]\n",
    "print(f'Feature is {\"present\" if is_present else \"not present\"}.')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Identifying datasets that measured specific genes\n",
    "\n",
    "Similarly, we can determine the datasets that measured a specific gene or set of genes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-04-05T14:16:05.523101Z",
     "iopub.status.busy": "2023-04-05T14:16:05.522880Z",
     "iopub.status.idle": "2023-04-05T14:16:05.552412Z",
     "shell.execute_reply": "2023-04-05T14:16:05.551100Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>soma_joinid</th>\n",
       "      <th>collection_id</th>\n",
       "      <th>collection_name</th>\n",
       "      <th>collection_doi</th>\n",
       "      <th>dataset_id</th>\n",
       "      <th>dataset_title</th>\n",
       "      <th>dataset_h5ad_path</th>\n",
       "      <th>dataset_total_cell_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>335</th>\n",
       "      <td>335</td>\n",
       "      <td>e5f58829-1a66-40b5-a624-9046778e74f5</td>\n",
       "      <td>Tabula Sapiens</td>\n",
       "      <td>10.1126/science.abl4896</td>\n",
       "      <td>a68b64d8-aee3-4947-81b7-36b8fe5a44d2</td>\n",
       "      <td>Tabula Sapiens - Stromal</td>\n",
       "      <td>a68b64d8-aee3-4947-81b7-36b8fe5a44d2.h5ad</td>\n",
       "      <td>82478</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>336</th>\n",
       "      <td>336</td>\n",
       "      <td>e5f58829-1a66-40b5-a624-9046778e74f5</td>\n",
       "      <td>Tabula Sapiens</td>\n",
       "      <td>10.1126/science.abl4896</td>\n",
       "      <td>97a17473-e2b1-4f31-a544-44a60773e2dd</td>\n",
       "      <td>Tabula Sapiens - Epithelial</td>\n",
       "      <td>97a17473-e2b1-4f31-a544-44a60773e2dd.h5ad</td>\n",
       "      <td>104148</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>337</th>\n",
       "      <td>337</td>\n",
       "      <td>e5f58829-1a66-40b5-a624-9046778e74f5</td>\n",
       "      <td>Tabula Sapiens</td>\n",
       "      <td>10.1126/science.abl4896</td>\n",
       "      <td>c5d88abe-f23a-45fa-a534-788985e93dad</td>\n",
       "      <td>Tabula Sapiens - Immune</td>\n",
       "      <td>c5d88abe-f23a-45fa-a534-788985e93dad.h5ad</td>\n",
       "      <td>264824</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>338</th>\n",
       "      <td>338</td>\n",
       "      <td>e5f58829-1a66-40b5-a624-9046778e74f5</td>\n",
       "      <td>Tabula Sapiens</td>\n",
       "      <td>10.1126/science.abl4896</td>\n",
       "      <td>5a11f879-d1ef-458a-910c-9b0bdfca5ebf</td>\n",
       "      <td>Tabula Sapiens - Endothelial</td>\n",
       "      <td>5a11f879-d1ef-458a-910c-9b0bdfca5ebf.h5ad</td>\n",
       "      <td>31691</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>339</th>\n",
       "      <td>339</td>\n",
       "      <td>e5f58829-1a66-40b5-a624-9046778e74f5</td>\n",
       "      <td>Tabula Sapiens</td>\n",
       "      <td>10.1126/science.abl4896</td>\n",
       "      <td>53d208b0-2cfd-4366-9866-c3c6114081bc</td>\n",
       "      <td>Tabula Sapiens - All Cells</td>\n",
       "      <td>53d208b0-2cfd-4366-9866-c3c6114081bc.h5ad</td>\n",
       "      <td>483152</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>358</th>\n",
       "      <td>358</td>\n",
       "      <td>283d65eb-dd53-496d-adb7-7570c7caa443</td>\n",
       "      <td>Transcriptomic diversity of cell types across ...</td>\n",
       "      <td>10.1101/2022.10.12.511898</td>\n",
       "      <td>07b1d7c8-5c2e-42f7-9246-26f746cd6013</td>\n",
       "      <td>Dissection: Myelencephalon (medulla oblongata)...</td>\n",
       "      <td>07b1d7c8-5c2e-42f7-9246-26f746cd6013.h5ad</td>\n",
       "      <td>27210</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>371</th>\n",
       "      <td>371</td>\n",
       "      <td>283d65eb-dd53-496d-adb7-7570c7caa443</td>\n",
       "      <td>Transcriptomic diversity of cell types across ...</td>\n",
       "      <td>10.1101/2022.10.12.511898</td>\n",
       "      <td>7c1c3d47-3166-43e5-9a95-65ceb2d45f78</td>\n",
       "      <td>Dissection: Pons (Pn) - Pontine reticular form...</td>\n",
       "      <td>7c1c3d47-3166-43e5-9a95-65ceb2d45f78.h5ad</td>\n",
       "      <td>49512</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>372</th>\n",
       "      <td>372</td>\n",
       "      <td>283d65eb-dd53-496d-adb7-7570c7caa443</td>\n",
       "      <td>Transcriptomic diversity of cell types across ...</td>\n",
       "      <td>10.1101/2022.10.12.511898</td>\n",
       "      <td>9372df2d-13d6-4fac-980b-919a5b7eb483</td>\n",
       "      <td>Dissection: Midbrain (M) - Periaqueductal gray...</td>\n",
       "      <td>9372df2d-13d6-4fac-980b-919a5b7eb483.h5ad</td>\n",
       "      <td>33794</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>399</th>\n",
       "      <td>399</td>\n",
       "      <td>283d65eb-dd53-496d-adb7-7570c7caa443</td>\n",
       "      <td>Transcriptomic diversity of cell types across ...</td>\n",
       "      <td>10.1101/2022.10.12.511898</td>\n",
       "      <td>dd03ce70-3243-4c96-9561-330cc461e4d7</td>\n",
       "      <td>Dissection: Cerebral cortex (Cx) - Perirhinal ...</td>\n",
       "      <td>dd03ce70-3243-4c96-9561-330cc461e4d7.h5ad</td>\n",
       "      <td>23732</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>413</th>\n",
       "      <td>413</td>\n",
       "      <td>283d65eb-dd53-496d-adb7-7570c7caa443</td>\n",
       "      <td>Transcriptomic diversity of cell types across ...</td>\n",
       "      <td>10.1101/2022.10.12.511898</td>\n",
       "      <td>7a0a8891-9a22-4549-a55b-c2aca23c3a2a</td>\n",
       "      <td>Supercluster: Hippocampal CA1-3</td>\n",
       "      <td>7a0a8891-9a22-4549-a55b-c2aca23c3a2a.h5ad</td>\n",
       "      <td>74979</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>415</th>\n",
       "      <td>415</td>\n",
       "      <td>283d65eb-dd53-496d-adb7-7570c7caa443</td>\n",
       "      <td>Transcriptomic diversity of cell types across ...</td>\n",
       "      <td>10.1101/2022.10.12.511898</td>\n",
       "      <td>d2b5efc1-14c6-4b5f-bd98-40f9084872d7</td>\n",
       "      <td>Dissection: Tail of Hippocampus (HiT) - Caudal...</td>\n",
       "      <td>d2b5efc1-14c6-4b5f-bd98-40f9084872d7.h5ad</td>\n",
       "      <td>36886</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>419</th>\n",
       "      <td>419</td>\n",
       "      <td>283d65eb-dd53-496d-adb7-7570c7caa443</td>\n",
       "      <td>Transcriptomic diversity of cell types across ...</td>\n",
       "      <td>10.1101/2022.10.12.511898</td>\n",
       "      <td>f8dda921-5fb4-4c94-a654-c6fc346bfd6d</td>\n",
       "      <td>Dissection: Cerebral cortex (Cx) - Occipitotem...</td>\n",
       "      <td>f8dda921-5fb4-4c94-a654-c6fc346bfd6d.h5ad</td>\n",
       "      <td>31899</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>422</th>\n",
       "      <td>422</td>\n",
       "      <td>283d65eb-dd53-496d-adb7-7570c7caa443</td>\n",
       "      <td>Transcriptomic diversity of cell types across ...</td>\n",
       "      <td>10.1101/2022.10.12.511898</td>\n",
       "      <td>3a7f3ab4-a280-4b3b-b2c0-6dd05614a78c</td>\n",
       "      <td>Supercluster: Splatter</td>\n",
       "      <td>3a7f3ab4-a280-4b3b-b2c0-6dd05614a78c.h5ad</td>\n",
       "      <td>291833</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>424</th>\n",
       "      <td>424</td>\n",
       "      <td>283d65eb-dd53-496d-adb7-7570c7caa443</td>\n",
       "      <td>Transcriptomic diversity of cell types across ...</td>\n",
       "      <td>10.1101/2022.10.12.511898</td>\n",
       "      <td>bdb26abd-f4ba-4ea3-8862-c2340e7a4f55</td>\n",
       "      <td>Supercluster: CGE interneuron</td>\n",
       "      <td>bdb26abd-f4ba-4ea3-8862-c2340e7a4f55.h5ad</td>\n",
       "      <td>227671</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>426</th>\n",
       "      <td>426</td>\n",
       "      <td>283d65eb-dd53-496d-adb7-7570c7caa443</td>\n",
       "      <td>Transcriptomic diversity of cell types across ...</td>\n",
       "      <td>10.1101/2022.10.12.511898</td>\n",
       "      <td>5e5ab909-f73f-4b57-98a0-6d2c5662f6a4</td>\n",
       "      <td>Dissection: Midbrain (M) - Inferior colliculus...</td>\n",
       "      <td>5e5ab909-f73f-4b57-98a0-6d2c5662f6a4.h5ad</td>\n",
       "      <td>32306</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>434</th>\n",
       "      <td>434</td>\n",
       "      <td>283d65eb-dd53-496d-adb7-7570c7caa443</td>\n",
       "      <td>Transcriptomic diversity of cell types across ...</td>\n",
       "      <td>10.1101/2022.10.12.511898</td>\n",
       "      <td>35c8a04c-8639-4d15-8228-765d8d93fc96</td>\n",
       "      <td>Dissection: Hypothalamus (HTH) - supraoptic re...</td>\n",
       "      <td>35c8a04c-8639-4d15-8228-765d8d93fc96.h5ad</td>\n",
       "      <td>16753</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>436</th>\n",
       "      <td>436</td>\n",
       "      <td>283d65eb-dd53-496d-adb7-7570c7caa443</td>\n",
       "      <td>Transcriptomic diversity of cell types across ...</td>\n",
       "      <td>10.1101/2022.10.12.511898</td>\n",
       "      <td>3f56901c-dd4a-47d6-b60b-7b0c0111cfb2</td>\n",
       "      <td>Dissection: Head of hippocampus (HiH) - CA1-3</td>\n",
       "      <td>3f56901c-dd4a-47d6-b60b-7b0c0111cfb2.h5ad</td>\n",
       "      <td>37911</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>452</th>\n",
       "      <td>452</td>\n",
       "      <td>283d65eb-dd53-496d-adb7-7570c7caa443</td>\n",
       "      <td>Transcriptomic diversity of cell types across ...</td>\n",
       "      <td>10.1101/2022.10.12.511898</td>\n",
       "      <td>93131426-0124-4ab4-a013-9dfbcd99d467</td>\n",
       "      <td>Dissection: Epithalamus - ETH</td>\n",
       "      <td>93131426-0124-4ab4-a013-9dfbcd99d467.h5ad</td>\n",
       "      <td>24327</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>458</th>\n",
       "      <td>458</td>\n",
       "      <td>283d65eb-dd53-496d-adb7-7570c7caa443</td>\n",
       "      <td>Transcriptomic diversity of cell types across ...</td>\n",
       "      <td>10.1101/2022.10.12.511898</td>\n",
       "      <td>acae7679-d077-461c-b857-ee6ccfeb267f</td>\n",
       "      <td>Dissection: Head of hippocampus (HiH) - CA1</td>\n",
       "      <td>acae7679-d077-461c-b857-ee6ccfeb267f.h5ad</td>\n",
       "      <td>39147</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>460</th>\n",
       "      <td>460</td>\n",
       "      <td>283d65eb-dd53-496d-adb7-7570c7caa443</td>\n",
       "      <td>Transcriptomic diversity of cell types across ...</td>\n",
       "      <td>10.1101/2022.10.12.511898</td>\n",
       "      <td>c202b243-1aa1-4b16-bc9a-b36241f3b1e3</td>\n",
       "      <td>Supercluster: Amygdala excitatory</td>\n",
       "      <td>c202b243-1aa1-4b16-bc9a-b36241f3b1e3.h5ad</td>\n",
       "      <td>109452</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>462</th>\n",
       "      <td>462</td>\n",
       "      <td>283d65eb-dd53-496d-adb7-7570c7caa443</td>\n",
       "      <td>Transcriptomic diversity of cell types across ...</td>\n",
       "      <td>10.1101/2022.10.12.511898</td>\n",
       "      <td>0325478a-9b52-45b5-b40a-2e2ab0d72eb1</td>\n",
       "      <td>Supercluster: Upper-layer intratelencephalic</td>\n",
       "      <td>0325478a-9b52-45b5-b40a-2e2ab0d72eb1.h5ad</td>\n",
       "      <td>455006</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>464</th>\n",
       "      <td>464</td>\n",
       "      <td>283d65eb-dd53-496d-adb7-7570c7caa443</td>\n",
       "      <td>Transcriptomic diversity of cell types across ...</td>\n",
       "      <td>10.1101/2022.10.12.511898</td>\n",
       "      <td>c2aad8fc-b63b-4f9b-9cfd-baf7bc9c1771</td>\n",
       "      <td>Dissection: Cerebral cortex (Cx) - Temporal po...</td>\n",
       "      <td>c2aad8fc-b63b-4f9b-9cfd-baf7bc9c1771.h5ad</td>\n",
       "      <td>37642</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>465</th>\n",
       "      <td>465</td>\n",
       "      <td>283d65eb-dd53-496d-adb7-7570c7caa443</td>\n",
       "      <td>Transcriptomic diversity of cell types across ...</td>\n",
       "      <td>10.1101/2022.10.12.511898</td>\n",
       "      <td>c4b03352-af8d-492a-8d6b-40f304e0a122</td>\n",
       "      <td>Supercluster: Medium spiny neuron</td>\n",
       "      <td>c4b03352-af8d-492a-8d6b-40f304e0a122.h5ad</td>\n",
       "      <td>152189</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>482</th>\n",
       "      <td>482</td>\n",
       "      <td>283d65eb-dd53-496d-adb7-7570c7caa443</td>\n",
       "      <td>Transcriptomic diversity of cell types across ...</td>\n",
       "      <td>10.1101/2022.10.12.511898</td>\n",
       "      <td>fe1a73ab-a203-45fd-84e9-0f7fd19efcbd</td>\n",
       "      <td>Dissection: Amygdaloid complex (AMY) - basolat...</td>\n",
       "      <td>fe1a73ab-a203-45fd-84e9-0f7fd19efcbd.h5ad</td>\n",
       "      <td>35285</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     soma_joinid                         collection_id   \n",
       "335          335  e5f58829-1a66-40b5-a624-9046778e74f5  \\\n",
       "336          336  e5f58829-1a66-40b5-a624-9046778e74f5   \n",
       "337          337  e5f58829-1a66-40b5-a624-9046778e74f5   \n",
       "338          338  e5f58829-1a66-40b5-a624-9046778e74f5   \n",
       "339          339  e5f58829-1a66-40b5-a624-9046778e74f5   \n",
       "358          358  283d65eb-dd53-496d-adb7-7570c7caa443   \n",
       "371          371  283d65eb-dd53-496d-adb7-7570c7caa443   \n",
       "372          372  283d65eb-dd53-496d-adb7-7570c7caa443   \n",
       "399          399  283d65eb-dd53-496d-adb7-7570c7caa443   \n",
       "413          413  283d65eb-dd53-496d-adb7-7570c7caa443   \n",
       "415          415  283d65eb-dd53-496d-adb7-7570c7caa443   \n",
       "419          419  283d65eb-dd53-496d-adb7-7570c7caa443   \n",
       "422          422  283d65eb-dd53-496d-adb7-7570c7caa443   \n",
       "424          424  283d65eb-dd53-496d-adb7-7570c7caa443   \n",
       "426          426  283d65eb-dd53-496d-adb7-7570c7caa443   \n",
       "434          434  283d65eb-dd53-496d-adb7-7570c7caa443   \n",
       "436          436  283d65eb-dd53-496d-adb7-7570c7caa443   \n",
       "452          452  283d65eb-dd53-496d-adb7-7570c7caa443   \n",
       "458          458  283d65eb-dd53-496d-adb7-7570c7caa443   \n",
       "460          460  283d65eb-dd53-496d-adb7-7570c7caa443   \n",
       "462          462  283d65eb-dd53-496d-adb7-7570c7caa443   \n",
       "464          464  283d65eb-dd53-496d-adb7-7570c7caa443   \n",
       "465          465  283d65eb-dd53-496d-adb7-7570c7caa443   \n",
       "482          482  283d65eb-dd53-496d-adb7-7570c7caa443   \n",
       "\n",
       "                                       collection_name   \n",
       "335                                     Tabula Sapiens  \\\n",
       "336                                     Tabula Sapiens   \n",
       "337                                     Tabula Sapiens   \n",
       "338                                     Tabula Sapiens   \n",
       "339                                     Tabula Sapiens   \n",
       "358  Transcriptomic diversity of cell types across ...   \n",
       "371  Transcriptomic diversity of cell types across ...   \n",
       "372  Transcriptomic diversity of cell types across ...   \n",
       "399  Transcriptomic diversity of cell types across ...   \n",
       "413  Transcriptomic diversity of cell types across ...   \n",
       "415  Transcriptomic diversity of cell types across ...   \n",
       "419  Transcriptomic diversity of cell types across ...   \n",
       "422  Transcriptomic diversity of cell types across ...   \n",
       "424  Transcriptomic diversity of cell types across ...   \n",
       "426  Transcriptomic diversity of cell types across ...   \n",
       "434  Transcriptomic diversity of cell types across ...   \n",
       "436  Transcriptomic diversity of cell types across ...   \n",
       "452  Transcriptomic diversity of cell types across ...   \n",
       "458  Transcriptomic diversity of cell types across ...   \n",
       "460  Transcriptomic diversity of cell types across ...   \n",
       "462  Transcriptomic diversity of cell types across ...   \n",
       "464  Transcriptomic diversity of cell types across ...   \n",
       "465  Transcriptomic diversity of cell types across ...   \n",
       "482  Transcriptomic diversity of cell types across ...   \n",
       "\n",
       "                collection_doi                            dataset_id   \n",
       "335    10.1126/science.abl4896  a68b64d8-aee3-4947-81b7-36b8fe5a44d2  \\\n",
       "336    10.1126/science.abl4896  97a17473-e2b1-4f31-a544-44a60773e2dd   \n",
       "337    10.1126/science.abl4896  c5d88abe-f23a-45fa-a534-788985e93dad   \n",
       "338    10.1126/science.abl4896  5a11f879-d1ef-458a-910c-9b0bdfca5ebf   \n",
       "339    10.1126/science.abl4896  53d208b0-2cfd-4366-9866-c3c6114081bc   \n",
       "358  10.1101/2022.10.12.511898  07b1d7c8-5c2e-42f7-9246-26f746cd6013   \n",
       "371  10.1101/2022.10.12.511898  7c1c3d47-3166-43e5-9a95-65ceb2d45f78   \n",
       "372  10.1101/2022.10.12.511898  9372df2d-13d6-4fac-980b-919a5b7eb483   \n",
       "399  10.1101/2022.10.12.511898  dd03ce70-3243-4c96-9561-330cc461e4d7   \n",
       "413  10.1101/2022.10.12.511898  7a0a8891-9a22-4549-a55b-c2aca23c3a2a   \n",
       "415  10.1101/2022.10.12.511898  d2b5efc1-14c6-4b5f-bd98-40f9084872d7   \n",
       "419  10.1101/2022.10.12.511898  f8dda921-5fb4-4c94-a654-c6fc346bfd6d   \n",
       "422  10.1101/2022.10.12.511898  3a7f3ab4-a280-4b3b-b2c0-6dd05614a78c   \n",
       "424  10.1101/2022.10.12.511898  bdb26abd-f4ba-4ea3-8862-c2340e7a4f55   \n",
       "426  10.1101/2022.10.12.511898  5e5ab909-f73f-4b57-98a0-6d2c5662f6a4   \n",
       "434  10.1101/2022.10.12.511898  35c8a04c-8639-4d15-8228-765d8d93fc96   \n",
       "436  10.1101/2022.10.12.511898  3f56901c-dd4a-47d6-b60b-7b0c0111cfb2   \n",
       "452  10.1101/2022.10.12.511898  93131426-0124-4ab4-a013-9dfbcd99d467   \n",
       "458  10.1101/2022.10.12.511898  acae7679-d077-461c-b857-ee6ccfeb267f   \n",
       "460  10.1101/2022.10.12.511898  c202b243-1aa1-4b16-bc9a-b36241f3b1e3   \n",
       "462  10.1101/2022.10.12.511898  0325478a-9b52-45b5-b40a-2e2ab0d72eb1   \n",
       "464  10.1101/2022.10.12.511898  c2aad8fc-b63b-4f9b-9cfd-baf7bc9c1771   \n",
       "465  10.1101/2022.10.12.511898  c4b03352-af8d-492a-8d6b-40f304e0a122   \n",
       "482  10.1101/2022.10.12.511898  fe1a73ab-a203-45fd-84e9-0f7fd19efcbd   \n",
       "\n",
       "                                         dataset_title   \n",
       "335                           Tabula Sapiens - Stromal  \\\n",
       "336                        Tabula Sapiens - Epithelial   \n",
       "337                            Tabula Sapiens - Immune   \n",
       "338                       Tabula Sapiens - Endothelial   \n",
       "339                         Tabula Sapiens - All Cells   \n",
       "358  Dissection: Myelencephalon (medulla oblongata)...   \n",
       "371  Dissection: Pons (Pn) - Pontine reticular form...   \n",
       "372  Dissection: Midbrain (M) - Periaqueductal gray...   \n",
       "399  Dissection: Cerebral cortex (Cx) - Perirhinal ...   \n",
       "413                    Supercluster: Hippocampal CA1-3   \n",
       "415  Dissection: Tail of Hippocampus (HiT) - Caudal...   \n",
       "419  Dissection: Cerebral cortex (Cx) - Occipitotem...   \n",
       "422                             Supercluster: Splatter   \n",
       "424                      Supercluster: CGE interneuron   \n",
       "426  Dissection: Midbrain (M) - Inferior colliculus...   \n",
       "434  Dissection: Hypothalamus (HTH) - supraoptic re...   \n",
       "436      Dissection: Head of hippocampus (HiH) - CA1-3   \n",
       "452                      Dissection: Epithalamus - ETH   \n",
       "458        Dissection: Head of hippocampus (HiH) - CA1   \n",
       "460                  Supercluster: Amygdala excitatory   \n",
       "462       Supercluster: Upper-layer intratelencephalic   \n",
       "464  Dissection: Cerebral cortex (Cx) - Temporal po...   \n",
       "465                  Supercluster: Medium spiny neuron   \n",
       "482  Dissection: Amygdaloid complex (AMY) - basolat...   \n",
       "\n",
       "                             dataset_h5ad_path  dataset_total_cell_count  \n",
       "335  a68b64d8-aee3-4947-81b7-36b8fe5a44d2.h5ad                     82478  \n",
       "336  97a17473-e2b1-4f31-a544-44a60773e2dd.h5ad                    104148  \n",
       "337  c5d88abe-f23a-45fa-a534-788985e93dad.h5ad                    264824  \n",
       "338  5a11f879-d1ef-458a-910c-9b0bdfca5ebf.h5ad                     31691  \n",
       "339  53d208b0-2cfd-4366-9866-c3c6114081bc.h5ad                    483152  \n",
       "358  07b1d7c8-5c2e-42f7-9246-26f746cd6013.h5ad                     27210  \n",
       "371  7c1c3d47-3166-43e5-9a95-65ceb2d45f78.h5ad                     49512  \n",
       "372  9372df2d-13d6-4fac-980b-919a5b7eb483.h5ad                     33794  \n",
       "399  dd03ce70-3243-4c96-9561-330cc461e4d7.h5ad                     23732  \n",
       "413  7a0a8891-9a22-4549-a55b-c2aca23c3a2a.h5ad                     74979  \n",
       "415  d2b5efc1-14c6-4b5f-bd98-40f9084872d7.h5ad                     36886  \n",
       "419  f8dda921-5fb4-4c94-a654-c6fc346bfd6d.h5ad                     31899  \n",
       "422  3a7f3ab4-a280-4b3b-b2c0-6dd05614a78c.h5ad                    291833  \n",
       "424  bdb26abd-f4ba-4ea3-8862-c2340e7a4f55.h5ad                    227671  \n",
       "426  5e5ab909-f73f-4b57-98a0-6d2c5662f6a4.h5ad                     32306  \n",
       "434  35c8a04c-8639-4d15-8228-765d8d93fc96.h5ad                     16753  \n",
       "436  3f56901c-dd4a-47d6-b60b-7b0c0111cfb2.h5ad                     37911  \n",
       "452  93131426-0124-4ab4-a013-9dfbcd99d467.h5ad                     24327  \n",
       "458  acae7679-d077-461c-b857-ee6ccfeb267f.h5ad                     39147  \n",
       "460  c202b243-1aa1-4b16-bc9a-b36241f3b1e3.h5ad                    109452  \n",
       "462  0325478a-9b52-45b5-b40a-2e2ab0d72eb1.h5ad                    455006  \n",
       "464  c2aad8fc-b63b-4f9b-9cfd-baf7bc9c1771.h5ad                     37642  \n",
       "465  c4b03352-af8d-492a-8d6b-40f304e0a122.h5ad                    152189  \n",
       "482  fe1a73ab-a203-45fd-84e9-0f7fd19efcbd.h5ad                     35285  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Grab the feature's soma_joinid from the var dataframe\n",
    "var_joinid = var_df.loc[var_df.feature_id == \"ENSG00000286096\"].soma_joinid\n",
    "\n",
    "# The presence matrix is indexed by the joinids of the dataset and var dataframes,\n",
    "# so slice out the feature of interest by its joinid.\n",
    "dataset_joinids = presence_matrix[:, var_joinid].tocoo().row\n",
    "\n",
    "# From the datasets dataframe, slice out the datasets which have a joinid in the list\n",
    "datasets_df.loc[datasets_df.soma_joinid.isin(dataset_joinids)]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Identifying all genes measured in a dataset \n",
    "\n",
    "Finally, we can find the set of genes that were measured in the cells of a given dataset."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-04-05T14:16:05.555010Z",
     "iopub.status.busy": "2023-04-05T14:16:05.554874Z",
     "iopub.status.idle": "2023-04-05T14:16:05.565033Z",
     "shell.execute_reply": "2023-04-05T14:16:05.563931Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>soma_joinid</th>\n",
       "      <th>feature_id</th>\n",
       "      <th>feature_name</th>\n",
       "      <th>feature_length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000238009</td>\n",
       "      <td>RP11-34P13.7</td>\n",
       "      <td>3726</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000279457</td>\n",
       "      <td>WASH9P</td>\n",
       "      <td>1397</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000228463</td>\n",
       "      <td>AP006222.1</td>\n",
       "      <td>8224</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>ENSG00000237094</td>\n",
       "      <td>RP4-669L17.4</td>\n",
       "      <td>6204</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>ENSG00000230021</td>\n",
       "      <td>RP11-206L10.17</td>\n",
       "      <td>5495</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40210</th>\n",
       "      <td>40210</td>\n",
       "      <td>ENSG00000255669</td>\n",
       "      <td>RP11-885B4.2</td>\n",
       "      <td>1023</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40219</th>\n",
       "      <td>40219</td>\n",
       "      <td>ENSG00000255618</td>\n",
       "      <td>LINC02440</td>\n",
       "      <td>741</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40352</th>\n",
       "      <td>40352</td>\n",
       "      <td>ENSG00000261627</td>\n",
       "      <td>RP11-91I8.1</td>\n",
       "      <td>508</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40367</th>\n",
       "      <td>40367</td>\n",
       "      <td>ENSG00000233376</td>\n",
       "      <td>RP5-881L22.6</td>\n",
       "      <td>403</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40376</th>\n",
       "      <td>40376</td>\n",
       "      <td>ENSG00000267465</td>\n",
       "      <td>AC011525.4</td>\n",
       "      <td>3363</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>27211 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       soma_joinid       feature_id    feature_name  feature_length\n",
       "0                0  ENSG00000238009    RP11-34P13.7            3726\n",
       "1                1  ENSG00000279457          WASH9P            1397\n",
       "2                2  ENSG00000228463      AP006222.1            8224\n",
       "3                3  ENSG00000237094    RP4-669L17.4            6204\n",
       "4                4  ENSG00000230021  RP11-206L10.17            5495\n",
       "...            ...              ...             ...             ...\n",
       "40210        40210  ENSG00000255669    RP11-885B4.2            1023\n",
       "40219        40219  ENSG00000255618       LINC02440             741\n",
       "40352        40352  ENSG00000261627     RP11-91I8.1             508\n",
       "40367        40367  ENSG00000233376    RP5-881L22.6             403\n",
       "40376        40376  ENSG00000267465      AC011525.4            3363\n",
       "\n",
       "[27211 rows x 4 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Slice the dataset(s) of interest, and get the joinid(s)\n",
    "dataset_joinids = datasets_df.loc[\n",
    "    datasets_df.collection_id == \"17481d16-ee44-49e5-bcf0-28c0780d8c4a\"\n",
    "].soma_joinid\n",
    "\n",
    "# Slice the presence matrix by the first dimension, i.e., by dataset\n",
    "var_joinids = presence_matrix[dataset_joinids, :].tocoo().col\n",
    "\n",
    "# From the feature (var) dataframe, slice out features which have a joinid in the list.\n",
    "var_df.loc[var_df.soma_joinid.isin(var_joinids)]"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.10"
  },
  "vscode": {
   "interpreter": {
    "hash": "3da8ec1c162cd849e59e6ea2824b2e353dce799884e910aae99411be5277f953"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
