

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8">
  <meta name="generator" content="Docutils 0.19: https://docutils.sourceforge.io/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Exploring the Census Datasets table &mdash; cellxgene-census  documentation</title>
  

  
  <link rel="icon" type="image/png" sizes="32x32" href="../../_static/img/favicon_32x32_v2.png"/>
  <link rel="icon" type="image/png" sizes="16x16" href="../../_static/img/favicon_16x16_v2.png"/>
  <link rel="preconnect" href="https://fonts.googleapis.com">
  <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
  <link href="https://fonts.googleapis.com/css2?family=Inter:wght@400;500;600;700&display=swap" rel="stylesheet">
  <link rel="stylesheet" href="../../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/css/custom.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/nbsphinx-code-cells.css" type="text/css" />

  
  
  
  

  
  <!--[if lt IE 9]>
    <script src="../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../../" src="../../_static/documentation_options.js"></script>
        <script data-url_root="../../" id="documentation_options" src="../../_static/documentation_options.js"></script>
        <script src="../../_static/doctools.js"></script>
        <script src="../../_static/sphinx_highlight.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
        <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
        <script data-domain="chanzuckerberg.github.io/cellxgene-census" defer="defer" src="https://plausible.io/js/script.js"></script>
    
    <script type="text/javascript" src="../../_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="Genes measured in each cell (dataset presence matrix)" href="census_dataset_presence.html" />
    <link rel="prev" title="Exploring all data from a tissue" href="../analysis_demo/comp_bio_explore_and_load_lung_data.html" /> 
</head>

<body class="wy-body-for-nav">

  <div style="height: 100vh; overflow: hidden;">
     
    <!-- top 56px for navbar height -->
    <div class="navbar-cxg">
        <div style="padding: 0 16px; display: flex; height: inherit; align-items: center; justify-content: space-between;">
            <span style="display:flex; gap: 32px; align-items: center;">
                <a href="https://cellxgene.cziscience.com/">
                  <img src="../../_static/img/cellxgene-discover-logo.svg" />
                </a>
                <span class="navbar-cxg-nav-wrapper">
                  <span class="navbar-cxg-section">
                    <span class="navbar-cxg-nav-section-title">Application</span>
                    <span class="navbar-cxg-nav-item-container">
                      <span class="navbar-cxg-link">
                          <a class="navbar-cxg-anchor" href="https://cellxgene.cziscience.com/collections">Collections</a>
                      </span>
                      <span class="navbar-cxg-link">
                          <a class="navbar-cxg-anchor" href="https://cellxgene.cziscience.com/datasets">Datasets</a>
                      </span>
                      <span class="navbar-cxg-link">
                          <a class="navbar-cxg-anchor" href="https://cellxgene.cziscience.com/gene-expression">Gene Expression</a>
                      </span>
                      <span class="navbar-cxg-link">
                        <a class="navbar-cxg-anchor" href="https://cellxgene.cziscience.com/cellguide">Cell Guide</a>
                        <div style="height: 16px!important; display: flex;">
                          <span class="beta">BETA</span>
                        </div>
                      </span>
                    </span>
                  </span>
                  <hr class="navbar-divider"/>
                  <span class="navbar-cxg-section">
                    <span class="navbar-cxg-nav-section-title">Census</span>
                    <span class="navbar-cxg-nav-item-container">
                      <span class="navbar-cxg-link active-link">
                        <a class="navbar-cxg-anchor" href="/cellxgene-census/index.html">API</a>
                      </span>
                      <span class="navbar-cxg-link">
                        <a class="navbar-cxg-anchor" href="https://cellxgene.cziscience.com/census-models">Models</a>
                      </span>
                    </span>
                  </span>
                </span>
            </span>

            <span class="navbar-cxg-link">
              <a class="navbar-cxg-anchor" href="https://cellxgene.cziscience.com/docs">Help & Documentation</a>
            </span>
        </div>
    </div>
    <div style="width: 100%; height: 100%; overflow: auto; padding-top: 56px;">
        
        <!-- top 56px for navbar height -->
        <nav data-toggle="wy-nav-shift" class="wy-nav-side" style="top: 56px;" >
        <div class="wy-side-scroll">
            <div class="wy-side-nav-search" >
            

            
                <a href="../../index.html" class="icon icon-home"> cellxgene-census
            

            
            </a>

            
                
                
                <div class="version">
                    v1.12.0
                </div>
                
            

            

            
            </div>

            
            <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
            
                
                
                
                
                
                <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../../cellxgene_census_docsite_installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../cellxgene_census_docsite_quick_start.html">Quick start</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../articles.html">What‚Äôs new?</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../cellxgene_census_aws_open_data.html">Census in AWS ‚òÅÔ∏è</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../cellxgene_census_docsite_schema.html">Census data and schema</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../cellxgene_census_docsite_data_release_info.html">Census data releases</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../python-api.html">Python API</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../../examples.html">Python tutorials</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="../../examples.html#exporting-data">Exporting data</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../examples.html#new-using-integrated-embeddings-and-models">[NEW! üöÄ] Using integrated embeddings and models</a></li>
<li class="toctree-l2 current"><a class="reference internal" href="../../examples.html#understanding-census-data">Understanding Census data</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="../analysis_demo/comp_bio_census_info.html">Learning about the CZ CELLxGENE Census</a></li>
<li class="toctree-l3"><a class="reference internal" href="../analysis_demo/comp_bio_explore_and_load_lung_data.html">Exploring all data from a tissue</a></li>
<li class="toctree-l3 current"><a class="current reference internal" href="#">Exploring the Census Datasets table</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#Fetching-the-datasets-table">Fetching the datasets table</a></li>
<li class="toctree-l4"><a class="reference internal" href="#Fetching-the-expression-data-from-a-single-dataset">Fetching the expression data from a single dataset</a></li>
<li class="toctree-l4"><a class="reference internal" href="#Downloading-the-original-source-H5AD-file-of-a-dataset.">Downloading the original source H5AD file of a dataset.</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="census_dataset_presence.html">Genes measured in each cell (dataset presence matrix)</a></li>
<li class="toctree-l3"><a class="reference internal" href="census_summary_cell_counts.html">Exploring pre-calculated summary cell counts</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../examples.html#analyzing-census-data">Analyzing Census data</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../examples.html#scalable-computing">Scalable computing</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../examples.html#scalable-machine-learning">Scalable machine learning</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference external" href="https://chanzuckerberg.github.io/cellxgene-census/r/index.html">R API &amp; tutorials</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../cellxgene_census_docsite_FAQ.html">FAQ</a></li>
</ul>

                
            
            </div>
            
        </div>
        </nav>

        <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

        
        <nav class="wy-nav-top" aria-label="top navigation">
            
            <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
            <a href="../../index.html">cellxgene-census</a>
            
        </nav>


        <div class="wy-nav-content">
            
            <div class="rst-content">
            
            <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../../examples.html">Python tutorials</a></li>
      <li class="breadcrumb-item active">Exploring the Census Datasets table</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../../_sources/notebooks/api_demo/census_datasets.ipynb.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
            <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
            <div itemprop="articleBody">
                
  <section id="Exploring-the-Census-Datasets-table">
<h1>Exploring the Census Datasets table<a class="headerlink" href="#Exploring-the-Census-Datasets-table" title="Permalink to this heading">¬∂</a></h1>
<p>This tutorial demonstrates basic use of the <code class="docutils literal notranslate"><span class="pre">census_datasets</span></code> dataframe that contains metadata of the Census source datasets. This metadata can be joined to the cell metadata dataframe (<code class="docutils literal notranslate"><span class="pre">obs</span></code>) via the column <code class="docutils literal notranslate"><span class="pre">dataset_id</span></code>,</p>
<p><strong>Contents</strong></p>
<ol class="arabic simple">
<li><p>Fetching the datasets table.</p></li>
<li><p>Fetching the expression data from a single dataset.</p></li>
<li><p>Downloading the original source H5AD file of a dataset.</p></li>
</ol>
<p>‚ö†Ô∏è Note that the Census RNA data includes duplicate cells present across multiple datasets. Duplicate cells can be filtered in or out using the cell metadata variable <code class="docutils literal notranslate"><span class="pre">is_primary_data</span></code> which is described in the <a class="reference external" href="https://github.com/chanzuckerberg/cellxgene-census/blob/main/docs/cellxgene_census_schema.md#repeated-data">Census schema</a>.</p>
<section id="Fetching-the-datasets-table">
<h2>Fetching the datasets table<a class="headerlink" href="#Fetching-the-datasets-table" title="Permalink to this heading">¬∂</a></h2>
<p>Each Census contains a top-level dataframe itemizing the datasets contained therein. You can read this into a <code class="docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code>.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">cellxgene_census</span>

<span class="n">census</span> <span class="o">=</span> <span class="n">cellxgene_census</span><span class="o">.</span><span class="n">open_soma</span><span class="p">()</span>
<span class="n">census_datasets</span> <span class="o">=</span> <span class="n">census</span><span class="p">[</span><span class="s2">&quot;census_info&quot;</span><span class="p">][</span><span class="s2">&quot;datasets&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">read</span><span class="p">()</span><span class="o">.</span><span class="n">concat</span><span class="p">()</span><span class="o">.</span><span class="n">to_pandas</span><span class="p">()</span>

<span class="c1"># for convenience, indexing on the soma_joinid which links this to other census data.</span>
<span class="n">census_datasets</span> <span class="o">=</span> <span class="n">census_datasets</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s2">&quot;soma_joinid&quot;</span><span class="p">)</span>

<span class="n">census_datasets</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
The &#34;stable&#34; release is currently 2023-07-25. Specify &#39;census_version=&#34;2023-07-25&#34;&#39; in future calls to open_soma() to ensure data consistency.
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>collection_id</th>
      <th>collection_name</th>
      <th>collection_doi</th>
      <th>dataset_id</th>
      <th>dataset_title</th>
      <th>dataset_h5ad_path</th>
      <th>dataset_total_cell_count</th>
    </tr>
    <tr>
      <th>soma_joinid</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>e2c257e7-6f79-487c-b81c-39451cd4ab3c</td>
      <td>Spatial multiomics map of trophoblast developm...</td>
      <td>10.1038/s41586-023-05869-0</td>
      <td>f171db61-e57e-4535-a06a-35d8b6ef8f2b</td>
      <td>donor_p13_trophoblasts</td>
      <td>f171db61-e57e-4535-a06a-35d8b6ef8f2b.h5ad</td>
      <td>31497</td>
    </tr>
    <tr>
      <th>1</th>
      <td>e2c257e7-6f79-487c-b81c-39451cd4ab3c</td>
      <td>Spatial multiomics map of trophoblast developm...</td>
      <td>10.1038/s41586-023-05869-0</td>
      <td>ecf2e08e-2032-4a9e-b466-b65b395f4a02</td>
      <td>All donors trophoblasts</td>
      <td>ecf2e08e-2032-4a9e-b466-b65b395f4a02.h5ad</td>
      <td>67070</td>
    </tr>
    <tr>
      <th>2</th>
      <td>e2c257e7-6f79-487c-b81c-39451cd4ab3c</td>
      <td>Spatial multiomics map of trophoblast developm...</td>
      <td>10.1038/s41586-023-05869-0</td>
      <td>74cff64f-9da9-4b2a-9b3b-8a04a1598040</td>
      <td>All donors all cell states (in vivo)</td>
      <td>74cff64f-9da9-4b2a-9b3b-8a04a1598040.h5ad</td>
      <td>286326</td>
    </tr>
    <tr>
      <th>3</th>
      <td>f7cecffa-00b4-4560-a29a-8ad626b8ee08</td>
      <td>Mapping single-cell transcriptomes in the intr...</td>
      <td>10.1016/j.ccell.2022.11.001</td>
      <td>5af90777-6760-4003-9dba-8f945fec6fdf</td>
      <td>Single-cell transcriptomic datasets of Renal c...</td>
      <td>5af90777-6760-4003-9dba-8f945fec6fdf.h5ad</td>
      <td>270855</td>
    </tr>
    <tr>
      <th>4</th>
      <td>3f50314f-bdc9-40c6-8e4a-b0901ebfbe4c</td>
      <td>Single-cell sequencing links multiregional imm...</td>
      <td>10.1016/j.ccell.2021.03.007</td>
      <td>bd65a70f-b274-4133-b9dd-0d1431b6af34</td>
      <td>Single-cell sequencing links multiregional imm...</td>
      <td>bd65a70f-b274-4133-b9dd-0d1431b6af34.h5ad</td>
      <td>167283</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>588</th>
      <td>180bff9c-c8a5-4539-b13b-ddbc00d643e6</td>
      <td>Molecular characterization of selectively vuln...</td>
      <td>10.1038/s41593-020-00764-7</td>
      <td>f9ad5649-f372-43e1-a3a8-423383e5a8a2</td>
      <td>Molecular characterization of selectively vuln...</td>
      <td>f9ad5649-f372-43e1-a3a8-423383e5a8a2.h5ad</td>
      <td>8168</td>
    </tr>
    <tr>
      <th>589</th>
      <td>a72afd53-ab92-4511-88da-252fb0e26b9a</td>
      <td>Single-cell atlas of peripheral immune respons...</td>
      <td>10.1038/s41591-020-0944-y</td>
      <td>456e8b9b-f872-488b-871d-94534090a865</td>
      <td>Single-cell atlas of peripheral immune respons...</td>
      <td>456e8b9b-f872-488b-871d-94534090a865.h5ad</td>
      <td>44721</td>
    </tr>
    <tr>
      <th>590</th>
      <td>38833785-fac5-48fd-944a-0f62a4c23ed1</td>
      <td>Construction of a human cell landscape at sing...</td>
      <td>10.1038/s41586-020-2157-4</td>
      <td>2adb1f8a-a6b1-4909-8ee8-484814e2d4bf</td>
      <td>Construction of a human cell landscape at sing...</td>
      <td>2adb1f8a-a6b1-4909-8ee8-484814e2d4bf.h5ad</td>
      <td>598266</td>
    </tr>
    <tr>
      <th>591</th>
      <td>5d445965-6f1a-4b68-ba3a-b8f765155d3a</td>
      <td>A molecular cell atlas of the human lung from ...</td>
      <td>10.1038/s41586-020-2922-4</td>
      <td>e04daea4-4412-45b5-989e-76a9be070a89</td>
      <td>Krasnow Lab Human Lung Cell Atlas, Smart-seq2</td>
      <td>e04daea4-4412-45b5-989e-76a9be070a89.h5ad</td>
      <td>9409</td>
    </tr>
    <tr>
      <th>592</th>
      <td>5d445965-6f1a-4b68-ba3a-b8f765155d3a</td>
      <td>A molecular cell atlas of the human lung from ...</td>
      <td>10.1038/s41586-020-2922-4</td>
      <td>8c42cfd0-0b0a-46d5-910c-fc833d83c45e</td>
      <td>Krasnow Lab Human Lung Cell Atlas, 10X</td>
      <td>8c42cfd0-0b0a-46d5-910c-fc833d83c45e.h5ad</td>
      <td>65662</td>
    </tr>
  </tbody>
</table>
<p>593 rows √ó 7 columns</p>
</div></div>
</div>
<p>The sum cells across all datasets should match the number of cells across all SOMA experiments (human, mouse).</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Count cells across all experiments</span>
<span class="n">all_experiments</span> <span class="o">=</span> <span class="p">(</span>
    <span class="p">(</span><span class="n">organism_name</span><span class="p">,</span> <span class="n">organism_experiment</span><span class="p">)</span> <span class="k">for</span> <span class="n">organism_name</span><span class="p">,</span> <span class="n">organism_experiment</span> <span class="ow">in</span> <span class="n">census</span><span class="p">[</span><span class="s2">&quot;census_data&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">items</span><span class="p">()</span>
<span class="p">)</span>
<span class="n">experiments_total_cells</span> <span class="o">=</span> <span class="mi">0</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Count by experiment:&quot;</span><span class="p">)</span>
<span class="k">for</span> <span class="n">organism_name</span><span class="p">,</span> <span class="n">organism_experiment</span> <span class="ow">in</span> <span class="n">all_experiments</span><span class="p">:</span>
    <span class="n">num_cells</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">organism_experiment</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">column_names</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;soma_joinid&quot;</span><span class="p">])</span><span class="o">.</span><span class="n">concat</span><span class="p">()</span><span class="o">.</span><span class="n">to_pandas</span><span class="p">())</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="se">\t</span><span class="si">{</span><span class="n">num_cells</span><span class="si">}</span><span class="s2"> cells in </span><span class="si">{</span><span class="n">organism_name</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">experiments_total_cells</span> <span class="o">+=</span> <span class="n">num_cells</span>

<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">Found </span><span class="si">{</span><span class="n">experiments_total_cells</span><span class="si">}</span><span class="s2"> cells in all experiments.&quot;</span><span class="p">)</span>

<span class="c1"># Count cells across all datasets</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Found </span><span class="si">{</span><span class="n">census_datasets</span><span class="o">.</span><span class="n">dataset_total_cell_count</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span><span class="si">}</span><span class="s2"> cells in all datasets.&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Count by experiment:
        5255245 cells in mus_musculus
        56400873 cells in homo_sapiens

Found 61656118 cells in all experiments.
Found 61656118 cells in all datasets.
</pre></div></div>
</div>
</section>
<section id="Fetching-the-expression-data-from-a-single-dataset">
<h2>Fetching the expression data from a single dataset<a class="headerlink" href="#Fetching-the-expression-data-from-a-single-dataset" title="Permalink to this heading">¬∂</a></h2>
<p>Lets pick one dataset to slice out of the census, and turn into an <a class="reference external" href="https://anndata.readthedocs.io/en/latest/">AnnData</a> in-memory object. This can be used with the <a class="reference external" href="https://scanpy.readthedocs.io/en/stable/">ScanPy</a> toolchain. You can also save this AnnData locally using the AnnData <a class="reference external" href="https://anndata.readthedocs.io/en/latest/api.html#writing">write</a> API.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">census_datasets</span><span class="p">[</span><span class="n">census_datasets</span><span class="o">.</span><span class="n">dataset_id</span> <span class="o">==</span> <span class="s2">&quot;0bd1a1de-3aee-40e0-b2ec-86c7a30c7149&quot;</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>collection_id</th>
      <th>collection_name</th>
      <th>collection_doi</th>
      <th>dataset_id</th>
      <th>dataset_title</th>
      <th>dataset_h5ad_path</th>
      <th>dataset_total_cell_count</th>
    </tr>
    <tr>
      <th>soma_joinid</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>522</th>
      <td>0b9d8a04-bb9d-44da-aa27-705bb65b54eb</td>
      <td>Tabula Muris Senis</td>
      <td>10.1038/s41586-020-2496-1</td>
      <td>0bd1a1de-3aee-40e0-b2ec-86c7a30c7149</td>
      <td>Bone marrow - A single-cell transcriptomic atl...</td>
      <td>0bd1a1de-3aee-40e0-b2ec-86c7a30c7149.h5ad</td>
      <td>40220</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<p>Create a query on the mouse experiment, ‚ÄúRNA‚Äù measurement, for the dataset_id.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span> <span class="o">=</span> <span class="n">cellxgene_census</span><span class="o">.</span><span class="n">get_anndata</span><span class="p">(</span>
    <span class="n">census</span><span class="p">,</span> <span class="n">organism</span><span class="o">=</span><span class="s2">&quot;Mus musculus&quot;</span><span class="p">,</span> <span class="n">obs_value_filter</span><span class="o">=</span><span class="s2">&quot;dataset_id == &#39;0bd1a1de-3aee-40e0-b2ec-86c7a30c7149&#39;&quot;</span>
<span class="p">)</span>

<span class="n">adata</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
AnnData object with n_obs √ó n_vars = 40220 √ó 52392
    obs: &#39;soma_joinid&#39;, &#39;dataset_id&#39;, &#39;assay&#39;, &#39;assay_ontology_term_id&#39;, &#39;cell_type&#39;, &#39;cell_type_ontology_term_id&#39;, &#39;development_stage&#39;, &#39;development_stage_ontology_term_id&#39;, &#39;disease&#39;, &#39;disease_ontology_term_id&#39;, &#39;donor_id&#39;, &#39;is_primary_data&#39;, &#39;self_reported_ethnicity&#39;, &#39;self_reported_ethnicity_ontology_term_id&#39;, &#39;sex&#39;, &#39;sex_ontology_term_id&#39;, &#39;suspension_type&#39;, &#39;tissue&#39;, &#39;tissue_ontology_term_id&#39;, &#39;tissue_general&#39;, &#39;tissue_general_ontology_term_id&#39;
    var: &#39;soma_joinid&#39;, &#39;feature_id&#39;, &#39;feature_name&#39;, &#39;feature_length&#39;
</pre></div></div>
</div>
</section>
<section id="Downloading-the-original-source-H5AD-file-of-a-dataset.">
<h2>Downloading the original source H5AD file of a dataset.<a class="headerlink" href="#Downloading-the-original-source-H5AD-file-of-a-dataset." title="Permalink to this heading">¬∂</a></h2>
<p>You can download the original H5AD file for any given dataset. This is the same H5AD you can download from the <a class="reference external" href="https://cellxgene.cziscience.com/">CZ CELLxGENE Discover</a>, and may contain additional data-submitter provided information which was not included in the Census.</p>
<p>To do this you can fetch the location in the cloud or directly download to your system using the <code class="docutils literal notranslate"><span class="pre">cellxgene-census</span></code></p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Option 1: Direct download</span>
<span class="n">cellxgene_census</span><span class="o">.</span><span class="n">download_source_h5ad</span><span class="p">(</span>
    <span class="s2">&quot;0bd1a1de-3aee-40e0-b2ec-86c7a30c7149&quot;</span><span class="p">,</span> <span class="n">to_path</span><span class="o">=</span><span class="s2">&quot;Tabula_Muris_Senis-bone_marrow.h5ad&quot;</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Option 2: Get location and download via preferred method</span>
<span class="n">uri</span> <span class="o">=</span> <span class="n">cellxgene_census</span><span class="o">.</span><span class="n">get_source_h5ad_uri</span><span class="p">(</span><span class="s2">&quot;0bd1a1de-3aee-40e0-b2ec-86c7a30c7149&quot;</span><span class="p">)</span>
<span class="n">uri</span>

<span class="c1"># you can now download the H5AD in shell via AWS CLI e.g. `aws s3 cp uri ./`</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
{&#39;uri&#39;: &#39;s3://cellxgene-data-public/cell-census/2023-07-25/h5ads/0bd1a1de-3aee-40e0-b2ec-86c7a30c7149.h5ad&#39;,
 &#39;s3_region&#39;: &#39;us-west-2&#39;}
</pre></div></div>
</div>
<p>Close the census</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">census</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</pre></div>
</div>
</div>
</section>
</section>


            </div>
            
            </div>
            <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="../analysis_demo/comp_bio_explore_and_load_lung_data.html" class="btn btn-neutral float-left" title="Exploring all data from a tissue" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="census_dataset_presence.html" class="btn btn-neutral float-right" title="Genes measured in each cell (dataset presence matrix)" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2022, Chan Zuckerberg Initiative.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
            </div>
        </div>

        </section>

    </div>
    <!-- Newsletter Banner -->
    <div role="banner" id="newsletter-banner">
      <span id="newsletter-subscribe-button" role="button">Subscribe</span>&nbsp;to our newsletter to receive updates about new features.
      <div id="newsletter-banner-close-button" role="button">X</div>
    </div>
    <!-- Newsletter Modal -->
    <dialog id="newsletter-modal">
      <div id="newsletter-header">
        <img id="newsletter-logo" src="../../_static/img/cellxGene-newsletter-logo.svg" />
        <div id="newsletter-close-button" role="button">X</div>
      </div>
      <div id="newsletter-content">
        <div id="newsletter-callout">Join Our Newsletter</div>
        <div id="newsletter-description">Get a quarterly email with the latest CELLxGENE features and data.</div>
        <!-- HubSpot Form target -->
        <div id="newsletter-form-container"></div>
      </div>

      <div id="newsletter-footnote">Unsubscribe at any time.</div>
    </dialog>
  </div>
  

  <script>
    // (thuang): 30 days
    const NEWSLETTER_BANNER_DISMISSED_TTL_MS = 30 * 24 * 60 * 60 * 1000;
    const NEWSLETTER_BANNER_DISMISSED_KEY = "newsletterBannerDismissed"

    var script = document.createElement('script');
    script.src = 'https://js.hsforms.net/forms/v2.js';
    script.defer = true;
    document.head.appendChild(script);

    // Run the code once the script is loaded
    script.onload = async function() {
      await hbspt.forms.create({
        region: "na1",
        portalId: "7272273",
        formId: "eb65b811-0451-414d-8304-7b9b6f468ce5",
        target: '#newsletter-form-container',
        onFormReady() {
          // get element by type "email"
          const emailInput = document.querySelector('#email-eb65b811-0451-414d-8304-7b9b6f468ce5');
          emailInput.setAttribute('placeholder', 'Enter email address');

          // remove the label element for emailInput
          const emailLabel = document.querySelector('#label-email-eb65b811-0451-414d-8304-7b9b6f468ce5');
          emailLabel.remove();
        },
        submitText: 'Subscribe',
      });
    };

    checkNewsletterBanner();

    document.querySelector('#newsletter-banner-close-button').addEventListener('click', () => {
      document.querySelector('#newsletter-banner').remove();
      localStorage.setItem(NEWSLETTER_BANNER_DISMISSED_KEY, Date.now());
    });

    const modal = document.querySelector('#newsletter-modal');

    document.querySelector('#newsletter-subscribe-button').addEventListener('click', () => {
      modal.showModal();
    });

    document.querySelector('#newsletter-close-button').addEventListener('click', () => {
      modal.close();
    });

    function checkNewsletterBanner() {
      /**
       * (thuang): Use LocalStorage to store dismissed state for 30 days
       * NOTE: Currently Census doc page doesn't share the same domain as the main site,
       * so dismissing the banner on the main site won't dismiss it on the Census doc page.
       * And vice versa.
       */
      const newsletterBannerDismissed = localStorage.getItem('newsletterBannerDismissed');

      if (newsletterBannerDismissed) {
        return;
      }

      if (newsletterBannerDismissed && Date.now() - newsletterBannerDismissed > NEWSLETTER_BANNER_DISMISSED_TTL_MS) {
        localStorage.removeItem(NEWSLETTER_BANNER_DISMISSED_KEY);
      }

      const newsletterBanner = document.querySelector('#newsletter-banner');

      if (!newsletterBannerDismissed) {
        newsletterBanner.style.display = 'flex';
      }
    }
  </script>

  
  
    
   

</body>
</html>