

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8">
  <meta name="generator" content="Docutils 0.19: https://docutils.sourceforge.io/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Querying and fetching the single-cell data and cell/gene metadata. &mdash; cellxgene-census  documentation</title>
  

  
  <link rel="icon" type="image/png" sizes="32x32" href="../../_static/img/favicon_32x32_v2.png"/>
  <link rel="icon" type="image/png" sizes="16x16" href="../../_static/img/favicon_16x16_v2.png"/>
  <link rel="preconnect" href="https://fonts.googleapis.com">
  <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
  <link href="https://fonts.googleapis.com/css2?family=Inter:wght@400;500;600;700&display=swap" rel="stylesheet">
  <link rel="stylesheet" href="../../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/css/custom.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/nbsphinx-code-cells.css" type="text/css" />

  
  
  
  

  
  <!--[if lt IE 9]>
    <script src="../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../../" src="../../_static/documentation_options.js"></script>
        <script data-url_root="../../" id="documentation_options" src="../../_static/documentation_options.js"></script>
        <script src="../../_static/doctools.js"></script>
        <script src="../../_static/sphinx_highlight.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
        <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
        <script data-domain="chanzuckerberg.github.io/cellxgene-census" defer="defer" src="https://plausible.io/js/script.js"></script>
    
    <script type="text/javascript" src="../../_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="Generating citations for Census slices" href="census_citation_generation.html" />
    <link rel="prev" title="Python tutorials" href="../../examples.html" /> 
</head>

<body class="wy-body-for-nav">

  <div style="height: 100vh; overflow: hidden;">
     
    <!-- top 56px for navbar height -->
    <div class="navbar-cxg">
        <div style="padding: 0 16px; display: flex; height: inherit; align-items: center; justify-content: space-between;">
            <span style="display:flex; gap: 32px; align-items: center;">
                <a href="https://cellxgene.cziscience.com/">
                  <img src="../../_static/img/cellxgene-discover-logo.svg" />
                </a>
                <span class="navbar-cxg-nav-wrapper">
                  <span class="navbar-cxg-section">
                    <span class="navbar-cxg-nav-section-title">Application</span>
                    <span class="navbar-cxg-nav-item-container">
                      <span class="navbar-cxg-link">
                          <a class="navbar-cxg-anchor" href="https://cellxgene.cziscience.com/collections">Collections</a>
                      </span>
                      <span class="navbar-cxg-link">
                          <a class="navbar-cxg-anchor" href="https://cellxgene.cziscience.com/datasets">Datasets</a>
                      </span>
                      <span class="navbar-cxg-link">
                          <a class="navbar-cxg-anchor" href="https://cellxgene.cziscience.com/gene-expression">Gene Expression</a>
                      </span>
                      <span class="navbar-cxg-link">
                        <a class="navbar-cxg-anchor" href="https://cellxgene.cziscience.com/cellguide">Cell Guide</a>
                        <div style="height: 16px!important; display: flex;">
                          <span class="beta">BETA</span>
                        </div>
                      </span>
                    </span>
                  </span>
                  <hr class="navbar-divider"/>
                  <span class="navbar-cxg-section">
                    <span class="navbar-cxg-nav-section-title">Census</span>
                    <span class="navbar-cxg-nav-item-container">
                      <span class="navbar-cxg-link active-link">
                        <a class="navbar-cxg-anchor" href="/cellxgene-census/index.html">API</a>
                      </span>
                      <span class="navbar-cxg-link">
                        <a class="navbar-cxg-anchor" href="https://cellxgene.cziscience.com/census-models">Models</a>
                      </span>
                    </span>
                  </span>
                </span>
            </span>

            <span class="navbar-cxg-link">
              <a class="navbar-cxg-anchor" href="https://cellxgene.cziscience.com/docs">Help & Documentation</a>
            </span>
        </div>
    </div>
    <div style="width: 100%; height: 100%; overflow: auto; padding-top: 56px;">
        
        <!-- top 56px for navbar height -->
        <nav data-toggle="wy-nav-shift" class="wy-nav-side" style="top: 56px;" >
        <div class="wy-side-scroll">
            <div class="wy-side-nav-search" >
            

            
                <a href="../../index.html" class="icon icon-home"> cellxgene-census
            

            
            </a>

            
                
                
                <div class="version">
                    v1.13.1
                </div>
                
            

            

            
            </div>

            
            <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
            
                
                
                
                
                
                <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../../cellxgene_census_docsite_installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../cellxgene_census_docsite_quick_start.html">Quick start</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../articles.html">What‚Äôs new?</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../cellxgene_census_aws_open_data.html">Census in AWS ‚òÅÔ∏è</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../cellxgene_census_docsite_schema.html">Census data and schema</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../cellxgene_census_docsite_data_release_info.html">Census data releases</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../python-api.html">Python API</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../../examples.html">Python tutorials</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="../../examples.html#exporting-data">Exporting data</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="../../cellxgene_census_docsite_quick_start.html">Quick start</a></li>
<li class="toctree-l3 current"><a class="current reference internal" href="#">Querying and fetching the single-cell data and cell/gene metadata.</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#Opening-the-census">Opening the census</a></li>
<li class="toctree-l4"><a class="reference internal" href="#Querying-expression-data">Querying expression data</a></li>
<li class="toctree-l4"><a class="reference internal" href="#Querying-cell-metadata-(obs)">Querying cell metadata (obs)</a></li>
<li class="toctree-l4"><a class="reference internal" href="#Querying-gene-metadata-(var)">Querying gene metadata (var)</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="census_citation_generation.html">Generating citations for Census slices</a></li>
<li class="toctree-l3"><a class="reference internal" href="census_gget_demo.html">Querying data using the gget cellxgene module</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../examples.html#new-using-integrated-embeddings-and-models">[NEW! üöÄ] Using integrated embeddings and models</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../examples.html#understanding-census-data">Understanding Census data</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../examples.html#analyzing-census-data">Analyzing Census data</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../examples.html#scalable-computing">Scalable computing</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../examples.html#scalable-machine-learning">Scalable machine learning</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference external" href="https://chanzuckerberg.github.io/cellxgene-census/r/index.html">R API &amp; tutorials</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../cellxgene_census_docsite_FAQ.html">FAQ</a></li>
</ul>

                
            
            </div>
            
        </div>
        </nav>

        <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

        
        <nav class="wy-nav-top" aria-label="top navigation">
            
            <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
            <a href="../../index.html">cellxgene-census</a>
            
        </nav>


        <div class="wy-nav-content">
            
            <div class="rst-content">
            
            <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../../examples.html">Python tutorials</a></li>
      <li class="breadcrumb-item active">Querying and fetching the single-cell data and cell/gene metadata.</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../../_sources/notebooks/api_demo/census_query_extract.ipynb.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
            <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
            <div itemprop="articleBody">
                
  <section id="Querying-and-fetching-the-single-cell-data-and-cell/gene-metadata.">
<h1>Querying and fetching the single-cell data and cell/gene metadata.<a class="headerlink" href="#Querying-and-fetching-the-single-cell-data-and-cell/gene-metadata." title="Permalink to this heading">¬∂</a></h1>
<p>This tutorial showcases the easiest ways to query the expression data and cell/gene metadata from the Census, and load them into common in-memory Python objects, including <code class="docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code> and <code class="docutils literal notranslate"><span class="pre">anndata.AnnData</span></code>.</p>
<p><strong>Contents</strong></p>
<ol class="arabic simple">
<li><p>Opening the census.</p></li>
<li><p>Querying expression data.</p></li>
<li><p>Querying cell metadata (obs).</p></li>
<li><p>Querying gene metadata (var).</p></li>
</ol>
<p>‚ö†Ô∏è Note that the Census RNA data includes duplicate cells present across multiple datasets. Duplicate cells can be filtered in or out using the cell metadata variable <code class="docutils literal notranslate"><span class="pre">is_primary_data</span></code> which is described in the <a class="reference external" href="https://github.com/chanzuckerberg/cellxgene-census/blob/main/docs/cellxgene_census_schema.md#repeated-data">Census schema</a>.</p>
<section id="Opening-the-census">
<h2>Opening the census<a class="headerlink" href="#Opening-the-census" title="Permalink to this heading">¬∂</a></h2>
<p>The <code class="docutils literal notranslate"><span class="pre">cellxgene_census</span></code> python package contains a convenient API to open the latest version of the Census.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">cellxgene_census</span>

<span class="n">census</span> <span class="o">=</span> <span class="n">cellxgene_census</span><span class="o">.</span><span class="n">open_soma</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
The &#34;stable&#34; release is currently 2023-07-25. Specify &#39;census_version=&#34;2023-07-25&#34;&#39; in future calls to open_soma() to ensure data consistency.
</pre></div></div>
</div>
<p>You can learn more about the <code class="docutils literal notranslate"><span class="pre">cellxgene_census</span></code> methods by accessing their corresponding documentation via <code class="docutils literal notranslate"><span class="pre">help()</span></code>. For example <code class="docutils literal notranslate"><span class="pre">help(cellxgene_census.open_soma)</span></code>.</p>
</section>
<section id="Querying-expression-data">
<h2>Querying expression data<a class="headerlink" href="#Querying-expression-data" title="Permalink to this heading">¬∂</a></h2>
<p>A convenient way to query and fetch expression data is to use the <code class="docutils literal notranslate"><span class="pre">get_anndata</span></code> method of the <code class="docutils literal notranslate"><span class="pre">cellxgene_census</span></code> API. This is a method that combines the column selection and value filtering we described above to obtain slices of the expression data based on metadata queries.</p>
<p>The method will return an <code class="docutils literal notranslate"><span class="pre">anndata.AnnData</span></code> object, it takes as an input a census object, the string for an organism, and for both cell and gene metadata we can specify filters and column selection as described above but with the following arguments:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">column_names</span></code> ‚Äî a dictionary with two keys <code class="docutils literal notranslate"><span class="pre">obs</span></code> and <code class="docutils literal notranslate"><span class="pre">var</span></code> whose values are lists of strings indicating the columns to select for cell and gene metadata respectively.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">obs_value_filter</span></code> ‚Äî python expression with selection conditions to fetch <strong>cells</strong> meeting a criteria. For full details see <a class="reference external" href="https://tiledb-inc-tiledb.readthedocs-hosted.com/projects/tiledb-py/en/stable/python-api.html#query-condition">tiledb.QueryCondition</a>.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">var_value_filter</span></code> ‚Äî python expression with selection conditions to fetch <strong>genes</strong> meeting a criteria. Details as above. For full details see <a class="reference external" href="https://tiledb-inc-tiledb.readthedocs-hosted.com/projects/tiledb-py/en/stable/python-api.html#query-condition">tiledb.QueryCondition</a>.</p></li>
</ul>
<p>For example if we want to fetch the expression data for:</p>
<ul class="simple">
<li><p>Genes <code class="docutils literal notranslate"><span class="pre">&quot;ENSG00000161798&quot;</span></code> and <code class="docutils literal notranslate"><span class="pre">&quot;ENSG00000188229&quot;</span></code>.</p></li>
<li><p>All <code class="docutils literal notranslate"><span class="pre">&quot;B</span> <span class="pre">cells&quot;</span></code> of <code class="docutils literal notranslate"><span class="pre">&quot;lung&quot;</span></code> with <code class="docutils literal notranslate"><span class="pre">&quot;COVID-19&quot;</span></code> from non-duplicated cells.</p></li>
<li><p>With all gene metadata and adding <code class="docutils literal notranslate"><span class="pre">sex</span></code> cell metadata.</p></li>
</ul>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span> <span class="o">=</span> <span class="n">cellxgene_census</span><span class="o">.</span><span class="n">get_anndata</span><span class="p">(</span>
    <span class="n">census</span><span class="o">=</span><span class="n">census</span><span class="p">,</span>
    <span class="n">organism</span><span class="o">=</span><span class="s2">&quot;Homo sapiens&quot;</span><span class="p">,</span>
    <span class="n">var_value_filter</span><span class="o">=</span><span class="s2">&quot;feature_id in [&#39;ENSG00000161798&#39;, &#39;ENSG00000188229&#39;]&quot;</span><span class="p">,</span>
    <span class="n">obs_value_filter</span><span class="o">=</span><span class="s2">&quot;cell_type == &#39;B cell&#39; and tissue_general == &#39;lung&#39; and disease == &#39;COVID-19&#39; and is_primary_data == True&quot;</span><span class="p">,</span>
    <span class="n">column_names</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;obs&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;sex&quot;</span><span class="p">]},</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<p>And now we can take a look at the results.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
AnnData object with n_obs √ó n_vars = 2313 √ó 2
    obs: &#39;sex&#39;, &#39;cell_type&#39;, &#39;tissue_general&#39;, &#39;disease&#39;, &#39;is_primary_data&#39;
    var: &#39;soma_joinid&#39;, &#39;feature_id&#39;, &#39;feature_name&#39;, &#39;feature_length&#39;
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>sex</th>
      <th>cell_type</th>
      <th>tissue_general</th>
      <th>disease</th>
      <th>is_primary_data</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>male</td>
      <td>B cell</td>
      <td>lung</td>
      <td>COVID-19</td>
      <td>True</td>
    </tr>
    <tr>
      <th>1</th>
      <td>male</td>
      <td>B cell</td>
      <td>lung</td>
      <td>COVID-19</td>
      <td>True</td>
    </tr>
    <tr>
      <th>2</th>
      <td>unknown</td>
      <td>B cell</td>
      <td>lung</td>
      <td>COVID-19</td>
      <td>True</td>
    </tr>
    <tr>
      <th>3</th>
      <td>male</td>
      <td>B cell</td>
      <td>lung</td>
      <td>COVID-19</td>
      <td>True</td>
    </tr>
    <tr>
      <th>4</th>
      <td>unknown</td>
      <td>B cell</td>
      <td>lung</td>
      <td>COVID-19</td>
      <td>True</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>2308</th>
      <td>male</td>
      <td>B cell</td>
      <td>lung</td>
      <td>COVID-19</td>
      <td>True</td>
    </tr>
    <tr>
      <th>2309</th>
      <td>male</td>
      <td>B cell</td>
      <td>lung</td>
      <td>COVID-19</td>
      <td>True</td>
    </tr>
    <tr>
      <th>2310</th>
      <td>male</td>
      <td>B cell</td>
      <td>lung</td>
      <td>COVID-19</td>
      <td>True</td>
    </tr>
    <tr>
      <th>2311</th>
      <td>male</td>
      <td>B cell</td>
      <td>lung</td>
      <td>COVID-19</td>
      <td>True</td>
    </tr>
    <tr>
      <th>2312</th>
      <td>male</td>
      <td>B cell</td>
      <td>lung</td>
      <td>COVID-19</td>
      <td>True</td>
    </tr>
  </tbody>
</table>
<p>2313 rows √ó 5 columns</p>
</div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span><span class="o">.</span><span class="n">var</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>soma_joinid</th>
      <th>feature_id</th>
      <th>feature_name</th>
      <th>feature_length</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>8626</td>
      <td>ENSG00000161798</td>
      <td>AQP5</td>
      <td>1884</td>
    </tr>
    <tr>
      <th>1</th>
      <td>27047</td>
      <td>ENSG00000188229</td>
      <td>TUBB4B</td>
      <td>2037</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<p>For a full description of <code class="docutils literal notranslate"><span class="pre">get_anndata()</span></code> refer to <code class="docutils literal notranslate"><span class="pre">help(cellxgene_census.get_anndata)</span></code></p>
<p>Don‚Äôt forget to close the census!</p>
</section>
<section id="Querying-cell-metadata-(obs)">
<h2>Querying cell metadata (obs)<a class="headerlink" href="#Querying-cell-metadata-(obs)" title="Permalink to this heading">¬∂</a></h2>
<p>The human gene metadata of the Census, for RNA assays, is located at <code class="docutils literal notranslate"><span class="pre">census[&quot;census_data&quot;][&quot;homo_sapiens&quot;].obs</span></code>. This is a <code class="docutils literal notranslate"><span class="pre">SOMADataFrame</span></code> and as such it can be materialized as a <code class="docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code> via the methods <code class="docutils literal notranslate"><span class="pre">read().concat().to_pandas()</span></code>.</p>
<p>The mouse cell metadata is at <code class="docutils literal notranslate"><span class="pre">census[&quot;census_data&quot;][&quot;mus_musculus&quot;].obs</span></code>.</p>
<p>For slicing the cell metadata there are two relevant arguments that can be passed through <code class="docutils literal notranslate"><span class="pre">read()</span></code>:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">column_names</span></code> ‚Äî list of strings indicating what metadata columns to fetch.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">value_filter</span></code> ‚Äî Python expression with selection conditions to fetch rows, it is similar to <a class="reference external" href="https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.query.html">pandas.DataFrame.query()</a>, for full details see <a class="reference external" href="https://tiledb-inc-tiledb.readthedocs-hosted.com/projects/tiledb-py/en/stable/python-api.html#query-condition">tiledb.QueryCondition</a> shortly:</p>
<ul>
<li><p>Expressions are one or more comparisons</p></li>
<li><p>Comparisons are one of <code class="docutils literal notranslate"><span class="pre">&lt;column&gt;</span> <span class="pre">&lt;op&gt;</span> <span class="pre">&lt;value&gt;</span></code> or <code class="docutils literal notranslate"><span class="pre">&lt;column&gt;</span> <span class="pre">&lt;op&gt;</span> <span class="pre">&lt;column&gt;</span></code></p></li>
<li><p>Expressions can combine comparisons using and, or, &amp; or |</p></li>
<li><p>op is one of &lt; | &gt; | &lt;= | &gt;= | == | != or in</p></li>
</ul>
</li>
</ul>
<p>To learn what metadata columns are available for fetching and filtering we can directly look at the keys of the cell metadata.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">keys</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">census</span><span class="p">[</span><span class="s2">&quot;census_data&quot;</span><span class="p">][</span><span class="s2">&quot;homo_sapiens&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span>

<span class="n">keys</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
[&#39;soma_joinid&#39;,
 &#39;dataset_id&#39;,
 &#39;assay&#39;,
 &#39;assay_ontology_term_id&#39;,
 &#39;cell_type&#39;,
 &#39;cell_type_ontology_term_id&#39;,
 &#39;development_stage&#39;,
 &#39;development_stage_ontology_term_id&#39;,
 &#39;disease&#39;,
 &#39;disease_ontology_term_id&#39;,
 &#39;donor_id&#39;,
 &#39;is_primary_data&#39;,
 &#39;self_reported_ethnicity&#39;,
 &#39;self_reported_ethnicity_ontology_term_id&#39;,
 &#39;sex&#39;,
 &#39;sex_ontology_term_id&#39;,
 &#39;suspension_type&#39;,
 &#39;tissue&#39;,
 &#39;tissue_ontology_term_id&#39;,
 &#39;tissue_general&#39;,
 &#39;tissue_general_ontology_term_id&#39;]
</pre></div></div>
</div>
<p><code class="docutils literal notranslate"><span class="pre">soma_joinid</span></code> is a special <code class="docutils literal notranslate"><span class="pre">SOMADataFrame</span></code> column that is used for join operations. The definition for all other columns can be found at the <a class="reference external" href="https://github.com/chanzuckerberg/cellxgene-census/blob/main/docs/cell_census_schema.md#cell-metadata--census_objcensus_dataorganismobs--somadataframe">Census schema</a>.</p>
<p>All of these can be used to fetch specific columns or specific rows matching a condition. For the latter we need to know the values we are looking for <em>a priori</em>.</p>
<p>For example let‚Äôs see what are the possible values available for <code class="docutils literal notranslate"><span class="pre">sex</span></code>. To this we can load all cell metadata but fetching only for the column <code class="docutils literal notranslate"><span class="pre">sex</span></code>.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sex_cell_metadata</span> <span class="o">=</span> <span class="n">census</span><span class="p">[</span><span class="s2">&quot;census_data&quot;</span><span class="p">][</span><span class="s2">&quot;homo_sapiens&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">column_names</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;sex&quot;</span><span class="p">])</span><span class="o">.</span><span class="n">concat</span><span class="p">()</span><span class="o">.</span><span class="n">to_pandas</span><span class="p">()</span>

<span class="n">sex_cell_metadata</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>sex</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>unknown</td>
    </tr>
    <tr>
      <th>669</th>
      <td>female</td>
    </tr>
    <tr>
      <th>385437</th>
      <td>male</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<p>As you can see there are only three different values for <code class="docutils literal notranslate"><span class="pre">sex</span></code>, that is <code class="docutils literal notranslate"><span class="pre">&quot;male&quot;</span></code>, <code class="docutils literal notranslate"><span class="pre">&quot;female&quot;</span></code> and <code class="docutils literal notranslate"><span class="pre">&quot;unknown&quot;</span></code>.</p>
<p>With this information we can fetch all cell metatadata for a specific <code class="docutils literal notranslate"><span class="pre">sex</span></code> value, for example <code class="docutils literal notranslate"><span class="pre">&quot;unknown&quot;</span></code>.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">cell_metadata_all_unknown_sex</span> <span class="o">=</span> <span class="p">(</span>
    <span class="n">census</span><span class="p">[</span><span class="s2">&quot;census_data&quot;</span><span class="p">][</span><span class="s2">&quot;homo_sapiens&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">value_filter</span><span class="o">=</span><span class="s2">&quot;sex == &#39;unknown&#39;&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">concat</span><span class="p">()</span><span class="o">.</span><span class="n">to_pandas</span><span class="p">()</span>
<span class="p">)</span>

<span class="n">cell_metadata_all_unknown_sex</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>soma_joinid</th>
      <th>dataset_id</th>
      <th>assay</th>
      <th>assay_ontology_term_id</th>
      <th>cell_type</th>
      <th>cell_type_ontology_term_id</th>
      <th>development_stage</th>
      <th>development_stage_ontology_term_id</th>
      <th>disease</th>
      <th>disease_ontology_term_id</th>
      <th>...</th>
      <th>is_primary_data</th>
      <th>self_reported_ethnicity</th>
      <th>self_reported_ethnicity_ontology_term_id</th>
      <th>sex</th>
      <th>sex_ontology_term_id</th>
      <th>suspension_type</th>
      <th>tissue</th>
      <th>tissue_ontology_term_id</th>
      <th>tissue_general</th>
      <th>tissue_general_ontology_term_id</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>0</td>
      <td>f171db61-e57e-4535-a06a-35d8b6ef8f2b</td>
      <td>10x 3' v3</td>
      <td>EFO:0009922</td>
      <td>syncytiotrophoblast cell</td>
      <td>CL:0000525</td>
      <td>9th week post-fertilization human stage</td>
      <td>HsapDv:0000046</td>
      <td>normal</td>
      <td>PATO:0000461</td>
      <td>...</td>
      <td>False</td>
      <td>unknown</td>
      <td>unknown</td>
      <td>unknown</td>
      <td>unknown</td>
      <td>nucleus</td>
      <td>decidua basalis</td>
      <td>UBERON:0000453</td>
      <td>placenta</td>
      <td>UBERON:0001987</td>
    </tr>
    <tr>
      <th>1</th>
      <td>1</td>
      <td>f171db61-e57e-4535-a06a-35d8b6ef8f2b</td>
      <td>10x 3' v3</td>
      <td>EFO:0009922</td>
      <td>placental villous trophoblast</td>
      <td>CL:2000060</td>
      <td>9th week post-fertilization human stage</td>
      <td>HsapDv:0000046</td>
      <td>normal</td>
      <td>PATO:0000461</td>
      <td>...</td>
      <td>False</td>
      <td>unknown</td>
      <td>unknown</td>
      <td>unknown</td>
      <td>unknown</td>
      <td>nucleus</td>
      <td>decidua basalis</td>
      <td>UBERON:0000453</td>
      <td>placenta</td>
      <td>UBERON:0001987</td>
    </tr>
    <tr>
      <th>2</th>
      <td>2</td>
      <td>f171db61-e57e-4535-a06a-35d8b6ef8f2b</td>
      <td>10x 3' v3</td>
      <td>EFO:0009922</td>
      <td>syncytiotrophoblast cell</td>
      <td>CL:0000525</td>
      <td>9th week post-fertilization human stage</td>
      <td>HsapDv:0000046</td>
      <td>normal</td>
      <td>PATO:0000461</td>
      <td>...</td>
      <td>False</td>
      <td>unknown</td>
      <td>unknown</td>
      <td>unknown</td>
      <td>unknown</td>
      <td>nucleus</td>
      <td>decidua basalis</td>
      <td>UBERON:0000453</td>
      <td>placenta</td>
      <td>UBERON:0001987</td>
    </tr>
    <tr>
      <th>3</th>
      <td>3</td>
      <td>f171db61-e57e-4535-a06a-35d8b6ef8f2b</td>
      <td>10x 3' v3</td>
      <td>EFO:0009922</td>
      <td>syncytiotrophoblast cell</td>
      <td>CL:0000525</td>
      <td>9th week post-fertilization human stage</td>
      <td>HsapDv:0000046</td>
      <td>normal</td>
      <td>PATO:0000461</td>
      <td>...</td>
      <td>False</td>
      <td>unknown</td>
      <td>unknown</td>
      <td>unknown</td>
      <td>unknown</td>
      <td>nucleus</td>
      <td>decidua basalis</td>
      <td>UBERON:0000453</td>
      <td>placenta</td>
      <td>UBERON:0001987</td>
    </tr>
    <tr>
      <th>4</th>
      <td>4</td>
      <td>f171db61-e57e-4535-a06a-35d8b6ef8f2b</td>
      <td>10x 3' v3</td>
      <td>EFO:0009922</td>
      <td>extravillous trophoblast</td>
      <td>CL:0008036</td>
      <td>9th week post-fertilization human stage</td>
      <td>HsapDv:0000046</td>
      <td>normal</td>
      <td>PATO:0000461</td>
      <td>...</td>
      <td>False</td>
      <td>unknown</td>
      <td>unknown</td>
      <td>unknown</td>
      <td>unknown</td>
      <td>nucleus</td>
      <td>decidua basalis</td>
      <td>UBERON:0000453</td>
      <td>placenta</td>
      <td>UBERON:0001987</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>3251329</th>
      <td>56274573</td>
      <td>2adb1f8a-a6b1-4909-8ee8-484814e2d4bf</td>
      <td>microwell-seq</td>
      <td>EFO:0030002</td>
      <td>cord blood hematopoietic stem cell</td>
      <td>CL:2000095</td>
      <td>newborn human stage</td>
      <td>HsapDv:0000082</td>
      <td>normal</td>
      <td>PATO:0000461</td>
      <td>...</td>
      <td>True</td>
      <td>Han Chinese</td>
      <td>HANCESTRO:0027</td>
      <td>unknown</td>
      <td>unknown</td>
      <td>cell</td>
      <td>umbilical cord blood</td>
      <td>UBERON:0012168</td>
      <td>blood</td>
      <td>UBERON:0000178</td>
    </tr>
    <tr>
      <th>3251330</th>
      <td>56274574</td>
      <td>2adb1f8a-a6b1-4909-8ee8-484814e2d4bf</td>
      <td>microwell-seq</td>
      <td>EFO:0030002</td>
      <td>cord blood hematopoietic stem cell</td>
      <td>CL:2000095</td>
      <td>newborn human stage</td>
      <td>HsapDv:0000082</td>
      <td>normal</td>
      <td>PATO:0000461</td>
      <td>...</td>
      <td>True</td>
      <td>Han Chinese</td>
      <td>HANCESTRO:0027</td>
      <td>unknown</td>
      <td>unknown</td>
      <td>cell</td>
      <td>umbilical cord blood</td>
      <td>UBERON:0012168</td>
      <td>blood</td>
      <td>UBERON:0000178</td>
    </tr>
    <tr>
      <th>3251331</th>
      <td>56274575</td>
      <td>2adb1f8a-a6b1-4909-8ee8-484814e2d4bf</td>
      <td>microwell-seq</td>
      <td>EFO:0030002</td>
      <td>cord blood hematopoietic stem cell</td>
      <td>CL:2000095</td>
      <td>newborn human stage</td>
      <td>HsapDv:0000082</td>
      <td>normal</td>
      <td>PATO:0000461</td>
      <td>...</td>
      <td>True</td>
      <td>Han Chinese</td>
      <td>HANCESTRO:0027</td>
      <td>unknown</td>
      <td>unknown</td>
      <td>cell</td>
      <td>umbilical cord blood</td>
      <td>UBERON:0012168</td>
      <td>blood</td>
      <td>UBERON:0000178</td>
    </tr>
    <tr>
      <th>3251332</th>
      <td>56274576</td>
      <td>2adb1f8a-a6b1-4909-8ee8-484814e2d4bf</td>
      <td>microwell-seq</td>
      <td>EFO:0030002</td>
      <td>cord blood hematopoietic stem cell</td>
      <td>CL:2000095</td>
      <td>newborn human stage</td>
      <td>HsapDv:0000082</td>
      <td>normal</td>
      <td>PATO:0000461</td>
      <td>...</td>
      <td>True</td>
      <td>Han Chinese</td>
      <td>HANCESTRO:0027</td>
      <td>unknown</td>
      <td>unknown</td>
      <td>cell</td>
      <td>umbilical cord blood</td>
      <td>UBERON:0012168</td>
      <td>blood</td>
      <td>UBERON:0000178</td>
    </tr>
    <tr>
      <th>3251333</th>
      <td>56274577</td>
      <td>2adb1f8a-a6b1-4909-8ee8-484814e2d4bf</td>
      <td>microwell-seq</td>
      <td>EFO:0030002</td>
      <td>cord blood hematopoietic stem cell</td>
      <td>CL:2000095</td>
      <td>newborn human stage</td>
      <td>HsapDv:0000082</td>
      <td>normal</td>
      <td>PATO:0000461</td>
      <td>...</td>
      <td>True</td>
      <td>Han Chinese</td>
      <td>HANCESTRO:0027</td>
      <td>unknown</td>
      <td>unknown</td>
      <td>cell</td>
      <td>umbilical cord blood</td>
      <td>UBERON:0012168</td>
      <td>blood</td>
      <td>UBERON:0000178</td>
    </tr>
  </tbody>
</table>
<p>3251334 rows √ó 21 columns</p>
</div></div>
</div>
<p>You can use both <code class="docutils literal notranslate"><span class="pre">column_names</span></code> and <code class="docutils literal notranslate"><span class="pre">value_filter</span></code> to perform specific queries. For example let‚Äôs fetch the <code class="docutils literal notranslate"><span class="pre">disease</span></code> columns for the <code class="docutils literal notranslate"><span class="pre">cell_type</span></code> <code class="docutils literal notranslate"><span class="pre">&quot;B</span> <span class="pre">cell&quot;</span></code> in the <code class="docutils literal notranslate"><span class="pre">tissue_general</span></code> <code class="docutils literal notranslate"><span class="pre">&quot;lung&quot;</span></code> and from non-duplicated cells.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">cell_metadata_b_cell</span> <span class="o">=</span> <span class="p">(</span>
    <span class="n">census</span><span class="p">[</span><span class="s2">&quot;census_data&quot;</span><span class="p">][</span><span class="s2">&quot;homo_sapiens&quot;</span><span class="p">]</span>
    <span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">read</span><span class="p">(</span>
        <span class="n">value_filter</span><span class="o">=</span><span class="s2">&quot;cell_type == &#39;B cell&#39; and tissue_general == &#39;lung&#39; and is_primary_data==True&quot;</span><span class="p">,</span>
        <span class="n">column_names</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;disease&quot;</span><span class="p">],</span>
    <span class="p">)</span>
    <span class="o">.</span><span class="n">concat</span><span class="p">()</span>
    <span class="o">.</span><span class="n">to_pandas</span><span class="p">()</span>
<span class="p">)</span>

<span class="n">cell_metadata_b_cell</span><span class="o">.</span><span class="n">value_counts</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
disease                                cell_type  tissue_general  is_primary_data
lung adenocarcinoma                    B cell     lung            True               42720
squamous cell lung carcinoma           B cell     lung            True               10631
non-small cell lung carcinoma          B cell     lung            True                8742
normal                                 B cell     lung            True                8187
COVID-19                               B cell     lung            True                2313
chronic obstructive pulmonary disease  B cell     lung            True                2083
lung large cell carcinoma              B cell     lung            True                1534
pulmonary emphysema                    B cell     lung            True                1512
pulmonary fibrosis                     B cell     lung            True                1474
pleomorphic carcinoma                  B cell     lung            True                1210
interstitial lung disease              B cell     lung            True                 332
small cell lung carcinoma              B cell     lung            True                 204
lymphangioleiomyomatosis               B cell     lung            True                 133
pneumonia                              B cell     lung            True                  50
Name: count, dtype: int64
</pre></div></div>
</div>
</section>
<section id="Querying-gene-metadata-(var)">
<h2>Querying gene metadata (var)<a class="headerlink" href="#Querying-gene-metadata-(var)" title="Permalink to this heading">¬∂</a></h2>
<p>The human gene metadata of the Census is located at <code class="docutils literal notranslate"><span class="pre">census[&quot;census_data&quot;][&quot;homo_sapiens&quot;].ms[&quot;RNA&quot;].var</span></code>. Similarly to the cell metadata, it is a <code class="docutils literal notranslate"><span class="pre">SOMADataFrame</span></code> and thus we can also use its method <code class="docutils literal notranslate"><span class="pre">read()</span></code>.</p>
<p>The mouse gene metadata is at <code class="docutils literal notranslate"><span class="pre">census[&quot;census_data&quot;][&quot;mus_musculus&quot;].ms[&quot;RNA&quot;].var</span></code>.</p>
<p>Let‚Äôs take a look at the metadata available for column selection and row filtering.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">keys</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">census</span><span class="p">[</span><span class="s2">&quot;census_data&quot;</span><span class="p">][</span><span class="s2">&quot;homo_sapiens&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">ms</span><span class="p">[</span><span class="s2">&quot;RNA&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span>

<span class="n">keys</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
[&#39;soma_joinid&#39;, &#39;feature_id&#39;, &#39;feature_name&#39;, &#39;feature_length&#39;]
</pre></div></div>
</div>
<p>With the exception of <code class="docutils literal notranslate"><span class="pre">soma_joinid</span></code> these columns are defined in the <a class="reference external" href="https://github.com/chanzuckerberg/cellxgene-census/blob/main/docs/cell_census_schema_0.1.0.md">Census schema</a>. Similarly to the cell metadata, we can use the same operations to learn and fetch gene metadata.</p>
<p>For example, to get the <code class="docutils literal notranslate"><span class="pre">feature_name</span></code> and <code class="docutils literal notranslate"><span class="pre">feature_length</span></code> of the genes <code class="docutils literal notranslate"><span class="pre">&quot;ENSG00000161798&quot;</span></code> and <code class="docutils literal notranslate"><span class="pre">&quot;ENSG00000188229&quot;</span></code> we can do the following.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">gene_metadata</span> <span class="o">=</span> <span class="p">(</span>
    <span class="n">census</span><span class="p">[</span><span class="s2">&quot;census_data&quot;</span><span class="p">][</span><span class="s2">&quot;homo_sapiens&quot;</span><span class="p">]</span>
    <span class="o">.</span><span class="n">ms</span><span class="p">[</span><span class="s2">&quot;RNA&quot;</span><span class="p">]</span>
    <span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">read</span><span class="p">(</span>
        <span class="n">value_filter</span><span class="o">=</span><span class="s2">&quot;feature_id in [&#39;ENSG00000161798&#39;, &#39;ENSG00000188229&#39;]&quot;</span><span class="p">,</span>
        <span class="n">column_names</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;feature_name&quot;</span><span class="p">,</span> <span class="s2">&quot;feature_length&quot;</span><span class="p">],</span>
    <span class="p">)</span>
    <span class="o">.</span><span class="n">concat</span><span class="p">()</span>
    <span class="o">.</span><span class="n">to_pandas</span><span class="p">()</span>
<span class="p">)</span>

<span class="n">gene_metadata</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>feature_name</th>
      <th>feature_length</th>
      <th>feature_id</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>AQP5</td>
      <td>1884</td>
      <td>ENSG00000161798</td>
    </tr>
    <tr>
      <th>1</th>
      <td>TUBB4B</td>
      <td>2037</td>
      <td>ENSG00000188229</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">census</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</pre></div>
</div>
</div>
</section>
</section>


            </div>
            
            </div>
            <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="../../examples.html" class="btn btn-neutral float-left" title="Python tutorials" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="census_citation_generation.html" class="btn btn-neutral float-right" title="Generating citations for Census slices" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2022, Chan Zuckerberg Initiative.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
            </div>
        </div>

        </section>

    </div>
    <!-- Newsletter Banner -->
    <div role="banner" id="newsletter-banner">
      <span id="newsletter-subscribe-button" role="button">Subscribe</span>&nbsp;to our newsletter to receive updates about new features.
      <div id="newsletter-banner-close-button" role="button">X</div>
    </div>
    <!-- Newsletter Modal -->
    <dialog id="newsletter-modal">
      <div id="newsletter-header">
        <img id="newsletter-logo" src="../../_static/img/cellxGene-newsletter-logo.svg" />
        <div id="newsletter-close-button" role="button">X</div>
      </div>
      <div id="newsletter-content">
        <div id="newsletter-callout">Join Our Newsletter</div>
        <div id="newsletter-description">Get a quarterly email with the latest CELLxGENE features and data.</div>
        <!-- HubSpot Form target -->
        <div id="newsletter-form-container"></div>
      </div>

      <div id="newsletter-footnote">Unsubscribe at any time.</div>
    </dialog>
  </div>
  

  <script>
    // (thuang): 30 days
    const NEWSLETTER_BANNER_DISMISSED_TTL_MS = 30 * 24 * 60 * 60 * 1000;
    const NEWSLETTER_BANNER_DISMISSED_KEY = "newsletterBannerDismissed"

    var script = document.createElement('script');
    script.src = 'https://js.hsforms.net/forms/v2.js';
    script.defer = true;
    document.head.appendChild(script);

    // Run the code once the script is loaded
    script.onload = async function() {
      await hbspt.forms.create({
        region: "na1",
        portalId: "7272273",
        formId: "eb65b811-0451-414d-8304-7b9b6f468ce5",
        target: '#newsletter-form-container',
        onFormReady() {
          // get element by type "email"
          const emailInput = document.querySelector('#email-eb65b811-0451-414d-8304-7b9b6f468ce5');
          emailInput.setAttribute('placeholder', 'Enter email address');

          // remove the label element for emailInput
          const emailLabel = document.querySelector('#label-email-eb65b811-0451-414d-8304-7b9b6f468ce5');
          emailLabel.remove();
        },
        submitText: 'Subscribe',
      });
    };

    checkNewsletterBanner();

    document.querySelector('#newsletter-banner-close-button').addEventListener('click', () => {
      document.querySelector('#newsletter-banner').remove();
      localStorage.setItem(NEWSLETTER_BANNER_DISMISSED_KEY, Date.now());
    });

    const modal = document.querySelector('#newsletter-modal');

    document.querySelector('#newsletter-subscribe-button').addEventListener('click', () => {
      modal.showModal();
    });

    document.querySelector('#newsletter-close-button').addEventListener('click', () => {
      modal.close();
    });

    function checkNewsletterBanner() {
      /**
       * (thuang): Use LocalStorage to store dismissed state for 30 days
       * NOTE: Currently Census doc page doesn't share the same domain as the main site,
       * so dismissing the banner on the main site won't dismiss it on the Census doc page.
       * And vice versa.
       */
      const newsletterBannerDismissed = localStorage.getItem('newsletterBannerDismissed');

      if (newsletterBannerDismissed) {
        return;
      }

      if (newsletterBannerDismissed && Date.now() - newsletterBannerDismissed > NEWSLETTER_BANNER_DISMISSED_TTL_MS) {
        localStorage.removeItem(NEWSLETTER_BANNER_DISMISSED_KEY);
      }

      const newsletterBanner = document.querySelector('#newsletter-banner');

      if (!newsletterBannerDismissed) {
        newsletterBanner.style.display = 'flex';
      }
    }
  </script>

  
  
    
   

</body>
</html>