{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "7879c75b",
   "metadata": {},
   "source": [
    "# Training a PyTorch Model\n",
    "\n",
    "This tutorial demonstrates training a simple Logistic Regression model using Census data, PyTorch, and [TileDB-SOMA-ML](https://github.com/single-cell-data/TileDB-SOMA-ML) (which provides a PyTorch loader for SOMA datasets like Census).\n",
    "\n",
    "We assume basic familiarity with PyTorch and the Census API. See the [Querying and fetching the single-cell data and cell/gene metadata](https://chanzuckerberg.github.io/cellxgene-census/notebooks/api_demo/census_query_extract.html) notebook tutorial for a quick primer on Census API usage.\n",
    "\n",
    "**Contents**\n",
    "\n",
    "* [Create an ExperimentDataset](#data-pipe)\n",
    "* [Split the dataset](#split)\n",
    "* [Create the DataLoader](#data-loader)\n",
    "* [Define the model](#model)\n",
    "* [Train the model](#train)\n",
    "* [Make predictions with the model](#predict)\n",
    "\n",
    "[CZI CELLxGENE Census]: https://chanzuckerberg.github.io/cellxgene-census/"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "1114d8e7-172b-4097-b873-e353e9911fe8",
   "metadata": {},
   "source": [
    "## Open the Census\n",
    "\n",
    "First, obtain a handle to the Census data, in the usual manner:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "c3dd549f",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-10-09T18:20:21.600206Z",
     "start_time": "2023-10-09T18:20:19.390343Z"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "The \"stable\" release is currently 2025-01-30. Specify 'census_version=\"2025-01-30\"' in future calls to open_soma() to ensure data consistency.\n"
     ]
    }
   ],
   "source": [
    "import cellxgene_census\n",
    "\n",
    "census = cellxgene_census.open_soma()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2e06cb34",
   "metadata": {},
   "source": [
    "## Create an ExperimentDataset <a id=\"data-pipe\"></a>\n",
    "\n",
    "To set up a PyTorch dataset from a Census slice, open an [`ExperimentAxisQuery`] and create a TileDB-SOMA-ML ExperimentDataset for it. We will also prepare a scikit-learn [`LabelEncoder`] to help with mapping between Census cell types (categorical metadata) and PyTorch tensors.\n",
    "\n",
    "[`ExperimentAxisQuery`]: https://tiledbsoma.readthedocs.io/en/stable/python-tiledbsoma-experimentaxisquery.html\n",
    "[`LabelEncoder`]: https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.LabelEncoder.html"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "b4cfc7d7",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "import tiledbsoma as soma\n",
    "from tiledbsoma_ml import ExperimentDataset\n",
    "from sklearn.preprocessing import LabelEncoder\n",
    "\n",
    "experiment = census[\"census_data\"][\"homo_sapiens\"]\n",
    "with experiment.axis_query(\n",
    "    measurement_name=\"RNA\", \n",
    "    obs_query=soma.AxisQuery(\n",
    "        value_filter=\"tissue_general == 'tongue' and is_primary_data == True\"\n",
    "    )\n",
    ") as query:\n",
    "    experiment_dataset = ExperimentDataset(\n",
    "        query,\n",
    "        layer_name=\"raw\",\n",
    "        obs_column_names=[\"cell_type\"],\n",
    "        batch_size=128,\n",
    "        shuffle=True,\n",
    "        seed=111,\n",
    "    )\n",
    "    obs_df = query.obs(column_names=[\"cell_type\"]).concat().to_pandas()\n",
    "    cell_type_encoder = LabelEncoder().fit(obs_df[\"cell_type\"].unique())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "599cfe83",
   "metadata": {},
   "source": [
    "### `ExperimentDataset` class explained\n",
    "\n",
    "This class provides an implementation of PyTorch's `IterableDataset` interface for accessing data from abstract sources. It encapsulates streaming the result set of a SOMA [`ExperimentAxisQuery`] in a series of \"batches,\" each consisting of a NumPy `ndarray` (batch of X results) and a Pandas `DataFrame` (corresponding batch of obs results). Most importantly, it avoids loading large result sets into memory all at once.\n",
    "\n",
    "The constructor requires the [`ExperimentAxisQuery`] defining the desired slice of Census data, and the name of the X layer to access.\n",
    "\n",
    "`obs_column_names` sets the columns to be included in the DataFrame for each batch, to be used as data labels or model inputs.\n",
    "\n",
    "The `batch_size` parameter specifies the target number of rows (cells) in each batch.\n",
    "\n",
    "The `shuffle` flag supports randomizing the ordering of the training data for each training epoch (default: `True`). This should be used instead of [`DataLoader`]'s `shuffle` flag, as the implementation is more efficient.\n",
    "\n",
    "[`ExperimentAxisQuery`]: https://tiledbsoma.readthedocs.io/en/stable/python-tiledbsoma-experimentaxisquery.html\n",
    "[`DataLoader`]: https://pytorch.org/docs/stable/data.html"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1eb8e653",
   "metadata": {},
   "source": [
    "You can inspect the shape of the full dataset, without causing the full dataset to be loaded. The `shape` property returns the number of batches on the first dimension:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "b943c146",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(303, 61888)"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "experiment_dataset.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "932aa3fd-0b7f-410e-b259-b65269cbb42d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(38754,)"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "experiment_dataset.query_ids.obs_joinids.shape # total result count"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2a46d45e",
   "metadata": {},
   "source": [
    "## Split the dataset <a id=\"split\"></a>\n",
    "\n",
    "You may split the query results into the typical training and test sets using the `ExperimentDataset.random_split()` method:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "02102906",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "((243, 61888), (61, 61888))"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "train_dataset, test_dataset = experiment_dataset.random_split(\n",
    "    0.8, 0.2,\n",
    "    seed=111,\n",
    ")\n",
    "train_dataset.shape, test_dataset.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "f1fa0114",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "((31003,), (7751,))"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "train_dataset.query_ids.obs_joinids.shape, test_dataset.query_ids.obs_joinids.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "92458740",
   "metadata": {},
   "source": [
    "## Create the DataLoader <a id=\"data-loader\"></a>\n",
    "\n",
    "Now you can prepare a PyTorch [`DataLoader`](https://pytorch.org/docs/stable/data.html#torch.utils.data.DataLoader) on the training data. While not shown here, the DataLoader can be configured with multiple worker processes to stream the result batches as quickly as possible."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "779a213f",
   "metadata": {},
   "outputs": [],
   "source": [
    "from tiledbsoma_ml import experiment_dataloader\n",
    "\n",
    "train_dataloader = experiment_dataloader(train_dataset)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "14e66e6d",
   "metadata": {},
   "source": [
    "(Instantiating a PyTorch `DataLoader` object directly is not recommended, as `experiment_dataloader()` enforces correct and performant usage.)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7ffe0bc3",
   "metadata": {},
   "source": [
    "## Define the model <a id=\"model\"></a>\n",
    "\n",
    "With the training data retrieval code now in place, we can move on to defining a simple logistic regression model, using PyTorch's `torch.nn.Linear` class:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "8ed00d8f",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "import torch\n",
    "\n",
    "class LogisticRegression(torch.nn.Module):\n",
    "    def __init__(self, input_dim, output_dim):\n",
    "        super(LogisticRegression, self).__init__()\n",
    "        self.linear = torch.nn.Linear(input_dim, output_dim)\n",
    "\n",
    "    def forward(self, x):\n",
    "        outputs = torch.sigmoid(self.linear(x))\n",
    "        return outputs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "17b35caf",
   "metadata": {},
   "source": [
    "Next, we define a function to train the model for a single epoch:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "f9cb9807",
   "metadata": {},
   "outputs": [],
   "source": [
    "def train_epoch(model, train_dataloader, loss_fn, optimizer, device):\n",
    "    model.train()\n",
    "    train_loss = 0\n",
    "    train_correct = 0\n",
    "    train_total = 0\n",
    "\n",
    "    for X_batch, obs_batch in train_dataloader:\n",
    "        optimizer.zero_grad()\n",
    "\n",
    "        # convert X_batch numpy ndarray into PyTorch tensor\n",
    "        X_batch = torch.from_numpy(X_batch).float().to(device)\n",
    "\n",
    "        # Perform prediction\n",
    "        outputs = model(X_batch)\n",
    "\n",
    "        # Determine the predicted label\n",
    "        probabilities = torch.nn.functional.softmax(outputs, 1)\n",
    "        predictions = torch.argmax(probabilities, axis=1)\n",
    "\n",
    "        # Compute the loss using encoded labels and perform back propagation\n",
    "        label_batch = torch.from_numpy(cell_type_encoder.transform(obs_batch['cell_type'])).to(device)\n",
    "        train_correct += (predictions == label_batch).sum().item()\n",
    "        train_total += len(predictions)\n",
    "\n",
    "        loss = loss_fn(outputs, label_batch.long())\n",
    "        train_loss += loss.item()\n",
    "        loss.backward()\n",
    "        optimizer.step()\n",
    "\n",
    "    train_loss /= train_total\n",
    "    train_accuracy = train_correct / train_total\n",
    "    return train_loss, train_accuracy"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4d69717b",
   "metadata": {},
   "source": [
    "The `X_batch` ndarray (later tensor) has the slice of the RNA X matrix for the current batch of results. Similarly, the `obs_batch` dataframe has the slice of `obs` with `cell_type` metadata. We use `cell_type_encoder` to encode these categorical cell types into integer-valued tensors for comparison with the model predictions."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "588d5db3",
   "metadata": {},
   "source": [
    "## Train the model <a id=\"train\"></a>\n",
    "\n",
    "Finally, we are ready to train the model. Here we instantiate the model, a loss function, and an optimization method, then iterate through the desired number of training epochs. Note how the `train_dataloader` is passed into `train_epoch`, where for each epoch it will provide a new iterator through the training dataset."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "0df30c3f",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Epoch 1: Train Loss: 0.0198963 Accuracy 0.4256\n",
      "Epoch 2: Train Loss: 0.0185954 Accuracy 0.6008\n",
      "Epoch 3: Train Loss: 0.0183252 Accuracy 0.7415\n",
      "Epoch 4: Train Loss: 0.0182038 Accuracy 0.8212\n",
      "Epoch 5: Train Loss: 0.0180737 Accuracy 0.8420\n",
      "Epoch 6: Train Loss: 0.0179905 Accuracy 0.8501\n",
      "Epoch 7: Train Loss: 0.0179427 Accuracy 0.8534\n",
      "Epoch 8: Train Loss: 0.0179072 Accuracy 0.8564\n",
      "Epoch 9: Train Loss: 0.0178783 Accuracy 0.8590\n",
      "Epoch 10: Train Loss: 0.0178454 Accuracy 0.8625\n",
      "Epoch 11: Train Loss: 0.0178180 Accuracy 0.8664\n",
      "Epoch 12: Train Loss: 0.0177930 Accuracy 0.8702\n",
      "Epoch 13: Train Loss: 0.0177638 Accuracy 0.8713\n",
      "Epoch 14: Train Loss: 0.0177478 Accuracy 0.8724\n",
      "Epoch 15: Train Loss: 0.0177347 Accuracy 0.8743\n",
      "Epoch 16: Train Loss: 0.0177222 Accuracy 0.8759\n",
      "Epoch 17: Train Loss: 0.0177065 Accuracy 0.8754\n",
      "Epoch 18: Train Loss: 0.0176974 Accuracy 0.8766\n",
      "Epoch 19: Train Loss: 0.0176894 Accuracy 0.8774\n",
      "Epoch 20: Train Loss: 0.0176822 Accuracy 0.8777\n",
      "Epoch 21: Train Loss: 0.0176752 Accuracy 0.8793\n",
      "Epoch 22: Train Loss: 0.0176681 Accuracy 0.8790\n",
      "Epoch 23: Train Loss: 0.0176624 Accuracy 0.8804\n",
      "Epoch 24: Train Loss: 0.0176570 Accuracy 0.8812\n",
      "Epoch 25: Train Loss: 0.0176558 Accuracy 0.8829\n"
     ]
    }
   ],
   "source": [
    "device = torch.device(\"cuda\") if torch.cuda.is_available() else torch.device(\"cpu\")\n",
    "\n",
    "# The size of the input dimension is the number of genes\n",
    "input_dim = experiment_dataset.shape[1]\n",
    "\n",
    "# The size of the output dimension is the number of distinct cell_type values\n",
    "output_dim = len(cell_type_encoder.classes_)\n",
    "\n",
    "model = LogisticRegression(input_dim, output_dim).to(device)\n",
    "loss_fn = torch.nn.CrossEntropyLoss()\n",
    "optimizer = torch.optim.Adam(model.parameters(), lr=1e-5)\n",
    "\n",
    "for epoch in range(25):\n",
    "    train_loss, train_accuracy = train_epoch(model, train_dataloader, loss_fn, optimizer, device)\n",
    "    print(f\"Epoch {epoch + 1}: Train Loss: {train_loss:.7f} Accuracy {train_accuracy:.4f}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3813df12",
   "metadata": {},
   "source": [
    "## Make predictions with the model <a id=\"predict\"></a>\n",
    "\n",
    "To make predictions with the model, we first create a new `DataLoader` using the `test_dataset`, which provides the \"test\" split of the original dataset. For this example, we will only make predictions on a single batch of data from the test split."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "be58abca",
   "metadata": {},
   "outputs": [],
   "source": [
    "test_dataloader = experiment_dataloader(test_dataset)\n",
    "X_batch, obs_batch = next(iter(test_dataloader))\n",
    "X_batch = torch.from_numpy(X_batch)\n",
    "true_cell_types = torch.from_numpy(cell_type_encoder.transform(obs_batch['cell_type']))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "73c84158",
   "metadata": {},
   "source": [
    "Next, we invoke the model on the `X_batch` input data and extract the predictions:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "65f4bf4d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "tensor([ 5,  8,  8, 20, 20, 20, 20,  5,  5, 20,  5,  8, 19,  9, 22,  5,  6,  5,\n",
       "        20, 20,  5,  9,  5, 20, 20,  1,  8,  5,  1,  0,  8,  5, 20,  5, 20,  5,\n",
       "         8, 20, 20,  5, 20,  1, 20, 20,  5, 20, 10,  9,  5,  5, 20,  5,  6,  9,\n",
       "         2, 12, 20,  8, 20,  5,  5, 20,  5,  5, 20, 20,  5,  1,  5,  5,  9, 20,\n",
       "         5, 20, 20,  5,  5,  5,  6, 12,  5,  5, 20,  1,  5,  5,  5, 20, 20, 20,\n",
       "         5,  5, 20, 22, 20,  5,  5,  5,  5,  5, 20,  2,  1,  5,  5,  5,  5,  5,\n",
       "        20,  5,  5, 20,  0, 22,  5,  1, 20,  2, 20, 20, 19,  5,  5,  5,  5,  5,\n",
       "        16, 19], device='cuda:0')"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.eval()\n",
    "\n",
    "model.to(device)\n",
    "outputs = model(X_batch.to(device))\n",
    "\n",
    "probabilities = torch.nn.functional.softmax(outputs, 1)\n",
    "predictions = torch.argmax(probabilities, axis=1)\n",
    "predictions"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3dc9de3f",
   "metadata": {},
   "source": [
    "The predictions are returned as the encoded values of `cell_type` label. To recover the original cell type labels as strings, we decode using the same `LabelEncoder` used for training.\n",
    "\n",
    "At inference time, if the model inputs are not obtained via an `ExperimentDataset`, one could pickle the encoder at training time and save it along with the model. Then, at inference time it can be unpickled and used as shown below."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "b1f20813",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['basal cell', 'fibroblast', 'fibroblast',\n",
       "       'stratified squamous epithelial cell',\n",
       "       'stratified squamous epithelial cell',\n",
       "       'stratified squamous epithelial cell',\n",
       "       'stratified squamous epithelial cell', 'basal cell', 'basal cell',\n",
       "       'stratified squamous epithelial cell', 'basal cell', 'fibroblast',\n",
       "       'salivary gland cell', 'macrophage', 'tongue muscle cell',\n",
       "       'basal cell', 'endothelial cell', 'basal cell',\n",
       "       'stratified squamous epithelial cell',\n",
       "       'stratified squamous epithelial cell', 'basal cell', 'macrophage',\n",
       "       'basal cell', 'stratified squamous epithelial cell',\n",
       "       'stratified squamous epithelial cell',\n",
       "       'CD4-positive, alpha-beta T cell', 'fibroblast', 'basal cell',\n",
       "       'CD4-positive, alpha-beta T cell', 'B cell', 'fibroblast',\n",
       "       'basal cell', 'stratified squamous epithelial cell', 'basal cell',\n",
       "       'stratified squamous epithelial cell', 'basal cell', 'fibroblast',\n",
       "       'stratified squamous epithelial cell',\n",
       "       'stratified squamous epithelial cell', 'basal cell',\n",
       "       'stratified squamous epithelial cell',\n",
       "       'CD4-positive, alpha-beta T cell',\n",
       "       'stratified squamous epithelial cell',\n",
       "       'stratified squamous epithelial cell', 'basal cell',\n",
       "       'stratified squamous epithelial cell', 'mast cell', 'macrophage',\n",
       "       'basal cell', 'basal cell', 'stratified squamous epithelial cell',\n",
       "       'basal cell', 'endothelial cell', 'macrophage',\n",
       "       'CD8-positive, alpha-beta T cell', 'mural cell',\n",
       "       'stratified squamous epithelial cell', 'fibroblast',\n",
       "       'stratified squamous epithelial cell', 'basal cell', 'basal cell',\n",
       "       'stratified squamous epithelial cell', 'basal cell', 'basal cell',\n",
       "       'stratified squamous epithelial cell',\n",
       "       'stratified squamous epithelial cell', 'basal cell',\n",
       "       'CD4-positive, alpha-beta T cell', 'basal cell', 'basal cell',\n",
       "       'macrophage', 'stratified squamous epithelial cell', 'basal cell',\n",
       "       'stratified squamous epithelial cell',\n",
       "       'stratified squamous epithelial cell', 'basal cell', 'basal cell',\n",
       "       'basal cell', 'endothelial cell', 'mural cell', 'basal cell',\n",
       "       'basal cell', 'stratified squamous epithelial cell',\n",
       "       'CD4-positive, alpha-beta T cell', 'basal cell', 'basal cell',\n",
       "       'basal cell', 'stratified squamous epithelial cell',\n",
       "       'stratified squamous epithelial cell',\n",
       "       'stratified squamous epithelial cell', 'basal cell', 'basal cell',\n",
       "       'stratified squamous epithelial cell', 'tongue muscle cell',\n",
       "       'stratified squamous epithelial cell', 'basal cell', 'basal cell',\n",
       "       'basal cell', 'basal cell', 'basal cell',\n",
       "       'stratified squamous epithelial cell',\n",
       "       'CD8-positive, alpha-beta T cell',\n",
       "       'CD4-positive, alpha-beta T cell', 'basal cell', 'basal cell',\n",
       "       'basal cell', 'basal cell', 'basal cell',\n",
       "       'stratified squamous epithelial cell', 'basal cell', 'basal cell',\n",
       "       'stratified squamous epithelial cell', 'B cell',\n",
       "       'tongue muscle cell', 'basal cell',\n",
       "       'CD4-positive, alpha-beta T cell',\n",
       "       'stratified squamous epithelial cell',\n",
       "       'CD8-positive, alpha-beta T cell',\n",
       "       'stratified squamous epithelial cell',\n",
       "       'stratified squamous epithelial cell', 'salivary gland cell',\n",
       "       'basal cell', 'basal cell', 'basal cell', 'basal cell',\n",
       "       'basal cell', 'neutrophil', 'salivary gland cell'], dtype=object)"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "predicted_cell_types = cell_type_encoder.inverse_transform(predictions.cpu())\n",
    "predicted_cell_types"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6721e8f1",
   "metadata": {},
   "source": [
    "Finally, we create a Pandas DataFrame to examine the predictions:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "a6317e11",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>true cell type</th>\n",
       "      <th>predicted cell type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>stratified squamous epithelial cell</td>\n",
       "      <td>basal cell</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>fibroblast</td>\n",
       "      <td>fibroblast</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>basal cell</td>\n",
       "      <td>fibroblast</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>stratified squamous epithelial cell</td>\n",
       "      <td>stratified squamous epithelial cell</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>stratified squamous epithelial cell</td>\n",
       "      <td>stratified squamous epithelial cell</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>123</th>\n",
       "      <td>basal cell</td>\n",
       "      <td>basal cell</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>124</th>\n",
       "      <td>basal cell</td>\n",
       "      <td>basal cell</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>125</th>\n",
       "      <td>basal cell</td>\n",
       "      <td>basal cell</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>126</th>\n",
       "      <td>basal cell</td>\n",
       "      <td>neutrophil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>127</th>\n",
       "      <td>salivary gland cell</td>\n",
       "      <td>salivary gland cell</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>128 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                          true cell type                  predicted cell type\n",
       "0    stratified squamous epithelial cell                           basal cell\n",
       "1                             fibroblast                           fibroblast\n",
       "2                             basal cell                           fibroblast\n",
       "3    stratified squamous epithelial cell  stratified squamous epithelial cell\n",
       "4    stratified squamous epithelial cell  stratified squamous epithelial cell\n",
       "..                                   ...                                  ...\n",
       "123                           basal cell                           basal cell\n",
       "124                           basal cell                           basal cell\n",
       "125                           basal cell                           basal cell\n",
       "126                           basal cell                           neutrophil\n",
       "127                  salivary gland cell                  salivary gland cell\n",
       "\n",
       "[128 rows x 2 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "batch_cmp_df = pd.DataFrame({\n",
    "    \"true cell type\": cell_type_encoder.inverse_transform(true_cell_types.ravel().numpy()),\n",
    "    \"predicted cell type\": predicted_cell_types,\n",
    "})\n",
    "batch_cmp_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "e45e1b86",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>predicted cell type</th>\n",
       "      <th>B cell</th>\n",
       "      <th>CD4-positive, alpha-beta T cell</th>\n",
       "      <th>CD8-positive, alpha-beta T cell</th>\n",
       "      <th>basal cell</th>\n",
       "      <th>endothelial cell</th>\n",
       "      <th>fibroblast</th>\n",
       "      <th>macrophage</th>\n",
       "      <th>mast cell</th>\n",
       "      <th>mural cell</th>\n",
       "      <th>neutrophil</th>\n",
       "      <th>salivary gland cell</th>\n",
       "      <th>stratified squamous epithelial cell</th>\n",
       "      <th>tongue muscle cell</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>true cell type</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>B cell</th>\n",
       "      <td>2</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CD4-positive, alpha-beta T cell</th>\n",
       "      <td></td>\n",
       "      <td>6</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CD8-positive, alpha-beta T cell</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>3</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Schwann cell</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>basal cell</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>45</td>\n",
       "      <td></td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>4</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>endothelial cell</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>3</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>fibroblast</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>6</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>macrophage</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>2</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mast cell</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>monocyte</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>3</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mural cell</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>myoepithelial cell</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>regulatory T cell</th>\n",
       "      <td></td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>salivary gland cell</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>2</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>3</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>stratified squamous epithelial cell</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>5</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>34</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>tongue muscle cell</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "predicted cell type                 B cell CD4-positive, alpha-beta T cell  \\\n",
       "true cell type                                                               \n",
       "B cell                                   2                                   \n",
       "CD4-positive, alpha-beta T cell                                          6   \n",
       "CD8-positive, alpha-beta T cell                                              \n",
       "Schwann cell                                                                 \n",
       "basal cell                                                                   \n",
       "endothelial cell                                                             \n",
       "fibroblast                                                                   \n",
       "macrophage                                                                   \n",
       "mast cell                                                                    \n",
       "monocyte                                                                     \n",
       "mural cell                                                                   \n",
       "myoepithelial cell                                                           \n",
       "regulatory T cell                                                        1   \n",
       "salivary gland cell                                                          \n",
       "stratified squamous epithelial cell                                          \n",
       "tongue muscle cell                                                           \n",
       "\n",
       "predicted cell type                 CD8-positive, alpha-beta T cell  \\\n",
       "true cell type                                                        \n",
       "B cell                                                                \n",
       "CD4-positive, alpha-beta T cell                                       \n",
       "CD8-positive, alpha-beta T cell                                   3   \n",
       "Schwann cell                                                          \n",
       "basal cell                                                            \n",
       "endothelial cell                                                      \n",
       "fibroblast                                                            \n",
       "macrophage                                                            \n",
       "mast cell                                                             \n",
       "monocyte                                                              \n",
       "mural cell                                                            \n",
       "myoepithelial cell                                                    \n",
       "regulatory T cell                                                     \n",
       "salivary gland cell                                                   \n",
       "stratified squamous epithelial cell                                   \n",
       "tongue muscle cell                                                    \n",
       "\n",
       "predicted cell type                 basal cell endothelial cell fibroblast  \\\n",
       "true cell type                                                               \n",
       "B cell                                                                       \n",
       "CD4-positive, alpha-beta T cell                                              \n",
       "CD8-positive, alpha-beta T cell                                              \n",
       "Schwann cell                                                                 \n",
       "basal cell                                  45                           1   \n",
       "endothelial cell                                              3              \n",
       "fibroblast                                                               6   \n",
       "macrophage                                                                   \n",
       "mast cell                                                                    \n",
       "monocyte                                                                     \n",
       "mural cell                                                                   \n",
       "myoepithelial cell                           1                               \n",
       "regulatory T cell                                                            \n",
       "salivary gland cell                          2                               \n",
       "stratified squamous epithelial cell          5                               \n",
       "tongue muscle cell                                                           \n",
       "\n",
       "predicted cell type                 macrophage mast cell mural cell  \\\n",
       "true cell type                                                        \n",
       "B cell                                                                \n",
       "CD4-positive, alpha-beta T cell                                       \n",
       "CD8-positive, alpha-beta T cell                                       \n",
       "Schwann cell                                                          \n",
       "basal cell                                                            \n",
       "endothelial cell                                                      \n",
       "fibroblast                                                            \n",
       "macrophage                                   2                        \n",
       "mast cell                                              1              \n",
       "monocyte                                     3                        \n",
       "mural cell                                                        1   \n",
       "myoepithelial cell                                                    \n",
       "regulatory T cell                                                     \n",
       "salivary gland cell                                               1   \n",
       "stratified squamous epithelial cell                                   \n",
       "tongue muscle cell                                                    \n",
       "\n",
       "predicted cell type                 neutrophil salivary gland cell  \\\n",
       "true cell type                                                       \n",
       "B cell                                                               \n",
       "CD4-positive, alpha-beta T cell                                      \n",
       "CD8-positive, alpha-beta T cell                                      \n",
       "Schwann cell                                                         \n",
       "basal cell                                   1                       \n",
       "endothelial cell                                                     \n",
       "fibroblast                                                           \n",
       "macrophage                                                           \n",
       "mast cell                                                            \n",
       "monocyte                                                             \n",
       "mural cell                                                           \n",
       "myoepithelial cell                                                   \n",
       "regulatory T cell                                                    \n",
       "salivary gland cell                                              3   \n",
       "stratified squamous epithelial cell                                  \n",
       "tongue muscle cell                                                   \n",
       "\n",
       "predicted cell type                 stratified squamous epithelial cell  \\\n",
       "true cell type                                                            \n",
       "B cell                                                                    \n",
       "CD4-positive, alpha-beta T cell                                           \n",
       "CD8-positive, alpha-beta T cell                                           \n",
       "Schwann cell                                                              \n",
       "basal cell                                                            4   \n",
       "endothelial cell                                                          \n",
       "fibroblast                                                                \n",
       "macrophage                                                                \n",
       "mast cell                                                                 \n",
       "monocyte                                                                  \n",
       "mural cell                                                                \n",
       "myoepithelial cell                                                        \n",
       "regulatory T cell                                                         \n",
       "salivary gland cell                                                       \n",
       "stratified squamous epithelial cell                                  34   \n",
       "tongue muscle cell                                                        \n",
       "\n",
       "predicted cell type                 tongue muscle cell  \n",
       "true cell type                                          \n",
       "B cell                                                  \n",
       "CD4-positive, alpha-beta T cell                         \n",
       "CD8-positive, alpha-beta T cell                         \n",
       "Schwann cell                                         1  \n",
       "basal cell                                              \n",
       "endothelial cell                                        \n",
       "fibroblast                                              \n",
       "macrophage                                              \n",
       "mast cell                                               \n",
       "monocyte                                                \n",
       "mural cell                                              \n",
       "myoepithelial cell                                      \n",
       "regulatory T cell                                       \n",
       "salivary gland cell                                     \n",
       "stratified squamous epithelial cell                     \n",
       "tongue muscle cell                                   2  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.crosstab(\n",
    "    batch_cmp_df[\"true cell type\"],\n",
    "    batch_cmp_df[\"predicted cell type\"],\n",
    ").replace(0, '')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fb07f10c-91bb-46b2-b957-3bcc064151ae",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
