

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8">
  <meta name="generator" content="Docutils 0.19: https://docutils.sourceforge.io/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Training a PyTorch Model &mdash; cellxgene-census  documentation</title>
  

  
  <link rel="icon" type="image/png" sizes="32x32" href="../../_static/img/favicon_32x32_v2.png"/>
  <link rel="icon" type="image/png" sizes="16x16" href="../../_static/img/favicon_16x16_v2.png"/>
  <link rel="preconnect" href="https://fonts.googleapis.com">
  <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
  <link href="https://fonts.googleapis.com/css2?family=Inter:wght@400;500;600;700&display=swap" rel="stylesheet">
  <link rel="stylesheet" href="../../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/css/custom.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/nbsphinx-code-cells.css" type="text/css" />

  
  
  
  

  
  <!--[if lt IE 9]>
    <script src="../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../../" src="../../_static/documentation_options.js"></script>
        <script src="../../_static/jquery.js"></script>
        <script src="../../_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script data-url_root="../../" id="documentation_options" src="../../_static/documentation_options.js"></script>
        <script src="../../_static/doctools.js"></script>
        <script src="../../_static/sphinx_highlight.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
        <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
        <script data-domain="chanzuckerberg.github.io/cellxgene-census" defer="defer" src="https://plausible.io/js/script.js"></script>
    
    <script type="text/javascript" src="../../_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="Train an scVI model using Census data" href="pytorch_scvi.html" />
    <link rel="prev" title="Computing on X using online (incremental) algorithms" href="../api_demo/census_compute_over_X.html" /> 
</head>

<body class="wy-body-for-nav">

  <div style="height: 100vh; overflow: hidden;">
     
    <!-- top 56px for navbar height -->
    <div class="navbar-cxg">
        <div style="padding: 0 16px; display: flex; height: inherit; align-items: center; justify-content: space-between;">
            <span style="display:flex; gap: 32px; align-items: center;">
                <a href="https://cellxgene.cziscience.com/">
                  <img src="../../_static/img/cellxgene-discover-logo.svg" />
                </a>
                <span class="navbar-cxg-nav-wrapper">
                  <span class="navbar-cxg-section">
                    <span class="navbar-cxg-nav-section-title">Application</span>
                    <span class="navbar-cxg-nav-item-container">
                      <span class="navbar-cxg-link">
                          <a class="navbar-cxg-anchor" href="https://cellxgene.cziscience.com/collections">Collections</a>
                      </span>
                      <span class="navbar-cxg-link">
                          <a class="navbar-cxg-anchor" href="https://cellxgene.cziscience.com/datasets">Datasets</a>
                      </span>
                      <span class="navbar-cxg-link">
                          <a class="navbar-cxg-anchor" href="https://cellxgene.cziscience.com/gene-expression">Gene Expression</a>
                      </span>
                      <span class="navbar-cxg-link">
                        <a class="navbar-cxg-anchor" href="https://cellxgene.cziscience.com/cellguide">Cell Guide</a>
                      </span>
                      <span class="navbar-cxg-link">
                        <a class="navbar-cxg-anchor" href="https://cellxgene.cziscience.com/differential-expression">Differential Expression</a>
                        <div style="height: 16px!important; display: flex;">
                          <span class="beta">NEW</span>
                        </div>
                      </span>
                    </span>
                  </span>
                  <hr class="navbar-divider"/>
                  <span class="navbar-cxg-section">
                    <span class="navbar-cxg-nav-section-title">Census</span>
                    <span class="navbar-cxg-nav-item-container">
                      <span class="navbar-cxg-link active-link">
                        <a class="navbar-cxg-anchor" href="/cellxgene-census/index.html">API</a>
                      </span>
                      <span class="navbar-cxg-link">
                        <a class="navbar-cxg-anchor" href="https://cellxgene.cziscience.com/census-models">Models</a>
                      </span>
                    </span>
                  </span>
                </span>
            </span>

            <span class="navbar-cxg-link">
              <a class="navbar-cxg-anchor" href="https://cellxgene.cziscience.com/docs">Help & Documentation</a>
            </span>
        </div>
    </div>
    <div style="width: 100%; height: 100%; overflow: auto; padding-top: 56px;">
        
        <!-- top 56px for navbar height -->
        <nav data-toggle="wy-nav-shift" class="wy-nav-side" style="top: 56px;" >
        <div class="wy-side-scroll">
            <div class="wy-side-nav-search" >
            

            
                <a href="../../index.html" class="icon icon-home"> cellxgene-census
            

            
            </a>

            

            
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

            
            </div>

            
            <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
            
                
                
                
                
                
                <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../../cellxgene_census_docsite_installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../cellxgene_census_docsite_quick_start.html">Quick start</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../articles.html">What‚Äôs new?</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../cellxgene_census_aws_open_data.html">Census in AWS ‚òÅÔ∏è</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../cellxgene_census_docsite_schema.html">Census data and schema</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../cellxgene_census_docsite_data_release_info.html">Census data releases</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../python-api.html">Python API</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../../examples.html">Python tutorials</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="../../examples.html#exporting-data">Exporting data</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../examples.html#new-accessing-spatial-transcriptomics-data-with-census">[NEW! üöÄ] Accessing spatial transcriptomics data with Census</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../examples.html#using-integrated-embeddings-and-models">Using integrated embeddings and models</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../examples.html#understanding-census-data">Understanding Census data</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../examples.html#analyzing-census-data">Analyzing Census data</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../examples.html#scalable-computing">Scalable computing</a></li>
<li class="toctree-l2 current"><a class="reference internal" href="../../examples.html#scalable-machine-learning">Scalable machine learning</a><ul class="current">
<li class="toctree-l3 current"><a class="current reference internal" href="#">Training a PyTorch Model</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#Open-the-Census">Open the Census</a></li>
<li class="toctree-l4"><a class="reference internal" href="#Create-an-ExperimentDataset">Create an ExperimentDataset</a></li>
<li class="toctree-l4"><a class="reference internal" href="#Split-the-dataset">Split the dataset</a></li>
<li class="toctree-l4"><a class="reference internal" href="#Create-the-DataLoader">Create the DataLoader</a></li>
<li class="toctree-l4"><a class="reference internal" href="#Define-the-model">Define the model</a></li>
<li class="toctree-l4"><a class="reference internal" href="#Train-the-model">Train the model</a></li>
<li class="toctree-l4"><a class="reference internal" href="#Make-predictions-with-the-model">Make predictions with the model</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="pytorch_scvi.html">Train an scVI model using Census data</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference external" href="https://chanzuckerberg.github.io/cellxgene-census/r/index.html">R API &amp; tutorials</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../cellxgene_census_docsite_FAQ.html">FAQ</a></li>
<li class="toctree-l1"><a class="reference external" href="https://github.com/chanzuckerberg/cellxgene-census/releases">Release notes</a></li>
</ul>

                
            
            </div>
            
        </div>
        </nav>

        <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

        
        <nav class="wy-nav-top" aria-label="top navigation">
            
            <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
            <a href="../../index.html">cellxgene-census</a>
            
        </nav>


        <div class="wy-nav-content">
            
            <div class="rst-content">
            
            <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../../examples.html">Python tutorials</a></li>
      <li class="breadcrumb-item active">Training a PyTorch Model</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../../_sources/notebooks/experimental/pytorch.ipynb.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
            <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
            <div itemprop="articleBody">
                
  <section id="Training-a-PyTorch-Model">
<h1>Training a PyTorch Model<a class="headerlink" href="#Training-a-PyTorch-Model" title="Permalink to this heading">ÔÉÅ</a></h1>
<p>This tutorial demonstrates training a simple Logistic Regression model using Census data, PyTorch, and <a class="reference external" href="https://github.com/single-cell-data/TileDB-SOMA-ML">TileDB-SOMA-ML</a>, which provides a PyTorch loader for SOMA datasets like Census. <em>Note: an earlier version of this notebook used a prototype loader in the</em> <code class="docutils literal notranslate"><span class="pre">cellxgene_census.experimental</span></code> <em>API, now replaced by TileDB-SOMA-ML.</em></p>
<p>We assume basic familiarity with PyTorch and the Census API. See the <a class="reference external" href="https://chanzuckerberg.github.io/cellxgene-census/notebooks/api_demo/census_query_extract.html">Querying and fetching the single-cell data and cell/gene metadata</a> notebook tutorial for a quick primer on Census API usage.</p>
<p><strong>Contents</strong></p>
<ul class="simple">
<li><p>Create an ExperimentDataset</p></li>
<li><p>Split the dataset</p></li>
<li><p>Create the DataLoader</p></li>
<li><p>Define the model</p></li>
<li><p>Train the model</p></li>
<li><p>Make predictions with the model</p></li>
</ul>
<section id="Open-the-Census">
<h2>Open the Census<a class="headerlink" href="#Open-the-Census" title="Permalink to this heading">ÔÉÅ</a></h2>
<p>First, obtain a handle to the Census data, in the usual manner:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">cellxgene_census</span>

<span class="n">census</span> <span class="o">=</span> <span class="n">cellxgene_census</span><span class="o">.</span><span class="n">open_soma</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
The &#34;stable&#34; release is currently 2025-01-30. Specify &#39;census_version=&#34;2025-01-30&#34;&#39; in future calls to open_soma() to ensure data consistency.
</pre></div></div>
</div>
</section>
<section id="Create-an-ExperimentDataset">
<h2>Create an ExperimentDataset<a class="headerlink" href="#Create-an-ExperimentDataset" title="Permalink to this heading">ÔÉÅ</a></h2>
<p>To set up a PyTorch dataset from a Census slice, open an <a class="reference external" href="https://tiledbsoma.readthedocs.io/en/stable/python-tiledbsoma-experimentaxisquery.html">ExperimentAxisQuery</a> and create a TileDB-SOMA-ML <a class="reference external" href="https://single-cell-data.github.io/TileDB-SOMA-ML/#tiledbsoma_ml.ExperimentDataset">ExperimentDataset</a> for it. We will also prepare a scikit-learn <a class="reference external" href="https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.LabelEncoder.html">LabelEncoder</a> to help with mapping between Census cell
types (categorical metadata) and PyTorch tensors.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">tiledbsoma</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">soma</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">sklearn.preprocessing</span><span class="w"> </span><span class="kn">import</span> <span class="n">LabelEncoder</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">tiledbsoma_ml</span><span class="w"> </span><span class="kn">import</span> <span class="n">ExperimentDataset</span>

<span class="n">experiment</span> <span class="o">=</span> <span class="n">census</span><span class="p">[</span><span class="s2">&quot;census_data&quot;</span><span class="p">][</span><span class="s2">&quot;homo_sapiens&quot;</span><span class="p">]</span>
<span class="k">with</span> <span class="n">experiment</span><span class="o">.</span><span class="n">axis_query</span><span class="p">(</span>
    <span class="n">measurement_name</span><span class="o">=</span><span class="s2">&quot;RNA&quot;</span><span class="p">,</span>
    <span class="n">obs_query</span><span class="o">=</span><span class="n">soma</span><span class="o">.</span><span class="n">AxisQuery</span><span class="p">(</span><span class="n">value_filter</span><span class="o">=</span><span class="s2">&quot;tissue_general == &#39;tongue&#39; and is_primary_data == True&quot;</span><span class="p">),</span>
<span class="p">)</span> <span class="k">as</span> <span class="n">query</span><span class="p">:</span>
    <span class="n">experiment_dataset</span> <span class="o">=</span> <span class="n">ExperimentDataset</span><span class="p">(</span>
        <span class="n">query</span><span class="p">,</span>
        <span class="n">layer_name</span><span class="o">=</span><span class="s2">&quot;raw&quot;</span><span class="p">,</span>
        <span class="n">obs_column_names</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;cell_type&quot;</span><span class="p">],</span>
        <span class="n">batch_size</span><span class="o">=</span><span class="mi">128</span><span class="p">,</span>
        <span class="n">shuffle</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
        <span class="n">seed</span><span class="o">=</span><span class="mi">111</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">obs_df</span> <span class="o">=</span> <span class="n">query</span><span class="o">.</span><span class="n">obs</span><span class="p">(</span><span class="n">column_names</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;cell_type&quot;</span><span class="p">])</span><span class="o">.</span><span class="n">concat</span><span class="p">()</span><span class="o">.</span><span class="n">to_pandas</span><span class="p">()</span>
    <span class="n">cell_type_encoder</span> <span class="o">=</span> <span class="n">LabelEncoder</span><span class="p">()</span><span class="o">.</span><span class="n">fit</span><span class="p">(</span><span class="n">obs_df</span><span class="p">[</span><span class="s2">&quot;cell_type&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">())</span>
</pre></div>
</div>
</div>
<section id="ExperimentDataset-class-explained">
<h3><a class="reference external" href="https://single-cell-data.github.io/TileDB-SOMA-ML/#tiledbsoma_ml.ExperimentDataset">ExperimentDataset</a> class explained<a class="headerlink" href="#ExperimentDataset-class-explained" title="Permalink to this heading">ÔÉÅ</a></h3>
<p>This class provides an implementation of PyTorch‚Äôs <code class="docutils literal notranslate"><span class="pre">IterableDataset</span></code> interface for accessing data from abstract sources. It encapsulates streaming the result set of a SOMA <a class="reference external" href="https://tiledbsoma.readthedocs.io/en/stable/python-tiledbsoma-experimentaxisquery.html">ExperimentAxisQuery</a> in a series of ‚Äúbatches,‚Äù each consisting of a NumPy <code class="docutils literal notranslate"><span class="pre">ndarray</span></code> (batch of X results) and a Pandas <code class="docutils literal notranslate"><span class="pre">DataFrame</span></code> (corresponding batch of obs results). Most importantly, it avoids loading large result sets into
memory all at once.</p>
<p>The constructor requires the <a class="reference external" href="https://tiledbsoma.readthedocs.io/en/stable/python-tiledbsoma-experimentaxisquery.html">ExperimentAxisQuery</a> defining the desired slice of Census data, and the name of the X layer to access.</p>
<p><code class="docutils literal notranslate"><span class="pre">obs_column_names</span></code> sets the columns to be included in the DataFrame for each batch, to be used as data labels or model inputs.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">batch_size</span></code> parameter specifies the target number of rows (cells) in each batch.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">shuffle</span></code> flag supports randomizing the ordering of the training data for each training epoch (default: <code class="docutils literal notranslate"><span class="pre">True</span></code>). This should be used instead of <a class="reference external" href="https://pytorch.org/docs/stable/data.html">DataLoader</a>‚Äôs <code class="docutils literal notranslate"><span class="pre">shuffle</span></code> flag, as the implementation is more efficient.</p>
<p>You can inspect the shape of the full dataset, without causing the full dataset to be loaded. The <code class="docutils literal notranslate"><span class="pre">shape</span></code> property returns the number of batches on the first dimension:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">experiment_dataset</span><span class="o">.</span><span class="n">shape</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
(303, 61888)
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">experiment_dataset</span><span class="o">.</span><span class="n">query_ids</span><span class="o">.</span><span class="n">obs_joinids</span><span class="o">.</span><span class="n">shape</span>  <span class="c1"># total result count</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
(38754,)
</pre></div></div>
</div>
</section>
</section>
<section id="Split-the-dataset">
<h2>Split the dataset<a class="headerlink" href="#Split-the-dataset" title="Permalink to this heading">ÔÉÅ</a></h2>
<p>You may split the query results into the typical training and test sets using the <a class="reference external" href="https://single-cell-data.github.io/TileDB-SOMA-ML/#tiledbsoma_ml.ExperimentDataset.random_split">ExperimentDataset.random_split()</a> method:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">train_dataset</span><span class="p">,</span> <span class="n">test_dataset</span> <span class="o">=</span> <span class="n">experiment_dataset</span><span class="o">.</span><span class="n">random_split</span><span class="p">(</span>
    <span class="mf">0.8</span><span class="p">,</span>
    <span class="mf">0.2</span><span class="p">,</span>
    <span class="n">seed</span><span class="o">=</span><span class="mi">111</span><span class="p">,</span>
<span class="p">)</span>
<span class="n">train_dataset</span><span class="o">.</span><span class="n">shape</span><span class="p">,</span> <span class="n">test_dataset</span><span class="o">.</span><span class="n">shape</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
((243, 61888), (61, 61888))
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">train_dataset</span><span class="o">.</span><span class="n">query_ids</span><span class="o">.</span><span class="n">obs_joinids</span><span class="o">.</span><span class="n">shape</span><span class="p">,</span> <span class="n">test_dataset</span><span class="o">.</span><span class="n">query_ids</span><span class="o">.</span><span class="n">obs_joinids</span><span class="o">.</span><span class="n">shape</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
((31003,), (7751,))
</pre></div></div>
</div>
</section>
<section id="Create-the-DataLoader">
<h2>Create the DataLoader<a class="headerlink" href="#Create-the-DataLoader" title="Permalink to this heading">ÔÉÅ</a></h2>
<p>Now you can prepare a PyTorch <a class="reference external" href="https://pytorch.org/docs/stable/data.html#torch.utils.data.DataLoader">DataLoader</a> on the training data. While not shown here, the DataLoader can be configured with multiple worker processes to stream the result batches as quickly as possible.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">tiledbsoma_ml</span><span class="w"> </span><span class="kn">import</span> <span class="n">experiment_dataloader</span>

<span class="n">train_dataloader</span> <span class="o">=</span> <span class="n">experiment_dataloader</span><span class="p">(</span><span class="n">train_dataset</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>(Instantiating a PyTorch DataLoader object directly is not recommended, as <a class="reference external" href="https://single-cell-data.github.io/TileDB-SOMA-ML/#tiledbsoma_ml.experiment_dataloader">experiment_dataloader()</a> enforces correct and performant usage.)</p>
</section>
<section id="Define-the-model">
<h2>Define the model<a class="headerlink" href="#Define-the-model" title="Permalink to this heading">ÔÉÅ</a></h2>
<p>With the training data retrieval code now in place, we can move on to defining a simple logistic regression model, using PyTorch‚Äôs <code class="docutils literal notranslate"><span class="pre">torch.nn.Linear</span></code> class:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">torch</span>


<span class="k">class</span><span class="w"> </span><span class="nc">LogisticRegression</span><span class="p">(</span><span class="n">torch</span><span class="o">.</span><span class="n">nn</span><span class="o">.</span><span class="n">Module</span><span class="p">):</span>
    <span class="k">def</span><span class="w"> </span><span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">input_dim</span><span class="p">,</span> <span class="n">output_dim</span><span class="p">):</span>
        <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="fm">__init__</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">linear</span> <span class="o">=</span> <span class="n">torch</span><span class="o">.</span><span class="n">nn</span><span class="o">.</span><span class="n">Linear</span><span class="p">(</span><span class="n">input_dim</span><span class="p">,</span> <span class="n">output_dim</span><span class="p">)</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">forward</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">x</span><span class="p">):</span>
        <span class="n">outputs</span> <span class="o">=</span> <span class="n">torch</span><span class="o">.</span><span class="n">sigmoid</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">linear</span><span class="p">(</span><span class="n">x</span><span class="p">))</span>
        <span class="k">return</span> <span class="n">outputs</span>
</pre></div>
</div>
</div>
<p>Next, we define a function to train the model for a single epoch:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span><span class="w"> </span><span class="nf">train_epoch</span><span class="p">(</span><span class="n">model</span><span class="p">,</span> <span class="n">train_dataloader</span><span class="p">,</span> <span class="n">loss_fn</span><span class="p">,</span> <span class="n">optimizer</span><span class="p">,</span> <span class="n">device</span><span class="p">):</span>
    <span class="n">model</span><span class="o">.</span><span class="n">train</span><span class="p">()</span>
    <span class="n">train_loss</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">train_correct</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">train_total</span> <span class="o">=</span> <span class="mi">0</span>

    <span class="k">for</span> <span class="n">X_batch</span><span class="p">,</span> <span class="n">obs_batch</span> <span class="ow">in</span> <span class="n">train_dataloader</span><span class="p">:</span>
        <span class="n">optimizer</span><span class="o">.</span><span class="n">zero_grad</span><span class="p">()</span>

        <span class="c1"># convert X_batch numpy ndarray into PyTorch tensor</span>
        <span class="n">X_batch</span> <span class="o">=</span> <span class="n">torch</span><span class="o">.</span><span class="n">from_numpy</span><span class="p">(</span><span class="n">X_batch</span><span class="p">)</span><span class="o">.</span><span class="n">float</span><span class="p">()</span><span class="o">.</span><span class="n">to</span><span class="p">(</span><span class="n">device</span><span class="p">)</span>

        <span class="c1"># Perform prediction</span>
        <span class="n">outputs</span> <span class="o">=</span> <span class="n">model</span><span class="p">(</span><span class="n">X_batch</span><span class="p">)</span>

        <span class="c1"># Determine the predicted label</span>
        <span class="n">probabilities</span> <span class="o">=</span> <span class="n">torch</span><span class="o">.</span><span class="n">nn</span><span class="o">.</span><span class="n">functional</span><span class="o">.</span><span class="n">softmax</span><span class="p">(</span><span class="n">outputs</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
        <span class="n">predictions</span> <span class="o">=</span> <span class="n">torch</span><span class="o">.</span><span class="n">argmax</span><span class="p">(</span><span class="n">probabilities</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>

        <span class="c1"># Compute the loss using encoded labels and perform back propagation</span>
        <span class="n">label_batch</span> <span class="o">=</span> <span class="n">torch</span><span class="o">.</span><span class="n">from_numpy</span><span class="p">(</span><span class="n">cell_type_encoder</span><span class="o">.</span><span class="n">transform</span><span class="p">(</span><span class="n">obs_batch</span><span class="p">[</span><span class="s2">&quot;cell_type&quot;</span><span class="p">]))</span><span class="o">.</span><span class="n">to</span><span class="p">(</span><span class="n">device</span><span class="p">)</span>
        <span class="n">train_correct</span> <span class="o">+=</span> <span class="p">(</span><span class="n">predictions</span> <span class="o">==</span> <span class="n">label_batch</span><span class="p">)</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span><span class="o">.</span><span class="n">item</span><span class="p">()</span>
        <span class="n">train_total</span> <span class="o">+=</span> <span class="nb">len</span><span class="p">(</span><span class="n">predictions</span><span class="p">)</span>

        <span class="n">loss</span> <span class="o">=</span> <span class="n">loss_fn</span><span class="p">(</span><span class="n">outputs</span><span class="p">,</span> <span class="n">label_batch</span><span class="o">.</span><span class="n">long</span><span class="p">())</span>
        <span class="n">train_loss</span> <span class="o">+=</span> <span class="n">loss</span><span class="o">.</span><span class="n">item</span><span class="p">()</span>
        <span class="n">loss</span><span class="o">.</span><span class="n">backward</span><span class="p">()</span>
        <span class="n">optimizer</span><span class="o">.</span><span class="n">step</span><span class="p">()</span>

    <span class="n">train_loss</span> <span class="o">/=</span> <span class="n">train_total</span>
    <span class="n">train_accuracy</span> <span class="o">=</span> <span class="n">train_correct</span> <span class="o">/</span> <span class="n">train_total</span>
    <span class="k">return</span> <span class="n">train_loss</span><span class="p">,</span> <span class="n">train_accuracy</span>
</pre></div>
</div>
</div>
<p>The <code class="docutils literal notranslate"><span class="pre">X_batch</span></code> ndarray (later tensor) has the slice of the RNA X matrix for the current batch of results. Similarly, the <code class="docutils literal notranslate"><span class="pre">obs_batch</span></code> dataframe has the slice of <code class="docutils literal notranslate"><span class="pre">obs</span></code> with <code class="docutils literal notranslate"><span class="pre">cell_type</span></code> metadata. We use <code class="docutils literal notranslate"><span class="pre">cell_type_encoder</span></code> to encode these categorical cell types into integer-valued tensors for comparison with the model predictions.</p>
</section>
<section id="Train-the-model">
<h2>Train the model<a class="headerlink" href="#Train-the-model" title="Permalink to this heading">ÔÉÅ</a></h2>
<p>Finally, we are ready to train the model. Here we instantiate the model, a loss function, and an optimization method, then iterate through the desired number of training epochs. Note how the <code class="docutils literal notranslate"><span class="pre">train_dataloader</span></code> is passed into <code class="docutils literal notranslate"><span class="pre">train_epoch</span></code>, where for each epoch it will provide a new iterator through the training dataset.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">device</span> <span class="o">=</span> <span class="n">torch</span><span class="o">.</span><span class="n">device</span><span class="p">(</span><span class="s2">&quot;cuda&quot;</span><span class="p">)</span> <span class="k">if</span> <span class="n">torch</span><span class="o">.</span><span class="n">cuda</span><span class="o">.</span><span class="n">is_available</span><span class="p">()</span> <span class="k">else</span> <span class="n">torch</span><span class="o">.</span><span class="n">device</span><span class="p">(</span><span class="s2">&quot;cpu&quot;</span><span class="p">)</span>

<span class="c1"># The size of the input dimension is the number of genes</span>
<span class="n">input_dim</span> <span class="o">=</span> <span class="n">experiment_dataset</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>

<span class="c1"># The size of the output dimension is the number of distinct cell_type values</span>
<span class="n">output_dim</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">cell_type_encoder</span><span class="o">.</span><span class="n">classes_</span><span class="p">)</span>

<span class="n">model</span> <span class="o">=</span> <span class="n">LogisticRegression</span><span class="p">(</span><span class="n">input_dim</span><span class="p">,</span> <span class="n">output_dim</span><span class="p">)</span><span class="o">.</span><span class="n">to</span><span class="p">(</span><span class="n">device</span><span class="p">)</span>
<span class="n">loss_fn</span> <span class="o">=</span> <span class="n">torch</span><span class="o">.</span><span class="n">nn</span><span class="o">.</span><span class="n">CrossEntropyLoss</span><span class="p">()</span>
<span class="n">optimizer</span> <span class="o">=</span> <span class="n">torch</span><span class="o">.</span><span class="n">optim</span><span class="o">.</span><span class="n">Adam</span><span class="p">(</span><span class="n">model</span><span class="o">.</span><span class="n">parameters</span><span class="p">(),</span> <span class="n">lr</span><span class="o">=</span><span class="mf">1e-5</span><span class="p">)</span>

<span class="k">for</span> <span class="n">epoch</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">25</span><span class="p">):</span>
    <span class="n">train_loss</span><span class="p">,</span> <span class="n">train_accuracy</span> <span class="o">=</span> <span class="n">train_epoch</span><span class="p">(</span><span class="n">model</span><span class="p">,</span> <span class="n">train_dataloader</span><span class="p">,</span> <span class="n">loss_fn</span><span class="p">,</span> <span class="n">optimizer</span><span class="p">,</span> <span class="n">device</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Epoch </span><span class="si">{</span><span class="n">epoch</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="mi">1</span><span class="si">}</span><span class="s2">: Train Loss: </span><span class="si">{</span><span class="n">train_loss</span><span class="si">:</span><span class="s2">.7f</span><span class="si">}</span><span class="s2"> Accuracy </span><span class="si">{</span><span class="n">train_accuracy</span><span class="si">:</span><span class="s2">.4f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Epoch 1: Train Loss: 0.0196637 Accuracy 0.4806
Epoch 2: Train Loss: 0.0186214 Accuracy 0.7383
Epoch 3: Train Loss: 0.0183390 Accuracy 0.8257
Epoch 4: Train Loss: 0.0181079 Accuracy 0.8503
Epoch 5: Train Loss: 0.0180089 Accuracy 0.8603
Epoch 6: Train Loss: 0.0179505 Accuracy 0.8641
Epoch 7: Train Loss: 0.0179011 Accuracy 0.8678
Epoch 8: Train Loss: 0.0178689 Accuracy 0.8694
Epoch 9: Train Loss: 0.0178441 Accuracy 0.8718
Epoch 10: Train Loss: 0.0178233 Accuracy 0.8728
Epoch 11: Train Loss: 0.0178058 Accuracy 0.8741
Epoch 12: Train Loss: 0.0177893 Accuracy 0.8755
Epoch 13: Train Loss: 0.0177776 Accuracy 0.8737
Epoch 14: Train Loss: 0.0177634 Accuracy 0.8771
Epoch 15: Train Loss: 0.0177532 Accuracy 0.8783
Epoch 16: Train Loss: 0.0177437 Accuracy 0.8793
Epoch 17: Train Loss: 0.0177350 Accuracy 0.8802
Epoch 18: Train Loss: 0.0177275 Accuracy 0.8811
Epoch 19: Train Loss: 0.0177217 Accuracy 0.8816
Epoch 20: Train Loss: 0.0177154 Accuracy 0.8823
Epoch 21: Train Loss: 0.0177091 Accuracy 0.8838
Epoch 22: Train Loss: 0.0177030 Accuracy 0.8840
Epoch 23: Train Loss: 0.0176972 Accuracy 0.8847
Epoch 24: Train Loss: 0.0176921 Accuracy 0.8851
Epoch 25: Train Loss: 0.0176756 Accuracy 0.8856
</pre></div></div>
</div>
</section>
<section id="Make-predictions-with-the-model">
<h2>Make predictions with the model<a class="headerlink" href="#Make-predictions-with-the-model" title="Permalink to this heading">ÔÉÅ</a></h2>
<p>To make predictions with the model, we first create a new <code class="docutils literal notranslate"><span class="pre">DataLoader</span></code> using the <code class="docutils literal notranslate"><span class="pre">test_dataset</span></code>, which provides the ‚Äútest‚Äù split of the original dataset. For this example, we will only make predictions on a single batch of data from the test split.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">test_dataloader</span> <span class="o">=</span> <span class="n">experiment_dataloader</span><span class="p">(</span><span class="n">test_dataset</span><span class="p">)</span>
<span class="n">X_batch</span><span class="p">,</span> <span class="n">obs_batch</span> <span class="o">=</span> <span class="nb">next</span><span class="p">(</span><span class="nb">iter</span><span class="p">(</span><span class="n">test_dataloader</span><span class="p">))</span>
<span class="n">X_batch</span> <span class="o">=</span> <span class="n">torch</span><span class="o">.</span><span class="n">from_numpy</span><span class="p">(</span><span class="n">X_batch</span><span class="p">)</span>
<span class="n">true_cell_types</span> <span class="o">=</span> <span class="n">torch</span><span class="o">.</span><span class="n">from_numpy</span><span class="p">(</span><span class="n">cell_type_encoder</span><span class="o">.</span><span class="n">transform</span><span class="p">(</span><span class="n">obs_batch</span><span class="p">[</span><span class="s2">&quot;cell_type&quot;</span><span class="p">]))</span>
</pre></div>
</div>
</div>
<p>Next, we invoke the model on the <code class="docutils literal notranslate"><span class="pre">X_batch</span></code> input data and extract the predictions:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">model</span><span class="o">.</span><span class="n">eval</span><span class="p">()</span>

<span class="n">model</span><span class="o">.</span><span class="n">to</span><span class="p">(</span><span class="n">device</span><span class="p">)</span>
<span class="n">outputs</span> <span class="o">=</span> <span class="n">model</span><span class="p">(</span><span class="n">X_batch</span><span class="o">.</span><span class="n">to</span><span class="p">(</span><span class="n">device</span><span class="p">))</span>

<span class="n">probabilities</span> <span class="o">=</span> <span class="n">torch</span><span class="o">.</span><span class="n">nn</span><span class="o">.</span><span class="n">functional</span><span class="o">.</span><span class="n">softmax</span><span class="p">(</span><span class="n">outputs</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
<span class="n">predictions</span> <span class="o">=</span> <span class="n">torch</span><span class="o">.</span><span class="n">argmax</span><span class="p">(</span><span class="n">probabilities</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="n">predictions</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
tensor([ 5,  8,  6, 20, 20, 20, 20,  5,  5, 20,  5,  8, 19,  9, 22,  5,  6,  5,
        20,  5,  5,  9,  5, 20, 20,  1,  8,  5,  1,  0,  8,  5, 20,  5, 20,  5,
         8, 20, 20,  5, 20,  1, 20, 20,  5, 20, 10,  9,  5,  5, 20,  5,  6,  9,
         2,  5, 20,  8, 20,  5,  5,  5,  5,  5, 20, 20,  5,  1,  5,  5,  9, 20,
         5, 20, 20,  5,  5,  5,  6, 12,  5,  5, 20,  1,  5,  5,  5, 20, 20, 20,
        20,  5, 20, 22, 20,  5,  5,  5,  5,  5, 20,  2,  1, 19,  5,  5,  5,  5,
        20,  5,  5, 20,  0, 22,  5,  1, 20,  2, 20, 20, 19,  5,  5,  5,  5,  5,
        16, 19], device=&#39;cuda:0&#39;)
</pre></div></div>
</div>
<p>The predictions are returned as the encoded values of <code class="docutils literal notranslate"><span class="pre">cell_type</span></code> label. To recover the original cell type labels as strings, we decode using the same <code class="docutils literal notranslate"><span class="pre">LabelEncoder</span></code> used for training.</p>
<p>At inference time, if the model inputs are not obtained via an <code class="docutils literal notranslate"><span class="pre">ExperimentDataset</span></code>, one could pickle the encoder at training time and save it along with the model. Then, at inference time it can be unpickled and used as shown below.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[13]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">predicted_cell_types</span> <span class="o">=</span> <span class="n">cell_type_encoder</span><span class="o">.</span><span class="n">inverse_transform</span><span class="p">(</span><span class="n">predictions</span><span class="o">.</span><span class="n">cpu</span><span class="p">())</span>
<span class="n">predicted_cell_types</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[13]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
array([&#39;basal cell&#39;, &#39;fibroblast&#39;, &#39;endothelial cell&#39;,
       &#39;stratified squamous epithelial cell&#39;,
       &#39;stratified squamous epithelial cell&#39;,
       &#39;stratified squamous epithelial cell&#39;,
       &#39;stratified squamous epithelial cell&#39;, &#39;basal cell&#39;, &#39;basal cell&#39;,
       &#39;stratified squamous epithelial cell&#39;, &#39;basal cell&#39;, &#39;fibroblast&#39;,
       &#39;salivary gland cell&#39;, &#39;macrophage&#39;, &#39;tongue muscle cell&#39;,
       &#39;basal cell&#39;, &#39;endothelial cell&#39;, &#39;basal cell&#39;,
       &#39;stratified squamous epithelial cell&#39;, &#39;basal cell&#39;, &#39;basal cell&#39;,
       &#39;macrophage&#39;, &#39;basal cell&#39;, &#39;stratified squamous epithelial cell&#39;,
       &#39;stratified squamous epithelial cell&#39;,
       &#39;CD4-positive, alpha-beta T cell&#39;, &#39;fibroblast&#39;, &#39;basal cell&#39;,
       &#39;CD4-positive, alpha-beta T cell&#39;, &#39;B cell&#39;, &#39;fibroblast&#39;,
       &#39;basal cell&#39;, &#39;stratified squamous epithelial cell&#39;, &#39;basal cell&#39;,
       &#39;stratified squamous epithelial cell&#39;, &#39;basal cell&#39;, &#39;fibroblast&#39;,
       &#39;stratified squamous epithelial cell&#39;,
       &#39;stratified squamous epithelial cell&#39;, &#39;basal cell&#39;,
       &#39;stratified squamous epithelial cell&#39;,
       &#39;CD4-positive, alpha-beta T cell&#39;,
       &#39;stratified squamous epithelial cell&#39;,
       &#39;stratified squamous epithelial cell&#39;, &#39;basal cell&#39;,
       &#39;stratified squamous epithelial cell&#39;, &#39;mast cell&#39;, &#39;macrophage&#39;,
       &#39;basal cell&#39;, &#39;basal cell&#39;, &#39;stratified squamous epithelial cell&#39;,
       &#39;basal cell&#39;, &#39;endothelial cell&#39;, &#39;macrophage&#39;,
       &#39;CD8-positive, alpha-beta T cell&#39;, &#39;basal cell&#39;,
       &#39;stratified squamous epithelial cell&#39;, &#39;fibroblast&#39;,
       &#39;stratified squamous epithelial cell&#39;, &#39;basal cell&#39;, &#39;basal cell&#39;,
       &#39;basal cell&#39;, &#39;basal cell&#39;, &#39;basal cell&#39;,
       &#39;stratified squamous epithelial cell&#39;,
       &#39;stratified squamous epithelial cell&#39;, &#39;basal cell&#39;,
       &#39;CD4-positive, alpha-beta T cell&#39;, &#39;basal cell&#39;, &#39;basal cell&#39;,
       &#39;macrophage&#39;, &#39;stratified squamous epithelial cell&#39;, &#39;basal cell&#39;,
       &#39;stratified squamous epithelial cell&#39;,
       &#39;stratified squamous epithelial cell&#39;, &#39;basal cell&#39;, &#39;basal cell&#39;,
       &#39;basal cell&#39;, &#39;endothelial cell&#39;, &#39;mural cell&#39;, &#39;basal cell&#39;,
       &#39;basal cell&#39;, &#39;stratified squamous epithelial cell&#39;,
       &#39;CD4-positive, alpha-beta T cell&#39;, &#39;basal cell&#39;, &#39;basal cell&#39;,
       &#39;basal cell&#39;, &#39;stratified squamous epithelial cell&#39;,
       &#39;stratified squamous epithelial cell&#39;,
       &#39;stratified squamous epithelial cell&#39;,
       &#39;stratified squamous epithelial cell&#39;, &#39;basal cell&#39;,
       &#39;stratified squamous epithelial cell&#39;, &#39;tongue muscle cell&#39;,
       &#39;stratified squamous epithelial cell&#39;, &#39;basal cell&#39;, &#39;basal cell&#39;,
       &#39;basal cell&#39;, &#39;basal cell&#39;, &#39;basal cell&#39;,
       &#39;stratified squamous epithelial cell&#39;,
       &#39;CD8-positive, alpha-beta T cell&#39;,
       &#39;CD4-positive, alpha-beta T cell&#39;, &#39;salivary gland cell&#39;,
       &#39;basal cell&#39;, &#39;basal cell&#39;, &#39;basal cell&#39;, &#39;basal cell&#39;,
       &#39;stratified squamous epithelial cell&#39;, &#39;basal cell&#39;, &#39;basal cell&#39;,
       &#39;stratified squamous epithelial cell&#39;, &#39;B cell&#39;,
       &#39;tongue muscle cell&#39;, &#39;basal cell&#39;,
       &#39;CD4-positive, alpha-beta T cell&#39;,
       &#39;stratified squamous epithelial cell&#39;,
       &#39;CD8-positive, alpha-beta T cell&#39;,
       &#39;stratified squamous epithelial cell&#39;,
       &#39;stratified squamous epithelial cell&#39;, &#39;salivary gland cell&#39;,
       &#39;basal cell&#39;, &#39;basal cell&#39;, &#39;basal cell&#39;, &#39;basal cell&#39;,
       &#39;basal cell&#39;, &#39;neutrophil&#39;, &#39;salivary gland cell&#39;], dtype=object)
</pre></div></div>
</div>
<p>Finally, we create a Pandas DataFrame to examine the predictions:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pd</span>

<span class="n">batch_cmp_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span>
    <span class="p">{</span>
        <span class="s2">&quot;true cell type&quot;</span><span class="p">:</span> <span class="n">cell_type_encoder</span><span class="o">.</span><span class="n">inverse_transform</span><span class="p">(</span><span class="n">true_cell_types</span><span class="o">.</span><span class="n">ravel</span><span class="p">()</span><span class="o">.</span><span class="n">numpy</span><span class="p">()),</span>
        <span class="s2">&quot;predicted cell type&quot;</span><span class="p">:</span> <span class="n">predicted_cell_types</span><span class="p">,</span>
    <span class="p">}</span>
<span class="p">)</span>
<span class="n">batch_cmp_df</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>true cell type</th>
      <th>predicted cell type</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>stratified squamous epithelial cell</td>
      <td>basal cell</td>
    </tr>
    <tr>
      <th>1</th>
      <td>fibroblast</td>
      <td>fibroblast</td>
    </tr>
    <tr>
      <th>2</th>
      <td>basal cell</td>
      <td>endothelial cell</td>
    </tr>
    <tr>
      <th>3</th>
      <td>stratified squamous epithelial cell</td>
      <td>stratified squamous epithelial cell</td>
    </tr>
    <tr>
      <th>4</th>
      <td>stratified squamous epithelial cell</td>
      <td>stratified squamous epithelial cell</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>123</th>
      <td>basal cell</td>
      <td>basal cell</td>
    </tr>
    <tr>
      <th>124</th>
      <td>basal cell</td>
      <td>basal cell</td>
    </tr>
    <tr>
      <th>125</th>
      <td>basal cell</td>
      <td>basal cell</td>
    </tr>
    <tr>
      <th>126</th>
      <td>basal cell</td>
      <td>neutrophil</td>
    </tr>
    <tr>
      <th>127</th>
      <td>salivary gland cell</td>
      <td>salivary gland cell</td>
    </tr>
  </tbody>
</table>
<p>128 rows √ó 2 columns</p>
</div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[15]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">pd</span><span class="o">.</span><span class="n">crosstab</span><span class="p">(</span>
    <span class="n">batch_cmp_df</span><span class="p">[</span><span class="s2">&quot;true cell type&quot;</span><span class="p">],</span>
    <span class="n">batch_cmp_df</span><span class="p">[</span><span class="s2">&quot;predicted cell type&quot;</span><span class="p">],</span>
<span class="p">)</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[15]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th>predicted cell type</th>
      <th>B cell</th>
      <th>CD4-positive, alpha-beta T cell</th>
      <th>CD8-positive, alpha-beta T cell</th>
      <th>basal cell</th>
      <th>endothelial cell</th>
      <th>fibroblast</th>
      <th>macrophage</th>
      <th>mast cell</th>
      <th>mural cell</th>
      <th>neutrophil</th>
      <th>salivary gland cell</th>
      <th>stratified squamous epithelial cell</th>
      <th>tongue muscle cell</th>
    </tr>
    <tr>
      <th>true cell type</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>B cell</th>
      <td>2</td>
      <td></td>
      <td></td>
      <td></td>
      <td></td>
      <td></td>
      <td></td>
      <td></td>
      <td></td>
      <td></td>
      <td></td>
      <td></td>
      <td></td>
    </tr>
    <tr>
      <th>CD4-positive, alpha-beta T cell</th>
      <td></td>
      <td>6</td>
      <td></td>
      <td></td>
      <td></td>
      <td></td>
      <td></td>
      <td></td>
      <td></td>
      <td></td>
      <td></td>
      <td></td>
      <td></td>
    </tr>
    <tr>
      <th>CD8-positive, alpha-beta T cell</th>
      <td></td>
      <td></td>
      <td>3</td>
      <td></td>
      <td></td>
      <td></td>
      <td></td>
      <td></td>
      <td></td>
      <td></td>
      <td></td>
      <td></td>
      <td></td>
    </tr>
    <tr>
      <th>Schwann cell</th>
      <td></td>
      <td></td>
      <td></td>
      <td></td>
      <td></td>
      <td></td>
      <td></td>
      <td></td>
      <td></td>
      <td></td>
      <td></td>
      <td></td>
      <td>1</td>
    </tr>
    <tr>
      <th>basal cell</th>
      <td></td>
      <td></td>
      <td></td>
      <td>46</td>
      <td>1</td>
      <td></td>
      <td></td>
      <td></td>
      <td></td>
      <td>1</td>
      <td></td>
      <td>3</td>
      <td></td>
    </tr>
    <tr>
      <th>endothelial cell</th>
      <td></td>
      <td></td>
      <td></td>
      <td></td>
      <td>3</td>
      <td></td>
      <td></td>
      <td></td>
      <td></td>
      <td></td>
      <td></td>
      <td></td>
      <td></td>
    </tr>
    <tr>
      <th>fibroblast</th>
      <td></td>
      <td></td>
      <td></td>
      <td></td>
      <td></td>
      <td>6</td>
      <td></td>
      <td></td>
      <td></td>
      <td></td>
      <td></td>
      <td></td>
      <td></td>
    </tr>
    <tr>
      <th>macrophage</th>
      <td></td>
      <td></td>
      <td></td>
      <td></td>
      <td></td>
      <td></td>
      <td>2</td>
      <td></td>
      <td></td>
      <td></td>
      <td></td>
      <td></td>
      <td></td>
    </tr>
    <tr>
      <th>mast cell</th>
      <td></td>
      <td></td>
      <td></td>
      <td></td>
      <td></td>
      <td></td>
      <td></td>
      <td>1</td>
      <td></td>
      <td></td>
      <td></td>
      <td></td>
      <td></td>
    </tr>
    <tr>
      <th>monocyte</th>
      <td></td>
      <td></td>
      <td></td>
      <td></td>
      <td></td>
      <td></td>
      <td>3</td>
      <td></td>
      <td></td>
      <td></td>
      <td></td>
      <td></td>
      <td></td>
    </tr>
    <tr>
      <th>mural cell</th>
      <td></td>
      <td></td>
      <td></td>
      <td></td>
      <td></td>
      <td></td>
      <td></td>
      <td></td>
      <td>1</td>
      <td></td>
      <td></td>
      <td></td>
      <td></td>
    </tr>
    <tr>
      <th>myoepithelial cell</th>
      <td></td>
      <td></td>
      <td></td>
      <td>1</td>
      <td></td>
      <td></td>
      <td></td>
      <td></td>
      <td></td>
      <td></td>
      <td></td>
      <td></td>
      <td></td>
    </tr>
    <tr>
      <th>regulatory T cell</th>
      <td></td>
      <td>1</td>
      <td></td>
      <td></td>
      <td></td>
      <td></td>
      <td></td>
      <td></td>
      <td></td>
      <td></td>
      <td></td>
      <td></td>
      <td></td>
    </tr>
    <tr>
      <th>salivary gland cell</th>
      <td></td>
      <td></td>
      <td></td>
      <td>2</td>
      <td></td>
      <td></td>
      <td></td>
      <td></td>
      <td></td>
      <td></td>
      <td>4</td>
      <td></td>
      <td></td>
    </tr>
    <tr>
      <th>stratified squamous epithelial cell</th>
      <td></td>
      <td></td>
      <td></td>
      <td>5</td>
      <td></td>
      <td></td>
      <td></td>
      <td></td>
      <td></td>
      <td></td>
      <td></td>
      <td>34</td>
      <td></td>
    </tr>
    <tr>
      <th>tongue muscle cell</th>
      <td></td>
      <td></td>
      <td></td>
      <td></td>
      <td></td>
      <td></td>
      <td></td>
      <td></td>
      <td></td>
      <td></td>
      <td></td>
      <td></td>
      <td>2</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>
</pre></div>
</div>
</div>
</section>
</section>


            </div>
            
            </div>
            <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="../api_demo/census_compute_over_X.html" class="btn btn-neutral float-left" title="Computing on X using online (incremental) algorithms" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="pytorch_scvi.html" class="btn btn-neutral float-right" title="Train an scVI model using Census data" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2022-Present, Chan Zuckerberg Initiative.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
            </div>
        </div>

        </section>

    </div>
    <!-- Newsletter Banner -->
    <div role="banner" id="newsletter-banner">
      <span id="newsletter-subscribe-button" role="button">Subscribe</span>&nbsp;to our newsletter to receive updates about new features.
      <div id="newsletter-banner-close-button" role="button">X</div>
    </div>
    <!-- Newsletter Modal -->
    <dialog id="newsletter-modal">
      <div id="newsletter-header">
        <img id="newsletter-logo" src="../../_static/img/cellxGene-newsletter-logo.svg" />
        <div id="newsletter-close-button" role="button">X</div>
      </div>
      <div id="newsletter-content">
        <div id="newsletter-callout">Join Our Newsletter</div>
        <div id="newsletter-description">Get a quarterly email with the latest CELLxGENE features and data.</div>
        <!-- HubSpot Form target -->
        <div id="newsletter-form-container"></div>
      </div>

      <div id="newsletter-footnote">Unsubscribe at any time.</div>
    </dialog>
    
  

    <script>
    // (thuang): 30 days
    const NEWSLETTER_BANNER_DISMISSED_TTL_MS = 30 * 24 * 60 * 60 * 1000;
    const NEWSLETTER_BANNER_DISMISSED_KEY = "newsletterBannerDismissed"

    var script = document.createElement('script');
    script.src = 'https://js.hsforms.net/forms/v2.js';
    script.defer = true;
    document.head.appendChild(script);

    // Run the code once the script is loaded
    script.onload = async function() {
      await hbspt.forms.create({
        region: "na1",
        portalId: "7272273",
        formId: "eb65b811-0451-414d-8304-7b9b6f468ce5",
        target: '#newsletter-form-container',
        onFormReady() {
          // get element by type "email"
          const emailInput = document.querySelector('#email-eb65b811-0451-414d-8304-7b9b6f468ce5');
          emailInput.setAttribute('placeholder', 'Enter email address');

          // remove the label element for emailInput
          const emailLabel = document.querySelector('#label-email-eb65b811-0451-414d-8304-7b9b6f468ce5');
          emailLabel.remove();
        },
        submitText: 'Subscribe',
      });
    };

    checkNewsletterBanner();

    document.querySelector('#newsletter-banner-close-button').addEventListener('click', () => {
      document.querySelector('#newsletter-banner').remove();
      localStorage.setItem(NEWSLETTER_BANNER_DISMISSED_KEY, Date.now());
    });

    const modal = document.querySelector('#newsletter-modal');

    document.querySelector('#newsletter-subscribe-button').addEventListener('click', () => {
      modal.showModal();
    });

    document.querySelector('#newsletter-close-button').addEventListener('click', () => {
      modal.close();
    });

    function checkNewsletterBanner() {
      /**
       * (thuang): Use LocalStorage to store dismissed state for 30 days
       * NOTE: Currently Census doc page doesn't share the same domain as the main site,
       * so dismissing the banner on the main site won't dismiss it on the Census doc page.
       * And vice versa.
       */
      const newsletterBannerDismissed = localStorage.getItem('newsletterBannerDismissed');

      if (newsletterBannerDismissed) {
        return;
      }

      if (newsletterBannerDismissed && Date.now() - newsletterBannerDismissed > NEWSLETTER_BANNER_DISMISSED_TTL_MS) {
        localStorage.removeItem(NEWSLETTER_BANNER_DISMISSED_KEY);
      }

      const newsletterBanner = document.querySelector('#newsletter-banner');

      if (!newsletterBannerDismissed) {
        newsletterBanner.style.display = 'flex';
      }
    }
  </script>
  
  
    
   

</body>
</html>