

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8">
  <meta name="generator" content="Docutils 0.19: https://docutils.sourceforge.io/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Benchmarks of single-cell Census models &mdash; cellxgene-census  documentation</title>
  

  
  <link rel="icon" type="image/png" sizes="32x32" href="../../_static/img/favicon_32x32_v2.png"/>
  <link rel="icon" type="image/png" sizes="16x16" href="../../_static/img/favicon_16x16_v2.png"/>
  <link rel="preconnect" href="https://fonts.googleapis.com">
  <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
  <link href="https://fonts.googleapis.com/css2?family=Inter:wght@400;500;600;700&display=swap" rel="stylesheet">
  <link rel="stylesheet" href="../../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/css/custom.css" type="text/css" />

  
  
  
  

  
  <!--[if lt IE 9]>
    <script src="../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../../" src="../../_static/documentation_options.js"></script>
        <script src="../../_static/jquery.js"></script>
        <script src="../../_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script data-url_root="../../" id="documentation_options" src="../../_static/documentation_options.js"></script>
        <script src="../../_static/doctools.js"></script>
        <script src="../../_static/sphinx_highlight.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script data-domain="chanzuckerberg.github.io/cellxgene-census" defer="defer" src="https://plausible.io/js/script.js"></script>
    
    <script type="text/javascript" src="../../_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="CZ CELLxGENE Discover Census in AWS" href="../../cellxgene_census_aws_open_data.html" />
    <link rel="prev" title="First stable iteration of Census (SOMA) PyTorch loaders" href="20240709-pytorch.html" /> 
</head>

<body class="wy-body-for-nav">

  <div style="height: 100vh; overflow: hidden;">
     
    <!-- top 56px for navbar height -->
    <div class="navbar-cxg">
        <div style="padding: 0 16px; display: flex; height: inherit; align-items: center; justify-content: space-between;">
            <span style="display:flex; gap: 32px; align-items: center;">
                <a href="https://cellxgene.cziscience.com/">
                  <img src="../../_static/img/cellxgene-discover-logo.svg" />
                </a>
                <span class="navbar-cxg-nav-wrapper">
                  <span class="navbar-cxg-section">
                    <span class="navbar-cxg-nav-section-title">Application</span>
                    <span class="navbar-cxg-nav-item-container">
                      <span class="navbar-cxg-link">
                          <a class="navbar-cxg-anchor" href="https://cellxgene.cziscience.com/collections">Collections</a>
                      </span>
                      <span class="navbar-cxg-link">
                          <a class="navbar-cxg-anchor" href="https://cellxgene.cziscience.com/datasets">Datasets</a>
                      </span>
                      <span class="navbar-cxg-link">
                          <a class="navbar-cxg-anchor" href="https://cellxgene.cziscience.com/gene-expression">Gene Expression</a>
                      </span>
                      <span class="navbar-cxg-link">
                        <a class="navbar-cxg-anchor" href="https://cellxgene.cziscience.com/cellguide">Cell Guide</a>
                      </span>
                      <span class="navbar-cxg-link">
                        <a class="navbar-cxg-anchor" href="https://cellxgene.cziscience.com/differential-expression">Differential Expression</a>
                        <div style="height: 16px!important; display: flex;">
                          <span class="beta">NEW</span>
                        </div>
                      </span>
                    </span>
                  </span>
                  <hr class="navbar-divider"/>
                  <span class="navbar-cxg-section">
                    <span class="navbar-cxg-nav-section-title">Census</span>
                    <span class="navbar-cxg-nav-item-container">
                      <span class="navbar-cxg-link active-link">
                        <a class="navbar-cxg-anchor" href="/cellxgene-census/index.html">API</a>
                      </span>
                      <span class="navbar-cxg-link">
                        <a class="navbar-cxg-anchor" href="https://cellxgene.cziscience.com/census-models">Models</a>
                      </span>
                    </span>
                  </span>
                </span>
            </span>

            <span class="navbar-cxg-link">
              <a class="navbar-cxg-anchor" href="https://cellxgene.cziscience.com/docs">Help & Documentation</a>
            </span>
        </div>
    </div>
    <div style="width: 100%; height: 100%; overflow: auto; padding-top: 56px;">
        
        <!-- top 56px for navbar height -->
        <nav data-toggle="wy-nav-shift" class="wy-nav-side" style="top: 56px;" >
        <div class="wy-side-scroll">
            <div class="wy-side-nav-search" >
            

            
                <a href="../../index.html" class="icon icon-home"> cellxgene-census
            

            
            </a>

            
                
                
                <div class="version">
                    1.16.2
                </div>
                
            

            
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

            
            </div>

            
            <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
            
                
                
                
                
                
                <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../../cellxgene_census_docsite_installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../cellxgene_census_docsite_quick_start.html">Quick start</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../../articles.html">What’s new?</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="../../articles.html#id1">2023</a></li>
<li class="toctree-l2 current"><a class="reference internal" href="../../articles.html#id2">2024</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="20240404-categoricals.html">Census supports categoricals for cell metadata</a></li>
<li class="toctree-l3"><a class="reference internal" href="20240709-pytorch.html">First stable iteration of Census (SOMA) PyTorch loaders</a></li>
<li class="toctree-l3 current"><a class="current reference internal" href="#">Benchmarks of single-cell Census models</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#accessing-the-embeddings-included-in-the-benchmark">Accessing the embeddings included in the benchmark</a></li>
<li class="toctree-l4"><a class="reference internal" href="#benchmarks-of-census-embeddings">Benchmarks of Census Embeddings</a></li>
<li class="toctree-l4"><a class="reference internal" href="#source-data">Source data</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../cellxgene_census_aws_open_data.html">Census in AWS ☁️</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../cellxgene_census_docsite_schema.html">Census data and schema</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../cellxgene_census_docsite_data_release_info.html">Census data releases</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../python-api.html">Python API</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../examples.html">Python tutorials</a></li>
<li class="toctree-l1"><a class="reference external" href="https://chanzuckerberg.github.io/cellxgene-census/r/index.html">R API &amp; tutorials</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../cellxgene_census_docsite_FAQ.html">FAQ</a></li>
<li class="toctree-l1"><a class="reference external" href="https://github.com/chanzuckerberg/cellxgene-census/releases">Release notes</a></li>
</ul>

                
            
            </div>
            
        </div>
        </nav>

        <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

        
        <nav class="wy-nav-top" aria-label="top navigation">
            
            <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
            <a href="../../index.html">cellxgene-census</a>
            
        </nav>


        <div class="wy-nav-content">
            
            <div class="rst-content">
            
            <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../../articles.html">What’s new?</a></li>
      <li class="breadcrumb-item active">Benchmarks of single-cell Census models</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../../_sources/articles/2024/20240710_embedding_metrics_dec_2023_lts.md.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
            <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
            <div itemprop="articleBody">
                
  <section id="benchmarks-of-single-cell-census-models">
<h1>Benchmarks of single-cell Census models<a class="headerlink" href="#benchmarks-of-single-cell-census-models" title="Permalink to this heading"></a></h1>
<p><em>Published:</em> <em>July 11th, 2024</em></p>
<p><em>Updated:</em> <em>July 19th, 2024</em>. Model names fixed.</p>
<p><em>By:</em> <em><a class="reference external" href="mailto:ebezzi&#37;&#52;&#48;chanzuckerberg&#46;com">Emanuele Bezzi</a>, <a class="reference external" href="mailto:pgarcia-nieto&#37;&#52;&#48;chanzuckerberg&#46;com">Pablo Garcia-Nieto</a></em></p>
<p>In 2023, the Census team released a series of cells embeddings (available at the <a class="reference external" href="https://cellxgene.cziscience.com/census-models">Census Model page</a>) compatible with the <a class="reference external" href="https://chanzuckerberg.github.io/cellxgene-census/cellxgene_census_docsite_data_release_info.html#lts-2023-12-15">Census LTS version <code class="docutils literal notranslate"><span class="pre">census_version=&quot;2023-12-15&quot;</span></code></a>, so that users can access and download for any slice of Census data.</p>
<p>These embeddings were generated via different large-scale models; in this article we present the results of light benchmarking of them. We hope that these benchmarks provide an initial picture to users on, 1) the strength of biological signal captured by these embeddings and, 2) the level of batch correction they exert.</p>
<p>We advise our users to consider these benchmarks as first-pass information and we recommend further benchmarking for a more comprehensive view of the embeddings and for task-oriented applications.</p>
<p>The benchmarks were run on the following embeddings:</p>
<ul class="simple">
<li><p>scVI latent spaces from a model trained on all Census data.</p></li>
<li><p>Fine-tuned Geneformer.</p></li>
<li><p>scGPT.</p></li>
<li><p>Universal Cell Embeddings (UCE).</p></li>
</ul>
<p>For more details on each model please see the <a class="reference external" href="https://cellxgene.cziscience.com/census-models">Census Model page</a>.</p>
<section id="accessing-the-embeddings-included-in-the-benchmark">
<h2>Accessing the embeddings included in the benchmark<a class="headerlink" href="#accessing-the-embeddings-included-in-the-benchmark" title="Permalink to this heading"></a></h2>
<p>Please the <a class="reference external" href="https://cellxgene.cziscience.com/census-models">Census Model page</a> for full details. Shortly, you can see the embeddings available for the Census LTS version <code class="docutils literal notranslate"><span class="pre">census_version=&quot;2023-12-15&quot;</span></code> using the Census API as follows.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">cellxgene_census.experimental.get_all_available_embeddings</span>
<span class="n">cellxgene_census</span><span class="o">.</span><span class="n">experimental</span><span class="o">.</span><span class="n">get_all_available_embeddings</span><span class="p">(</span><span class="n">census_version</span><span class="o">=</span><span class="s2">&quot;2023-12-15&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>With the exception of NMF factors, all other human embeddings were included in the benchmarks below. If you would want to access the embeddings for any slice of data you can utilize the parameter <code class="docutils literal notranslate"><span class="pre">obs_embeddings</span></code> from  the<code class="docutils literal notranslate"><span class="pre">get_anndata()</span></code> method of the Census API, for example:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">cellxgene_census</span>
<span class="n">census</span> <span class="o">=</span> <span class="n">cellxgene_census</span><span class="o">.</span><span class="n">open_soma</span><span class="p">(</span><span class="n">census_version</span><span class="o">=</span><span class="s2">&quot;2023-12-15&quot;</span><span class="p">)</span>
<span class="n">adata</span> <span class="o">=</span> <span class="n">cellxgene_census</span><span class="o">.</span><span class="n">get_anndata</span><span class="p">(</span>
    <span class="n">census</span><span class="p">,</span>
    <span class="n">organism</span> <span class="o">=</span> <span class="s2">&quot;homo_sapiens&quot;</span><span class="p">,</span>
    <span class="n">measurement_name</span> <span class="o">=</span> <span class="s2">&quot;RNA&quot;</span><span class="p">,</span>
    <span class="n">obs_value_filter</span> <span class="o">=</span> <span class="s2">&quot;tissue_general == &#39;central nervous system&#39;&quot;</span><span class="p">,</span>
    <span class="n">obs_embeddings</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;scvi&quot;</span><span class="p">]</span>
<span class="p">)</span>
</pre></div>
</div>
</section>
<section id="benchmarks-of-census-embeddings">
<h2>Benchmarks of Census Embeddings<a class="headerlink" href="#benchmarks-of-census-embeddings" title="Permalink to this heading"></a></h2>
<section id="about-the-benchmarks">
<h3>About the benchmarks<a class="headerlink" href="#about-the-benchmarks" title="Permalink to this heading"></a></h3>
<p>We executed a series of benchmarks falling into two general types: one to assess the level of biological signal contained in the embeddings, and the second to measure the level of correction for batch effects. In general, the utility of the embeddings increases as a function of these two set of benchmarks.</p>
<p>For each of the type, the benchmarks can be further subdivided by their “mode”. A series of metrics assess the embedding space, and the others assess the capacity of the embeddings to predict labels.</p>
<p>The table below shows a breakdown of the benchmarks we used in this report.</p>
<table class="custom-table">
  <thead>
      <tr>
        <th>Type</th>
        <th>Mode</th>
        <th>Metric</th>
        <th>Description</th>
      </tr>
  </thead>
  <tbody>
      <tr>
        <td rowspan="6">Bio-conservation</td>
        <td rowspan="3">Embedding<br>Space</td>
        <td><code>leiden_nmi</code></td>
        <td>Normalized Mutual Information of biological labels and leiden clusters. Described in <a href="https://scib-metrics.readthedocs.io/en/stable/generated/scib_metrics.nmi_ari_cluster_labels_leiden.html">Luecken et al.</a> and implemented in <a href="https://scib-metrics.readthedocs.io/en/stable/generated/scib_metrics.nmi_ari_cluster_labels_leiden.html">scib-metrics.</a></td>
      </tr>
      <tr>
        <td><code>leiden_ari</code></td>
        <td>Adjusted Rand Index of biological labels and leiden clusters. Described in <a href="https://scib-metrics.readthedocs.io/en/stable/generated/scib_metrics.nmi_ari_cluster_labels_leiden.html">Luecken et al.</a> and implemented in <a href="https://scib-metrics.readthedocs.io/en/stable/generated/scib_metrics.nmi_ari_cluster_labels_leiden.html">scib-metrics.</a></td>
      </tr>
      <tr>
        <td><code>silhouette_label</code></td>
        <td>Silhouette score with respect to biological labels. Described in <a href="https://scib-metrics.readthedocs.io/en/stable/generated/scib_metrics.nmi_ari_cluster_labels_leiden.html">Luecken et al.</a> and implemented in <a href="https://scib-metrics.readthedocs.io/en/stable/generated/scib_metrics.silhouette_label.html">scib-metrics.</a></td>
      </tr>
      <tr>
           <td rowspan="3">Label<br>Classifier</td>
        <td><code>classifier_svm</code></td>
        <td>Accuracy of biological label prediction using a SVM (60/40 train/test split). Implemented <a href="https://github.com/chanzuckerberg/cellxgene-census/blob/f44637ba33567400820407f4f7b9984e52966156/tools/models/metrics/run-scib.py#L36">here</a>.</td>
      </tr>
      <tr>
<td><code>classifier_forest</code></td>
        <td>Accuracy of biological label prediction using a Random Forest classifier (60/40 train/test split). Implemented <a href="https://github.com/chanzuckerberg/cellxgene-census/blob/f44637ba33567400820407f4f7b9984e52966156/tools/models/metrics/run-scib.py#L39">here</a>.</td>
      </tr>
      <tr>
<td><code>classifier_lr</code></td>
        <td>Accuracy of biological label prediction using a Logistic regression classifier (60/40 train/test split). Implemented <a href="https://github.com/chanzuckerberg/cellxgene-census/blob/f44637ba33567400820407f4f7b9984e52966156/tools/models/metrics/run-scib.py#L39">here</a>.</td>
      </tr>
      <tr>
        <td rowspan="5">Batch-correction</td>
        <td rowspan="2">Embedding<br>Space</td>
        <td><code>silhouette_batch</code></td>
        <td>1- silhouette score with respect to biological labels. Described in <a href="https://scib-metrics.readthedocs.io/en/stable/generated/scib_metrics.nmi_ari_cluster_labels_leiden.html">Luecken et al.</a> and implemented in <a href="https://scib-metrics.readthedocs.io/en/stable/generated/scib_metrics.nmi_ari_cluster_labels_leiden.html">scib-metrics.</a></td>
      </tr>
      <tr>
        <td><code>entropy</code></td>
        <td>Average of neighborhood entropy of batch labels per cell. Implemented <a href="https://github.com/chanzuckerberg/cellxgene-census/blob/f44637ba33567400820407f4f7b9984e52966156/tools/models/metrics/run-scib.py#L86">here</a>.</td>
      </tr>
      <tr>
           <td rowspan="3">Label<br>Classifier</td>
        <td><code>classifier_svm</code></td>
        <td>1 - accuracy of batch label prediction using a SVM (60/40 train/test split). Implemented <a href="https://github.com/chanzuckerberg/cellxgene-census/blob/f44637ba33567400820407f4f7b9984e52966156/tools/models/metrics/run-scib.py#L45">here</a>.</td>
      </tr>
      <tr>
<td><code>classifier_forest</code></td>
        <td>1 - accuracy of batch label prediction using a Random Forest classifier (60/40 train/test split). Implemented <a href="https://github.com/chanzuckerberg/cellxgene-census/blob/f44637ba33567400820407f4f7b9984e52966156/tools/models/metrics/run-scib.py#L48">here</a>.</td>
      </tr>
      <tr>
<td><code>classifier_lr</code></td>
        <td>1 - accuracy of batch label prediction using a Logistic regression classifier (60/40 train/test split). Implemented <a href="https://github.com/chanzuckerberg/cellxgene-census/blob/f44637ba33567400820407f4f7b9984e52966156/tools/models/metrics/run-scib.py#L42">here</a>.</td>
      </tr>
  </tbody>
</table>
<p><strong>Table 1:</strong> List of benchmarks.</p>
</section>
<section id="benchmark-results">
<h3>Benchmark results<a class="headerlink" href="#benchmark-results" title="Permalink to this heading"></a></h3>
<p>As reminder the benchmarks were run on the following embeddings:</p>
<ul class="simple">
<li><p>scVI latent spaces from a model trained on all Census data.</p></li>
<li><p>Fine-tuned Geneformer.</p></li>
<li><p>scGPT.</p></li>
<li><p>Universal Cell Embeddings (UCE).</p></li>
</ul>
<section id="summary">
<h4>Summary<a class="headerlink" href="#summary" title="Permalink to this heading"></a></h4>
<p>The following are averages for all the metrics shown in the following sections.</p>
<figure class="align-center" id="id1" style="width: 90%">
<img alt="Bio-conservation single-cell Census benchmark" src="../../_images/20240710_metrics_0_summary.png" />
<figcaption>
<p><span class="caption-text"><strong>Figure 1. Summary of all benchmarks.</strong> Numerical averages across the metric types and modes from all bio- and batch-labels across the tissues in this report.</span><a class="headerlink" href="#id1" title="Permalink to this image"></a></p>
</figcaption>
</figure>
</section>
<section id="bio-conservation">
<h4>Bio-conservation<a class="headerlink" href="#bio-conservation" title="Permalink to this heading"></a></h4>
<p>The bio-conservation metrics were run the in following biological labels in a Census cells from Adipose Tissue and Spinal Cord:</p>
<ul class="simple">
<li><p>Cell subclass: a higher definition of a cell type with maximum of 73 unique labels, as defined on the CELLxGENE collection page.</p></li>
<li><p>Cell class: an even higher definition of a cell type with a maximum of 22 unique labels, also defined on the CELLxGENE collection page.</p></li>
</ul>
<figure class="align-center" id="id2" style="width: 90%">
<img alt="Bio-conservation single-cell Census benchmark" src="../../_images/20240710_metrics_1_bio_emb.png" />
<figcaption>
<p><span class="caption-text"><strong>Figure 2. Bio-conservation metrics on the embedding space.</strong> Higher values signify better performance, max value for all metrics is 1.</span><a class="headerlink" href="#id2" title="Permalink to this image"></a></p>
</figcaption>
</figure>
<figure class="align-center" id="id3" style="width: 90%">
<img alt="Bio-conservation single-cell Census benchmark" src="../../_images/20240710_metrics_2_bio_classifier.png" />
<figcaption>
<p><span class="caption-text"><strong>Figure 3. Bio-conservation metrics based on label classifiers.</strong> Values represent label prediction accuracy. Higher values signify better performance, max value for all metrics is 1.</span><a class="headerlink" href="#id3" title="Permalink to this image"></a></p>
</figcaption>
</figure>
</section>
<section id="batch-correction">
<h4>Batch-correction<a class="headerlink" href="#batch-correction" title="Permalink to this heading"></a></h4>
<p>The batch-correction metrics were run the in following batch labels in a Census cells from Adipose Tissue and Spinal Cord:</p>
<ul class="simple">
<li><p>Assay: the sequencing technology.</p></li>
<li><p>Dataset: the dataset from which the cell originated from.</p></li>
<li><p>Suspension type: cell vs nucleus.</p></li>
<li><p>Batch: the concatenation of values for all of the above.</p></li>
</ul>
<figure class="align-center" id="id4" style="width: 90%">
<img alt="Batch-correction single-cell Census benchmark" src="../../_images/20240710_metrics_3_batch_emb.png" />
<figcaption>
<p><span class="caption-text"><strong>Figure 4. Batch-correction metrics on the embedding space.</strong> Higher values signify better performance, max value for <code class="docutils literal notranslate"><span class="pre">silhouette_batch</span></code> is 1, <code class="docutils literal notranslate"><span class="pre">entropy</span></code> values should only be compared within the tissue/label combination and not across.</span><a class="headerlink" href="#id4" title="Permalink to this image"></a></p>
</figcaption>
</figure>
<figure class="align-center" id="id5" style="width: 90%">
<img alt="Batch-correction single-cell Census benchmark" src="../../_images/20240710_metrics_4_batch_classifier.png" />
<figcaption>
<p><span class="caption-text"><strong>Figure 5. Batch-correction metrics based on label classifiers.</strong> Values represent <strong>1 - label prediction accuracy</strong>. In theory higher values signify better performance indicating that prediction of batch labels is not accurate. However foundation models may be designed to learn <em>all</em> information including technical variation, please refer to the original publications of the models to learn more about them.</span><a class="headerlink" href="#id5" title="Permalink to this image"></a></p>
</figcaption>
</figure>
</section>
</section>
</section>
<section id="source-data">
<h2>Source data<a class="headerlink" href="#source-data" title="Permalink to this heading"></a></h2>
<p>All data was obtained from the Census API, to fetch the data used in this report you can execute the following in Python. To get the cell subclass and cell class please refer to the <a class="reference external" href="https://github.com/chanzuckerberg/cellxgene-ontology-guide/tree/main">CellxGene Ontology Guide API</a>.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">cellxgene_census</span>

<span class="n">val_filters</span> <span class="o">=</span> <span class="p">{</span>
   <span class="s2">&quot;adipose&quot;</span><span class="p">:</span> <span class="s2">&quot;tissue_general == &#39;adipose tissue&#39; and is_primary_data == True&quot;</span><span class="p">,</span>
   <span class="s2">&quot;spinal&quot;</span><span class="p">:</span> <span class="s2">&quot;tissue_general == &#39;spinal cord&#39; and is_primary_data == True&quot;</span><span class="p">,</span>
<span class="p">}</span>

<span class="n">embedding_names</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;geneformer&quot;</span><span class="p">,</span> <span class="s2">&quot;scgpt&quot;</span><span class="p">,</span> <span class="s2">&quot;scvi&quot;</span><span class="p">,</span> <span class="s2">&quot;uce&quot;</span><span class="p">]</span>
<span class="n">embedding_names</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;scvi&quot;</span><span class="p">]</span>
<span class="n">column_names</span> <span class="o">=</span> <span class="p">{</span>
   <span class="s2">&quot;obs&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;cell_type_ontology_term_id&quot;</span><span class="p">,</span> <span class="s2">&quot;cell_type&quot;</span><span class="p">,</span> <span class="s2">&quot;assay&quot;</span><span class="p">,</span> <span class="s2">&quot;suspension_type&quot;</span><span class="p">,</span> <span class="s2">&quot;dataset_id&quot;</span><span class="p">,</span> <span class="s2">&quot;soma_joinid&quot;</span><span class="p">]</span>
<span class="p">}</span>

<span class="n">census</span> <span class="o">=</span> <span class="n">cellxgene_census</span><span class="o">.</span><span class="n">open_soma</span><span class="p">(</span><span class="n">census_version</span><span class="o">=</span><span class="s2">&quot;2023-12-15&quot;</span><span class="p">)</span>

<span class="n">adatas</span> <span class="o">=</span> <span class="p">[]</span>
<span class="k">for</span> <span class="n">tissue</span> <span class="ow">in</span> <span class="n">val_filters</span><span class="p">:</span>
    <span class="n">adatas</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
       <span class="n">cellxgene_census</span><span class="o">.</span><span class="n">get_anndata</span><span class="p">(</span>
           <span class="n">census</span><span class="p">,</span>
           <span class="n">organism</span><span class="o">=</span><span class="s2">&quot;homo_sapiens&quot;</span><span class="p">,</span>
           <span class="n">measurement_name</span><span class="o">=</span><span class="s2">&quot;RNA&quot;</span><span class="p">,</span>
           <span class="n">obs_value_filter</span><span class="o">=</span> <span class="n">val_filters</span><span class="p">[</span><span class="n">tissue</span><span class="p">],</span>
           <span class="n">obs_embeddings</span><span class="o">=</span><span class="n">embedding_names</span><span class="p">,</span>
           <span class="n">column_names</span><span class="o">=</span><span class="n">column_names</span><span class="p">,</span>
        <span class="p">)</span>
    <span class="p">)</span>
</pre></div>
</div>
<section id="batch-label-counts">
<h3>Batch label counts<a class="headerlink" href="#batch-label-counts" title="Permalink to this heading"></a></h3>
<p>The following shows the batch label counts per tissue:</p>
<section id="adipose-tissue">
<h4>Adipose tissue<a class="headerlink" href="#adipose-tissue" title="Permalink to this heading"></a></h4>
<table class="custom-table">
  <thead>
      <tr>
        <th>Type</th>
        <th>Label</th>
        <th>Count</th>
      </tr>
  </thead>
  <tbody>
      <tr>
          <td rowspan="4">Assay</td>
          <td>10x 3' v3</td>
          <td>91947</td>
      </tr>
      <tr>
          <td>10x 5' transcription profiling</td>
          <td>2121</td>
      </tr>
      <tr>
          <td>microwell-seq</td>
          <td>5916</td>
      </tr>
      <tr>
          <td>Smart-seq2</td>
          <td>651</td>
      </tr>
      <tr>
          <td rowspan="2">Suspension Type</td>
          <td>nucleus</td>
          <td>72335</td>
      </tr>
      <tr>
          <td>cell</td>
          <td>23756</td>
      </tr>
      <tr>
          <td rowspan="4">Dataset</td>
          <td>9d8e5dca-03a3-457d-b7fb-844c75735c83</td>
          <td>72335</td>
      </tr>
      <tr>
          <td>53d208b0-2cfd-4366-9866-c3c6114081bc</td>
          <td>20263</td>
      </tr>
      <tr>
          <td>5af90777-6760-4003-9dba-8f945fec6fdf</td>
          <td>2121</td>
      </tr>
      <tr>
          <td>2adb1f8a-a6b1-4909-8ee8-484814e2d4bf</td>
          <td>1372</td>
      </tr>
   </tbody>
</table>
</section>
<section id="spinal-cord">
<h4>Spinal cord<a class="headerlink" href="#spinal-cord" title="Permalink to this heading"></a></h4>
<table class="custom-table">
  <thead>
      <tr>
        <th>Type</th>
        <th>Label</th>
        <th>Count</th>
      </tr>
  </thead>
  <tbody>
      <tr>
          <td rowspan="2">Assay</td>
          <td>10x 3' v3</td>
          <td>43840</td>
      </tr>
      <tr>
          <td>microwell-seq</td>
          <td>5916</td>
      </tr>
      <tr>
          <td rowspan="2">Suspension Type</td>
          <td>nucleus</td>
          <td>43840</td>
      </tr>
      <tr>
          <td>cell</td>
          <td>5916</td>
      </tr>
      <tr>
          <td rowspan="3">Dataset</td>
          <td>090da8ea-46e8-40df-bffc-1f78e1538d27</td>
          <td>24190</td>
      </tr>
      <tr>
          <td>c05e6940-729c-47bd-a2a6-6ce3730c4919</td>
          <td>19650</td>
      </tr>
      <tr>
          <td>2adb1f8a-a6b1-4909-8ee8-484814e2d4bf</td>
          <td>5916</td>
      </tr>
   </tbody>
</table>
</section>
</section>
</section>
</section>


            </div>
            
            </div>
            <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="20240709-pytorch.html" class="btn btn-neutral float-left" title="First stable iteration of Census (SOMA) PyTorch loaders" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="../../cellxgene_census_aws_open_data.html" class="btn btn-neutral float-right" title="CZ CELLxGENE Discover Census in AWS" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2022, Chan Zuckerberg Initiative.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
            </div>
        </div>

        </section>

    </div>
    <!-- Newsletter Banner -->
    <div role="banner" id="newsletter-banner">
      <span id="newsletter-subscribe-button" role="button">Subscribe</span>&nbsp;to our newsletter to receive updates about new features.
      <div id="newsletter-banner-close-button" role="button">X</div>
    </div>
    <!-- Newsletter Modal -->
    <dialog id="newsletter-modal">
      <div id="newsletter-header">
        <img id="newsletter-logo" src="../../_static/img/cellxGene-newsletter-logo.svg" />
        <div id="newsletter-close-button" role="button">X</div>
      </div>
      <div id="newsletter-content">
        <div id="newsletter-callout">Join Our Newsletter</div>
        <div id="newsletter-description">Get a quarterly email with the latest CELLxGENE features and data.</div>
        <!-- HubSpot Form target -->
        <div id="newsletter-form-container"></div>
      </div>

      <div id="newsletter-footnote">Unsubscribe at any time.</div>
    </dialog>
    
  

    <script>
    // (thuang): 30 days
    const NEWSLETTER_BANNER_DISMISSED_TTL_MS = 30 * 24 * 60 * 60 * 1000;
    const NEWSLETTER_BANNER_DISMISSED_KEY = "newsletterBannerDismissed"

    var script = document.createElement('script');
    script.src = 'https://js.hsforms.net/forms/v2.js';
    script.defer = true;
    document.head.appendChild(script);

    // Run the code once the script is loaded
    script.onload = async function() {
      await hbspt.forms.create({
        region: "na1",
        portalId: "7272273",
        formId: "eb65b811-0451-414d-8304-7b9b6f468ce5",
        target: '#newsletter-form-container',
        onFormReady() {
          // get element by type "email"
          const emailInput = document.querySelector('#email-eb65b811-0451-414d-8304-7b9b6f468ce5');
          emailInput.setAttribute('placeholder', 'Enter email address');

          // remove the label element for emailInput
          const emailLabel = document.querySelector('#label-email-eb65b811-0451-414d-8304-7b9b6f468ce5');
          emailLabel.remove();
        },
        submitText: 'Subscribe',
      });
    };

    checkNewsletterBanner();

    document.querySelector('#newsletter-banner-close-button').addEventListener('click', () => {
      document.querySelector('#newsletter-banner').remove();
      localStorage.setItem(NEWSLETTER_BANNER_DISMISSED_KEY, Date.now());
    });

    const modal = document.querySelector('#newsletter-modal');

    document.querySelector('#newsletter-subscribe-button').addEventListener('click', () => {
      modal.showModal();
    });

    document.querySelector('#newsletter-close-button').addEventListener('click', () => {
      modal.close();
    });

    function checkNewsletterBanner() {
      /**
       * (thuang): Use LocalStorage to store dismissed state for 30 days
       * NOTE: Currently Census doc page doesn't share the same domain as the main site,
       * so dismissing the banner on the main site won't dismiss it on the Census doc page.
       * And vice versa.
       */
      const newsletterBannerDismissed = localStorage.getItem('newsletterBannerDismissed');

      if (newsletterBannerDismissed) {
        return;
      }

      if (newsletterBannerDismissed && Date.now() - newsletterBannerDismissed > NEWSLETTER_BANNER_DISMISSED_TTL_MS) {
        localStorage.removeItem(NEWSLETTER_BANNER_DISMISSED_KEY);
      }

      const newsletterBanner = document.querySelector('#newsletter-banner');

      if (!newsletterBannerDismissed) {
        newsletterBanner.style.display = 'flex';
      }
    }
  </script>
  
  
    
   

</body>
</html>