

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8">
  <meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Normalizing full-length gene sequencing data &mdash; cellxgene-census  documentation</title>
  

  
  <link rel="icon" type="image/png" sizes="32x32" href="../../_static/img/favicon_32x32_v2.png"/>
  <link rel="icon" type="image/png" sizes="16x16" href="../../_static/img/favicon_16x16_v2.png"/>
  <link rel="preconnect" href="https://fonts.googleapis.com">
  <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
  <link href="https://fonts.googleapis.com/css2?family=Inter:wght@400;500;600;700&display=swap" rel="stylesheet">
  <link rel="stylesheet" href="../../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/css/custom.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/nbsphinx-code-cells.css" type="text/css" />

  
  
  
  

  
  <!--[if lt IE 9]>
    <script src="../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../../" src="../../_static/documentation_options.js"></script>
        <script data-url_root="../../" id="documentation_options" src="../../_static/documentation_options.js"></script>
        <script src="../../_static/doctools.js"></script>
        <script src="../../_static/sphinx_highlight.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
        <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
        <script data-domain="chanzuckerberg.github.io/cellxgene-census" defer="defer" src="https://plausible.io/js/script.js"></script>
    
    <script type="text/javascript" src="../../_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="Summarizing cell and gene metadata" href="comp_bio_summarize_axis_query.html" />
    <link rel="prev" title="Exploring all data from a tissue" href="comp_bio_explore_and_load_lung_data.html" /> 
</head>

<body class="wy-body-for-nav">

  <div style="height: 100vh; overflow: hidden;">
     
    <!-- top 48px for navbar height -->
    <div class="navbar-cxg">
        <div style="padding: 0 16px; display: flex; height: inherit; align-items: center; justify-content: space-between;">
            <span style="display:flex; gap: 32px; align-items: center;">
                <a href="https://cellxgene.cziscience.com/">
                  <img src="../../_static/img/cellxgene-discover-logo.svg" />
                </a>
                <span style="display:flex; gap: 16px;">
                    <span class="navbar-cxg-link">
                        <a class="navbar-cxg-anchor" href="https://cellxgene.cziscience.com/collections">Collections</a>
                    </span>
                    <span class="navbar-cxg-link">
                        <a class="navbar-cxg-anchor" href="https://cellxgene.cziscience.com/datasets">Datasets</a>
                    </span>
                    <span class="navbar-cxg-link">
                        <a class="navbar-cxg-anchor" href="https://cellxgene.cziscience.com/gene-expression">Gene Expression</a>
                    </span>
                    <span class="navbar-cxg-link">
                      <a class="navbar-cxg-anchor" href="https://cellxgene.cziscience.com/cellguide">Cell Guide</a>
                      <div style="height: 16px!important; display: flex;">
                        <span class="beta">BETA</span>
                      </div>
                    </span>
                    <hr class="navbar-divider">
                    <span class="navbar-cxg-link">
                      <a class="navbar-cxg-anchor" href="/cellxgene-census/index.html">Census</a>
                  </span>
                </span>
            </span>

            <span class="navbar-cxg-link">
              <a class="navbar-cxg-anchor" href="https://cellxgene.cziscience.com/docs">Help & Documentation</a>
            </span>
        </div>
    </div>
    <div style="width: 100%; height: 100%; overflow: auto; padding-top: 48px;">
        
        <!-- top 48px for navbar height -->
        <nav data-toggle="wy-nav-shift" class="wy-nav-side" style="top: 48px;" >
        <div class="wy-side-scroll">
            <div class="wy-side-nav-search" >
            

            
                <a href="../../index.html" class="icon icon-home"> cellxgene-census
            

            
            </a>

            
                
                
                <div class="version">
                    v1.5.0
                </div>
                
            

            

            
            </div>

            
            <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
            
                
                
                
                
                
                <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../../cellxgene_census_docsite_installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../cellxgene_census_docsite_quick_start.html">Quick start</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../cellxgene_census_docsite_schema.html">Census data and schema</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../cellxgene_census_docsite_data_release_info.html">Census data releases</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../python-api.html">Python API</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../../examples.html">Python tutorials</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="../../examples.html#api">API</a></li>
<li class="toctree-l2 current"><a class="reference internal" href="../../examples.html#analysis">Analysis</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="comp_bio_census_info.html">Learning about the CZ CELLxGENE Census</a></li>
<li class="toctree-l3"><a class="reference internal" href="comp_bio_data_integration_scvi.html">Integrating multi-dataset slices of data</a></li>
<li class="toctree-l3"><a class="reference internal" href="comp_bio_explore_and_load_lung_data.html">Exploring all data from a tissue</a></li>
<li class="toctree-l3 current"><a class="current reference internal" href="#">Normalizing full-length gene sequencing data</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#Opening-the-census">Opening the census</a></li>
<li class="toctree-l4"><a class="reference internal" href="#Fetching-full-length-example-sequencing-data-(Smart-Seq)">Fetching full-length example sequencing data (Smart-Seq)</a></li>
<li class="toctree-l4"><a class="reference internal" href="#Normalizing-expression-to-account-for-gene-length">Normalizing expression to account for gene length</a></li>
<li class="toctree-l4"><a class="reference internal" href="#Validation-through-clustering-exploration">Validation through clustering exploration</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="comp_bio_summarize_axis_query.html">Summarizing cell and gene metadata</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../examples.html#experimental">Experimental</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference external" href="https://chanzuckerberg.github.io/cellxgene-census/r/index.html">R API &amp; tutorials</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../cellxgene_census_docsite_FAQ.html">FAQ</a></li>
</ul>

                
            
            </div>
            
        </div>
        </nav>

        <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

        
        <nav class="wy-nav-top" aria-label="top navigation">
            
            <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
            <a href="../../index.html">cellxgene-census</a>
            
        </nav>


        <div class="wy-nav-content">
            
            <div class="rst-content">
            
            <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../../examples.html">Python tutorials</a></li>
      <li class="breadcrumb-item active">Normalizing full-length gene sequencing data</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../../_sources/notebooks/analysis_demo/comp_bio_normalizing_full_gene_sequencing.ipynb.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
            <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
            <div itemprop="articleBody">
                
  <section id="Normalizing-full-length-gene-sequencing-data">
<h1>Normalizing full-length gene sequencing data<a class="headerlink" href="#Normalizing-full-length-gene-sequencing-data" title="Permalink to this heading">¶</a></h1>
<p>This tutorial shows you how to fetch full-length gene sequencing data from the Census and normalize it to account for gene length.</p>
<p><strong>Contents</strong></p>
<ol class="arabic simple">
<li><p>Opening the census</p></li>
<li><p>Fetching example full-length sequencing data (Smart-Seq2)</p></li>
<li><p>Normalizing expression to account for gene length</p></li>
<li><p>Validation through clustering exploration</p></li>
</ol>
<section id="Opening-the-census">
<h2>Opening the census<a class="headerlink" href="#Opening-the-census" title="Permalink to this heading">¶</a></h2>
<p>First we open the Census, if you are not familiar with the basics of the Census API you should take a look at notebook <a class="reference external" href="https://chanzuckerberg.github.io/cellxgene-census/notebooks/analysis_demo/comp_bio_census_info.html">Learning about the CZ CELLxGENE Census</a></p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">cellxgene_census</span>
<span class="kn">import</span> <span class="nn">scanpy</span> <span class="k">as</span> <span class="nn">sc</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">from</span> <span class="nn">scipy.sparse</span> <span class="kn">import</span> <span class="n">csr_matrix</span>

<span class="n">census</span> <span class="o">=</span> <span class="n">cellxgene_census</span><span class="o">.</span><span class="n">open_soma</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
The &#34;stable&#34; release is currently 2023-07-25. Specify &#39;census_version=&#34;2023-07-25&#34;&#39; in future calls to open_soma() to ensure data consistency.
</pre></div></div>
</div>
<p>You can learn more about the all of the <code class="docutils literal notranslate"><span class="pre">cellxgene_census</span></code> methods by accessing their corresponding documention via <code class="docutils literal notranslate"><span class="pre">help()</span></code>. For example <code class="docutils literal notranslate"><span class="pre">help(cellxgene_census.open_soma)</span></code>.</p>
</section>
<section id="Fetching-full-length-example-sequencing-data-(Smart-Seq)">
<h2>Fetching full-length example sequencing data (Smart-Seq)<a class="headerlink" href="#Fetching-full-length-example-sequencing-data-(Smart-Seq)" title="Permalink to this heading">¶</a></h2>
<p>Let’s get some example data, in this case we’ll fetch all cells from a relatively small dataset derived from the Smart-Seq2 technology which performs full-length gene sequencing:</p>
<ul class="simple">
<li><p>Collection: <a class="reference external" href="https://cellxgene.cziscience.com/collections/0b9d8a04-bb9d-44da-aa27-705bb65b54eb">Tabula Muris Senis</a></p></li>
<li><p>Dataset: <a class="reference external" href="https://cellxgene.cziscience.com/e/524179b0-b406-4723-9c46-293ffa77ca81.cxg/">Liver - A single-cell transcriptomic atlas characterizes ageing tissues in the mouse - Smart-seq2</a></p></li>
</ul>
<p>Let’s first find this dataset’s id by using the dataset table of the Census</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">liver_dataset</span> <span class="o">=</span> <span class="p">(</span>
    <span class="n">census</span><span class="p">[</span><span class="s2">&quot;census_info&quot;</span><span class="p">][</span><span class="s2">&quot;datasets&quot;</span><span class="p">]</span>
    <span class="o">.</span><span class="n">read</span><span class="p">(</span>
        <span class="n">value_filter</span><span class="o">=</span><span class="s2">&quot;dataset_title == &#39;Liver - A single-cell transcriptomic atlas characterizes ageing tissues in the mouse - Smart-seq2&#39;&quot;</span>
    <span class="p">)</span>
    <span class="o">.</span><span class="n">concat</span><span class="p">()</span>
    <span class="o">.</span><span class="n">to_pandas</span><span class="p">()</span>
<span class="p">)</span>
<span class="n">liver_dataset</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>soma_joinid</th>
      <th>collection_id</th>
      <th>collection_name</th>
      <th>collection_doi</th>
      <th>dataset_id</th>
      <th>dataset_title</th>
      <th>dataset_h5ad_path</th>
      <th>dataset_total_cell_count</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>525</td>
      <td>0b9d8a04-bb9d-44da-aa27-705bb65b54eb</td>
      <td>Tabula Muris Senis</td>
      <td>10.1038/s41586-020-2496-1</td>
      <td>4546e757-34d0-4d17-be06-538318925fcd</td>
      <td>Liver - A single-cell transcriptomic atlas cha...</td>
      <td>4546e757-34d0-4d17-be06-538318925fcd.h5ad</td>
      <td>2859</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<p>Now we can use this id to fetch the data.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">liver_dataset_id</span> <span class="o">=</span> <span class="s2">&quot;4546e757-34d0-4d17-be06-538318925fcd&quot;</span>
<span class="n">liver_adata</span> <span class="o">=</span> <span class="n">cellxgene_census</span><span class="o">.</span><span class="n">get_anndata</span><span class="p">(</span>
    <span class="n">census</span><span class="p">,</span> <span class="n">organism</span><span class="o">=</span><span class="s2">&quot;Mus musculus&quot;</span><span class="p">,</span> <span class="n">obs_value_filter</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;dataset_id==&#39;</span><span class="si">{</span><span class="n">liver_dataset_id</span><span class="si">}</span><span class="s2">&#39;&quot;</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<p>Let’s make sure this data only contains Smart-Seq2 cells</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">liver_adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;assay&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">value_counts</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
assay
Smart-seq2    2859
Name: count, dtype: int64
</pre></div></div>
</div>
<p>Great! As you can see this a small dataset only containing <strong>2,859</strong> cells. Now let’s proceed to normalize by gene lengths.</p>
<p>For good practice let’s add the number of genes expressed by cell, and the total counts per cell.</p>
<p>Don’t forget to close the census</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">census</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</pre></div>
</div>
</div>
</section>
<section id="Normalizing-expression-to-account-for-gene-length">
<h2>Normalizing expression to account for gene length<a class="headerlink" href="#Normalizing-expression-to-account-for-gene-length" title="Permalink to this heading">¶</a></h2>
<p>By default <code class="docutils literal notranslate"><span class="pre">cellxgene_census.get_anndata()</span></code> fetches all genes in the Census. So let’s first identify the genes that were measured in this dataset and subset the <code class="docutils literal notranslate"><span class="pre">AnnData</span></code> to only include those.</p>
<p>To this goal we can use the “Dataset Presence Matrix” in <code class="docutils literal notranslate"><span class="pre">census[&quot;census_data&quot;][&quot;mus_musculus&quot;].ms[&quot;RNA&quot;][&quot;feature_dataset_presence_matrix&quot;]</span></code>. This is a boolean matrix <code class="docutils literal notranslate"><span class="pre">N</span> <span class="pre">x</span> <span class="pre">M</span></code> where <code class="docutils literal notranslate"><span class="pre">N</span></code> is the number of datasets and <code class="docutils literal notranslate"><span class="pre">M</span></code> is the number of genes in the Census, <code class="docutils literal notranslate"><span class="pre">True</span></code> indicates that a gene was measured in a dataset.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">liver_adata</span><span class="o">.</span><span class="n">n_vars</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
52392
</pre></div></div>
</div>
<p>Let’s get the genes measured in this dataset.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">liver_dataset_id</span> <span class="o">=</span> <span class="n">liver_dataset</span><span class="p">[</span><span class="s2">&quot;soma_joinid&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>

<span class="n">presence_matrix</span> <span class="o">=</span> <span class="n">cellxgene_census</span><span class="o">.</span><span class="n">get_presence_matrix</span><span class="p">(</span><span class="n">census</span><span class="p">,</span> <span class="s2">&quot;Mus musculus&quot;</span><span class="p">,</span> <span class="s2">&quot;RNA&quot;</span><span class="p">)</span>
<span class="n">presence_matrix</span> <span class="o">=</span> <span class="n">presence_matrix</span><span class="p">[</span><span class="n">liver_dataset_id</span><span class="p">,</span> <span class="p">:]</span>
<span class="n">gene_presence</span> <span class="o">=</span> <span class="n">presence_matrix</span><span class="o">.</span><span class="n">nonzero</span><span class="p">()[</span><span class="mi">1</span><span class="p">]</span>

<span class="n">liver_adata</span> <span class="o">=</span> <span class="n">liver_adata</span><span class="p">[:,</span> <span class="n">gene_presence</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">liver_adata</span><span class="o">.</span><span class="n">n_vars</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
17992
</pre></div></div>
</div>
<p>We can see that out of the all genes in the Census <strong>17,992</strong> were measured in this dataset.</p>
<p>Now let’s normalize these genes by gene length. We can easily do this because the Census has gene lengths included in the gene metadata under <code class="docutils literal notranslate"><span class="pre">feature_length</span></code>.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">liver_adata</span><span class="o">.</span><span class="n">X</span><span class="p">[:</span><span class="mi">5</span><span class="p">,</span> <span class="p">:</span><span class="mi">5</span><span class="p">]</span><span class="o">.</span><span class="n">toarray</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
array([[0.000e+00, 0.000e+00, 0.000e+00, 0.000e+00, 0.000e+00],
       [0.000e+00, 0.000e+00, 5.590e+02, 0.000e+00, 0.000e+00],
       [0.000e+00, 0.000e+00, 1.969e+03, 0.000e+00, 8.280e+02],
       [0.000e+00, 0.000e+00, 0.000e+00, 0.000e+00, 1.000e+00],
       [0.000e+00, 2.250e+03, 0.000e+00, 0.000e+00, 5.400e+01]],
      dtype=float32)
</pre></div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">gene_lengths</span> <span class="o">=</span> <span class="n">liver_adata</span><span class="o">.</span><span class="n">var</span><span class="p">[[</span><span class="s2">&quot;feature_length&quot;</span><span class="p">]]</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">()</span>
<span class="n">liver_adata</span><span class="o">.</span><span class="n">X</span> <span class="o">=</span> <span class="n">csr_matrix</span><span class="p">((</span><span class="n">liver_adata</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">T</span> <span class="o">/</span> <span class="n">gene_lengths</span><span class="p">)</span><span class="o">.</span><span class="n">T</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">liver_adata</span><span class="o">.</span><span class="n">X</span><span class="p">[:</span><span class="mi">5</span><span class="p">,</span> <span class="p">:</span><span class="mi">5</span><span class="p">]</span><span class="o">.</span><span class="n">toarray</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
array([[0.00000000e+00, 0.00000000e+00, 0.00000000e+00, 0.00000000e+00,
        0.00000000e+00],
       [0.00000000e+00, 0.00000000e+00, 6.58654413e-02, 0.00000000e+00,
        0.00000000e+00],
       [0.00000000e+00, 0.00000000e+00, 2.32001885e-01, 0.00000000e+00,
        2.74444813e-01],
       [0.00000000e+00, 0.00000000e+00, 0.00000000e+00, 0.00000000e+00,
        3.31455088e-04],
       [0.00000000e+00, 4.71500419e-01, 0.00000000e+00, 0.00000000e+00,
        1.78985747e-02]])
</pre></div></div>
</div>
<p>All done! You can see that we now have real numbers instead of integers.</p>
</section>
<section id="Validation-through-clustering-exploration">
<h2>Validation through clustering exploration<a class="headerlink" href="#Validation-through-clustering-exploration" title="Permalink to this heading">¶</a></h2>
<p>Let’s perform some basic clustering analysis to see if cell types cluster as expected using the normalized counts.</p>
<p>First we do some basic filtering of cells and genes.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">filter_cells</span><span class="p">(</span><span class="n">liver_adata</span><span class="p">,</span> <span class="n">min_genes</span><span class="o">=</span><span class="mi">500</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">filter_genes</span><span class="p">(</span><span class="n">liver_adata</span><span class="p">,</span> <span class="n">min_cells</span><span class="o">=</span><span class="mi">5</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>Then we normalize to account for sequencing depth and transform data to log scale.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[13]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">normalize_total</span><span class="p">(</span><span class="n">liver_adata</span><span class="p">,</span> <span class="n">target_sum</span><span class="o">=</span><span class="mf">1e4</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">log1p</span><span class="p">(</span><span class="n">liver_adata</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>Then we subset to highly variable genes.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">highly_variable_genes</span><span class="p">(</span><span class="n">liver_adata</span><span class="p">,</span> <span class="n">n_top_genes</span><span class="o">=</span><span class="mi">1000</span><span class="p">)</span>
<span class="n">liver_adata</span> <span class="o">=</span> <span class="n">liver_adata</span><span class="p">[:,</span> <span class="n">liver_adata</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">highly_variable</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
</pre></div>
</div>
</div>
<p>And finally we scale values across the gene axis.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[15]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">scale</span><span class="p">(</span><span class="n">liver_adata</span><span class="p">,</span> <span class="n">max_value</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>Now we can proceed to do clustering analysis</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[16]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">pca</span><span class="p">(</span><span class="n">liver_adata</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">neighbors</span><span class="p">(</span><span class="n">liver_adata</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">liver_adata</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">liver_adata</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;cell_type&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/home/ssm-user/cellxgene-census/venv/lib/python3.10/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm
/home/ssm-user/cellxgene-census/venv/lib/python3.10/site-packages/scanpy/plotting/_tools/scatterplots.py:392: UserWarning: No data for colormapping provided via &#39;c&#39;. Parameters &#39;cmap&#39; will be ignored
  cax = scatter(
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/notebooks_analysis_demo_comp_bio_normalizing_full_gene_sequencing_31_1.png" src="../../_images/notebooks_analysis_demo_comp_bio_normalizing_full_gene_sequencing_31_1.png" />
</div>
</div>
<p>With a few exceptions we can see that all cells from the same cell type cluster near each other which serves as a sanity check for the gene-length normalization that we applied.</p>
</section>
</section>


            </div>
            
            </div>
            <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="comp_bio_explore_and_load_lung_data.html" class="btn btn-neutral float-left" title="Exploring all data from a tissue" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="comp_bio_summarize_axis_query.html" class="btn btn-neutral float-right" title="Summarizing cell and gene metadata" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2022-2023 Chan Zuckerberg Initiative.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
            </div>
        </div>

        </section>

    </div>
    <!-- Newsletter Banner -->
    <div role="banner" id="newsletter-banner">
      <span id="newsletter-subscribe-button" role="button">Subscribe</span>&nbsp;to our newsletter to receive updates about new features.
      <div id="newsletter-banner-close-button" role="button">X</div>
    </div>
    <!-- Newsletter Modal -->
    <dialog id="newsletter-modal">
      <div id="newsletter-header">
        <img id="newsletter-logo" src="../../_static/img/cellxGene-newsletter-logo.svg" />
        <div id="newsletter-close-button" role="button">X</div>
      </div>
      <div id="newsletter-content">
        <div id="newsletter-callout">Join Our Newsletter</div>
        <div id="newsletter-description">Get a quarterly email with the latest CELLxGENE features and data.</div>
        <!-- HubSpot Form target -->
        <div id="newsletter-form-container"></div>
      </div>

      <div id="newsletter-footnote">Unsubscribe at any time.</div>
    </dialog>
  </div>
  

  <script>
    // (thuang): 30 days
    const NEWSLETTER_BANNER_DISMISSED_TTL_MS = 30 * 24 * 60 * 60 * 1000;
    const NEWSLETTER_BANNER_DISMISSED_KEY = "newsletterBannerDismissed"

    var script = document.createElement('script');
    script.src = 'https://js.hsforms.net/forms/v2.js';
    script.defer = true;
    document.head.appendChild(script);

    // Run the code once the script is loaded
    script.onload = async function() {
      await hbspt.forms.create({
        region: "na1",
        portalId: "7272273",
        formId: "eb65b811-0451-414d-8304-7b9b6f468ce5",
        target: '#newsletter-form-container',
        onFormReady() {
          // get element by type "email"
          const emailInput = document.querySelector('#email-eb65b811-0451-414d-8304-7b9b6f468ce5');
          emailInput.setAttribute('placeholder', 'Enter email address');

          // remove the label element for emailInput
          const emailLabel = document.querySelector('#label-email-eb65b811-0451-414d-8304-7b9b6f468ce5');
          emailLabel.remove();
        },
        submitText: 'Subscribe',
      });
    };

    checkNewsletterBanner();

    document.querySelector('#newsletter-banner-close-button').addEventListener('click', () => {
      document.querySelector('#newsletter-banner').remove();
      localStorage.setItem(NEWSLETTER_BANNER_DISMISSED_KEY, Date.now());
    });

    const modal = document.querySelector('#newsletter-modal');

    document.querySelector('#newsletter-subscribe-button').addEventListener('click', () => {
      modal.showModal();
    });

    document.querySelector('#newsletter-close-button').addEventListener('click', () => {
      modal.close();
    });

    function checkNewsletterBanner() {
      /**
       * (thuang): Use LocalStorage to store dismissed state for 30 days
       * NOTE: Currently Census doc page doesn't share the same domain as the main site,
       * so dismissing the banner on the main site won't dismiss it on the Census doc page.
       * And vice versa.
       */
      const newsletterBannerDismissed = localStorage.getItem('newsletterBannerDismissed');

      if (newsletterBannerDismissed) {
        return;
      }

      if (newsletterBannerDismissed && Date.now() - newsletterBannerDismissed > NEWSLETTER_BANNER_DISMISSED_TTL_MS) {
        localStorage.removeItem(NEWSLETTER_BANNER_DISMISSED_KEY);
      }

      const newsletterBanner = document.querySelector('#newsletter-banner');

      if (!newsletterBannerDismissed) {
        newsletterBanner.style.display = 'flex';
      }
    }
  </script>

  
  
    
   

</body>
</html>