

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8">
  <meta name="generator" content="Docutils 0.19: https://docutils.sourceforge.io/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Exploring all data from a tissue &mdash; cellxgene-census  documentation</title>
  

  
  <link rel="icon" type="image/png" sizes="32x32" href="../../_static/img/favicon_32x32_v2.png"/>
  <link rel="icon" type="image/png" sizes="16x16" href="../../_static/img/favicon_16x16_v2.png"/>
  <link rel="preconnect" href="https://fonts.googleapis.com">
  <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
  <link href="https://fonts.googleapis.com/css2?family=Inter:wght@400;500;600;700&display=swap" rel="stylesheet">
  <link rel="stylesheet" href="../../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/css/custom.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/nbsphinx-code-cells.css" type="text/css" />

  
  
  
  

  
  <!--[if lt IE 9]>
    <script src="../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../../" src="../../_static/documentation_options.js"></script>
        <script src="../../_static/jquery.js"></script>
        <script src="../../_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script data-url_root="../../" id="documentation_options" src="../../_static/documentation_options.js"></script>
        <script src="../../_static/doctools.js"></script>
        <script src="../../_static/sphinx_highlight.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
        <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
        <script data-domain="chanzuckerberg.github.io/cellxgene-census" defer="defer" src="https://plausible.io/js/script.js"></script>
    
    <script type="text/javascript" src="../../_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="Exploring the Census Datasets table" href="../api_demo/census_datasets.html" />
    <link rel="prev" title="Learning about the CZ CELLxGENE Census" href="comp_bio_census_info.html" /> 
</head>

<body class="wy-body-for-nav">

  <div style="height: 100vh; overflow: hidden;">
     
    <!-- top 56px for navbar height -->
    <div class="navbar-cxg">
        <div style="padding: 0 16px; display: flex; height: inherit; align-items: center; justify-content: space-between;">
            <span style="display:flex; gap: 32px; align-items: center;">
                <a href="https://cellxgene.cziscience.com/">
                  <img src="../../_static/img/cellxgene-discover-logo.svg" />
                </a>
                <span class="navbar-cxg-nav-wrapper">
                  <span class="navbar-cxg-section">
                    <span class="navbar-cxg-nav-section-title">Application</span>
                    <span class="navbar-cxg-nav-item-container">
                      <span class="navbar-cxg-link">
                          <a class="navbar-cxg-anchor" href="https://cellxgene.cziscience.com/collections">Collections</a>
                      </span>
                      <span class="navbar-cxg-link">
                          <a class="navbar-cxg-anchor" href="https://cellxgene.cziscience.com/datasets">Datasets</a>
                      </span>
                      <span class="navbar-cxg-link">
                          <a class="navbar-cxg-anchor" href="https://cellxgene.cziscience.com/gene-expression">Gene Expression</a>
                      </span>
                      <span class="navbar-cxg-link">
                        <a class="navbar-cxg-anchor" href="https://cellxgene.cziscience.com/cellguide">Cell Guide</a>
                        <div style="height: 16px!important; display: flex;">
                          <span class="beta">BETA</span>
                        </div>
                      </span>
                    </span>
                  </span>
                  <hr class="navbar-divider"/>
                  <span class="navbar-cxg-section">
                    <span class="navbar-cxg-nav-section-title">Census</span>
                    <span class="navbar-cxg-nav-item-container">
                      <span class="navbar-cxg-link active-link">
                        <a class="navbar-cxg-anchor" href="/cellxgene-census/index.html">API</a>
                      </span>
                      <span class="navbar-cxg-link">
                        <a class="navbar-cxg-anchor" href="https://cellxgene.cziscience.com/census-models">Models</a>
                      </span>
                    </span>
                  </span>
                </span>
            </span>

            <span class="navbar-cxg-link">
              <a class="navbar-cxg-anchor" href="https://cellxgene.cziscience.com/docs">Help & Documentation</a>
            </span>
        </div>
    </div>
    <div style="width: 100%; height: 100%; overflow: auto; padding-top: 56px;">
        
        <!-- top 56px for navbar height -->
        <nav data-toggle="wy-nav-shift" class="wy-nav-side" style="top: 56px;" >
        <div class="wy-side-scroll">
            <div class="wy-side-nav-search" >
            

            
                <a href="../../index.html" class="icon icon-home"> cellxgene-census
            

            
            </a>

            
                
                
                <div class="version">
                    v1.14.1
                </div>
                
            

            
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

            
            </div>

            
            <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
            
                
                
                
                
                
                <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../../cellxgene_census_docsite_installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../cellxgene_census_docsite_quick_start.html">Quick start</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../articles.html">What‚Äôs new?</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../cellxgene_census_aws_open_data.html">Census in AWS ‚òÅÔ∏è</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../cellxgene_census_docsite_schema.html">Census data and schema</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../cellxgene_census_docsite_data_release_info.html">Census data releases</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../python-api.html">Python API</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../../examples.html">Python tutorials</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="../../examples.html#exporting-data">Exporting data</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../examples.html#new-using-integrated-embeddings-and-models">[NEW! üöÄ] Using integrated embeddings and models</a></li>
<li class="toctree-l2 current"><a class="reference internal" href="../../examples.html#understanding-census-data">Understanding Census data</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="../api_demo/census_duplicated_cells.html">Understanding and filtering out duplicate cells</a></li>
<li class="toctree-l3"><a class="reference internal" href="comp_bio_census_info.html">Learning about the CZ CELLxGENE Census</a></li>
<li class="toctree-l3 current"><a class="current reference internal" href="#">Exploring all data from a tissue</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#Learning-about-the-lung-data-in-the-Census">Learning about the lung data in the Census</a></li>
<li class="toctree-l4"><a class="reference internal" href="#Fetching-all-single-cell-human-lung-data-from-the-Census">Fetching all single-cell human lung data from the Census</a></li>
<li class="toctree-l4"><a class="reference internal" href="#Calculating-QC-metrics-of-the-lung-data">Calculating QC metrics of the lung data</a></li>
<li class="toctree-l4"><a class="reference internal" href="#Creating-a-normalized-expression-layer-and-embeddings">Creating a normalized expression layer and embeddings</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../api_demo/census_datasets.html">Exploring the Census Datasets table</a></li>
<li class="toctree-l3"><a class="reference internal" href="../api_demo/census_dataset_presence.html">Genes measured in each cell (dataset presence matrix)</a></li>
<li class="toctree-l3"><a class="reference internal" href="../api_demo/census_summary_cell_counts.html">Exploring pre-calculated summary cell counts</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../examples.html#analyzing-census-data">Analyzing Census data</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../examples.html#scalable-computing">Scalable computing</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../examples.html#scalable-machine-learning">Scalable machine learning</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference external" href="https://chanzuckerberg.github.io/cellxgene-census/r/index.html">R API &amp; tutorials</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../cellxgene_census_docsite_FAQ.html">FAQ</a></li>
</ul>

                
            
            </div>
            
        </div>
        </nav>

        <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

        
        <nav class="wy-nav-top" aria-label="top navigation">
            
            <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
            <a href="../../index.html">cellxgene-census</a>
            
        </nav>


        <div class="wy-nav-content">
            
            <div class="rst-content">
            
            <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../../examples.html">Python tutorials</a></li>
      <li class="breadcrumb-item active">Exploring all data from a tissue</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../../_sources/notebooks/analysis_demo/comp_bio_explore_and_load_lung_data.ipynb.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
            <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
            <div itemprop="articleBody">
                
  <section id="Exploring-all-data-from-a-tissue">
<h1>Exploring all data from a tissue<a class="headerlink" href="#Exploring-all-data-from-a-tissue" title="Permalink to this heading">ÔÉÅ</a></h1>
<p>This tutorial provides a series of examples for how to explore and query the Census in the context of a single tissue, lung. We will summarize cell and gene metadata, then fetch the single-cell expression counts and perform some basic data explorations via <a class="reference external" href="https://scanpy.readthedocs.io/en/stable/">Scanpy</a></p>
<p><strong>Contents</strong></p>
<ol class="arabic simple">
<li><p>Learning about the human lung data.</p>
<ol class="arabic simple">
<li><p>Learning about cells of the lung.</p></li>
<li><p>Learning about genes of the lung .</p></li>
</ol>
</li>
<li><p>Fetching all single-cell human lung data from the Census.</p></li>
<li><p>Calculating QC metrics of the lung data.</p></li>
<li><p>Creating a normalized expression layer and embeddings.</p></li>
</ol>
<p>‚ö†Ô∏è Note that the Census RNA data includes duplicate cells present across multiple datasets. Duplicate cells can be filtered in or out using the cell metadata variable <code class="docutils literal notranslate"><span class="pre">is_primary_data</span></code> which is described in the <a class="reference external" href="https://github.com/chanzuckerberg/cellxgene-census/blob/main/docs/cellxgene_census_schema.md#repeated-data">Census schema</a>.</p>
<section id="Learning-about-the-lung-data-in-the-Census">
<h2>Learning about the lung data in the Census<a class="headerlink" href="#Learning-about-the-lung-data-in-the-Census" title="Permalink to this heading">ÔÉÅ</a></h2>
<p>First we will open the Census. If you are not familiar with the basics of the Census API you should take a look at notebook <a class="reference external" href="https://cellxgene-census.readthedocs.io/en/latest/notebooks/analysis_demo/comp_bio_census_info.html">Learning about the CZ CELLxGENE Census</a></p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">cellxgene_census</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">scanpy</span> <span class="k">as</span> <span class="nn">sc</span>

<span class="n">census</span> <span class="o">=</span> <span class="n">cellxgene_census</span><span class="o">.</span><span class="n">open_soma</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
The &#34;stable&#34; release is currently 2023-07-25. Specify &#39;census_version=&#34;2023-07-25&#34;&#39; in future calls to open_soma() to ensure data consistency.
</pre></div></div>
</div>
<p>Let‚Äôs first take a look at the number of cells from human lung:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">summary_table</span> <span class="o">=</span> <span class="n">census</span><span class="p">[</span><span class="s2">&quot;census_info&quot;</span><span class="p">][</span><span class="s2">&quot;summary_cell_counts&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">read</span><span class="p">()</span><span class="o">.</span><span class="n">concat</span><span class="p">()</span><span class="o">.</span><span class="n">to_pandas</span><span class="p">()</span>

<span class="n">summary_table</span><span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="s2">&quot;organism == &#39;Homo sapiens&#39; &amp; category == &#39;tissue_general&#39; &amp; label ==&#39;lung&#39;&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>soma_joinid</th>
      <th>organism</th>
      <th>category</th>
      <th>ontology_term_id</th>
      <th>unique_cell_count</th>
      <th>total_cell_count</th>
      <th>label</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>980</th>
      <td>980</td>
      <td>Homo sapiens</td>
      <td>tissue_general</td>
      <td>UBERON:0002048</td>
      <td>2907156</td>
      <td>6011592</td>
      <td>lung</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<p>There you can see the total of cells of under <code class="docutils literal notranslate"><span class="pre">total_cell_count</span></code> and the unique number cells under <code class="docutils literal notranslate"><span class="pre">unique_cell_count</span></code> (i.e. after removing cells that were included in multiple datasets).</p>
<p>Let‚Äôs now take a look at the cell and gene information of this slice of the Census.</p>
<section id="Learning-about-cells-of-lung-data">
<h3>Learning about cells of lung data<a class="headerlink" href="#Learning-about-cells-of-lung-data" title="Permalink to this heading">ÔÉÅ</a></h3>
<p>Let‚Äôs load the cell metadata for all lung cells and select only the unique cells using <code class="docutils literal notranslate"><span class="pre">is_primary_data</span></code>.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">lung_obs</span> <span class="o">=</span> <span class="p">(</span>
    <span class="n">census</span><span class="p">[</span><span class="s2">&quot;census_data&quot;</span><span class="p">][</span><span class="s2">&quot;homo_sapiens&quot;</span><span class="p">]</span>
    <span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">value_filter</span><span class="o">=</span><span class="s2">&quot;tissue_general == &#39;lung&#39; and is_primary_data == True&quot;</span><span class="p">)</span>
    <span class="o">.</span><span class="n">concat</span><span class="p">()</span>
    <span class="o">.</span><span class="n">to_pandas</span><span class="p">()</span>
<span class="p">)</span>
<span class="n">lung_obs</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>soma_joinid</th>
      <th>dataset_id</th>
      <th>assay</th>
      <th>assay_ontology_term_id</th>
      <th>cell_type</th>
      <th>cell_type_ontology_term_id</th>
      <th>development_stage</th>
      <th>development_stage_ontology_term_id</th>
      <th>disease</th>
      <th>disease_ontology_term_id</th>
      <th>...</th>
      <th>is_primary_data</th>
      <th>self_reported_ethnicity</th>
      <th>self_reported_ethnicity_ontology_term_id</th>
      <th>sex</th>
      <th>sex_ontology_term_id</th>
      <th>suspension_type</th>
      <th>tissue</th>
      <th>tissue_ontology_term_id</th>
      <th>tissue_general</th>
      <th>tissue_general_ontology_term_id</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>5945423</td>
      <td>9f222629-9e39-47d0-b83f-e08d610c7479</td>
      <td>10x 3' v2</td>
      <td>EFO:0009899</td>
      <td>native cell</td>
      <td>CL:0000003</td>
      <td>unknown</td>
      <td>unknown</td>
      <td>normal</td>
      <td>PATO:0000461</td>
      <td>...</td>
      <td>True</td>
      <td>unknown</td>
      <td>unknown</td>
      <td>unknown</td>
      <td>unknown</td>
      <td>cell</td>
      <td>lung</td>
      <td>UBERON:0002048</td>
      <td>lung</td>
      <td>UBERON:0002048</td>
    </tr>
    <tr>
      <th>1</th>
      <td>5945426</td>
      <td>9f222629-9e39-47d0-b83f-e08d610c7479</td>
      <td>Drop-seq</td>
      <td>EFO:0008722</td>
      <td>ciliated columnar cell of tracheobronchial tree</td>
      <td>CL:0002145</td>
      <td>57-year-old human stage</td>
      <td>HsapDv:0000151</td>
      <td>pulmonary fibrosis</td>
      <td>MONDO:0002771</td>
      <td>...</td>
      <td>True</td>
      <td>unknown</td>
      <td>unknown</td>
      <td>male</td>
      <td>PATO:0000384</td>
      <td>cell</td>
      <td>lung</td>
      <td>UBERON:0002048</td>
      <td>lung</td>
      <td>UBERON:0002048</td>
    </tr>
    <tr>
      <th>2</th>
      <td>5945428</td>
      <td>9f222629-9e39-47d0-b83f-e08d610c7479</td>
      <td>10x 3' transcription profiling</td>
      <td>EFO:0030003</td>
      <td>CD8-positive, alpha-beta T cell</td>
      <td>CL:0000625</td>
      <td>unknown</td>
      <td>unknown</td>
      <td>squamous cell lung carcinoma</td>
      <td>MONDO:0005097</td>
      <td>...</td>
      <td>True</td>
      <td>unknown</td>
      <td>unknown</td>
      <td>unknown</td>
      <td>unknown</td>
      <td>cell</td>
      <td>lung</td>
      <td>UBERON:0002048</td>
      <td>lung</td>
      <td>UBERON:0002048</td>
    </tr>
    <tr>
      <th>3</th>
      <td>5945432</td>
      <td>9f222629-9e39-47d0-b83f-e08d610c7479</td>
      <td>10x 3' v2</td>
      <td>EFO:0009899</td>
      <td>CD4-positive, alpha-beta T cell</td>
      <td>CL:0000624</td>
      <td>unknown</td>
      <td>unknown</td>
      <td>lung adenocarcinoma</td>
      <td>MONDO:0005061</td>
      <td>...</td>
      <td>True</td>
      <td>unknown</td>
      <td>unknown</td>
      <td>unknown</td>
      <td>unknown</td>
      <td>cell</td>
      <td>lung</td>
      <td>UBERON:0002048</td>
      <td>lung</td>
      <td>UBERON:0002048</td>
    </tr>
    <tr>
      <th>4</th>
      <td>5945441</td>
      <td>9f222629-9e39-47d0-b83f-e08d610c7479</td>
      <td>10x 3' v2</td>
      <td>EFO:0009899</td>
      <td>CD8-positive, alpha-beta T cell</td>
      <td>CL:0000625</td>
      <td>unknown</td>
      <td>unknown</td>
      <td>lung adenocarcinoma</td>
      <td>MONDO:0005061</td>
      <td>...</td>
      <td>True</td>
      <td>unknown</td>
      <td>unknown</td>
      <td>unknown</td>
      <td>unknown</td>
      <td>cell</td>
      <td>lung</td>
      <td>UBERON:0002048</td>
      <td>lung</td>
      <td>UBERON:0002048</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>2907151</th>
      <td>56400868</td>
      <td>8c42cfd0-0b0a-46d5-910c-fc833d83c45e</td>
      <td>10x 3' v2</td>
      <td>EFO:0009899</td>
      <td>pericyte</td>
      <td>CL:0000669</td>
      <td>51-year-old human stage</td>
      <td>HsapDv:0000145</td>
      <td>normal</td>
      <td>PATO:0000461</td>
      <td>...</td>
      <td>True</td>
      <td>unknown</td>
      <td>unknown</td>
      <td>female</td>
      <td>PATO:0000383</td>
      <td>cell</td>
      <td>lung</td>
      <td>UBERON:0002048</td>
      <td>lung</td>
      <td>UBERON:0002048</td>
    </tr>
    <tr>
      <th>2907152</th>
      <td>56400869</td>
      <td>8c42cfd0-0b0a-46d5-910c-fc833d83c45e</td>
      <td>10x 3' v2</td>
      <td>EFO:0009899</td>
      <td>pericyte</td>
      <td>CL:0000669</td>
      <td>51-year-old human stage</td>
      <td>HsapDv:0000145</td>
      <td>normal</td>
      <td>PATO:0000461</td>
      <td>...</td>
      <td>True</td>
      <td>unknown</td>
      <td>unknown</td>
      <td>female</td>
      <td>PATO:0000383</td>
      <td>cell</td>
      <td>lung</td>
      <td>UBERON:0002048</td>
      <td>lung</td>
      <td>UBERON:0002048</td>
    </tr>
    <tr>
      <th>2907153</th>
      <td>56400870</td>
      <td>8c42cfd0-0b0a-46d5-910c-fc833d83c45e</td>
      <td>10x 3' v2</td>
      <td>EFO:0009899</td>
      <td>pericyte</td>
      <td>CL:0000669</td>
      <td>51-year-old human stage</td>
      <td>HsapDv:0000145</td>
      <td>normal</td>
      <td>PATO:0000461</td>
      <td>...</td>
      <td>True</td>
      <td>unknown</td>
      <td>unknown</td>
      <td>female</td>
      <td>PATO:0000383</td>
      <td>cell</td>
      <td>lung</td>
      <td>UBERON:0002048</td>
      <td>lung</td>
      <td>UBERON:0002048</td>
    </tr>
    <tr>
      <th>2907154</th>
      <td>56400871</td>
      <td>8c42cfd0-0b0a-46d5-910c-fc833d83c45e</td>
      <td>10x 3' v2</td>
      <td>EFO:0009899</td>
      <td>pericyte</td>
      <td>CL:0000669</td>
      <td>51-year-old human stage</td>
      <td>HsapDv:0000145</td>
      <td>normal</td>
      <td>PATO:0000461</td>
      <td>...</td>
      <td>True</td>
      <td>unknown</td>
      <td>unknown</td>
      <td>female</td>
      <td>PATO:0000383</td>
      <td>cell</td>
      <td>lung</td>
      <td>UBERON:0002048</td>
      <td>lung</td>
      <td>UBERON:0002048</td>
    </tr>
    <tr>
      <th>2907155</th>
      <td>56400872</td>
      <td>8c42cfd0-0b0a-46d5-910c-fc833d83c45e</td>
      <td>10x 3' v2</td>
      <td>EFO:0009899</td>
      <td>pericyte</td>
      <td>CL:0000669</td>
      <td>51-year-old human stage</td>
      <td>HsapDv:0000145</td>
      <td>normal</td>
      <td>PATO:0000461</td>
      <td>...</td>
      <td>True</td>
      <td>unknown</td>
      <td>unknown</td>
      <td>female</td>
      <td>PATO:0000383</td>
      <td>cell</td>
      <td>lung</td>
      <td>UBERON:0002048</td>
      <td>lung</td>
      <td>UBERON:0002048</td>
    </tr>
  </tbody>
</table>
<p>2907156 rows √ó 21 columns</p>
</div></div>
</div>
<p>You can see that the number or rows represents the total number of unique lung cells in the Census. Now let‚Äôs take a deeper dive into the characteristics of these cells.</p>
<section id="Datasets">
<h4>Datasets<a class="headerlink" href="#Datasets" title="Permalink to this heading">ÔÉÅ</a></h4>
<p>First let‚Äôs start by looking at what are the datasets and collections from <a class="reference external" href="https://cellxgene.cziscience.com/collections">CELLxGENE Discover</a> contributing to lung. For this we will use the dataset table at <code class="docutils literal notranslate"><span class="pre">census[&quot;census-info&quot;][&quot;datasets&quot;]</span></code> that contains metadata of all datasets used to build this Census.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">census_datasets</span> <span class="o">=</span> <span class="p">(</span>
    <span class="n">census</span><span class="p">[</span><span class="s2">&quot;census_info&quot;</span><span class="p">][</span><span class="s2">&quot;datasets&quot;</span><span class="p">]</span>
    <span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">column_names</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;collection_name&quot;</span><span class="p">,</span> <span class="s2">&quot;dataset_title&quot;</span><span class="p">,</span> <span class="s2">&quot;dataset_id&quot;</span><span class="p">,</span> <span class="s2">&quot;soma_joinid&quot;</span><span class="p">])</span>
    <span class="o">.</span><span class="n">concat</span><span class="p">()</span>
    <span class="o">.</span><span class="n">to_pandas</span><span class="p">()</span>
<span class="p">)</span>
<span class="n">census_datasets</span> <span class="o">=</span> <span class="n">census_datasets</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s2">&quot;dataset_id&quot;</span><span class="p">)</span>
<span class="n">census_datasets</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>collection_name</th>
      <th>dataset_title</th>
      <th>soma_joinid</th>
    </tr>
    <tr>
      <th>dataset_id</th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>f171db61-e57e-4535-a06a-35d8b6ef8f2b</th>
      <td>Spatial multiomics map of trophoblast developm...</td>
      <td>donor_p13_trophoblasts</td>
      <td>0</td>
    </tr>
    <tr>
      <th>ecf2e08e-2032-4a9e-b466-b65b395f4a02</th>
      <td>Spatial multiomics map of trophoblast developm...</td>
      <td>All donors trophoblasts</td>
      <td>1</td>
    </tr>
    <tr>
      <th>74cff64f-9da9-4b2a-9b3b-8a04a1598040</th>
      <td>Spatial multiomics map of trophoblast developm...</td>
      <td>All donors all cell states (in vivo)</td>
      <td>2</td>
    </tr>
    <tr>
      <th>5af90777-6760-4003-9dba-8f945fec6fdf</th>
      <td>Mapping single-cell transcriptomes in the intr...</td>
      <td>Single-cell transcriptomic datasets of Renal c...</td>
      <td>3</td>
    </tr>
    <tr>
      <th>bd65a70f-b274-4133-b9dd-0d1431b6af34</th>
      <td>Single-cell sequencing links multiregional imm...</td>
      <td>Single-cell sequencing links multiregional imm...</td>
      <td>4</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>f9ad5649-f372-43e1-a3a8-423383e5a8a2</th>
      <td>Molecular characterization of selectively vuln...</td>
      <td>Molecular characterization of selectively vuln...</td>
      <td>588</td>
    </tr>
    <tr>
      <th>456e8b9b-f872-488b-871d-94534090a865</th>
      <td>Single-cell atlas of peripheral immune respons...</td>
      <td>Single-cell atlas of peripheral immune respons...</td>
      <td>589</td>
    </tr>
    <tr>
      <th>2adb1f8a-a6b1-4909-8ee8-484814e2d4bf</th>
      <td>Construction of a human cell landscape at sing...</td>
      <td>Construction of a human cell landscape at sing...</td>
      <td>590</td>
    </tr>
    <tr>
      <th>e04daea4-4412-45b5-989e-76a9be070a89</th>
      <td>A molecular cell atlas of the human lung from ...</td>
      <td>Krasnow Lab Human Lung Cell Atlas, Smart-seq2</td>
      <td>591</td>
    </tr>
    <tr>
      <th>8c42cfd0-0b0a-46d5-910c-fc833d83c45e</th>
      <td>A molecular cell atlas of the human lung from ...</td>
      <td>Krasnow Lab Human Lung Cell Atlas, 10X</td>
      <td>592</td>
    </tr>
  </tbody>
</table>
<p>593 rows √ó 3 columns</p>
</div></div>
</div>
<p>The <code class="docutils literal notranslate"><span class="pre">obs</span></code> cell metadata <code class="docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code> contains a column <code class="docutils literal notranslate"><span class="pre">dataset_id</span></code> that can be used for joining to the <code class="docutils literal notranslate"><span class="pre">census_dataset</span></code> <code class="docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code> we just created.</p>
<p>So let‚Äôs take a look at the cell counts per <code class="docutils literal notranslate"><span class="pre">dataset_id</span></code> of the lung slice and then join to the dataset table to append the human-readable labels.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">dataset_cell_counts</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">lung_obs</span><span class="p">[[</span><span class="s2">&quot;dataset_id&quot;</span><span class="p">]]</span><span class="o">.</span><span class="n">value_counts</span><span class="p">())</span>
<span class="n">dataset_cell_counts</span> <span class="o">=</span> <span class="n">dataset_cell_counts</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">{</span><span class="mi">0</span><span class="p">:</span> <span class="s2">&quot;cell_counts&quot;</span><span class="p">})</span>
<span class="n">dataset_cell_counts</span> <span class="o">=</span> <span class="n">dataset_cell_counts</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">census_datasets</span><span class="p">,</span> <span class="n">on</span><span class="o">=</span><span class="s2">&quot;dataset_id&quot;</span><span class="p">)</span>

<span class="n">dataset_cell_counts</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>count</th>
      <th>collection_name</th>
      <th>dataset_title</th>
      <th>soma_joinid</th>
    </tr>
    <tr>
      <th>dataset_id</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>1e6a6ef9-7ec9-4c90-bbfb-2ad3c3165fd1</th>
      <td>1028006</td>
      <td>High-resolution single-cell atlas reveals dive...</td>
      <td>The single-cell lung cancer atlas (LuCA) -- ex...</td>
      <td>314</td>
    </tr>
    <tr>
      <th>9f222629-9e39-47d0-b83f-e08d610c7479</th>
      <td>784630</td>
      <td>The integrated Human Lung Cell Atlas</td>
      <td>An integrated cell atlas of the human lung in ...</td>
      <td>56</td>
    </tr>
    <tr>
      <th>f7c1c579-2dc0-47e2-ba19-8165c5a0e353</th>
      <td>217738</td>
      <td>A human cell atlas of fetal gene expression</td>
      <td>Survey of human embryonic development</td>
      <td>483</td>
    </tr>
    <tr>
      <th>d8da613f-e681-4c69-b463-e94f5e66847f</th>
      <td>116313</td>
      <td>A molecular single-cell lung atlas of lethal C...</td>
      <td>A molecular single-cell lung atlas of lethal C...</td>
      <td>80</td>
    </tr>
    <tr>
      <th>576f193c-75d0-4a11-bd25-8676587e6dc2</th>
      <td>90384</td>
      <td>HTAN MSK - Single cell profiling reveals novel...</td>
      <td>Combined samples</td>
      <td>377</td>
    </tr>
    <tr>
      <th>d41f45c1-1b7b-4573-a998-ac5c5acb1647</th>
      <td>82991</td>
      <td>HTAN MSK - Transcriptional connectivity of reg...</td>
      <td>Transcriptional connectivity of regulatory T c...</td>
      <td>58</td>
    </tr>
    <tr>
      <th>3dc61ca1-ce40-46b6-8337-f27260fd9a03</th>
      <td>71752</td>
      <td>A human fetal lung cell atlas uncovers proxima...</td>
      <td>All cells</td>
      <td>325</td>
    </tr>
    <tr>
      <th>8c42cfd0-0b0a-46d5-910c-fc833d83c45e</th>
      <td>60993</td>
      <td>A molecular cell atlas of the human lung from ...</td>
      <td>Krasnow Lab Human Lung Cell Atlas, 10X</td>
      <td>592</td>
    </tr>
    <tr>
      <th>2672b679-8048-4f5e-9786-f1b196ccfd08</th>
      <td>57019</td>
      <td>scRNA-seq assessment of the human lung, spleen...</td>
      <td>Lung Parenchyma</td>
      <td>416</td>
    </tr>
    <tr>
      <th>9dbab10c-118d-496b-966a-67f1763a6b7d</th>
      <td>49014</td>
      <td>COVID-19 immune features revealed by a large-s...</td>
      <td>Large-scale single-cell analysis reveals criti...</td>
      <td>482</td>
    </tr>
    <tr>
      <th>9968be68-ab65-4a38-9e1a-c9b6abece194</th>
      <td>47909</td>
      <td>Charting human development using a multi-endod...</td>
      <td>Developing Human Atlas</td>
      <td>78</td>
    </tr>
    <tr>
      <th>3de0ad6d-4378-4f62-b37b-ec0b75a50d94</th>
      <td>46500</td>
      <td>LungMAP ‚Äî Human data from a broad age healthy ...</td>
      <td>Single-cell multiomic profiling of human lungs...</td>
      <td>456</td>
    </tr>
    <tr>
      <th>2f132ec9-24b5-422f-9be0-ccef03b4fe28</th>
      <td>39778</td>
      <td>SARS-CoV-2 receptor ACE2 and TMPRSS2 are prima...</td>
      <td>Lung</td>
      <td>312</td>
    </tr>
    <tr>
      <th>1e5bd3b8-6a0e-4959-8d69-cafed30fe814</th>
      <td>35699</td>
      <td>Emphysema Cell Atlas</td>
      <td>immune cells</td>
      <td>130</td>
    </tr>
    <tr>
      <th>53d208b0-2cfd-4366-9866-c3c6114081bc</th>
      <td>35682</td>
      <td>Tabula Sapiens</td>
      <td>Tabula Sapiens - All Cells</td>
      <td>475</td>
    </tr>
    <tr>
      <th>1b9d8702-5af8-4142-85ed-020eb06ec4f6</th>
      <td>35419</td>
      <td>Cross-tissue immune cell analysis reveals tiss...</td>
      <td>Global</td>
      <td>411</td>
    </tr>
    <tr>
      <th>4ed927e9-c099-49af-b8ce-a2652d069333</th>
      <td>35284</td>
      <td>Single-nucleus cross-tissue molecular referenc...</td>
      <td>Single-nucleus cross-tissue molecular referenc...</td>
      <td>367</td>
    </tr>
    <tr>
      <th>2adb1f8a-a6b1-4909-8ee8-484814e2d4bf</th>
      <td>33698</td>
      <td>Construction of a human cell landscape at sing...</td>
      <td>Construction of a human cell landscape at sing...</td>
      <td>590</td>
    </tr>
    <tr>
      <th>4b6af54a-4a21-46e0-bc8d-673c0561a836</th>
      <td>18386</td>
      <td>Emphysema Cell Atlas</td>
      <td>non-immune cells</td>
      <td>128</td>
    </tr>
    <tr>
      <th>01209dce-3575-4bed-b1df-129f57fbc031</th>
      <td>11059</td>
      <td>Single-cell transcriptomics of human T cells r...</td>
      <td>Single-cell transcriptomics of human T cells r...</td>
      <td>531</td>
    </tr>
    <tr>
      <th>e04daea4-4412-45b5-989e-76a9be070a89</th>
      <td>8657</td>
      <td>A molecular cell atlas of the human lung from ...</td>
      <td>Krasnow Lab Human Lung Cell Atlas, Smart-seq2</td>
      <td>591</td>
    </tr>
    <tr>
      <th>f9846bb4-784d-4582-92c1-3f279e4c6f0c</th>
      <td>176</td>
      <td>A human fetal lung cell atlas uncovers proxima...</td>
      <td>Fibroblast and smooth muscle</td>
      <td>317</td>
    </tr>
    <tr>
      <th>f64e1be1-de15-4d27-8da4-82225cd4c035</th>
      <td>55</td>
      <td>HTAN MSK - Single cell profiling reveals novel...</td>
      <td>Immune cells</td>
      <td>370</td>
    </tr>
    <tr>
      <th>810ac45f-8969-4698-b42c-652f802f75c2</th>
      <td>10</td>
      <td>A human fetal lung cell atlas uncovers proxima...</td>
      <td>Endothelium</td>
      <td>320</td>
    </tr>
    <tr>
      <th>0ba16f4b-cb87-4fa3-9363-19fc51eec6e7</th>
      <td>4</td>
      <td>A human fetal lung cell atlas uncovers proxima...</td>
      <td>Myeloid</td>
      <td>326</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<p>These are all the datasets lung cells whose counts are reprensented in the column <code class="docutils literal notranslate"><span class="pre">cell_counts</span></code>. The top collections with lung data are:</p>
<ol class="arabic simple">
<li><p><a class="reference external" href="https://cellxgene.cziscience.com/collections/6f6d381a-7701-4781-935c-db10d30de293">The integrated Human Lung Cell Atlas</a>.</p></li>
<li><p><a class="reference external" href="https://cellxgene.cziscience.com/collections/c114c20f-1ef4-49a5-9c2e-d965787fb90c">A human cell atlas of fetal gene expression</a>.</p></li>
<li><p><a class="reference external" href="https://cellxgene.cziscience.com/collections/edb893ee-4066-4128-9aec-5eb2b03f8287">High-resolution single-cell atlas reveals diversity and plasticity of tumor-associated neutrophils in non-small cell lung cancer</a>.</p></li>
<li><p><a class="reference external" href="https://cellxgene.cziscience.com/collections/62e8f058-9c37-48bc-9200-e767f318a8ec">HTAN MSK - Single cell profiling reveals novel tumor and myeloid subpopulations in small cell lung cancer</a>.</p></li>
<li><p><a class="reference external" href="https://cellxgene.cziscience.com/collections/2d2e2acd-dade-489f-a2da-6c11aa654028">A human fetal lung cell atlas uncovers proximal-distal gradients of differentiation and key regulators of epithelial fates.</a>.</p></li>
</ol>
</section>
<section id="Assays">
<h4>Assays<a class="headerlink" href="#Assays" title="Permalink to this heading">ÔÉÅ</a></h4>
<p>Let‚Äôs use similar logic to take a look at all the assays available for human lung data. This tells us that most assays are from 10x technologies and sci-RNA-seq.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">lung_obs</span><span class="p">[[</span><span class="s2">&quot;assay&quot;</span><span class="p">]]</span><span class="o">.</span><span class="n">value_counts</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
assay
10x 3&#39; v2                                   1236968
10x 3&#39; v3                                    702074
10x 5&#39; v1                                    262323
sci-RNA-seq                                  217738
BD Rhapsody Whole Transcriptome Analysis     122902
10x 3&#39; transcription profiling                97432
Drop-seq                                      65220
single cell library construction              58981
10x 5&#39; v2                                     41852
microwell-seq                                 33698
Smart-seq2                                    25662
inDrop                                        25652
10x 3&#39; v1                                      8638
Seq-Well                                       8016
Name: count, dtype: int64
</pre></div></div>
</div>
</section>
<section id="Disease">
<h4>Disease<a class="headerlink" href="#Disease" title="Permalink to this heading">ÔÉÅ</a></h4>
<p>And now let‚Äôs take a look at diseased cell counts, with <code class="docutils literal notranslate"><span class="pre">normal</span></code> indicating non-diseased cells.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">lung_obs</span><span class="p">[[</span><span class="s2">&quot;disease&quot;</span><span class="p">]]</span><span class="o">.</span><span class="n">value_counts</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
disease
normal                                   1164084
lung adenocarcinoma                       772120
COVID-19                                  331019
squamous cell lung carcinoma              209675
non-small cell lung carcinoma             120796
chronic obstructive pulmonary disease      55254
pulmonary fibrosis                         51343
interstitial lung disease                  45714
pneumonia                                  31923
pulmonary emphysema                        31792
small cell lung carcinoma                  31540
lung large cell carcinoma                  21167
cystic fibrosis                            17590
lymphangioleiomyomatosis                   12374
pleomorphic carcinoma                      10765
Name: count, dtype: int64
</pre></div></div>
</div>
</section>
<section id="Sex">
<h4>Sex<a class="headerlink" href="#Sex" title="Permalink to this heading">ÔÉÅ</a></h4>
<p>There doesn‚Äôt seem to be strong biases for sex.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">lung_obs</span><span class="p">[[</span><span class="s2">&quot;sex&quot;</span><span class="p">]]</span><span class="o">.</span><span class="n">value_counts</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
sex
male       1402565
female     1122990
unknown     381601
Name: count, dtype: int64
</pre></div></div>
</div>
</section>
<section id="Cell-vs-nucleus">
<h4>Cell vs nucleus<a class="headerlink" href="#Cell-vs-nucleus" title="Permalink to this heading">ÔÉÅ</a></h4>
<p>The majority of data are from cells and not nucleus.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">lung_obs</span><span class="p">[[</span><span class="s2">&quot;suspension_type&quot;</span><span class="p">]]</span><span class="o">.</span><span class="n">value_counts</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
suspension_type
cell               2468587
nucleus             438569
Name: count, dtype: int64
</pre></div></div>
</div>
</section>
<section id="Cell-types">
<h4>Cell types<a class="headerlink" href="#Cell-types" title="Permalink to this heading">ÔÉÅ</a></h4>
<p>Let‚Äôs take a look at the counts of the top 20 cell types.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">lung_obs</span><span class="p">[[</span><span class="s2">&quot;cell_type&quot;</span><span class="p">]]</span><span class="o">.</span><span class="n">value_counts</span><span class="p">()</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">20</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
cell_type
alveolar macrophage                                291507
native cell                                        263362
CD4-positive, alpha-beta T cell                    211456
CD8-positive, alpha-beta T cell                    189471
macrophage                                         154415
type II pneumocyte                                 128463
epithelial cell of lower respiratory tract         105090
classical monocyte                                 102303
natural killer cell                                 95953
T cell                                              92846
stromal cell                                        87714
B cell                                              81125
malignant cell                                      75917
plasma cell                                         64551
epithelial cell                                     59353
fibroblast                                          45305
capillary endothelial cell                          39416
regulatory T cell                                   36381
ciliated columnar cell of tracheobronchial tree     36049
epithelial cell of lung                             35467
Name: count, dtype: int64
</pre></div></div>
</div>
</section>
<section id="Sub-tissues">
<h4>Sub-tissues<a class="headerlink" href="#Sub-tissues" title="Permalink to this heading">ÔÉÅ</a></h4>
<p>We can look at the original tissue annotations that were mapped to ‚Äúlung‚Äù.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">lung_obs</span><span class="p">[[</span><span class="s2">&quot;tissue&quot;</span><span class="p">]]</span><span class="o">.</span><span class="n">value_counts</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
tissue
lung                         2576327
lung parenchyma               147410
alveolus of lung               54085
lingula of left lung           35284
upper lobe of right lung       32099
lower lobe of left lung        17854
right lung                     12880
upper lobe of left lung        10113
left lung                       9276
lower lobe of right lung        7981
middle lobe of right lung       3847
Name: count, dtype: int64
</pre></div></div>
</div>
</section>
</section>
<section id="Learning-about-genes-of-lung-data">
<h3>Learning about genes of lung data<a class="headerlink" href="#Learning-about-genes-of-lung-data" title="Permalink to this heading">ÔÉÅ</a></h3>
<p>Let‚Äôs load the gene metadata of the Census.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">lung_var</span> <span class="o">=</span> <span class="n">census</span><span class="p">[</span><span class="s2">&quot;census_data&quot;</span><span class="p">][</span><span class="s2">&quot;homo_sapiens&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">ms</span><span class="p">[</span><span class="s2">&quot;RNA&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">read</span><span class="p">()</span><span class="o">.</span><span class="n">concat</span><span class="p">()</span><span class="o">.</span><span class="n">to_pandas</span><span class="p">()</span>
<span class="n">lung_var</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>soma_joinid</th>
      <th>feature_id</th>
      <th>feature_name</th>
      <th>feature_length</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>0</td>
      <td>ENSG00000121410</td>
      <td>A1BG</td>
      <td>3999</td>
    </tr>
    <tr>
      <th>1</th>
      <td>1</td>
      <td>ENSG00000268895</td>
      <td>A1BG-AS1</td>
      <td>3374</td>
    </tr>
    <tr>
      <th>2</th>
      <td>2</td>
      <td>ENSG00000148584</td>
      <td>A1CF</td>
      <td>9603</td>
    </tr>
    <tr>
      <th>3</th>
      <td>3</td>
      <td>ENSG00000175899</td>
      <td>A2M</td>
      <td>6318</td>
    </tr>
    <tr>
      <th>4</th>
      <td>4</td>
      <td>ENSG00000245105</td>
      <td>A2M-AS1</td>
      <td>2948</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>60659</th>
      <td>60659</td>
      <td>ENSG00000288719</td>
      <td>RP4-669P10.21</td>
      <td>4252</td>
    </tr>
    <tr>
      <th>60660</th>
      <td>60660</td>
      <td>ENSG00000288720</td>
      <td>RP11-852E15.3</td>
      <td>7007</td>
    </tr>
    <tr>
      <th>60661</th>
      <td>60661</td>
      <td>ENSG00000288721</td>
      <td>RP5-973N23.5</td>
      <td>7765</td>
    </tr>
    <tr>
      <th>60662</th>
      <td>60662</td>
      <td>ENSG00000288723</td>
      <td>RP11-553N16.6</td>
      <td>1015</td>
    </tr>
    <tr>
      <th>60663</th>
      <td>60663</td>
      <td>ENSG00000288724</td>
      <td>RP13-546I2.2</td>
      <td>625</td>
    </tr>
  </tbody>
</table>
<p>60664 rows √ó 4 columns</p>
</div></div>
</div>
<p>You can see the total number of genes represented by the number of rows. This number is actually misleading because it is the join of all genes in the Census. However we know that the lung data comes from a subset of datasets.</p>
<p>So let‚Äôs take a look at the number of genes that were measured in each of those datasets.</p>
<p>To accomplish this we can use the ‚Äúdataset presence matrix‚Äù at <code class="docutils literal notranslate"><span class="pre">census[&quot;census_data&quot;][&quot;homo_sapiens&quot;].ms[&quot;RNA&quot;][&quot;feature_dataset_presence_matrix&quot;]</span></code>. This is a boolean matrix <code class="docutils literal notranslate"><span class="pre">N</span> <span class="pre">x</span> <span class="pre">M</span></code> where <code class="docutils literal notranslate"><span class="pre">N</span></code> is the number of datasets and <code class="docutils literal notranslate"><span class="pre">M</span></code> is the number of genes in the Census.</p>
<p>So we can select the rows corresponding to the lung datasets and perform a row-wise sum.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[13]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">presence_matrix</span> <span class="o">=</span> <span class="n">cellxgene_census</span><span class="o">.</span><span class="n">get_presence_matrix</span><span class="p">(</span><span class="n">census</span><span class="p">,</span> <span class="s2">&quot;Homo sapiens&quot;</span><span class="p">,</span> <span class="s2">&quot;RNA&quot;</span><span class="p">)</span>
<span class="n">presence_matrix</span> <span class="o">=</span> <span class="n">presence_matrix</span><span class="p">[</span><span class="n">dataset_cell_counts</span><span class="o">.</span><span class="n">soma_joinid</span><span class="p">,</span> <span class="p">:]</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">presence_matrix</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">A1</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
array([17811, 50259, 44150, 34265, 22447, 23642, 26347, 20921, 24672,
       27705, 27243, 26323, 27181, 23203, 57042, 32610, 29620, 26454,
       23705, 38676, 47307, 23740, 22552, 20594, 19952], dtype=uint64)
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[15]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">genes_measured</span> <span class="o">=</span> <span class="n">presence_matrix</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">A1</span>
<span class="n">dataset_cell_counts</span><span class="p">[</span><span class="s2">&quot;genes_measured&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">genes_measured</span>
<span class="n">dataset_cell_counts</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[15]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>count</th>
      <th>collection_name</th>
      <th>dataset_title</th>
      <th>soma_joinid</th>
      <th>genes_measured</th>
    </tr>
    <tr>
      <th>dataset_id</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>1e6a6ef9-7ec9-4c90-bbfb-2ad3c3165fd1</th>
      <td>1028006</td>
      <td>High-resolution single-cell atlas reveals dive...</td>
      <td>The single-cell lung cancer atlas (LuCA) -- ex...</td>
      <td>314</td>
      <td>17811</td>
    </tr>
    <tr>
      <th>9f222629-9e39-47d0-b83f-e08d610c7479</th>
      <td>784630</td>
      <td>The integrated Human Lung Cell Atlas</td>
      <td>An integrated cell atlas of the human lung in ...</td>
      <td>56</td>
      <td>50259</td>
    </tr>
    <tr>
      <th>f7c1c579-2dc0-47e2-ba19-8165c5a0e353</th>
      <td>217738</td>
      <td>A human cell atlas of fetal gene expression</td>
      <td>Survey of human embryonic development</td>
      <td>483</td>
      <td>44150</td>
    </tr>
    <tr>
      <th>d8da613f-e681-4c69-b463-e94f5e66847f</th>
      <td>116313</td>
      <td>A molecular single-cell lung atlas of lethal C...</td>
      <td>A molecular single-cell lung atlas of lethal C...</td>
      <td>80</td>
      <td>34265</td>
    </tr>
    <tr>
      <th>576f193c-75d0-4a11-bd25-8676587e6dc2</th>
      <td>90384</td>
      <td>HTAN MSK - Single cell profiling reveals novel...</td>
      <td>Combined samples</td>
      <td>377</td>
      <td>22447</td>
    </tr>
    <tr>
      <th>d41f45c1-1b7b-4573-a998-ac5c5acb1647</th>
      <td>82991</td>
      <td>HTAN MSK - Transcriptional connectivity of reg...</td>
      <td>Transcriptional connectivity of regulatory T c...</td>
      <td>58</td>
      <td>23642</td>
    </tr>
    <tr>
      <th>3dc61ca1-ce40-46b6-8337-f27260fd9a03</th>
      <td>71752</td>
      <td>A human fetal lung cell atlas uncovers proxima...</td>
      <td>All cells</td>
      <td>325</td>
      <td>26347</td>
    </tr>
    <tr>
      <th>8c42cfd0-0b0a-46d5-910c-fc833d83c45e</th>
      <td>60993</td>
      <td>A molecular cell atlas of the human lung from ...</td>
      <td>Krasnow Lab Human Lung Cell Atlas, 10X</td>
      <td>592</td>
      <td>20921</td>
    </tr>
    <tr>
      <th>2672b679-8048-4f5e-9786-f1b196ccfd08</th>
      <td>57019</td>
      <td>scRNA-seq assessment of the human lung, spleen...</td>
      <td>Lung Parenchyma</td>
      <td>416</td>
      <td>24672</td>
    </tr>
    <tr>
      <th>9dbab10c-118d-496b-966a-67f1763a6b7d</th>
      <td>49014</td>
      <td>COVID-19 immune features revealed by a large-s...</td>
      <td>Large-scale single-cell analysis reveals criti...</td>
      <td>482</td>
      <td>27705</td>
    </tr>
    <tr>
      <th>9968be68-ab65-4a38-9e1a-c9b6abece194</th>
      <td>47909</td>
      <td>Charting human development using a multi-endod...</td>
      <td>Developing Human Atlas</td>
      <td>78</td>
      <td>27243</td>
    </tr>
    <tr>
      <th>3de0ad6d-4378-4f62-b37b-ec0b75a50d94</th>
      <td>46500</td>
      <td>LungMAP ‚Äî Human data from a broad age healthy ...</td>
      <td>Single-cell multiomic profiling of human lungs...</td>
      <td>456</td>
      <td>26323</td>
    </tr>
    <tr>
      <th>2f132ec9-24b5-422f-9be0-ccef03b4fe28</th>
      <td>39778</td>
      <td>SARS-CoV-2 receptor ACE2 and TMPRSS2 are prima...</td>
      <td>Lung</td>
      <td>312</td>
      <td>27181</td>
    </tr>
    <tr>
      <th>1e5bd3b8-6a0e-4959-8d69-cafed30fe814</th>
      <td>35699</td>
      <td>Emphysema Cell Atlas</td>
      <td>immune cells</td>
      <td>130</td>
      <td>23203</td>
    </tr>
    <tr>
      <th>53d208b0-2cfd-4366-9866-c3c6114081bc</th>
      <td>35682</td>
      <td>Tabula Sapiens</td>
      <td>Tabula Sapiens - All Cells</td>
      <td>475</td>
      <td>57042</td>
    </tr>
    <tr>
      <th>1b9d8702-5af8-4142-85ed-020eb06ec4f6</th>
      <td>35419</td>
      <td>Cross-tissue immune cell analysis reveals tiss...</td>
      <td>Global</td>
      <td>411</td>
      <td>32610</td>
    </tr>
    <tr>
      <th>4ed927e9-c099-49af-b8ce-a2652d069333</th>
      <td>35284</td>
      <td>Single-nucleus cross-tissue molecular referenc...</td>
      <td>Single-nucleus cross-tissue molecular referenc...</td>
      <td>367</td>
      <td>29620</td>
    </tr>
    <tr>
      <th>2adb1f8a-a6b1-4909-8ee8-484814e2d4bf</th>
      <td>33698</td>
      <td>Construction of a human cell landscape at sing...</td>
      <td>Construction of a human cell landscape at sing...</td>
      <td>590</td>
      <td>26454</td>
    </tr>
    <tr>
      <th>4b6af54a-4a21-46e0-bc8d-673c0561a836</th>
      <td>18386</td>
      <td>Emphysema Cell Atlas</td>
      <td>non-immune cells</td>
      <td>128</td>
      <td>23705</td>
    </tr>
    <tr>
      <th>01209dce-3575-4bed-b1df-129f57fbc031</th>
      <td>11059</td>
      <td>Single-cell transcriptomics of human T cells r...</td>
      <td>Single-cell transcriptomics of human T cells r...</td>
      <td>531</td>
      <td>38676</td>
    </tr>
    <tr>
      <th>e04daea4-4412-45b5-989e-76a9be070a89</th>
      <td>8657</td>
      <td>A molecular cell atlas of the human lung from ...</td>
      <td>Krasnow Lab Human Lung Cell Atlas, Smart-seq2</td>
      <td>591</td>
      <td>47307</td>
    </tr>
    <tr>
      <th>f9846bb4-784d-4582-92c1-3f279e4c6f0c</th>
      <td>176</td>
      <td>A human fetal lung cell atlas uncovers proxima...</td>
      <td>Fibroblast and smooth muscle</td>
      <td>317</td>
      <td>23740</td>
    </tr>
    <tr>
      <th>f64e1be1-de15-4d27-8da4-82225cd4c035</th>
      <td>55</td>
      <td>HTAN MSK - Single cell profiling reveals novel...</td>
      <td>Immune cells</td>
      <td>370</td>
      <td>22552</td>
    </tr>
    <tr>
      <th>810ac45f-8969-4698-b42c-652f802f75c2</th>
      <td>10</td>
      <td>A human fetal lung cell atlas uncovers proxima...</td>
      <td>Endothelium</td>
      <td>320</td>
      <td>20594</td>
    </tr>
    <tr>
      <th>0ba16f4b-cb87-4fa3-9363-19fc51eec6e7</th>
      <td>4</td>
      <td>A human fetal lung cell atlas uncovers proxima...</td>
      <td>Myeloid</td>
      <td>326</td>
      <td>19952</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<p>You can see the genes measured in each dataset represented in <code class="docutils literal notranslate"><span class="pre">genes_measured</span></code>. Now lets get the <strong>genes that were measured in all datasets</strong>.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[16]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">var_somaid</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">nonzero</span><span class="p">(</span><span class="n">presence_matrix</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">A1</span> <span class="o">==</span> <span class="n">presence_matrix</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">])[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[17]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">lung_var</span> <span class="o">=</span> <span class="n">lung_var</span><span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;soma_joinid in </span><span class="si">{</span><span class="n">var_somaid</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="n">lung_var</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[17]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>soma_joinid</th>
      <th>feature_id</th>
      <th>feature_name</th>
      <th>feature_length</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>0</td>
      <td>ENSG00000121410</td>
      <td>A1BG</td>
      <td>3999</td>
    </tr>
    <tr>
      <th>3</th>
      <td>3</td>
      <td>ENSG00000175899</td>
      <td>A2M</td>
      <td>6318</td>
    </tr>
    <tr>
      <th>8</th>
      <td>8</td>
      <td>ENSG00000128274</td>
      <td>A4GALT</td>
      <td>3358</td>
    </tr>
    <tr>
      <th>10</th>
      <td>10</td>
      <td>ENSG00000094914</td>
      <td>AAAS</td>
      <td>4727</td>
    </tr>
    <tr>
      <th>11</th>
      <td>11</td>
      <td>ENSG00000081760</td>
      <td>AACS</td>
      <td>16039</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>29951</th>
      <td>29951</td>
      <td>ENSG00000177272</td>
      <td>KCNA3</td>
      <td>2476</td>
    </tr>
    <tr>
      <th>30157</th>
      <td>30157</td>
      <td>ENSG00000184709</td>
      <td>LRRC26</td>
      <td>1209</td>
    </tr>
    <tr>
      <th>30185</th>
      <td>30185</td>
      <td>ENSG00000087250</td>
      <td>MT3</td>
      <td>1679</td>
    </tr>
    <tr>
      <th>30202</th>
      <td>30202</td>
      <td>ENSG00000136352</td>
      <td>NKX2-1</td>
      <td>3165</td>
    </tr>
    <tr>
      <th>30512</th>
      <td>30512</td>
      <td>ENSG00000231439</td>
      <td>WASIR2</td>
      <td>1054</td>
    </tr>
  </tbody>
</table>
<p>11595 rows √ó 4 columns</p>
</div></div>
</div>
<p>The number of rows represents the genes that were measured in all lung datasets.</p>
</section>
<section id="Summary-of-lung-metadata">
<h3>Summary of lung metadata<a class="headerlink" href="#Summary-of-lung-metadata" title="Permalink to this heading">ÔÉÅ</a></h3>
<p>In the previous sections, using the Census we learned the following information:</p>
<ul class="simple">
<li><p>The total number of unique lung cells and their composition for:</p>
<ul>
<li><p>Number of datasets.</p></li>
<li><p>Number sequencing technologies, most of which are 10x</p></li>
<li><p>Mostly human data, but some diseases exist, primarily ‚Äúlung adenocarcinoma‚Äù and ‚ÄúCOVID-19 infected‚Äù</p></li>
<li><p>No sex biases.</p></li>
<li><p>Mostly data from cells (~80%) rather than nucleus (~20%)</p></li>
</ul>
</li>
<li><p>A total of <strong>~12k</strong> genes were measured across all cells.</p></li>
</ul>
</section>
</section>
<section id="Fetching-all-single-cell-human-lung-data-from-the-Census">
<h2>Fetching all single-cell human lung data from the Census<a class="headerlink" href="#Fetching-all-single-cell-human-lung-data-from-the-Census" title="Permalink to this heading">ÔÉÅ</a></h2>
<p>Since loading the entire lung data is resource-intensive, for the sake of this exercise let‚Äôs load a subset of the lung data into an <code class="docutils literal notranslate"><span class="pre">anndata.AnnData</span></code> object and perform some exploratory analysis.</p>
<p>We will subset to 100,000 random unique cells using the <code class="docutils literal notranslate"><span class="pre">lung_obs</span></code> <code class="docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code> we previously created.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[18]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">lung_cell_subsampled_n</span> <span class="o">=</span> <span class="mi">100000</span>
<span class="n">lung_cell_subsampled_ids</span> <span class="o">=</span> <span class="n">lung_obs</span><span class="p">[</span><span class="s2">&quot;soma_joinid&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">sample</span><span class="p">(</span><span class="n">lung_cell_subsampled_n</span><span class="p">,</span> <span class="n">random_state</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
</pre></div>
</div>
</div>
<p>Now we can directly use the values of <code class="docutils literal notranslate"><span class="pre">soma_joinid</span></code> for querying the Census data and obtaining an <code class="docutils literal notranslate"><span class="pre">AnnData</span></code> object.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[19]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">lung_gene_ids</span> <span class="o">=</span> <span class="n">lung_var</span><span class="p">[</span><span class="s2">&quot;soma_joinid&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">()</span>
<span class="n">lung_adata</span> <span class="o">=</span> <span class="n">cellxgene_census</span><span class="o">.</span><span class="n">get_anndata</span><span class="p">(</span>
    <span class="n">census</span><span class="p">,</span>
    <span class="n">organism</span><span class="o">=</span><span class="s2">&quot;Homo sapiens&quot;</span><span class="p">,</span>
    <span class="n">obs_coords</span><span class="o">=</span><span class="n">lung_cell_subsampled_ids</span><span class="p">,</span>
    <span class="n">var_coords</span><span class="o">=</span><span class="n">lung_gene_ids</span><span class="p">,</span>
<span class="p">)</span>

<span class="n">lung_adata</span><span class="o">.</span><span class="n">var_names</span> <span class="o">=</span> <span class="n">lung_adata</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s2">&quot;feature_name&quot;</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[20]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">lung_adata</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[20]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
AnnData object with n_obs √ó n_vars = 100000 √ó 11595
    obs: &#39;soma_joinid&#39;, &#39;dataset_id&#39;, &#39;assay&#39;, &#39;assay_ontology_term_id&#39;, &#39;cell_type&#39;, &#39;cell_type_ontology_term_id&#39;, &#39;development_stage&#39;, &#39;development_stage_ontology_term_id&#39;, &#39;disease&#39;, &#39;disease_ontology_term_id&#39;, &#39;donor_id&#39;, &#39;is_primary_data&#39;, &#39;self_reported_ethnicity&#39;, &#39;self_reported_ethnicity_ontology_term_id&#39;, &#39;sex&#39;, &#39;sex_ontology_term_id&#39;, &#39;suspension_type&#39;, &#39;tissue&#39;, &#39;tissue_ontology_term_id&#39;, &#39;tissue_general&#39;, &#39;tissue_general_ontology_term_id&#39;
    var: &#39;soma_joinid&#39;, &#39;feature_id&#39;, &#39;feature_name&#39;, &#39;feature_length&#39;
</pre></div></div>
</div>
<p>We are done with the census, so close it</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[21]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">census</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
<span class="k">del</span> <span class="n">census</span>
</pre></div>
</div>
</div>
</section>
<section id="Calculating-QC-metrics-of-the-lung-data">
<h2>Calculating QC metrics of the lung data<a class="headerlink" href="#Calculating-QC-metrics-of-the-lung-data" title="Permalink to this heading">ÔÉÅ</a></h2>
<p>Now let‚Äôs take a look at some QC metrics</p>
<p><strong>Top genes per cell</strong></p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[22]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">highest_expr_genes</span><span class="p">(</span><span class="n">lung_adata</span><span class="p">,</span> <span class="n">n_top</span><span class="o">=</span><span class="mi">20</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/notebooks_analysis_demo_comp_bio_explore_and_load_lung_data_39_0.png" src="../../_images/notebooks_analysis_demo_comp_bio_explore_and_load_lung_data_39_0.png" />
</div>
</div>
<p><strong>Number of sequenced genes by assay</strong></p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[23]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">calculate_qc_metrics</span><span class="p">(</span><span class="n">lung_adata</span><span class="p">,</span> <span class="n">percent_top</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">log1p</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">violin</span><span class="p">(</span><span class="n">lung_adata</span><span class="p">,</span> <span class="s2">&quot;n_genes_by_counts&quot;</span><span class="p">,</span> <span class="n">groupby</span><span class="o">=</span><span class="s2">&quot;assay&quot;</span><span class="p">,</span> <span class="n">jitter</span><span class="o">=</span><span class="mf">0.4</span><span class="p">,</span> <span class="n">rotation</span><span class="o">=</span><span class="mi">90</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/notebooks_analysis_demo_comp_bio_explore_and_load_lung_data_41_0.png" src="../../_images/notebooks_analysis_demo_comp_bio_explore_and_load_lung_data_41_0.png" />
</div>
</div>
<p><strong>Total counts by assay</strong></p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[24]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">violin</span><span class="p">(</span><span class="n">lung_adata</span><span class="p">,</span> <span class="s2">&quot;total_counts&quot;</span><span class="p">,</span> <span class="n">groupby</span><span class="o">=</span><span class="s2">&quot;assay&quot;</span><span class="p">,</span> <span class="n">jitter</span><span class="o">=</span><span class="mf">0.4</span><span class="p">,</span> <span class="n">rotation</span><span class="o">=</span><span class="mi">90</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/notebooks_analysis_demo_comp_bio_explore_and_load_lung_data_43_0.png" src="../../_images/notebooks_analysis_demo_comp_bio_explore_and_load_lung_data_43_0.png" />
</div>
</div>
<p>You can see that Smart-Seq2 is an outlier for the total counts per cell, so let‚Äôs exlcude it to see how the rest of the assays look like</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[25]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">violin</span><span class="p">(</span>
    <span class="n">lung_adata</span><span class="p">[</span><span class="n">lung_adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;assay&quot;</span><span class="p">]</span> <span class="o">!=</span> <span class="s2">&quot;Smart-seq2&quot;</span><span class="p">,],</span>
    <span class="s2">&quot;total_counts&quot;</span><span class="p">,</span>
    <span class="n">groupby</span><span class="o">=</span><span class="s2">&quot;assay&quot;</span><span class="p">,</span>
    <span class="n">jitter</span><span class="o">=</span><span class="mf">0.4</span><span class="p">,</span>
    <span class="n">rotation</span><span class="o">=</span><span class="mi">90</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/notebooks_analysis_demo_comp_bio_explore_and_load_lung_data_45_0.png" src="../../_images/notebooks_analysis_demo_comp_bio_explore_and_load_lung_data_45_0.png" />
</div>
</div>
</section>
<section id="Creating-a-normalized-expression-layer-and-embeddings">
<h2>Creating a normalized expression layer and embeddings<a class="headerlink" href="#Creating-a-normalized-expression-layer-and-embeddings" title="Permalink to this heading">ÔÉÅ</a></h2>
<p>Let‚Äôs perform a bread and butter normalization and take a look at UMAP embeddings, but for all the data below we‚Äôll exclude Smart-seq2 as this requires an extra step to normalize based on gene lengths</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[26]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">lung_adata</span> <span class="o">=</span> <span class="n">lung_adata</span><span class="p">[</span><span class="n">lung_adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;assay&quot;</span><span class="p">]</span> <span class="o">!=</span> <span class="s2">&quot;Smart-seq2&quot;</span><span class="p">,]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="n">lung_adata</span><span class="o">.</span><span class="n">layers</span><span class="p">[</span><span class="s2">&quot;counts&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">lung_adata</span><span class="o">.</span><span class="n">X</span>
</pre></div>
</div>
</div>
<p>Now let‚Äôs do some basic normalization:</p>
<ul class="simple">
<li><p>Normalize by sequencing depth</p></li>
<li><p>Transform to log-scale</p></li>
<li><p>Select 500 highly variable genes</p></li>
<li><p>Scale values across the gene axis</p></li>
</ul>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[27]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">normalize_total</span><span class="p">(</span><span class="n">lung_adata</span><span class="p">,</span> <span class="n">target_sum</span><span class="o">=</span><span class="mf">1e4</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">log1p</span><span class="p">(</span><span class="n">lung_adata</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">highly_variable_genes</span><span class="p">(</span><span class="n">lung_adata</span><span class="p">,</span> <span class="n">n_top_genes</span><span class="o">=</span><span class="mi">500</span><span class="p">,</span> <span class="n">flavor</span><span class="o">=</span><span class="s2">&quot;seurat_v3&quot;</span><span class="p">,</span> <span class="n">layer</span><span class="o">=</span><span class="s2">&quot;counts&quot;</span><span class="p">)</span>
<span class="n">lung_adata</span> <span class="o">=</span> <span class="n">lung_adata</span><span class="p">[:,</span> <span class="n">lung_adata</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">highly_variable</span><span class="p">]</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">scale</span><span class="p">(</span><span class="n">lung_adata</span><span class="p">,</span> <span class="n">max_value</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/home/ssm-user/cellxgene-census/venv/lib/python3.10/site-packages/scanpy/preprocessing/_highly_variable_genes.py:62: UserWarning: `flavor=&#39;seurat_v3&#39;` expects raw count data, but non-integers were found.
  warnings.warn(
/home/ssm-user/cellxgene-census/venv/lib/python3.10/site-packages/scanpy/preprocessing/_simple.py:843: UserWarning: Received a view of an AnnData. Making a copy.
  view_to_actual(adata)
</pre></div></div>
</div>
<p>And reduce dimensionality by obtaining UMAP embeddings.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[28]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">pca</span><span class="p">(</span><span class="n">lung_adata</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">neighbors</span><span class="p">(</span><span class="n">lung_adata</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">lung_adata</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/home/ssm-user/cellxgene-census/venv/lib/python3.10/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm
</pre></div></div>
</div>
<p>And plot these embeddings.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[29]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">n_cell_types</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">lung_adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;cell_type&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">())</span>

<span class="kn">from</span> <span class="nn">random</span> <span class="kn">import</span> <span class="n">randint</span>

<span class="n">colors</span> <span class="o">=</span> <span class="p">[]</span>

<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">lung_adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;cell_type&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">())):</span>
    <span class="n">colors</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;#</span><span class="si">%06X</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="n">randint</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mh">0xFFFFFF</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[30]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">lung_adata</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;cell_type&quot;</span><span class="p">,</span> <span class="n">palette</span><span class="o">=</span><span class="n">colors</span><span class="p">,</span> <span class="n">legend_loc</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/home/ssm-user/cellxgene-census/venv/lib/python3.10/site-packages/scanpy/plotting/_tools/scatterplots.py:392: UserWarning: No data for colormapping provided via &#39;c&#39;. Parameters &#39;cmap&#39; will be ignored
  cax = scatter(
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/notebooks_analysis_demo_comp_bio_explore_and_load_lung_data_54_1.png" src="../../_images/notebooks_analysis_demo_comp_bio_explore_and_load_lung_data_54_1.png" />
</div>
</div>
<p>Let‚Äôs color by assay.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[31]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">lung_adata</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;assay&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/home/ssm-user/cellxgene-census/venv/lib/python3.10/site-packages/scanpy/plotting/_tools/scatterplots.py:392: UserWarning: No data for colormapping provided via &#39;c&#39;. Parameters &#39;cmap&#39; will be ignored
  cax = scatter(
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/notebooks_analysis_demo_comp_bio_explore_and_load_lung_data_56_1.png" src="../../_images/notebooks_analysis_demo_comp_bio_explore_and_load_lung_data_56_1.png" />
</div>
</div>
<p>Given the high number of cell types it makes it hard to visualize, so let‚Äôs look at the top 20 most abundant cell types.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[32]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">top_cell_types</span> <span class="o">=</span> <span class="n">lung_adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;cell_type&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">value_counts</span><span class="p">()</span>
<span class="n">top_cell_types</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">top_cell_types</span><span class="o">.</span><span class="n">reset_index</span><span class="p">()</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">20</span><span class="p">)[</span><span class="s2">&quot;cell_type&quot;</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[33]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">lung_adata_top_cell_types</span> <span class="o">=</span> <span class="n">lung_adata</span><span class="p">[[</span><span class="n">i</span> <span class="ow">in</span> <span class="n">top_cell_types</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">lung_adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;cell_type&quot;</span><span class="p">]],</span> <span class="p">:]</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">lung_adata_top_cell_types</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;cell_type&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/home/ssm-user/cellxgene-census/venv/lib/python3.10/site-packages/scanpy/plotting/_tools/scatterplots.py:392: UserWarning: No data for colormapping provided via &#39;c&#39;. Parameters &#39;cmap&#39; will be ignored
  cax = scatter(
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/notebooks_analysis_demo_comp_bio_explore_and_load_lung_data_59_1.png" src="../../_images/notebooks_analysis_demo_comp_bio_explore_and_load_lung_data_59_1.png" />
</div>
</div>
<p>Let‚Äôs color by assay of this subset of the data.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[34]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">lung_adata_top_cell_types</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;assay&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/home/ssm-user/cellxgene-census/venv/lib/python3.10/site-packages/scanpy/plotting/_tools/scatterplots.py:392: UserWarning: No data for colormapping provided via &#39;c&#39;. Parameters &#39;cmap&#39; will be ignored
  cax = scatter(
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/notebooks_analysis_demo_comp_bio_explore_and_load_lung_data_61_1.png" src="../../_images/notebooks_analysis_demo_comp_bio_explore_and_load_lung_data_61_1.png" />
</div>
</div>
</section>
</section>


            </div>
            
            </div>
            <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="comp_bio_census_info.html" class="btn btn-neutral float-left" title="Learning about the CZ CELLxGENE Census" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="../api_demo/census_datasets.html" class="btn btn-neutral float-right" title="Exploring the Census Datasets table" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2022, Chan Zuckerberg Initiative.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
            </div>
        </div>

        </section>

    </div>

  

  
  
    
   

</body>
</html>