

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8">
  <meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Exploring all data from a tissue &mdash; cellxgene-census 0.5.0 documentation</title>
  

  
  <link rel="preconnect" href="https://fonts.googleapis.com">
  <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
  <link href="https://fonts.googleapis.com/css2?family=Inter:wght@400;500;600;700&display=swap" rel="stylesheet">
  <link rel="stylesheet" href="../../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/css/custom.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/nbsphinx-code-cells.css" type="text/css" />

  
  
  
  

  
  <!--[if lt IE 9]>
    <script src="../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../../" src="../../_static/documentation_options.js"></script>
        <script data-url_root="../../" id="documentation_options" src="../../_static/documentation_options.js"></script>
        <script src="../../_static/doctools.js"></script>
        <script src="../../_static/sphinx_highlight.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
        <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
        <script data-domain="cellxgene-census.readthedocs.io" defer="defer" src="https://plausible.io/js/script.js"></script>
    
    <script type="text/javascript" src="../../_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="Normalizing full-length gene sequencing data" href="comp_bio_normalizing_full_gene_sequencing.html" />
    <link rel="prev" title="Integrating multi-dataset slices of data" href="comp_bio_data_integration_scvi.html" /> 
</head>

<body class="wy-body-for-nav">

  <div style="height: 100vh; overflow: hidden;">
     
    <!-- top 48px for navbar height -->
    <div class="navbar-cxg">
        <div style="padding: 0 16px; display: flex; height: inherit; align-items: center; justify-content: space-between;">
            <span style="display:flex; gap: 32px; align-items: center;">
                <a href="https://cellxgene.cziscience.com/">
                  <img src="../../_static/img/cellxgene-discover-logo.svg" />
                </a>
                <span style="display:flex; gap: 16px;">
                    <span class="navbar-cxg-link">
                        <a class="navbar-cxg-anchor" href="https://cellxgene.cziscience.com/collections">Collections</a>
                    </span>
                    <span class="navbar-cxg-link">
                        <a class="navbar-cxg-anchor" href="https://cellxgene.cziscience.com/datasets">Datasets</a>    
                    </span>
                    <span class="navbar-cxg-link">
                        <a class="navbar-cxg-anchor" href="https://cellxgene.cziscience.com/gene-expression">Gene Expression</a>    
                    </span>
                </span>
            </span>

            <span class="navbar-cxg-link">
              <a class="navbar-cxg-anchor" href="https://cellxgene.cziscience.com/docs">Help & Documentation</a>    
            </span>
        </div>
    </div>
    <div style="width: 100%; height: 100%; overflow: auto; padding-top: 48px;">
        
        <!-- top 48px for navbar height -->
        <nav data-toggle="wy-nav-shift" class="wy-nav-side" style="top: 48px;" >
        <div class="wy-side-scroll">
            <div class="wy-side-nav-search" >
            

            
                <a href="../../index.html" class="icon icon-home"> cellxgene-census
            

            
            </a>

            
                
                
                <div class="version">
                    0.5
                </div>
                
            

            
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

            
            </div>

            
            <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
            
                
                
                
                
                
                <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../../cellxgene_census_docsite_installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../cellxgene_census_docsite_quick_start.html">Quick start</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../../examples.html">Tutorials</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="../../examples.html#api">API</a></li>
<li class="toctree-l2 current"><a class="reference internal" href="../../examples.html#analysis">Analysis</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="comp_bio_census_info.html">Learning about the CZ CELLxGENE Census</a></li>
<li class="toctree-l3"><a class="reference internal" href="comp_bio_data_integration_scvi.html">Integrating multi-dataset slices of data</a></li>
<li class="toctree-l3 current"><a class="current reference internal" href="#">Exploring all data from a tissue</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#Learning-about-the-lung-data-in-the-Census">Learning about the lung data in the Census</a></li>
<li class="toctree-l4"><a class="reference internal" href="#Fetching-all-single-cell-human-lung-data-from-the-Census">Fetching all single-cell human lung data from the Census</a></li>
<li class="toctree-l4"><a class="reference internal" href="#Calculating-QC-metrics-of-the-lung-data">Calculating QC metrics of the lung data</a></li>
<li class="toctree-l4"><a class="reference internal" href="#Creating-a-normalized-expression-layer-and-embeddings">Creating a normalized expression layer and embeddings</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="comp_bio_normalizing_full_gene_sequencing.html">Normalizing full-length gene sequencing data</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../cellxgene_census_docsite_schema.html">Census data and schema</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../cellxgene_census_docsite_data_release_info.html">Census data releases</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../python-api.html">Python API</a></li>
<li class="toctree-l1"><a class="reference external" href="https://chanzuckerberg.github.io/cellxgene-census/r/index.html">R API</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../cellxgene_census_docsite_FAQ.html">FAQ</a></li>
</ul>

                
            
            </div>
            
        </div>
        </nav>

        <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

        
        <nav class="wy-nav-top" aria-label="top navigation">
            
            <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
            <a href="../../index.html">cellxgene-census</a>
            
        </nav>


        <div class="wy-nav-content">
            
            <div class="rst-content">
            
            <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../../examples.html">Tutorials</a></li>
      <li class="breadcrumb-item active">Exploring all data from a tissue</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../../_sources/notebooks/analysis_demo/comp_bio_explore_and_load_lung_data.ipynb.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
            <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
            <div itemprop="articleBody">
                
  <section id="Exploring-all-data-from-a-tissue">
<h1>Exploring all data from a tissue<a class="headerlink" href="#Exploring-all-data-from-a-tissue" title="Permalink to this heading">¶</a></h1>
<p>This tutorial provides a series of examples for how to explore and query the Census in the context of a single tissue, lung. We will summarize cell and gene metadata, then fetch the single-cell expression counts and perform some basic data explorations via <a class="reference external" href="https://scanpy.readthedocs.io/en/stable/">Scanpy</a></p>
<p><strong>Contents</strong></p>
<ol class="arabic simple">
<li><p>Learning about the human lung data.</p>
<ol class="arabic simple">
<li><p>Learning about cells of the lung.</p></li>
<li><p>Learning about genes of the lung .</p></li>
</ol>
</li>
<li><p>Fetching all single-cell human lung data from the Census.</p></li>
<li><p>Calculating QC metrics of the lung data.</p></li>
<li><p>Creating a normalized expression layer and embeddings.</p></li>
</ol>
<section id="Learning-about-the-lung-data-in-the-Census">
<h2>Learning about the lung data in the Census<a class="headerlink" href="#Learning-about-the-lung-data-in-the-Census" title="Permalink to this heading">¶</a></h2>
<p>First we will open the Census. If you are not familiar with the basics of the Census API you should take a look at notebook <a class="reference external" href="https://cellxgene-census.readthedocs.io/en/latest/notebooks/analysis_demo/comp_bio_census_info.html">Learning about the CZ CELLxGENE Census</a></p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>import cellxgene_census
import scanpy as sc
import pandas as pd
import numpy as np
from scipy.sparse import coo_matrix

census = cellxgene_census.open_soma()
</pre></div>
</div>
</div>
<p>Let’s first take a look at the number of cells from human lung:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>summary_table = census[&quot;census_info&quot;][&quot;summary_cell_counts&quot;].read().concat().to_pandas()

summary_table.query(&quot;organism == &#39;Homo sapiens&#39; &amp; category == &#39;tissue_general&#39; &amp; label ==&#39;lung&#39;&quot;)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>soma_joinid</th>
      <th>organism</th>
      <th>category</th>
      <th>ontology_term_id</th>
      <th>unique_cell_count</th>
      <th>total_cell_count</th>
      <th>label</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>944</th>
      <td>944</td>
      <td>Homo sapiens</td>
      <td>tissue_general</td>
      <td>UBERON:0002048</td>
      <td>2295237</td>
      <td>4573212</td>
      <td>lung</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<p>There you can see the total of cells of under <code class="docutils literal notranslate"><span class="pre">total_cell_count</span></code> and the unique number cells under <code class="docutils literal notranslate"><span class="pre">unique_cell_count</span></code> (i.e. after removing cells that were included in multiple datasets).</p>
<p>Let’s now take a look at the cell and gene information of this slice of the Census.</p>
<section id="Learning-about-cells-of-lung-data">
<h3>Learning about cells of lung data<a class="headerlink" href="#Learning-about-cells-of-lung-data" title="Permalink to this heading">¶</a></h3>
<p>Let’s load the cell metadata for all lung cells and select only the unique cells using <code class="docutils literal notranslate"><span class="pre">is_primary_data</span></code>.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>lung_obs = (
    census[&quot;census_data&quot;][&quot;homo_sapiens&quot;]
    .obs.read(value_filter=&quot;tissue_general == &#39;lung&#39; and is_primary_data == True&quot;)
    .concat()
    .to_pandas()
)
lung_obs
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>soma_joinid</th>
      <th>dataset_id</th>
      <th>assay</th>
      <th>assay_ontology_term_id</th>
      <th>cell_type</th>
      <th>cell_type_ontology_term_id</th>
      <th>development_stage</th>
      <th>development_stage_ontology_term_id</th>
      <th>disease</th>
      <th>disease_ontology_term_id</th>
      <th>...</th>
      <th>is_primary_data</th>
      <th>self_reported_ethnicity</th>
      <th>self_reported_ethnicity_ontology_term_id</th>
      <th>sex</th>
      <th>sex_ontology_term_id</th>
      <th>suspension_type</th>
      <th>tissue</th>
      <th>tissue_ontology_term_id</th>
      <th>tissue_general</th>
      <th>tissue_general_ontology_term_id</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>145215</td>
      <td>d8da613f-e681-4c69-b463-e94f5e66847f</td>
      <td>10x 3' v3</td>
      <td>EFO:0009922</td>
      <td>lung ciliated cell</td>
      <td>CL:1000271</td>
      <td>70-year-old human stage</td>
      <td>HsapDv:0000164</td>
      <td>normal</td>
      <td>PATO:0000461</td>
      <td>...</td>
      <td>True</td>
      <td>European</td>
      <td>HANCESTRO:0005</td>
      <td>female</td>
      <td>PATO:0000383</td>
      <td>nucleus</td>
      <td>lung</td>
      <td>UBERON:0002048</td>
      <td>lung</td>
      <td>UBERON:0002048</td>
    </tr>
    <tr>
      <th>1</th>
      <td>145216</td>
      <td>d8da613f-e681-4c69-b463-e94f5e66847f</td>
      <td>10x 3' v3</td>
      <td>EFO:0009922</td>
      <td>alveolar macrophage</td>
      <td>CL:0000583</td>
      <td>70-year-old human stage</td>
      <td>HsapDv:0000164</td>
      <td>normal</td>
      <td>PATO:0000461</td>
      <td>...</td>
      <td>True</td>
      <td>European</td>
      <td>HANCESTRO:0005</td>
      <td>female</td>
      <td>PATO:0000383</td>
      <td>nucleus</td>
      <td>lung</td>
      <td>UBERON:0002048</td>
      <td>lung</td>
      <td>UBERON:0002048</td>
    </tr>
    <tr>
      <th>2</th>
      <td>145217</td>
      <td>d8da613f-e681-4c69-b463-e94f5e66847f</td>
      <td>10x 3' v3</td>
      <td>EFO:0009922</td>
      <td>lung ciliated cell</td>
      <td>CL:1000271</td>
      <td>70-year-old human stage</td>
      <td>HsapDv:0000164</td>
      <td>normal</td>
      <td>PATO:0000461</td>
      <td>...</td>
      <td>True</td>
      <td>European</td>
      <td>HANCESTRO:0005</td>
      <td>female</td>
      <td>PATO:0000383</td>
      <td>nucleus</td>
      <td>lung</td>
      <td>UBERON:0002048</td>
      <td>lung</td>
      <td>UBERON:0002048</td>
    </tr>
    <tr>
      <th>3</th>
      <td>145218</td>
      <td>d8da613f-e681-4c69-b463-e94f5e66847f</td>
      <td>10x 3' v3</td>
      <td>EFO:0009922</td>
      <td>lung ciliated cell</td>
      <td>CL:1000271</td>
      <td>70-year-old human stage</td>
      <td>HsapDv:0000164</td>
      <td>normal</td>
      <td>PATO:0000461</td>
      <td>...</td>
      <td>True</td>
      <td>European</td>
      <td>HANCESTRO:0005</td>
      <td>female</td>
      <td>PATO:0000383</td>
      <td>nucleus</td>
      <td>lung</td>
      <td>UBERON:0002048</td>
      <td>lung</td>
      <td>UBERON:0002048</td>
    </tr>
    <tr>
      <th>4</th>
      <td>145219</td>
      <td>d8da613f-e681-4c69-b463-e94f5e66847f</td>
      <td>10x 3' v3</td>
      <td>EFO:0009922</td>
      <td>lung ciliated cell</td>
      <td>CL:1000271</td>
      <td>70-year-old human stage</td>
      <td>HsapDv:0000164</td>
      <td>normal</td>
      <td>PATO:0000461</td>
      <td>...</td>
      <td>True</td>
      <td>European</td>
      <td>HANCESTRO:0005</td>
      <td>female</td>
      <td>PATO:0000383</td>
      <td>nucleus</td>
      <td>lung</td>
      <td>UBERON:0002048</td>
      <td>lung</td>
      <td>UBERON:0002048</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>2295232</th>
      <td>45561670</td>
      <td>f64e1be1-de15-4d27-8da4-82225cd4c035</td>
      <td>10x 3' v3</td>
      <td>EFO:0009922</td>
      <td>T cell</td>
      <td>CL:0000084</td>
      <td>78-year-old human stage</td>
      <td>HsapDv:0000172</td>
      <td>lung adenocarcinoma</td>
      <td>MONDO:0005061</td>
      <td>...</td>
      <td>True</td>
      <td>European</td>
      <td>HANCESTRO:0005</td>
      <td>male</td>
      <td>PATO:0000384</td>
      <td>cell</td>
      <td>lung</td>
      <td>UBERON:0002048</td>
      <td>lung</td>
      <td>UBERON:0002048</td>
    </tr>
    <tr>
      <th>2295233</th>
      <td>45562154</td>
      <td>f64e1be1-de15-4d27-8da4-82225cd4c035</td>
      <td>10x 3' v2</td>
      <td>EFO:0009899</td>
      <td>T cell</td>
      <td>CL:0000084</td>
      <td>67-year-old human stage</td>
      <td>HsapDv:0000161</td>
      <td>lung adenocarcinoma</td>
      <td>MONDO:0005061</td>
      <td>...</td>
      <td>True</td>
      <td>European</td>
      <td>HANCESTRO:0005</td>
      <td>female</td>
      <td>PATO:0000383</td>
      <td>cell</td>
      <td>lung</td>
      <td>UBERON:0002048</td>
      <td>lung</td>
      <td>UBERON:0002048</td>
    </tr>
    <tr>
      <th>2295234</th>
      <td>45565290</td>
      <td>f64e1be1-de15-4d27-8da4-82225cd4c035</td>
      <td>10x 3' v2</td>
      <td>EFO:0009899</td>
      <td>T cell</td>
      <td>CL:0000084</td>
      <td>65-year-old human stage</td>
      <td>HsapDv:0000159</td>
      <td>lung adenocarcinoma</td>
      <td>MONDO:0005061</td>
      <td>...</td>
      <td>True</td>
      <td>European</td>
      <td>HANCESTRO:0005</td>
      <td>female</td>
      <td>PATO:0000383</td>
      <td>cell</td>
      <td>lung</td>
      <td>UBERON:0002048</td>
      <td>lung</td>
      <td>UBERON:0002048</td>
    </tr>
    <tr>
      <th>2295235</th>
      <td>45566613</td>
      <td>f64e1be1-de15-4d27-8da4-82225cd4c035</td>
      <td>10x 3' v2</td>
      <td>EFO:0009899</td>
      <td>T cell</td>
      <td>CL:0000084</td>
      <td>80-year-old human stage</td>
      <td>HsapDv:0000206</td>
      <td>lung adenocarcinoma</td>
      <td>MONDO:0005061</td>
      <td>...</td>
      <td>True</td>
      <td>European</td>
      <td>HANCESTRO:0005</td>
      <td>male</td>
      <td>PATO:0000384</td>
      <td>cell</td>
      <td>lung</td>
      <td>UBERON:0002048</td>
      <td>lung</td>
      <td>UBERON:0002048</td>
    </tr>
    <tr>
      <th>2295236</th>
      <td>45567413</td>
      <td>f64e1be1-de15-4d27-8da4-82225cd4c035</td>
      <td>10x 3' v3</td>
      <td>EFO:0009922</td>
      <td>T cell</td>
      <td>CL:0000084</td>
      <td>68-year-old human stage</td>
      <td>HsapDv:0000162</td>
      <td>lung adenocarcinoma</td>
      <td>MONDO:0005061</td>
      <td>...</td>
      <td>True</td>
      <td>European</td>
      <td>HANCESTRO:0005</td>
      <td>male</td>
      <td>PATO:0000384</td>
      <td>cell</td>
      <td>lung</td>
      <td>UBERON:0002048</td>
      <td>lung</td>
      <td>UBERON:0002048</td>
    </tr>
  </tbody>
</table>
<p>2295237 rows × 21 columns</p>
</div></div>
</div>
<p>You can see that the number or rows represents the total number of unique lung cells in the Census. Now let’s take a deeper dive into the characteristics of these cells.</p>
<section id="Datasets">
<h4>Datasets<a class="headerlink" href="#Datasets" title="Permalink to this heading">¶</a></h4>
<p>First let’s start by looking at what are the datasets and collections from <a class="reference external" href="https://cellxgene.cziscience.com/collections">CELLxGENE Discover</a> contributing to lung. For this we will use the dataset table at <code class="docutils literal notranslate"><span class="pre">census[&quot;census-info&quot;][&quot;datasets&quot;]</span></code> that contains metadata of all datasets used to build this Census.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>census_datasets = (
    census[&quot;census_info&quot;][&quot;datasets&quot;]
    .read(column_names=[&quot;collection_name&quot;, &quot;dataset_title&quot;, &quot;dataset_id&quot;, &quot;soma_joinid&quot;])
    .concat()
    .to_pandas()
)
census_datasets = census_datasets.set_index(&quot;dataset_id&quot;)
census_datasets
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>collection_name</th>
      <th>dataset_title</th>
      <th>soma_joinid</th>
    </tr>
    <tr>
      <th>dataset_id</th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>f512b8b6-369d-4a85-a695-116e0806857f</th>
      <td>Transcriptional Programming of Normal and Infl...</td>
      <td>Skin</td>
      <td>0</td>
    </tr>
    <tr>
      <th>36c867a7-be10-4e69-9b39-5de12b0af6da</th>
      <td>Single-Cell Analysis of Crohn’s Disease Lesion...</td>
      <td>Ileum</td>
      <td>1</td>
    </tr>
    <tr>
      <th>58b01044-c5e5-4b0f-8a2d-6ebf951e01ff</th>
      <td>Humoral immunity at the brain borders in homeo...</td>
      <td>A scRNA-seq atlas of immune cells at the CNS b...</td>
      <td>2</td>
    </tr>
    <tr>
      <th>456e8b9b-f872-488b-871d-94534090a865</th>
      <td>Single-cell atlas of peripheral immune respons...</td>
      <td>Single-cell atlas of peripheral immune respons...</td>
      <td>3</td>
    </tr>
    <tr>
      <th>d8da613f-e681-4c69-b463-e94f5e66847f</th>
      <td>A molecular single-cell lung atlas of lethal C...</td>
      <td>A molecular single-cell lung atlas of lethal C...</td>
      <td>4</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>30498543-4fdd-4f86-9e1b-05c1a1454a6a</th>
      <td>COVID-19 mRNA vaccine elicits a potent adaptiv...</td>
      <td>B cells -- CV19 infection, vaccination and HC</td>
      <td>524</td>
    </tr>
    <tr>
      <th>b5191f01-f67d-44b8-bc8d-511a4ecd07bb</th>
      <td>COVID-19 mRNA vaccine elicits a potent adaptiv...</td>
      <td>innate T cells -- CV19 infection, vaccination ...</td>
      <td>525</td>
    </tr>
    <tr>
      <th>e463dae9-3fc1-476d-870e-d98a04c56cd6</th>
      <td>COVID-19 mRNA vaccine elicits a potent adaptiv...</td>
      <td>M cells -- CV19 infection, vaccination and HC</td>
      <td>526</td>
    </tr>
    <tr>
      <th>1b699e04-1127-42ea-998b-011ace4a5b81</th>
      <td>COVID-19 mRNA vaccine elicits a potent adaptiv...</td>
      <td>T cells -- CV19 infection, vaccination and HC</td>
      <td>527</td>
    </tr>
    <tr>
      <th>87ce26ed-e5d1-44b4-81cc-cc5b709a169f</th>
      <td>COVID-19 mRNA vaccine elicits a potent adaptiv...</td>
      <td>Circulating Immune cells -- CV19 infection, va...</td>
      <td>528</td>
    </tr>
  </tbody>
</table>
<p>529 rows × 3 columns</p>
</div></div>
</div>
<p>The <code class="docutils literal notranslate"><span class="pre">obs</span></code> cell metadata <code class="docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code> contains a column <code class="docutils literal notranslate"><span class="pre">dataset_id</span></code> that can be used for joining to the <code class="docutils literal notranslate"><span class="pre">census_dataset</span></code> <code class="docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code> we just created.</p>
<p>So let’s take a look at the cell counts per <code class="docutils literal notranslate"><span class="pre">dataset_id</span></code> of the lung slice and then join to the dataset table to append the human-readable labels.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>dataset_cell_counts = pd.DataFrame(lung_obs[[&quot;dataset_id&quot;]].value_counts())
dataset_cell_counts = dataset_cell_counts.rename(columns={0: &quot;cell_counts&quot;})
dataset_cell_counts = dataset_cell_counts.merge(census_datasets, on=&quot;dataset_id&quot;)

dataset_cell_counts
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>count</th>
      <th>collection_name</th>
      <th>dataset_title</th>
      <th>soma_joinid</th>
    </tr>
    <tr>
      <th>dataset_id</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>1e6a6ef9-7ec9-4c90-bbfb-2ad3c3165fd1</th>
      <td>1028006</td>
      <td>High-resolution single-cell atlas reveals dive...</td>
      <td>The single-cell lung cancer atlas (LuCA) -- ex...</td>
      <td>157</td>
    </tr>
    <tr>
      <th>066943a2-fdac-4b29-b348-40cede398e4e</th>
      <td>254931</td>
      <td>The integrated Human Lung Cell Atlas</td>
      <td>The integrated Human Lung Cell Atlas</td>
      <td>79</td>
    </tr>
    <tr>
      <th>f7c1c579-2dc0-47e2-ba19-8165c5a0e353</th>
      <td>217738</td>
      <td>A human cell atlas of fetal gene expression</td>
      <td>Survey of human embryonic development</td>
      <td>94</td>
    </tr>
    <tr>
      <th>d8da613f-e681-4c69-b463-e94f5e66847f</th>
      <td>116313</td>
      <td>A molecular single-cell lung atlas of lethal C...</td>
      <td>A molecular single-cell lung atlas of lethal C...</td>
      <td>4</td>
    </tr>
    <tr>
      <th>576f193c-75d0-4a11-bd25-8676587e6dc2</th>
      <td>90384</td>
      <td>HTAN MSK - Single cell profiling reveals novel...</td>
      <td>Combined samples</td>
      <td>519</td>
    </tr>
    <tr>
      <th>3dc61ca1-ce40-46b6-8337-f27260fd9a03</th>
      <td>71752</td>
      <td>A human fetal lung cell atlas uncovers proxima...</td>
      <td>All cells</td>
      <td>75</td>
    </tr>
    <tr>
      <th>8c42cfd0-0b0a-46d5-910c-fc833d83c45e</th>
      <td>60993</td>
      <td>A molecular cell atlas of the human lung from ...</td>
      <td>Krasnow Lab Human Lung Cell Atlas, 10X</td>
      <td>177</td>
    </tr>
    <tr>
      <th>2672b679-8048-4f5e-9786-f1b196ccfd08</th>
      <td>57019</td>
      <td>scRNA-seq assessment of the human lung, spleen...</td>
      <td>Lung Parenchyma</td>
      <td>317</td>
    </tr>
    <tr>
      <th>9dbab10c-118d-496b-966a-67f1763a6b7d</th>
      <td>49014</td>
      <td>COVID-19 immune features revealed by a large-s...</td>
      <td>Large-scale single-cell analysis reveals criti...</td>
      <td>63</td>
    </tr>
    <tr>
      <th>9968be68-ab65-4a38-9e1a-c9b6abece194</th>
      <td>47909</td>
      <td>Charting human development using a multi-endod...</td>
      <td>Developing Human Atlas</td>
      <td>508</td>
    </tr>
    <tr>
      <th>3de0ad6d-4378-4f62-b37b-ec0b75a50d94</th>
      <td>46500</td>
      <td>LungMAP — Human data from a broad age healthy ...</td>
      <td>Single-cell multiomic profiling of human lungs...</td>
      <td>188</td>
    </tr>
    <tr>
      <th>2f132ec9-24b5-422f-9be0-ccef03b4fe28</th>
      <td>39778</td>
      <td>SARS-CoV-2 receptor ACE2 and TMPRSS2 are prima...</td>
      <td>Lung</td>
      <td>104</td>
    </tr>
    <tr>
      <th>1e5bd3b8-6a0e-4959-8d69-cafed30fe814</th>
      <td>35699</td>
      <td>Emphysema Cell Atlas</td>
      <td>immune cells</td>
      <td>51</td>
    </tr>
    <tr>
      <th>53d208b0-2cfd-4366-9866-c3c6114081bc</th>
      <td>35682</td>
      <td>Tabula Sapiens</td>
      <td>Tabula Sapiens - All Cells</td>
      <td>339</td>
    </tr>
    <tr>
      <th>1b9d8702-5af8-4142-85ed-020eb06ec4f6</th>
      <td>35419</td>
      <td>Cross-tissue immune cell analysis reveals tiss...</td>
      <td>Global</td>
      <td>314</td>
    </tr>
    <tr>
      <th>4ed927e9-c099-49af-b8ce-a2652d069333</th>
      <td>35284</td>
      <td>Single-nucleus cross-tissue molecular referenc...</td>
      <td>Single-nucleus cross-tissue molecular referenc...</td>
      <td>62</td>
    </tr>
    <tr>
      <th>2adb1f8a-a6b1-4909-8ee8-484814e2d4bf</th>
      <td>33698</td>
      <td>Construction of a human cell landscape at sing...</td>
      <td>Construction of a human cell landscape at sing...</td>
      <td>244</td>
    </tr>
    <tr>
      <th>4b6af54a-4a21-46e0-bc8d-673c0561a836</th>
      <td>15495</td>
      <td>Emphysema Cell Atlas</td>
      <td>non-immune cells</td>
      <td>53</td>
    </tr>
    <tr>
      <th>01209dce-3575-4bed-b1df-129f57fbc031</th>
      <td>11059</td>
      <td>Single-cell transcriptomics of human T cells r...</td>
      <td>Single-cell transcriptomics of human T cells r...</td>
      <td>186</td>
    </tr>
    <tr>
      <th>e04daea4-4412-45b5-989e-76a9be070a89</th>
      <td>8657</td>
      <td>A molecular cell atlas of the human lung from ...</td>
      <td>Krasnow Lab Human Lung Cell Atlas, Smart-seq2</td>
      <td>176</td>
    </tr>
    <tr>
      <th>214bf9eb-93db-48c8-8e3c-9bb22fa3bc63</th>
      <td>3662</td>
      <td>Emphysema Cell Atlas</td>
      <td>AT2 cells</td>
      <td>52</td>
    </tr>
    <tr>
      <th>f9846bb4-784d-4582-92c1-3f279e4c6f0c</th>
      <td>176</td>
      <td>A human fetal lung cell atlas uncovers proxima...</td>
      <td>Fibroblast and smooth muscle</td>
      <td>71</td>
    </tr>
    <tr>
      <th>f64e1be1-de15-4d27-8da4-82225cd4c035</th>
      <td>55</td>
      <td>HTAN MSK - Single cell profiling reveals novel...</td>
      <td>Immune cells</td>
      <td>520</td>
    </tr>
    <tr>
      <th>810ac45f-8969-4698-b42c-652f802f75c2</th>
      <td>10</td>
      <td>A human fetal lung cell atlas uncovers proxima...</td>
      <td>Endothelium</td>
      <td>77</td>
    </tr>
    <tr>
      <th>0ba16f4b-cb87-4fa3-9363-19fc51eec6e7</th>
      <td>4</td>
      <td>A human fetal lung cell atlas uncovers proxima...</td>
      <td>Myeloid</td>
      <td>74</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<p>These are all the datasets lung cells whose counts are reprensented in the column <code class="docutils literal notranslate"><span class="pre">cell_counts</span></code>. The top collections with lung data are:</p>
<ol class="arabic simple">
<li><p><a class="reference external" href="https://cellxgene.cziscience.com/collections/6f6d381a-7701-4781-935c-db10d30de293">The integrated Human Lung Cell Atlas</a>.</p></li>
<li><p><a class="reference external" href="https://cellxgene.cziscience.com/collections/c114c20f-1ef4-49a5-9c2e-d965787fb90c">A human cell atlas of fetal gene expression</a>.</p></li>
<li><p><a class="reference external" href="https://cellxgene.cziscience.com/collections/edb893ee-4066-4128-9aec-5eb2b03f8287">High-resolution single-cell atlas reveals diversity and plasticity of tumor-associated neutrophils in non-small cell lung cancer</a>.</p></li>
<li><p><a class="reference external" href="https://cellxgene.cziscience.com/collections/62e8f058-9c37-48bc-9200-e767f318a8ec">HTAN MSK - Single cell profiling reveals novel tumor and myeloid subpopulations in small cell lung cancer</a>.</p></li>
<li><p><a class="reference external" href="https://cellxgene.cziscience.com/collections/2d2e2acd-dade-489f-a2da-6c11aa654028">A human fetal lung cell atlas uncovers proximal-distal gradients of differentiation and key regulators of epithelial fates.</a>.</p></li>
</ol>
</section>
<section id="Assays">
<h4>Assays<a class="headerlink" href="#Assays" title="Permalink to this heading">¶</a></h4>
<p>Let’s use similar logic to take a look at all the assays available for human lung data. This tells us that most assays are from 10x technologies and sci-RNA-seq.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>lung_obs[[&quot;assay&quot;]].value_counts()
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
assay
10x 3&#39; v2                                   1087211
10x 3&#39; v3                                    436291
10x 5&#39; v1                                    218556
sci-RNA-seq                                  217738
BD Rhapsody Whole Transcriptome Analysis     122902
single cell library construction              58981
10x 5&#39; v2                                     45798
microwell-seq                                 33698
Smart-seq2                                    25662
inDrop                                        25652
10x 3&#39; v1                                     11558
Drop-seq                                       7333
10x 3&#39; transcription profiling                 3857
Name: count, dtype: int64
</pre></div></div>
</div>
</section>
<section id="Disease">
<h4>Disease<a class="headerlink" href="#Disease" title="Permalink to this heading">¶</a></h4>
<p>And now let’s take a look at diseased cell counts, with <code class="docutils literal notranslate"><span class="pre">normal</span></code> indicating non-diseased cells.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>lung_obs[[&quot;disease&quot;]].value_counts()
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
disease
normal                                   1111566
lung adenocarcinoma                       626322
squamous cell lung carcinoma              189044
COVID-19                                  128650
non-small cell lung carcinoma             120796
chronic obstructive pulmonary disease      55254
pulmonary emphysema                        32065
small cell lung carcinoma                  31540
Name: count, dtype: int64
</pre></div></div>
</div>
</section>
<section id="Sex">
<h4>Sex<a class="headerlink" href="#Sex" title="Permalink to this heading">¶</a></h4>
<p>There doesn’t seem to be strong biases for sex.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>lung_obs[[&quot;sex&quot;]].value_counts()
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
sex
male       1267756
female     1019958
unknown       7523
Name: count, dtype: int64
</pre></div></div>
</div>
</section>
<section id="Cell-vs-nucleus">
<h4>Cell vs nucleus<a class="headerlink" href="#Cell-vs-nucleus" title="Permalink to this heading">¶</a></h4>
<p>The majority of data are from cells and not nucleus.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>lung_obs[[&quot;suspension_type&quot;]].value_counts()
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
suspension_type
cell               1879402
nucleus             415835
Name: count, dtype: int64
</pre></div></div>
</div>
</section>
<section id="Cell-types">
<h4>Cell types<a class="headerlink" href="#Cell-types" title="Permalink to this heading">¶</a></h4>
<p>Let’s take a look at the counts of the top 20 cell types.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>lung_obs[[&quot;cell_type&quot;]].value_counts().head(20)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
cell_type
alveolar macrophage                           244502
CD4-positive, alpha-beta T cell               185787
macrophage                                    154415
CD8-positive, alpha-beta T cell               152746
type II pneumocyte                            143252
epithelial cell of lower respiratory tract    101959
classical monocyte                             90969
natural killer cell                            90715
stromal cell                                   87587
malignant cell                                 75917
B cell                                         61804
T cell                                         53703
fibroblast                                     41934
plasma cell                                    41907
fibroblast of lung                             41294
epithelial cell of lung                        38426
epithelial cell                                37498
regulatory T cell                              36381
mesenchymal cell                               33235
monocyte                                       33154
Name: count, dtype: int64
</pre></div></div>
</div>
</section>
<section id="Sub-tissues">
<h4>Sub-tissues<a class="headerlink" href="#Sub-tissues" title="Permalink to this heading">¶</a></h4>
<p>We can look at the original tissue annotations that were mapped to “lung”.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>lung_obs[[&quot;tissue&quot;]].value_counts()
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
tissue
lung                       1882088
lung parenchyma             311950
alveolus of lung             54856
lingula of left lung         35284
lower lobe of left lung      11059
Name: count, dtype: int64
</pre></div></div>
</div>
</section>
</section>
<section id="Learning-about-genes-of-lung-data">
<h3>Learning about genes of lung data<a class="headerlink" href="#Learning-about-genes-of-lung-data" title="Permalink to this heading">¶</a></h3>
<p>Let’s load the gene metadata of the Census.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>lung_var = census[&quot;census_data&quot;][&quot;homo_sapiens&quot;].ms[&quot;RNA&quot;].var.read().concat().to_pandas()
lung_var
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>soma_joinid</th>
      <th>feature_id</th>
      <th>feature_name</th>
      <th>feature_length</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>0</td>
      <td>ENSG00000238009</td>
      <td>RP11-34P13.7</td>
      <td>3726</td>
    </tr>
    <tr>
      <th>1</th>
      <td>1</td>
      <td>ENSG00000279457</td>
      <td>WASH9P</td>
      <td>1397</td>
    </tr>
    <tr>
      <th>2</th>
      <td>2</td>
      <td>ENSG00000228463</td>
      <td>AP006222.1</td>
      <td>8224</td>
    </tr>
    <tr>
      <th>3</th>
      <td>3</td>
      <td>ENSG00000237094</td>
      <td>RP4-669L17.4</td>
      <td>6204</td>
    </tr>
    <tr>
      <th>4</th>
      <td>4</td>
      <td>ENSG00000230021</td>
      <td>RP11-206L10.17</td>
      <td>5495</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>60659</th>
      <td>60659</td>
      <td>ENSG00000288719</td>
      <td>RP4-669P10.21</td>
      <td>4252</td>
    </tr>
    <tr>
      <th>60660</th>
      <td>60660</td>
      <td>ENSG00000288720</td>
      <td>RP11-852E15.3</td>
      <td>7007</td>
    </tr>
    <tr>
      <th>60661</th>
      <td>60661</td>
      <td>ENSG00000288721</td>
      <td>RP5-973N23.5</td>
      <td>7765</td>
    </tr>
    <tr>
      <th>60662</th>
      <td>60662</td>
      <td>ENSG00000288723</td>
      <td>RP11-553N16.6</td>
      <td>1015</td>
    </tr>
    <tr>
      <th>60663</th>
      <td>60663</td>
      <td>ENSG00000288724</td>
      <td>RP13-546I2.2</td>
      <td>625</td>
    </tr>
  </tbody>
</table>
<p>60664 rows × 4 columns</p>
</div></div>
</div>
<p>You can see the total number of genes represented by the number of rows. This number is actually misleading because it is the join of all genes in the Census. However we know that the lung data comes from a subset of datasets.</p>
<p>So let’s take a look at the number of genes that were measured in each of those datasets.</p>
<p>To accomplish this we can use the “dataset presence matrix” at <code class="docutils literal notranslate"><span class="pre">census[&quot;census_data&quot;][&quot;homo_sapiens&quot;].ms[&quot;RNA&quot;][&quot;feature_dataset_presence_matrix&quot;]</span></code>. This is a boolean matrix <code class="docutils literal notranslate"><span class="pre">N</span> <span class="pre">x</span> <span class="pre">M</span></code> where <code class="docutils literal notranslate"><span class="pre">N</span></code> is the number of datasets and <code class="docutils literal notranslate"><span class="pre">M</span></code> is the number of genes in the Census.</p>
<p>So we can select the rows corresponding to the lung datasets and perform a row-wise sum.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[13]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>presence_matrix = cellxgene_census.get_presence_matrix(census, &quot;Homo sapiens&quot;, &quot;RNA&quot;)
presence_matrix = presence_matrix[dataset_cell_counts.soma_joinid, :]
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>presence_matrix.sum(axis=1).A1
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
array([17811, 28024, 44150, 34265, 22447, 26347, 20921, 24672, 27705,
       27243, 26323, 27181, 23203, 57030, 32610, 29620, 26454, 23705,
       38676, 47307, 21641, 23740, 22552, 20594, 19952], dtype=uint64)
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[15]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>genes_measured = presence_matrix.sum(axis=1).A1
dataset_cell_counts[&quot;genes_measured&quot;] = genes_measured
dataset_cell_counts
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[15]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>count</th>
      <th>collection_name</th>
      <th>dataset_title</th>
      <th>soma_joinid</th>
      <th>genes_measured</th>
    </tr>
    <tr>
      <th>dataset_id</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>1e6a6ef9-7ec9-4c90-bbfb-2ad3c3165fd1</th>
      <td>1028006</td>
      <td>High-resolution single-cell atlas reveals dive...</td>
      <td>The single-cell lung cancer atlas (LuCA) -- ex...</td>
      <td>157</td>
      <td>17811</td>
    </tr>
    <tr>
      <th>066943a2-fdac-4b29-b348-40cede398e4e</th>
      <td>254931</td>
      <td>The integrated Human Lung Cell Atlas</td>
      <td>The integrated Human Lung Cell Atlas</td>
      <td>79</td>
      <td>28024</td>
    </tr>
    <tr>
      <th>f7c1c579-2dc0-47e2-ba19-8165c5a0e353</th>
      <td>217738</td>
      <td>A human cell atlas of fetal gene expression</td>
      <td>Survey of human embryonic development</td>
      <td>94</td>
      <td>44150</td>
    </tr>
    <tr>
      <th>d8da613f-e681-4c69-b463-e94f5e66847f</th>
      <td>116313</td>
      <td>A molecular single-cell lung atlas of lethal C...</td>
      <td>A molecular single-cell lung atlas of lethal C...</td>
      <td>4</td>
      <td>34265</td>
    </tr>
    <tr>
      <th>576f193c-75d0-4a11-bd25-8676587e6dc2</th>
      <td>90384</td>
      <td>HTAN MSK - Single cell profiling reveals novel...</td>
      <td>Combined samples</td>
      <td>519</td>
      <td>22447</td>
    </tr>
    <tr>
      <th>3dc61ca1-ce40-46b6-8337-f27260fd9a03</th>
      <td>71752</td>
      <td>A human fetal lung cell atlas uncovers proxima...</td>
      <td>All cells</td>
      <td>75</td>
      <td>26347</td>
    </tr>
    <tr>
      <th>8c42cfd0-0b0a-46d5-910c-fc833d83c45e</th>
      <td>60993</td>
      <td>A molecular cell atlas of the human lung from ...</td>
      <td>Krasnow Lab Human Lung Cell Atlas, 10X</td>
      <td>177</td>
      <td>20921</td>
    </tr>
    <tr>
      <th>2672b679-8048-4f5e-9786-f1b196ccfd08</th>
      <td>57019</td>
      <td>scRNA-seq assessment of the human lung, spleen...</td>
      <td>Lung Parenchyma</td>
      <td>317</td>
      <td>24672</td>
    </tr>
    <tr>
      <th>9dbab10c-118d-496b-966a-67f1763a6b7d</th>
      <td>49014</td>
      <td>COVID-19 immune features revealed by a large-s...</td>
      <td>Large-scale single-cell analysis reveals criti...</td>
      <td>63</td>
      <td>27705</td>
    </tr>
    <tr>
      <th>9968be68-ab65-4a38-9e1a-c9b6abece194</th>
      <td>47909</td>
      <td>Charting human development using a multi-endod...</td>
      <td>Developing Human Atlas</td>
      <td>508</td>
      <td>27243</td>
    </tr>
    <tr>
      <th>3de0ad6d-4378-4f62-b37b-ec0b75a50d94</th>
      <td>46500</td>
      <td>LungMAP — Human data from a broad age healthy ...</td>
      <td>Single-cell multiomic profiling of human lungs...</td>
      <td>188</td>
      <td>26323</td>
    </tr>
    <tr>
      <th>2f132ec9-24b5-422f-9be0-ccef03b4fe28</th>
      <td>39778</td>
      <td>SARS-CoV-2 receptor ACE2 and TMPRSS2 are prima...</td>
      <td>Lung</td>
      <td>104</td>
      <td>27181</td>
    </tr>
    <tr>
      <th>1e5bd3b8-6a0e-4959-8d69-cafed30fe814</th>
      <td>35699</td>
      <td>Emphysema Cell Atlas</td>
      <td>immune cells</td>
      <td>51</td>
      <td>23203</td>
    </tr>
    <tr>
      <th>53d208b0-2cfd-4366-9866-c3c6114081bc</th>
      <td>35682</td>
      <td>Tabula Sapiens</td>
      <td>Tabula Sapiens - All Cells</td>
      <td>339</td>
      <td>57030</td>
    </tr>
    <tr>
      <th>1b9d8702-5af8-4142-85ed-020eb06ec4f6</th>
      <td>35419</td>
      <td>Cross-tissue immune cell analysis reveals tiss...</td>
      <td>Global</td>
      <td>314</td>
      <td>32610</td>
    </tr>
    <tr>
      <th>4ed927e9-c099-49af-b8ce-a2652d069333</th>
      <td>35284</td>
      <td>Single-nucleus cross-tissue molecular referenc...</td>
      <td>Single-nucleus cross-tissue molecular referenc...</td>
      <td>62</td>
      <td>29620</td>
    </tr>
    <tr>
      <th>2adb1f8a-a6b1-4909-8ee8-484814e2d4bf</th>
      <td>33698</td>
      <td>Construction of a human cell landscape at sing...</td>
      <td>Construction of a human cell landscape at sing...</td>
      <td>244</td>
      <td>26454</td>
    </tr>
    <tr>
      <th>4b6af54a-4a21-46e0-bc8d-673c0561a836</th>
      <td>15495</td>
      <td>Emphysema Cell Atlas</td>
      <td>non-immune cells</td>
      <td>53</td>
      <td>23705</td>
    </tr>
    <tr>
      <th>01209dce-3575-4bed-b1df-129f57fbc031</th>
      <td>11059</td>
      <td>Single-cell transcriptomics of human T cells r...</td>
      <td>Single-cell transcriptomics of human T cells r...</td>
      <td>186</td>
      <td>38676</td>
    </tr>
    <tr>
      <th>e04daea4-4412-45b5-989e-76a9be070a89</th>
      <td>8657</td>
      <td>A molecular cell atlas of the human lung from ...</td>
      <td>Krasnow Lab Human Lung Cell Atlas, Smart-seq2</td>
      <td>176</td>
      <td>47307</td>
    </tr>
    <tr>
      <th>214bf9eb-93db-48c8-8e3c-9bb22fa3bc63</th>
      <td>3662</td>
      <td>Emphysema Cell Atlas</td>
      <td>AT2 cells</td>
      <td>52</td>
      <td>21641</td>
    </tr>
    <tr>
      <th>f9846bb4-784d-4582-92c1-3f279e4c6f0c</th>
      <td>176</td>
      <td>A human fetal lung cell atlas uncovers proxima...</td>
      <td>Fibroblast and smooth muscle</td>
      <td>71</td>
      <td>23740</td>
    </tr>
    <tr>
      <th>f64e1be1-de15-4d27-8da4-82225cd4c035</th>
      <td>55</td>
      <td>HTAN MSK - Single cell profiling reveals novel...</td>
      <td>Immune cells</td>
      <td>520</td>
      <td>22552</td>
    </tr>
    <tr>
      <th>810ac45f-8969-4698-b42c-652f802f75c2</th>
      <td>10</td>
      <td>A human fetal lung cell atlas uncovers proxima...</td>
      <td>Endothelium</td>
      <td>77</td>
      <td>20594</td>
    </tr>
    <tr>
      <th>0ba16f4b-cb87-4fa3-9363-19fc51eec6e7</th>
      <td>4</td>
      <td>A human fetal lung cell atlas uncovers proxima...</td>
      <td>Myeloid</td>
      <td>74</td>
      <td>19952</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<p>You can see the genes measured in each dataset represented in <code class="docutils literal notranslate"><span class="pre">genes_measured</span></code>. Now lets get the <strong>genes that were measured in all datasets</strong>.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[16]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>var_somaid = np.nonzero(presence_matrix.sum(axis=0).A1 == presence_matrix.shape[0])[0].tolist()
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[17]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>lung_var = lung_var.query(f&quot;soma_joinid in {var_somaid}&quot;)
lung_var
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[17]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>soma_joinid</th>
      <th>feature_id</th>
      <th>feature_name</th>
      <th>feature_length</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>6</th>
      <td>6</td>
      <td>ENSG00000177757</td>
      <td>FAM87B</td>
      <td>1947</td>
    </tr>
    <tr>
      <th>8</th>
      <td>8</td>
      <td>ENSG00000230368</td>
      <td>FAM41C</td>
      <td>1971</td>
    </tr>
    <tr>
      <th>10</th>
      <td>10</td>
      <td>ENSG00000187634</td>
      <td>SAMD11</td>
      <td>4172</td>
    </tr>
    <tr>
      <th>11</th>
      <td>11</td>
      <td>ENSG00000188976</td>
      <td>NOC2L</td>
      <td>5540</td>
    </tr>
    <tr>
      <th>12</th>
      <td>12</td>
      <td>ENSG00000187961</td>
      <td>KLHL17</td>
      <td>3402</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>27743</th>
      <td>27743</td>
      <td>ENSG00000160838</td>
      <td>LRRC71</td>
      <td>2649</td>
    </tr>
    <tr>
      <th>30433</th>
      <td>30433</td>
      <td>ENSG00000136918</td>
      <td>WDR38</td>
      <td>1387</td>
    </tr>
    <tr>
      <th>30816</th>
      <td>30816</td>
      <td>ENSG00000149021</td>
      <td>SCGB1A1</td>
      <td>584</td>
    </tr>
    <tr>
      <th>31272</th>
      <td>31272</td>
      <td>ENSG00000187848</td>
      <td>P2RX2</td>
      <td>1975</td>
    </tr>
    <tr>
      <th>32538</th>
      <td>32538</td>
      <td>ENSG00000160472</td>
      <td>TMEM190</td>
      <td>605</td>
    </tr>
  </tbody>
</table>
<p>11516 rows × 4 columns</p>
</div></div>
</div>
<p>The number of rows represents the genes that were measured in all lung datasets.</p>
</section>
<section id="Summary-of-lung-metadata">
<h3>Summary of lung metadata<a class="headerlink" href="#Summary-of-lung-metadata" title="Permalink to this heading">¶</a></h3>
<p>In the previous sections, using the Census we learned the following information:</p>
<ul class="simple">
<li><p>The total number of unique lung cells and their composition for:</p>
<ul>
<li><p>Number of datasets.</p></li>
<li><p>Number sequencing technologies, most of which are 10x</p></li>
<li><p>Mostly human data, but some diseases exist, primarily “lung adenocarcinoma” and “COVID-19 infected”</p></li>
<li><p>No sex biases.</p></li>
<li><p>Mostly data from cells (~80%) rather than nucleus (~20%)</p></li>
</ul>
</li>
<li><p>A total of <strong>~12k</strong> genes were measured across all cells.</p></li>
</ul>
</section>
</section>
<section id="Fetching-all-single-cell-human-lung-data-from-the-Census">
<h2>Fetching all single-cell human lung data from the Census<a class="headerlink" href="#Fetching-all-single-cell-human-lung-data-from-the-Census" title="Permalink to this heading">¶</a></h2>
<p>Since loading the entire lung data is resource-intensive, for the sake of this exercise let’s load a subset of the lung data into an <code class="docutils literal notranslate"><span class="pre">anndata.AnnData</span></code> object and perform some exploratory analysis.</p>
<p>We will subset to 100,000 random unique cells using the <code class="docutils literal notranslate"><span class="pre">lung_obs</span></code> <code class="docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code> we previously created.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[18]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>lung_cell_subsampled_n = 100000
lung_cell_subsampled_ids = lung_obs[&quot;soma_joinid&quot;].sample(lung_cell_subsampled_n, random_state=1).tolist()
</pre></div>
</div>
</div>
<p>Now we can directly use the values of <code class="docutils literal notranslate"><span class="pre">soma_joinid</span></code> for querying the Census data and obtaining an <code class="docutils literal notranslate"><span class="pre">AnnData</span></code> object.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[19]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>lung_gene_ids = lung_var[&quot;soma_joinid&quot;].to_numpy()
lung_adata = cellxgene_census.get_anndata(
    census,
    organism=&quot;Homo sapiens&quot;,
    obs_coords=lung_cell_subsampled_ids,
    var_coords=lung_gene_ids,
)

lung_adata.var_names = lung_adata.var[&quot;feature_name&quot;]
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[20]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>lung_adata
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[20]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
AnnData object with n_obs × n_vars = 100000 × 11516
    obs: &#39;soma_joinid&#39;, &#39;dataset_id&#39;, &#39;assay&#39;, &#39;assay_ontology_term_id&#39;, &#39;cell_type&#39;, &#39;cell_type_ontology_term_id&#39;, &#39;development_stage&#39;, &#39;development_stage_ontology_term_id&#39;, &#39;disease&#39;, &#39;disease_ontology_term_id&#39;, &#39;donor_id&#39;, &#39;is_primary_data&#39;, &#39;self_reported_ethnicity&#39;, &#39;self_reported_ethnicity_ontology_term_id&#39;, &#39;sex&#39;, &#39;sex_ontology_term_id&#39;, &#39;suspension_type&#39;, &#39;tissue&#39;, &#39;tissue_ontology_term_id&#39;, &#39;tissue_general&#39;, &#39;tissue_general_ontology_term_id&#39;
    var: &#39;soma_joinid&#39;, &#39;feature_id&#39;, &#39;feature_name&#39;, &#39;feature_length&#39;
</pre></div></div>
</div>
<p>We are done with the census, so close it</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[21]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>census.close()
del census
</pre></div>
</div>
</div>
</section>
<section id="Calculating-QC-metrics-of-the-lung-data">
<h2>Calculating QC metrics of the lung data<a class="headerlink" href="#Calculating-QC-metrics-of-the-lung-data" title="Permalink to this heading">¶</a></h2>
<p>Now let’s take a look at some QC metrics</p>
<p><strong>Top genes per cell</strong></p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[22]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>sc.pl.highest_expr_genes(lung_adata, n_top=20)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/notebooks_analysis_demo_comp_bio_explore_and_load_lung_data_39_0.png" src="../../_images/notebooks_analysis_demo_comp_bio_explore_and_load_lung_data_39_0.png" />
</div>
</div>
<p><strong>Number of sequenced genes by assay</strong></p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[23]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>sc.pp.calculate_qc_metrics(lung_adata, percent_top=None, log1p=False, inplace=True)
sc.pl.violin(lung_adata, &quot;n_genes_by_counts&quot;, groupby=&quot;assay&quot;, jitter=0.4, rotation=90)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/notebooks_analysis_demo_comp_bio_explore_and_load_lung_data_41_0.png" src="../../_images/notebooks_analysis_demo_comp_bio_explore_and_load_lung_data_41_0.png" />
</div>
</div>
<p><strong>Total counts by assay</strong></p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[24]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>sc.pl.violin(lung_adata, &quot;total_counts&quot;, groupby=&quot;assay&quot;, jitter=0.4, rotation=90)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/notebooks_analysis_demo_comp_bio_explore_and_load_lung_data_43_0.png" src="../../_images/notebooks_analysis_demo_comp_bio_explore_and_load_lung_data_43_0.png" />
</div>
</div>
<p>You can see that Smart-Seq2 is an outlier for the total counts per cell, so let’s exlcude it to see how the rest of the assays look like</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[25]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>sc.pl.violin(
    lung_adata[lung_adata.obs[&quot;assay&quot;] != &quot;Smart-seq2&quot;,],
    &quot;total_counts&quot;,
    groupby=&quot;assay&quot;,
    jitter=0.4,
    rotation=90,
)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/notebooks_analysis_demo_comp_bio_explore_and_load_lung_data_45_0.png" src="../../_images/notebooks_analysis_demo_comp_bio_explore_and_load_lung_data_45_0.png" />
</div>
</div>
</section>
<section id="Creating-a-normalized-expression-layer-and-embeddings">
<h2>Creating a normalized expression layer and embeddings<a class="headerlink" href="#Creating-a-normalized-expression-layer-and-embeddings" title="Permalink to this heading">¶</a></h2>
<p>Let’s perform a bread and butter normalization and take a look at UMAP embeddings, but for all the data below we’ll exclude Smart-seq2 as this requires an extra step to normalize based on gene lengths</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[26]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>lung_adata = lung_adata[lung_adata.obs[&quot;assay&quot;] != &quot;Smart-seq2&quot;,].copy()
lung_adata.layers[&quot;counts&quot;] = lung_adata.X
</pre></div>
</div>
</div>
<p>Now let’s do some basic normalization:</p>
<ul class="simple">
<li><p>Normalize by sequencing depth</p></li>
<li><p>Transform to log-scale</p></li>
<li><p>Select 500 highly variable genes</p></li>
<li><p>Scale values across the gene axis</p></li>
</ul>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[27]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>sc.pp.normalize_total(lung_adata, target_sum=1e4)
sc.pp.log1p(lung_adata)
sc.pp.highly_variable_genes(lung_adata, n_top_genes=500, flavor=&quot;seurat_v3&quot;, layer=&quot;counts&quot;)
lung_adata = lung_adata[:, lung_adata.var.highly_variable]
sc.pp.scale(lung_adata, max_value=10)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/Users/pgarcianieto/opt/miniconda3/envs/census_latest/lib/python3.10/site-packages/scanpy/preprocessing/_highly_variable_genes.py:62: UserWarning: `flavor=&#39;seurat_v3&#39;` expects raw count data, but non-integers were found.
  warnings.warn(
/Users/pgarcianieto/opt/miniconda3/envs/census_latest/lib/python3.10/site-packages/scanpy/preprocessing/_simple.py:843: UserWarning: Received a view of an AnnData. Making a copy.
  view_to_actual(adata)
</pre></div></div>
</div>
<p>And reduce dimensionality by obtaining UMAP embeddings.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[28]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>sc.tl.pca(lung_adata)
sc.pp.neighbors(lung_adata)
sc.tl.umap(lung_adata)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
OMP: Info #276: omp_set_nested routine deprecated, please use omp_set_max_active_levels instead.
</pre></div></div>
</div>
<p>And plot these embeddings.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[29]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>n_cell_types = len(lung_adata.obs[&quot;cell_type&quot;].drop_duplicates())

from random import randint

colors = []

for i in range(len(lung_adata.obs[&quot;cell_type&quot;].drop_duplicates())):
    colors.append(&quot;#%06X&quot; % randint(0, 0xFFFFFF))
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[30]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>sc.pl.umap(lung_adata, color=&quot;cell_type&quot;, palette=colors, legend_loc=None)
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/Users/pgarcianieto/opt/miniconda3/envs/census_latest/lib/python3.10/site-packages/scanpy/plotting/_tools/scatterplots.py:392: UserWarning: No data for colormapping provided via &#39;c&#39;. Parameters &#39;cmap&#39; will be ignored
  cax = scatter(
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/notebooks_analysis_demo_comp_bio_explore_and_load_lung_data_54_1.png" src="../../_images/notebooks_analysis_demo_comp_bio_explore_and_load_lung_data_54_1.png" />
</div>
</div>
<p>Let’s color by assay.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[31]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>sc.pl.umap(lung_adata, color=&quot;assay&quot;)
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/Users/pgarcianieto/opt/miniconda3/envs/census_latest/lib/python3.10/site-packages/scanpy/plotting/_tools/scatterplots.py:392: UserWarning: No data for colormapping provided via &#39;c&#39;. Parameters &#39;cmap&#39; will be ignored
  cax = scatter(
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/notebooks_analysis_demo_comp_bio_explore_and_load_lung_data_56_1.png" src="../../_images/notebooks_analysis_demo_comp_bio_explore_and_load_lung_data_56_1.png" />
</div>
</div>
<p>Given the high number of cell types it makes it hard to visualize, so let’s look at the top 20 most abundant cell types.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[32]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>top_cell_types = lung_adata.obs[&quot;cell_type&quot;].value_counts()
top_cell_types = list(top_cell_types.reset_index().head(20)[&quot;cell_type&quot;])
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[33]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>lung_adata_top_cell_types = lung_adata[[i in top_cell_types for i in lung_adata.obs[&quot;cell_type&quot;]], :]
sc.pl.umap(lung_adata_top_cell_types, color=&quot;cell_type&quot;)
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/Users/pgarcianieto/opt/miniconda3/envs/census_latest/lib/python3.10/site-packages/scanpy/plotting/_tools/scatterplots.py:392: UserWarning: No data for colormapping provided via &#39;c&#39;. Parameters &#39;cmap&#39; will be ignored
  cax = scatter(
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/notebooks_analysis_demo_comp_bio_explore_and_load_lung_data_59_1.png" src="../../_images/notebooks_analysis_demo_comp_bio_explore_and_load_lung_data_59_1.png" />
</div>
</div>
<p>Let’s color by assay of this subset of the data.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[34]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>sc.pl.umap(lung_adata_top_cell_types, color=&quot;assay&quot;)
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/Users/pgarcianieto/opt/miniconda3/envs/census_latest/lib/python3.10/site-packages/scanpy/plotting/_tools/scatterplots.py:392: UserWarning: No data for colormapping provided via &#39;c&#39;. Parameters &#39;cmap&#39; will be ignored
  cax = scatter(
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/notebooks_analysis_demo_comp_bio_explore_and_load_lung_data_61_1.png" src="../../_images/notebooks_analysis_demo_comp_bio_explore_and_load_lung_data_61_1.png" />
</div>
</div>
</section>
</section>


            </div>
            
            </div>
            <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="comp_bio_data_integration_scvi.html" class="btn btn-neutral float-left" title="Integrating multi-dataset slices of data" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="comp_bio_normalizing_full_gene_sequencing.html" class="btn btn-neutral float-right" title="Normalizing full-length gene sequencing data" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2022-2023 Chan Zuckerberg Initiative.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
            </div>
        </div>

        </section>

    </div>
  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>