

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8">
  <meta name="generator" content="Docutils 0.19: https://docutils.sourceforge.io/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Learning about the CZ CELLxGENE Census &mdash; cellxgene-census  documentation</title>
  

  
  <link rel="icon" type="image/png" sizes="32x32" href="../../_static/img/favicon_32x32_v2.png"/>
  <link rel="icon" type="image/png" sizes="16x16" href="../../_static/img/favicon_16x16_v2.png"/>
  <link rel="preconnect" href="https://fonts.googleapis.com">
  <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
  <link href="https://fonts.googleapis.com/css2?family=Inter:wght@400;500;600;700&display=swap" rel="stylesheet">
  <link rel="stylesheet" href="../../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/css/custom.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/nbsphinx-code-cells.css" type="text/css" />

  
  
  
  

  
  <!--[if lt IE 9]>
    <script src="../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../../" src="../../_static/documentation_options.js"></script>
        <script data-url_root="../../" id="documentation_options" src="../../_static/documentation_options.js"></script>
        <script src="../../_static/doctools.js"></script>
        <script src="../../_static/sphinx_highlight.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
        <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
        <script data-domain="chanzuckerberg.github.io/cellxgene-census" defer="defer" src="https://plausible.io/js/script.js"></script>
    
    <script type="text/javascript" src="../../_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="Integrating multi-dataset slices of data" href="comp_bio_data_integration_scvi.html" />
    <link rel="prev" title="Exploring pre-calculated summary cell counts" href="../api_demo/census_summary_cell_counts.html" /> 
</head>

<body class="wy-body-for-nav">

  <div style="height: 100vh; overflow: hidden;">
     
    <!-- top 48px for navbar height -->
    <div class="navbar-cxg">
        <div style="padding: 0 16px; display: flex; height: inherit; align-items: center; justify-content: space-between;">
            <span style="display:flex; gap: 32px; align-items: center;">
                <a href="https://cellxgene.cziscience.com/">
                  <img src="../../_static/img/cellxgene-discover-logo.svg" />
                </a>
                <span style="display:flex; gap: 16px;">
                    <span class="navbar-cxg-link">
                        <a class="navbar-cxg-anchor" href="https://cellxgene.cziscience.com/collections">Collections</a>
                    </span>
                    <span class="navbar-cxg-link">
                        <a class="navbar-cxg-anchor" href="https://cellxgene.cziscience.com/datasets">Datasets</a>
                    </span>
                    <span class="navbar-cxg-link">
                        <a class="navbar-cxg-anchor" href="https://cellxgene.cziscience.com/gene-expression">Gene Expression</a>
                    </span>
                    <span class="navbar-cxg-link">
                      <a class="navbar-cxg-anchor" href="https://cellxgene.cziscience.com/cellguide">Cell Guide</a>
                      <div style="height: 16px!important; display: flex;">
                        <span class="beta">BETA</span>
                      </div>
                    </span>
                    <hr class="navbar-divider">
                    <span class="navbar-cxg-link">
                      <a class="navbar-cxg-anchor" href="/cellxgene-census/index.html">Census</a>
                  </span>
                </span>
            </span>

            <span class="navbar-cxg-link">
              <a class="navbar-cxg-anchor" href="https://cellxgene.cziscience.com/docs">Help & Documentation</a>
            </span>
        </div>
    </div>
    <div style="width: 100%; height: 100%; overflow: auto; padding-top: 48px;">
        
        <!-- top 48px for navbar height -->
        <nav data-toggle="wy-nav-shift" class="wy-nav-side" style="top: 48px;" >
        <div class="wy-side-scroll">
            <div class="wy-side-nav-search" >
            

            
                <a href="../../index.html" class="icon icon-home"> cellxgene-census
            

            
            </a>

            
                
                
                <div class="version">
                    v1.8.0
                </div>
                
            

            

            
            </div>

            
            <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
            
                
                
                
                
                
                <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../../cellxgene_census_docsite_installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../cellxgene_census_docsite_quick_start.html">Quick start</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../articles.html">What’s new?</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../cellxgene_census_aws_open_data.html">Census in AWS ☁️</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../cellxgene_census_docsite_schema.html">Census data and schema</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../cellxgene_census_docsite_data_release_info.html">Census data releases</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../python-api.html">Python API</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../../examples.html">Python tutorials</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="../../examples.html#api">API</a></li>
<li class="toctree-l2 current"><a class="reference internal" href="../../examples.html#analysis">Analysis</a><ul class="current">
<li class="toctree-l3 current"><a class="current reference internal" href="#">Learning about the CZ CELLxGENE Census</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#Opening-the-Census">Opening the Census</a></li>
<li class="toctree-l4"><a class="reference internal" href="#Census-organization">Census organization</a></li>
<li class="toctree-l4"><a class="reference internal" href="#Cell-metadata">Cell metadata</a></li>
<li class="toctree-l4"><a class="reference internal" href="#Gene-metadata">Gene metadata</a></li>
<li class="toctree-l4"><a class="reference internal" href="#Census-summary-content-tables">Census summary content tables</a></li>
<li class="toctree-l4"><a class="reference internal" href="#Understanding-Census-contents-beyond-the-summary-tables">Understanding Census contents beyond the summary tables</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="comp_bio_data_integration_scvi.html">Integrating multi-dataset slices of data</a></li>
<li class="toctree-l3"><a class="reference internal" href="comp_bio_explore_and_load_lung_data.html">Exploring all data from a tissue</a></li>
<li class="toctree-l3"><a class="reference internal" href="comp_bio_normalizing_full_gene_sequencing.html">Normalizing full-length gene sequencing data</a></li>
<li class="toctree-l3"><a class="reference internal" href="comp_bio_summarize_axis_query.html">Summarizing cell and gene metadata</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../examples.html#experimental">Experimental</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference external" href="https://chanzuckerberg.github.io/cellxgene-census/r/index.html">R API &amp; tutorials</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../cellxgene_census_docsite_FAQ.html">FAQ</a></li>
</ul>

                
            
            </div>
            
        </div>
        </nav>

        <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

        
        <nav class="wy-nav-top" aria-label="top navigation">
            
            <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
            <a href="../../index.html">cellxgene-census</a>
            
        </nav>


        <div class="wy-nav-content">
            
            <div class="rst-content">
            
            <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../../examples.html">Python tutorials</a></li>
      <li class="breadcrumb-item active">Learning about the CZ CELLxGENE Census</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../../_sources/notebooks/analysis_demo/comp_bio_census_info.ipynb.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
            <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
            <div itemprop="articleBody">
                
  <section id="Learning-about-the-CZ-CELLxGENE-Census">
<h1>Learning about the CZ CELLxGENE Census<a class="headerlink" href="#Learning-about-the-CZ-CELLxGENE-Census" title="Permalink to this heading">¶</a></h1>
<p>This notebook showcases the Census contents and how to obtain high-level information about it. It covers the organization of data within the Census, what cell and gene metadata are available, and it provides simple demonstrations to summarize cell counts across cell metadata.</p>
<p><strong>Contents</strong></p>
<ul class="simple">
<li><p>Opening the census</p></li>
<li><p>Census organization</p></li>
<li><p>Cell metadata</p></li>
<li><p>Gene metadata</p></li>
<li><p>Census summary content tables</p></li>
<li><p>Understanding Census contents beyond the summary tables</p></li>
</ul>
<p>⚠️ Note that the Census RNA data includes duplicate cells present across multiple datasets. Duplicate cells can be filtered in or out using the cell metadata variable <code class="docutils literal notranslate"><span class="pre">is_primary_data</span></code> which is described in the <a class="reference external" href="https://github.com/chanzuckerberg/cellxgene-census/blob/main/docs/cellxgene_census_schema.md#repeated-data">Census schema</a>.</p>
<section id="Opening-the-Census">
<h2>Opening the Census<a class="headerlink" href="#Opening-the-Census" title="Permalink to this heading">¶</a></h2>
<p>The <code class="docutils literal notranslate"><span class="pre">cellxgene_census</span></code> python package contains a convenient API to open the latest version of the Census. If you open the census, you should close it. <code class="docutils literal notranslate"><span class="pre">open_soma()</span></code> returns a context, so you can open/close it in several ways, like a Python file handle. The context manager is preferred, as it will automatically close upon an error raise.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">cellxgene_census</span>

<span class="c1"># Preferred: use a Python context manager</span>
<span class="k">with</span> <span class="n">cellxgene_census</span><span class="o">.</span><span class="n">open_soma</span><span class="p">()</span> <span class="k">as</span> <span class="n">census</span><span class="p">:</span>
    <span class="o">...</span>

<span class="c1"># or</span>
<span class="n">census</span> <span class="o">=</span> <span class="n">cellxgene_census</span><span class="o">.</span><span class="n">open_soma</span><span class="p">()</span>
<span class="o">...</span>
<span class="n">census</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
The &#34;stable&#34; release is currently 2023-07-25. Specify &#39;census_version=&#34;2023-07-25&#34;&#39; in future calls to open_soma() to ensure data consistency.
The &#34;stable&#34; release is currently 2023-07-25. Specify &#39;census_version=&#34;2023-07-25&#34;&#39; in future calls to open_soma() to ensure data consistency.
</pre></div></div>
</div>
<p>You can learn more about the <code class="docutils literal notranslate"><span class="pre">cellxgene_census</span></code> methods by accessing their corresponding documentation via <code class="docutils literal notranslate"><span class="pre">help()</span></code>. For example <code class="docutils literal notranslate"><span class="pre">help(cellxgene_census.open_soma)</span></code>.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">census</span> <span class="o">=</span> <span class="n">cellxgene_census</span><span class="o">.</span><span class="n">open_soma</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
The &#34;stable&#34; release is currently 2023-07-25. Specify &#39;census_version=&#34;2023-07-25&#34;&#39; in future calls to open_soma() to ensure data consistency.
</pre></div></div>
</div>
</section>
<section id="Census-organization">
<h2>Census organization<a class="headerlink" href="#Census-organization" title="Permalink to this heading">¶</a></h2>
<p>The <a class="reference external" href="https://chanzuckerberg.github.io/cellxgene-census/cellxgene_census_docsite_schema.html">Census schema</a> defines the structure of the Census. In short, you can think of the Census as a structured collection of items that stores different pieces of information. All of these items and the parent collection are SOMA objects of various types and can all be accessed with the <a class="reference external" href="https://github.com/single-cell-data/TileDB-SOMA">TileDB-SOMA API</a>
(<a class="reference external" href="https://tiledbsoma.readthedocs.io/en/latest/">documentation</a>).</p>
<p>The <code class="docutils literal notranslate"><span class="pre">cellxgene_census</span></code> package contains some convenient wrappers of the <code class="docutils literal notranslate"><span class="pre">TileDB-SOMA</span></code> API. An example of this is the function we used to open the Census: <code class="docutils literal notranslate"><span class="pre">cellxgene_census.open_soma()</span></code></p>
<section id="Main-Census-components">
<h3>Main Census components<a class="headerlink" href="#Main-Census-components" title="Permalink to this heading">¶</a></h3>
<p>With the command above you created <code class="docutils literal notranslate"><span class="pre">census</span></code>, which is a <code class="docutils literal notranslate"><span class="pre">SOMACollection</span></code>. It is analogous to a Python dictionary, and it has two items: <code class="docutils literal notranslate"><span class="pre">census_info</span></code> and <code class="docutils literal notranslate"><span class="pre">census_data</span></code>.</p>
<section id="Census-summary-info">
<h4>Census summary info<a class="headerlink" href="#Census-summary-info" title="Permalink to this heading">¶</a></h4>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">census[&quot;census_info&quot;]</span></code> A collection of tables providing information of the census as a whole.</p>
<ul>
<li><p><code class="docutils literal notranslate"><span class="pre">census[&quot;census_info&quot;][&quot;summary&quot;]</span></code>: A data frame with high-level information of this Census, e.g. build date, total cell count, etc.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">census[&quot;census_info&quot;][&quot;datasets&quot;]</span></code>: A data frame with all datasets from <a class="reference external" href="https://cellxgene.cziscience.com/">CELLxGENE Discover</a> used to create the Census.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">census[&quot;census_info&quot;][&quot;summary_cell_counts&quot;]</span></code>: A data frame with cell counts stratified by <strong>relevant</strong> cell metadata</p></li>
</ul>
</li>
</ul>
</section>
<section id="Census-data">
<h4>Census data<a class="headerlink" href="#Census-data" title="Permalink to this heading">¶</a></h4>
<p>Data for each organism is stored in independent <code class="docutils literal notranslate"><span class="pre">SOMAExperiment</span></code> objects which are a specialized form of a <code class="docutils literal notranslate"><span class="pre">SOMACollection</span></code>. Each of these store a data matrix (cell by genes), cell metadata, gene metadata, and some other useful components not covered in this notebook.</p>
<p>This is how the data is organized for one organism – <em>Homo sapiens</em>:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">census_obj[&quot;census_data&quot;][&quot;homo_sapiens&quot;].obs</span></code>: Cell metadata</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">census_obj[&quot;census_data&quot;][&quot;homo_sapiens&quot;].ms[&quot;RNA&quot;].X:</span></code> Data matrices, currently only raw counts exist <code class="docutils literal notranslate"><span class="pre">X[&quot;raw&quot;]</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">census_obj[&quot;census_data&quot;][&quot;homo_sapiens&quot;].ms[&quot;RNA&quot;].var:</span></code> Gene Metadata</p></li>
</ul>
</section>
</section>
</section>
<section id="Cell-metadata">
<h2>Cell metadata<a class="headerlink" href="#Cell-metadata" title="Permalink to this heading">¶</a></h2>
<p>You can obtain all cell metadata variables by directly querying the columns of the corresponding <code class="docutils literal notranslate"><span class="pre">SOMADataFrame</span></code>.</p>
<p>All of these variables can be used for querying the Census in case you want to work with specific cells.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">keys</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">census</span><span class="p">[</span><span class="s2">&quot;census_data&quot;</span><span class="p">][</span><span class="s2">&quot;homo_sapiens&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span>

<span class="n">keys</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
[&#39;soma_joinid&#39;,
 &#39;dataset_id&#39;,
 &#39;assay&#39;,
 &#39;assay_ontology_term_id&#39;,
 &#39;cell_type&#39;,
 &#39;cell_type_ontology_term_id&#39;,
 &#39;development_stage&#39;,
 &#39;development_stage_ontology_term_id&#39;,
 &#39;disease&#39;,
 &#39;disease_ontology_term_id&#39;,
 &#39;donor_id&#39;,
 &#39;is_primary_data&#39;,
 &#39;self_reported_ethnicity&#39;,
 &#39;self_reported_ethnicity_ontology_term_id&#39;,
 &#39;sex&#39;,
 &#39;sex_ontology_term_id&#39;,
 &#39;suspension_type&#39;,
 &#39;tissue&#39;,
 &#39;tissue_ontology_term_id&#39;,
 &#39;tissue_general&#39;,
 &#39;tissue_general_ontology_term_id&#39;]
</pre></div></div>
</div>
<p>All of these variables are defined in the <a class="reference external" href="https://github.com/chanzuckerberg/single-cell-curation/blob/main/schema/3.0.0/schema.md#obs-cell-metadata">CELLxGENE dataset schema</a> except for the following:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">soma_joinid</span></code>: a SOMA-defined value use for join operations.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">dataset_id</span></code>: the dataset id as encoded in <code class="docutils literal notranslate"><span class="pre">census[&quot;census-info&quot;][&quot;datasets&quot;]</span></code>.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">tissue_general</span></code> and <code class="docutils literal notranslate"><span class="pre">tissue_general_ontology_term_id</span></code>: the high-level tissue mapping.</p></li>
</ul>
</section>
<section id="Gene-metadata">
<h2>Gene metadata<a class="headerlink" href="#Gene-metadata" title="Permalink to this heading">¶</a></h2>
<p>Similarly, we can obtain all gene metadata variables by directly querying the columns of the corresponding <code class="docutils literal notranslate"><span class="pre">SOMADataFrame</span></code>.</p>
<p>These are the variables you can use for querying the Census in case there are specific genes you are interested in.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">keys</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">census</span><span class="p">[</span><span class="s2">&quot;census_data&quot;</span><span class="p">][</span><span class="s2">&quot;homo_sapiens&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">ms</span><span class="p">[</span><span class="s2">&quot;RNA&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span>

<span class="n">keys</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
[&#39;soma_joinid&#39;, &#39;feature_id&#39;, &#39;feature_name&#39;, &#39;feature_length&#39;]
</pre></div></div>
</div>
<p>All of these variables are defined in the <a class="reference external" href="https://github.com/chanzuckerberg/single-cell-curation/blob/main/schema/3.0.0/schema.md#var-and-rawvar-gene-metadata">CELLxGENE dataset schema</a> except for the following:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">soma_joinid</span></code>: a SOMA-defined value use for join operations.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">feature_length</span></code>: the length in base pairs of the gene.</p></li>
</ul>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">census_info</span> <span class="o">=</span> <span class="n">census</span><span class="p">[</span><span class="s2">&quot;census_info&quot;</span><span class="p">][</span><span class="s2">&quot;summary&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">read</span><span class="p">()</span><span class="o">.</span><span class="n">concat</span><span class="p">()</span><span class="o">.</span><span class="n">to_pandas</span><span class="p">()</span>

<span class="n">census_info</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>soma_joinid</th>
      <th>label</th>
      <th>value</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>0</td>
      <td>census_schema_version</td>
      <td>1.0.0</td>
    </tr>
    <tr>
      <th>1</th>
      <td>1</td>
      <td>census_build_date</td>
      <td>2023-07-25</td>
    </tr>
    <tr>
      <th>2</th>
      <td>2</td>
      <td>dataset_schema_version</td>
      <td>3.0.0</td>
    </tr>
    <tr>
      <th>3</th>
      <td>3</td>
      <td>total_cell_count</td>
      <td>61656118</td>
    </tr>
    <tr>
      <th>4</th>
      <td>4</td>
      <td>unique_cell_count</td>
      <td>37447773</td>
    </tr>
    <tr>
      <th>5</th>
      <td>5</td>
      <td>number_donors_homo_sapiens</td>
      <td>13035</td>
    </tr>
    <tr>
      <th>6</th>
      <td>6</td>
      <td>number_donors_mus_musculus</td>
      <td>1417</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
</section>
<section id="Census-summary-content-tables">
<h2>Census summary content tables<a class="headerlink" href="#Census-summary-content-tables" title="Permalink to this heading">¶</a></h2>
<p>You can take a quick look at the high-level Census information by looking at <code class="docutils literal notranslate"><span class="pre">census[&quot;census_info&quot;][&quot;summary&quot;]</span></code></p>
<p>Of special interest are the <code class="docutils literal notranslate"><span class="pre">label</span></code>-<code class="docutils literal notranslate"><span class="pre">value</span></code> combinations for :</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">total_cell_count</span></code> is the total number of cells in the Census.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">unique_cell_count</span></code> is the number of unique cells, as some cells may be present twice due to meta-analysis or consortia-like data.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">number_donors_homo_sapiens</span></code> and <code class="docutils literal notranslate"><span class="pre">number_donors_mus_musculus</span></code> are the number of individuals for human and mouse. These are not guaranteed to be unique as one individual ID may be present or identical in different datasets.</p></li>
</ul>
<section id="Cell-counts-by-cell-metadata">
<h3>Cell counts by cell metadata<a class="headerlink" href="#Cell-counts-by-cell-metadata" title="Permalink to this heading">¶</a></h3>
<p>By looking at <code class="docutils literal notranslate"><span class="pre">census[&quot;summary_cell_counts&quot;]</span></code> you can get a general idea of cell counts stratified by <strong>some relevant</strong> cell metadata. Not all cell metadata is included in this table, you can take a look at all cell and gene metadata available in the sections below “Cell metadata” and “Gene metadata”.</p>
<p>The line below retrieves this table and casts it into a <code class="docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code>.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">census_counts</span> <span class="o">=</span> <span class="n">census</span><span class="p">[</span><span class="s2">&quot;census_info&quot;</span><span class="p">][</span><span class="s2">&quot;summary_cell_counts&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">read</span><span class="p">()</span><span class="o">.</span><span class="n">concat</span><span class="p">()</span><span class="o">.</span><span class="n">to_pandas</span><span class="p">()</span>

<span class="n">census_counts</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>soma_joinid</th>
      <th>organism</th>
      <th>category</th>
      <th>ontology_term_id</th>
      <th>unique_cell_count</th>
      <th>total_cell_count</th>
      <th>label</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>0</td>
      <td>Homo sapiens</td>
      <td>all</td>
      <td>na</td>
      <td>33364242</td>
      <td>56400873</td>
      <td>na</td>
    </tr>
    <tr>
      <th>1</th>
      <td>1</td>
      <td>Homo sapiens</td>
      <td>assay</td>
      <td>EFO:0008722</td>
      <td>264166</td>
      <td>279635</td>
      <td>Drop-seq</td>
    </tr>
    <tr>
      <th>2</th>
      <td>2</td>
      <td>Homo sapiens</td>
      <td>assay</td>
      <td>EFO:0008780</td>
      <td>25652</td>
      <td>51304</td>
      <td>inDrop</td>
    </tr>
    <tr>
      <th>3</th>
      <td>3</td>
      <td>Homo sapiens</td>
      <td>assay</td>
      <td>EFO:0008919</td>
      <td>89477</td>
      <td>206754</td>
      <td>Seq-Well</td>
    </tr>
    <tr>
      <th>4</th>
      <td>4</td>
      <td>Homo sapiens</td>
      <td>assay</td>
      <td>EFO:0008931</td>
      <td>78750</td>
      <td>188248</td>
      <td>Smart-seq2</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>1357</th>
      <td>1357</td>
      <td>Mus musculus</td>
      <td>tissue_general</td>
      <td>UBERON:0002113</td>
      <td>179684</td>
      <td>208324</td>
      <td>kidney</td>
    </tr>
    <tr>
      <th>1358</th>
      <td>1358</td>
      <td>Mus musculus</td>
      <td>tissue_general</td>
      <td>UBERON:0002365</td>
      <td>15577</td>
      <td>31154</td>
      <td>exocrine gland</td>
    </tr>
    <tr>
      <th>1359</th>
      <td>1359</td>
      <td>Mus musculus</td>
      <td>tissue_general</td>
      <td>UBERON:0002367</td>
      <td>37715</td>
      <td>130135</td>
      <td>prostate gland</td>
    </tr>
    <tr>
      <th>1360</th>
      <td>1360</td>
      <td>Mus musculus</td>
      <td>tissue_general</td>
      <td>UBERON:0002368</td>
      <td>13322</td>
      <td>26644</td>
      <td>endocrine gland</td>
    </tr>
    <tr>
      <th>1361</th>
      <td>1361</td>
      <td>Mus musculus</td>
      <td>tissue_general</td>
      <td>UBERON:0002371</td>
      <td>90225</td>
      <td>144962</td>
      <td>bone marrow</td>
    </tr>
  </tbody>
</table>
<p>1362 rows × 7 columns</p>
</div></div>
</div>
<p>For each combination of <code class="docutils literal notranslate"><span class="pre">organism</span></code> and values for each <code class="docutils literal notranslate"><span class="pre">category</span></code> of cell metadata you can take a look at <code class="docutils literal notranslate"><span class="pre">total_cell_count</span></code> and <code class="docutils literal notranslate"><span class="pre">unique_cell_count</span></code> for the cell counts of that combination.</p>
<p>The values for each <code class="docutils literal notranslate"><span class="pre">category</span></code> are specified in <code class="docutils literal notranslate"><span class="pre">ontology_term_id</span></code> and <code class="docutils literal notranslate"><span class="pre">label</span></code>, which are the value’s IDs and labels, respectively.</p>
<section id="Example:-cell-metadata-included-in-the-summary-counts-table">
<h4>Example: cell metadata included in the summary counts table<a class="headerlink" href="#Example:-cell-metadata-included-in-the-summary-counts-table" title="Permalink to this heading">¶</a></h4>
<p>To get all the available cell metadata in the summary counts table you can do the following. Remember this is not all the cell metadata available, as some variables were omitted in the creation of this table.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">census_counts</span><span class="p">[[</span><span class="s2">&quot;organism&quot;</span><span class="p">,</span> <span class="s2">&quot;category&quot;</span><span class="p">]]</span><span class="o">.</span><span class="n">value_counts</span><span class="p">(</span><span class="n">sort</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
organism      category
Homo sapiens  all                          1
              assay                       19
              cell_type                  613
              disease                     64
              self_reported_ethnicity     26
              sex                          3
              suspension_type              1
              tissue                     220
              tissue_general              54
Mus musculus  all                          1
              assay                        9
              cell_type                  248
              disease                      5
              self_reported_ethnicity      1
              sex                          3
              suspension_type              1
              tissue                      66
              tissue_general              27
Name: count, dtype: int64
</pre></div></div>
</div>
</section>
<section id="Example:-cell-counts-for-each-sequencing-assay-in-human-data">
<h4>Example: cell counts for each sequencing assay in human data<a class="headerlink" href="#Example:-cell-counts-for-each-sequencing-assay-in-human-data" title="Permalink to this heading">¶</a></h4>
<p>To get the cell counts for each sequencing assay type in human data, you can perform the following <code class="docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code> operations:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">census_human_assays</span> <span class="o">=</span> <span class="n">census_counts</span><span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="s2">&quot;organism == &#39;Homo sapiens&#39; &amp; category == &#39;assay&#39;&quot;</span><span class="p">)</span>
<span class="n">census_human_assays</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="s2">&quot;total_cell_count&quot;</span><span class="p">,</span> <span class="n">ascending</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>soma_joinid</th>
      <th>organism</th>
      <th>category</th>
      <th>ontology_term_id</th>
      <th>unique_cell_count</th>
      <th>total_cell_count</th>
      <th>label</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>10</th>
      <td>10</td>
      <td>Homo sapiens</td>
      <td>assay</td>
      <td>EFO:0009922</td>
      <td>11845077</td>
      <td>25597563</td>
      <td>10x 3' v3</td>
    </tr>
    <tr>
      <th>7</th>
      <td>7</td>
      <td>Homo sapiens</td>
      <td>assay</td>
      <td>EFO:0009899</td>
      <td>7559102</td>
      <td>12638794</td>
      <td>10x 3' v2</td>
    </tr>
    <tr>
      <th>14</th>
      <td>14</td>
      <td>Homo sapiens</td>
      <td>assay</td>
      <td>EFO:0011025</td>
      <td>3872375</td>
      <td>6139786</td>
      <td>10x 5' v1</td>
    </tr>
    <tr>
      <th>13</th>
      <td>13</td>
      <td>Homo sapiens</td>
      <td>assay</td>
      <td>EFO:0010550</td>
      <td>4062980</td>
      <td>5064268</td>
      <td>sci-RNA-seq</td>
    </tr>
    <tr>
      <th>8</th>
      <td>8</td>
      <td>Homo sapiens</td>
      <td>assay</td>
      <td>EFO:0009900</td>
      <td>2930054</td>
      <td>3139770</td>
      <td>10x 5' v2</td>
    </tr>
    <tr>
      <th>17</th>
      <td>17</td>
      <td>Homo sapiens</td>
      <td>assay</td>
      <td>EFO:0030004</td>
      <td>915037</td>
      <td>1084235</td>
      <td>10x 5' transcription profiling</td>
    </tr>
    <tr>
      <th>16</th>
      <td>16</td>
      <td>Homo sapiens</td>
      <td>assay</td>
      <td>EFO:0030003</td>
      <td>744798</td>
      <td>811422</td>
      <td>10x 3' transcription profiling</td>
    </tr>
    <tr>
      <th>15</th>
      <td>15</td>
      <td>Homo sapiens</td>
      <td>assay</td>
      <td>EFO:0030002</td>
      <td>625175</td>
      <td>642559</td>
      <td>microwell-seq</td>
    </tr>
    <tr>
      <th>1</th>
      <td>1</td>
      <td>Homo sapiens</td>
      <td>assay</td>
      <td>EFO:0008722</td>
      <td>264166</td>
      <td>279635</td>
      <td>Drop-seq</td>
    </tr>
    <tr>
      <th>3</th>
      <td>3</td>
      <td>Homo sapiens</td>
      <td>assay</td>
      <td>EFO:0008919</td>
      <td>89477</td>
      <td>206754</td>
      <td>Seq-Well</td>
    </tr>
    <tr>
      <th>4</th>
      <td>4</td>
      <td>Homo sapiens</td>
      <td>assay</td>
      <td>EFO:0008931</td>
      <td>78750</td>
      <td>188248</td>
      <td>Smart-seq2</td>
    </tr>
    <tr>
      <th>18</th>
      <td>18</td>
      <td>Homo sapiens</td>
      <td>assay</td>
      <td>EFO:0700003</td>
      <td>146278</td>
      <td>177276</td>
      <td>BD Rhapsody Whole Transcriptome Analysis</td>
    </tr>
    <tr>
      <th>9</th>
      <td>9</td>
      <td>Homo sapiens</td>
      <td>assay</td>
      <td>EFO:0009901</td>
      <td>42397</td>
      <td>121394</td>
      <td>10x 3' v1</td>
    </tr>
    <tr>
      <th>12</th>
      <td>12</td>
      <td>Homo sapiens</td>
      <td>assay</td>
      <td>EFO:0010183</td>
      <td>58981</td>
      <td>117962</td>
      <td>single cell library construction</td>
    </tr>
    <tr>
      <th>19</th>
      <td>19</td>
      <td>Homo sapiens</td>
      <td>assay</td>
      <td>EFO:0700004</td>
      <td>96145</td>
      <td>96145</td>
      <td>BD Rhapsody Targeted mRNA</td>
    </tr>
    <tr>
      <th>2</th>
      <td>2</td>
      <td>Homo sapiens</td>
      <td>assay</td>
      <td>EFO:0008780</td>
      <td>25652</td>
      <td>51304</td>
      <td>inDrop</td>
    </tr>
    <tr>
      <th>6</th>
      <td>6</td>
      <td>Homo sapiens</td>
      <td>assay</td>
      <td>EFO:0008995</td>
      <td>0</td>
      <td>29128</td>
      <td>10x technology</td>
    </tr>
    <tr>
      <th>5</th>
      <td>5</td>
      <td>Homo sapiens</td>
      <td>assay</td>
      <td>EFO:0008953</td>
      <td>4693</td>
      <td>9386</td>
      <td>STRT-seq</td>
    </tr>
    <tr>
      <th>11</th>
      <td>11</td>
      <td>Homo sapiens</td>
      <td>assay</td>
      <td>EFO:0010010</td>
      <td>3105</td>
      <td>5244</td>
      <td>CEL-seq2</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
</section>
<section id="Example:-number-of-microglial-cells-in-the-Census">
<h4>Example: number of microglial cells in the Census<a class="headerlink" href="#Example:-number-of-microglial-cells-in-the-Census" title="Permalink to this heading">¶</a></h4>
<p>If you have a specific term from any of the categories shown above you can directly find out the number of cells for that term.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">census_counts</span><span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="s2">&quot;label == &#39;microglial cell&#39;&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>soma_joinid</th>
      <th>organism</th>
      <th>category</th>
      <th>ontology_term_id</th>
      <th>unique_cell_count</th>
      <th>total_cell_count</th>
      <th>label</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>69</th>
      <td>69</td>
      <td>Homo sapiens</td>
      <td>cell_type</td>
      <td>CL:0000129</td>
      <td>268114</td>
      <td>370771</td>
      <td>microglial cell</td>
    </tr>
    <tr>
      <th>1038</th>
      <td>1038</td>
      <td>Mus musculus</td>
      <td>cell_type</td>
      <td>CL:0000129</td>
      <td>48998</td>
      <td>62617</td>
      <td>microglial cell</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
</section>
</section>
</section>
<section id="Understanding-Census-contents-beyond-the-summary-tables">
<h2>Understanding Census contents beyond the summary tables<a class="headerlink" href="#Understanding-Census-contents-beyond-the-summary-tables" title="Permalink to this heading">¶</a></h2>
<p>While using the pre-computed tables in <code class="docutils literal notranslate"><span class="pre">census[&quot;census_info&quot;]</span></code> is an easy and quick way to understand the contents of the Census, it falls short if you want to learn more about certain slices of the Census.</p>
<p>For example, you may want to learn more about:</p>
<ul class="simple">
<li><p>What are the cell types available for human liver?</p></li>
<li><p>What are the total number of cells in all lung datasets stratified by sequencing technology?</p></li>
<li><p>What is the sex distribution of all cells from brain in mouse?</p></li>
<li><p>What are the diseases available for T cells?</p></li>
</ul>
<p>All of these questions can be answered by directly querying the cell metadata as shown in the examples below.</p>
<section id="Example:-all-cell-types-available-in-human">
<h3>Example: all cell types available in human<a class="headerlink" href="#Example:-all-cell-types-available-in-human" title="Permalink to this heading">¶</a></h3>
<p>To exemplify the process of accessing and slicing cell metadata for summary stats, let’s start with a trivial example and take a look at all human cell types available in the Census:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">human_cell_types</span> <span class="o">=</span> <span class="p">(</span>
    <span class="n">census</span><span class="p">[</span><span class="s2">&quot;census_data&quot;</span><span class="p">][</span><span class="s2">&quot;homo_sapiens&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">column_names</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;cell_type&quot;</span><span class="p">,</span> <span class="s2">&quot;is_primary_data&quot;</span><span class="p">])</span><span class="o">.</span><span class="n">concat</span><span class="p">()</span><span class="o">.</span><span class="n">to_pandas</span><span class="p">()</span>
<span class="p">)</span>
<span class="n">human_cell_types</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>cell_type</th>
      <th>is_primary_data</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>syncytiotrophoblast cell</td>
      <td>False</td>
    </tr>
    <tr>
      <th>1</th>
      <td>placental villous trophoblast</td>
      <td>False</td>
    </tr>
    <tr>
      <th>2</th>
      <td>syncytiotrophoblast cell</td>
      <td>False</td>
    </tr>
    <tr>
      <th>3</th>
      <td>syncytiotrophoblast cell</td>
      <td>False</td>
    </tr>
    <tr>
      <th>4</th>
      <td>extravillous trophoblast</td>
      <td>False</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>56400868</th>
      <td>pericyte</td>
      <td>True</td>
    </tr>
    <tr>
      <th>56400869</th>
      <td>pericyte</td>
      <td>True</td>
    </tr>
    <tr>
      <th>56400870</th>
      <td>pericyte</td>
      <td>True</td>
    </tr>
    <tr>
      <th>56400871</th>
      <td>pericyte</td>
      <td>True</td>
    </tr>
    <tr>
      <th>56400872</th>
      <td>pericyte</td>
      <td>True</td>
    </tr>
  </tbody>
</table>
<p>56400873 rows × 2 columns</p>
</div></div>
</div>
<p>The number of rows is the total number of cells for humans. Now, if you wish to get the cell counts per cell type we can perform some <code class="docutils literal notranslate"><span class="pre">pandas</span></code> operations on this object.</p>
<p>In addition, we will only focus on cells that are marked with <code class="docutils literal notranslate"><span class="pre">is_primary_data=True</span></code> as this ensures we de-duplicate cells that appear more than once in CELLxGENE Discover.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">human_cell_types</span> <span class="o">=</span> <span class="p">(</span>
    <span class="n">census</span><span class="p">[</span><span class="s2">&quot;census_data&quot;</span><span class="p">][</span><span class="s2">&quot;homo_sapiens&quot;</span><span class="p">]</span>
    <span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">column_names</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;cell_type&quot;</span><span class="p">],</span> <span class="n">value_filter</span><span class="o">=</span><span class="s2">&quot;is_primary_data == True&quot;</span><span class="p">)</span>
    <span class="o">.</span><span class="n">concat</span><span class="p">()</span>
    <span class="o">.</span><span class="n">to_pandas</span><span class="p">()</span>
<span class="p">)</span>

<span class="n">human_cell_types</span> <span class="o">=</span> <span class="n">human_cell_types</span><span class="p">[[</span><span class="s2">&quot;cell_type&quot;</span><span class="p">]]</span>
<span class="n">human_cell_types</span><span class="o">.</span><span class="n">shape</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
(33364242, 1)
</pre></div></div>
</div>
<p>This is the number of unique cells. Now let’s look at the counts per cell type:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">human_cell_type_counts</span> <span class="o">=</span> <span class="n">human_cell_types</span><span class="o">.</span><span class="n">value_counts</span><span class="p">()</span>
<span class="n">human_cell_type_counts</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
cell_type
neuron                                             2673669
glutamatergic neuron                               1541605
CD4-positive, alpha-beta T cell                    1258976
CD8-positive, alpha-beta T cell                    1235987
classical monocyte                                 1030996
                                                    ...
microfold cell of epithelium of small intestine         19
mature conventional dendritic cell                      17
serous cell of epithelium of bronchus                   15
sperm                                                   11
type N enteroendocrine cell                             10
Name: count, Length: 599, dtype: int64
</pre></div></div>
</div>
<p>This shows you that the most abundant cell types are “glutamatergic neuron”, “CD8-positive, alpha-beta T cell”, and “CD4-positive, alpha-beta T cell”.</p>
<p>Now let’s take a look at the number of unique cell types:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[13]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">human_cell_type_counts</span><span class="o">.</span><span class="n">shape</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[13]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
(599,)
</pre></div></div>
</div>
<p>That is the total number of different cell types for human.</p>
<p>All the information in this example can be quickly obtained from the summary table at <code class="docutils literal notranslate"><span class="pre">census[&quot;census-info&quot;][&quot;summary_cell_counts&quot;]</span></code>.</p>
<p>The examples below are more complex and can only be achieved by accessing the cell metadata.</p>
</section>
<section id="Example:-cell-types-available-in-human-liver">
<h3>Example: cell types available in human liver<a class="headerlink" href="#Example:-cell-types-available-in-human-liver" title="Permalink to this heading">¶</a></h3>
<p>Similar to the example above, we can learn what cell types are available for a specific tissue, e.g. liver.</p>
<p>To achieve this goal we just need to limit our cell metadata to that tissue. We will use the information in the cell metadata variable <code class="docutils literal notranslate"><span class="pre">tissue_general</span></code>. This variable contains the high-level tissue label for all cells in the Census:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">human_liver_cell_types</span> <span class="o">=</span> <span class="p">(</span>
    <span class="n">census</span><span class="p">[</span><span class="s2">&quot;census_data&quot;</span><span class="p">][</span><span class="s2">&quot;homo_sapiens&quot;</span><span class="p">]</span>
    <span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">column_names</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;cell_type&quot;</span><span class="p">],</span> <span class="n">value_filter</span><span class="o">=</span><span class="s2">&quot;is_primary_data == True and tissue_general == &#39;liver&#39;&quot;</span><span class="p">)</span>
    <span class="o">.</span><span class="n">concat</span><span class="p">()</span>
    <span class="o">.</span><span class="n">to_pandas</span><span class="p">()</span>
<span class="p">)</span>

<span class="n">human_liver_cell_types</span><span class="p">[</span><span class="s2">&quot;cell_type&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">value_counts</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
cell_type
T cell                               85739
hepatoblast                          58447
neoplastic cell                      52431
erythroblast                         45605
monocyte                             31388
                                     ...
pulmonary artery endothelial cell        1
germinal center B cell                   1
enteroendocrine cell                     1
type I pneumocyte                        1
group 2 innate lymphoid cell             1
Name: count, Length: 126, dtype: int64
</pre></div></div>
</div>
<p>These are the cell types and their cell counts in the human liver.</p>
</section>
<section id="Example:-diseased-T-cells-in-human-tissues">
<h3>Example: diseased T cells in human tissues<a class="headerlink" href="#Example:-diseased-T-cells-in-human-tissues" title="Permalink to this heading">¶</a></h3>
<p>In this example we are going to get the counts for all diseased cells annotated as T cells. For the sake of the example we will focus on “CD8-positive, alpha-beta T cell” and “CD4-positive, alpha-beta T cell”:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[15]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">t_cells_list</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;CD8-positive, alpha-beta T cell&quot;</span><span class="p">,</span> <span class="s2">&quot;CD4-positive, alpha-beta T cell&quot;</span><span class="p">]</span>

<span class="n">t_cells_diseased</span> <span class="o">=</span> <span class="p">(</span>
    <span class="n">census</span><span class="p">[</span><span class="s2">&quot;census_data&quot;</span><span class="p">][</span><span class="s2">&quot;homo_sapiens&quot;</span><span class="p">]</span>
    <span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">read</span><span class="p">(</span>
        <span class="n">column_names</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;disease&quot;</span><span class="p">,</span> <span class="s2">&quot;tissue_general&quot;</span><span class="p">],</span>
        <span class="n">value_filter</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;is_primary_data == True and cell_type in </span><span class="si">{</span><span class="n">t_cells_list</span><span class="si">}</span><span class="s2"> and disease != &#39;normal&#39;&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="o">.</span><span class="n">concat</span><span class="p">()</span>
    <span class="o">.</span><span class="n">to_pandas</span><span class="p">()</span>
<span class="p">)</span>

<span class="n">t_cells_diseased</span> <span class="o">=</span> <span class="n">t_cells_diseased</span><span class="p">[[</span><span class="s2">&quot;disease&quot;</span><span class="p">,</span> <span class="s2">&quot;tissue_general&quot;</span><span class="p">]]</span><span class="o">.</span><span class="n">value_counts</span><span class="p">(</span><span class="n">sort</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="n">t_cells_diseased</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[15]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
disease                                tissue_general
B-cell non-Hodgkin lymphoma            blood                  62499
COVID-19                               blood                 819428
                                       lung                   30578
                                       nose                      13
                                       respiratory system         4
                                       saliva                    41
Crohn disease                          colon                  17490
                                       small intestine        52029
Down syndrome                          bone marrow              181
breast cancer                          breast                  1850
chronic obstructive pulmonary disease  lung                    9382
chronic rhinitis                       nose                     909
clear cell renal carcinoma             blood                   6548
                                       kidney                 20540
                                       lymph node                36
cystic fibrosis                        lung                       7
follicular lymphoma                    lymph node              1089
influenza                              blood                   8871
interstitial lung disease              lung                    1803
kidney benign neoplasm                 blood                     20
                                       kidney                    10
kidney oncocytoma                      blood                     16
                                       kidney                  2408
lung adenocarcinoma                    adrenal gland            205
                                       brain                   3274
                                       liver                    507
                                       lung                  215013
                                       lymph node             24969
                                       pleural fluid          11558
lung large cell carcinoma              lung                    5922
lymphangioleiomyomatosis               lung                     513
non-small cell lung carcinoma          lung                   36573
nonpapillary renal cell carcinoma      adipose tissue           243
                                       adrenal gland           4828
                                       blood                    288
                                       blood clot              1717
                                       kidney                 69136
pleomorphic carcinoma                  lung                    1715
pneumonia                              lung                     856
pulmonary fibrosis                     lung                    1671
respiratory system disorder            blood                  34301
squamous cell lung carcinoma           lung                   52053
                                       lymph node               100
systemic lupus erythematosus           blood                 355471
Name: count, dtype: int64
</pre></div></div>
</div>
<p>These are the cell counts annotated with the indicated disease across human tissues for “CD8-positive, alpha-beta T cell” or “CD4-positive, alpha-beta T cell”.</p>
<p>And, don’t forget to close the census!</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[16]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">census</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
<span class="k">del</span> <span class="n">census</span>
</pre></div>
</div>
</div>
</section>
</section>
</section>


            </div>
            
            </div>
            <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="../api_demo/census_summary_cell_counts.html" class="btn btn-neutral float-left" title="Exploring pre-calculated summary cell counts" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="comp_bio_data_integration_scvi.html" class="btn btn-neutral float-right" title="Integrating multi-dataset slices of data" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2022-2023 Chan Zuckerberg Initiative Foundation.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
            </div>
        </div>

        </section>

    </div>
    <!-- Newsletter Banner -->
    <div role="banner" id="newsletter-banner">
      <span id="newsletter-subscribe-button" role="button">Subscribe</span>&nbsp;to our newsletter to receive updates about new features.
      <div id="newsletter-banner-close-button" role="button">X</div>
    </div>
    <!-- Newsletter Modal -->
    <dialog id="newsletter-modal">
      <div id="newsletter-header">
        <img id="newsletter-logo" src="../../_static/img/cellxGene-newsletter-logo.svg" />
        <div id="newsletter-close-button" role="button">X</div>
      </div>
      <div id="newsletter-content">
        <div id="newsletter-callout">Join Our Newsletter</div>
        <div id="newsletter-description">Get a quarterly email with the latest CELLxGENE features and data.</div>
        <!-- HubSpot Form target -->
        <div id="newsletter-form-container"></div>
      </div>

      <div id="newsletter-footnote">Unsubscribe at any time.</div>
    </dialog>
  </div>
  

  <script>
    // (thuang): 30 days
    const NEWSLETTER_BANNER_DISMISSED_TTL_MS = 30 * 24 * 60 * 60 * 1000;
    const NEWSLETTER_BANNER_DISMISSED_KEY = "newsletterBannerDismissed"

    var script = document.createElement('script');
    script.src = 'https://js.hsforms.net/forms/v2.js';
    script.defer = true;
    document.head.appendChild(script);

    // Run the code once the script is loaded
    script.onload = async function() {
      await hbspt.forms.create({
        region: "na1",
        portalId: "7272273",
        formId: "eb65b811-0451-414d-8304-7b9b6f468ce5",
        target: '#newsletter-form-container',
        onFormReady() {
          // get element by type "email"
          const emailInput = document.querySelector('#email-eb65b811-0451-414d-8304-7b9b6f468ce5');
          emailInput.setAttribute('placeholder', 'Enter email address');

          // remove the label element for emailInput
          const emailLabel = document.querySelector('#label-email-eb65b811-0451-414d-8304-7b9b6f468ce5');
          emailLabel.remove();
        },
        submitText: 'Subscribe',
      });
    };

    checkNewsletterBanner();

    document.querySelector('#newsletter-banner-close-button').addEventListener('click', () => {
      document.querySelector('#newsletter-banner').remove();
      localStorage.setItem(NEWSLETTER_BANNER_DISMISSED_KEY, Date.now());
    });

    const modal = document.querySelector('#newsletter-modal');

    document.querySelector('#newsletter-subscribe-button').addEventListener('click', () => {
      modal.showModal();
    });

    document.querySelector('#newsletter-close-button').addEventListener('click', () => {
      modal.close();
    });

    function checkNewsletterBanner() {
      /**
       * (thuang): Use LocalStorage to store dismissed state for 30 days
       * NOTE: Currently Census doc page doesn't share the same domain as the main site,
       * so dismissing the banner on the main site won't dismiss it on the Census doc page.
       * And vice versa.
       */
      const newsletterBannerDismissed = localStorage.getItem('newsletterBannerDismissed');

      if (newsletterBannerDismissed) {
        return;
      }

      if (newsletterBannerDismissed && Date.now() - newsletterBannerDismissed > NEWSLETTER_BANNER_DISMISSED_TTL_MS) {
        localStorage.removeItem(NEWSLETTER_BANNER_DISMISSED_KEY);
      }

      const newsletterBanner = document.querySelector('#newsletter-banner');

      if (!newsletterBannerDismissed) {
        newsletterBanner.style.display = 'flex';
      }
    }
  </script>

  
  
    
   

</body>
</html>