<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Access CELLxGENE collaboration embeddings (scVI, Geneformer) â€¢ cellxgene.census</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Access CELLxGENE collaboration embeddings (scVI, Geneformer)">
<script defer data-domain="chanzuckerberg.github.io/cellxgene-census" src="https://plausible.io/js/script.js"></script>
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">cellxgene.census</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.15.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="external-link nav-link" href="https://chanzuckerberg.github.io/cellxgene-census/">Main Census Site</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Use Census embeddings</h6></li>
    <li><a class="dropdown-item" href="../articles/census_access_maintained_embeddings.html">Access CELLxGENE collaboration embeddings (scVI, Geneformer)</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Explore and analyze the data</h6></li>
    <li><a class="dropdown-item" href="../articles/comp_bio_census_info.html">Learning about the CZ CELLxGENE Census</a></li>
    <li><a class="dropdown-item" href="../articles/comp_bio_summarize_axis_query.html">Summarizing cell and gene metadata</a></li>
    <li><a class="dropdown-item" href="../articles/comp_bio_data_integration.html">Integrating multi-dataset slices of data with Seurat</a></li>
    <li><a class="dropdown-item" href="../articles/comp_bio_normalizing_full_gene_sequencing.html">Normalizing full-length gene sequencing data</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>cellxgene.census capabilities</h6></li>
    <li><a class="dropdown-item" href="../articles/census_query_extract.html">Querying and fetching the single-cell data and cell/gene metadata</a></li>
    <li><a class="dropdown-item" href="../articles/census_citation_generation.html">Generating citations for Census slices</a></li>
    <li><a class="dropdown-item" href="../articles/census_datasets.html">Census Datasets example</a></li>
    <li><a class="dropdown-item" href="../articles/census_dataset_presence.html">Genes measured in each cell (dataset presence matrix)</a></li>
    <li><a class="dropdown-item" href="../articles/census_compute_over_X.html">Computing on X using online (incremental) algorithms</a></li>
    <li><a class="dropdown-item" href="../articles/census_access_maintained_embeddings.html">Access CELLxGENE collaboration embeddings (scVI, Geneformer)</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/chanzuckerberg/cellxgene-census/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">



<script src="census_access_maintained_embeddings_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Access CELLxGENE collaboration embeddings (scVI, Geneformer)</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/chanzuckerberg/cellxgene-census/tree/main/api/r/cellxgene.census/vignettes/census_access_maintained_embeddings.Rmd" class="external-link"><code>vignettes/census_access_maintained_embeddings.Rmd</code></a></small>
      <div class="d-none name"><code>census_access_maintained_embeddings.Rmd</code></div>
    </div>

    
    
<p>This notebook demonstrates basic access to CELLxGENE collaboration embeddings of CELLxGENE Discover Census. Currently, embeddings from scVI and a fine-tuned Geneformer model are maintained by CELLxGENE Discover. There are other CELLxGENE-hosted embeddings contributed by the community to CELLxGENE Discover, find out more about these in the <a href="https://cellxgene.cziscience.com/census-models" class="external-link">Census model page</a>.</p>
<p><strong>Contents</strong></p>
<p>Querying cells and loading embeddings as:</p>
<ol style="list-style-type: decimal">
<li>Seurat reductions</li>
<li>Bioconductor SingleCellExperiment reductions</li>
<li>sparseMatrix</li>
</ol>
<div class="section level2">
<h2 id="open-census">Open Census<a class="anchor" aria-label="anchor" href="#open-census"></a>
</h2>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/chanzuckerberg/cellxgene-census" class="external-link">"cellxgene.census"</a></span><span class="op">)</span></span>
<span><span class="va">census</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/open_soma.html">open_soma</a></span><span class="op">(</span>census_version <span class="op">=</span> <span class="st">"2023-12-15"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="load-embeddings-as-seurat-reductions">Load embeddings as Seurat reductions<a class="anchor" aria-label="anchor" href="#load-embeddings-as-seurat-reductions"></a>
</h2>
<p>The high-level <a href="https://chanzuckerberg.github.io/cellxgene-census/r/reference/get_seurat.html" class="external-link"><code>cellxgene.census::get_seurat()</code></a> function can both query the Census and load embeddings into <a href="https://satijalab.org/seurat/archive/v3.0/dim_reduction_vignette" class="external-link">dimensional reductions</a> of the Seurat object.</p>
<p>Here we will ask for a Seurat object with the expression data for all human cells of <code>tissue_general</code> equal to <code>'central nervous system'</code>, along with the scVI and geneformer embeddings (<code>obsm_layers</code>).</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://satijalab.org/seurat" class="external-link">"Seurat"</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">seurat_obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_seurat.html">get_seurat</a></span><span class="op">(</span></span>
<span>  <span class="va">census</span>,</span>
<span>  organism <span class="op">=</span> <span class="st">"homo_sapiens"</span>,</span>
<span>  obs_value_filter <span class="op">=</span> <span class="st">"tissue_general == 'central nervous system'"</span>,</span>
<span>  obs_column_names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"cell_type"</span><span class="op">)</span>,</span>
<span>  obsm_layers <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"scvi"</span>, <span class="st">"geneformer"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>With the embeddings stored as dimensional reductions on <code>seurat_obj</code>, we can take a quick look at the scVI embeddings in a 2D scatter plot via UMAP, colored by the Census <code>cell_type</code> annotations.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seurat_obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunUMAP.html" class="external-link">RunUMAP</a></span><span class="op">(</span></span>
<span>  <span class="va">seurat_obj</span>,</span>
<span>  reduction <span class="op">=</span> <span class="st">"scvi"</span>,</span>
<span>  dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/Embeddings.html" class="external-link">Embeddings</a></span><span class="op">(</span><span class="va">seurat_obj</span>, <span class="st">"scvi"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">seurat_obj</span>, reduction <span class="op">=</span> <span class="st">"umap"</span>, group.by <span class="op">=</span> <span class="st">"cell_type"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme</span><span class="op">(</span>legend.text <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>size <span class="op">=</span> <span class="fl">8</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="census_access_maintained_embeddings_files/figure-html/unnamed-chunk-4-1.png" width="672"></p>
</div>
<div class="section level2">
<h2 id="load-embeddings-as-singlecellexperiment-reductions">Load embeddings as SingleCellExperiment reductions<a class="anchor" aria-label="anchor" href="#load-embeddings-as-singlecellexperiment-reductions"></a>
</h2>
<p>Similarly, <a href="https://chanzuckerberg.github.io/cellxgene-census/r/reference/get_single_cell_experiment.html" class="external-link"><code>cellxgene.census::get_single_cell_experiment()</code></a> can query the Census and store embeddings as dimensionality reduction results on a Bioconductor <code>SingleCellExperiment</code> object.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st">"SingleCellExperiment"</span><span class="op">)</span></span>
<span><span class="va">sce_obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_single_cell_experiment.html">get_single_cell_experiment</a></span><span class="op">(</span></span>
<span>  <span class="va">census</span>,</span>
<span>  organism <span class="op">=</span> <span class="st">"homo_sapiens"</span>,</span>
<span>  obs_value_filter <span class="op">=</span> <span class="st">"tissue_general == 'central nervous system'"</span>,</span>
<span>  obs_column_names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"cell_type"</span><span class="op">)</span>,</span>
<span>  obsm_layers <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"scvi"</span>, <span class="st">"geneformer"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Then, we can view a UMAP of the Geneformer embeddings colored by <code>cell_type</code>.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce_obj</span> <span class="op">&lt;-</span> <span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/runUMAP.html" class="external-link">runUMAP</a></span><span class="op">(</span><span class="va">sce_obj</span>, dimred <span class="op">=</span> <span class="st">"geneformer"</span><span class="op">)</span></span>
<span><span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotReducedDim.html" class="external-link">plotReducedDim</a></span><span class="op">(</span><span class="va">sce_obj</span>, dimred <span class="op">=</span> <span class="st">"UMAP"</span>, colour_by <span class="op">=</span> <span class="st">"cell_type"</span><span class="op">)</span></span></code></pre></div>
<p><img src="census_access_maintained_embeddings_files/figure-html/unnamed-chunk-6-1.png" width="672"></p>
</div>
<div class="section level2">
<h2 id="load-embeddings-as-sparsematrix">Load embeddings as <code>sparseMatrix</code><a class="anchor" aria-label="anchor" href="#load-embeddings-as-sparsematrix"></a>
</h2>
<p>Lastly, we can use a <a href="https://single-cell-data.github.io/TileDB-SOMA/reference/SOMAExperimentAxisQuery.html#method-read-" class="external-link">SOMAExperimentAxisQuery</a> for lower-level access to the embeddingsâ€™ numerical data. This can be more performant for some use cases that donâ€™t need the other features of Seurat or SingleCellExperiment.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">query</span> <span class="op">&lt;-</span> <span class="va">census</span><span class="op">$</span><span class="fu">get</span><span class="op">(</span><span class="st">"census_data"</span><span class="op">)</span><span class="op">$</span><span class="fu">get</span><span class="op">(</span><span class="st">"homo_sapiens"</span><span class="op">)</span><span class="op">$</span><span class="fu">axis_query</span><span class="op">(</span></span>
<span>  <span class="st">"RNA"</span>,</span>
<span>  obs_query <span class="op">=</span> <span class="fu">tiledbsoma</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/tiledbsoma/man/SOMAAxisQuery.html" class="external-link">SOMAAxisQuery</a></span><span class="op">$</span><span class="fu">new</span><span class="op">(</span>value_filter <span class="op">=</span> <span class="st">"tissue == 'tongue'"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">embeddings</span> <span class="op">&lt;-</span> <span class="va">query</span><span class="op">$</span><span class="fu">to_sparse_matrix</span><span class="op">(</span><span class="st">"obsm"</span>, <span class="st">"geneformer"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">embeddings</span><span class="op">)</span></span>
<span><span class="co">#&gt; Formal class 'dgTMatrix' [package "Matrix"] with 6 slots</span></span>
<span><span class="co">#&gt;   ..@ i       : int [1:190464] 0 0 0 0 0 0 0 0 0 0 ...</span></span>
<span><span class="co">#&gt;   ..@ j       : int [1:190464] 0 1 2 3 4 5 6 7 8 9 ...</span></span>
<span><span class="co">#&gt;   ..@ Dim     : int [1:2] 372 512</span></span>
<span><span class="co">#&gt;   ..@ Dimnames:List of 2</span></span>
<span><span class="co">#&gt;   .. ..$ : chr [1:372] "51784858" "51784859" "51784860" "51784861" ...</span></span>
<span><span class="co">#&gt;   .. ..$ : chr [1:512] "0" "1" "2" "3" ...</span></span>
<span><span class="co">#&gt;   ..@ x       : num [1:190464] 0.1104 -1.2031 1.0078 0.0131 1.2422 ...</span></span>
<span><span class="co">#&gt;   ..@ factors : list()</span></span></code></pre></div>
<p>Each row of the embeddings <code>sparseMatrix</code> provides the fine-tuned Geneformer modelâ€™s 512-dimensional embedding vector for a cell, with the cell <code>soma_joinid</code>s in the row names. With different arguments, <a href="https://single-cell-data.github.io/TileDB-SOMA/reference/SOMAExperimentAxisQuery.html#method-SOMAExperimentAxisQuery-to_sparse_matrix" class="external-link"><code>SOMAExperimentAxisQuery$to_sparse_matrix()</code></a> can also be read the scVI embeddings or the expression data.</p>
<p>Still lower-level access is available through <a href="https://single-cell-data.github.io/TileDB-SOMA/reference/SOMAExperimentAxisQuery.html#method-read-" class="external-link"><code>SOMAExperimentAxisQuery$read()</code></a>, which streams Arrow tables. And other methods on <code>SOMAExperimentAxisQuery</code> can fetch metadata like <code>cell_type</code>:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="va">query</span><span class="op">$</span><span class="fu">obs</span><span class="op">(</span>column_names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"soma_joinid"</span>, <span class="st">"cell_type"</span><span class="op">)</span><span class="op">)</span><span class="op">$</span><span class="fu">concat</span><span class="op">(</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt;   soma_joinid  cell_type</span></span>
<span><span class="co">#&gt; 1    51784858 basal cell</span></span>
<span><span class="co">#&gt; 2    51784859 basal cell</span></span>
<span><span class="co">#&gt; 3    51784860 fibroblast</span></span>
<span><span class="co">#&gt; 4    51784861 fibroblast</span></span>
<span><span class="co">#&gt; 5    51784862 basal cell</span></span>
<span><span class="co">#&gt; 6    51784863 basal cell</span></span></code></pre></div>
<p>The <code>SOMAExperimentAxisQuery</code> loads only what you ask for from the Census, unlike the high-level <code><a href="../reference/get_seurat.html">get_seurat()</a></code> and <code><a href="../reference/get_single_cell_experiment.html">get_single_cell_experiment()</a></code> functions, which eagerly populate those objects based on your query.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">census</span><span class="op">$</span><span class="fu">close</span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Chan Zuckerberg Initiative Foundation.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

    </footer>
</div>





  </body>
</html>
