<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Genes measured in each cell (dataset presence matrix) • cellxgene.census</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Genes measured in each cell (dataset presence matrix)">
<script defer data-domain="chanzuckerberg.github.io/cellxgene-census" src="https://plausible.io/js/script.js"></script>
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">cellxgene.census</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.15.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="external-link nav-link" href="https://chanzuckerberg.github.io/cellxgene-census/">Main Census Site</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Use Census embeddings</h6></li>
    <li><a class="dropdown-item" href="../articles/census_access_maintained_embeddings.html">Access CELLxGENE collaboration embeddings (scVI, Geneformer)</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Explore and analyze the data</h6></li>
    <li><a class="dropdown-item" href="../articles/comp_bio_census_info.html">Learning about the CZ CELLxGENE Census</a></li>
    <li><a class="dropdown-item" href="../articles/comp_bio_summarize_axis_query.html">Summarizing cell and gene metadata</a></li>
    <li><a class="dropdown-item" href="../articles/comp_bio_data_integration.html">Integrating multi-dataset slices of data with Seurat</a></li>
    <li><a class="dropdown-item" href="../articles/comp_bio_normalizing_full_gene_sequencing.html">Normalizing full-length gene sequencing data</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>cellxgene.census capabilities</h6></li>
    <li><a class="dropdown-item" href="../articles/census_query_extract.html">Querying and fetching the single-cell data and cell/gene metadata</a></li>
    <li><a class="dropdown-item" href="../articles/census_citation_generation.html">Generating citations for Census slices</a></li>
    <li><a class="dropdown-item" href="../articles/census_datasets.html">Census Datasets example</a></li>
    <li><a class="dropdown-item" href="../articles/census_dataset_presence.html">Genes measured in each cell (dataset presence matrix)</a></li>
    <li><a class="dropdown-item" href="../articles/census_compute_over_X.html">Computing on X using online (incremental) algorithms</a></li>
    <li><a class="dropdown-item" href="../articles/census_access_maintained_embeddings.html">Access CELLxGENE collaboration embeddings (scVI, Geneformer)</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/chanzuckerberg/cellxgene-census/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">



<script src="census_dataset_presence_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Genes measured in each cell (dataset presence matrix)</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/chanzuckerberg/cellxgene-census/tree/main/api/r/cellxgene.census/vignettes/census_dataset_presence.Rmd" class="external-link"><code>vignettes/census_dataset_presence.Rmd</code></a></small>
      <div class="d-none name"><code>census_dataset_presence.Rmd</code></div>
    </div>

    
    
<!--
THIS VIGNETTE IS BASED ON:
https://github.com/chanzuckerberg/cellxgene-census/blob/main/api/python/notebooks/api_demo/census_dataset_presence.ipynb
-->
<p>The Census is a compilation of cells from multiple datasets that may differ by the sets of genes they measure. This notebook describes the way to identify the genes measured per dataset.</p>
<p>The presence matrix is a sparse boolean array, indicating which features (var) were present in each dataset. The array has dimensions [n_datasets, n_var], and is stored in the SOMA Measurement <code>varp</code> collection. The first dimension is indexed by the <code>soma_joinid</code> in the <code>census_datasets</code> dataframe. The second is indexed by the <code>soma_joinid</code> in the <code>var</code> dataframe of the measurement.</p>
<p>As a reminder the <code>obs</code> data frame has a column <code>dataset_id</code> that can be used to link any cell in the Census to the presence matrix.</p>
<p>Contents</p>
<ol style="list-style-type: decimal">
<li>Opening the Census.</li>
<li>Fetching the IDs of the Census datasets.</li>
<li>Fetching the dataset presence matrix.</li>
<li>Identifying genes measured in a specific dataset.</li>
<li>Identifying datasets that measured specific genes.</li>
<li>Identifying all genes measured in a dataset.</li>
</ol>
<div class="section level2">
<h2 id="opening-the-census">Opening the Census<a class="anchor" aria-label="anchor" href="#opening-the-census"></a>
</h2>
<p>The <code>cellxgene.census</code> R package contains a convenient API to open any version of the Census (by default, the newest stable version).</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/chanzuckerberg/cellxgene-census" class="external-link">"cellxgene.census"</a></span><span class="op">)</span></span>
<span><span class="va">census</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/open_soma.html">open_soma</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="fetching-the-ids-of-the-census-datasets">Fetching the IDs of the Census datasets<a class="anchor" aria-label="anchor" href="#fetching-the-ids-of-the-census-datasets"></a>
</h2>
<p>Let’s grab a table of all the datasets included in the Census and use this table in combination with the presence matrix below.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Grab the experiment containing human data, and the measurement therein with RNA</span></span>
<span><span class="va">human</span> <span class="op">&lt;-</span> <span class="va">census</span><span class="op">$</span><span class="fu">get</span><span class="op">(</span><span class="st">"census_data"</span><span class="op">)</span><span class="op">$</span><span class="fu">get</span><span class="op">(</span><span class="st">"homo_sapiens"</span><span class="op">)</span></span>
<span><span class="va">human_rna</span> <span class="op">&lt;-</span> <span class="va">human</span><span class="op">$</span><span class="va">ms</span><span class="op">$</span><span class="fu">get</span><span class="op">(</span><span class="st">"RNA"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># The census-wide datasets</span></span>
<span><span class="va">datasets_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="va">census</span><span class="op">$</span><span class="fu">get</span><span class="op">(</span><span class="st">"census_info"</span><span class="op">)</span><span class="op">$</span><span class="fu">get</span><span class="op">(</span><span class="st">"datasets"</span><span class="op">)</span><span class="op">$</span><span class="fu">read</span><span class="op">(</span><span class="op">)</span><span class="op">$</span><span class="fu">concat</span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">datasets_df</span><span class="op">)</span></span>
<span><span class="co">#&gt;    soma_joinid</span></span>
<span><span class="co">#&gt; 1            0</span></span>
<span><span class="co">#&gt; 2            1</span></span>
<span><span class="co">#&gt; 3            2</span></span>
<span><span class="co">#&gt; 4            3</span></span>
<span><span class="co">#&gt; 5            4</span></span>
<span><span class="co">#&gt; 6            5</span></span>
<span><span class="co">#&gt; 7            6</span></span>
<span><span class="co">#&gt; 8            7</span></span>
<span><span class="co">#&gt; 9            8</span></span>
<span><span class="co">#&gt; 10           9</span></span>
<span><span class="co">#&gt;                                                                                                                                                                                                                                                                                                            citation</span></span>
<span><span class="co">#&gt; 1             Publication: https://doi.org/10.1002/hep4.1854 Dataset Version: https://datasets.cellxgene.cziscience.com/fb76c95f-0391-4fac-9fb9-082ce2430b59.h5ad curated and distributed by CZ CELLxGENE Discover in Collection: https://cellxgene.cziscience.com/collections/44531dd9-1388-4416-a117-af0a99de2294</span></span>
<span><span class="co">#&gt; 2    Publication: https://doi.org/10.1126/sciimmunol.abe6291 Dataset Version: https://datasets.cellxgene.cziscience.com/b6737a5e-9069-4dd6-9a57-92e17a746df9.h5ad curated and distributed by CZ CELLxGENE Discover in Collection: https://cellxgene.cziscience.com/collections/3a2af25b-2338-4266-aad3-aa8d07473f50</span></span>
<span><span class="co">#&gt; 3    Publication: https://doi.org/10.1038/s41593-020-00764-7 Dataset Version: https://datasets.cellxgene.cziscience.com/0e02290f-b992-450b-8a19-554f73cd7f09.h5ad curated and distributed by CZ CELLxGENE Discover in Collection: https://cellxgene.cziscience.com/collections/180bff9c-c8a5-4539-b13b-ddbc00d643e6</span></span>
<span><span class="co">#&gt; 4    Publication: https://doi.org/10.1038/s41467-022-29450-x Dataset Version: https://datasets.cellxgene.cziscience.com/40832710-d7b1-43fb-b2c2-1cd2255bc3ac.h5ad curated and distributed by CZ CELLxGENE Discover in Collection: https://cellxgene.cziscience.com/collections/bf325905-5e8e-42e3-933d-9a9053e9af80</span></span>
<span><span class="co">#&gt; 5    Publication: https://doi.org/10.1038/s41590-021-01059-0 Dataset Version: https://datasets.cellxgene.cziscience.com/eb6c070c-ff67-4c1f-8d4d-65f9fe2119ee.h5ad curated and distributed by CZ CELLxGENE Discover in Collection: https://cellxgene.cziscience.com/collections/93eebe82-d8c3-41bc-a906-63b5b5f24a9d</span></span>
<span><span class="co">#&gt; 6  Publication: https://doi.org/10.1016/j.celrep.2019.12.082 Dataset Version: https://datasets.cellxgene.cziscience.com/650a47be-6666-4f70-ac47-8414c50bbd8e.h5ad curated and distributed by CZ CELLxGENE Discover in Collection: https://cellxgene.cziscience.com/collections/939769a8-d8d2-4d01-abfc-55699893fd49</span></span>
<span><span class="co">#&gt; 7    Publication: https://doi.org/10.1038/s41593-020-00764-7 Dataset Version: https://datasets.cellxgene.cziscience.com/1f0cd8ed-94c6-440c-bd5b-bad55e2666b1.h5ad curated and distributed by CZ CELLxGENE Discover in Collection: https://cellxgene.cziscience.com/collections/180bff9c-c8a5-4539-b13b-ddbc00d643e6</span></span>
<span><span class="co">#&gt; 8    Publication: https://doi.org/10.1016/j.cell.2022.11.005 Dataset Version: https://datasets.cellxgene.cziscience.com/086c8f4e-3fe7-46e5-8b6b-a8cb7f92dadd.h5ad curated and distributed by CZ CELLxGENE Discover in Collection: https://cellxgene.cziscience.com/collections/2d2e2acd-dade-489f-a2da-6c11aa654028</span></span>
<span><span class="co">#&gt; 9    Publication: https://doi.org/10.1016/j.jhep.2023.12.023 Dataset Version: https://datasets.cellxgene.cziscience.com/7e88c4aa-d95f-43e3-bc8b-cf02629d7301.h5ad curated and distributed by CZ CELLxGENE Discover in Collection: https://cellxgene.cziscience.com/collections/0c8a364b-97b5-4cc8-a593-23c38c6f0ac5</span></span>
<span><span class="co">#&gt; 10            Publication: https://doi.org/10.1002/hep4.1854 Dataset Version: https://datasets.cellxgene.cziscience.com/575dd70a-45b1-4a82-8acd-467d313e4c66.h5ad curated and distributed by CZ CELLxGENE Discover in Collection: https://cellxgene.cziscience.com/collections/44531dd9-1388-4416-a117-af0a99de2294</span></span>
<span><span class="co">#&gt;                           collection_id</span></span>
<span><span class="co">#&gt; 1  44531dd9-1388-4416-a117-af0a99de2294</span></span>
<span><span class="co">#&gt; 2  3a2af25b-2338-4266-aad3-aa8d07473f50</span></span>
<span><span class="co">#&gt; 3  180bff9c-c8a5-4539-b13b-ddbc00d643e6</span></span>
<span><span class="co">#&gt; 4  bf325905-5e8e-42e3-933d-9a9053e9af80</span></span>
<span><span class="co">#&gt; 5  93eebe82-d8c3-41bc-a906-63b5b5f24a9d</span></span>
<span><span class="co">#&gt; 6  939769a8-d8d2-4d01-abfc-55699893fd49</span></span>
<span><span class="co">#&gt; 7  180bff9c-c8a5-4539-b13b-ddbc00d643e6</span></span>
<span><span class="co">#&gt; 8  2d2e2acd-dade-489f-a2da-6c11aa654028</span></span>
<span><span class="co">#&gt; 9  0c8a364b-97b5-4cc8-a593-23c38c6f0ac5</span></span>
<span><span class="co">#&gt; 10 44531dd9-1388-4416-a117-af0a99de2294</span></span>
<span><span class="co">#&gt;                                                                                                                                         collection_name</span></span>
<span><span class="co">#&gt; 1                     Single-Cell, Single-Nucleus, and Spatial RNA Sequencing of the Human Liver Identifies Cholangiocyte and Mesenchymal Heterogeneity</span></span>
<span><span class="co">#&gt; 2                                       Single-cell analysis of human B cell maturation predicts how antibody class switching shapes selection dynamics</span></span>
<span><span class="co">#&gt; 3                                                                   Molecular characterization of selectively vulnerable neurons in Alzheimer's Disease</span></span>
<span><span class="co">#&gt; 4                   Single-cell Atlas of common variable immunodeficiency shows germinal center-associated epigenetic dysregulation in B-cell responses</span></span>
<span><span class="co">#&gt; 5  Single-cell proteo-genomic reference maps of the hematopoietic system enable the purification and massive profiling of precisely defined cell states</span></span>
<span><span class="co">#&gt; 6                             Integration of eQTL and a Single-Cell Atlas in the Human Eye Identifies Causal Genes for Age-Related Macular Degeneration</span></span>
<span><span class="co">#&gt; 7                                                                   Molecular characterization of selectively vulnerable neurons in Alzheimer's Disease</span></span>
<span><span class="co">#&gt; 8                            A human fetal lung cell atlas uncovers proximal-distal gradients of differentiation and key regulators of epithelial fates</span></span>
<span><span class="co">#&gt; 9                            Single-cell and spatial transcriptomics characterisation of the immunological landscape in the healthy and PSC human liver</span></span>
<span><span class="co">#&gt; 10                    Single-Cell, Single-Nucleus, and Spatial RNA Sequencing of the Human Liver Identifies Cholangiocyte and Mesenchymal Heterogeneity</span></span>
<span><span class="co">#&gt;                  collection_doi                           dataset_id</span></span>
<span><span class="co">#&gt; 1             10.1002/hep4.1854 0895c838-e550-48a3-a777-dbcd35d30272</span></span>
<span><span class="co">#&gt; 2    10.1126/sciimmunol.abe6291 00ff600e-6e2e-4d76-846f-0eec4f0ae417</span></span>
<span><span class="co">#&gt; 3    10.1038/s41593-020-00764-7 bdacc907-7c26-419f-8808-969eab3ca2e8</span></span>
<span><span class="co">#&gt; 4    10.1038/s41467-022-29450-x a5d95a42-0137-496f-8a60-101e17f263c8</span></span>
<span><span class="co">#&gt; 5    10.1038/s41590-021-01059-0 d3566d6a-a455-4a15-980f-45eb29114cab</span></span>
<span><span class="co">#&gt; 6  10.1016/j.celrep.2019.12.082 de17ac25-550a-4018-be75-bbb485a0636e</span></span>
<span><span class="co">#&gt; 7    10.1038/s41593-020-00764-7 9f1049ac-f8b7-45ad-8e31-6e96c3e5058f</span></span>
<span><span class="co">#&gt; 8    10.1016/j.cell.2022.11.005 703f00e6-b996-48e5-bc34-00c41b9876f4</span></span>
<span><span class="co">#&gt; 9    10.1016/j.jhep.2023.12.023 e347396c-a7ff-4691-9f7a-99a43555ca18</span></span>
<span><span class="co">#&gt; 10            10.1002/hep4.1854 524e045e-e74c-4e00-9884-d5c3bef3d862</span></span>
<span><span class="co">#&gt;                      dataset_version_id</span></span>
<span><span class="co">#&gt; 1  fb76c95f-0391-4fac-9fb9-082ce2430b59</span></span>
<span><span class="co">#&gt; 2  b6737a5e-9069-4dd6-9a57-92e17a746df9</span></span>
<span><span class="co">#&gt; 3  0e02290f-b992-450b-8a19-554f73cd7f09</span></span>
<span><span class="co">#&gt; 4  40832710-d7b1-43fb-b2c2-1cd2255bc3ac</span></span>
<span><span class="co">#&gt; 5  eb6c070c-ff67-4c1f-8d4d-65f9fe2119ee</span></span>
<span><span class="co">#&gt; 6  650a47be-6666-4f70-ac47-8414c50bbd8e</span></span>
<span><span class="co">#&gt; 7  1f0cd8ed-94c6-440c-bd5b-bad55e2666b1</span></span>
<span><span class="co">#&gt; 8  086c8f4e-3fe7-46e5-8b6b-a8cb7f92dadd</span></span>
<span><span class="co">#&gt; 9  7e88c4aa-d95f-43e3-bc8b-cf02629d7301</span></span>
<span><span class="co">#&gt; 10 575dd70a-45b1-4a82-8acd-467d313e4c66</span></span>
<span><span class="co">#&gt;                                                                                                dataset_title</span></span>
<span><span class="co">#&gt; 1                                                                               Healthy human liver: B cells</span></span>
<span><span class="co">#&gt; 2                                                                       Human tonsil nonlymphoid cells scRNA</span></span>
<span><span class="co">#&gt; 3  Molecular characterization of selectively vulnerable neurons in Alzheimer&lt;U+2019&gt;s Disease: SFG microglia</span></span>
<span><span class="co">#&gt; 4                                                                           Steady-state B cells - scRNA-seq</span></span>
<span><span class="co">#&gt; 5                                                           blood and bone marrow from a healthy young donor</span></span>
<span><span class="co">#&gt; 6                                                                                 Myeloid cells of human eye</span></span>
<span><span class="co">#&gt; 7   Molecular characterization of selectively vulnerable neurons in Alzheimer&lt;U+2019&gt;s Disease: EC microglia</span></span>
<span><span class="co">#&gt; 8                                                                                                        PNS</span></span>
<span><span class="co">#&gt; 9                                                      Stellate cells from human healthy donor liver samples</span></span>
<span><span class="co">#&gt; 10                                                               Healthy human liver: hepatic stellate cells</span></span>
<span><span class="co">#&gt;                            dataset_h5ad_path dataset_total_cell_count</span></span>
<span><span class="co">#&gt; 1  0895c838-e550-48a3-a777-dbcd35d30272.h5ad                      146</span></span>
<span><span class="co">#&gt; 2  00ff600e-6e2e-4d76-846f-0eec4f0ae417.h5ad                      363</span></span>
<span><span class="co">#&gt; 3  bdacc907-7c26-419f-8808-969eab3ca2e8.h5ad                     3799</span></span>
<span><span class="co">#&gt; 4  a5d95a42-0137-496f-8a60-101e17f263c8.h5ad                     1324</span></span>
<span><span class="co">#&gt; 5  d3566d6a-a455-4a15-980f-45eb29114cab.h5ad                    15502</span></span>
<span><span class="co">#&gt; 6  de17ac25-550a-4018-be75-bbb485a0636e.h5ad                      395</span></span>
<span><span class="co">#&gt; 7  9f1049ac-f8b7-45ad-8e31-6e96c3e5058f.h5ad                     5572</span></span>
<span><span class="co">#&gt; 8  703f00e6-b996-48e5-bc34-00c41b9876f4.h5ad                      649</span></span>
<span><span class="co">#&gt; 9  e347396c-a7ff-4691-9f7a-99a43555ca18.h5ad                     1417</span></span>
<span><span class="co">#&gt; 10 524e045e-e74c-4e00-9884-d5c3bef3d862.h5ad                     1374</span></span>
<span><span class="co">#&gt;  [ reached 'max' / getOption("max.print") -- omitted 802 rows ]</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="fetching-the-dataset-presence-matrix">Fetching the dataset presence matrix<a class="anchor" aria-label="anchor" href="#fetching-the-dataset-presence-matrix"></a>
</h2>
<p>Now let’s fetch the dataset presence matrix.</p>
<p>For convenience, read the entire presence matrix (for Homo sapiens) into a sparse matrix. There is a convenience function providing this capability:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">presence_matrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_presence_matrix.html">get_presence_matrix</a></span><span class="op">(</span><span class="va">census</span>, <span class="st">"Homo sapiens"</span>, <span class="st">"RNA"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">presence_matrix</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; NULL</span></span></code></pre></div>
<p>We also need the <code>var</code> dataframe, which is read into an R data frame for convenient manipulation:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">var_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="va">human_rna</span><span class="op">$</span><span class="va">var</span><span class="op">$</span><span class="fu">read</span><span class="op">(</span><span class="op">)</span><span class="op">$</span><span class="fu">concat</span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">var_df</span><span class="op">)</span></span>
<span><span class="co">#&gt;    soma_joinid      feature_id feature_name feature_length      nnz n_measured_obs</span></span>
<span><span class="co">#&gt; 1            0 ENSG00000000003       TSPAN6           4530  4530448       73855064</span></span>
<span><span class="co">#&gt; 2            1 ENSG00000000005         TNMD           1476   236059       61201828</span></span>
<span><span class="co">#&gt; 3            2 ENSG00000000419         DPM1           9276 17576462       74159149</span></span>
<span><span class="co">#&gt; 4            3 ENSG00000000457        SCYL3           6883  9117322       73988868</span></span>
<span><span class="co">#&gt; 5            4 ENSG00000000460     C1orf112           5970  6287794       73636201</span></span>
<span><span class="co">#&gt; 6            5 ENSG00000000938          FGR           3382  5667858       74061500</span></span>
<span><span class="co">#&gt; 7            6 ENSG00000000971          CFH          15284  4423029       74095467</span></span>
<span><span class="co">#&gt; 8            7 ENSG00000001036        FUCA2           2822 10129388       73988868</span></span>
<span><span class="co">#&gt; 9            8 ENSG00000001084         GCLC           8618 12662710       74226365</span></span>
<span><span class="co">#&gt; 10           9 ENSG00000001167         NFYA           6209  4866873       73988868</span></span>
<span><span class="co">#&gt; 11          10 ENSG00000001460        STPG1           8511  6138613       74132995</span></span>
<span><span class="co">#&gt; 12          11 ENSG00000001461       NIPAL3           9396 13079400       74197924</span></span>
<span><span class="co">#&gt; 13          12 ENSG00000001497        LAS1L          12555 10716777       73858963</span></span>
<span><span class="co">#&gt; 14          13 ENSG00000001561        ENPP4           4644  9509567       74091113</span></span>
<span><span class="co">#&gt; 15          14 ENSG00000001617       SEMA3F           4826  2055855       72941475</span></span>
<span><span class="co">#&gt; 16          15 ENSG00000001626         CFTR          16821  5419122       71475983</span></span>
<span><span class="co">#&gt;  [ reached 'max' / getOption("max.print") -- omitted 60514 rows ]</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="identifying-genes-measured-in-a-specific-dataset">Identifying genes measured in a specific dataset<a class="anchor" aria-label="anchor" href="#identifying-genes-measured-in-a-specific-dataset"></a>
</h2>
<p>Now that we have the dataset table, the genes metadata table, and the dataset presence matrix, we can check if a gene or set of genes were measured in a specific dataset.</p>
<p><strong>Important:</strong> the presence matrix is indexed by <code>soma_joinid</code>, and is <em>NOT</em> positionally indexed. In other words:</p>
<ul>
<li>the first dimension of the presence matrix is the dataset’s <code>soma_joinid</code>, as stored in the <code>census_datasets</code> dataframe.</li>
<li>the second dimension of the presence matrix is the feature’s <code>soma_joinid</code>, as stored in the <code>var</code> dataframe.</li>
</ul>
<p>The presence matrix has a method <code>$take()</code> that lets you slice it by <code>soma_joinid</code>s from <code>census_datasets</code> and <code>var</code>. And the full presence matrix, or slices of it, can then be exported to a regular matrix with the method <code>$get_one_based_matrix()</code></p>
<p>Let’s find out if the the gene <code>"ENSG00000286096"</code> was measured in the dataset with id <code>"97a17473-e2b1-4f31-a544-44a60773e2dd"</code>.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Get soma_joinid for datasets and genes of interest</span></span>
<span><span class="va">var_joinid</span> <span class="op">&lt;-</span> <span class="va">var_df</span><span class="op">$</span><span class="va">soma_joinid</span><span class="op">[</span><span class="va">var_df</span><span class="op">$</span><span class="va">feature_id</span> <span class="op">==</span> <span class="st">"ENSG00000286096"</span><span class="op">]</span></span>
<span><span class="va">dataset_joinid</span> <span class="op">&lt;-</span> <span class="va">datasets_df</span><span class="op">$</span><span class="va">soma_joinid</span><span class="op">[</span><span class="va">datasets_df</span><span class="op">$</span><span class="va">dataset_id</span> <span class="op">==</span> <span class="st">"97a17473-e2b1-4f31-a544-44a60773e2dd"</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># Slice presence matrix with datasets and genes of interest</span></span>
<span><span class="va">presence_matrix_slice</span> <span class="op">&lt;-</span> <span class="va">presence_matrix</span><span class="op">$</span><span class="fu">take</span><span class="op">(</span>i <span class="op">=</span> <span class="va">dataset_joinid</span>, j <span class="op">=</span> <span class="va">var_joinid</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Convert presence matrix to regular matrix</span></span>
<span><span class="va">presence_matrix_slice</span> <span class="op">&lt;-</span> <span class="va">presence_matrix_slice</span><span class="op">$</span><span class="fu">get_one_based_matrix</span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Find how if the gene is present in this dataset</span></span>
<span><span class="va">is_present</span> <span class="op">&lt;-</span> <span class="va">presence_matrix_slice</span><span class="op">[</span>, , drop <span class="op">=</span> <span class="cn">TRUE</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"Feature is"</span>, <span class="kw">if</span> <span class="op">(</span><span class="va">is_present</span><span class="op">)</span> <span class="st">"present."</span> <span class="kw">else</span> <span class="st">"not present."</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; Feature is present.</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="identifying-datasets-that-measured-specific-genes">Identifying datasets that measured specific genes<a class="anchor" aria-label="anchor" href="#identifying-datasets-that-measured-specific-genes"></a>
</h2>
<p>Similarly, we can determine the datasets that measured a specific gene or set of genes.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Grab the feature's soma_joinid from the var dataframe</span></span>
<span><span class="va">var_joinid</span> <span class="op">&lt;-</span> <span class="va">var_df</span><span class="op">$</span><span class="va">soma_joinid</span><span class="op">[</span><span class="va">var_df</span><span class="op">$</span><span class="va">feature_id</span> <span class="op">==</span> <span class="st">"ENSG00000286096"</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># The presence matrix is indexed by the joinids of the dataset and var dataframes,</span></span>
<span><span class="co"># so slice out the feature of interest by its joinid.</span></span>
<span><span class="va">presence_matrix_slice</span> <span class="op">&lt;-</span> <span class="va">presence_matrix</span><span class="op">$</span><span class="fu">take</span><span class="op">(</span>j <span class="op">=</span> <span class="va">var_joinid</span><span class="op">)</span><span class="op">$</span><span class="fu">get_one_based_matrix</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">measured_datasets</span> <span class="op">&lt;-</span> <span class="va">presence_matrix_slice</span><span class="op">[</span>, , drop <span class="op">=</span> <span class="cn">TRUE</span><span class="op">]</span> <span class="op">!=</span> <span class="fl">0</span></span>
<span><span class="va">dataset_joinids</span> <span class="op">&lt;-</span> <span class="va">datasets_df</span><span class="op">$</span><span class="va">soma_joinid</span><span class="op">[</span><span class="va">measured_datasets</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># From the datasets dataframe, slice out the datasets which have a joinid in the list</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">datasets_df</span><span class="op">[</span><span class="va">dataset_joinids</span>, <span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt;     soma_joinid</span></span>
<span><span class="co">#&gt; 264         263</span></span>
<span><span class="co">#&gt; 272         271</span></span>
<span><span class="co">#&gt; 299         298</span></span>
<span><span class="co">#&gt; 302         301</span></span>
<span><span class="co">#&gt; 316         315</span></span>
<span><span class="co">#&gt; 335         334</span></span>
<span><span class="co">#&gt; 350         349</span></span>
<span><span class="co">#&gt; 352         351</span></span>
<span><span class="co">#&gt; 353         352</span></span>
<span><span class="co">#&gt; 367         366</span></span>
<span><span class="co">#&gt;                                                                                                                                                                                                                                                                                                           citation</span></span>
<span><span class="co">#&gt; 264  Publication: https://doi.org/10.1038/s41586-020-2496-1 Dataset Version: https://datasets.cellxgene.cziscience.com/a87a3135-3ce7-405e-bc07-b7eac2c09cc6.h5ad curated and distributed by CZ CELLxGENE Discover in Collection: https://cellxgene.cziscience.com/collections/0b9d8a04-bb9d-44da-aa27-705bb65b54eb</span></span>
<span><span class="co">#&gt; 272 Publication: https://doi.org/10.1016/j.cell.2021.12.018 Dataset Version: https://datasets.cellxgene.cziscience.com/80650a81-3bde-49cf-873b-548425184f7a.h5ad curated and distributed by CZ CELLxGENE Discover in Collection: https://cellxgene.cziscience.com/collections/74e10dc4-cbb2-4605-a189-8a1cd8e44d8c</span></span>
<span><span class="co">#&gt; 299  Publication: https://doi.org/10.1101/2023.05.08.539485 Dataset Version: https://datasets.cellxgene.cziscience.com/5097433f-5cf0-4a5e-b252-b9f012e76d46.h5ad curated and distributed by CZ CELLxGENE Discover in Collection: https://cellxgene.cziscience.com/collections/1ca90a2d-2943-483d-b678-b809bf464c30</span></span>
<span><span class="co">#&gt; 302  Publication: https://doi.org/10.1101/2023.05.08.539485 Dataset Version: https://datasets.cellxgene.cziscience.com/d8df87f4-279a-4e85-a192-b3fe0df3143e.h5ad curated and distributed by CZ CELLxGENE Discover in Collection: https://cellxgene.cziscience.com/collections/1ca90a2d-2943-483d-b678-b809bf464c30</span></span>
<span><span class="co">#&gt; 316 Publication: https://doi.org/10.1016/j.cell.2020.08.013 Dataset Version: https://datasets.cellxgene.cziscience.com/74a8e0fe-08ee-4a93-9c2f-9cb70acd714b.h5ad curated and distributed by CZ CELLxGENE Discover in Collection: https://cellxgene.cziscience.com/collections/2f4c738f-e2f3-4553-9db2-0582a38ea4dc</span></span>
<span><span class="co">#&gt; 335 Publication: https://doi.org/10.1016/j.cell.2021.07.023 Dataset Version: https://datasets.cellxgene.cziscience.com/837efefb-05c5-4dfc-9e02-17842c1880f2.h5ad curated and distributed by CZ CELLxGENE Discover in Collection: https://cellxgene.cziscience.com/collections/35d0b748-3eed-43a5-a1c4-1dade5ec5ca0</span></span>
<span><span class="co">#&gt; 350  Publication: https://doi.org/10.1038/s41586-020-2797-4 Dataset Version: https://datasets.cellxgene.cziscience.com/0f5dba64-8621-420f-a404-2f6836cbe1ce.h5ad curated and distributed by CZ CELLxGENE Discover in Collection: https://cellxgene.cziscience.com/collections/b52eb423-5d0d-4645-b217-e1c6d38b2e72</span></span>
<span><span class="co">#&gt; 352  Publication: https://doi.org/10.1038/s41586-020-2496-1 Dataset Version: https://datasets.cellxgene.cziscience.com/1bd1f7fb-5efe-4886-abc5-aa08bf218a83.h5ad curated and distributed by CZ CELLxGENE Discover in Collection: https://cellxgene.cziscience.com/collections/0b9d8a04-bb9d-44da-aa27-705bb65b54eb</span></span>
<span><span class="co">#&gt; 353    Publication: https://doi.org/10.1126/science.add7046 Dataset Version: https://datasets.cellxgene.cziscience.com/fb6148b1-20ab-4e0e-bb6d-e5d8bb2147f2.h5ad curated and distributed by CZ CELLxGENE Discover in Collection: https://cellxgene.cziscience.com/collections/283d65eb-dd53-496d-adb7-7570c7caa443</span></span>
<span><span class="co">#&gt; 367    Publication: https://doi.org/10.1126/science.add7046 Dataset Version: https://datasets.cellxgene.cziscience.com/628158bf-037c-4198-8e4e-b97e2675506b.h5ad curated and distributed by CZ CELLxGENE Discover in Collection: https://cellxgene.cziscience.com/collections/283d65eb-dd53-496d-adb7-7570c7caa443</span></span>
<span><span class="co">#&gt;                            collection_id</span></span>
<span><span class="co">#&gt; 264 0b9d8a04-bb9d-44da-aa27-705bb65b54eb</span></span>
<span><span class="co">#&gt; 272 74e10dc4-cbb2-4605-a189-8a1cd8e44d8c</span></span>
<span><span class="co">#&gt; 299 1ca90a2d-2943-483d-b678-b809bf464c30</span></span>
<span><span class="co">#&gt; 302 1ca90a2d-2943-483d-b678-b809bf464c30</span></span>
<span><span class="co">#&gt; 316 2f4c738f-e2f3-4553-9db2-0582a38ea4dc</span></span>
<span><span class="co">#&gt; 335 35d0b748-3eed-43a5-a1c4-1dade5ec5ca0</span></span>
<span><span class="co">#&gt; 350 b52eb423-5d0d-4645-b217-e1c6d38b2e72</span></span>
<span><span class="co">#&gt; 352 0b9d8a04-bb9d-44da-aa27-705bb65b54eb</span></span>
<span><span class="co">#&gt; 353 283d65eb-dd53-496d-adb7-7570c7caa443</span></span>
<span><span class="co">#&gt; 367 283d65eb-dd53-496d-adb7-7570c7caa443</span></span>
<span><span class="co">#&gt;                                                                                    collection_name</span></span>
<span><span class="co">#&gt; 264                                                                             Tabula Muris Senis</span></span>
<span><span class="co">#&gt; 272 Spatial proteogenomics reveals distinct and evolutionarily conserved hepatic macrophage niches</span></span>
<span><span class="co">#&gt; 299                                    SEA-AD: Seattle Alzheimer&lt;U+2019&gt;s Disease Brain Cell Atlas</span></span>
<span><span class="co">#&gt; 302                                    SEA-AD: Seattle Alzheimer&lt;U+2019&gt;s Disease Brain Cell Atlas</span></span>
<span><span class="co">#&gt; 316                     Cell Types of the Human Retina and Its Organoids at Single-Cell Resolution</span></span>
<span><span class="co">#&gt; 335                   Impaired local intrinsic immunity to SARS-CoV-2 infection in severe COVID-19</span></span>
<span><span class="co">#&gt; 350                                                                 Cells of the adult human heart</span></span>
<span><span class="co">#&gt; 352                                                                             Tabula Muris Senis</span></span>
<span><span class="co">#&gt; 353                                                                    Human Brain Cell Atlas v1.0</span></span>
<span><span class="co">#&gt; 367                                                                    Human Brain Cell Atlas v1.0</span></span>
<span><span class="co">#&gt;                 collection_doi                           dataset_id</span></span>
<span><span class="co">#&gt; 264  10.1038/s41586-020-2496-1 93966790-bbfa-420f-aa85-bc5ca51d9c96</span></span>
<span><span class="co">#&gt; 272 10.1016/j.cell.2021.12.018 d1cbed97-d88f-4954-8925-13302fe30b39</span></span>
<span><span class="co">#&gt; 299  10.1101/2023.05.08.539485 f67f2cfa-ba45-4f77-8e26-64a15f666043</span></span>
<span><span class="co">#&gt; 302  10.1101/2023.05.08.539485 8c6ed88f-11bf-4159-bad7-ff41fb1e1eca</span></span>
<span><span class="co">#&gt; 316 10.1016/j.cell.2020.08.013 babbf9f3-f482-45a5-ba76-c41f4c2f503e</span></span>
<span><span class="co">#&gt; 335 10.1016/j.cell.2021.07.023 13b61a7d-5605-4948-ba48-02c588960143</span></span>
<span><span class="co">#&gt; 350  10.1038/s41586-020-2797-4 9d584fcb-a28a-4b91-a886-ceb66a88ef81</span></span>
<span><span class="co">#&gt; 352  10.1038/s41586-020-2496-1 0fb7916e-7a68-4a4c-a441-3ab3989f29a7</span></span>
<span><span class="co">#&gt; 353    10.1126/science.add7046 93131426-0124-4ab4-a013-9dfbcd99d467</span></span>
<span><span class="co">#&gt; 367    10.1126/science.add7046 7d3ab174-e433-40fc-a352-6fe71b1a19f9</span></span>
<span><span class="co">#&gt;                       dataset_version_id</span></span>
<span><span class="co">#&gt; 264 a87a3135-3ce7-405e-bc07-b7eac2c09cc6</span></span>
<span><span class="co">#&gt; 272 80650a81-3bde-49cf-873b-548425184f7a</span></span>
<span><span class="co">#&gt; 299 5097433f-5cf0-4a5e-b252-b9f012e76d46</span></span>
<span><span class="co">#&gt; 302 d8df87f4-279a-4e85-a192-b3fe0df3143e</span></span>
<span><span class="co">#&gt; 316 74a8e0fe-08ee-4a93-9c2f-9cb70acd714b</span></span>
<span><span class="co">#&gt; 335 837efefb-05c5-4dfc-9e02-17842c1880f2</span></span>
<span><span class="co">#&gt; 350 0f5dba64-8621-420f-a404-2f6836cbe1ce</span></span>
<span><span class="co">#&gt; 352 1bd1f7fb-5efe-4886-abc5-aa08bf218a83</span></span>
<span><span class="co">#&gt; 353 fb6148b1-20ab-4e0e-bb6d-e5d8bb2147f2</span></span>
<span><span class="co">#&gt; 367 628158bf-037c-4198-8e4e-b97e2675506b</span></span>
<span><span class="co">#&gt;                                                                                          dataset_title</span></span>
<span><span class="co">#&gt; 264 Bladder lumen - A single-cell transcriptomic atlas characterizes ageing tissues in the mouse - 10x</span></span>
<span><span class="co">#&gt; 272                                                       CD45 negative cells from human liver dataset</span></span>
<span><span class="co">#&gt; 299                                             Pax6 - MTG: Seattle Alzheimer's Disease Atlas (SEA-AD)</span></span>
<span><span class="co">#&gt; 302                                       Chandelier - MTG: Seattle Alzheimer's Disease Atlas (SEA-AD)</span></span>
<span><span class="co">#&gt; 316                 Fovea - Cell Types of the Human Retina and Its Organoids at Single-Cell Resolution</span></span>
<span><span class="co">#&gt; 335                                                                                        Nasopharynx</span></span>
<span><span class="co">#&gt; 350                                                Fibroblasts &lt;U+2014&gt; Cells of the adult human heart</span></span>
<span><span class="co">#&gt; 352          Lung - A single-cell transcriptomic atlas characterizes ageing tissues in the mouse - 10x</span></span>
<span><span class="co">#&gt; 353                                                                      Dissection: Epithalamus - ETH</span></span>
<span><span class="co">#&gt; 367      Dissection: Amygdaloid complex (AMY) - Basolateral nuclear group (BLN) - lateral nucleus - La</span></span>
<span><span class="co">#&gt;                             dataset_h5ad_path dataset_total_cell_count</span></span>
<span><span class="co">#&gt; 264 93966790-bbfa-420f-aa85-bc5ca51d9c96.h5ad                     8945</span></span>
<span><span class="co">#&gt; 272 d1cbed97-d88f-4954-8925-13302fe30b39.h5ad                    15481</span></span>
<span><span class="co">#&gt; 299 f67f2cfa-ba45-4f77-8e26-64a15f666043.h5ad                     8463</span></span>
<span><span class="co">#&gt; 302 8c6ed88f-11bf-4159-bad7-ff41fb1e1eca.h5ad                     9893</span></span>
<span><span class="co">#&gt; 316 babbf9f3-f482-45a5-ba76-c41f4c2f503e.h5ad                    19768</span></span>
<span><span class="co">#&gt; 335 13b61a7d-5605-4948-ba48-02c588960143.h5ad                    32588</span></span>
<span><span class="co">#&gt; 350 9d584fcb-a28a-4b91-a886-ceb66a88ef81.h5ad                    59341</span></span>
<span><span class="co">#&gt; 352 0fb7916e-7a68-4a4c-a441-3ab3989f29a7.h5ad                    24540</span></span>
<span><span class="co">#&gt; 353 93131426-0124-4ab4-a013-9dfbcd99d467.h5ad                    24327</span></span>
<span><span class="co">#&gt; 367 7d3ab174-e433-40fc-a352-6fe71b1a19f9.h5ad                    28984</span></span>
<span><span class="co">#&gt;  [ reached 'max' / getOption("max.print") -- omitted 32 rows ]</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="identifying-all-genes-measured-in-a-dataset">Identifying all genes measured in a dataset<a class="anchor" aria-label="anchor" href="#identifying-all-genes-measured-in-a-dataset"></a>
</h2>
<p>Finally, we can find the set of genes that were measured in the cells of a given dataset.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Slice the dataset(s) of interest, and get the joinid(s)</span></span>
<span><span class="va">dataset_joinids</span> <span class="op">&lt;-</span> <span class="va">datasets_df</span><span class="op">$</span><span class="va">soma_joinid</span><span class="op">[</span><span class="va">datasets_df</span><span class="op">$</span><span class="va">collection_id</span> <span class="op">==</span> <span class="st">"17481d16-ee44-49e5-bcf0-28c0780d8c4a"</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># Slice the presence matrix by the first dimension, i.e., by dataset</span></span>
<span><span class="va">presence_matrix_slice</span> <span class="op">&lt;-</span> <span class="va">presence_matrix</span><span class="op">$</span><span class="fu">take</span><span class="op">(</span>i <span class="op">=</span> <span class="va">dataset_joinids</span><span class="op">)</span><span class="op">$</span><span class="fu">get_one_based_matrix</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">genes_measured</span> <span class="op">&lt;-</span> <span class="fu">Matrix</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/colSums-methods.html" class="external-link">colSums</a></span><span class="op">(</span><span class="va">presence_matrix_slice</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0</span></span>
<span><span class="va">var_joinids</span> <span class="op">&lt;-</span> <span class="va">var_df</span><span class="op">$</span><span class="va">soma_joinid</span><span class="op">[</span><span class="va">genes_measured</span><span class="op">]</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">var_df</span><span class="op">[</span><span class="va">var_joinids</span>, <span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt;    soma_joinid      feature_id feature_name feature_length      nnz n_measured_obs</span></span>
<span><span class="co">#&gt; 1            0 ENSG00000000003       TSPAN6           4530  4530448       73855064</span></span>
<span><span class="co">#&gt; 2            1 ENSG00000000005         TNMD           1476   236059       61201828</span></span>
<span><span class="co">#&gt; 3            2 ENSG00000000419         DPM1           9276 17576462       74159149</span></span>
<span><span class="co">#&gt; 4            3 ENSG00000000457        SCYL3           6883  9117322       73988868</span></span>
<span><span class="co">#&gt; 5            4 ENSG00000000460     C1orf112           5970  6287794       73636201</span></span>
<span><span class="co">#&gt; 6            5 ENSG00000000938          FGR           3382  5667858       74061500</span></span>
<span><span class="co">#&gt; 7            6 ENSG00000000971          CFH          15284  4423029       74095467</span></span>
<span><span class="co">#&gt; 8            7 ENSG00000001036        FUCA2           2822 10129388       73988868</span></span>
<span><span class="co">#&gt; 9            8 ENSG00000001084         GCLC           8618 12662710       74226365</span></span>
<span><span class="co">#&gt; 10           9 ENSG00000001167         NFYA           6209  4866873       73988868</span></span>
<span><span class="co">#&gt; 11          10 ENSG00000001460        STPG1           8511  6138613       74132995</span></span>
<span><span class="co">#&gt; 12          11 ENSG00000001461       NIPAL3           9396 13079400       74197924</span></span>
<span><span class="co">#&gt; 13          12 ENSG00000001497        LAS1L          12555 10716777       73858963</span></span>
<span><span class="co">#&gt; 14          13 ENSG00000001561        ENPP4           4644  9509567       74091113</span></span>
<span><span class="co">#&gt; 15          14 ENSG00000001617       SEMA3F           4826  2055855       72941475</span></span>
<span><span class="co">#&gt; 16          15 ENSG00000001626         CFTR          16821  5419122       71475983</span></span>
<span><span class="co">#&gt;  [ reached 'max' / getOption("max.print") -- omitted 27130 rows ]</span></span></code></pre></div>
<div class="section level3">
<h3 id="close-the-census">Close the census<a class="anchor" aria-label="anchor" href="#close-the-census"></a>
</h3>
<p>After use, the census object should be closed to release memory and other resources.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">census</span><span class="op">$</span><span class="fu">close</span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>This also closes all SOMA objects accessed via the top-level <code>census</code>. Closing can be automated using <code>on.exit(census$close(), add = TRUE)</code> immediately after <code>census &lt;- open_soma()</code>.</p>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Chan Zuckerberg Initiative Foundation.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

    </footer>
</div>





  </body>
</html>
