<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="cellxgene.census">
<title>Integrating multi-dataset slices of data with Seurat • cellxgene.census</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Integrating multi-dataset slices of data with Seurat">
<meta property="og:description" content="cellxgene.census">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><script defer data-domain="chanzuckerberg.github.io/cellxgene-census" src="https://plausible.io/js/script.js"></script>
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">cellxgene.census</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.9.1</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="external-link nav-link" href="https://chanzuckerberg.github.io/cellxgene-census/">Main Census Site</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <h6 class="dropdown-header" data-toc-skip>Use Census embeddings</h6>
    <a class="dropdown-item" href="../articles/census_access_maintained_embeddings.html">Access CELLxGENE collaboration embeddings (scVI, Geneformer)</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Explore and analyze the data</h6>
    <a class="dropdown-item" href="../articles/comp_bio_census_info.html">Learning about the CZ CELLxGENE Census</a>
    <a class="dropdown-item" href="../articles/comp_bio_summarize_axis_query.html">Summarizing cell and gene metadata</a>
    <a class="dropdown-item" href="../articles/comp_bio_data_integration.html">Integrating multi-dataset slices of data with Seurat</a>
    <a class="dropdown-item" href="../articles/comp_bio_normalizing_full_gene_sequencing.html">Normalizing full-length gene sequencing data</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>cellxgene.census capabilities</h6>
    <a class="dropdown-item" href="../articles/census_query_extract.html">Querying and fetching the single-cell data and cell/gene metadata</a>
    <a class="dropdown-item" href="../articles/census_citation_generation.html">Generating citations for Census slices</a>
    <a class="dropdown-item" href="../articles/census_datasets.html">Census Datasets example</a>
    <a class="dropdown-item" href="../articles/census_dataset_presence.html">Genes measured in each cell (dataset presence matrix)</a>
    <a class="dropdown-item" href="../articles/census_compute_over_X.html">Computing on X using online (incremental) algorithms</a>
    <a class="dropdown-item" href="../articles/census_access_maintained_embeddings.html">Access CELLxGENE collaboration embeddings (scVI, Geneformer)</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/chanzuckerberg/cellxgene-census/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">

<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Integrating multi-dataset slices of data with Seurat</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/chanzuckerberg/cellxgene-census/tree/main/api/r/cellxgene.census/vignettes_/comp_bio_data_integration.Rmd" class="external-link"><code>vignettes/comp_bio_data_integration.Rmd</code></a></small>
      <div class="d-none name"><code>comp_bio_data_integration.Rmd</code></div>
    </div>

    
    
<!--
THIS VIGNETTE IS INSPIRED BY: https://github.com/chanzuckerberg/cellxgene-census/blob/main/api/python/notebooks/analysis_demo/comp_bio_data_integration_scvi.ipynb

-->
<p>The Census contains data from multiple studies providing an opportunity to perform inter-dataset analysis. To this end integration of data has to be performed first to account for batch effects.</p>
<p>This notebook provides a demonstration for integrating two Census datasets using <a href="https://satijalab.org/seurat/articles/integration_introduction.html" class="external-link"><code>Seurat</code></a>. <strong>The goal is not to provide an exhaustive guide on proper integration, but to showcase what information in the Census can inform data integration.</strong></p>
<p><strong>Contents</strong></p>
<ol style="list-style-type: decimal">
<li>Finding and fetching data from mouse liver (10X Genomics and Smart-Seq2).</li>
<li>Gene-length normalization of Smart-Seq2 data.</li>
<li>Integration with <code>Seurat</code>.
<ol style="list-style-type: decimal">
<li>Inspecting data prior to integration.</li>
<li>Integration across datasets using <code>dataset_id</code>.</li>
<li>Integration across datasets using <code>dataset_id</code> and controlling for batch using <code>donor_id</code>.</li>
<li>Integration across datasets using <code>dataset_id</code> and controlling for batch using <code>donor_id</code> + <code>assay_ontology_term_id</code> + <code>suspension_type</code>. .</li>
</ol>
</li>
</ol>
<div class="section level2">
<h2 id="finding-and-fetching-data-from-mouse-liver-10x-genomics-and-smart-seq2">Finding and fetching data from mouse liver (10X Genomics and Smart-Seq2)<a class="anchor" aria-label="anchor" href="#finding-and-fetching-data-from-mouse-liver-10x-genomics-and-smart-seq2"></a>
</h2>
<p>Let’s load all packages needed for this notebook.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/chanzuckerberg/cellxgene-census" class="external-link">"cellxgene.census"</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://satijalab.org/seurat" class="external-link">"Seurat"</a></span><span class="op">)</span></span></code></pre></div>
<p>Now we can open the Census.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">census</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/open_soma.html">open_soma</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>In this notebook we will use Tabula Muris Senis data from the liver as it contains cells from both 10X Genomics and Smart-Seq2 technologies.</p>
<p>Let’s query the <code>datasets</code> table of the Census by filtering on <code>collection_name</code> for “Tabula Muris Senis” and <code>dataset_title</code> for “liver”.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">census_datasets</span> <span class="op">&lt;-</span> <span class="va">census</span><span class="op">$</span><span class="fu">get</span><span class="op">(</span><span class="st">"census_info"</span><span class="op">)</span><span class="op">$</span><span class="fu">get</span><span class="op">(</span><span class="st">"datasets"</span><span class="op">)</span></span>
<span><span class="va">census_datasets</span> <span class="op">&lt;-</span> <span class="va">census_datasets</span><span class="op">$</span><span class="fu">read</span><span class="op">(</span>value_filter <span class="op">=</span> <span class="st">"collection_name == 'Tabula Muris Senis'"</span><span class="op">)</span></span>
<span><span class="va">census_datasets</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="va">census_datasets</span><span class="op">$</span><span class="fu">concat</span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Print rows with liver data</span></span>
<span><span class="va">census_datasets</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grep</a></span><span class="op">(</span><span class="st">"Liver"</span>, <span class="va">census_datasets</span><span class="op">$</span><span class="va">dataset_title</span><span class="op">)</span>, <span class="op">]</span></span>
<span><span class="co">#&gt;    soma_joinid                        collection_id    collection_name</span></span>
<span><span class="co">#&gt; 15         583 0b9d8a04-bb9d-44da-aa27-705bb65b54eb Tabula Muris Senis</span></span>
<span><span class="co">#&gt; 36         605 0b9d8a04-bb9d-44da-aa27-705bb65b54eb Tabula Muris Senis</span></span>
<span><span class="co">#&gt;               collection_doi                           dataset_id</span></span>
<span><span class="co">#&gt; 15 10.1038/s41586-020-2496-1 4546e757-34d0-4d17-be06-538318925fcd</span></span>
<span><span class="co">#&gt; 36 10.1038/s41586-020-2496-1 6202a243-b713-4e12-9ced-c387f8483dea</span></span>
<span><span class="co">#&gt;                      dataset_version_id</span></span>
<span><span class="co">#&gt; 15 0a851e26-a629-4e59-9b52-9b4d1ce4440b</span></span>
<span><span class="co">#&gt; 36 70f4f091-86a9-44e3-a92a-54cee98cc223</span></span>
<span><span class="co">#&gt;                                                                                        dataset_title</span></span>
<span><span class="co">#&gt; 15 Liver - A single-cell transcriptomic atlas characterizes ageing tissues in the mouse - Smart-seq2</span></span>
<span><span class="co">#&gt; 36        Liver - A single-cell transcriptomic atlas characterizes ageing tissues in the mouse - 10x</span></span>
<span><span class="co">#&gt;                            dataset_h5ad_path dataset_total_cell_count</span></span>
<span><span class="co">#&gt; 15 4546e757-34d0-4d17-be06-538318925fcd.h5ad                     2859</span></span>
<span><span class="co">#&gt; 36 6202a243-b713-4e12-9ced-c387f8483dea.h5ad                     7294</span></span></code></pre></div>
<p>Now we can use the values from <code>dataset_id</code> to query and load a Seurat object with all the cells from those datasets.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tabula_muris_liver_ids</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"4546e757-34d0-4d17-be06-538318925fcd"</span>, <span class="st">"6202a243-b713-4e12-9ced-c387f8483dea"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">seurat_obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_seurat.html">get_seurat</a></span><span class="op">(</span></span>
<span>  <span class="va">census</span>,</span>
<span>  organism <span class="op">=</span> <span class="st">"Mus musculus"</span>,</span>
<span>  obs_value_filter <span class="op">=</span> <span class="st">"dataset_id %in% tabula_muris_liver_ids"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>We can check the cell counts for both 10X Genomics and Smart-Seq2 data by looking at <code>assay</code> in the metadata.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">seurat_obj</span><span class="op">$</span><span class="va">assay</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;  10x 3' v2 Smart-seq2 </span></span>
<span><span class="co">#&gt;       7294       2859</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="gene-length-normalization-of-smart-seq2-data-">Gene-length normalization of Smart-Seq2 data.<a class="anchor" aria-label="anchor" href="#gene-length-normalization-of-smart-seq2-data-"></a>
</h2>
<p>Smart-seq2 read counts have to be normalized by gene length.</p>
<p>Lets first get the gene lengths from <code>var.feature_length</code>.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">smart_seq_gene_lengths</span> <span class="op">&lt;-</span> <span class="va">seurat_obj</span><span class="op">$</span><span class="va">RNA</span><span class="op">[[</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">feature_length</span></span></code></pre></div>
<p>Now we can use those to normalize Smart-seq data. So let’s split the object by assay.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seurat_obj.list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/SplitObject.html" class="external-link">SplitObject</a></span><span class="op">(</span><span class="va">seurat_obj</span>, split.by <span class="op">=</span> <span class="st">"assay"</span><span class="op">)</span></span></code></pre></div>
<p>An normalize the Smart-seq slice using the gene lengths and merge them back into a single object.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seurat_obj.list</span><span class="op">[[</span><span class="st">"Smart-seq2"</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="st">"RNA"</span><span class="op">]</span><span class="op">]</span><span class="op">@</span><span class="va">counts</span> <span class="op">&lt;-</span> <span class="va">seurat_obj.list</span><span class="op">[[</span><span class="st">"Smart-seq2"</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="st">"RNA"</span><span class="op">]</span><span class="op">]</span><span class="op">@</span><span class="va">counts</span> <span class="op">/</span> <span class="va">smart_seq_gene_lengths</span></span>
<span><span class="va">seurat_obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/merge.html" class="external-link">merge</a></span><span class="op">(</span><span class="va">seurat_obj.list</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, <span class="va">seurat_obj.list</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="integration-with-seurat">Integration with Seurat<a class="anchor" aria-label="anchor" href="#integration-with-seurat"></a>
</h2>
<p>Here we will use the native integration capabilities of Seurat.</p>
<p><strong>For comprehensive usage and best practices of Seurat intergation please refer to the <a href="https://satijalab.org/seurat/articles/integration_introduction.html" class="external-link">doc site</a> of <code>Seurat</code>.</strong></p>
<div class="section level3">
<h3 id="inspecting-data-prior-to-integration">Inspecting data prior to integration<a class="anchor" aria-label="anchor" href="#inspecting-data-prior-to-integration"></a>
</h3>
<p>Let’s take a look at the strength of batch effects in our data. For that we will perform embedding visualization via UMAP.</p>
<p>Let’s do basic data normalization and variable gene selection</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seurat_obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/SCTransform.html" class="external-link">SCTransform</a></span><span class="op">(</span><span class="va">seurat_obj</span><span class="op">)</span></span>
<span><span class="va">seurat_obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindVariableFeatures.html" class="external-link">FindVariableFeatures</a></span><span class="op">(</span><span class="va">seurat_obj</span>, selection.method <span class="op">=</span> <span class="st">"vst"</span>, nfeatures <span class="op">=</span> <span class="fl">2000</span><span class="op">)</span></span></code></pre></div>
<p>And now perform PCA and UMAP</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seurat_obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunPCA.html" class="external-link">RunPCA</a></span><span class="op">(</span><span class="va">seurat_obj</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SeuratObject/man/VariableFeatures.html" class="external-link">VariableFeatures</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">seurat_obj</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">seurat_obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunUMAP.html" class="external-link">RunUMAP</a></span><span class="op">(</span><span class="va">seurat_obj</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">30</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># By assay</span></span>
<span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">seurat_obj</span>, reduction <span class="op">=</span> <span class="st">"umap"</span>, group.by <span class="op">=</span> <span class="st">"assay"</span><span class="op">)</span></span>
<span><span class="va">p1</span></span></code></pre></div>
<p><img src="comp_bio_data_integration_files/figure-html/unnamed-chunk-12-1.png" width="480" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># By cell type</span></span>
<span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">seurat_obj</span>, reduction <span class="op">=</span> <span class="st">"umap"</span>, group.by <span class="op">=</span> <span class="st">"cell_type"</span><span class="op">)</span></span>
<span><span class="va">p2</span></span></code></pre></div>
<p><img src="comp_bio_data_integration_files/figure-html/unnamed-chunk-13-1.png" width="672" style="display: block; margin: auto;"></p>
<p>You can see that <strong>batch effects are strong</strong> as cells cluster primarily by <code>assay</code> and then by <code>cell_type</code>. Properly integrated embedding would in principle cluster primarily by <code>cell_type</code>, <code>assay</code> should at best randomly distributed.</p>
</div>
<div class="section level3">
<h3 id="data-integration-with-seurat">Data integration with Seurat<a class="anchor" aria-label="anchor" href="#data-integration-with-seurat"></a>
</h3>
<p>Whenever you query and fetch Census data from multiple datasets then integration needs to be performed as evidenced by the batch effects we observed.</p>
<p><strong>The paramaters for Seurat used in this notebook were selected to the model run quickly.</strong> For best practices on integration of single-cell data using <a href="https://satijalab.org/seurat/index.html" class="external-link"><code>Seurat</code></a> please refer to their documentation page.</p>
<p>seurat_d reading the article <a href="https://www.biorxiv.org/content/10.1101/2022.03.10.483747v1" class="external-link">An integrated cell atlas of the human lung in health and disease</a> by Sikkema et al. whom perfomed integration of 43 datasets from Lung.</p>
<p>Here we focus on the metadata from the Census that can be as batch information for integration.</p>
<div class="section level4">
<h4 id="integration-across-datasets-using-dataset_id">Integration across datasets using <code>dataset_id</code><a class="anchor" aria-label="anchor" href="#integration-across-datasets-using-dataset_id"></a>
</h4>
<p>All cells in the Census are annotated with the dataset they come from in <code>"dataset_id"</code>. This is a great place to start for integration.</p>
<p>So let’s run a Seurat integration pipeline. First we define our model with batch set as <code>dataset_id</code>.</p>
<p>Firs normalize and select variable genes seperated by our batch key <code>dataset_id</code></p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># split the dataset into a list of two seurat objects for each dataset</span></span>
<span><span class="va">seurat_obj.list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/SplitObject.html" class="external-link">SplitObject</a></span><span class="op">(</span><span class="va">seurat_obj</span>, split.by <span class="op">=</span> <span class="st">"dataset_id"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># normalize each dataset independently</span></span>
<span><span class="va">seurat_obj.list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span>X <span class="op">=</span> <span class="va">seurat_obj.list</span>, FUN <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">x</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/SCTransform.html" class="external-link">SCTransform</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># select features for integration</span></span>
<span><span class="va">features</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/SelectIntegrationFeatures.html" class="external-link">SelectIntegrationFeatures</a></span><span class="op">(</span>object.list <span class="op">=</span> <span class="va">seurat_obj.list</span><span class="op">)</span></span></code></pre></div>
<p>Now we perform integration.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seurat_obj.list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/PrepSCTIntegration.html" class="external-link">PrepSCTIntegration</a></span><span class="op">(</span><span class="va">seurat_obj.list</span>, anchor.features <span class="op">=</span> <span class="va">features</span><span class="op">)</span></span>
<span><span class="va">seurat_obj.anchors</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindIntegrationAnchors.html" class="external-link">FindIntegrationAnchors</a></span><span class="op">(</span>object.list <span class="op">=</span> <span class="va">seurat_obj.list</span>, anchor.features <span class="op">=</span> <span class="va">features</span>, normalization.method <span class="op">=</span> <span class="st">"SCT"</span><span class="op">)</span></span>
<span><span class="va">seurat_obj.combined</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/IntegrateData.html" class="external-link">IntegrateData</a></span><span class="op">(</span>anchorset <span class="op">=</span> <span class="va">seurat_obj.anchors</span>, normalization.method <span class="op">=</span> <span class="st">"SCT"</span><span class="op">)</span></span></code></pre></div>
<p>Let’s inspect the results by doing normalization and then UMAP visulization.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SeuratObject/man/DefaultAssay.html" class="external-link">DefaultAssay</a></span><span class="op">(</span><span class="va">seurat_obj.combined</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"integrated"</span></span>
<span></span>
<span><span class="co"># Run the standard workflow for visualization and clustering</span></span>
<span><span class="va">seurat_obj.combined</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/ScaleData.html" class="external-link">ScaleData</a></span><span class="op">(</span><span class="va">seurat_obj.combined</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">seurat_obj.combined</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunPCA.html" class="external-link">RunPCA</a></span><span class="op">(</span><span class="va">seurat_obj.combined</span>, npcs <span class="op">=</span> <span class="fl">30</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">seurat_obj.combined</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunUMAP.html" class="external-link">RunUMAP</a></span><span class="op">(</span><span class="va">seurat_obj.combined</span>, reduction <span class="op">=</span> <span class="st">"pca"</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">30</span><span class="op">)</span></span></code></pre></div>
<p>And plot the UMAP.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># By assay</span></span>
<span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">seurat_obj.combined</span>, reduction <span class="op">=</span> <span class="st">"umap"</span>, group.by <span class="op">=</span> <span class="st">"assay"</span><span class="op">)</span></span>
<span><span class="va">p1</span></span></code></pre></div>
<p><img src="comp_bio_data_integration_files/figure-html/unnamed-chunk-17-1.png" width="480" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># By cell type</span></span>
<span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">seurat_obj.combined</span>, reduction <span class="op">=</span> <span class="st">"umap"</span>, group.by <span class="op">=</span> <span class="st">"cell_type"</span><span class="op">)</span></span>
<span><span class="va">p2</span></span></code></pre></div>
<p><img src="comp_bio_data_integration_files/figure-html/unnamed-chunk-18-1.png" width="672" style="display: block; margin: auto;"></p>
<p>Great! You can see that the clustering is no longer mainly driven by assay, albeit still contributing to it.</p>
<p>Great! You can see that the clustering is no longer mainly driven by assay, albeit still contributing to it.</p>
</div>
<div class="section level4">
<h4 id="integration-across-datasets-using-dataset_id-and-controlling-for-batch-using-donor_id">Integration across datasets using <code>dataset_id</code> and controlling for batch using <code>donor_id</code><a class="anchor" aria-label="anchor" href="#integration-across-datasets-using-dataset_id-and-controlling-for-batch-using-donor_id"></a>
</h4>
<p>Similar to <code>dataset_id</code>, all cells in Census are annotated with <code>donor_id</code>. The definition of <code>donor_id</code> depends on the dataset and it is left to the discretion of data curators. However it is still rich in information and can be used as a batch variable during integration.</p>
<p>Because <code>donor_id</code> is not guaranteed to be unique across all cells of the Census, we strongly recommend concatenating <code>dataset_id</code> and <code>donor_id</code> and use that as the batch separator for Seurat</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># split the dataset into a list of two seurat objects for each dataset</span></span>
<span><span class="va">seurat_obj.list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/SplitObject.html" class="external-link">SplitObject</a></span><span class="op">(</span><span class="va">seurat_obj</span>, split.by <span class="op">=</span> <span class="st">"dataset_id"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># normalize each dataset independently controlling for batch</span></span>
<span><span class="va">seurat_obj.list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span>X <span class="op">=</span> <span class="va">seurat_obj.list</span>, FUN <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">x</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/SCTransform.html" class="external-link">SCTransform</a></span><span class="op">(</span><span class="va">x</span>, vars.to.regress <span class="op">=</span> <span class="st">"donor_id"</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># select features for integration</span></span>
<span><span class="va">features</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/SelectIntegrationFeatures.html" class="external-link">SelectIntegrationFeatures</a></span><span class="op">(</span>object.list <span class="op">=</span> <span class="va">seurat_obj.list</span><span class="op">)</span></span></code></pre></div>
<p>Now we perform integration.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seurat_obj.list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/PrepSCTIntegration.html" class="external-link">PrepSCTIntegration</a></span><span class="op">(</span><span class="va">seurat_obj.list</span>, anchor.features <span class="op">=</span> <span class="va">features</span><span class="op">)</span></span>
<span><span class="va">seurat_obj.anchors</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindIntegrationAnchors.html" class="external-link">FindIntegrationAnchors</a></span><span class="op">(</span>object.list <span class="op">=</span> <span class="va">seurat_obj.list</span>, anchor.features <span class="op">=</span> <span class="va">features</span>, normalization.method <span class="op">=</span> <span class="st">"SCT"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Finding all pairwise anchors</span></span>
<span><span class="co">#&gt; Running CCA</span></span>
<span><span class="co">#&gt; Merging objects</span></span>
<span><span class="co">#&gt; Finding neighborhoods</span></span>
<span><span class="co">#&gt; Finding anchors</span></span>
<span><span class="co">#&gt;  Found 7161 anchors</span></span>
<span><span class="co">#&gt; Filtering anchors</span></span>
<span><span class="co">#&gt;  Retained 4990 anchors</span></span>
<span><span class="va">seurat_obj.combined</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/IntegrateData.html" class="external-link">IntegrateData</a></span><span class="op">(</span>anchorset <span class="op">=</span> <span class="va">seurat_obj.anchors</span>, normalization.method <span class="op">=</span> <span class="st">"SCT"</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 1</span></span>
<span><span class="co">#&gt; Warning: Different cells and/or features from existing assay SCT</span></span>
<span><span class="co">#&gt; [1] 2</span></span>
<span><span class="co">#&gt; Warning: Different cells and/or features from existing assay SCT</span></span>
<span><span class="co">#&gt; Merging dataset 1 into 2</span></span>
<span><span class="co">#&gt; Extracting anchors for merged samples</span></span>
<span><span class="co">#&gt; Finding integration vectors</span></span>
<span><span class="co">#&gt; Finding integration vector weights</span></span>
<span><span class="co">#&gt; Integrating data</span></span>
<span><span class="co">#&gt; Warning: Assay integrated changing from Assay to SCTAssay</span></span>
<span></span>
<span><span class="co">#&gt; Warning: Different cells and/or features from existing assay SCT</span></span></code></pre></div>
<p>And inspect the new results by UMAP.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SeuratObject/man/DefaultAssay.html" class="external-link">DefaultAssay</a></span><span class="op">(</span><span class="va">seurat_obj.combined</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"integrated"</span></span>
<span></span>
<span><span class="co"># Run the standard workflow for visualization and clustering</span></span>
<span><span class="va">seurat_obj.combined</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunPCA.html" class="external-link">RunPCA</a></span><span class="op">(</span><span class="va">seurat_obj.combined</span>, npcs <span class="op">=</span> <span class="fl">30</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">seurat_obj.combined</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunUMAP.html" class="external-link">RunUMAP</a></span><span class="op">(</span><span class="va">seurat_obj.combined</span>, reduction <span class="op">=</span> <span class="st">"pca"</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">30</span><span class="op">)</span></span>
<span><span class="co">#&gt; 14:49:10 UMAP embedding parameters a = 0.9922 b = 1.112</span></span>
<span><span class="co">#&gt; 14:49:10 Read 10153 rows and found 30 numeric columns</span></span>
<span><span class="co">#&gt; 14:49:10 Using Annoy for neighbor search, n_neighbors = 30</span></span>
<span><span class="co">#&gt; 14:49:10 Building Annoy index with metric = cosine, n_trees = 50</span></span>
<span><span class="co">#&gt; 0%   10   20   30   40   50   60   70   80   90   100%</span></span>
<span><span class="co">#&gt; [----|----|----|----|----|----|----|----|----|----|</span></span>
<span><span class="co">#&gt; **************************************************|</span></span>
<span><span class="co">#&gt; 14:49:12 Writing NN index file to temp file /tmp/RtmpRa88jI/file926122f23120</span></span>
<span><span class="co">#&gt; 14:49:12 Searching Annoy index using 1 thread, search_k = 3000</span></span>
<span><span class="co">#&gt; 14:49:16 Annoy recall = 100%</span></span>
<span><span class="co">#&gt; 14:49:16 Commencing smooth kNN distance calibration using 1 thread with target n_neighbors = 30</span></span>
<span><span class="co">#&gt; 14:49:17 Initializing from normalized Laplacian + noise (using RSpectra)</span></span>
<span><span class="co">#&gt; 14:49:17 Commencing optimization for 200 epochs, with 409958 positive edges</span></span>
<span><span class="co">#&gt; 14:49:21 Optimization finished</span></span></code></pre></div>
<p>Plot the UMAP.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># By assay</span></span>
<span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">seurat_obj.combined</span>, reduction <span class="op">=</span> <span class="st">"umap"</span>, group.by <span class="op">=</span> <span class="st">"assay"</span><span class="op">)</span></span>
<span><span class="va">p1</span></span></code></pre></div>
<p><img src="comp_bio_data_integration_files/figure-html/unnamed-chunk-22-1.png" width="480" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># By cell type</span></span>
<span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">seurat_obj.combined</span>, reduction <span class="op">=</span> <span class="st">"umap"</span>, group.by <span class="op">=</span> <span class="st">"cell_type"</span><span class="op">)</span></span>
<span><span class="va">p2</span></span></code></pre></div>
<p><img src="comp_bio_data_integration_files/figure-html/unnamed-chunk-23-1.png" width="672" style="display: block; margin: auto;"></p>
<p>As you can see using <code>dataset_id</code> and <code>donor_id</code> as batch the cells now mostly cluster by cell type.</p>
</div>
<div class="section level4">
<h4 id="integration-across-datasets-using-dataset_id-and-controlling-for-batch-using-donor_id-assay_ontology_term_id-suspension_type-">Integration across datasets using <code>dataset_id</code> and controlling for batch using <code>donor_id</code> + <code>assay_ontology_term_id</code> + <code>suspension_type</code>.<a class="anchor" aria-label="anchor" href="#integration-across-datasets-using-dataset_id-and-controlling-for-batch-using-donor_id-assay_ontology_term_id-suspension_type-"></a>
</h4>
<p>In some cases one dataset may contain multiple assay types and/or multiple suspension types (cell vs nucleus), and for those it is important to consider these metadata as batches.</p>
<p>Therefore, the most comprehensive definition of batch in the Census can be accomplished by combining the cell metadata of <code>dataset_id</code>, <code>donor_id</code>, <code>assay_ontology_term_id</code> and <code>suspension_type</code>, the latter will encode the <code>EFO</code> ids for assay types.</p>
<p>In our example, the two datasets that we used only contain cells from one assay each, and one suspension type for all of them. Thus it would not make a difference to include these metadata as part of batch.</p>
<p>This how the implementation would look line</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># EXAMPLE, DON'T RUN.</span></span>
<span></span>
<span><span class="co"># split the dataset into a list of seurat objects for each dataset</span></span>
<span><span class="va">seurat_obj.list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/SplitObject.html" class="external-link">SplitObject</a></span><span class="op">(</span><span class="va">seurat_obj</span>, split.by <span class="op">=</span> <span class="st">"dataset_id"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># normalize each dataset independently controlling for batch</span></span>
<span><span class="va">seurat_obj.list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span>X <span class="op">=</span> <span class="va">seurat_obj.list</span>, FUN <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">x</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/SCTransform.html" class="external-link">SCTransform</a></span><span class="op">(</span><span class="va">x</span>, vars.to.regress <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"donor_id"</span>, <span class="st">"assay_ontology_term_id"</span>, <span class="st">"suspension_type"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># select features for integration</span></span>
<span><span class="va">features</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/SelectIntegrationFeatures.html" class="external-link">SelectIntegrationFeatures</a></span><span class="op">(</span>object.list <span class="op">=</span> <span class="va">seurat_obj.list</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># integrate</span></span>
<span><span class="va">seurat_obj.list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/PrepSCTIntegration.html" class="external-link">PrepSCTIntegration</a></span><span class="op">(</span><span class="va">seurat_obj.list</span>, anchor.features <span class="op">=</span> <span class="va">features</span><span class="op">)</span></span>
<span><span class="va">seurat_obj.anchors</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindIntegrationAnchors.html" class="external-link">FindIntegrationAnchors</a></span><span class="op">(</span>object.list <span class="op">=</span> <span class="va">seurat_obj.list</span>, anchor.features <span class="op">=</span> <span class="va">features</span>, normalization.method <span class="op">=</span> <span class="st">"SCT"</span><span class="op">)</span></span>
<span><span class="va">seurat_obj.combined</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/IntegrateData.html" class="external-link">IntegrateData</a></span><span class="op">(</span>anchorset <span class="op">=</span> <span class="va">seurat_obj.anchors</span>, normalization.method <span class="op">=</span> <span class="st">"SCT"</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Chan Zuckerberg Initiative Foundation.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
