{
  "cells": [
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "db7785c1-eb7f-4914-a230-5a8ddf3ae17b",
      "metadata": {
        "id": "db7785c1-eb7f-4914-a230-5a8ddf3ae17b"
      },
      "outputs": [],
      "source": [
        "import cellxgene_census\n",
        "import tiledbsoma as soma\n",
        "import pandas as pd\n",
        "import numpy as np"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "403eda44-f401-4f3c-b34c-29dbd3eb8a65",
      "metadata": {
        "id": "403eda44-f401-4f3c-b34c-29dbd3eb8a65"
      },
      "outputs": [],
      "source": [
        "DEFAULT_TILEDB_CONFIGURATION = {\n",
        "    # https://docs.tiledb.com/main/how-to/configuration#configuration-parameters\n",
        "    \"py.init_buffer_bytes\": 1 * 1024**3,\n",
        "    \"soma.init_buffer_bytes\": 1 * 1024**3,\n",
        "    # S3 requests should not be signed, since we want to allow anonymous access\n",
        "    \"vfs.s3.no_sign_request\": \"false\",\n",
        "    \"vfs.s3.region\": \"us-west-2\",\n",
        "}\n",
        "ctx = soma.options.SOMATileDBContext().replace(tiledb_config=DEFAULT_TILEDB_CONFIGURATION)\n",
        "census_old = cellxgene_census.open_soma(census_version=\"latest\")\n",
        "census_new = cellxgene_census.open_soma(uri=\"s3://bruce-tmp/census-schema-five-prod-QC-build/soma/\", context=ctx)"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "5ecb30e6",
      "metadata": {},
      "source": [
        "## QC results\n",
        "\n",
        "Light QC on test build shows no issues\n",
        "\n",
        "Checks\n",
        "- Verifying fidelity in data additions, in particular changes to `obs` and addition to `census[\"census_info\"][\"organisms\"]`.\n",
        "- Checking new assays added based on updated list of assays, and verifying the addition of expected datasets.\n",
        "- Validating the existence of  previously missing data due to fixed filter for multi-species datasets.\n",
        "- Validating calculation of normalized layer for SMART-like technologies"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "46166017-1bab-43d8-b265-ef25fce39044",
      "metadata": {
        "id": "46166017-1bab-43d8-b265-ef25fce39044"
      },
      "source": [
        "## Checking census[\"info\"]"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "1359d000-ef8a-4535-aafc-1397173262a4",
      "metadata": {
        "id": "1359d000-ef8a-4535-aafc-1397173262a4",
        "outputId": "fcabb94a-654c-4cd4-f084-b8fdd9ddd904"
      },
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>soma_joinid</th>\n",
              "      <th>label</th>\n",
              "      <th>value</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>0</td>\n",
              "      <td>census_schema_version</td>\n",
              "      <td>1.3.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>1</td>\n",
              "      <td>census_build_date</td>\n",
              "      <td>2024-03-18</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>2</td>\n",
              "      <td>dataset_schema_version</td>\n",
              "      <td>4.0.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>3</td>\n",
              "      <td>total_cell_count</td>\n",
              "      <td>111070547</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>4</td>\n",
              "      <td>unique_cell_count</td>\n",
              "      <td>57996085</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5</th>\n",
              "      <td>5</td>\n",
              "      <td>number_donors_homo_sapiens</td>\n",
              "      <td>16622</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>6</th>\n",
              "      <td>6</td>\n",
              "      <td>number_donors_mus_musculus</td>\n",
              "      <td>4146</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "   soma_joinid                       label       value\n",
              "0            0       census_schema_version       1.3.0\n",
              "1            1           census_build_date  2024-03-18\n",
              "2            2      dataset_schema_version       4.0.0\n",
              "3            3            total_cell_count   111070547\n",
              "4            4           unique_cell_count    57996085\n",
              "5            5  number_donors_homo_sapiens       16622\n",
              "6            6  number_donors_mus_musculus        4146"
            ]
          },
          "execution_count": 13,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "census_old[\"census_info\"][\"summary\"].read().concat().to_pandas()"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "951421c5-e6d7-4186-bb22-2665cd23af38",
      "metadata": {
        "id": "951421c5-e6d7-4186-bb22-2665cd23af38",
        "outputId": "6155c956-80e8-4c1c-d28d-73ae061301d3"
      },
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>soma_joinid</th>\n",
              "      <th>label</th>\n",
              "      <th>value</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>0</td>\n",
              "      <td>census_schema_version</td>\n",
              "      <td>2.0.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>1</td>\n",
              "      <td>census_build_date</td>\n",
              "      <td>test-build</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>2</td>\n",
              "      <td>dataset_schema_version</td>\n",
              "      <td>5.0.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>3</td>\n",
              "      <td>total_cell_count</td>\n",
              "      <td>114199045</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>4</td>\n",
              "      <td>unique_cell_count</td>\n",
              "      <td>59554288</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5</th>\n",
              "      <td>5</td>\n",
              "      <td>number_donors_homo_sapiens</td>\n",
              "      <td>17050</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>6</th>\n",
              "      <td>6</td>\n",
              "      <td>number_donors_mus_musculus</td>\n",
              "      <td>4146</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "   soma_joinid                       label       value\n",
              "0            0       census_schema_version       2.0.0\n",
              "1            1           census_build_date  test-build\n",
              "2            2      dataset_schema_version       5.0.0\n",
              "3            3            total_cell_count   114199045\n",
              "4            4           unique_cell_count    59554288\n",
              "5            5  number_donors_homo_sapiens       17050\n",
              "6            6  number_donors_mus_musculus        4146"
            ]
          },
          "execution_count": 14,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "census_new[\"census_info\"][\"summary\"].read().concat().to_pandas()"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "a364ddb3-5253-4818-a5ac-63b57841feb3",
      "metadata": {
        "id": "a364ddb3-5253-4818-a5ac-63b57841feb3",
        "outputId": "bf48a215-6b83-4187-b3b7-e3a25236f17b"
      },
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>soma_joinid</th>\n",
              "      <th>organism_ontology_term_id</th>\n",
              "      <th>organism_label</th>\n",
              "      <th>organism</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>0</td>\n",
              "      <td>NCBITaxon:9606</td>\n",
              "      <td>Homo sapiens</td>\n",
              "      <td>homo_sapiens</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>1</td>\n",
              "      <td>NCBITaxon:10090</td>\n",
              "      <td>Mus musculus</td>\n",
              "      <td>mus_musculus</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "   soma_joinid organism_ontology_term_id organism_label      organism\n",
              "0            0            NCBITaxon:9606   Homo sapiens  homo_sapiens\n",
              "1            1           NCBITaxon:10090   Mus musculus  mus_musculus"
            ]
          },
          "execution_count": 76,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "census_new[\"census_info\"][\"organisms\"].read().concat().to_pandas()"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "5fb96dfd-0d38-43fc-8fe3-93481a2da508",
      "metadata": {
        "id": "5fb96dfd-0d38-43fc-8fe3-93481a2da508",
        "outputId": "7062aec9-3ef6-496d-9829-abf2fe63633c"
      },
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>soma_joinid</th>\n",
              "      <th>citation</th>\n",
              "      <th>collection_id</th>\n",
              "      <th>collection_name</th>\n",
              "      <th>collection_doi</th>\n",
              "      <th>dataset_id</th>\n",
              "      <th>dataset_version_id</th>\n",
              "      <th>dataset_title</th>\n",
              "      <th>dataset_h5ad_path</th>\n",
              "      <th>dataset_total_cell_count</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>0</td>\n",
              "      <td>Publication: https://doi.org/10.1002/hep4.1854...</td>\n",
              "      <td>44531dd9-1388-4416-a117-af0a99de2294</td>\n",
              "      <td>Single-Cell, Single-Nucleus, and Spatial RNA S...</td>\n",
              "      <td>10.1002/hep4.1854</td>\n",
              "      <td>0895c838-e550-48a3-a777-dbcd35d30272</td>\n",
              "      <td>90c609bd-1439-4cbf-935f-201ecadf0297</td>\n",
              "      <td>Healthy human liver: B cells</td>\n",
              "      <td>0895c838-e550-48a3-a777-dbcd35d30272.h5ad</td>\n",
              "      <td>146</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>1</td>\n",
              "      <td>Publication: https://doi.org/10.1126/sciimmuno...</td>\n",
              "      <td>3a2af25b-2338-4266-aad3-aa8d07473f50</td>\n",
              "      <td>Single-cell analysis of human B cell maturatio...</td>\n",
              "      <td>10.1126/sciimmunol.abe6291</td>\n",
              "      <td>00ff600e-6e2e-4d76-846f-0eec4f0ae417</td>\n",
              "      <td>eefe1452-cf76-40ce-aeba-6c82395edfd8</td>\n",
              "      <td>Human tonsil nonlymphoid cells scRNA</td>\n",
              "      <td>00ff600e-6e2e-4d76-846f-0eec4f0ae417.h5ad</td>\n",
              "      <td>363</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>2</td>\n",
              "      <td>Publication: https://doi.org/10.1038/s41593-02...</td>\n",
              "      <td>180bff9c-c8a5-4539-b13b-ddbc00d643e6</td>\n",
              "      <td>Molecular characterization of selectively vuln...</td>\n",
              "      <td>10.1038/s41593-020-00764-7</td>\n",
              "      <td>bdacc907-7c26-419f-8808-969eab3ca2e8</td>\n",
              "      <td>a835c57c-3805-4c01-b22f-9bcfe1794d10</td>\n",
              "      <td>Molecular characterization of selectively vuln...</td>\n",
              "      <td>bdacc907-7c26-419f-8808-969eab3ca2e8.h5ad</td>\n",
              "      <td>3799</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>3</td>\n",
              "      <td>Publication: https://doi.org/10.1038/s41467-02...</td>\n",
              "      <td>bf325905-5e8e-42e3-933d-9a9053e9af80</td>\n",
              "      <td>Single-cell Atlas of common variable immunodef...</td>\n",
              "      <td>10.1038/s41467-022-29450-x</td>\n",
              "      <td>a5d95a42-0137-496f-8a60-101e17f263c8</td>\n",
              "      <td>dc8026ac-b646-43d6-b714-a2a895ca9acd</td>\n",
              "      <td>Steady-state B cells - scRNA-seq</td>\n",
              "      <td>a5d95a42-0137-496f-8a60-101e17f263c8.h5ad</td>\n",
              "      <td>1324</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>4</td>\n",
              "      <td>Publication: https://doi.org/10.1038/s41590-02...</td>\n",
              "      <td>93eebe82-d8c3-41bc-a906-63b5b5f24a9d</td>\n",
              "      <td>Single-cell proteo-genomic reference maps of t...</td>\n",
              "      <td>10.1038/s41590-021-01059-0</td>\n",
              "      <td>d3566d6a-a455-4a15-980f-45eb29114cab</td>\n",
              "      <td>6cb0bf3a-9e54-4c86-89b9-160869556e77</td>\n",
              "      <td>blood and bone marrow from a healthy young donor</td>\n",
              "      <td>d3566d6a-a455-4a15-980f-45eb29114cab.h5ad</td>\n",
              "      <td>15502</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>758</th>\n",
              "      <td>758</td>\n",
              "      <td>Publication: https://doi.org/10.1038/s41586-02...</td>\n",
              "      <td>45d5d2c3-bc28-4814-aed6-0bb6f0e11c82</td>\n",
              "      <td>A single-cell transcriptional timelapse of mou...</td>\n",
              "      <td>10.1038/s41586-024-07069-w</td>\n",
              "      <td>0bce33ed-455c-4e12-93f8-b7b04a2de4a1</td>\n",
              "      <td>23445882-1c14-4a99-a1a3-16c2e373a099</td>\n",
              "      <td>Whole dataset: Normalized subset 2</td>\n",
              "      <td>0bce33ed-455c-4e12-93f8-b7b04a2de4a1.h5ad</td>\n",
              "      <td>2863559</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>759</th>\n",
              "      <td>759</td>\n",
              "      <td>Publication: https://doi.org/10.1101/2023.05.0...</td>\n",
              "      <td>1ca90a2d-2943-483d-b678-b809bf464c30</td>\n",
              "      <td>SEA-AD: Seattle Alzheimer’s Disease Brain Cell...</td>\n",
              "      <td>10.1101/2023.05.08.539485</td>\n",
              "      <td>c2876b1b-06d8-4d96-a56b-5304f815b99a</td>\n",
              "      <td>18011876-b61b-4995-9f51-934cda890e82</td>\n",
              "      <td>Whole Taxonomy - MTG: Seattle Alzheimer's Dise...</td>\n",
              "      <td>c2876b1b-06d8-4d96-a56b-5304f815b99a.h5ad</td>\n",
              "      <td>1226855</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>760</th>\n",
              "      <td>760</td>\n",
              "      <td>Publication: https://doi.org/10.1101/2023.05.0...</td>\n",
              "      <td>1ca90a2d-2943-483d-b678-b809bf464c30</td>\n",
              "      <td>SEA-AD: Seattle Alzheimer’s Disease Brain Cell...</td>\n",
              "      <td>10.1101/2023.05.08.539485</td>\n",
              "      <td>6f7fd0f1-a2ed-4ff1-80d3-33dde731cbc3</td>\n",
              "      <td>788e2ad7-dca6-415a-b0f4-54bc968ff8db</td>\n",
              "      <td>Whole Taxonomy - DLPFC: Seattle Alzheimer's Di...</td>\n",
              "      <td>6f7fd0f1-a2ed-4ff1-80d3-33dde731cbc3.h5ad</td>\n",
              "      <td>1309414</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>761</th>\n",
              "      <td>761</td>\n",
              "      <td>Publication: https://doi.org/10.1038/s41586-02...</td>\n",
              "      <td>45d5d2c3-bc28-4814-aed6-0bb6f0e11c82</td>\n",
              "      <td>A single-cell transcriptional timelapse of mou...</td>\n",
              "      <td>10.1038/s41586-024-07069-w</td>\n",
              "      <td>dcfa2614-7ca7-4d82-814c-350626eccb26</td>\n",
              "      <td>53ff3a93-c576-4b91-9215-122585dad207</td>\n",
              "      <td>Major cell cluster: Mesoderm</td>\n",
              "      <td>dcfa2614-7ca7-4d82-814c-350626eccb26.h5ad</td>\n",
              "      <td>3267338</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>762</th>\n",
              "      <td>762</td>\n",
              "      <td>Publication: https://doi.org/10.1038/s41586-02...</td>\n",
              "      <td>45d5d2c3-bc28-4814-aed6-0bb6f0e11c82</td>\n",
              "      <td>A single-cell transcriptional timelapse of mou...</td>\n",
              "      <td>10.1038/s41586-024-07069-w</td>\n",
              "      <td>dcfd4feb-18a3-4b30-81d7-1b0c544a8ab3</td>\n",
              "      <td>54ab927a-095b-4ebe-974e-88e142e0e372</td>\n",
              "      <td>Whole dataset: Raw counts only</td>\n",
              "      <td>dcfd4feb-18a3-4b30-81d7-1b0c544a8ab3.h5ad</td>\n",
              "      <td>11441407</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>763 rows × 10 columns</p>\n",
              "</div>"
            ],
            "text/plain": [
              "     soma_joinid                                           citation  \\\n",
              "0              0  Publication: https://doi.org/10.1002/hep4.1854...   \n",
              "1              1  Publication: https://doi.org/10.1126/sciimmuno...   \n",
              "2              2  Publication: https://doi.org/10.1038/s41593-02...   \n",
              "3              3  Publication: https://doi.org/10.1038/s41467-02...   \n",
              "4              4  Publication: https://doi.org/10.1038/s41590-02...   \n",
              "..           ...                                                ...   \n",
              "758          758  Publication: https://doi.org/10.1038/s41586-02...   \n",
              "759          759  Publication: https://doi.org/10.1101/2023.05.0...   \n",
              "760          760  Publication: https://doi.org/10.1101/2023.05.0...   \n",
              "761          761  Publication: https://doi.org/10.1038/s41586-02...   \n",
              "762          762  Publication: https://doi.org/10.1038/s41586-02...   \n",
              "\n",
              "                            collection_id  \\\n",
              "0    44531dd9-1388-4416-a117-af0a99de2294   \n",
              "1    3a2af25b-2338-4266-aad3-aa8d07473f50   \n",
              "2    180bff9c-c8a5-4539-b13b-ddbc00d643e6   \n",
              "3    bf325905-5e8e-42e3-933d-9a9053e9af80   \n",
              "4    93eebe82-d8c3-41bc-a906-63b5b5f24a9d   \n",
              "..                                    ...   \n",
              "758  45d5d2c3-bc28-4814-aed6-0bb6f0e11c82   \n",
              "759  1ca90a2d-2943-483d-b678-b809bf464c30   \n",
              "760  1ca90a2d-2943-483d-b678-b809bf464c30   \n",
              "761  45d5d2c3-bc28-4814-aed6-0bb6f0e11c82   \n",
              "762  45d5d2c3-bc28-4814-aed6-0bb6f0e11c82   \n",
              "\n",
              "                                       collection_name  \\\n",
              "0    Single-Cell, Single-Nucleus, and Spatial RNA S...   \n",
              "1    Single-cell analysis of human B cell maturatio...   \n",
              "2    Molecular characterization of selectively vuln...   \n",
              "3    Single-cell Atlas of common variable immunodef...   \n",
              "4    Single-cell proteo-genomic reference maps of t...   \n",
              "..                                                 ...   \n",
              "758  A single-cell transcriptional timelapse of mou...   \n",
              "759  SEA-AD: Seattle Alzheimer’s Disease Brain Cell...   \n",
              "760  SEA-AD: Seattle Alzheimer’s Disease Brain Cell...   \n",
              "761  A single-cell transcriptional timelapse of mou...   \n",
              "762  A single-cell transcriptional timelapse of mou...   \n",
              "\n",
              "                 collection_doi                            dataset_id  \\\n",
              "0             10.1002/hep4.1854  0895c838-e550-48a3-a777-dbcd35d30272   \n",
              "1    10.1126/sciimmunol.abe6291  00ff600e-6e2e-4d76-846f-0eec4f0ae417   \n",
              "2    10.1038/s41593-020-00764-7  bdacc907-7c26-419f-8808-969eab3ca2e8   \n",
              "3    10.1038/s41467-022-29450-x  a5d95a42-0137-496f-8a60-101e17f263c8   \n",
              "4    10.1038/s41590-021-01059-0  d3566d6a-a455-4a15-980f-45eb29114cab   \n",
              "..                          ...                                   ...   \n",
              "758  10.1038/s41586-024-07069-w  0bce33ed-455c-4e12-93f8-b7b04a2de4a1   \n",
              "759   10.1101/2023.05.08.539485  c2876b1b-06d8-4d96-a56b-5304f815b99a   \n",
              "760   10.1101/2023.05.08.539485  6f7fd0f1-a2ed-4ff1-80d3-33dde731cbc3   \n",
              "761  10.1038/s41586-024-07069-w  dcfa2614-7ca7-4d82-814c-350626eccb26   \n",
              "762  10.1038/s41586-024-07069-w  dcfd4feb-18a3-4b30-81d7-1b0c544a8ab3   \n",
              "\n",
              "                       dataset_version_id  \\\n",
              "0    90c609bd-1439-4cbf-935f-201ecadf0297   \n",
              "1    eefe1452-cf76-40ce-aeba-6c82395edfd8   \n",
              "2    a835c57c-3805-4c01-b22f-9bcfe1794d10   \n",
              "3    dc8026ac-b646-43d6-b714-a2a895ca9acd   \n",
              "4    6cb0bf3a-9e54-4c86-89b9-160869556e77   \n",
              "..                                    ...   \n",
              "758  23445882-1c14-4a99-a1a3-16c2e373a099   \n",
              "759  18011876-b61b-4995-9f51-934cda890e82   \n",
              "760  788e2ad7-dca6-415a-b0f4-54bc968ff8db   \n",
              "761  53ff3a93-c576-4b91-9215-122585dad207   \n",
              "762  54ab927a-095b-4ebe-974e-88e142e0e372   \n",
              "\n",
              "                                         dataset_title  \\\n",
              "0                         Healthy human liver: B cells   \n",
              "1                 Human tonsil nonlymphoid cells scRNA   \n",
              "2    Molecular characterization of selectively vuln...   \n",
              "3                     Steady-state B cells - scRNA-seq   \n",
              "4     blood and bone marrow from a healthy young donor   \n",
              "..                                                 ...   \n",
              "758                 Whole dataset: Normalized subset 2   \n",
              "759  Whole Taxonomy - MTG: Seattle Alzheimer's Dise...   \n",
              "760  Whole Taxonomy - DLPFC: Seattle Alzheimer's Di...   \n",
              "761                       Major cell cluster: Mesoderm   \n",
              "762                     Whole dataset: Raw counts only   \n",
              "\n",
              "                             dataset_h5ad_path  dataset_total_cell_count  \n",
              "0    0895c838-e550-48a3-a777-dbcd35d30272.h5ad                       146  \n",
              "1    00ff600e-6e2e-4d76-846f-0eec4f0ae417.h5ad                       363  \n",
              "2    bdacc907-7c26-419f-8808-969eab3ca2e8.h5ad                      3799  \n",
              "3    a5d95a42-0137-496f-8a60-101e17f263c8.h5ad                      1324  \n",
              "4    d3566d6a-a455-4a15-980f-45eb29114cab.h5ad                     15502  \n",
              "..                                         ...                       ...  \n",
              "758  0bce33ed-455c-4e12-93f8-b7b04a2de4a1.h5ad                   2863559  \n",
              "759  c2876b1b-06d8-4d96-a56b-5304f815b99a.h5ad                   1226855  \n",
              "760  6f7fd0f1-a2ed-4ff1-80d3-33dde731cbc3.h5ad                   1309414  \n",
              "761  dcfa2614-7ca7-4d82-814c-350626eccb26.h5ad                   3267338  \n",
              "762  dcfd4feb-18a3-4b30-81d7-1b0c544a8ab3.h5ad                  11441407  \n",
              "\n",
              "[763 rows x 10 columns]"
            ]
          },
          "execution_count": 15,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "census_old[\"census_info\"][\"datasets\"].read().concat().to_pandas()"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "82ef0129-72ac-4173-967a-b8e4346abec5",
      "metadata": {
        "id": "82ef0129-72ac-4173-967a-b8e4346abec5",
        "outputId": "cc696dcd-7742-4b38-c1e0-c13caab94ec6"
      },
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>soma_joinid</th>\n",
              "      <th>citation</th>\n",
              "      <th>collection_id</th>\n",
              "      <th>collection_name</th>\n",
              "      <th>collection_doi</th>\n",
              "      <th>dataset_id</th>\n",
              "      <th>dataset_version_id</th>\n",
              "      <th>dataset_title</th>\n",
              "      <th>dataset_h5ad_path</th>\n",
              "      <th>dataset_total_cell_count</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>0</td>\n",
              "      <td>Publication: https://doi.org/10.1002/hep4.1854...</td>\n",
              "      <td>44531dd9-1388-4416-a117-af0a99de2294</td>\n",
              "      <td>Single-Cell, Single-Nucleus, and Spatial RNA S...</td>\n",
              "      <td>10.1002/hep4.1854</td>\n",
              "      <td>0895c838-e550-48a3-a777-dbcd35d30272</td>\n",
              "      <td>fb76c95f-0391-4fac-9fb9-082ce2430b59</td>\n",
              "      <td>Healthy human liver: B cells</td>\n",
              "      <td>0895c838-e550-48a3-a777-dbcd35d30272.h5ad</td>\n",
              "      <td>146</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>1</td>\n",
              "      <td>Publication: https://doi.org/10.1126/sciimmuno...</td>\n",
              "      <td>3a2af25b-2338-4266-aad3-aa8d07473f50</td>\n",
              "      <td>Single-cell analysis of human B cell maturatio...</td>\n",
              "      <td>10.1126/sciimmunol.abe6291</td>\n",
              "      <td>00ff600e-6e2e-4d76-846f-0eec4f0ae417</td>\n",
              "      <td>b6737a5e-9069-4dd6-9a57-92e17a746df9</td>\n",
              "      <td>Human tonsil nonlymphoid cells scRNA</td>\n",
              "      <td>00ff600e-6e2e-4d76-846f-0eec4f0ae417.h5ad</td>\n",
              "      <td>363</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>2</td>\n",
              "      <td>Publication: https://doi.org/10.1038/s41593-02...</td>\n",
              "      <td>180bff9c-c8a5-4539-b13b-ddbc00d643e6</td>\n",
              "      <td>Molecular characterization of selectively vuln...</td>\n",
              "      <td>10.1038/s41593-020-00764-7</td>\n",
              "      <td>bdacc907-7c26-419f-8808-969eab3ca2e8</td>\n",
              "      <td>0e02290f-b992-450b-8a19-554f73cd7f09</td>\n",
              "      <td>Molecular characterization of selectively vuln...</td>\n",
              "      <td>bdacc907-7c26-419f-8808-969eab3ca2e8.h5ad</td>\n",
              "      <td>3799</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>3</td>\n",
              "      <td>Publication: https://doi.org/10.1038/s41467-02...</td>\n",
              "      <td>bf325905-5e8e-42e3-933d-9a9053e9af80</td>\n",
              "      <td>Single-cell Atlas of common variable immunodef...</td>\n",
              "      <td>10.1038/s41467-022-29450-x</td>\n",
              "      <td>a5d95a42-0137-496f-8a60-101e17f263c8</td>\n",
              "      <td>40832710-d7b1-43fb-b2c2-1cd2255bc3ac</td>\n",
              "      <td>Steady-state B cells - scRNA-seq</td>\n",
              "      <td>a5d95a42-0137-496f-8a60-101e17f263c8.h5ad</td>\n",
              "      <td>1324</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>4</td>\n",
              "      <td>Publication: https://doi.org/10.1038/s41590-02...</td>\n",
              "      <td>93eebe82-d8c3-41bc-a906-63b5b5f24a9d</td>\n",
              "      <td>Single-cell proteo-genomic reference maps of t...</td>\n",
              "      <td>10.1038/s41590-021-01059-0</td>\n",
              "      <td>d3566d6a-a455-4a15-980f-45eb29114cab</td>\n",
              "      <td>eb6c070c-ff67-4c1f-8d4d-65f9fe2119ee</td>\n",
              "      <td>blood and bone marrow from a healthy young donor</td>\n",
              "      <td>d3566d6a-a455-4a15-980f-45eb29114cab.h5ad</td>\n",
              "      <td>15502</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>791</th>\n",
              "      <td>791</td>\n",
              "      <td>Publication: https://doi.org/10.1038/s41586-02...</td>\n",
              "      <td>45d5d2c3-bc28-4814-aed6-0bb6f0e11c82</td>\n",
              "      <td>A single-cell transcriptional timelapse of mou...</td>\n",
              "      <td>10.1038/s41586-024-07069-w</td>\n",
              "      <td>0bce33ed-455c-4e12-93f8-b7b04a2de4a1</td>\n",
              "      <td>ffeb40f8-d4b9-45c4-95cc-5e2674452ef8</td>\n",
              "      <td>Whole dataset: Normalized subset 2</td>\n",
              "      <td>0bce33ed-455c-4e12-93f8-b7b04a2de4a1.h5ad</td>\n",
              "      <td>2863559</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>792</th>\n",
              "      <td>792</td>\n",
              "      <td>Publication: https://doi.org/10.1101/2023.05.0...</td>\n",
              "      <td>1ca90a2d-2943-483d-b678-b809bf464c30</td>\n",
              "      <td>SEA-AD: Seattle Alzheimer’s Disease Brain Cell...</td>\n",
              "      <td>10.1101/2023.05.08.539485</td>\n",
              "      <td>c2876b1b-06d8-4d96-a56b-5304f815b99a</td>\n",
              "      <td>77dab54a-f2a8-42fc-8c1b-3fda90622ac7</td>\n",
              "      <td>Whole Taxonomy - MTG: Seattle Alzheimer's Dise...</td>\n",
              "      <td>c2876b1b-06d8-4d96-a56b-5304f815b99a.h5ad</td>\n",
              "      <td>1226855</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>793</th>\n",
              "      <td>793</td>\n",
              "      <td>Publication: https://doi.org/10.1101/2023.05.0...</td>\n",
              "      <td>1ca90a2d-2943-483d-b678-b809bf464c30</td>\n",
              "      <td>SEA-AD: Seattle Alzheimer’s Disease Brain Cell...</td>\n",
              "      <td>10.1101/2023.05.08.539485</td>\n",
              "      <td>6f7fd0f1-a2ed-4ff1-80d3-33dde731cbc3</td>\n",
              "      <td>b0cbf861-edd3-4add-a09a-c8698ed0cedf</td>\n",
              "      <td>Whole Taxonomy - DLPFC: Seattle Alzheimer's Di...</td>\n",
              "      <td>6f7fd0f1-a2ed-4ff1-80d3-33dde731cbc3.h5ad</td>\n",
              "      <td>1309414</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>794</th>\n",
              "      <td>794</td>\n",
              "      <td>Publication: https://doi.org/10.1038/s41586-02...</td>\n",
              "      <td>45d5d2c3-bc28-4814-aed6-0bb6f0e11c82</td>\n",
              "      <td>A single-cell transcriptional timelapse of mou...</td>\n",
              "      <td>10.1038/s41586-024-07069-w</td>\n",
              "      <td>dcfa2614-7ca7-4d82-814c-350626eccb26</td>\n",
              "      <td>4ef3a829-b36e-413f-9a32-56f5a91b1041</td>\n",
              "      <td>Major cell cluster: Mesoderm</td>\n",
              "      <td>dcfa2614-7ca7-4d82-814c-350626eccb26.h5ad</td>\n",
              "      <td>3267338</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>795</th>\n",
              "      <td>795</td>\n",
              "      <td>Publication: https://doi.org/10.1038/s41586-02...</td>\n",
              "      <td>45d5d2c3-bc28-4814-aed6-0bb6f0e11c82</td>\n",
              "      <td>A single-cell transcriptional timelapse of mou...</td>\n",
              "      <td>10.1038/s41586-024-07069-w</td>\n",
              "      <td>dcfd4feb-18a3-4b30-81d7-1b0c544a8ab3</td>\n",
              "      <td>9da4d19f-f6ac-4bf0-a47e-2935b1164569</td>\n",
              "      <td>Whole dataset: Raw counts only</td>\n",
              "      <td>dcfd4feb-18a3-4b30-81d7-1b0c544a8ab3.h5ad</td>\n",
              "      <td>11441407</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>796 rows × 10 columns</p>\n",
              "</div>"
            ],
            "text/plain": [
              "     soma_joinid                                           citation  \\\n",
              "0              0  Publication: https://doi.org/10.1002/hep4.1854...   \n",
              "1              1  Publication: https://doi.org/10.1126/sciimmuno...   \n",
              "2              2  Publication: https://doi.org/10.1038/s41593-02...   \n",
              "3              3  Publication: https://doi.org/10.1038/s41467-02...   \n",
              "4              4  Publication: https://doi.org/10.1038/s41590-02...   \n",
              "..           ...                                                ...   \n",
              "791          791  Publication: https://doi.org/10.1038/s41586-02...   \n",
              "792          792  Publication: https://doi.org/10.1101/2023.05.0...   \n",
              "793          793  Publication: https://doi.org/10.1101/2023.05.0...   \n",
              "794          794  Publication: https://doi.org/10.1038/s41586-02...   \n",
              "795          795  Publication: https://doi.org/10.1038/s41586-02...   \n",
              "\n",
              "                            collection_id  \\\n",
              "0    44531dd9-1388-4416-a117-af0a99de2294   \n",
              "1    3a2af25b-2338-4266-aad3-aa8d07473f50   \n",
              "2    180bff9c-c8a5-4539-b13b-ddbc00d643e6   \n",
              "3    bf325905-5e8e-42e3-933d-9a9053e9af80   \n",
              "4    93eebe82-d8c3-41bc-a906-63b5b5f24a9d   \n",
              "..                                    ...   \n",
              "791  45d5d2c3-bc28-4814-aed6-0bb6f0e11c82   \n",
              "792  1ca90a2d-2943-483d-b678-b809bf464c30   \n",
              "793  1ca90a2d-2943-483d-b678-b809bf464c30   \n",
              "794  45d5d2c3-bc28-4814-aed6-0bb6f0e11c82   \n",
              "795  45d5d2c3-bc28-4814-aed6-0bb6f0e11c82   \n",
              "\n",
              "                                       collection_name  \\\n",
              "0    Single-Cell, Single-Nucleus, and Spatial RNA S...   \n",
              "1    Single-cell analysis of human B cell maturatio...   \n",
              "2    Molecular characterization of selectively vuln...   \n",
              "3    Single-cell Atlas of common variable immunodef...   \n",
              "4    Single-cell proteo-genomic reference maps of t...   \n",
              "..                                                 ...   \n",
              "791  A single-cell transcriptional timelapse of mou...   \n",
              "792  SEA-AD: Seattle Alzheimer’s Disease Brain Cell...   \n",
              "793  SEA-AD: Seattle Alzheimer’s Disease Brain Cell...   \n",
              "794  A single-cell transcriptional timelapse of mou...   \n",
              "795  A single-cell transcriptional timelapse of mou...   \n",
              "\n",
              "                 collection_doi                            dataset_id  \\\n",
              "0             10.1002/hep4.1854  0895c838-e550-48a3-a777-dbcd35d30272   \n",
              "1    10.1126/sciimmunol.abe6291  00ff600e-6e2e-4d76-846f-0eec4f0ae417   \n",
              "2    10.1038/s41593-020-00764-7  bdacc907-7c26-419f-8808-969eab3ca2e8   \n",
              "3    10.1038/s41467-022-29450-x  a5d95a42-0137-496f-8a60-101e17f263c8   \n",
              "4    10.1038/s41590-021-01059-0  d3566d6a-a455-4a15-980f-45eb29114cab   \n",
              "..                          ...                                   ...   \n",
              "791  10.1038/s41586-024-07069-w  0bce33ed-455c-4e12-93f8-b7b04a2de4a1   \n",
              "792   10.1101/2023.05.08.539485  c2876b1b-06d8-4d96-a56b-5304f815b99a   \n",
              "793   10.1101/2023.05.08.539485  6f7fd0f1-a2ed-4ff1-80d3-33dde731cbc3   \n",
              "794  10.1038/s41586-024-07069-w  dcfa2614-7ca7-4d82-814c-350626eccb26   \n",
              "795  10.1038/s41586-024-07069-w  dcfd4feb-18a3-4b30-81d7-1b0c544a8ab3   \n",
              "\n",
              "                       dataset_version_id  \\\n",
              "0    fb76c95f-0391-4fac-9fb9-082ce2430b59   \n",
              "1    b6737a5e-9069-4dd6-9a57-92e17a746df9   \n",
              "2    0e02290f-b992-450b-8a19-554f73cd7f09   \n",
              "3    40832710-d7b1-43fb-b2c2-1cd2255bc3ac   \n",
              "4    eb6c070c-ff67-4c1f-8d4d-65f9fe2119ee   \n",
              "..                                    ...   \n",
              "791  ffeb40f8-d4b9-45c4-95cc-5e2674452ef8   \n",
              "792  77dab54a-f2a8-42fc-8c1b-3fda90622ac7   \n",
              "793  b0cbf861-edd3-4add-a09a-c8698ed0cedf   \n",
              "794  4ef3a829-b36e-413f-9a32-56f5a91b1041   \n",
              "795  9da4d19f-f6ac-4bf0-a47e-2935b1164569   \n",
              "\n",
              "                                         dataset_title  \\\n",
              "0                         Healthy human liver: B cells   \n",
              "1                 Human tonsil nonlymphoid cells scRNA   \n",
              "2    Molecular characterization of selectively vuln...   \n",
              "3                     Steady-state B cells - scRNA-seq   \n",
              "4     blood and bone marrow from a healthy young donor   \n",
              "..                                                 ...   \n",
              "791                 Whole dataset: Normalized subset 2   \n",
              "792  Whole Taxonomy - MTG: Seattle Alzheimer's Dise...   \n",
              "793  Whole Taxonomy - DLPFC: Seattle Alzheimer's Di...   \n",
              "794                       Major cell cluster: Mesoderm   \n",
              "795                     Whole dataset: Raw counts only   \n",
              "\n",
              "                             dataset_h5ad_path  dataset_total_cell_count  \n",
              "0    0895c838-e550-48a3-a777-dbcd35d30272.h5ad                       146  \n",
              "1    00ff600e-6e2e-4d76-846f-0eec4f0ae417.h5ad                       363  \n",
              "2    bdacc907-7c26-419f-8808-969eab3ca2e8.h5ad                      3799  \n",
              "3    a5d95a42-0137-496f-8a60-101e17f263c8.h5ad                      1324  \n",
              "4    d3566d6a-a455-4a15-980f-45eb29114cab.h5ad                     15502  \n",
              "..                                         ...                       ...  \n",
              "791  0bce33ed-455c-4e12-93f8-b7b04a2de4a1.h5ad                   2863559  \n",
              "792  c2876b1b-06d8-4d96-a56b-5304f815b99a.h5ad                   1226855  \n",
              "793  6f7fd0f1-a2ed-4ff1-80d3-33dde731cbc3.h5ad                   1309414  \n",
              "794  dcfa2614-7ca7-4d82-814c-350626eccb26.h5ad                   3267338  \n",
              "795  dcfd4feb-18a3-4b30-81d7-1b0c544a8ab3.h5ad                  11441407  \n",
              "\n",
              "[796 rows x 10 columns]"
            ]
          },
          "execution_count": 16,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "census_new[\"census_info\"][\"datasets\"].read().concat().to_pandas()"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "cec7070d-c0ff-4b0f-b9b4-2905c3c51f15",
      "metadata": {
        "id": "cec7070d-c0ff-4b0f-b9b4-2905c3c51f15"
      },
      "source": [
        "## Missing dataset due to incorrect filter of multi-species data"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "7450fa70-6e73-402f-8d21-bf8a4faf158a",
      "metadata": {
        "id": "7450fa70-6e73-402f-8d21-bf8a4faf158a",
        "outputId": "c31d2643-4ad6-4af4-f6a8-3f258aed2f33"
      },
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>dataset_id</th>\n",
              "      <th>dataset_title</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>116</th>\n",
              "      <td>ae5341b8-60fb-4fac-86db-86e49ee66287</td>\n",
              "      <td>Individual Single-Cell RNA-seq PBMC Data from Guo et al.</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>248</th>\n",
              "      <td>055ca631-6ffb-40de-815e-b931e10718c0</td>\n",
              "      <td>Individual Single-Cell RNA-seq PBMC Data from Wilk et al.</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>411</th>\n",
              "      <td>4c4cd77c-8fee-4836-9145-16562a8782fe</td>\n",
              "      <td>Individual Single-Cell RNA-seq PBMC Data from Lee et al.</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>455</th>\n",
              "      <td>59b69042-47c2-47fd-ad03-d21beb99818f</td>\n",
              "      <td>Individual Single-Cell RNA-seq PBMC Data from Arunachalam et al.</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "                               dataset_id  \\\n",
              "116  ae5341b8-60fb-4fac-86db-86e49ee66287   \n",
              "248  055ca631-6ffb-40de-815e-b931e10718c0   \n",
              "411  4c4cd77c-8fee-4836-9145-16562a8782fe   \n",
              "455  59b69042-47c2-47fd-ad03-d21beb99818f   \n",
              "\n",
              "                                                        dataset_title  \n",
              "116          Individual Single-Cell RNA-seq PBMC Data from Guo et al.  \n",
              "248         Individual Single-Cell RNA-seq PBMC Data from Wilk et al.  \n",
              "411          Individual Single-Cell RNA-seq PBMC Data from Lee et al.  \n",
              "455  Individual Single-Cell RNA-seq PBMC Data from Arunachalam et al.  "
            ]
          },
          "execution_count": 126,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "pd.set_option('display.max_colwidth', None)\n",
        "datasets = census_old[\"census_info\"][\"datasets\"].read(column_names=[\"dataset_id\", \"dataset_title\"]).concat().to_pandas()\n",
        "datasets[datasets[\"dataset_title\"].str.contains(\"Individual\")]"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "2fe6736a-eff4-4031-bc4a-0c96ef4d4c36",
      "metadata": {
        "id": "2fe6736a-eff4-4031-bc4a-0c96ef4d4c36",
        "outputId": "6a8c7e23-c748-4f93-93f5-c9c44d5ec093"
      },
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>dataset_id</th>\n",
              "      <th>dataset_title</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>139</th>\n",
              "      <td>ae5341b8-60fb-4fac-86db-86e49ee66287</td>\n",
              "      <td>Individual Single-Cell RNA-seq PBMC Data from Guo et al.</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>276</th>\n",
              "      <td>055ca631-6ffb-40de-815e-b931e10718c0</td>\n",
              "      <td>Individual Single-Cell RNA-seq PBMC Data from Wilk et al.</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>440</th>\n",
              "      <td>4c4cd77c-8fee-4836-9145-16562a8782fe</td>\n",
              "      <td>Individual Single-Cell RNA-seq PBMC Data from Lee et al.</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>484</th>\n",
              "      <td>59b69042-47c2-47fd-ad03-d21beb99818f</td>\n",
              "      <td>Individual Single-Cell RNA-seq PBMC Data from Arunachalam et al.</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>542</th>\n",
              "      <td>5e717147-0f75-4de1-8bd2-6fda01b8d75f</td>\n",
              "      <td>Individual Single-Cell RNA-seq PBMC Data from Schulte-Schrepping et al.</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "                               dataset_id  \\\n",
              "139  ae5341b8-60fb-4fac-86db-86e49ee66287   \n",
              "276  055ca631-6ffb-40de-815e-b931e10718c0   \n",
              "440  4c4cd77c-8fee-4836-9145-16562a8782fe   \n",
              "484  59b69042-47c2-47fd-ad03-d21beb99818f   \n",
              "542  5e717147-0f75-4de1-8bd2-6fda01b8d75f   \n",
              "\n",
              "                                                               dataset_title  \n",
              "139                 Individual Single-Cell RNA-seq PBMC Data from Guo et al.  \n",
              "276                Individual Single-Cell RNA-seq PBMC Data from Wilk et al.  \n",
              "440                 Individual Single-Cell RNA-seq PBMC Data from Lee et al.  \n",
              "484         Individual Single-Cell RNA-seq PBMC Data from Arunachalam et al.  \n",
              "542  Individual Single-Cell RNA-seq PBMC Data from Schulte-Schrepping et al.  "
            ]
          },
          "execution_count": 127,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "datasets = census_new[\"census_info\"][\"datasets\"].read(column_names=[\"dataset_id\", \"dataset_title\"]).concat().to_pandas()\n",
        "datasets[datasets[\"dataset_title\"].str.contains(\"Individual\")]"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "610e5109-919c-4b42-b712-a761b8c7f267",
      "metadata": {
        "id": "610e5109-919c-4b42-b712-a761b8c7f267"
      },
      "source": [
        "**PASS the new Census has the previously missing dataset**"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "2e139975-b0c5-41e3-9e2f-64dd6c6d9648",
      "metadata": {
        "id": "2e139975-b0c5-41e3-9e2f-64dd6c6d9648"
      },
      "source": [
        "## Checking list of all assays and veryfying new ones"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "3e08ddf7-1f9d-4185-8272-1858abcffcc6",
      "metadata": {
        "id": "3e08ddf7-1f9d-4185-8272-1858abcffcc6"
      },
      "outputs": [],
      "source": [
        "assays_old = census_old[\"census_data\"][\"homo_sapiens\"].obs.read(column_names=[\"assay\"]).concat().to_pandas().value_counts().reset_index()"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "5df8ec89-8fb8-49ff-a6fc-e239427e3929",
      "metadata": {
        "id": "5df8ec89-8fb8-49ff-a6fc-e239427e3929"
      },
      "outputs": [],
      "source": [
        "assays_new = census_new[\"census_data\"][\"homo_sapiens\"].obs.read(column_names=[\"assay\"]).concat().to_pandas().value_counts().reset_index()"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "92ac549e-e01d-4297-8dea-8dc811f23357",
      "metadata": {
        "id": "92ac549e-e01d-4297-8dea-8dc811f23357",
        "outputId": "de1bfa0b-25c4-402b-8b56-2267415cb856"
      },
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>assay</th>\n",
              "      <th>count_x</th>\n",
              "      <th>count_y</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>10x 3' transcription profiling</td>\n",
              "      <td>811422.0</td>\n",
              "      <td>819131</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>10x 3' v1</td>\n",
              "      <td>121394.0</td>\n",
              "      <td>121394</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>10x 3' v2</td>\n",
              "      <td>14281666.0</td>\n",
              "      <td>14811657</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>10x 3' v3</td>\n",
              "      <td>36599802.0</td>\n",
              "      <td>38401688</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>10x 5' transcription profiling</td>\n",
              "      <td>1282251.0</td>\n",
              "      <td>1282251</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5</th>\n",
              "      <td>10x 5' v1</td>\n",
              "      <td>6450367.0</td>\n",
              "      <td>6450367</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>6</th>\n",
              "      <td>10x 5' v2</td>\n",
              "      <td>3381738.0</td>\n",
              "      <td>3381738</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>7</th>\n",
              "      <td>BD Rhapsody Targeted mRNA</td>\n",
              "      <td>96145.0</td>\n",
              "      <td>96145</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>8</th>\n",
              "      <td>BD Rhapsody Whole Transcriptome Analysis</td>\n",
              "      <td>177276.0</td>\n",
              "      <td>177276</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>9</th>\n",
              "      <td>CEL-seq2</td>\n",
              "      <td>5244.0</td>\n",
              "      <td>7370</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>10</th>\n",
              "      <td>Drop-seq</td>\n",
              "      <td>336562.0</td>\n",
              "      <td>336562</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>11</th>\n",
              "      <td>GEXSCOPE technology</td>\n",
              "      <td>117962.0</td>\n",
              "      <td>117962</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>12</th>\n",
              "      <td>MARS-seq</td>\n",
              "      <td>54753.0</td>\n",
              "      <td>54753</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>13</th>\n",
              "      <td>SPLiT-seq</td>\n",
              "      <td>NaN</td>\n",
              "      <td>69174</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>14</th>\n",
              "      <td>STRT-seq</td>\n",
              "      <td>9386.0</td>\n",
              "      <td>9386</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>15</th>\n",
              "      <td>ScaleBio single cell RNA sequencing</td>\n",
              "      <td>NaN</td>\n",
              "      <td>685024</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>16</th>\n",
              "      <td>Seq-Well</td>\n",
              "      <td>206754.0</td>\n",
              "      <td>206754</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>17</th>\n",
              "      <td>Seq-Well S3</td>\n",
              "      <td>NaN</td>\n",
              "      <td>32588</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>18</th>\n",
              "      <td>Smart-seq v4</td>\n",
              "      <td>63920.0</td>\n",
              "      <td>63920</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>19</th>\n",
              "      <td>Smart-seq2</td>\n",
              "      <td>190792.0</td>\n",
              "      <td>190792</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>20</th>\n",
              "      <td>inDrop</td>\n",
              "      <td>51304.0</td>\n",
              "      <td>51304</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>21</th>\n",
              "      <td>microwell-seq</td>\n",
              "      <td>642559.0</td>\n",
              "      <td>642559</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>22</th>\n",
              "      <td>sci-RNA-seq</td>\n",
              "      <td>5064268.0</td>\n",
              "      <td>5064268</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "                                       assay     count_x   count_y\n",
              "0             10x 3' transcription profiling    811422.0    819131\n",
              "1                                  10x 3' v1    121394.0    121394\n",
              "2                                  10x 3' v2  14281666.0  14811657\n",
              "3                                  10x 3' v3  36599802.0  38401688\n",
              "4             10x 5' transcription profiling   1282251.0   1282251\n",
              "5                                  10x 5' v1   6450367.0   6450367\n",
              "6                                  10x 5' v2   3381738.0   3381738\n",
              "7                  BD Rhapsody Targeted mRNA     96145.0     96145\n",
              "8   BD Rhapsody Whole Transcriptome Analysis    177276.0    177276\n",
              "9                                   CEL-seq2      5244.0      7370\n",
              "10                                  Drop-seq    336562.0    336562\n",
              "11                       GEXSCOPE technology    117962.0    117962\n",
              "12                                  MARS-seq     54753.0     54753\n",
              "13                                 SPLiT-seq         NaN     69174\n",
              "14                                  STRT-seq      9386.0      9386\n",
              "15       ScaleBio single cell RNA sequencing         NaN    685024\n",
              "16                                  Seq-Well    206754.0    206754\n",
              "17                               Seq-Well S3         NaN     32588\n",
              "18                              Smart-seq v4     63920.0     63920\n",
              "19                                Smart-seq2    190792.0    190792\n",
              "20                                    inDrop     51304.0     51304\n",
              "21                             microwell-seq    642559.0    642559\n",
              "22                               sci-RNA-seq   5064268.0   5064268"
            ]
          },
          "execution_count": 40,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "pd.merge(assasy_old, assays_new, how=\"outer\", on=\"assay\")"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "6cbffa71-54d4-4d4f-b2d9-b5044e52ebb8",
      "metadata": {
        "id": "6cbffa71-54d4-4d4f-b2d9-b5044e52ebb8"
      },
      "source": [
        "## Normalized layer for SMART-like data"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "4c38ae9f-6772-4bb0-a73d-adf80fd8113a",
      "metadata": {
        "id": "4c38ae9f-6772-4bb0-a73d-adf80fd8113a"
      },
      "outputs": [],
      "source": [
        "smart_like = pd.read_csv(\n",
        "    \"https://raw.githubusercontent.com/chanzuckerberg/cellxgene-census/bkmartinjr/993-schema-five/docs/census_accepted_assays_full_gene.csv\",\n",
        "    names=[\"assay_ontology_term_id\", \"assay\"]\n",
        ")"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "e2bff18f-af73-4844-8835-7b4fd54f339a",
      "metadata": {
        "id": "e2bff18f-af73-4844-8835-7b4fd54f339a",
        "outputId": "6feac678-beda-43fd-d616-6f0d1cc6d6f6"
      },
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>assay</th>\n",
              "      <th>count</th>\n",
              "      <th>assay_ontology_term_id</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>FL-cDNA</td>\n",
              "      <td>NaN</td>\n",
              "      <td>EFO:0003755</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>full length single cell RNA sequencing</td>\n",
              "      <td>NaN</td>\n",
              "      <td>EFO:0008441</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>FRISCR</td>\n",
              "      <td>NaN</td>\n",
              "      <td>EFO:0008747</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>Hi-SCL</td>\n",
              "      <td>NaN</td>\n",
              "      <td>EFO:0008763</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>MATQ-seq</td>\n",
              "      <td>NaN</td>\n",
              "      <td>EFO:0008797</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5</th>\n",
              "      <td>Quartz-seq</td>\n",
              "      <td>NaN</td>\n",
              "      <td>EFO:0008877</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>6</th>\n",
              "      <td>Smart-seq</td>\n",
              "      <td>NaN</td>\n",
              "      <td>EFO:0008930</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>7</th>\n",
              "      <td>Smart-seq2</td>\n",
              "      <td>190792.0</td>\n",
              "      <td>EFO:0008931</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>8</th>\n",
              "      <td>SUPeR-seq</td>\n",
              "      <td>NaN</td>\n",
              "      <td>EFO:0008956</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>9</th>\n",
              "      <td>full length single nucleus RNA sequencing</td>\n",
              "      <td>NaN</td>\n",
              "      <td>EFO:0009810</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>10</th>\n",
              "      <td>SCRB-seq</td>\n",
              "      <td>NaN</td>\n",
              "      <td>EFO:0010004</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>11</th>\n",
              "      <td>Smart-3Seq</td>\n",
              "      <td>NaN</td>\n",
              "      <td>EFO:0010022</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>12</th>\n",
              "      <td>Fluidigm C1-based library preparation</td>\n",
              "      <td>NaN</td>\n",
              "      <td>EFO:0010058</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>13</th>\n",
              "      <td>Smart-like</td>\n",
              "      <td>NaN</td>\n",
              "      <td>EFO:0010184</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>14</th>\n",
              "      <td>TruSeq</td>\n",
              "      <td>NaN</td>\n",
              "      <td>EFO:0022396</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>15</th>\n",
              "      <td>Smart-seq3</td>\n",
              "      <td>NaN</td>\n",
              "      <td>EFO:0022488</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>16</th>\n",
              "      <td>SCOPE-chip</td>\n",
              "      <td>NaN</td>\n",
              "      <td>EFO:0030031</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>17</th>\n",
              "      <td>mcSCRB-seq</td>\n",
              "      <td>NaN</td>\n",
              "      <td>EFO:0030061</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>18</th>\n",
              "      <td>Smart-seq v4</td>\n",
              "      <td>63920.0</td>\n",
              "      <td>EFO:0700016</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "                                        assay     count assay_ontology_term_id\n",
              "0                                     FL-cDNA       NaN            EFO:0003755\n",
              "1      full length single cell RNA sequencing       NaN            EFO:0008441\n",
              "2                                      FRISCR       NaN            EFO:0008747\n",
              "3                                      Hi-SCL       NaN            EFO:0008763\n",
              "4                                    MATQ-seq       NaN            EFO:0008797\n",
              "5                                  Quartz-seq       NaN            EFO:0008877\n",
              "6                                   Smart-seq       NaN            EFO:0008930\n",
              "7                                  Smart-seq2  190792.0            EFO:0008931\n",
              "8                                   SUPeR-seq       NaN            EFO:0008956\n",
              "9   full length single nucleus RNA sequencing       NaN            EFO:0009810\n",
              "10                                   SCRB-seq       NaN            EFO:0010004\n",
              "11                                 Smart-3Seq       NaN            EFO:0010022\n",
              "12      Fluidigm C1-based library preparation       NaN            EFO:0010058\n",
              "13                                 Smart-like       NaN            EFO:0010184\n",
              "14                                     TruSeq       NaN            EFO:0022396\n",
              "15                                 Smart-seq3       NaN            EFO:0022488\n",
              "16                                 SCOPE-chip       NaN            EFO:0030031\n",
              "17                                 mcSCRB-seq       NaN            EFO:0030061\n",
              "18                               Smart-seq v4   63920.0            EFO:0700016"
            ]
          },
          "execution_count": 70,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "pd.merge(assays_new, smart_like, how=\"right\", on=\"assay\")"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "552c1c43-b4a5-42ef-846f-2a2ce9e62495",
      "metadata": {
        "id": "552c1c43-b4a5-42ef-846f-2a2ce9e62495"
      },
      "outputs": [],
      "source": [
        "# get a few cells from each assay present\n",
        "smartseq1_ids = census_new[\"census_data\"][\"homo_sapiens\"].obs.read(column_names=[\"soma_joinid\"], value_filter = f\"assay == 'Smart-seq2'\").concat().to_pandas()\n",
        "smartseq4_ids = census_new[\"census_data\"][\"homo_sapiens\"].obs.read(column_names=[\"soma_joinid\"], value_filter = f\"assay == 'Smart-seq v4'\").concat().to_pandas()\n",
        "\n",
        "smartseq = smartseq1_ids[\"soma_joinid\"].tolist()[:1000] + smartseq4_ids[\"soma_joinid\"].tolist()[:1000]\n",
        "\n",
        "adata = cellxgene_census.get_anndata(\n",
        "    census_new,\n",
        "    organism=\"homo_sapiens\",\n",
        "    obs_coords=smartseq,\n",
        "    X_layers=[\"normalized\"]\n",
        ")\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "87c0ca70-e0aa-4d7e-9bb9-6869d13ef3ff",
      "metadata": {
        "id": "87c0ca70-e0aa-4d7e-9bb9-6869d13ef3ff",
        "outputId": "5a61655f-7058-4246-c1a3-27f7102059fd"
      },
      "outputs": [
        {
          "data": {
            "text/plain": [
              "<2000x60528 sparse matrix of type '<class 'numpy.float32'>'\n",
              "\twith 10132245 stored elements in Compressed Sparse Row format>"
            ]
          },
          "execution_count": 102,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "adata.layers[\"normalized\"]"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "f4e94760-15f5-40e6-a779-3f317c40c201",
      "metadata": {
        "id": "f4e94760-15f5-40e6-a779-3f317c40c201",
        "outputId": "3009d87c-b98c-42fa-831d-31020ccf94fe"
      },
      "outputs": [
        {
          "data": {
            "text/plain": [
              "0         Smart-seq2\n",
              "1000    Smart-seq v4\n",
              "Name: assay, dtype: object"
            ]
          },
          "execution_count": 98,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "adata.obs[\"assay\"].drop_duplicates()"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "d8b14a0d-680a-41b8-b95f-83b142adbab3",
      "metadata": {
        "id": "d8b14a0d-680a-41b8-b95f-83b142adbab3",
        "outputId": "0855a0ca-912c-454a-e02d-183fa39bda7a"
      },
      "outputs": [
        {
          "data": {
            "text/plain": [
              "2.1492724e-08"
            ]
          },
          "execution_count": 103,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "min(adata.layers[\"normalized\"].data)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "b21bf46e-59a9-4f11-a21c-713e9b0da7c1",
      "metadata": {
        "id": "b21bf46e-59a9-4f11-a21c-713e9b0da7c1",
        "outputId": "821be60f-b623-49ed-baff-5188b5d21c96"
      },
      "outputs": [
        {
          "data": {
            "text/plain": [
              "0.580307"
            ]
          },
          "execution_count": 104,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "max(adata.layers[\"normalized\"].data)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "ccacd915-536b-4cbd-a280-75d4e393d1e0",
      "metadata": {
        "id": "ccacd915-536b-4cbd-a280-75d4e393d1e0",
        "outputId": "45b8b9a2-9533-490d-943f-4c842ec93e12"
      },
      "outputs": [
        {
          "data": {
            "text/plain": [
              "0.99999154"
            ]
          },
          "execution_count": 117,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "adata.layers[\"normalized\"].sum(axis=1).min()"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "66d3f0e1-ff1b-42f1-a970-cfb6f334a28b",
      "metadata": {
        "id": "66d3f0e1-ff1b-42f1-a970-cfb6f334a28b",
        "outputId": "4051b4e0-9f61-4a6b-e632-e9be9eeeb9fa"
      },
      "outputs": [
        {
          "data": {
            "text/plain": [
              "1.000007"
            ]
          },
          "execution_count": 118,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "adata.layers[\"normalized\"].sum(axis=1).max()"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "86fc9f7b-e24a-4446-add2-7485a1c7af1b",
      "metadata": {
        "id": "86fc9f7b-e24a-4446-add2-7485a1c7af1b"
      },
      "outputs": [],
      "source": [
        "adata.layers[\"normalized_local\"] = adata.X.copy()\n",
        "adata.layers[\"normalized_local\"] = adata.layers[\"normalized_local\"].multiply(1 / adata.var[\"feature_length\"].values[None,:])\n",
        "adata.layers[\"normalized_local\"] = adata.layers[\"normalized_local\"].multiply( 1 / adata.layers[\"normalized_local\"].sum(1).A)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "6f833d7e-760c-4e15-8f24-b74860cb6493",
      "metadata": {
        "id": "6f833d7e-760c-4e15-8f24-b74860cb6493",
        "outputId": "4493f1f3-e42c-478b-88cd-b13e63b0d122"
      },
      "outputs": [
        {
          "data": {
            "text/plain": [
              "array([6.70588850e-04, 5.30853111e-04, 6.05805790e-05, ...,\n",
              "       1.72411111e-04, 1.03791338e-04, 3.12075306e-06])"
            ]
          },
          "execution_count": 120,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "adata.layers[\"normalized_local\"].data"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "322af3b0-da87-41f3-b09c-fd010ecd1599",
      "metadata": {
        "id": "322af3b0-da87-41f3-b09c-fd010ecd1599",
        "outputId": "218216e2-f283-48f0-e882-3a8d9bb6141a"
      },
      "outputs": [
        {
          "data": {
            "text/plain": [
              "array([6.70582056e-04, 5.30853868e-04, 6.05806708e-05, ...,\n",
              "       1.72410160e-04, 1.03792176e-04, 3.12074553e-06], dtype=float32)"
            ]
          },
          "execution_count": 121,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "adata.layers[\"normalized\"].data"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "f7673f43-7b87-4d88-8f14-43afdf0f08c6",
      "metadata": {
        "id": "f7673f43-7b87-4d88-8f14-43afdf0f08c6",
        "outputId": "2340fcf2-b068-417a-bcbb-03171d1ba193"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Decimal points:  10  values differing: n =  6853148 , fracion =  0.6763701430433235\n",
            "Decimal points:  9  values differing: n =  3399477 , fracion =  0.33551073824211713\n",
            "Decimal points:  8  values differing: n =  742827 , fracion =  0.07331316998355251\n",
            "Decimal points:  7  values differing: n =  100107 , fracion =  0.009880041392603515\n",
            "Decimal points:  6  values differing: n =  10993 , fracion =  0.001084952051593699\n",
            "Decimal points:  5  values differing: n =  1087 , fracion =  0.0001072812589904804\n",
            "Decimal points:  4  values differing: n =  110 , fracion =  1.0856429152670509e-05\n",
            "Decimal points:  3  values differing: n =  9 , fracion =  8.882532943094053e-07\n",
            "Decimal points:  2  values differing: n =  1 , fracion =  9.869481047882281e-08\n",
            "Decimal points:  1  values differing: n =  0 , fracion =  0.0\n",
            "Decimal points:  0  values differing: n =  0 , fracion =  0.0\n"
          ]
        }
      ],
      "source": [
        "pre_calc = adata.layers[\"normalized\"].data.copy()\n",
        "fly_calc = adata.layers[\"normalized_local\"].data.copy()\n",
        "for r in reversed(range(11)):\n",
        "    n_differing = np.nonzero(np.round(pre_calc.copy(), r).astype(np.float32) != np.round(fly_calc.copy(), r).astype(np.float32))[0].shape[0]\n",
        "    percent_differing = n_differing / fly_calc.shape[0]\n",
        "    print(\"Decimal points: \", r, \" values differing: n = \", n_differing, \", fracion = \", percent_differing)"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "3ea4f96c-0fa5-4a8e-83f7-133c5820c032",
      "metadata": {
        "id": "3ea4f96c-0fa5-4a8e-83f7-133c5820c032"
      },
      "source": [
        "## Checking schema 5 changes\n",
        "\n",
        "###  **Breaking change**. Updated the requirements for `cell_type` to annotate `\"unknown\"` as the label when the `cell_type_ontology_term_id` value is  `\"unknown\"`."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "1092e75e-1d2c-46b1-9e26-84fef485dadb",
      "metadata": {
        "id": "1092e75e-1d2c-46b1-9e26-84fef485dadb"
      },
      "outputs": [],
      "source": [
        "cell_type_old = census_old[\"census_data\"][\"homo_sapiens\"].obs.read(column_names=[\"cell_type\"]).concat().to_pandas().value_counts().reset_index()\n",
        "cell_type_new = census_new[\"census_data\"][\"homo_sapiens\"].obs.read(column_names=[\"cell_type\"]).concat().to_pandas().value_counts().reset_index()\n",
        "merged = pd.merge(cell_type_old, cell_type_new, how=\"outer\", on=\"cell_type\")"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "4c31e760-31ec-4b4b-9d9d-6b94a5e58537",
      "metadata": {
        "id": "4c31e760-31ec-4b4b-9d9d-6b94a5e58537",
        "outputId": "8da6a794-deeb-4a59-adaf-d9ee5c14255c"
      },
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>cell_type</th>\n",
              "      <th>count_x</th>\n",
              "      <th>count_y</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>492</th>\n",
              "      <td>native cell</td>\n",
              "      <td>1491279.0</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "       cell_type    count_x  count_y\n",
              "492  native cell  1491279.0      NaN"
            ]
          },
          "execution_count": 55,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "merged.loc[merged[\"cell_type\"] == \"native cell\",]"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "383de224-0e8e-4b58-91f6-d3d75a5b9142",
      "metadata": {
        "id": "383de224-0e8e-4b58-91f6-d3d75a5b9142",
        "outputId": "5df8c138-29de-4ba4-d59e-d495630218fc"
      },
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>cell_type</th>\n",
              "      <th>count_x</th>\n",
              "      <th>count_y</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>681</th>\n",
              "      <td>unknown</td>\n",
              "      <td>NaN</td>\n",
              "      <td>1570991.0</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "    cell_type  count_x    count_y\n",
              "681   unknown      NaN  1570991.0"
            ]
          },
          "execution_count": 56,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "merged.loc[merged[\"cell_type\"] == \"unknown\",]"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "6d057019-416c-4201-8706-8d8997e1802f",
      "metadata": {
        "id": "6d057019-416c-4201-8706-8d8997e1802f"
      },
      "source": [
        "**PASS**"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "faf46ebd-1d89-4194-a3a2-26d59470fee5",
      "metadata": {
        "id": "faf46ebd-1d89-4194-a3a2-26d59470fee5"
      },
      "source": [
        "### **Breaking change**. Updated the requirements for `cell_type_ontology_term_id` to replace `\"CL:0000003\"` for *native cell* with `\"unknown\"` to indicate that the cell type is unknown.\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "a3e87abf-5c34-4ba6-9da5-337b40184518",
      "metadata": {
        "id": "a3e87abf-5c34-4ba6-9da5-337b40184518"
      },
      "outputs": [],
      "source": [
        "cell_type_old = census_old[\"census_data\"][\"homo_sapiens\"].obs.read(column_names=[\"cell_type_ontology_term_id\"]).concat().to_pandas().value_counts().reset_index()\n",
        "cell_type_new = census_new[\"census_data\"][\"homo_sapiens\"].obs.read(column_names=[\"cell_type_ontology_term_id\"]).concat().to_pandas().value_counts().reset_index()\n",
        "merged = pd.merge(cell_type_old, cell_type_new, how=\"outer\", on=\"cell_type_ontology_term_id\")"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "9c522f67-414d-420b-89ec-e8bd49115d88",
      "metadata": {
        "id": "9c522f67-414d-420b-89ec-e8bd49115d88",
        "outputId": "a05fc674-d997-492b-a72e-fabbd74e97c9"
      },
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>cell_type_ontology_term_id</th>\n",
              "      <th>count_x</th>\n",
              "      <th>count_y</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>CL:0000003</td>\n",
              "      <td>1491279.0</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "  cell_type_ontology_term_id    count_x  count_y\n",
              "1                 CL:0000003  1491279.0      NaN"
            ]
          },
          "execution_count": 58,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "merged.loc[merged[\"cell_type_ontology_term_id\"] == \"CL:0000003\",]"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "be7faf48-9078-41cf-9fa8-e939d6566911",
      "metadata": {
        "id": "be7faf48-9078-41cf-9fa8-e939d6566911",
        "outputId": "63adc204-b769-436c-c7fe-61a30655084f"
      },
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>cell_type_ontology_term_id</th>\n",
              "      <th>count_x</th>\n",
              "      <th>count_y</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>689</th>\n",
              "      <td>unknown</td>\n",
              "      <td>NaN</td>\n",
              "      <td>1570991.0</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "    cell_type_ontology_term_id  count_x    count_y\n",
              "689                    unknown      NaN  1570991.0"
            ]
          },
          "execution_count": 59,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "merged.loc[merged[\"cell_type_ontology_term_id\"] == \"unknown\",]"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "81522340-f182-4e26-9990-920ccb8b34b4",
      "metadata": {
        "id": "81522340-f182-4e26-9990-920ccb8b34b4"
      },
      "source": [
        "**PASS**"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "6052b39d-86e4-442c-a8a4-351f8cd0e657",
      "metadata": {
        "id": "6052b39d-86e4-442c-a8a4-351f8cd0e657"
      },
      "outputs": [],
      "source": []
    }
  ],
  "metadata": {
    "colab": {
      "provenance": []
    },
    "kernelspec": {
      "display_name": "Python 3 (ipykernel)",
      "language": "python",
      "name": "python3"
    },
    "language_info": {
      "codemirror_mode": {
        "name": "ipython",
        "version": 3
      },
      "file_extension": ".py",
      "mimetype": "text/x-python",
      "name": "python",
      "nbconvert_exporter": "python",
      "pygments_lexer": "ipython3",
      "version": "3.10.14"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 5
}
