{
  "cells": [
    {
      "cell_type": "code",
      "execution_count": 1,
      "id": "db7785c1-eb7f-4914-a230-5a8ddf3ae17b",
      "metadata": {
        "id": "db7785c1-eb7f-4914-a230-5a8ddf3ae17b"
      },
      "outputs": [],
      "source": [
        "import cellxgene_census\n",
        "import tiledbsoma as soma\n",
        "import pandas as pd\n",
        "import numpy as np\n",
        "\n",
        "from scipy.sparse import csr_matrix"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "46d31458",
      "metadata": {},
      "source": [
        "## QC results\n",
        "\n",
        "Light QC on test build shows no issues\n",
        "\n",
        "Checks\n",
        "\n",
        "✅ High-level data is accesible and looks in order for `census[\"census_info\"]` and `census[\"census_data\"]`\n",
        "\n",
        "✅ Expected increase in cells from latest LTS 154M to 211M\n",
        "\n",
        "✅ Similar expected increase in other categories (datasets, donors, unique cells)\n",
        "\n",
        "❌ Duplicated rows for `homo_sapiens` and `mus_musculus` in `census_new[\"census_info\"][\"organisms\"]`\n",
        "\n",
        "✅ All assays increased in N cells, no decreases \n",
        "\n",
        "✅ Normalized layer was properly created\n",
        "\n",
        "✅ Normalized layer was properly created for full-gene sequencing asssays\n",
        "\n",
        "✅ Spot checking obs and var contents across all organisms shows compliance with Discover schema 7.0.0\n",
        "\n",
        "✅ Spot checking expression matrices across all organisms shows expected values\n",
        "\n",
        "✅ Anndata exporting works for all organisms\n",
        " \n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 2,
      "id": "403eda44-f401-4f3c-b34c-29dbd3eb8a65",
      "metadata": {
        "id": "403eda44-f401-4f3c-b34c-29dbd3eb8a65"
      },
      "outputs": [],
      "source": [
        "DEFAULT_TILEDB_CONFIGURATION = {\n",
        "    # https://docs.tiledb.com/main/how-to/configuration#configuration-parameters\n",
        "    \"py.init_buffer_bytes\": 1 * 1024**3,\n",
        "    \"soma.init_buffer_bytes\": 1 * 1024**3,\n",
        "    # S3 requests should not be signed, since we want to allow anonymous access\n",
        "    \"vfs.s3.no_sign_request\": \"false\",\n",
        "    \"vfs.s3.region\": \"us-west-2\",\n",
        "}\n",
        "ctx = soma.options.SOMATileDBContext().replace(tiledb_config=DEFAULT_TILEDB_CONFIGURATION)\n",
        "census_old = cellxgene_census.open_soma(census_version=\"2025-01-30\")\n",
        "census_new = cellxgene_census.open_soma(uri=\"s3://cellxgene-census-dev/cell-census/2025-10-27/soma\", context=ctx)"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "46166017-1bab-43d8-b265-ef25fce39044",
      "metadata": {
        "id": "46166017-1bab-43d8-b265-ef25fce39044"
      },
      "source": [
        "## Checking census[\"info\"]"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 3,
      "id": "1359d000-ef8a-4535-aafc-1397173262a4",
      "metadata": {
        "id": "1359d000-ef8a-4535-aafc-1397173262a4",
        "outputId": "fcabb94a-654c-4cd4-f084-b8fdd9ddd904"
      },
      "outputs": [
        {
          "data": {
            "application/vnd.microsoft.datawrangler.viewer.v0+json": {
              "columns": [
                {
                  "name": "index",
                  "rawType": "int64",
                  "type": "integer"
                },
                {
                  "name": "soma_joinid",
                  "rawType": "int64",
                  "type": "integer"
                },
                {
                  "name": "label",
                  "rawType": "object",
                  "type": "string"
                },
                {
                  "name": "value",
                  "rawType": "object",
                  "type": "string"
                }
              ],
              "ref": "9ce4f05b-41af-4902-b651-6e6d44463e22",
              "rows": [
                [
                  "0",
                  "0",
                  "census_schema_version",
                  "2.1.0"
                ],
                [
                  "1",
                  "1",
                  "census_build_date",
                  "spatial-build-rc3-io2-2025-01-30T21:20:53Z"
                ],
                [
                  "2",
                  "2",
                  "dataset_schema_version",
                  "5.2.0"
                ],
                [
                  "3",
                  "3",
                  "total_cell_count",
                  "154437194"
                ],
                [
                  "4",
                  "4",
                  "unique_cell_count",
                  "85809959"
                ],
                [
                  "5",
                  "5",
                  "number_donors_homo_sapiens",
                  "390"
                ],
                [
                  "6",
                  "6",
                  "number_donors_mus_musculus",
                  "172"
                ]
              ],
              "shape": {
                "columns": 3,
                "rows": 7
              }
            },
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>soma_joinid</th>\n",
              "      <th>label</th>\n",
              "      <th>value</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>0</td>\n",
              "      <td>census_schema_version</td>\n",
              "      <td>2.1.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>1</td>\n",
              "      <td>census_build_date</td>\n",
              "      <td>spatial-build-rc3-io2-2025-01-30T21:20:53Z</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>2</td>\n",
              "      <td>dataset_schema_version</td>\n",
              "      <td>5.2.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>3</td>\n",
              "      <td>total_cell_count</td>\n",
              "      <td>154437194</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>4</td>\n",
              "      <td>unique_cell_count</td>\n",
              "      <td>85809959</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5</th>\n",
              "      <td>5</td>\n",
              "      <td>number_donors_homo_sapiens</td>\n",
              "      <td>390</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>6</th>\n",
              "      <td>6</td>\n",
              "      <td>number_donors_mus_musculus</td>\n",
              "      <td>172</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "   soma_joinid                       label  \\\n",
              "0            0       census_schema_version   \n",
              "1            1           census_build_date   \n",
              "2            2      dataset_schema_version   \n",
              "3            3            total_cell_count   \n",
              "4            4           unique_cell_count   \n",
              "5            5  number_donors_homo_sapiens   \n",
              "6            6  number_donors_mus_musculus   \n",
              "\n",
              "                                        value  \n",
              "0                                       2.1.0  \n",
              "1  spatial-build-rc3-io2-2025-01-30T21:20:53Z  \n",
              "2                                       5.2.0  \n",
              "3                                   154437194  \n",
              "4                                    85809959  \n",
              "5                                         390  \n",
              "6                                         172  "
            ]
          },
          "execution_count": 3,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "census_old[\"census_info\"][\"summary\"].read().concat().to_pandas()"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 4,
      "id": "951421c5-e6d7-4186-bb22-2665cd23af38",
      "metadata": {
        "id": "951421c5-e6d7-4186-bb22-2665cd23af38",
        "outputId": "6155c956-80e8-4c1c-d28d-73ae061301d3"
      },
      "outputs": [
        {
          "data": {
            "application/vnd.microsoft.datawrangler.viewer.v0+json": {
              "columns": [
                {
                  "name": "index",
                  "rawType": "int64",
                  "type": "integer"
                },
                {
                  "name": "soma_joinid",
                  "rawType": "int64",
                  "type": "integer"
                },
                {
                  "name": "label",
                  "rawType": "object",
                  "type": "string"
                },
                {
                  "name": "value",
                  "rawType": "object",
                  "type": "string"
                }
              ],
              "ref": "1ba56de9-80d4-44f7-8a5e-9953ddfc818c",
              "rows": [
                [
                  "0",
                  "0",
                  "census_schema_version",
                  "2.4.0"
                ],
                [
                  "1",
                  "1",
                  "census_build_date",
                  "2025-10-27"
                ],
                [
                  "2",
                  "2",
                  "dataset_schema_version",
                  "7.0.0"
                ],
                [
                  "3",
                  "3",
                  "total_cell_count",
                  "211913318"
                ],
                [
                  "4",
                  "4",
                  "unique_cell_count",
                  "124710875"
                ]
              ],
              "shape": {
                "columns": 3,
                "rows": 5
              }
            },
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>soma_joinid</th>\n",
              "      <th>label</th>\n",
              "      <th>value</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>0</td>\n",
              "      <td>census_schema_version</td>\n",
              "      <td>2.4.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>1</td>\n",
              "      <td>census_build_date</td>\n",
              "      <td>2025-10-27</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>2</td>\n",
              "      <td>dataset_schema_version</td>\n",
              "      <td>7.0.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>3</td>\n",
              "      <td>total_cell_count</td>\n",
              "      <td>211913318</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>4</td>\n",
              "      <td>unique_cell_count</td>\n",
              "      <td>124710875</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "   soma_joinid                   label       value\n",
              "0            0   census_schema_version       2.4.0\n",
              "1            1       census_build_date  2025-10-27\n",
              "2            2  dataset_schema_version       7.0.0\n",
              "3            3        total_cell_count   211913318\n",
              "4            4       unique_cell_count   124710875"
            ]
          },
          "execution_count": 4,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "census_new[\"census_info\"][\"summary\"].read().concat().to_pandas()"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 5,
      "id": "a364ddb3-5253-4818-a5ac-63b57841feb3",
      "metadata": {
        "id": "a364ddb3-5253-4818-a5ac-63b57841feb3",
        "outputId": "bf48a215-6b83-4187-b3b7-e3a25236f17b"
      },
      "outputs": [
        {
          "data": {
            "application/vnd.microsoft.datawrangler.viewer.v0+json": {
              "columns": [
                {
                  "name": "index",
                  "rawType": "int64",
                  "type": "integer"
                },
                {
                  "name": "soma_joinid",
                  "rawType": "int64",
                  "type": "integer"
                },
                {
                  "name": "organism_ontology_term_id",
                  "rawType": "object",
                  "type": "string"
                },
                {
                  "name": "organism_label",
                  "rawType": "object",
                  "type": "string"
                },
                {
                  "name": "organism",
                  "rawType": "object",
                  "type": "string"
                }
              ],
              "ref": "ac487be6-e1a3-419f-b305-02f1993ce385",
              "rows": [
                [
                  "0",
                  "0",
                  "NCBITaxon:9606",
                  "Homo sapiens",
                  "homo_sapiens"
                ],
                [
                  "1",
                  "1",
                  "NCBITaxon:10090",
                  "Mus musculus",
                  "mus_musculus"
                ],
                [
                  "2",
                  "2",
                  "NCBITaxon:9606",
                  "Homo sapiens",
                  "homo_sapiens"
                ],
                [
                  "3",
                  "3",
                  "NCBITaxon:10090",
                  "Mus musculus",
                  "mus_musculus"
                ],
                [
                  "4",
                  "4",
                  "NCBITaxon:9483",
                  "Callithrix jacchus",
                  "callithrix_jacchus"
                ],
                [
                  "5",
                  "5",
                  "NCBITaxon:9544",
                  "Macaca mulatta",
                  "macaca_mulatta"
                ],
                [
                  "6",
                  "6",
                  "NCBITaxon:9598",
                  "Pan troglodytes",
                  "pan_troglodytes"
                ]
              ],
              "shape": {
                "columns": 4,
                "rows": 7
              }
            },
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>soma_joinid</th>\n",
              "      <th>organism_ontology_term_id</th>\n",
              "      <th>organism_label</th>\n",
              "      <th>organism</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>0</td>\n",
              "      <td>NCBITaxon:9606</td>\n",
              "      <td>Homo sapiens</td>\n",
              "      <td>homo_sapiens</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>1</td>\n",
              "      <td>NCBITaxon:10090</td>\n",
              "      <td>Mus musculus</td>\n",
              "      <td>mus_musculus</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>2</td>\n",
              "      <td>NCBITaxon:9606</td>\n",
              "      <td>Homo sapiens</td>\n",
              "      <td>homo_sapiens</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>3</td>\n",
              "      <td>NCBITaxon:10090</td>\n",
              "      <td>Mus musculus</td>\n",
              "      <td>mus_musculus</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>4</td>\n",
              "      <td>NCBITaxon:9483</td>\n",
              "      <td>Callithrix jacchus</td>\n",
              "      <td>callithrix_jacchus</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5</th>\n",
              "      <td>5</td>\n",
              "      <td>NCBITaxon:9544</td>\n",
              "      <td>Macaca mulatta</td>\n",
              "      <td>macaca_mulatta</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>6</th>\n",
              "      <td>6</td>\n",
              "      <td>NCBITaxon:9598</td>\n",
              "      <td>Pan troglodytes</td>\n",
              "      <td>pan_troglodytes</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "   soma_joinid organism_ontology_term_id      organism_label  \\\n",
              "0            0            NCBITaxon:9606        Homo sapiens   \n",
              "1            1           NCBITaxon:10090        Mus musculus   \n",
              "2            2            NCBITaxon:9606        Homo sapiens   \n",
              "3            3           NCBITaxon:10090        Mus musculus   \n",
              "4            4            NCBITaxon:9483  Callithrix jacchus   \n",
              "5            5            NCBITaxon:9544      Macaca mulatta   \n",
              "6            6            NCBITaxon:9598     Pan troglodytes   \n",
              "\n",
              "             organism  \n",
              "0        homo_sapiens  \n",
              "1        mus_musculus  \n",
              "2        homo_sapiens  \n",
              "3        mus_musculus  \n",
              "4  callithrix_jacchus  \n",
              "5      macaca_mulatta  \n",
              "6     pan_troglodytes  "
            ]
          },
          "execution_count": 5,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "census_new[\"census_info\"][\"organisms\"].read().concat().to_pandas()"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 6,
      "id": "5fb96dfd-0d38-43fc-8fe3-93481a2da508",
      "metadata": {
        "id": "5fb96dfd-0d38-43fc-8fe3-93481a2da508",
        "outputId": "7062aec9-3ef6-496d-9829-abf2fe63633c"
      },
      "outputs": [
        {
          "data": {
            "application/vnd.microsoft.datawrangler.viewer.v0+json": {
              "columns": [
                {
                  "name": "index",
                  "rawType": "int64",
                  "type": "integer"
                },
                {
                  "name": "soma_joinid",
                  "rawType": "int64",
                  "type": "integer"
                },
                {
                  "name": "citation",
                  "rawType": "object",
                  "type": "string"
                },
                {
                  "name": "collection_id",
                  "rawType": "object",
                  "type": "string"
                },
                {
                  "name": "collection_name",
                  "rawType": "object",
                  "type": "string"
                },
                {
                  "name": "collection_doi",
                  "rawType": "object",
                  "type": "string"
                },
                {
                  "name": "collection_doi_label",
                  "rawType": "object",
                  "type": "string"
                },
                {
                  "name": "dataset_id",
                  "rawType": "object",
                  "type": "string"
                },
                {
                  "name": "dataset_version_id",
                  "rawType": "object",
                  "type": "string"
                },
                {
                  "name": "dataset_title",
                  "rawType": "object",
                  "type": "string"
                },
                {
                  "name": "dataset_h5ad_path",
                  "rawType": "object",
                  "type": "string"
                },
                {
                  "name": "dataset_total_cell_count",
                  "rawType": "int64",
                  "type": "integer"
                }
              ],
              "ref": "e78c148e-8bc9-4159-ac55-1059e2ffec5b",
              "rows": [
                [
                  "0",
                  "0",
                  "Publication: https://doi.org/10.1016/j.isci.2022.104097 Dataset Version: https://datasets.cellxgene.cziscience.com/f76861bb-becb-4eb7-82fc-782dc96ccc7f.h5ad curated and distributed by CZ CELLxGENE Discover in Collection: https://cellxgene.cziscience.com/collections/8e880741-bf9a-4c8e-9227-934204631d2a",
                  "8e880741-bf9a-4c8e-9227-934204631d2a",
                  "High Resolution Slide-seqV2 Spatial Transcriptomics Enables Discovery of Disease-Specific Cell Neighborhoods and Pathways",
                  "10.1016/j.isci.2022.104097",
                  "Marshall et al. (2022) iScience",
                  "4eb29386-de81-452f-b3c0-e00844e8c7fd",
                  "f76861bb-becb-4eb7-82fc-782dc96ccc7f",
                  "Spatial transcriptomics in mouse: Puck_191112_05",
                  "4eb29386-de81-452f-b3c0-e00844e8c7fd.h5ad",
                  "10888"
                ],
                [
                  "1",
                  "1",
                  "Publication: https://doi.org/10.1016/j.isci.2022.104097 Dataset Version: https://datasets.cellxgene.cziscience.com/7d7ec1b6-6e3f-4aaa-9442-4b22f3424396.h5ad curated and distributed by CZ CELLxGENE Discover in Collection: https://cellxgene.cziscience.com/collections/8e880741-bf9a-4c8e-9227-934204631d2a",
                  "8e880741-bf9a-4c8e-9227-934204631d2a",
                  "High Resolution Slide-seqV2 Spatial Transcriptomics Enables Discovery of Disease-Specific Cell Neighborhoods and Pathways",
                  "10.1016/j.isci.2022.104097",
                  "Marshall et al. (2022) iScience",
                  "78d59e4a-82eb-4a61-a1dc-da974d7ea54b",
                  "7d7ec1b6-6e3f-4aaa-9442-4b22f3424396",
                  "Spatial transcriptomics in mouse: Puck_191112_08",
                  "78d59e4a-82eb-4a61-a1dc-da974d7ea54b.h5ad",
                  "10250"
                ],
                [
                  "2",
                  "2",
                  "Publication: https://doi.org/10.1016/j.isci.2022.104097 Dataset Version: https://datasets.cellxgene.cziscience.com/de54aed8-4f73-48f6-9229-418a840e2d82.h5ad curated and distributed by CZ CELLxGENE Discover in Collection: https://cellxgene.cziscience.com/collections/8e880741-bf9a-4c8e-9227-934204631d2a",
                  "8e880741-bf9a-4c8e-9227-934204631d2a",
                  "High Resolution Slide-seqV2 Spatial Transcriptomics Enables Discovery of Disease-Specific Cell Neighborhoods and Pathways",
                  "10.1016/j.isci.2022.104097",
                  "Marshall et al. (2022) iScience",
                  "add5eb84-5fc9-4f01-982e-a346dd42ee82",
                  "de54aed8-4f73-48f6-9229-418a840e2d82",
                  "Spatial transcriptomics in mouse: Puck_191109_20",
                  "add5eb84-5fc9-4f01-982e-a346dd42ee82.h5ad",
                  "12906"
                ],
                [
                  "3",
                  "3",
                  "Publication: https://doi.org/10.1016/j.isci.2022.104097 Dataset Version: https://datasets.cellxgene.cziscience.com/abe4fce1-0859-4a56-ad1e-734d79f0e6c8.h5ad curated and distributed by CZ CELLxGENE Discover in Collection: https://cellxgene.cziscience.com/collections/8e880741-bf9a-4c8e-9227-934204631d2a",
                  "8e880741-bf9a-4c8e-9227-934204631d2a",
                  "High Resolution Slide-seqV2 Spatial Transcriptomics Enables Discovery of Disease-Specific Cell Neighborhoods and Pathways",
                  "10.1016/j.isci.2022.104097",
                  "Marshall et al. (2022) iScience",
                  "b020294c-ab82-4547-b5a7-63d8ffa575ed",
                  "abe4fce1-0859-4a56-ad1e-734d79f0e6c8",
                  "Spatial transcriptomics in mouse: Puck_191112_13",
                  "b020294c-ab82-4547-b5a7-63d8ffa575ed.h5ad",
                  "15161"
                ],
                [
                  "4",
                  "4",
                  "Publication: https://doi.org/10.1038/s41591-024-03215-z Dataset Version: https://datasets.cellxgene.cziscience.com/863fc5e4-bd4a-4681-9c3d-0ee7ef54e327.h5ad curated and distributed by CZ CELLxGENE Discover in Collection: https://cellxgene.cziscience.com/collections/a96133de-e951-4e2d-ace6-59db8b3bfb1d",
                  "a96133de-e951-4e2d-ace6-59db8b3bfb1d",
                  "HTAN/HTAPP Broad - Spatio-molecular dissection of the breast cancer metastatic microenvironment",
                  "10.1038/s41591-024-03215-z",
                  "Klughammer et al. (2024) Nat Med",
                  "d7476ae2-e320-4703-8304-da5c42627e71",
                  "863fc5e4-bd4a-4681-9c3d-0ee7ef54e327",
                  "HTAPP-330-SMP-1082 scRNA-seq",
                  "d7476ae2-e320-4703-8304-da5c42627e71.h5ad",
                  "565"
                ],
                [
                  "5",
                  "5",
                  "Publication: https://doi.org/10.1016/j.isci.2022.104097 Dataset Version: https://datasets.cellxgene.cziscience.com/d5e8239f-a6a3-475a-b60d-d6991ad9a108.h5ad curated and distributed by CZ CELLxGENE Discover in Collection: https://cellxgene.cziscience.com/collections/8e880741-bf9a-4c8e-9227-934204631d2a",
                  "8e880741-bf9a-4c8e-9227-934204631d2a",
                  "High Resolution Slide-seqV2 Spatial Transcriptomics Enables Discovery of Disease-Specific Cell Neighborhoods and Pathways",
                  "10.1016/j.isci.2022.104097",
                  "Marshall et al. (2022) iScience",
                  "530c9bff-c552-45b9-be77-0de605e6858b",
                  "d5e8239f-a6a3-475a-b60d-d6991ad9a108",
                  "Spatial transcriptomics in mouse: Puck_191109_14",
                  "530c9bff-c552-45b9-be77-0de605e6858b.h5ad",
                  "12351"
                ],
                [
                  "6",
                  "6",
                  "Publication: https://doi.org/10.1002/hep4.1854 Dataset Version: https://datasets.cellxgene.cziscience.com/17616a35-1b3d-4754-9553-8ddb71267d63.h5ad curated and distributed by CZ CELLxGENE Discover in Collection: https://cellxgene.cziscience.com/collections/44531dd9-1388-4416-a117-af0a99de2294",
                  "44531dd9-1388-4416-a117-af0a99de2294",
                  "Single-Cell, Single-Nucleus, and Spatial RNA Sequencing of the Human Liver Identifies Cholangiocyte and Mesenchymal Heterogeneity",
                  "10.1002/hep4.1854",
                  "Andrews et al. (2022) Hepatology Communications",
                  "0895c838-e550-48a3-a777-dbcd35d30272",
                  "17616a35-1b3d-4754-9553-8ddb71267d63",
                  "Healthy human liver: B cells",
                  "0895c838-e550-48a3-a777-dbcd35d30272.h5ad",
                  "146"
                ],
                [
                  "7",
                  "7",
                  "Publication: https://doi.org/10.1038/s41586-024-07944-6 Dataset Version: https://datasets.cellxgene.cziscience.com/463451bb-78a0-447f-9555-b05d11472d09.h5ad curated and distributed by CZ CELLxGENE Discover in Collection: https://cellxgene.cziscience.com/collections/fc19ae6c-d7c1-4dce-b703-62c5d52061b4",
                  "fc19ae6c-d7c1-4dce-b703-62c5d52061b4",
                  "A spatial human thymus cell atlas mapped to a continuous tissue axis",
                  "10.1038/s41586-024-07944-6",
                  "Yayon et al. (2024) Nature",
                  "fbd69faa-b0c5-45ba-89c9-da938a7f5a14",
                  "463451bb-78a0-447f-9555-b05d11472d09",
                  "thymus scRNA-seq atlas - myeloid p2 subset",
                  "fbd69faa-b0c5-45ba-89c9-da938a7f5a14.h5ad",
                  "843"
                ],
                [
                  "8",
                  "8",
                  "Publication: https://doi.org/10.1016/j.isci.2022.104097 Dataset Version: https://datasets.cellxgene.cziscience.com/3d859447-b8be-4ba7-9a62-698b5f0744ac.h5ad curated and distributed by CZ CELLxGENE Discover in Collection: https://cellxgene.cziscience.com/collections/8e880741-bf9a-4c8e-9227-934204631d2a",
                  "8e880741-bf9a-4c8e-9227-934204631d2a",
                  "High Resolution Slide-seqV2 Spatial Transcriptomics Enables Discovery of Disease-Specific Cell Neighborhoods and Pathways",
                  "10.1016/j.isci.2022.104097",
                  "Marshall et al. (2022) iScience",
                  "fb83600a-7823-426e-b06b-1fcd666a3cd8",
                  "3d859447-b8be-4ba7-9a62-698b5f0744ac",
                  "Spatial transcriptomics in mouse: Puck_191109_18",
                  "fb83600a-7823-426e-b06b-1fcd666a3cd8.h5ad",
                  "19156"
                ],
                [
                  "9",
                  "9",
                  "Publication: https://doi.org/10.1126/sciimmunol.abe6291 Dataset Version: https://datasets.cellxgene.cziscience.com/3fa969b4-25e8-4e0b-b2ec-86cd9ddd2ccb.h5ad curated and distributed by CZ CELLxGENE Discover in Collection: https://cellxgene.cziscience.com/collections/3a2af25b-2338-4266-aad3-aa8d07473f50",
                  "3a2af25b-2338-4266-aad3-aa8d07473f50",
                  "Single-cell analysis of human B cell maturation predicts how antibody class switching shapes selection dynamics",
                  "10.1126/sciimmunol.abe6291",
                  "King et al. (2021) Sci. Immunol.",
                  "00ff600e-6e2e-4d76-846f-0eec4f0ae417",
                  "3fa969b4-25e8-4e0b-b2ec-86cd9ddd2ccb",
                  "Human tonsil nonlymphoid cells scRNA",
                  "00ff600e-6e2e-4d76-846f-0eec4f0ae417.h5ad",
                  "363"
                ],
                [
                  "10",
                  "10",
                  "Publication: https://doi.org/10.1016/j.isci.2022.104097 Dataset Version: https://datasets.cellxgene.cziscience.com/120b612b-a9ac-4c70-a07e-05c5f054544f.h5ad curated and distributed by CZ CELLxGENE Discover in Collection: https://cellxgene.cziscience.com/collections/8e880741-bf9a-4c8e-9227-934204631d2a",
                  "8e880741-bf9a-4c8e-9227-934204631d2a",
                  "High Resolution Slide-seqV2 Spatial Transcriptomics Enables Discovery of Disease-Specific Cell Neighborhoods and Pathways",
                  "10.1016/j.isci.2022.104097",
                  "Marshall et al. (2022) iScience",
                  "70fa6bd9-cc35-4779-9a05-fa8918c0ff8b",
                  "120b612b-a9ac-4c70-a07e-05c5f054544f",
                  "Spatial transcriptomics in mouse: Puck_191109_19",
                  "70fa6bd9-cc35-4779-9a05-fa8918c0ff8b.h5ad",
                  "22659"
                ],
                [
                  "11",
                  "11",
                  "Publication: https://doi.org/10.1016/j.isci.2022.104097 Dataset Version: https://datasets.cellxgene.cziscience.com/80bcbbcc-7424-4c97-a796-34c528cb9041.h5ad curated and distributed by CZ CELLxGENE Discover in Collection: https://cellxgene.cziscience.com/collections/8e880741-bf9a-4c8e-9227-934204631d2a",
                  "8e880741-bf9a-4c8e-9227-934204631d2a",
                  "High Resolution Slide-seqV2 Spatial Transcriptomics Enables Discovery of Disease-Specific Cell Neighborhoods and Pathways",
                  "10.1016/j.isci.2022.104097",
                  "Marshall et al. (2022) iScience",
                  "1c224f30-a9c4-48c6-951d-8d86c0a2886e",
                  "80bcbbcc-7424-4c97-a796-34c528cb9041",
                  "Spatial transcriptomics in mouse: Puck_191223_13",
                  "1c224f30-a9c4-48c6-951d-8d86c0a2886e.h5ad",
                  "16787"
                ],
                [
                  "12",
                  "12",
                  "Publication: https://doi.org/10.1016/j.isci.2022.104097 Dataset Version: https://datasets.cellxgene.cziscience.com/87e44ad8-b2da-4478-8e80-56bd3b4058e6.h5ad curated and distributed by CZ CELLxGENE Discover in Collection: https://cellxgene.cziscience.com/collections/8e880741-bf9a-4c8e-9227-934204631d2a",
                  "8e880741-bf9a-4c8e-9227-934204631d2a",
                  "High Resolution Slide-seqV2 Spatial Transcriptomics Enables Discovery of Disease-Specific Cell Neighborhoods and Pathways",
                  "10.1016/j.isci.2022.104097",
                  "Marshall et al. (2022) iScience",
                  "3483835a-fdd2-4589-8de3-530e2a13ad14",
                  "87e44ad8-b2da-4478-8e80-56bd3b4058e6",
                  "Spatial transcriptomics in mouse: Puck_191204_18",
                  "3483835a-fdd2-4589-8de3-530e2a13ad14.h5ad",
                  "27814"
                ],
                [
                  "13",
                  "13",
                  "Publication: https://doi.org/10.1038/s41588-023-01435-6 Dataset Version: https://datasets.cellxgene.cziscience.com/98fc8857-18a3-492f-b440-0628d323c07c.h5ad curated and distributed by CZ CELLxGENE Discover in Collection: https://cellxgene.cziscience.com/collections/d74b6979-efba-47cd-990a-9d80ccf29055",
                  "d74b6979-efba-47cd-990a-9d80ccf29055",
                  "Spatial transcriptomic maps of whole mouse embryos reveal principles of neural tube patterning",
                  "10.1038/s41588-023-01435-6",
                  "Sampath Kumar et al. (2023) Nat Genet",
                  "c233842b-9430-41b8-8ec6-4342428e43f3",
                  "98fc8857-18a3-492f-b440-0628d323c07c",
                  "mouse_trunkTbx6WT_KO",
                  "c233842b-9430-41b8-8ec6-4342428e43f3.h5ad",
                  "499"
                ],
                [
                  "14",
                  "14",
                  "Publication: https://doi.org/10.1016/j.isci.2022.104097 Dataset Version: https://datasets.cellxgene.cziscience.com/8369a5ce-8199-4517-8a43-df21bb543413.h5ad curated and distributed by CZ CELLxGENE Discover in Collection: https://cellxgene.cziscience.com/collections/8e880741-bf9a-4c8e-9227-934204631d2a",
                  "8e880741-bf9a-4c8e-9227-934204631d2a",
                  "High Resolution Slide-seqV2 Spatial Transcriptomics Enables Discovery of Disease-Specific Cell Neighborhoods and Pathways",
                  "10.1016/j.isci.2022.104097",
                  "Marshall et al. (2022) iScience",
                  "ce19be17-f066-452f-922e-b1f232ce8af5",
                  "8369a5ce-8199-4517-8a43-df21bb543413",
                  "Spatial transcriptomics in mouse: Puck_191109_10",
                  "ce19be17-f066-452f-922e-b1f232ce8af5.h5ad",
                  "15592"
                ],
                [
                  "15",
                  "15",
                  "Publication: https://doi.org/10.1016/j.isci.2022.104097 Dataset Version: https://datasets.cellxgene.cziscience.com/eabd8ee3-be20-4ab0-8b34-50563b5ec879.h5ad curated and distributed by CZ CELLxGENE Discover in Collection: https://cellxgene.cziscience.com/collections/8e880741-bf9a-4c8e-9227-934204631d2a",
                  "8e880741-bf9a-4c8e-9227-934204631d2a",
                  "High Resolution Slide-seqV2 Spatial Transcriptomics Enables Discovery of Disease-Specific Cell Neighborhoods and Pathways",
                  "10.1016/j.isci.2022.104097",
                  "Marshall et al. (2022) iScience",
                  "c5ac9bb9-9597-4c0e-a985-1c3947288785",
                  "eabd8ee3-be20-4ab0-8b34-50563b5ec879",
                  "Spatial transcriptomics in mouse: Puck_191112_07",
                  "c5ac9bb9-9597-4c0e-a985-1c3947288785.h5ad",
                  "21181"
                ],
                [
                  "16",
                  "16",
                  "Publication: https://doi.org/10.1016/j.isci.2022.104097 Dataset Version: https://datasets.cellxgene.cziscience.com/5ea87310-c584-4c7d-b2ba-44f31c2e7ea8.h5ad curated and distributed by CZ CELLxGENE Discover in Collection: https://cellxgene.cziscience.com/collections/8e880741-bf9a-4c8e-9227-934204631d2a",
                  "8e880741-bf9a-4c8e-9227-934204631d2a",
                  "High Resolution Slide-seqV2 Spatial Transcriptomics Enables Discovery of Disease-Specific Cell Neighborhoods and Pathways",
                  "10.1016/j.isci.2022.104097",
                  "Marshall et al. (2022) iScience",
                  "72ed5f5b-e685-474d-b436-3bcc8c597e8e",
                  "5ea87310-c584-4c7d-b2ba-44f31c2e7ea8",
                  "Spatial transcriptomics in mouse: Puck_191112_04",
                  "72ed5f5b-e685-474d-b436-3bcc8c597e8e.h5ad",
                  "13822"
                ],
                [
                  "17",
                  "17",
                  "Publication: https://doi.org/10.1016/j.isci.2022.104097 Dataset Version: https://datasets.cellxgene.cziscience.com/5acbe8e6-cc23-49a0-b74e-ac1d6921e295.h5ad curated and distributed by CZ CELLxGENE Discover in Collection: https://cellxgene.cziscience.com/collections/8e880741-bf9a-4c8e-9227-934204631d2a",
                  "8e880741-bf9a-4c8e-9227-934204631d2a",
                  "High Resolution Slide-seqV2 Spatial Transcriptomics Enables Discovery of Disease-Specific Cell Neighborhoods and Pathways",
                  "10.1016/j.isci.2022.104097",
                  "Marshall et al. (2022) iScience",
                  "b24705ee-87c0-45f2-adfb-2819fdc16225",
                  "5acbe8e6-cc23-49a0-b74e-ac1d6921e295",
                  "Spatial transcriptomics in mouse: Puck_200104_04",
                  "b24705ee-87c0-45f2-adfb-2819fdc16225.h5ad",
                  "25817"
                ],
                [
                  "18",
                  "18",
                  "Publication: https://doi.org/10.1038/s41593-020-00764-7 Dataset Version: https://datasets.cellxgene.cziscience.com/017837df-b8be-4a4f-bc08-c5f14bd8815a.h5ad curated and distributed by CZ CELLxGENE Discover in Collection: https://cellxgene.cziscience.com/collections/180bff9c-c8a5-4539-b13b-ddbc00d643e6",
                  "180bff9c-c8a5-4539-b13b-ddbc00d643e6",
                  "Molecular characterization of selectively vulnerable neurons in Alzheimer's Disease",
                  "10.1038/s41593-020-00764-7",
                  "Leng et al. (2021) Nat Neurosci",
                  "bdacc907-7c26-419f-8808-969eab3ca2e8",
                  "017837df-b8be-4a4f-bc08-c5f14bd8815a",
                  "Molecular characterization of selectively vulnerable neurons in Alzheimer’s Disease: SFG microglia",
                  "bdacc907-7c26-419f-8808-969eab3ca2e8.h5ad",
                  "3799"
                ],
                [
                  "19",
                  "19",
                  "Publication: https://doi.org/10.1038/s41588-023-01435-6 Dataset Version: https://datasets.cellxgene.cziscience.com/09224950-b662-4db3-88f1-17d1525bc2f6.h5ad curated and distributed by CZ CELLxGENE Discover in Collection: https://cellxgene.cziscience.com/collections/d74b6979-efba-47cd-990a-9d80ccf29055",
                  "d74b6979-efba-47cd-990a-9d80ccf29055",
                  "Spatial transcriptomic maps of whole mouse embryos reveal principles of neural tube patterning",
                  "10.1038/s41588-023-01435-6",
                  "Sampath Kumar et al. (2023) Nat Genet",
                  "6c0f2e18-3793-4d90-88b5-12d9d6f0f4e2",
                  "09224950-b662-4db3-88f1-17d1525bc2f6",
                  "E9.0: Puck_220718_02",
                  "6c0f2e18-3793-4d90-88b5-12d9d6f0f4e2.h5ad",
                  "703"
                ],
                [
                  "20",
                  "20",
                  "Publication: https://doi.org/10.1016/j.isci.2022.104097 Dataset Version: https://datasets.cellxgene.cziscience.com/0aeba889-34c7-427c-8c0b-5abf70fda5e8.h5ad curated and distributed by CZ CELLxGENE Discover in Collection: https://cellxgene.cziscience.com/collections/8e880741-bf9a-4c8e-9227-934204631d2a",
                  "8e880741-bf9a-4c8e-9227-934204631d2a",
                  "High Resolution Slide-seqV2 Spatial Transcriptomics Enables Discovery of Disease-Specific Cell Neighborhoods and Pathways",
                  "10.1016/j.isci.2022.104097",
                  "Marshall et al. (2022) iScience",
                  "90ef2377-e4d0-41c7-924f-6f9ee7a9237d",
                  "0aeba889-34c7-427c-8c0b-5abf70fda5e8",
                  "Spatial transcriptomics in the healthy human kidney: Puck_200113_07",
                  "90ef2377-e4d0-41c7-924f-6f9ee7a9237d.h5ad",
                  "11456"
                ],
                [
                  "21",
                  "21",
                  "Publication: https://doi.org/10.1101/2024.09.10.612293 Dataset Version: https://datasets.cellxgene.cziscience.com/cb46feff-d5bd-41e6-91c5-ce4daddc066e.h5ad curated and distributed by CZ CELLxGENE Discover in Collection: https://cellxgene.cziscience.com/collections/8f17ac63-aaba-44b5-9b78-60f121da4c2f",
                  "8f17ac63-aaba-44b5-9b78-60f121da4c2f",
                  "A Coding and Non-Coding Atlas of the Human Arterial Cell",
                  "10.1101/2024.09.10.612293",
                  "Zhao et al. (2024) bioRxiv",
                  "b2005457-dede-4434-a9c3-dbe41fcd542e",
                  "cb46feff-d5bd-41e6-91c5-ce4daddc066e",
                  "slide-seqV2 analysis of aorta",
                  "b2005457-dede-4434-a9c3-dbe41fcd542e.h5ad",
                  "6142"
                ],
                [
                  "22",
                  "22",
                  "Publication: https://doi.org/10.1038/s41467-022-29450-x Dataset Version: https://datasets.cellxgene.cziscience.com/f223d2b0-d878-443b-8bef-f182fe1f22d1.h5ad curated and distributed by CZ CELLxGENE Discover in Collection: https://cellxgene.cziscience.com/collections/bf325905-5e8e-42e3-933d-9a9053e9af80",
                  "bf325905-5e8e-42e3-933d-9a9053e9af80",
                  "Single-cell Atlas of common variable immunodeficiency shows germinal center-associated epigenetic dysregulation in B-cell responses",
                  "10.1038/s41467-022-29450-x",
                  "Rodríguez-Ubreva et al. (2022) Nat Commun",
                  "a5d95a42-0137-496f-8a60-101e17f263c8",
                  "f223d2b0-d878-443b-8bef-f182fe1f22d1",
                  "Steady-state B cells - scRNA-seq",
                  "a5d95a42-0137-496f-8a60-101e17f263c8.h5ad",
                  "1324"
                ],
                [
                  "23",
                  "23",
                  "Publication: https://doi.org/10.1016/j.isci.2022.104097 Dataset Version: https://datasets.cellxgene.cziscience.com/0effbd1e-4397-407f-8ab7-7c087828e307.h5ad curated and distributed by CZ CELLxGENE Discover in Collection: https://cellxgene.cziscience.com/collections/8e880741-bf9a-4c8e-9227-934204631d2a",
                  "8e880741-bf9a-4c8e-9227-934204631d2a",
                  "High Resolution Slide-seqV2 Spatial Transcriptomics Enables Discovery of Disease-Specific Cell Neighborhoods and Pathways",
                  "10.1016/j.isci.2022.104097",
                  "Marshall et al. (2022) iScience",
                  "2fb24a91-55b9-4cc4-9a11-d46c373d53c1",
                  "0effbd1e-4397-407f-8ab7-7c087828e307",
                  "Spatial transcriptomics in mouse: Puck_191109_08",
                  "2fb24a91-55b9-4cc4-9a11-d46c373d53c1.h5ad",
                  "30182"
                ],
                [
                  "24",
                  "24",
                  "Publication: https://doi.org/10.1038/s41590-021-01059-0 Dataset Version: https://datasets.cellxgene.cziscience.com/8b1e0d38-c48a-4e60-84c4-744a1d8ad2a1.h5ad curated and distributed by CZ CELLxGENE Discover in Collection: https://cellxgene.cziscience.com/collections/93eebe82-d8c3-41bc-a906-63b5b5f24a9d",
                  "93eebe82-d8c3-41bc-a906-63b5b5f24a9d",
                  "Single-cell proteo-genomic reference maps of the hematopoietic system enable the purification and massive profiling of precisely defined cell states",
                  "10.1038/s41590-021-01059-0",
                  "Triana et al. (2021) Nat Immunol",
                  "d3566d6a-a455-4a15-980f-45eb29114cab",
                  "8b1e0d38-c48a-4e60-84c4-744a1d8ad2a1",
                  "blood and bone marrow from a healthy young donor",
                  "d3566d6a-a455-4a15-980f-45eb29114cab.h5ad",
                  "15502"
                ],
                [
                  "25",
                  "25",
                  "Publication: https://doi.org/10.1016/j.isci.2022.104097 Dataset Version: https://datasets.cellxgene.cziscience.com/24218d17-2680-4214-9c92-b9bb370dbbbd.h5ad curated and distributed by CZ CELLxGENE Discover in Collection: https://cellxgene.cziscience.com/collections/8e880741-bf9a-4c8e-9227-934204631d2a",
                  "8e880741-bf9a-4c8e-9227-934204631d2a",
                  "High Resolution Slide-seqV2 Spatial Transcriptomics Enables Discovery of Disease-Specific Cell Neighborhoods and Pathways",
                  "10.1016/j.isci.2022.104097",
                  "Marshall et al. (2022) iScience",
                  "5e73d977-8e54-4406-a5f0-711bf0d3c228",
                  "24218d17-2680-4214-9c92-b9bb370dbbbd",
                  "Spatial transcriptomics in mouse: Puck_191204_09",
                  "5e73d977-8e54-4406-a5f0-711bf0d3c228.h5ad",
                  "17138"
                ],
                [
                  "26",
                  "26",
                  "Publication: https://doi.org/10.1016/j.immuni.2022.09.002 Dataset Version: https://datasets.cellxgene.cziscience.com/6a01686c-f4d6-4fa0-b255-da0dcfcd1804.h5ad curated and distributed by CZ CELLxGENE Discover in Collection: https://cellxgene.cziscience.com/collections/02b01703-bf1b-48de-b99a-23bef8cccc81",
                  "02b01703-bf1b-48de-b99a-23bef8cccc81",
                  "Spatially mapping T cell receptors and transcriptomes",
                  "10.1016/j.immuni.2022.09.002",
                  "Liu et al. (2022) Immunity",
                  "c62ac8e5-72d1-48d1-8bb5-9e66fd2b1e19",
                  "6a01686c-f4d6-4fa0-b255-da0dcfcd1804",
                  "Normal Spleen OT1 Puck_200727_02",
                  "c62ac8e5-72d1-48d1-8bb5-9e66fd2b1e19.h5ad",
                  "11510"
                ],
                [
                  "27",
                  "27",
                  "Publication: https://doi.org/10.1016/j.isci.2022.104097 Dataset Version: https://datasets.cellxgene.cziscience.com/7a361462-363f-4dc9-8d84-1605fabc9691.h5ad curated and distributed by CZ CELLxGENE Discover in Collection: https://cellxgene.cziscience.com/collections/8e880741-bf9a-4c8e-9227-934204631d2a",
                  "8e880741-bf9a-4c8e-9227-934204631d2a",
                  "High Resolution Slide-seqV2 Spatial Transcriptomics Enables Discovery of Disease-Specific Cell Neighborhoods and Pathways",
                  "10.1016/j.isci.2022.104097",
                  "Marshall et al. (2022) iScience",
                  "10eed666-34e0-4b98-9384-179993873aa8",
                  "7a361462-363f-4dc9-8d84-1605fabc9691",
                  "Spatial transcriptomics in mouse: Puck_200104_03",
                  "10eed666-34e0-4b98-9384-179993873aa8.h5ad",
                  "16027"
                ],
                [
                  "28",
                  "28",
                  "Publication: https://doi.org/10.1038/s41588-023-01435-6 Dataset Version: https://datasets.cellxgene.cziscience.com/28758f93-4149-48c8-a4ff-5d0ff28b9d16.h5ad curated and distributed by CZ CELLxGENE Discover in Collection: https://cellxgene.cziscience.com/collections/d74b6979-efba-47cd-990a-9d80ccf29055",
                  "d74b6979-efba-47cd-990a-9d80ccf29055",
                  "Spatial transcriptomic maps of whole mouse embryos reveal principles of neural tube patterning",
                  "10.1038/s41588-023-01435-6",
                  "Sampath Kumar et al. (2023) Nat Genet",
                  "e545cafe-80f4-494e-bbb1-e95c86955756",
                  "28758f93-4149-48c8-a4ff-5d0ff28b9d16",
                  "E9.0: Puck_220718_10",
                  "e545cafe-80f4-494e-bbb1-e95c86955756.h5ad",
                  "1336"
                ],
                [
                  "29",
                  "29",
                  "Publication: https://doi.org/10.1016/j.isci.2022.104097 Dataset Version: https://datasets.cellxgene.cziscience.com/67855e83-c9c1-4b9c-b6bb-d5cef75f212a.h5ad curated and distributed by CZ CELLxGENE Discover in Collection: https://cellxgene.cziscience.com/collections/8e880741-bf9a-4c8e-9227-934204631d2a",
                  "8e880741-bf9a-4c8e-9227-934204631d2a",
                  "High Resolution Slide-seqV2 Spatial Transcriptomics Enables Discovery of Disease-Specific Cell Neighborhoods and Pathways",
                  "10.1016/j.isci.2022.104097",
                  "Marshall et al. (2022) iScience",
                  "9098fb44-ac81-4ac6-be8e-464f10ec7128",
                  "67855e83-c9c1-4b9c-b6bb-d5cef75f212a",
                  "Spatial transcriptomics in mouse: Puck_191223_10",
                  "9098fb44-ac81-4ac6-be8e-464f10ec7128.h5ad",
                  "14961"
                ],
                [
                  "30",
                  "30",
                  "Publication: https://doi.org/10.1016/j.immuni.2022.09.002 Dataset Version: https://datasets.cellxgene.cziscience.com/23faf7da-1792-4069-b20c-a035fd301698.h5ad curated and distributed by CZ CELLxGENE Discover in Collection: https://cellxgene.cziscience.com/collections/02b01703-bf1b-48de-b99a-23bef8cccc81",
                  "02b01703-bf1b-48de-b99a-23bef8cccc81",
                  "Spatially mapping T cell receptors and transcriptomes",
                  "10.1016/j.immuni.2022.09.002",
                  "Liu et al. (2022) Immunity",
                  "76bb43ff-e2f5-4513-a0f0-1059c860c3b7",
                  "23faf7da-1792-4069-b20c-a035fd301698",
                  "Cutaneous melanoma, lymph node metastasis Puck_220408_08",
                  "76bb43ff-e2f5-4513-a0f0-1059c860c3b7.h5ad",
                  "19034"
                ],
                [
                  "31",
                  "31",
                  "Publication: https://doi.org/10.1016/j.celrep.2019.12.082 Dataset Version: https://datasets.cellxgene.cziscience.com/3e706d81-4387-4405-9401-795671be1d8f.h5ad curated and distributed by CZ CELLxGENE Discover in Collection: https://cellxgene.cziscience.com/collections/939769a8-d8d2-4d01-abfc-55699893fd49",
                  "939769a8-d8d2-4d01-abfc-55699893fd49",
                  "Integration of eQTL and a Single-Cell Atlas in the Human Eye Identifies Causal Genes for Age-Related Macular Degeneration",
                  "10.1016/j.celrep.2019.12.082",
                  "Orozco et al. (2020) Cell Reports",
                  "de17ac25-550a-4018-be75-bbb485a0636e",
                  "3e706d81-4387-4405-9401-795671be1d8f",
                  "Myeloid cells of human eye",
                  "de17ac25-550a-4018-be75-bbb485a0636e.h5ad",
                  "395"
                ],
                [
                  "32",
                  "32",
                  "Publication: https://doi.org/10.1016/j.isci.2022.104097 Dataset Version: https://datasets.cellxgene.cziscience.com/6fc5cbd5-ad26-4518-a19f-843f37fcb893.h5ad curated and distributed by CZ CELLxGENE Discover in Collection: https://cellxgene.cziscience.com/collections/8e880741-bf9a-4c8e-9227-934204631d2a",
                  "8e880741-bf9a-4c8e-9227-934204631d2a",
                  "High Resolution Slide-seqV2 Spatial Transcriptomics Enables Discovery of Disease-Specific Cell Neighborhoods and Pathways",
                  "10.1016/j.isci.2022.104097",
                  "Marshall et al. (2022) iScience",
                  "707acde8-4e99-4e31-8da9-782ded87d136",
                  "6fc5cbd5-ad26-4518-a19f-843f37fcb893",
                  "Spatial transcriptomics in mouse: Puck_191204_24",
                  "707acde8-4e99-4e31-8da9-782ded87d136.h5ad",
                  "27220"
                ],
                [
                  "33",
                  "33",
                  "Publication: https://doi.org/10.1016/j.isci.2022.104097 Dataset Version: https://datasets.cellxgene.cziscience.com/bfa2695d-0d7f-43d5-8126-5f25415701d6.h5ad curated and distributed by CZ CELLxGENE Discover in Collection: https://cellxgene.cziscience.com/collections/8e880741-bf9a-4c8e-9227-934204631d2a",
                  "8e880741-bf9a-4c8e-9227-934204631d2a",
                  "High Resolution Slide-seqV2 Spatial Transcriptomics Enables Discovery of Disease-Specific Cell Neighborhoods and Pathways",
                  "10.1016/j.isci.2022.104097",
                  "Marshall et al. (2022) iScience",
                  "3679ae7d-d70e-4c09-a8bb-78b8ae1006ff",
                  "bfa2695d-0d7f-43d5-8126-5f25415701d6",
                  "Spatial transcriptomics in the healthy human kidney: Puck_200131_25",
                  "3679ae7d-d70e-4c09-a8bb-78b8ae1006ff.h5ad",
                  "10701"
                ],
                [
                  "34",
                  "34",
                  "Publication: https://doi.org/10.1016/j.isci.2022.104097 Dataset Version: https://datasets.cellxgene.cziscience.com/e0256600-a917-4ec7-abb9-34131f1e1e71.h5ad curated and distributed by CZ CELLxGENE Discover in Collection: https://cellxgene.cziscience.com/collections/8e880741-bf9a-4c8e-9227-934204631d2a",
                  "8e880741-bf9a-4c8e-9227-934204631d2a",
                  "High Resolution Slide-seqV2 Spatial Transcriptomics Enables Discovery of Disease-Specific Cell Neighborhoods and Pathways",
                  "10.1016/j.isci.2022.104097",
                  "Marshall et al. (2022) iScience",
                  "9246671b-f814-4252-8f1a-57f4f5e28f81",
                  "e0256600-a917-4ec7-abb9-34131f1e1e71",
                  "Spatial transcriptomics in mouse: Puck_200210_04",
                  "9246671b-f814-4252-8f1a-57f4f5e28f81.h5ad",
                  "20424"
                ],
                [
                  "35",
                  "35",
                  "Publication: https://doi.org/10.1016/j.isci.2022.104097 Dataset Version: https://datasets.cellxgene.cziscience.com/ca32f1db-a218-443f-b367-d5c0a39fbf09.h5ad curated and distributed by CZ CELLxGENE Discover in Collection: https://cellxgene.cziscience.com/collections/8e880741-bf9a-4c8e-9227-934204631d2a",
                  "8e880741-bf9a-4c8e-9227-934204631d2a",
                  "High Resolution Slide-seqV2 Spatial Transcriptomics Enables Discovery of Disease-Specific Cell Neighborhoods and Pathways",
                  "10.1016/j.isci.2022.104097",
                  "Marshall et al. (2022) iScience",
                  "f9f63905-34b2-4c61-b747-1df6c7a35a6a",
                  "ca32f1db-a218-443f-b367-d5c0a39fbf09",
                  "Spatial transcriptomics in mouse: Puck_200127_07",
                  "f9f63905-34b2-4c61-b747-1df6c7a35a6a.h5ad",
                  "17363"
                ],
                [
                  "36",
                  "36",
                  "Publication: https://doi.org/10.1016/j.isci.2022.104097 Dataset Version: https://datasets.cellxgene.cziscience.com/1b667779-e452-45f7-87a1-a8a5ad19bab2.h5ad curated and distributed by CZ CELLxGENE Discover in Collection: https://cellxgene.cziscience.com/collections/8e880741-bf9a-4c8e-9227-934204631d2a",
                  "8e880741-bf9a-4c8e-9227-934204631d2a",
                  "High Resolution Slide-seqV2 Spatial Transcriptomics Enables Discovery of Disease-Specific Cell Neighborhoods and Pathways",
                  "10.1016/j.isci.2022.104097",
                  "Marshall et al. (2022) iScience",
                  "34cd019c-c79d-4f4f-8f95-8934062dc132",
                  "1b667779-e452-45f7-87a1-a8a5ad19bab2",
                  "Spatial transcriptomics in mouse: Puck_191206_03",
                  "34cd019c-c79d-4f4f-8f95-8934062dc132.h5ad",
                  "18542"
                ],
                [
                  "37",
                  "37",
                  "Publication: https://doi.org/10.1038/s41593-020-00764-7 Dataset Version: https://datasets.cellxgene.cziscience.com/bb8f672d-593d-4820-b963-bc7ba00eae18.h5ad curated and distributed by CZ CELLxGENE Discover in Collection: https://cellxgene.cziscience.com/collections/180bff9c-c8a5-4539-b13b-ddbc00d643e6",
                  "180bff9c-c8a5-4539-b13b-ddbc00d643e6",
                  "Molecular characterization of selectively vulnerable neurons in Alzheimer's Disease",
                  "10.1038/s41593-020-00764-7",
                  "Leng et al. (2021) Nat Neurosci",
                  "9f1049ac-f8b7-45ad-8e31-6e96c3e5058f",
                  "bb8f672d-593d-4820-b963-bc7ba00eae18",
                  "Molecular characterization of selectively vulnerable neurons in Alzheimer’s Disease: EC microglia",
                  "9f1049ac-f8b7-45ad-8e31-6e96c3e5058f.h5ad",
                  "5572"
                ],
                [
                  "38",
                  "38",
                  "Publication: https://doi.org/10.1016/j.isci.2022.104097 Dataset Version: https://datasets.cellxgene.cziscience.com/09168c5d-0f16-4375-be0e-b527d2561919.h5ad curated and distributed by CZ CELLxGENE Discover in Collection: https://cellxgene.cziscience.com/collections/8e880741-bf9a-4c8e-9227-934204631d2a",
                  "8e880741-bf9a-4c8e-9227-934204631d2a",
                  "High Resolution Slide-seqV2 Spatial Transcriptomics Enables Discovery of Disease-Specific Cell Neighborhoods and Pathways",
                  "10.1016/j.isci.2022.104097",
                  "Marshall et al. (2022) iScience",
                  "a5a6c83a-3afe-4a87-8c32-691118c46e51",
                  "09168c5d-0f16-4375-be0e-b527d2561919",
                  "Spatial transcriptomics in the healthy human kidney: Puck_200113_09",
                  "a5a6c83a-3afe-4a87-8c32-691118c46e51.h5ad",
                  "13780"
                ],
                [
                  "39",
                  "39",
                  "Publication: https://doi.org/10.1016/j.isci.2022.104097 Dataset Version: https://datasets.cellxgene.cziscience.com/9019b24b-d6ed-4923-a7fe-b40445f853b2.h5ad curated and distributed by CZ CELLxGENE Discover in Collection: https://cellxgene.cziscience.com/collections/8e880741-bf9a-4c8e-9227-934204631d2a",
                  "8e880741-bf9a-4c8e-9227-934204631d2a",
                  "High Resolution Slide-seqV2 Spatial Transcriptomics Enables Discovery of Disease-Specific Cell Neighborhoods and Pathways",
                  "10.1016/j.isci.2022.104097",
                  "Marshall et al. (2022) iScience",
                  "a21a47fe-1516-47d0-aeca-fd90d729fa4a",
                  "9019b24b-d6ed-4923-a7fe-b40445f853b2",
                  "Spatial transcriptomics in mouse: Puck_200131_23",
                  "a21a47fe-1516-47d0-aeca-fd90d729fa4a.h5ad",
                  "12533"
                ],
                [
                  "40",
                  "40",
                  "Publication: https://doi.org/10.1016/j.cell.2022.11.005 Dataset Version: https://datasets.cellxgene.cziscience.com/33a5762d-7dd0-4e48-b475-ef7715fe15f6.h5ad curated and distributed by CZ CELLxGENE Discover in Collection: https://cellxgene.cziscience.com/collections/2d2e2acd-dade-489f-a2da-6c11aa654028",
                  "2d2e2acd-dade-489f-a2da-6c11aa654028",
                  "A human fetal lung cell atlas uncovers proximal-distal gradients of differentiation and key regulators of epithelial fates",
                  "10.1016/j.cell.2022.11.005",
                  "He et al. (2022) Cell",
                  "703f00e6-b996-48e5-bc34-00c41b9876f4",
                  "33a5762d-7dd0-4e48-b475-ef7715fe15f6",
                  "PNS",
                  "703f00e6-b996-48e5-bc34-00c41b9876f4.h5ad",
                  "649"
                ],
                [
                  "41",
                  "41",
                  "Publication: https://doi.org/10.1016/j.isci.2022.104097 Dataset Version: https://datasets.cellxgene.cziscience.com/f26c5734-29c4-4e55-93c3-deb3bb68ca51.h5ad curated and distributed by CZ CELLxGENE Discover in Collection: https://cellxgene.cziscience.com/collections/8e880741-bf9a-4c8e-9227-934204631d2a",
                  "8e880741-bf9a-4c8e-9227-934204631d2a",
                  "High Resolution Slide-seqV2 Spatial Transcriptomics Enables Discovery of Disease-Specific Cell Neighborhoods and Pathways",
                  "10.1016/j.isci.2022.104097",
                  "Marshall et al. (2022) iScience",
                  "286fead0-7be7-46db-b358-7abc6d39adf7",
                  "f26c5734-29c4-4e55-93c3-deb3bb68ca51",
                  "Spatial transcriptomics in mouse: Puck_191109_09",
                  "286fead0-7be7-46db-b358-7abc6d39adf7.h5ad",
                  "29011"
                ],
                [
                  "42",
                  "42",
                  "Publication: https://doi.org/10.1016/j.isci.2022.104097 Dataset Version: https://datasets.cellxgene.cziscience.com/a6554b87-fa7d-4896-972d-79231c56e3a3.h5ad curated and distributed by CZ CELLxGENE Discover in Collection: https://cellxgene.cziscience.com/collections/8e880741-bf9a-4c8e-9227-934204631d2a",
                  "8e880741-bf9a-4c8e-9227-934204631d2a",
                  "High Resolution Slide-seqV2 Spatial Transcriptomics Enables Discovery of Disease-Specific Cell Neighborhoods and Pathways",
                  "10.1016/j.isci.2022.104097",
                  "Marshall et al. (2022) iScience",
                  "0738f538-ff2f-4346-b2eb-72704c291188",
                  "a6554b87-fa7d-4896-972d-79231c56e3a3",
                  "Spatial transcriptomics in the healthy human kidney: Puck_200113_12",
                  "0738f538-ff2f-4346-b2eb-72704c291188.h5ad",
                  "14010"
                ],
                [
                  "43",
                  "43",
                  "Publication: https://doi.org/10.1038/s41588-023-01435-6 Dataset Version: https://datasets.cellxgene.cziscience.com/034ae94c-d804-4073-8f8f-77761a2bb88b.h5ad curated and distributed by CZ CELLxGENE Discover in Collection: https://cellxgene.cziscience.com/collections/d74b6979-efba-47cd-990a-9d80ccf29055",
                  "d74b6979-efba-47cd-990a-9d80ccf29055",
                  "Spatial transcriptomic maps of whole mouse embryos reveal principles of neural tube patterning",
                  "10.1038/s41588-023-01435-6",
                  "Sampath Kumar et al. (2023) Nat Genet",
                  "63521245-fe33-4e0f-870a-8f5d083f69f9",
                  "034ae94c-d804-4073-8f8f-77761a2bb88b",
                  "E9.0: Puck_220615_39",
                  "63521245-fe33-4e0f-870a-8f5d083f69f9.h5ad",
                  "2567"
                ],
                [
                  "44",
                  "44",
                  "Publication: https://doi.org/10.1016/j.isci.2022.104097 Dataset Version: https://datasets.cellxgene.cziscience.com/7aa01b94-92aa-42fd-8a47-b73d0c8e2c60.h5ad curated and distributed by CZ CELLxGENE Discover in Collection: https://cellxgene.cziscience.com/collections/8e880741-bf9a-4c8e-9227-934204631d2a",
                  "8e880741-bf9a-4c8e-9227-934204631d2a",
                  "High Resolution Slide-seqV2 Spatial Transcriptomics Enables Discovery of Disease-Specific Cell Neighborhoods and Pathways",
                  "10.1016/j.isci.2022.104097",
                  "Marshall et al. (2022) iScience",
                  "68bcbb10-cc0d-435b-ba76-ab03c040967b",
                  "7aa01b94-92aa-42fd-8a47-b73d0c8e2c60",
                  "Spatial transcriptomics in mouse: Puck_200131_22",
                  "68bcbb10-cc0d-435b-ba76-ab03c040967b.h5ad",
                  "12541"
                ],
                [
                  "45",
                  "45",
                  "Publication: https://doi.org/10.1016/j.isci.2022.104097 Dataset Version: https://datasets.cellxgene.cziscience.com/dc0c7c46-7f06-4d96-aebb-26d38ad8f57a.h5ad curated and distributed by CZ CELLxGENE Discover in Collection: https://cellxgene.cziscience.com/collections/8e880741-bf9a-4c8e-9227-934204631d2a",
                  "8e880741-bf9a-4c8e-9227-934204631d2a",
                  "High Resolution Slide-seqV2 Spatial Transcriptomics Enables Discovery of Disease-Specific Cell Neighborhoods and Pathways",
                  "10.1016/j.isci.2022.104097",
                  "Marshall et al. (2022) iScience",
                  "c44e22ae-b746-4c09-9826-ca0650e828d4",
                  "dc0c7c46-7f06-4d96-aebb-26d38ad8f57a",
                  "Spatial transcriptomics in mouse: Puck_200127_10",
                  "c44e22ae-b746-4c09-9826-ca0650e828d4.h5ad",
                  "12132"
                ],
                [
                  "46",
                  "46",
                  "Publication: https://doi.org/10.1038/s41588-023-01435-6 Dataset Version: https://datasets.cellxgene.cziscience.com/b4421aa2-864a-4db4-b847-950d81a08fbb.h5ad curated and distributed by CZ CELLxGENE Discover in Collection: https://cellxgene.cziscience.com/collections/d74b6979-efba-47cd-990a-9d80ccf29055",
                  "d74b6979-efba-47cd-990a-9d80ccf29055",
                  "Spatial transcriptomic maps of whole mouse embryos reveal principles of neural tube patterning",
                  "10.1038/s41588-023-01435-6",
                  "Sampath Kumar et al. (2023) Nat Genet",
                  "024593dd-d237-4eaa-aff2-9c9d87be595e",
                  "b4421aa2-864a-4db4-b847-950d81a08fbb",
                  "E9.0: Puck_220615_38",
                  "024593dd-d237-4eaa-aff2-9c9d87be595e.h5ad",
                  "2872"
                ],
                [
                  "47",
                  "47",
                  "Publication: https://doi.org/10.1016/j.isci.2022.104097 Dataset Version: https://datasets.cellxgene.cziscience.com/84c11aae-6b53-44bc-830b-53365bcd9d93.h5ad curated and distributed by CZ CELLxGENE Discover in Collection: https://cellxgene.cziscience.com/collections/8e880741-bf9a-4c8e-9227-934204631d2a",
                  "8e880741-bf9a-4c8e-9227-934204631d2a",
                  "High Resolution Slide-seqV2 Spatial Transcriptomics Enables Discovery of Disease-Specific Cell Neighborhoods and Pathways",
                  "10.1016/j.isci.2022.104097",
                  "Marshall et al. (2022) iScience",
                  "34893c3b-1776-4a4c-9ea0-9f86ad168ae4",
                  "84c11aae-6b53-44bc-830b-53365bcd9d93",
                  "Spatial transcriptomics in mouse: Puck_191112_02",
                  "34893c3b-1776-4a4c-9ea0-9f86ad168ae4.h5ad",
                  "20568"
                ],
                [
                  "48",
                  "48",
                  "Publication: https://doi.org/10.1038/s41588-023-01435-6 Dataset Version: https://datasets.cellxgene.cziscience.com/25167967-ba95-4423-9008-49499ec92aff.h5ad curated and distributed by CZ CELLxGENE Discover in Collection: https://cellxgene.cziscience.com/collections/d74b6979-efba-47cd-990a-9d80ccf29055",
                  "d74b6979-efba-47cd-990a-9d80ccf29055",
                  "Spatial transcriptomic maps of whole mouse embryos reveal principles of neural tube patterning",
                  "10.1038/s41588-023-01435-6",
                  "Sampath Kumar et al. (2023) Nat Genet",
                  "fe70db78-88b8-4bf7-8449-7aa177d51978",
                  "25167967-ba95-4423-9008-49499ec92aff",
                  "E9.0: Puck_220718_36",
                  "fe70db78-88b8-4bf7-8449-7aa177d51978.h5ad",
                  "3374"
                ],
                [
                  "49",
                  "49",
                  "Publication: https://doi.org/10.1038/s41588-023-01435-6 Dataset Version: https://datasets.cellxgene.cziscience.com/cc385c8d-50cb-4040-a949-bb657c7b2c05.h5ad curated and distributed by CZ CELLxGENE Discover in Collection: https://cellxgene.cziscience.com/collections/d74b6979-efba-47cd-990a-9d80ccf29055",
                  "d74b6979-efba-47cd-990a-9d80ccf29055",
                  "Spatial transcriptomic maps of whole mouse embryos reveal principles of neural tube patterning",
                  "10.1038/s41588-023-01435-6",
                  "Sampath Kumar et al. (2023) Nat Genet",
                  "b09f9b6a-7c4b-4345-9298-153bdc41191a",
                  "cc385c8d-50cb-4040-a949-bb657c7b2c05",
                  "E9.0: Puck_220615_36",
                  "b09f9b6a-7c4b-4345-9298-153bdc41191a.h5ad",
                  "3076"
                ]
              ],
              "shape": {
                "columns": 11,
                "rows": 1573
              }
            },
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>soma_joinid</th>\n",
              "      <th>citation</th>\n",
              "      <th>collection_id</th>\n",
              "      <th>collection_name</th>\n",
              "      <th>collection_doi</th>\n",
              "      <th>collection_doi_label</th>\n",
              "      <th>dataset_id</th>\n",
              "      <th>dataset_version_id</th>\n",
              "      <th>dataset_title</th>\n",
              "      <th>dataset_h5ad_path</th>\n",
              "      <th>dataset_total_cell_count</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>0</td>\n",
              "      <td>Publication: https://doi.org/10.1016/j.isci.20...</td>\n",
              "      <td>8e880741-bf9a-4c8e-9227-934204631d2a</td>\n",
              "      <td>High Resolution Slide-seqV2 Spatial Transcript...</td>\n",
              "      <td>10.1016/j.isci.2022.104097</td>\n",
              "      <td>Marshall et al. (2022) iScience</td>\n",
              "      <td>4eb29386-de81-452f-b3c0-e00844e8c7fd</td>\n",
              "      <td>f76861bb-becb-4eb7-82fc-782dc96ccc7f</td>\n",
              "      <td>Spatial transcriptomics in mouse: Puck_191112_05</td>\n",
              "      <td>4eb29386-de81-452f-b3c0-e00844e8c7fd.h5ad</td>\n",
              "      <td>10888</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>1</td>\n",
              "      <td>Publication: https://doi.org/10.1016/j.isci.20...</td>\n",
              "      <td>8e880741-bf9a-4c8e-9227-934204631d2a</td>\n",
              "      <td>High Resolution Slide-seqV2 Spatial Transcript...</td>\n",
              "      <td>10.1016/j.isci.2022.104097</td>\n",
              "      <td>Marshall et al. (2022) iScience</td>\n",
              "      <td>78d59e4a-82eb-4a61-a1dc-da974d7ea54b</td>\n",
              "      <td>7d7ec1b6-6e3f-4aaa-9442-4b22f3424396</td>\n",
              "      <td>Spatial transcriptomics in mouse: Puck_191112_08</td>\n",
              "      <td>78d59e4a-82eb-4a61-a1dc-da974d7ea54b.h5ad</td>\n",
              "      <td>10250</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>2</td>\n",
              "      <td>Publication: https://doi.org/10.1016/j.isci.20...</td>\n",
              "      <td>8e880741-bf9a-4c8e-9227-934204631d2a</td>\n",
              "      <td>High Resolution Slide-seqV2 Spatial Transcript...</td>\n",
              "      <td>10.1016/j.isci.2022.104097</td>\n",
              "      <td>Marshall et al. (2022) iScience</td>\n",
              "      <td>add5eb84-5fc9-4f01-982e-a346dd42ee82</td>\n",
              "      <td>de54aed8-4f73-48f6-9229-418a840e2d82</td>\n",
              "      <td>Spatial transcriptomics in mouse: Puck_191109_20</td>\n",
              "      <td>add5eb84-5fc9-4f01-982e-a346dd42ee82.h5ad</td>\n",
              "      <td>12906</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>3</td>\n",
              "      <td>Publication: https://doi.org/10.1016/j.isci.20...</td>\n",
              "      <td>8e880741-bf9a-4c8e-9227-934204631d2a</td>\n",
              "      <td>High Resolution Slide-seqV2 Spatial Transcript...</td>\n",
              "      <td>10.1016/j.isci.2022.104097</td>\n",
              "      <td>Marshall et al. (2022) iScience</td>\n",
              "      <td>b020294c-ab82-4547-b5a7-63d8ffa575ed</td>\n",
              "      <td>abe4fce1-0859-4a56-ad1e-734d79f0e6c8</td>\n",
              "      <td>Spatial transcriptomics in mouse: Puck_191112_13</td>\n",
              "      <td>b020294c-ab82-4547-b5a7-63d8ffa575ed.h5ad</td>\n",
              "      <td>15161</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>4</td>\n",
              "      <td>Publication: https://doi.org/10.1038/s41591-02...</td>\n",
              "      <td>a96133de-e951-4e2d-ace6-59db8b3bfb1d</td>\n",
              "      <td>HTAN/HTAPP Broad - Spatio-molecular dissection...</td>\n",
              "      <td>10.1038/s41591-024-03215-z</td>\n",
              "      <td>Klughammer et al. (2024) Nat Med</td>\n",
              "      <td>d7476ae2-e320-4703-8304-da5c42627e71</td>\n",
              "      <td>863fc5e4-bd4a-4681-9c3d-0ee7ef54e327</td>\n",
              "      <td>HTAPP-330-SMP-1082 scRNA-seq</td>\n",
              "      <td>d7476ae2-e320-4703-8304-da5c42627e71.h5ad</td>\n",
              "      <td>565</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1568</th>\n",
              "      <td>1568</td>\n",
              "      <td>Publication: https://doi.org/10.1038/s41586-02...</td>\n",
              "      <td>45d5d2c3-bc28-4814-aed6-0bb6f0e11c82</td>\n",
              "      <td>A single-cell transcriptional timelapse of mou...</td>\n",
              "      <td>10.1038/s41586-024-07069-w</td>\n",
              "      <td>Qiu et al. (2024) Nature</td>\n",
              "      <td>dcfa2614-7ca7-4d82-814c-350626eccb26</td>\n",
              "      <td>3002a659-a1a9-4406-9976-99e658e1fbb5</td>\n",
              "      <td>Major cell cluster: Mesoderm</td>\n",
              "      <td>dcfa2614-7ca7-4d82-814c-350626eccb26.h5ad</td>\n",
              "      <td>3267338</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1569</th>\n",
              "      <td>1569</td>\n",
              "      <td>Publication: https://doi.org/10.1126/science.a...</td>\n",
              "      <td>e5f58829-1a66-40b5-a624-9046778e74f5</td>\n",
              "      <td>Tabula Sapiens</td>\n",
              "      <td>10.1126/science.abl4896</td>\n",
              "      <td>The Tabula Sapiens Consortium* et al. (2022) S...</td>\n",
              "      <td>53d208b0-2cfd-4366-9866-c3c6114081bc</td>\n",
              "      <td>10df7690-6d10-4029-a47e-0f071bb2df83</td>\n",
              "      <td>Tabula Sapiens - All Cells</td>\n",
              "      <td>53d208b0-2cfd-4366-9866-c3c6114081bc.h5ad</td>\n",
              "      <td>1136218</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1570</th>\n",
              "      <td>1570</td>\n",
              "      <td>Publication: https://doi.org/10.1038/s41586-02...</td>\n",
              "      <td>45d5d2c3-bc28-4814-aed6-0bb6f0e11c82</td>\n",
              "      <td>A single-cell transcriptional timelapse of mou...</td>\n",
              "      <td>10.1038/s41586-024-07069-w</td>\n",
              "      <td>Qiu et al. (2024) Nature</td>\n",
              "      <td>dcfd4feb-18a3-4b30-81d7-1b0c544a8ab3</td>\n",
              "      <td>3817734b-0f82-433b-8c38-55b214200fff</td>\n",
              "      <td>Whole dataset: Raw counts only</td>\n",
              "      <td>dcfd4feb-18a3-4b30-81d7-1b0c544a8ab3.h5ad</td>\n",
              "      <td>11441407</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1571</th>\n",
              "      <td>1571</td>\n",
              "      <td>Publication: https://doi.org/10.1038/s41593-02...</td>\n",
              "      <td>1ca90a2d-2943-483d-b678-b809bf464c30</td>\n",
              "      <td>SEA-AD: Seattle Alzheimer’s Disease Brain Cell...</td>\n",
              "      <td>10.1038/s41593-024-01774-5</td>\n",
              "      <td>Gabitto et al. (2024) Nat Neurosci</td>\n",
              "      <td>c2876b1b-06d8-4d96-a56b-5304f815b99a</td>\n",
              "      <td>c32964d2-3339-441f-8e56-7177234c7876</td>\n",
              "      <td>Whole Taxonomy - MTG: Seattle Alzheimer's Dise...</td>\n",
              "      <td>c2876b1b-06d8-4d96-a56b-5304f815b99a.h5ad</td>\n",
              "      <td>1226855</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1572</th>\n",
              "      <td>1572</td>\n",
              "      <td>Publication: https://doi.org/10.1038/s41593-02...</td>\n",
              "      <td>1ca90a2d-2943-483d-b678-b809bf464c30</td>\n",
              "      <td>SEA-AD: Seattle Alzheimer’s Disease Brain Cell...</td>\n",
              "      <td>10.1038/s41593-024-01774-5</td>\n",
              "      <td>Gabitto et al. (2024) Nat Neurosci</td>\n",
              "      <td>6f7fd0f1-a2ed-4ff1-80d3-33dde731cbc3</td>\n",
              "      <td>d3427e8c-c55d-4d4e-b15b-1a8774cd3a4b</td>\n",
              "      <td>Whole Taxonomy - DLPFC: Seattle Alzheimer's Di...</td>\n",
              "      <td>6f7fd0f1-a2ed-4ff1-80d3-33dde731cbc3.h5ad</td>\n",
              "      <td>1309414</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>1573 rows × 11 columns</p>\n",
              "</div>"
            ],
            "text/plain": [
              "      soma_joinid                                           citation  \\\n",
              "0               0  Publication: https://doi.org/10.1016/j.isci.20...   \n",
              "1               1  Publication: https://doi.org/10.1016/j.isci.20...   \n",
              "2               2  Publication: https://doi.org/10.1016/j.isci.20...   \n",
              "3               3  Publication: https://doi.org/10.1016/j.isci.20...   \n",
              "4               4  Publication: https://doi.org/10.1038/s41591-02...   \n",
              "...           ...                                                ...   \n",
              "1568         1568  Publication: https://doi.org/10.1038/s41586-02...   \n",
              "1569         1569  Publication: https://doi.org/10.1126/science.a...   \n",
              "1570         1570  Publication: https://doi.org/10.1038/s41586-02...   \n",
              "1571         1571  Publication: https://doi.org/10.1038/s41593-02...   \n",
              "1572         1572  Publication: https://doi.org/10.1038/s41593-02...   \n",
              "\n",
              "                             collection_id  \\\n",
              "0     8e880741-bf9a-4c8e-9227-934204631d2a   \n",
              "1     8e880741-bf9a-4c8e-9227-934204631d2a   \n",
              "2     8e880741-bf9a-4c8e-9227-934204631d2a   \n",
              "3     8e880741-bf9a-4c8e-9227-934204631d2a   \n",
              "4     a96133de-e951-4e2d-ace6-59db8b3bfb1d   \n",
              "...                                    ...   \n",
              "1568  45d5d2c3-bc28-4814-aed6-0bb6f0e11c82   \n",
              "1569  e5f58829-1a66-40b5-a624-9046778e74f5   \n",
              "1570  45d5d2c3-bc28-4814-aed6-0bb6f0e11c82   \n",
              "1571  1ca90a2d-2943-483d-b678-b809bf464c30   \n",
              "1572  1ca90a2d-2943-483d-b678-b809bf464c30   \n",
              "\n",
              "                                        collection_name  \\\n",
              "0     High Resolution Slide-seqV2 Spatial Transcript...   \n",
              "1     High Resolution Slide-seqV2 Spatial Transcript...   \n",
              "2     High Resolution Slide-seqV2 Spatial Transcript...   \n",
              "3     High Resolution Slide-seqV2 Spatial Transcript...   \n",
              "4     HTAN/HTAPP Broad - Spatio-molecular dissection...   \n",
              "...                                                 ...   \n",
              "1568  A single-cell transcriptional timelapse of mou...   \n",
              "1569                                     Tabula Sapiens   \n",
              "1570  A single-cell transcriptional timelapse of mou...   \n",
              "1571  SEA-AD: Seattle Alzheimer’s Disease Brain Cell...   \n",
              "1572  SEA-AD: Seattle Alzheimer’s Disease Brain Cell...   \n",
              "\n",
              "                  collection_doi  \\\n",
              "0     10.1016/j.isci.2022.104097   \n",
              "1     10.1016/j.isci.2022.104097   \n",
              "2     10.1016/j.isci.2022.104097   \n",
              "3     10.1016/j.isci.2022.104097   \n",
              "4     10.1038/s41591-024-03215-z   \n",
              "...                          ...   \n",
              "1568  10.1038/s41586-024-07069-w   \n",
              "1569     10.1126/science.abl4896   \n",
              "1570  10.1038/s41586-024-07069-w   \n",
              "1571  10.1038/s41593-024-01774-5   \n",
              "1572  10.1038/s41593-024-01774-5   \n",
              "\n",
              "                                   collection_doi_label  \\\n",
              "0                       Marshall et al. (2022) iScience   \n",
              "1                       Marshall et al. (2022) iScience   \n",
              "2                       Marshall et al. (2022) iScience   \n",
              "3                       Marshall et al. (2022) iScience   \n",
              "4                      Klughammer et al. (2024) Nat Med   \n",
              "...                                                 ...   \n",
              "1568                           Qiu et al. (2024) Nature   \n",
              "1569  The Tabula Sapiens Consortium* et al. (2022) S...   \n",
              "1570                           Qiu et al. (2024) Nature   \n",
              "1571                 Gabitto et al. (2024) Nat Neurosci   \n",
              "1572                 Gabitto et al. (2024) Nat Neurosci   \n",
              "\n",
              "                                dataset_id  \\\n",
              "0     4eb29386-de81-452f-b3c0-e00844e8c7fd   \n",
              "1     78d59e4a-82eb-4a61-a1dc-da974d7ea54b   \n",
              "2     add5eb84-5fc9-4f01-982e-a346dd42ee82   \n",
              "3     b020294c-ab82-4547-b5a7-63d8ffa575ed   \n",
              "4     d7476ae2-e320-4703-8304-da5c42627e71   \n",
              "...                                    ...   \n",
              "1568  dcfa2614-7ca7-4d82-814c-350626eccb26   \n",
              "1569  53d208b0-2cfd-4366-9866-c3c6114081bc   \n",
              "1570  dcfd4feb-18a3-4b30-81d7-1b0c544a8ab3   \n",
              "1571  c2876b1b-06d8-4d96-a56b-5304f815b99a   \n",
              "1572  6f7fd0f1-a2ed-4ff1-80d3-33dde731cbc3   \n",
              "\n",
              "                        dataset_version_id  \\\n",
              "0     f76861bb-becb-4eb7-82fc-782dc96ccc7f   \n",
              "1     7d7ec1b6-6e3f-4aaa-9442-4b22f3424396   \n",
              "2     de54aed8-4f73-48f6-9229-418a840e2d82   \n",
              "3     abe4fce1-0859-4a56-ad1e-734d79f0e6c8   \n",
              "4     863fc5e4-bd4a-4681-9c3d-0ee7ef54e327   \n",
              "...                                    ...   \n",
              "1568  3002a659-a1a9-4406-9976-99e658e1fbb5   \n",
              "1569  10df7690-6d10-4029-a47e-0f071bb2df83   \n",
              "1570  3817734b-0f82-433b-8c38-55b214200fff   \n",
              "1571  c32964d2-3339-441f-8e56-7177234c7876   \n",
              "1572  d3427e8c-c55d-4d4e-b15b-1a8774cd3a4b   \n",
              "\n",
              "                                          dataset_title  \\\n",
              "0      Spatial transcriptomics in mouse: Puck_191112_05   \n",
              "1      Spatial transcriptomics in mouse: Puck_191112_08   \n",
              "2      Spatial transcriptomics in mouse: Puck_191109_20   \n",
              "3      Spatial transcriptomics in mouse: Puck_191112_13   \n",
              "4                          HTAPP-330-SMP-1082 scRNA-seq   \n",
              "...                                                 ...   \n",
              "1568                       Major cell cluster: Mesoderm   \n",
              "1569                         Tabula Sapiens - All Cells   \n",
              "1570                     Whole dataset: Raw counts only   \n",
              "1571  Whole Taxonomy - MTG: Seattle Alzheimer's Dise...   \n",
              "1572  Whole Taxonomy - DLPFC: Seattle Alzheimer's Di...   \n",
              "\n",
              "                              dataset_h5ad_path  dataset_total_cell_count  \n",
              "0     4eb29386-de81-452f-b3c0-e00844e8c7fd.h5ad                     10888  \n",
              "1     78d59e4a-82eb-4a61-a1dc-da974d7ea54b.h5ad                     10250  \n",
              "2     add5eb84-5fc9-4f01-982e-a346dd42ee82.h5ad                     12906  \n",
              "3     b020294c-ab82-4547-b5a7-63d8ffa575ed.h5ad                     15161  \n",
              "4     d7476ae2-e320-4703-8304-da5c42627e71.h5ad                       565  \n",
              "...                                         ...                       ...  \n",
              "1568  dcfa2614-7ca7-4d82-814c-350626eccb26.h5ad                   3267338  \n",
              "1569  53d208b0-2cfd-4366-9866-c3c6114081bc.h5ad                   1136218  \n",
              "1570  dcfd4feb-18a3-4b30-81d7-1b0c544a8ab3.h5ad                  11441407  \n",
              "1571  c2876b1b-06d8-4d96-a56b-5304f815b99a.h5ad                   1226855  \n",
              "1572  6f7fd0f1-a2ed-4ff1-80d3-33dde731cbc3.h5ad                   1309414  \n",
              "\n",
              "[1573 rows x 11 columns]"
            ]
          },
          "execution_count": 6,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "census_old[\"census_info\"][\"datasets\"].read().concat().to_pandas()"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 7,
      "id": "82ef0129-72ac-4173-967a-b8e4346abec5",
      "metadata": {
        "id": "82ef0129-72ac-4173-967a-b8e4346abec5",
        "outputId": "cc696dcd-7742-4b38-c1e0-c13caab94ec6"
      },
      "outputs": [
        {
          "data": {
            "application/vnd.microsoft.datawrangler.viewer.v0+json": {
              "columns": [
                {
                  "name": "index",
                  "rawType": "int64",
                  "type": "integer"
                },
                {
                  "name": "soma_joinid",
                  "rawType": "int64",
                  "type": "integer"
                },
                {
                  "name": "citation",
                  "rawType": "object",
                  "type": "string"
                },
                {
                  "name": "collection_id",
                  "rawType": "object",
                  "type": "string"
                },
                {
                  "name": "collection_name",
                  "rawType": "object",
                  "type": "string"
                },
                {
                  "name": "collection_doi",
                  "rawType": "object",
                  "type": "string"
                },
                {
                  "name": "collection_doi_label",
                  "rawType": "object",
                  "type": "string"
                },
                {
                  "name": "dataset_id",
                  "rawType": "object",
                  "type": "string"
                },
                {
                  "name": "dataset_version_id",
                  "rawType": "object",
                  "type": "string"
                },
                {
                  "name": "dataset_title",
                  "rawType": "object",
                  "type": "string"
                },
                {
                  "name": "dataset_h5ad_path",
                  "rawType": "object",
                  "type": "string"
                },
                {
                  "name": "dataset_total_cell_count",
                  "rawType": "int64",
                  "type": "integer"
                }
              ],
              "ref": "f5544430-e16c-4206-aabd-1167bedcfcd8",
              "rows": [
                [
                  "0",
                  "0",
                  "Publication: https://doi.org/10.1016/j.isci.2022.104097 Dataset Version: https://datasets.cellxgene.cziscience.com/66699060-0389-4fbd-b3a5-196b3b4e32d6.h5ad curated and distributed by CZ CELLxGENE Discover in Collection: https://cellxgene.cziscience.com/collections/8e880741-bf9a-4c8e-9227-934204631d2a",
                  "8e880741-bf9a-4c8e-9227-934204631d2a",
                  "High Resolution Slide-seqV2 Spatial Transcriptomics Enables Discovery of Disease-Specific Cell Neighborhoods and Pathways",
                  "10.1016/j.isci.2022.104097",
                  "Marshall et al. (2022) iScience",
                  "4eb29386-de81-452f-b3c0-e00844e8c7fd",
                  "66699060-0389-4fbd-b3a5-196b3b4e32d6",
                  "Spatial transcriptomics in mouse: Puck_191112_05",
                  "4eb29386-de81-452f-b3c0-e00844e8c7fd.h5ad",
                  "10888"
                ],
                [
                  "1",
                  "1",
                  "Publication: https://doi.org/10.1016/j.isci.2022.104097 Dataset Version: https://datasets.cellxgene.cziscience.com/f64950a2-a3c8-490a-8431-7121eeb4f5f4.h5ad curated and distributed by CZ CELLxGENE Discover in Collection: https://cellxgene.cziscience.com/collections/8e880741-bf9a-4c8e-9227-934204631d2a",
                  "8e880741-bf9a-4c8e-9227-934204631d2a",
                  "High Resolution Slide-seqV2 Spatial Transcriptomics Enables Discovery of Disease-Specific Cell Neighborhoods and Pathways",
                  "10.1016/j.isci.2022.104097",
                  "Marshall et al. (2022) iScience",
                  "78d59e4a-82eb-4a61-a1dc-da974d7ea54b",
                  "f64950a2-a3c8-490a-8431-7121eeb4f5f4",
                  "Spatial transcriptomics in mouse: Puck_191112_08",
                  "78d59e4a-82eb-4a61-a1dc-da974d7ea54b.h5ad",
                  "10250"
                ],
                [
                  "2",
                  "2",
                  "Publication: https://doi.org/10.1016/j.isci.2022.104097 Dataset Version: https://datasets.cellxgene.cziscience.com/781a724a-b0f5-46c4-9a13-e6293ef4364f.h5ad curated and distributed by CZ CELLxGENE Discover in Collection: https://cellxgene.cziscience.com/collections/8e880741-bf9a-4c8e-9227-934204631d2a",
                  "8e880741-bf9a-4c8e-9227-934204631d2a",
                  "High Resolution Slide-seqV2 Spatial Transcriptomics Enables Discovery of Disease-Specific Cell Neighborhoods and Pathways",
                  "10.1016/j.isci.2022.104097",
                  "Marshall et al. (2022) iScience",
                  "add5eb84-5fc9-4f01-982e-a346dd42ee82",
                  "781a724a-b0f5-46c4-9a13-e6293ef4364f",
                  "Spatial transcriptomics in mouse: Puck_191109_20",
                  "add5eb84-5fc9-4f01-982e-a346dd42ee82.h5ad",
                  "12906"
                ],
                [
                  "3",
                  "3",
                  "Publication: https://doi.org/10.1016/j.isci.2022.104097 Dataset Version: https://datasets.cellxgene.cziscience.com/96a79598-297b-4ade-a6c1-a431ab243548.h5ad curated and distributed by CZ CELLxGENE Discover in Collection: https://cellxgene.cziscience.com/collections/8e880741-bf9a-4c8e-9227-934204631d2a",
                  "8e880741-bf9a-4c8e-9227-934204631d2a",
                  "High Resolution Slide-seqV2 Spatial Transcriptomics Enables Discovery of Disease-Specific Cell Neighborhoods and Pathways",
                  "10.1016/j.isci.2022.104097",
                  "Marshall et al. (2022) iScience",
                  "b020294c-ab82-4547-b5a7-63d8ffa575ed",
                  "96a79598-297b-4ade-a6c1-a431ab243548",
                  "Spatial transcriptomics in mouse: Puck_191112_13",
                  "b020294c-ab82-4547-b5a7-63d8ffa575ed.h5ad",
                  "15161"
                ],
                [
                  "4",
                  "4",
                  "Publication: https://doi.org/10.1038/s41591-024-03215-z Dataset Version: https://datasets.cellxgene.cziscience.com/ac9fe945-6784-48fa-a2d5-a8646196d37e.h5ad curated and distributed by CZ CELLxGENE Discover in Collection: https://cellxgene.cziscience.com/collections/a96133de-e951-4e2d-ace6-59db8b3bfb1d",
                  "a96133de-e951-4e2d-ace6-59db8b3bfb1d",
                  "HTAN/HTAPP Broad - Spatio-molecular dissection of the breast cancer metastatic microenvironment",
                  "10.1038/s41591-024-03215-z",
                  "Klughammer et al. (2024) Nat Med",
                  "d7476ae2-e320-4703-8304-da5c42627e71",
                  "ac9fe945-6784-48fa-a2d5-a8646196d37e",
                  "HTAPP-330-SMP-1082 scRNA-seq",
                  "d7476ae2-e320-4703-8304-da5c42627e71.h5ad",
                  "565"
                ],
                [
                  "5",
                  "5",
                  "Publication: https://doi.org/10.1002/hep4.1854 Dataset Version: https://datasets.cellxgene.cziscience.com/85954cff-a901-4de2-bd7f-a23b0077812b.h5ad curated and distributed by CZ CELLxGENE Discover in Collection: https://cellxgene.cziscience.com/collections/44531dd9-1388-4416-a117-af0a99de2294",
                  "44531dd9-1388-4416-a117-af0a99de2294",
                  "Single-Cell, Single-Nucleus, and Spatial RNA Sequencing of the Human Liver Identifies Cholangiocyte and Mesenchymal Heterogeneity",
                  "10.1002/hep4.1854",
                  "Andrews et al. (2022) Hepatol Commun",
                  "0895c838-e550-48a3-a777-dbcd35d30272",
                  "85954cff-a901-4de2-bd7f-a23b0077812b",
                  "Healthy human liver: B cells",
                  "0895c838-e550-48a3-a777-dbcd35d30272.h5ad",
                  "146"
                ],
                [
                  "6",
                  "6",
                  "Publication: https://doi.org/10.1016/j.isci.2022.104097 Dataset Version: https://datasets.cellxgene.cziscience.com/3b0f15f6-abcc-4889-8daa-2f75c106099a.h5ad curated and distributed by CZ CELLxGENE Discover in Collection: https://cellxgene.cziscience.com/collections/8e880741-bf9a-4c8e-9227-934204631d2a",
                  "8e880741-bf9a-4c8e-9227-934204631d2a",
                  "High Resolution Slide-seqV2 Spatial Transcriptomics Enables Discovery of Disease-Specific Cell Neighborhoods and Pathways",
                  "10.1016/j.isci.2022.104097",
                  "Marshall et al. (2022) iScience",
                  "530c9bff-c552-45b9-be77-0de605e6858b",
                  "3b0f15f6-abcc-4889-8daa-2f75c106099a",
                  "Spatial transcriptomics in mouse: Puck_191109_14",
                  "530c9bff-c552-45b9-be77-0de605e6858b.h5ad",
                  "12351"
                ],
                [
                  "7",
                  "7",
                  "Publication: https://doi.org/10.1038/s41586-024-07944-6 Dataset Version: https://datasets.cellxgene.cziscience.com/6c49bd07-0978-4b24-9358-f9a11746276a.h5ad curated and distributed by CZ CELLxGENE Discover in Collection: https://cellxgene.cziscience.com/collections/fc19ae6c-d7c1-4dce-b703-62c5d52061b4",
                  "fc19ae6c-d7c1-4dce-b703-62c5d52061b4",
                  "A spatial human thymus cell atlas mapped to a continuous tissue axis",
                  "10.1038/s41586-024-07944-6",
                  "Yayon et al. (2024) Nature",
                  "fbd69faa-b0c5-45ba-89c9-da938a7f5a14",
                  "6c49bd07-0978-4b24-9358-f9a11746276a",
                  "thymus scRNA-seq atlas - myeloid p2 subset",
                  "fbd69faa-b0c5-45ba-89c9-da938a7f5a14.h5ad",
                  "843"
                ],
                [
                  "8",
                  "8",
                  "Publication: https://doi.org/10.1016/j.isci.2022.104097 Dataset Version: https://datasets.cellxgene.cziscience.com/d7149896-5be5-4b37-ad54-249a7825fe3e.h5ad curated and distributed by CZ CELLxGENE Discover in Collection: https://cellxgene.cziscience.com/collections/8e880741-bf9a-4c8e-9227-934204631d2a",
                  "8e880741-bf9a-4c8e-9227-934204631d2a",
                  "High Resolution Slide-seqV2 Spatial Transcriptomics Enables Discovery of Disease-Specific Cell Neighborhoods and Pathways",
                  "10.1016/j.isci.2022.104097",
                  "Marshall et al. (2022) iScience",
                  "fb83600a-7823-426e-b06b-1fcd666a3cd8",
                  "d7149896-5be5-4b37-ad54-249a7825fe3e",
                  "Spatial transcriptomics in mouse: Puck_191109_18",
                  "fb83600a-7823-426e-b06b-1fcd666a3cd8.h5ad",
                  "19156"
                ],
                [
                  "9",
                  "9",
                  "Publication: https://doi.org/10.1126/sciimmunol.abe6291 Dataset Version: https://datasets.cellxgene.cziscience.com/09d1124d-68e5-4ca7-9452-e3012b2858d6.h5ad curated and distributed by CZ CELLxGENE Discover in Collection: https://cellxgene.cziscience.com/collections/3a2af25b-2338-4266-aad3-aa8d07473f50",
                  "3a2af25b-2338-4266-aad3-aa8d07473f50",
                  "Single-cell analysis of human B cell maturation predicts how antibody class switching shapes selection dynamics",
                  "10.1126/sciimmunol.abe6291",
                  "King et al. (2021) Sci. Immunol.",
                  "00ff600e-6e2e-4d76-846f-0eec4f0ae417",
                  "09d1124d-68e5-4ca7-9452-e3012b2858d6",
                  "Human tonsil nonlymphoid cells scRNA",
                  "00ff600e-6e2e-4d76-846f-0eec4f0ae417.h5ad",
                  "363"
                ],
                [
                  "10",
                  "10",
                  "Publication: https://doi.org/10.1016/j.isci.2022.104097 Dataset Version: https://datasets.cellxgene.cziscience.com/9396b087-fb6c-4180-aa32-27fad712d305.h5ad curated and distributed by CZ CELLxGENE Discover in Collection: https://cellxgene.cziscience.com/collections/8e880741-bf9a-4c8e-9227-934204631d2a",
                  "8e880741-bf9a-4c8e-9227-934204631d2a",
                  "High Resolution Slide-seqV2 Spatial Transcriptomics Enables Discovery of Disease-Specific Cell Neighborhoods and Pathways",
                  "10.1016/j.isci.2022.104097",
                  "Marshall et al. (2022) iScience",
                  "70fa6bd9-cc35-4779-9a05-fa8918c0ff8b",
                  "9396b087-fb6c-4180-aa32-27fad712d305",
                  "Spatial transcriptomics in mouse: Puck_191109_19",
                  "70fa6bd9-cc35-4779-9a05-fa8918c0ff8b.h5ad",
                  "22659"
                ],
                [
                  "11",
                  "11",
                  "Publication: https://doi.org/10.1016/j.isci.2022.104097 Dataset Version: https://datasets.cellxgene.cziscience.com/4ca924bb-5012-4bb9-b74e-fef159a112fc.h5ad curated and distributed by CZ CELLxGENE Discover in Collection: https://cellxgene.cziscience.com/collections/8e880741-bf9a-4c8e-9227-934204631d2a",
                  "8e880741-bf9a-4c8e-9227-934204631d2a",
                  "High Resolution Slide-seqV2 Spatial Transcriptomics Enables Discovery of Disease-Specific Cell Neighborhoods and Pathways",
                  "10.1016/j.isci.2022.104097",
                  "Marshall et al. (2022) iScience",
                  "1c224f30-a9c4-48c6-951d-8d86c0a2886e",
                  "4ca924bb-5012-4bb9-b74e-fef159a112fc",
                  "Spatial transcriptomics in mouse: Puck_191223_13",
                  "1c224f30-a9c4-48c6-951d-8d86c0a2886e.h5ad",
                  "16787"
                ],
                [
                  "12",
                  "12",
                  "Publication: https://doi.org/10.1038/s41588-023-01435-6 Dataset Version: https://datasets.cellxgene.cziscience.com/89c9fcbd-4935-4a13-818e-b0d436ae3864.h5ad curated and distributed by CZ CELLxGENE Discover in Collection: https://cellxgene.cziscience.com/collections/d74b6979-efba-47cd-990a-9d80ccf29055",
                  "d74b6979-efba-47cd-990a-9d80ccf29055",
                  "Spatial transcriptomic maps of whole mouse embryos reveal principles of neural tube patterning",
                  "10.1038/s41588-023-01435-6",
                  "Sampath Kumar et al. (2023) Nat Genet",
                  "c233842b-9430-41b8-8ec6-4342428e43f3",
                  "89c9fcbd-4935-4a13-818e-b0d436ae3864",
                  "mouse_trunkTbx6WT_KO",
                  "c233842b-9430-41b8-8ec6-4342428e43f3.h5ad",
                  "499"
                ],
                [
                  "13",
                  "13",
                  "Publication: https://doi.org/10.1016/j.isci.2022.104097 Dataset Version: https://datasets.cellxgene.cziscience.com/094c4311-ed15-47f1-936d-13e405efe52a.h5ad curated and distributed by CZ CELLxGENE Discover in Collection: https://cellxgene.cziscience.com/collections/8e880741-bf9a-4c8e-9227-934204631d2a",
                  "8e880741-bf9a-4c8e-9227-934204631d2a",
                  "High Resolution Slide-seqV2 Spatial Transcriptomics Enables Discovery of Disease-Specific Cell Neighborhoods and Pathways",
                  "10.1016/j.isci.2022.104097",
                  "Marshall et al. (2022) iScience",
                  "3483835a-fdd2-4589-8de3-530e2a13ad14",
                  "094c4311-ed15-47f1-936d-13e405efe52a",
                  "Spatial transcriptomics in mouse: Puck_191204_18",
                  "3483835a-fdd2-4589-8de3-530e2a13ad14.h5ad",
                  "27814"
                ],
                [
                  "14",
                  "14",
                  "Publication: https://doi.org/10.1016/j.isci.2022.104097 Dataset Version: https://datasets.cellxgene.cziscience.com/53f89023-3ea8-484a-982b-0a2f1ae304d4.h5ad curated and distributed by CZ CELLxGENE Discover in Collection: https://cellxgene.cziscience.com/collections/8e880741-bf9a-4c8e-9227-934204631d2a",
                  "8e880741-bf9a-4c8e-9227-934204631d2a",
                  "High Resolution Slide-seqV2 Spatial Transcriptomics Enables Discovery of Disease-Specific Cell Neighborhoods and Pathways",
                  "10.1016/j.isci.2022.104097",
                  "Marshall et al. (2022) iScience",
                  "ce19be17-f066-452f-922e-b1f232ce8af5",
                  "53f89023-3ea8-484a-982b-0a2f1ae304d4",
                  "Spatial transcriptomics in mouse: Puck_191109_10",
                  "ce19be17-f066-452f-922e-b1f232ce8af5.h5ad",
                  "15592"
                ],
                [
                  "15",
                  "15",
                  "Publication: https://doi.org/10.1016/j.isci.2022.104097 Dataset Version: https://datasets.cellxgene.cziscience.com/1db2b7c9-5566-410d-a061-4812034610f0.h5ad curated and distributed by CZ CELLxGENE Discover in Collection: https://cellxgene.cziscience.com/collections/8e880741-bf9a-4c8e-9227-934204631d2a",
                  "8e880741-bf9a-4c8e-9227-934204631d2a",
                  "High Resolution Slide-seqV2 Spatial Transcriptomics Enables Discovery of Disease-Specific Cell Neighborhoods and Pathways",
                  "10.1016/j.isci.2022.104097",
                  "Marshall et al. (2022) iScience",
                  "c5ac9bb9-9597-4c0e-a985-1c3947288785",
                  "1db2b7c9-5566-410d-a061-4812034610f0",
                  "Spatial transcriptomics in mouse: Puck_191112_07",
                  "c5ac9bb9-9597-4c0e-a985-1c3947288785.h5ad",
                  "21181"
                ],
                [
                  "16",
                  "16",
                  "Publication: https://doi.org/10.1016/j.isci.2022.104097 Dataset Version: https://datasets.cellxgene.cziscience.com/6fa10d74-637c-4c2b-9c5b-20b02f887e09.h5ad curated and distributed by CZ CELLxGENE Discover in Collection: https://cellxgene.cziscience.com/collections/8e880741-bf9a-4c8e-9227-934204631d2a",
                  "8e880741-bf9a-4c8e-9227-934204631d2a",
                  "High Resolution Slide-seqV2 Spatial Transcriptomics Enables Discovery of Disease-Specific Cell Neighborhoods and Pathways",
                  "10.1016/j.isci.2022.104097",
                  "Marshall et al. (2022) iScience",
                  "72ed5f5b-e685-474d-b436-3bcc8c597e8e",
                  "6fa10d74-637c-4c2b-9c5b-20b02f887e09",
                  "Spatial transcriptomics in mouse: Puck_191112_04",
                  "72ed5f5b-e685-474d-b436-3bcc8c597e8e.h5ad",
                  "13822"
                ],
                [
                  "17",
                  "17",
                  "Publication: https://doi.org/10.1016/j.labinv.2023.100258 Dataset Version: https://datasets.cellxgene.cziscience.com/f4040a55-2315-4b01-8e8e-e841e4782918.h5ad curated and distributed by CZ CELLxGENE Discover in Collection: https://cellxgene.cziscience.com/collections/bfd55632-15e2-4dde-a1b9-02ade511fc4d",
                  "bfd55632-15e2-4dde-a1b9-02ade511fc4d",
                  "Spatial Transcriptomics Reveal Pitfalls and Opportunities for the Detection of Rare High-Plasticity Breast Cancer Subtypes",
                  "10.1016/j.labinv.2023.100258",
                  "Coutant et al. (2023) Laboratory Investigation",
                  "60d5175f-f477-48b2-8396-ad76be2c3310",
                  "f4040a55-2315-4b01-8e8e-e841e4782918",
                  "GSM6592057_M10",
                  "60d5175f-f477-48b2-8396-ad76be2c3310.h5ad",
                  "4992"
                ],
                [
                  "18",
                  "18",
                  "Publication: https://doi.org/10.1038/s41593-020-00764-7 Dataset Version: https://datasets.cellxgene.cziscience.com/dda14efc-dea6-4d1c-ac42-c87797de8c30.h5ad curated and distributed by CZ CELLxGENE Discover in Collection: https://cellxgene.cziscience.com/collections/180bff9c-c8a5-4539-b13b-ddbc00d643e6",
                  "180bff9c-c8a5-4539-b13b-ddbc00d643e6",
                  "Molecular characterization of selectively vulnerable neurons in Alzheimer's Disease",
                  "10.1038/s41593-020-00764-7",
                  "Leng et al. (2021) Nat Neurosci",
                  "bdacc907-7c26-419f-8808-969eab3ca2e8",
                  "dda14efc-dea6-4d1c-ac42-c87797de8c30",
                  "Molecular characterization of selectively vulnerable neurons in Alzheimer’s Disease: SFG microglia",
                  "bdacc907-7c26-419f-8808-969eab3ca2e8.h5ad",
                  "3799"
                ],
                [
                  "19",
                  "19",
                  "Publication: https://doi.org/10.1038/s41588-023-01435-6 Dataset Version: https://datasets.cellxgene.cziscience.com/5e9d9d03-4972-4dfb-b02d-248a3b16e064.h5ad curated and distributed by CZ CELLxGENE Discover in Collection: https://cellxgene.cziscience.com/collections/d74b6979-efba-47cd-990a-9d80ccf29055",
                  "d74b6979-efba-47cd-990a-9d80ccf29055",
                  "Spatial transcriptomic maps of whole mouse embryos reveal principles of neural tube patterning",
                  "10.1038/s41588-023-01435-6",
                  "Sampath Kumar et al. (2023) Nat Genet",
                  "6c0f2e18-3793-4d90-88b5-12d9d6f0f4e2",
                  "5e9d9d03-4972-4dfb-b02d-248a3b16e064",
                  "E9.0: Puck_220718_02",
                  "6c0f2e18-3793-4d90-88b5-12d9d6f0f4e2.h5ad",
                  "703"
                ],
                [
                  "20",
                  "20",
                  "Publication: https://doi.org/10.1016/j.isci.2022.104097 Dataset Version: https://datasets.cellxgene.cziscience.com/cecd8b2e-d19e-4e00-b624-781ba8edef60.h5ad curated and distributed by CZ CELLxGENE Discover in Collection: https://cellxgene.cziscience.com/collections/8e880741-bf9a-4c8e-9227-934204631d2a",
                  "8e880741-bf9a-4c8e-9227-934204631d2a",
                  "High Resolution Slide-seqV2 Spatial Transcriptomics Enables Discovery of Disease-Specific Cell Neighborhoods and Pathways",
                  "10.1016/j.isci.2022.104097",
                  "Marshall et al. (2022) iScience",
                  "b24705ee-87c0-45f2-adfb-2819fdc16225",
                  "cecd8b2e-d19e-4e00-b624-781ba8edef60",
                  "Spatial transcriptomics in mouse: Puck_200104_04",
                  "b24705ee-87c0-45f2-adfb-2819fdc16225.h5ad",
                  "25817"
                ],
                [
                  "21",
                  "21",
                  "Publication: https://doi.org/10.1038/s41467-022-29450-x Dataset Version: https://datasets.cellxgene.cziscience.com/963e004b-77ac-451d-a73c-bd9e1e14c232.h5ad curated and distributed by CZ CELLxGENE Discover in Collection: https://cellxgene.cziscience.com/collections/bf325905-5e8e-42e3-933d-9a9053e9af80",
                  "bf325905-5e8e-42e3-933d-9a9053e9af80",
                  "Single-cell Atlas of common variable immunodeficiency shows germinal center-associated epigenetic dysregulation in B-cell responses",
                  "10.1038/s41467-022-29450-x",
                  "Rodríguez-Ubreva et al. (2022) Nat Commun",
                  "a5d95a42-0137-496f-8a60-101e17f263c8",
                  "963e004b-77ac-451d-a73c-bd9e1e14c232",
                  "Steady-state B cells - scRNA-seq",
                  "a5d95a42-0137-496f-8a60-101e17f263c8.h5ad",
                  "1324"
                ],
                [
                  "22",
                  "22",
                  "Publication: https://doi.org/10.1016/j.isci.2022.104097 Dataset Version: https://datasets.cellxgene.cziscience.com/2c2f209b-09e1-4eb7-9aed-93ac1d03d813.h5ad curated and distributed by CZ CELLxGENE Discover in Collection: https://cellxgene.cziscience.com/collections/8e880741-bf9a-4c8e-9227-934204631d2a",
                  "8e880741-bf9a-4c8e-9227-934204631d2a",
                  "High Resolution Slide-seqV2 Spatial Transcriptomics Enables Discovery of Disease-Specific Cell Neighborhoods and Pathways",
                  "10.1016/j.isci.2022.104097",
                  "Marshall et al. (2022) iScience",
                  "90ef2377-e4d0-41c7-924f-6f9ee7a9237d",
                  "2c2f209b-09e1-4eb7-9aed-93ac1d03d813",
                  "Spatial transcriptomics in the healthy human kidney: Puck_200113_07",
                  "90ef2377-e4d0-41c7-924f-6f9ee7a9237d.h5ad",
                  "11456"
                ],
                [
                  "23",
                  "23",
                  "Publication: https://doi.org/10.1101/2024.09.10.612293 Dataset Version: https://datasets.cellxgene.cziscience.com/fd7cb98d-0096-448c-b1fa-2806e4abdb66.h5ad curated and distributed by CZ CELLxGENE Discover in Collection: https://cellxgene.cziscience.com/collections/8f17ac63-aaba-44b5-9b78-60f121da4c2f",
                  "8f17ac63-aaba-44b5-9b78-60f121da4c2f",
                  "A Coding and Non-Coding Atlas of the Human Arterial Cell",
                  "10.1101/2024.09.10.612293",
                  "Zhao et al. (2024) bioRxiv",
                  "b2005457-dede-4434-a9c3-dbe41fcd542e",
                  "fd7cb98d-0096-448c-b1fa-2806e4abdb66",
                  "slide-seqV2 analysis of aorta",
                  "b2005457-dede-4434-a9c3-dbe41fcd542e.h5ad",
                  "6142"
                ],
                [
                  "24",
                  "24",
                  "Publication: https://doi.org/10.1016/j.isci.2022.104097 Dataset Version: https://datasets.cellxgene.cziscience.com/596d0795-d454-4d23-83d5-fecfa6dc165e.h5ad curated and distributed by CZ CELLxGENE Discover in Collection: https://cellxgene.cziscience.com/collections/8e880741-bf9a-4c8e-9227-934204631d2a",
                  "8e880741-bf9a-4c8e-9227-934204631d2a",
                  "High Resolution Slide-seqV2 Spatial Transcriptomics Enables Discovery of Disease-Specific Cell Neighborhoods and Pathways",
                  "10.1016/j.isci.2022.104097",
                  "Marshall et al. (2022) iScience",
                  "2fb24a91-55b9-4cc4-9a11-d46c373d53c1",
                  "596d0795-d454-4d23-83d5-fecfa6dc165e",
                  "Spatial transcriptomics in mouse: Puck_191109_08",
                  "2fb24a91-55b9-4cc4-9a11-d46c373d53c1.h5ad",
                  "30182"
                ],
                [
                  "25",
                  "25",
                  "Publication: https://doi.org/10.1016/j.isci.2022.104097 Dataset Version: https://datasets.cellxgene.cziscience.com/9344d507-7e68-4d44-b319-0ea1da84da06.h5ad curated and distributed by CZ CELLxGENE Discover in Collection: https://cellxgene.cziscience.com/collections/8e880741-bf9a-4c8e-9227-934204631d2a",
                  "8e880741-bf9a-4c8e-9227-934204631d2a",
                  "High Resolution Slide-seqV2 Spatial Transcriptomics Enables Discovery of Disease-Specific Cell Neighborhoods and Pathways",
                  "10.1016/j.isci.2022.104097",
                  "Marshall et al. (2022) iScience",
                  "5e73d977-8e54-4406-a5f0-711bf0d3c228",
                  "9344d507-7e68-4d44-b319-0ea1da84da06",
                  "Spatial transcriptomics in mouse: Puck_191204_09",
                  "5e73d977-8e54-4406-a5f0-711bf0d3c228.h5ad",
                  "17138"
                ],
                [
                  "26",
                  "26",
                  "Publication: https://doi.org/10.1016/j.immuni.2022.09.002 Dataset Version: https://datasets.cellxgene.cziscience.com/a6baa044-e38d-48cc-bc8c-651747a46474.h5ad curated and distributed by CZ CELLxGENE Discover in Collection: https://cellxgene.cziscience.com/collections/02b01703-bf1b-48de-b99a-23bef8cccc81",
                  "02b01703-bf1b-48de-b99a-23bef8cccc81",
                  "Spatially mapping T cell receptors and transcriptomes",
                  "10.1016/j.immuni.2022.09.002",
                  "Liu et al. (2022) Immunity",
                  "c62ac8e5-72d1-48d1-8bb5-9e66fd2b1e19",
                  "a6baa044-e38d-48cc-bc8c-651747a46474",
                  "Normal Spleen OT1 Puck_200727_02",
                  "c62ac8e5-72d1-48d1-8bb5-9e66fd2b1e19.h5ad",
                  "11510"
                ],
                [
                  "27",
                  "27",
                  "Publication: https://doi.org/10.1038/s41590-021-01059-0 Dataset Version: https://datasets.cellxgene.cziscience.com/e4ace604-4f9a-4dd7-99b1-98a18c78c531.h5ad curated and distributed by CZ CELLxGENE Discover in Collection: https://cellxgene.cziscience.com/collections/93eebe82-d8c3-41bc-a906-63b5b5f24a9d",
                  "93eebe82-d8c3-41bc-a906-63b5b5f24a9d",
                  "Single-cell proteo-genomic reference maps of the hematopoietic system enable the purification and massive profiling of precisely defined cell states",
                  "10.1038/s41590-021-01059-0",
                  "Triana et al. (2021) Nat Immunol",
                  "d3566d6a-a455-4a15-980f-45eb29114cab",
                  "e4ace604-4f9a-4dd7-99b1-98a18c78c531",
                  "blood and bone marrow from a healthy young donor",
                  "d3566d6a-a455-4a15-980f-45eb29114cab.h5ad",
                  "15502"
                ],
                [
                  "28",
                  "28",
                  "Publication: https://doi.org/10.1016/j.isci.2022.104097 Dataset Version: https://datasets.cellxgene.cziscience.com/01e21c00-81c0-40ce-a6ee-553dad5fd863.h5ad curated and distributed by CZ CELLxGENE Discover in Collection: https://cellxgene.cziscience.com/collections/8e880741-bf9a-4c8e-9227-934204631d2a",
                  "8e880741-bf9a-4c8e-9227-934204631d2a",
                  "High Resolution Slide-seqV2 Spatial Transcriptomics Enables Discovery of Disease-Specific Cell Neighborhoods and Pathways",
                  "10.1016/j.isci.2022.104097",
                  "Marshall et al. (2022) iScience",
                  "10eed666-34e0-4b98-9384-179993873aa8",
                  "01e21c00-81c0-40ce-a6ee-553dad5fd863",
                  "Spatial transcriptomics in mouse: Puck_200104_03",
                  "10eed666-34e0-4b98-9384-179993873aa8.h5ad",
                  "16027"
                ],
                [
                  "29",
                  "29",
                  "Publication: https://doi.org/10.1016/j.celrep.2019.12.082 Dataset Version: https://datasets.cellxgene.cziscience.com/f394dd68-701c-4cc9-b6a3-6d7a18ffd221.h5ad curated and distributed by CZ CELLxGENE Discover in Collection: https://cellxgene.cziscience.com/collections/939769a8-d8d2-4d01-abfc-55699893fd49",
                  "939769a8-d8d2-4d01-abfc-55699893fd49",
                  "Integration of eQTL and a Single-Cell Atlas in the Human Eye Identifies Causal Genes for Age-Related Macular Degeneration",
                  "10.1016/j.celrep.2019.12.082",
                  "Orozco et al. (2020) Cell Reports",
                  "de17ac25-550a-4018-be75-bbb485a0636e",
                  "f394dd68-701c-4cc9-b6a3-6d7a18ffd221",
                  "Myeloid cells of human eye",
                  "de17ac25-550a-4018-be75-bbb485a0636e.h5ad",
                  "395"
                ],
                [
                  "30",
                  "30",
                  "Publication: https://doi.org/10.1101/2024.10.23.619925 Dataset Version: https://datasets.cellxgene.cziscience.com/77ab5a70-e848-49e8-9a64-ee81e6f9e772.h5ad curated and distributed by CZ CELLxGENE Discover in Collection: https://cellxgene.cziscience.com/collections/bdac7a53-fe34-4f04-8c46-f9bd5297c099",
                  "bdac7a53-fe34-4f04-8c46-f9bd5297c099",
                  "Profiling of epithelial functional states and fibroblast phenotypes in hormone therapy-naïve localised prostate cancer",
                  "10.1101/2024.10.23.619925",
                  "Apostolov et al. (2024) bioRxiv",
                  "04557650-a691-4306-a56c-232fbfeb236c",
                  "77ab5a70-e848-49e8-9a64-ee81e6f9e772",
                  "Single-cell atlas of 24 hormone therapy naive localised prostate cancers -- Cycling subset",
                  "04557650-a691-4306-a56c-232fbfeb236c.h5ad",
                  "268"
                ],
                [
                  "31",
                  "31",
                  "Publication: https://doi.org/10.1038/s41588-023-01435-6 Dataset Version: https://datasets.cellxgene.cziscience.com/263a81f8-bff0-4f8a-854b-4586472c7e06.h5ad curated and distributed by CZ CELLxGENE Discover in Collection: https://cellxgene.cziscience.com/collections/d74b6979-efba-47cd-990a-9d80ccf29055",
                  "d74b6979-efba-47cd-990a-9d80ccf29055",
                  "Spatial transcriptomic maps of whole mouse embryos reveal principles of neural tube patterning",
                  "10.1038/s41588-023-01435-6",
                  "Sampath Kumar et al. (2023) Nat Genet",
                  "e545cafe-80f4-494e-bbb1-e95c86955756",
                  "263a81f8-bff0-4f8a-854b-4586472c7e06",
                  "E9.0: Puck_220718_10",
                  "e545cafe-80f4-494e-bbb1-e95c86955756.h5ad",
                  "1336"
                ],
                [
                  "32",
                  "32",
                  "Publication: https://doi.org/10.1016/j.isci.2022.104097 Dataset Version: https://datasets.cellxgene.cziscience.com/2273a60c-7a4f-42a2-9e04-dc2da182f4e6.h5ad curated and distributed by CZ CELLxGENE Discover in Collection: https://cellxgene.cziscience.com/collections/8e880741-bf9a-4c8e-9227-934204631d2a",
                  "8e880741-bf9a-4c8e-9227-934204631d2a",
                  "High Resolution Slide-seqV2 Spatial Transcriptomics Enables Discovery of Disease-Specific Cell Neighborhoods and Pathways",
                  "10.1016/j.isci.2022.104097",
                  "Marshall et al. (2022) iScience",
                  "9098fb44-ac81-4ac6-be8e-464f10ec7128",
                  "2273a60c-7a4f-42a2-9e04-dc2da182f4e6",
                  "Spatial transcriptomics in mouse: Puck_191223_10",
                  "9098fb44-ac81-4ac6-be8e-464f10ec7128.h5ad",
                  "14961"
                ],
                [
                  "33",
                  "33",
                  "Publication: https://doi.org/10.1016/j.immuni.2022.09.002 Dataset Version: https://datasets.cellxgene.cziscience.com/243d1f06-6875-4e55-95d1-2e66f0ef9f96.h5ad curated and distributed by CZ CELLxGENE Discover in Collection: https://cellxgene.cziscience.com/collections/02b01703-bf1b-48de-b99a-23bef8cccc81",
                  "02b01703-bf1b-48de-b99a-23bef8cccc81",
                  "Spatially mapping T cell receptors and transcriptomes",
                  "10.1016/j.immuni.2022.09.002",
                  "Liu et al. (2022) Immunity",
                  "76bb43ff-e2f5-4513-a0f0-1059c860c3b7",
                  "243d1f06-6875-4e55-95d1-2e66f0ef9f96",
                  "Cutaneous melanoma, lymph node metastasis Puck_220408_08",
                  "76bb43ff-e2f5-4513-a0f0-1059c860c3b7.h5ad",
                  "19034"
                ],
                [
                  "34",
                  "34",
                  "Publication: https://doi.org/10.1101/2024.10.23.619925 Dataset Version: https://datasets.cellxgene.cziscience.com/a1d93898-7c7d-4ef8-aa35-af9bf863c564.h5ad curated and distributed by CZ CELLxGENE Discover in Collection: https://cellxgene.cziscience.com/collections/bdac7a53-fe34-4f04-8c46-f9bd5297c099",
                  "bdac7a53-fe34-4f04-8c46-f9bd5297c099",
                  "Profiling of epithelial functional states and fibroblast phenotypes in hormone therapy-naïve localised prostate cancer",
                  "10.1101/2024.10.23.619925",
                  "Apostolov et al. (2024) bioRxiv",
                  "182f8181-172b-46c3-9bc3-575bf3bce91b",
                  "a1d93898-7c7d-4ef8-aa35-af9bf863c564",
                  "Single-cell atlas of 24 hormone therapy naive localised prostate cancers -- B-cells subset",
                  "182f8181-172b-46c3-9bc3-575bf3bce91b.h5ad",
                  "706"
                ],
                [
                  "35",
                  "35",
                  "Publication: https://doi.org/10.1016/j.isci.2022.104097 Dataset Version: https://datasets.cellxgene.cziscience.com/c19b56a6-fc92-4898-9bdf-ced146a33608.h5ad curated and distributed by CZ CELLxGENE Discover in Collection: https://cellxgene.cziscience.com/collections/8e880741-bf9a-4c8e-9227-934204631d2a",
                  "8e880741-bf9a-4c8e-9227-934204631d2a",
                  "High Resolution Slide-seqV2 Spatial Transcriptomics Enables Discovery of Disease-Specific Cell Neighborhoods and Pathways",
                  "10.1016/j.isci.2022.104097",
                  "Marshall et al. (2022) iScience",
                  "3679ae7d-d70e-4c09-a8bb-78b8ae1006ff",
                  "c19b56a6-fc92-4898-9bdf-ced146a33608",
                  "Spatial transcriptomics in the healthy human kidney: Puck_200131_25",
                  "3679ae7d-d70e-4c09-a8bb-78b8ae1006ff.h5ad",
                  "10701"
                ],
                [
                  "36",
                  "36",
                  "Publication: https://doi.org/10.1038/s41586-021-03465-8 Dataset Version: https://datasets.cellxgene.cziscience.com/2349b8a9-a4f0-485c-856c-670528ac86a5.h5ad curated and distributed by CZ CELLxGENE Discover in Collection: https://cellxgene.cziscience.com/collections/367d95c0-0eb0-4dae-8276-9407239421ee",
                  "367d95c0-0eb0-4dae-8276-9407239421ee",
                  "Evolution of cellular diversity in primary motor cortex of human, marmoset monkey, and mouse",
                  "10.1038/s41586-021-03465-8",
                  "Bakken et al. (2021) Nature",
                  "9b686bb6-1427-4e13-b451-7ee961115cf9",
                  "2349b8a9-a4f0-485c-856c-670528ac86a5",
                  "Evolution of cellular diversity in primary motor cortex of human - non-neurons",
                  "9b686bb6-1427-4e13-b451-7ee961115cf9.h5ad",
                  "1705"
                ],
                [
                  "37",
                  "37",
                  "Publication: https://doi.org/10.1038/s41593-020-00764-7 Dataset Version: https://datasets.cellxgene.cziscience.com/9d1f1b35-8357-4d54-ac4f-84f1dde3f37c.h5ad curated and distributed by CZ CELLxGENE Discover in Collection: https://cellxgene.cziscience.com/collections/180bff9c-c8a5-4539-b13b-ddbc00d643e6",
                  "180bff9c-c8a5-4539-b13b-ddbc00d643e6",
                  "Molecular characterization of selectively vulnerable neurons in Alzheimer's Disease",
                  "10.1038/s41593-020-00764-7",
                  "Leng et al. (2021) Nat Neurosci",
                  "9f1049ac-f8b7-45ad-8e31-6e96c3e5058f",
                  "9d1f1b35-8357-4d54-ac4f-84f1dde3f37c",
                  "Molecular characterization of selectively vulnerable neurons in Alzheimer’s Disease: EC microglia",
                  "9f1049ac-f8b7-45ad-8e31-6e96c3e5058f.h5ad",
                  "5572"
                ],
                [
                  "38",
                  "38",
                  "Publication: https://doi.org/10.1016/j.isci.2022.104097 Dataset Version: https://datasets.cellxgene.cziscience.com/01458ecc-c1e7-4fe7-960d-750597bd124c.h5ad curated and distributed by CZ CELLxGENE Discover in Collection: https://cellxgene.cziscience.com/collections/8e880741-bf9a-4c8e-9227-934204631d2a",
                  "8e880741-bf9a-4c8e-9227-934204631d2a",
                  "High Resolution Slide-seqV2 Spatial Transcriptomics Enables Discovery of Disease-Specific Cell Neighborhoods and Pathways",
                  "10.1016/j.isci.2022.104097",
                  "Marshall et al. (2022) iScience",
                  "707acde8-4e99-4e31-8da9-782ded87d136",
                  "01458ecc-c1e7-4fe7-960d-750597bd124c",
                  "Spatial transcriptomics in mouse: Puck_191204_24",
                  "707acde8-4e99-4e31-8da9-782ded87d136.h5ad",
                  "27220"
                ],
                [
                  "39",
                  "39",
                  "Publication: https://doi.org/10.1016/j.isci.2022.104097 Dataset Version: https://datasets.cellxgene.cziscience.com/81faa4ce-75a1-4bdc-b79c-0a66212664ca.h5ad curated and distributed by CZ CELLxGENE Discover in Collection: https://cellxgene.cziscience.com/collections/8e880741-bf9a-4c8e-9227-934204631d2a",
                  "8e880741-bf9a-4c8e-9227-934204631d2a",
                  "High Resolution Slide-seqV2 Spatial Transcriptomics Enables Discovery of Disease-Specific Cell Neighborhoods and Pathways",
                  "10.1016/j.isci.2022.104097",
                  "Marshall et al. (2022) iScience",
                  "9246671b-f814-4252-8f1a-57f4f5e28f81",
                  "81faa4ce-75a1-4bdc-b79c-0a66212664ca",
                  "Spatial transcriptomics in mouse: Puck_200210_04",
                  "9246671b-f814-4252-8f1a-57f4f5e28f81.h5ad",
                  "20424"
                ],
                [
                  "40",
                  "40",
                  "Publication: https://doi.org/10.1016/j.isci.2022.104097 Dataset Version: https://datasets.cellxgene.cziscience.com/bca2a9d3-86dd-4011-95a9-621f4809d573.h5ad curated and distributed by CZ CELLxGENE Discover in Collection: https://cellxgene.cziscience.com/collections/8e880741-bf9a-4c8e-9227-934204631d2a",
                  "8e880741-bf9a-4c8e-9227-934204631d2a",
                  "High Resolution Slide-seqV2 Spatial Transcriptomics Enables Discovery of Disease-Specific Cell Neighborhoods and Pathways",
                  "10.1016/j.isci.2022.104097",
                  "Marshall et al. (2022) iScience",
                  "f9f63905-34b2-4c61-b747-1df6c7a35a6a",
                  "bca2a9d3-86dd-4011-95a9-621f4809d573",
                  "Spatial transcriptomics in mouse: Puck_200127_07",
                  "f9f63905-34b2-4c61-b747-1df6c7a35a6a.h5ad",
                  "17363"
                ],
                [
                  "41",
                  "41",
                  "Publication: https://doi.org/10.1016/j.isci.2022.104097 Dataset Version: https://datasets.cellxgene.cziscience.com/83749899-a5b5-4a3e-9966-1bc9802ef3af.h5ad curated and distributed by CZ CELLxGENE Discover in Collection: https://cellxgene.cziscience.com/collections/8e880741-bf9a-4c8e-9227-934204631d2a",
                  "8e880741-bf9a-4c8e-9227-934204631d2a",
                  "High Resolution Slide-seqV2 Spatial Transcriptomics Enables Discovery of Disease-Specific Cell Neighborhoods and Pathways",
                  "10.1016/j.isci.2022.104097",
                  "Marshall et al. (2022) iScience",
                  "34cd019c-c79d-4f4f-8f95-8934062dc132",
                  "83749899-a5b5-4a3e-9966-1bc9802ef3af",
                  "Spatial transcriptomics in mouse: Puck_191206_03",
                  "34cd019c-c79d-4f4f-8f95-8934062dc132.h5ad",
                  "18542"
                ],
                [
                  "42",
                  "42",
                  "Publication: https://doi.org/10.1016/j.isci.2022.104097 Dataset Version: https://datasets.cellxgene.cziscience.com/2446d4bd-2fd7-421a-bfd3-6133cef1a18b.h5ad curated and distributed by CZ CELLxGENE Discover in Collection: https://cellxgene.cziscience.com/collections/8e880741-bf9a-4c8e-9227-934204631d2a",
                  "8e880741-bf9a-4c8e-9227-934204631d2a",
                  "High Resolution Slide-seqV2 Spatial Transcriptomics Enables Discovery of Disease-Specific Cell Neighborhoods and Pathways",
                  "10.1016/j.isci.2022.104097",
                  "Marshall et al. (2022) iScience",
                  "a5a6c83a-3afe-4a87-8c32-691118c46e51",
                  "2446d4bd-2fd7-421a-bfd3-6133cef1a18b",
                  "Spatial transcriptomics in the healthy human kidney: Puck_200113_09",
                  "a5a6c83a-3afe-4a87-8c32-691118c46e51.h5ad",
                  "13780"
                ],
                [
                  "43",
                  "43",
                  "Publication: https://doi.org/10.1016/j.cell.2022.11.005 Dataset Version: https://datasets.cellxgene.cziscience.com/4e4be92c-985c-43c1-a24c-c6b2c0dca70a.h5ad curated and distributed by CZ CELLxGENE Discover in Collection: https://cellxgene.cziscience.com/collections/2d2e2acd-dade-489f-a2da-6c11aa654028",
                  "2d2e2acd-dade-489f-a2da-6c11aa654028",
                  "A human fetal lung cell atlas uncovers proximal-distal gradients of differentiation and key regulators of epithelial fates",
                  "10.1016/j.cell.2022.11.005",
                  "He et al. (2022) Cell",
                  "703f00e6-b996-48e5-bc34-00c41b9876f4",
                  "4e4be92c-985c-43c1-a24c-c6b2c0dca70a",
                  "PNS",
                  "703f00e6-b996-48e5-bc34-00c41b9876f4.h5ad",
                  "649"
                ],
                [
                  "44",
                  "44",
                  "Publication: https://doi.org/10.1016/j.isci.2022.104097 Dataset Version: https://datasets.cellxgene.cziscience.com/ab849e19-b792-4b70-b05c-9972c5b10192.h5ad curated and distributed by CZ CELLxGENE Discover in Collection: https://cellxgene.cziscience.com/collections/8e880741-bf9a-4c8e-9227-934204631d2a",
                  "8e880741-bf9a-4c8e-9227-934204631d2a",
                  "High Resolution Slide-seqV2 Spatial Transcriptomics Enables Discovery of Disease-Specific Cell Neighborhoods and Pathways",
                  "10.1016/j.isci.2022.104097",
                  "Marshall et al. (2022) iScience",
                  "a21a47fe-1516-47d0-aeca-fd90d729fa4a",
                  "ab849e19-b792-4b70-b05c-9972c5b10192",
                  "Spatial transcriptomics in mouse: Puck_200131_23",
                  "a21a47fe-1516-47d0-aeca-fd90d729fa4a.h5ad",
                  "12533"
                ],
                [
                  "45",
                  "45",
                  "Publication: https://doi.org/10.1016/j.labinv.2023.100258 Dataset Version: https://datasets.cellxgene.cziscience.com/69f59798-b26a-49e5-846a-cc4c2ce002d9.h5ad curated and distributed by CZ CELLxGENE Discover in Collection: https://cellxgene.cziscience.com/collections/bfd55632-15e2-4dde-a1b9-02ade511fc4d",
                  "bfd55632-15e2-4dde-a1b9-02ade511fc4d",
                  "Spatial Transcriptomics Reveal Pitfalls and Opportunities for the Detection of Rare High-Plasticity Breast Cancer Subtypes",
                  "10.1016/j.labinv.2023.100258",
                  "Coutant et al. (2023) Laboratory Investigation",
                  "4db76400-e867-4fd5-9f49-fdc489b38269",
                  "69f59798-b26a-49e5-846a-cc4c2ce002d9",
                  "GSM6592056_M9",
                  "4db76400-e867-4fd5-9f49-fdc489b38269.h5ad",
                  "4992"
                ],
                [
                  "46",
                  "46",
                  "Publication: https://doi.org/10.1016/j.isci.2022.104097 Dataset Version: https://datasets.cellxgene.cziscience.com/ed8b8429-27f3-4d5e-b0e7-56cde4173ddf.h5ad curated and distributed by CZ CELLxGENE Discover in Collection: https://cellxgene.cziscience.com/collections/8e880741-bf9a-4c8e-9227-934204631d2a",
                  "8e880741-bf9a-4c8e-9227-934204631d2a",
                  "High Resolution Slide-seqV2 Spatial Transcriptomics Enables Discovery of Disease-Specific Cell Neighborhoods and Pathways",
                  "10.1016/j.isci.2022.104097",
                  "Marshall et al. (2022) iScience",
                  "0738f538-ff2f-4346-b2eb-72704c291188",
                  "ed8b8429-27f3-4d5e-b0e7-56cde4173ddf",
                  "Spatial transcriptomics in the healthy human kidney: Puck_200113_12",
                  "0738f538-ff2f-4346-b2eb-72704c291188.h5ad",
                  "14010"
                ],
                [
                  "47",
                  "47",
                  "Publication: https://doi.org/10.1016/j.isci.2022.104097 Dataset Version: https://datasets.cellxgene.cziscience.com/dd81e833-ed45-42fe-a7fa-5335a31a3cad.h5ad curated and distributed by CZ CELLxGENE Discover in Collection: https://cellxgene.cziscience.com/collections/8e880741-bf9a-4c8e-9227-934204631d2a",
                  "8e880741-bf9a-4c8e-9227-934204631d2a",
                  "High Resolution Slide-seqV2 Spatial Transcriptomics Enables Discovery of Disease-Specific Cell Neighborhoods and Pathways",
                  "10.1016/j.isci.2022.104097",
                  "Marshall et al. (2022) iScience",
                  "286fead0-7be7-46db-b358-7abc6d39adf7",
                  "dd81e833-ed45-42fe-a7fa-5335a31a3cad",
                  "Spatial transcriptomics in mouse: Puck_191109_09",
                  "286fead0-7be7-46db-b358-7abc6d39adf7.h5ad",
                  "29011"
                ],
                [
                  "48",
                  "48",
                  "Publication: https://doi.org/10.1038/s41588-023-01435-6 Dataset Version: https://datasets.cellxgene.cziscience.com/2238fde9-1ca2-49a7-9c20-664b94d24d97.h5ad curated and distributed by CZ CELLxGENE Discover in Collection: https://cellxgene.cziscience.com/collections/d74b6979-efba-47cd-990a-9d80ccf29055",
                  "d74b6979-efba-47cd-990a-9d80ccf29055",
                  "Spatial transcriptomic maps of whole mouse embryos reveal principles of neural tube patterning",
                  "10.1038/s41588-023-01435-6",
                  "Sampath Kumar et al. (2023) Nat Genet",
                  "63521245-fe33-4e0f-870a-8f5d083f69f9",
                  "2238fde9-1ca2-49a7-9c20-664b94d24d97",
                  "E9.0: Puck_220615_39",
                  "63521245-fe33-4e0f-870a-8f5d083f69f9.h5ad",
                  "2567"
                ],
                [
                  "49",
                  "49",
                  "Publication: https://doi.org/10.1016/j.isci.2022.104097 Dataset Version: https://datasets.cellxgene.cziscience.com/bab79800-d439-4172-8dd9-46c46e425dff.h5ad curated and distributed by CZ CELLxGENE Discover in Collection: https://cellxgene.cziscience.com/collections/8e880741-bf9a-4c8e-9227-934204631d2a",
                  "8e880741-bf9a-4c8e-9227-934204631d2a",
                  "High Resolution Slide-seqV2 Spatial Transcriptomics Enables Discovery of Disease-Specific Cell Neighborhoods and Pathways",
                  "10.1016/j.isci.2022.104097",
                  "Marshall et al. (2022) iScience",
                  "68bcbb10-cc0d-435b-ba76-ab03c040967b",
                  "bab79800-d439-4172-8dd9-46c46e425dff",
                  "Spatial transcriptomics in mouse: Puck_200131_22",
                  "68bcbb10-cc0d-435b-ba76-ab03c040967b.h5ad",
                  "12541"
                ]
              ],
              "shape": {
                "columns": 11,
                "rows": 1815
              }
            },
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>soma_joinid</th>\n",
              "      <th>citation</th>\n",
              "      <th>collection_id</th>\n",
              "      <th>collection_name</th>\n",
              "      <th>collection_doi</th>\n",
              "      <th>collection_doi_label</th>\n",
              "      <th>dataset_id</th>\n",
              "      <th>dataset_version_id</th>\n",
              "      <th>dataset_title</th>\n",
              "      <th>dataset_h5ad_path</th>\n",
              "      <th>dataset_total_cell_count</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>0</td>\n",
              "      <td>Publication: https://doi.org/10.1016/j.isci.20...</td>\n",
              "      <td>8e880741-bf9a-4c8e-9227-934204631d2a</td>\n",
              "      <td>High Resolution Slide-seqV2 Spatial Transcript...</td>\n",
              "      <td>10.1016/j.isci.2022.104097</td>\n",
              "      <td>Marshall et al. (2022) iScience</td>\n",
              "      <td>4eb29386-de81-452f-b3c0-e00844e8c7fd</td>\n",
              "      <td>66699060-0389-4fbd-b3a5-196b3b4e32d6</td>\n",
              "      <td>Spatial transcriptomics in mouse: Puck_191112_05</td>\n",
              "      <td>4eb29386-de81-452f-b3c0-e00844e8c7fd.h5ad</td>\n",
              "      <td>10888</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>1</td>\n",
              "      <td>Publication: https://doi.org/10.1016/j.isci.20...</td>\n",
              "      <td>8e880741-bf9a-4c8e-9227-934204631d2a</td>\n",
              "      <td>High Resolution Slide-seqV2 Spatial Transcript...</td>\n",
              "      <td>10.1016/j.isci.2022.104097</td>\n",
              "      <td>Marshall et al. (2022) iScience</td>\n",
              "      <td>78d59e4a-82eb-4a61-a1dc-da974d7ea54b</td>\n",
              "      <td>f64950a2-a3c8-490a-8431-7121eeb4f5f4</td>\n",
              "      <td>Spatial transcriptomics in mouse: Puck_191112_08</td>\n",
              "      <td>78d59e4a-82eb-4a61-a1dc-da974d7ea54b.h5ad</td>\n",
              "      <td>10250</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>2</td>\n",
              "      <td>Publication: https://doi.org/10.1016/j.isci.20...</td>\n",
              "      <td>8e880741-bf9a-4c8e-9227-934204631d2a</td>\n",
              "      <td>High Resolution Slide-seqV2 Spatial Transcript...</td>\n",
              "      <td>10.1016/j.isci.2022.104097</td>\n",
              "      <td>Marshall et al. (2022) iScience</td>\n",
              "      <td>add5eb84-5fc9-4f01-982e-a346dd42ee82</td>\n",
              "      <td>781a724a-b0f5-46c4-9a13-e6293ef4364f</td>\n",
              "      <td>Spatial transcriptomics in mouse: Puck_191109_20</td>\n",
              "      <td>add5eb84-5fc9-4f01-982e-a346dd42ee82.h5ad</td>\n",
              "      <td>12906</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>3</td>\n",
              "      <td>Publication: https://doi.org/10.1016/j.isci.20...</td>\n",
              "      <td>8e880741-bf9a-4c8e-9227-934204631d2a</td>\n",
              "      <td>High Resolution Slide-seqV2 Spatial Transcript...</td>\n",
              "      <td>10.1016/j.isci.2022.104097</td>\n",
              "      <td>Marshall et al. (2022) iScience</td>\n",
              "      <td>b020294c-ab82-4547-b5a7-63d8ffa575ed</td>\n",
              "      <td>96a79598-297b-4ade-a6c1-a431ab243548</td>\n",
              "      <td>Spatial transcriptomics in mouse: Puck_191112_13</td>\n",
              "      <td>b020294c-ab82-4547-b5a7-63d8ffa575ed.h5ad</td>\n",
              "      <td>15161</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>4</td>\n",
              "      <td>Publication: https://doi.org/10.1038/s41591-02...</td>\n",
              "      <td>a96133de-e951-4e2d-ace6-59db8b3bfb1d</td>\n",
              "      <td>HTAN/HTAPP Broad - Spatio-molecular dissection...</td>\n",
              "      <td>10.1038/s41591-024-03215-z</td>\n",
              "      <td>Klughammer et al. (2024) Nat Med</td>\n",
              "      <td>d7476ae2-e320-4703-8304-da5c42627e71</td>\n",
              "      <td>ac9fe945-6784-48fa-a2d5-a8646196d37e</td>\n",
              "      <td>HTAPP-330-SMP-1082 scRNA-seq</td>\n",
              "      <td>d7476ae2-e320-4703-8304-da5c42627e71.h5ad</td>\n",
              "      <td>565</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1810</th>\n",
              "      <td>1810</td>\n",
              "      <td>Publication: https://doi.org/10.1038/s41586-02...</td>\n",
              "      <td>45d5d2c3-bc28-4814-aed6-0bb6f0e11c82</td>\n",
              "      <td>A single-cell transcriptional timelapse of mou...</td>\n",
              "      <td>10.1038/s41586-024-07069-w</td>\n",
              "      <td>Qiu et al. (2024) Nature</td>\n",
              "      <td>dcfa2614-7ca7-4d82-814c-350626eccb26</td>\n",
              "      <td>ca20ef35-13c0-4850-bec4-ba00e6bbd6f9</td>\n",
              "      <td>Major cell cluster: Mesoderm</td>\n",
              "      <td>dcfa2614-7ca7-4d82-814c-350626eccb26.h5ad</td>\n",
              "      <td>3267338</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1811</th>\n",
              "      <td>1811</td>\n",
              "      <td>Publication: https://doi.org/10.1126/science.a...</td>\n",
              "      <td>e5f58829-1a66-40b5-a624-9046778e74f5</td>\n",
              "      <td>Tabula Sapiens</td>\n",
              "      <td>10.1126/science.abl4896</td>\n",
              "      <td>The Tabula Sapiens Consortium* et al. (2022) S...</td>\n",
              "      <td>53d208b0-2cfd-4366-9866-c3c6114081bc</td>\n",
              "      <td>5a495302-b7cd-4bf9-853e-95627b00bb03</td>\n",
              "      <td>Tabula Sapiens - All Cells</td>\n",
              "      <td>53d208b0-2cfd-4366-9866-c3c6114081bc.h5ad</td>\n",
              "      <td>1136218</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1812</th>\n",
              "      <td>1812</td>\n",
              "      <td>Publication: https://doi.org/10.1038/s41586-02...</td>\n",
              "      <td>45d5d2c3-bc28-4814-aed6-0bb6f0e11c82</td>\n",
              "      <td>A single-cell transcriptional timelapse of mou...</td>\n",
              "      <td>10.1038/s41586-024-07069-w</td>\n",
              "      <td>Qiu et al. (2024) Nature</td>\n",
              "      <td>dcfd4feb-18a3-4b30-81d7-1b0c544a8ab3</td>\n",
              "      <td>a5a85963-8004-41a1-8eb5-ca65266d89c3</td>\n",
              "      <td>Whole dataset: Raw counts only</td>\n",
              "      <td>dcfd4feb-18a3-4b30-81d7-1b0c544a8ab3.h5ad</td>\n",
              "      <td>11441407</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1813</th>\n",
              "      <td>1813</td>\n",
              "      <td>Publication: https://doi.org/10.1038/s41593-02...</td>\n",
              "      <td>1ca90a2d-2943-483d-b678-b809bf464c30</td>\n",
              "      <td>SEA-AD: Seattle Alzheimer’s Disease Brain Cell...</td>\n",
              "      <td>10.1038/s41593-024-01774-5</td>\n",
              "      <td>Gabitto et al. (2024) Nat Neurosci</td>\n",
              "      <td>c2876b1b-06d8-4d96-a56b-5304f815b99a</td>\n",
              "      <td>92b37feb-aa2c-40d7-bd90-0a9b5ddb3b27</td>\n",
              "      <td>Whole Taxonomy - MTG: Seattle Alzheimer's Dise...</td>\n",
              "      <td>c2876b1b-06d8-4d96-a56b-5304f815b99a.h5ad</td>\n",
              "      <td>1378211</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1814</th>\n",
              "      <td>1814</td>\n",
              "      <td>Publication: https://doi.org/10.1038/s41593-02...</td>\n",
              "      <td>1ca90a2d-2943-483d-b678-b809bf464c30</td>\n",
              "      <td>SEA-AD: Seattle Alzheimer’s Disease Brain Cell...</td>\n",
              "      <td>10.1038/s41593-024-01774-5</td>\n",
              "      <td>Gabitto et al. (2024) Nat Neurosci</td>\n",
              "      <td>6f7fd0f1-a2ed-4ff1-80d3-33dde731cbc3</td>\n",
              "      <td>dad4819b-4c14-439c-b32a-2c8d68bd22e1</td>\n",
              "      <td>Whole Taxonomy - DLPFC: Seattle Alzheimer's Di...</td>\n",
              "      <td>6f7fd0f1-a2ed-4ff1-80d3-33dde731cbc3.h5ad</td>\n",
              "      <td>1395601</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>1815 rows × 11 columns</p>\n",
              "</div>"
            ],
            "text/plain": [
              "      soma_joinid                                           citation  \\\n",
              "0               0  Publication: https://doi.org/10.1016/j.isci.20...   \n",
              "1               1  Publication: https://doi.org/10.1016/j.isci.20...   \n",
              "2               2  Publication: https://doi.org/10.1016/j.isci.20...   \n",
              "3               3  Publication: https://doi.org/10.1016/j.isci.20...   \n",
              "4               4  Publication: https://doi.org/10.1038/s41591-02...   \n",
              "...           ...                                                ...   \n",
              "1810         1810  Publication: https://doi.org/10.1038/s41586-02...   \n",
              "1811         1811  Publication: https://doi.org/10.1126/science.a...   \n",
              "1812         1812  Publication: https://doi.org/10.1038/s41586-02...   \n",
              "1813         1813  Publication: https://doi.org/10.1038/s41593-02...   \n",
              "1814         1814  Publication: https://doi.org/10.1038/s41593-02...   \n",
              "\n",
              "                             collection_id  \\\n",
              "0     8e880741-bf9a-4c8e-9227-934204631d2a   \n",
              "1     8e880741-bf9a-4c8e-9227-934204631d2a   \n",
              "2     8e880741-bf9a-4c8e-9227-934204631d2a   \n",
              "3     8e880741-bf9a-4c8e-9227-934204631d2a   \n",
              "4     a96133de-e951-4e2d-ace6-59db8b3bfb1d   \n",
              "...                                    ...   \n",
              "1810  45d5d2c3-bc28-4814-aed6-0bb6f0e11c82   \n",
              "1811  e5f58829-1a66-40b5-a624-9046778e74f5   \n",
              "1812  45d5d2c3-bc28-4814-aed6-0bb6f0e11c82   \n",
              "1813  1ca90a2d-2943-483d-b678-b809bf464c30   \n",
              "1814  1ca90a2d-2943-483d-b678-b809bf464c30   \n",
              "\n",
              "                                        collection_name  \\\n",
              "0     High Resolution Slide-seqV2 Spatial Transcript...   \n",
              "1     High Resolution Slide-seqV2 Spatial Transcript...   \n",
              "2     High Resolution Slide-seqV2 Spatial Transcript...   \n",
              "3     High Resolution Slide-seqV2 Spatial Transcript...   \n",
              "4     HTAN/HTAPP Broad - Spatio-molecular dissection...   \n",
              "...                                                 ...   \n",
              "1810  A single-cell transcriptional timelapse of mou...   \n",
              "1811                                     Tabula Sapiens   \n",
              "1812  A single-cell transcriptional timelapse of mou...   \n",
              "1813  SEA-AD: Seattle Alzheimer’s Disease Brain Cell...   \n",
              "1814  SEA-AD: Seattle Alzheimer’s Disease Brain Cell...   \n",
              "\n",
              "                  collection_doi  \\\n",
              "0     10.1016/j.isci.2022.104097   \n",
              "1     10.1016/j.isci.2022.104097   \n",
              "2     10.1016/j.isci.2022.104097   \n",
              "3     10.1016/j.isci.2022.104097   \n",
              "4     10.1038/s41591-024-03215-z   \n",
              "...                          ...   \n",
              "1810  10.1038/s41586-024-07069-w   \n",
              "1811     10.1126/science.abl4896   \n",
              "1812  10.1038/s41586-024-07069-w   \n",
              "1813  10.1038/s41593-024-01774-5   \n",
              "1814  10.1038/s41593-024-01774-5   \n",
              "\n",
              "                                   collection_doi_label  \\\n",
              "0                       Marshall et al. (2022) iScience   \n",
              "1                       Marshall et al. (2022) iScience   \n",
              "2                       Marshall et al. (2022) iScience   \n",
              "3                       Marshall et al. (2022) iScience   \n",
              "4                      Klughammer et al. (2024) Nat Med   \n",
              "...                                                 ...   \n",
              "1810                           Qiu et al. (2024) Nature   \n",
              "1811  The Tabula Sapiens Consortium* et al. (2022) S...   \n",
              "1812                           Qiu et al. (2024) Nature   \n",
              "1813                 Gabitto et al. (2024) Nat Neurosci   \n",
              "1814                 Gabitto et al. (2024) Nat Neurosci   \n",
              "\n",
              "                                dataset_id  \\\n",
              "0     4eb29386-de81-452f-b3c0-e00844e8c7fd   \n",
              "1     78d59e4a-82eb-4a61-a1dc-da974d7ea54b   \n",
              "2     add5eb84-5fc9-4f01-982e-a346dd42ee82   \n",
              "3     b020294c-ab82-4547-b5a7-63d8ffa575ed   \n",
              "4     d7476ae2-e320-4703-8304-da5c42627e71   \n",
              "...                                    ...   \n",
              "1810  dcfa2614-7ca7-4d82-814c-350626eccb26   \n",
              "1811  53d208b0-2cfd-4366-9866-c3c6114081bc   \n",
              "1812  dcfd4feb-18a3-4b30-81d7-1b0c544a8ab3   \n",
              "1813  c2876b1b-06d8-4d96-a56b-5304f815b99a   \n",
              "1814  6f7fd0f1-a2ed-4ff1-80d3-33dde731cbc3   \n",
              "\n",
              "                        dataset_version_id  \\\n",
              "0     66699060-0389-4fbd-b3a5-196b3b4e32d6   \n",
              "1     f64950a2-a3c8-490a-8431-7121eeb4f5f4   \n",
              "2     781a724a-b0f5-46c4-9a13-e6293ef4364f   \n",
              "3     96a79598-297b-4ade-a6c1-a431ab243548   \n",
              "4     ac9fe945-6784-48fa-a2d5-a8646196d37e   \n",
              "...                                    ...   \n",
              "1810  ca20ef35-13c0-4850-bec4-ba00e6bbd6f9   \n",
              "1811  5a495302-b7cd-4bf9-853e-95627b00bb03   \n",
              "1812  a5a85963-8004-41a1-8eb5-ca65266d89c3   \n",
              "1813  92b37feb-aa2c-40d7-bd90-0a9b5ddb3b27   \n",
              "1814  dad4819b-4c14-439c-b32a-2c8d68bd22e1   \n",
              "\n",
              "                                          dataset_title  \\\n",
              "0      Spatial transcriptomics in mouse: Puck_191112_05   \n",
              "1      Spatial transcriptomics in mouse: Puck_191112_08   \n",
              "2      Spatial transcriptomics in mouse: Puck_191109_20   \n",
              "3      Spatial transcriptomics in mouse: Puck_191112_13   \n",
              "4                          HTAPP-330-SMP-1082 scRNA-seq   \n",
              "...                                                 ...   \n",
              "1810                       Major cell cluster: Mesoderm   \n",
              "1811                         Tabula Sapiens - All Cells   \n",
              "1812                     Whole dataset: Raw counts only   \n",
              "1813  Whole Taxonomy - MTG: Seattle Alzheimer's Dise...   \n",
              "1814  Whole Taxonomy - DLPFC: Seattle Alzheimer's Di...   \n",
              "\n",
              "                              dataset_h5ad_path  dataset_total_cell_count  \n",
              "0     4eb29386-de81-452f-b3c0-e00844e8c7fd.h5ad                     10888  \n",
              "1     78d59e4a-82eb-4a61-a1dc-da974d7ea54b.h5ad                     10250  \n",
              "2     add5eb84-5fc9-4f01-982e-a346dd42ee82.h5ad                     12906  \n",
              "3     b020294c-ab82-4547-b5a7-63d8ffa575ed.h5ad                     15161  \n",
              "4     d7476ae2-e320-4703-8304-da5c42627e71.h5ad                       565  \n",
              "...                                         ...                       ...  \n",
              "1810  dcfa2614-7ca7-4d82-814c-350626eccb26.h5ad                   3267338  \n",
              "1811  53d208b0-2cfd-4366-9866-c3c6114081bc.h5ad                   1136218  \n",
              "1812  dcfd4feb-18a3-4b30-81d7-1b0c544a8ab3.h5ad                  11441407  \n",
              "1813  c2876b1b-06d8-4d96-a56b-5304f815b99a.h5ad                   1378211  \n",
              "1814  6f7fd0f1-a2ed-4ff1-80d3-33dde731cbc3.h5ad                   1395601  \n",
              "\n",
              "[1815 rows x 11 columns]"
            ]
          },
          "execution_count": 7,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "census_new[\"census_info\"][\"datasets\"].read().concat().to_pandas()"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "2e139975-b0c5-41e3-9e2f-64dd6c6d9648",
      "metadata": {
        "id": "2e139975-b0c5-41e3-9e2f-64dd6c6d9648"
      },
      "source": [
        "## Checking list of all assays and veryfying new ones"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 8,
      "id": "3e08ddf7-1f9d-4185-8272-1858abcffcc6",
      "metadata": {
        "id": "3e08ddf7-1f9d-4185-8272-1858abcffcc6"
      },
      "outputs": [],
      "source": [
        "assays_old = census_old[\"census_data\"][\"homo_sapiens\"].obs.read(column_names=[\"assay\"]).concat().to_pandas().value_counts().reset_index()"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 9,
      "id": "5df8ec89-8fb8-49ff-a6fc-e239427e3929",
      "metadata": {
        "id": "5df8ec89-8fb8-49ff-a6fc-e239427e3929"
      },
      "outputs": [],
      "source": [
        "assays_new = census_new[\"census_data\"][\"homo_sapiens\"].obs.read(column_names=[\"assay\"]).concat().to_pandas().value_counts().reset_index()"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 10,
      "id": "92ac549e-e01d-4297-8dea-8dc811f23357",
      "metadata": {
        "id": "92ac549e-e01d-4297-8dea-8dc811f23357",
        "outputId": "de1bfa0b-25c4-402b-8b56-2267415cb856"
      },
      "outputs": [
        {
          "data": {
            "application/vnd.microsoft.datawrangler.viewer.v0+json": {
              "columns": [
                {
                  "name": "index",
                  "rawType": "int64",
                  "type": "integer"
                },
                {
                  "name": "assay",
                  "rawType": "object",
                  "type": "string"
                },
                {
                  "name": "count_x",
                  "rawType": "float64",
                  "type": "float"
                },
                {
                  "name": "count_y",
                  "rawType": "int64",
                  "type": "integer"
                },
                {
                  "name": "delta",
                  "rawType": "float64",
                  "type": "float"
                }
              ],
              "ref": "a9bac230-ad3d-4fa4-9561-90615da65365",
              "rows": [
                [
                  "3",
                  "10x 3' v3",
                  "59668147.0",
                  "96655665",
                  "36987518.0"
                ],
                [
                  "2",
                  "10x 3' v2",
                  "22750589.0",
                  "24507159",
                  "1756570.0"
                ],
                [
                  "6",
                  "10x 5' v2",
                  "4363694.0",
                  "8748986",
                  "4385292.0"
                ],
                [
                  "5",
                  "10x 5' v1",
                  "7448617.0",
                  "7683010",
                  "234393.0"
                ],
                [
                  "36",
                  "sci-RNA-seq3",
                  "5064268.0",
                  "5230353",
                  "166085.0"
                ],
                [
                  "8",
                  "10x multiome",
                  null,
                  "3033426",
                  null
                ],
                [
                  "4",
                  "10x 5' transcription profiling",
                  "1968545.0",
                  "2125643",
                  "157098.0"
                ],
                [
                  "14",
                  "Drop-seq",
                  "1048377.0",
                  "1048377",
                  "0.0"
                ],
                [
                  "23",
                  "ScaleBio single cell RNA sequencing",
                  "700524.0",
                  "700524",
                  "0.0"
                ],
                [
                  "0",
                  "10x 3' transcription profiling",
                  "665642.0",
                  "665642",
                  "0.0"
                ],
                [
                  "33",
                  "microwell-seq",
                  "642559.0",
                  "642559",
                  "0.0"
                ],
                [
                  "11",
                  "BD Rhapsody Whole Transcriptome Analysis",
                  "269298.0",
                  "269298",
                  "0.0"
                ],
                [
                  "28",
                  "Smart-seq2",
                  "260350.0",
                  "267948",
                  "7598.0"
                ],
                [
                  "24",
                  "Seq-Well",
                  "255232.0",
                  "255232",
                  "0.0"
                ],
                [
                  "1",
                  "10x 3' v1",
                  "213017.0",
                  "232475",
                  "19458.0"
                ],
                [
                  "31",
                  "inDrop",
                  "51304.0",
                  "136019",
                  "84715.0"
                ],
                [
                  "30",
                  "TruDrop",
                  "133507.0",
                  "133507",
                  "0.0"
                ],
                [
                  "15",
                  "GEXSCOPE technology",
                  "117962.0",
                  "117962",
                  "0.0"
                ],
                [
                  "7",
                  "10x gene expression flex",
                  "86356.0",
                  "117356",
                  "31000.0"
                ],
                [
                  "10",
                  "BD Rhapsody Targeted mRNA",
                  "96145.0",
                  "96145",
                  "0.0"
                ],
                [
                  "21",
                  "SPLiT-seq",
                  "81936.0",
                  "81936",
                  "0.0"
                ],
                [
                  "17",
                  "MARS-seq",
                  "70146.0",
                  "70146",
                  "0.0"
                ],
                [
                  "27",
                  "Smart-seq v4",
                  "63920.0",
                  "63920",
                  "0.0"
                ],
                [
                  "35",
                  "particle-templated instant partition sequencing",
                  null,
                  "46500",
                  null
                ],
                [
                  "25",
                  "Seq-Well S3",
                  "41560.0",
                  "41560",
                  "0.0"
                ],
                [
                  "32",
                  "mCT-seq",
                  null,
                  "37121",
                  null
                ],
                [
                  "16",
                  "HIVE CLX Single-Cell RNAseq Solution",
                  null,
                  "31000",
                  null
                ],
                [
                  "13",
                  "CEL-seq2",
                  "25416.0",
                  "25416",
                  "0.0"
                ],
                [
                  "9",
                  "Asteria scRNA-seq kit",
                  null,
                  "15764",
                  null
                ],
                [
                  "18",
                  "Parse Evercode Whole Transcriptome v2",
                  null,
                  "15500",
                  null
                ],
                [
                  "20",
                  "SORT-seq",
                  "11640.0",
                  "11640",
                  "0.0"
                ],
                [
                  "22",
                  "STRT-seq",
                  "11357.0",
                  "9386",
                  "-1971.0"
                ],
                [
                  "29",
                  "Smart-seq3",
                  "5005.0",
                  "5005",
                  "0.0"
                ],
                [
                  "19",
                  "Quartz-seq",
                  "2882.0",
                  "2882",
                  "0.0"
                ],
                [
                  "34",
                  "modified STRT-seq",
                  null,
                  "1971",
                  null
                ],
                [
                  "12",
                  "CEL-seq",
                  null,
                  "796",
                  null
                ],
                [
                  "26",
                  "Smart-seq",
                  "172.0",
                  "172",
                  "0.0"
                ]
              ],
              "shape": {
                "columns": 4,
                "rows": 37
              }
            },
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>assay</th>\n",
              "      <th>count_x</th>\n",
              "      <th>count_y</th>\n",
              "      <th>delta</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>10x 3' v3</td>\n",
              "      <td>59668147.0</td>\n",
              "      <td>96655665</td>\n",
              "      <td>36987518.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>10x 3' v2</td>\n",
              "      <td>22750589.0</td>\n",
              "      <td>24507159</td>\n",
              "      <td>1756570.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>6</th>\n",
              "      <td>10x 5' v2</td>\n",
              "      <td>4363694.0</td>\n",
              "      <td>8748986</td>\n",
              "      <td>4385292.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5</th>\n",
              "      <td>10x 5' v1</td>\n",
              "      <td>7448617.0</td>\n",
              "      <td>7683010</td>\n",
              "      <td>234393.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>36</th>\n",
              "      <td>sci-RNA-seq3</td>\n",
              "      <td>5064268.0</td>\n",
              "      <td>5230353</td>\n",
              "      <td>166085.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>8</th>\n",
              "      <td>10x multiome</td>\n",
              "      <td>NaN</td>\n",
              "      <td>3033426</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>10x 5' transcription profiling</td>\n",
              "      <td>1968545.0</td>\n",
              "      <td>2125643</td>\n",
              "      <td>157098.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>14</th>\n",
              "      <td>Drop-seq</td>\n",
              "      <td>1048377.0</td>\n",
              "      <td>1048377</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>23</th>\n",
              "      <td>ScaleBio single cell RNA sequencing</td>\n",
              "      <td>700524.0</td>\n",
              "      <td>700524</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>10x 3' transcription profiling</td>\n",
              "      <td>665642.0</td>\n",
              "      <td>665642</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>33</th>\n",
              "      <td>microwell-seq</td>\n",
              "      <td>642559.0</td>\n",
              "      <td>642559</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>11</th>\n",
              "      <td>BD Rhapsody Whole Transcriptome Analysis</td>\n",
              "      <td>269298.0</td>\n",
              "      <td>269298</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>28</th>\n",
              "      <td>Smart-seq2</td>\n",
              "      <td>260350.0</td>\n",
              "      <td>267948</td>\n",
              "      <td>7598.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>24</th>\n",
              "      <td>Seq-Well</td>\n",
              "      <td>255232.0</td>\n",
              "      <td>255232</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>10x 3' v1</td>\n",
              "      <td>213017.0</td>\n",
              "      <td>232475</td>\n",
              "      <td>19458.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>31</th>\n",
              "      <td>inDrop</td>\n",
              "      <td>51304.0</td>\n",
              "      <td>136019</td>\n",
              "      <td>84715.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>30</th>\n",
              "      <td>TruDrop</td>\n",
              "      <td>133507.0</td>\n",
              "      <td>133507</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>15</th>\n",
              "      <td>GEXSCOPE technology</td>\n",
              "      <td>117962.0</td>\n",
              "      <td>117962</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>7</th>\n",
              "      <td>10x gene expression flex</td>\n",
              "      <td>86356.0</td>\n",
              "      <td>117356</td>\n",
              "      <td>31000.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>10</th>\n",
              "      <td>BD Rhapsody Targeted mRNA</td>\n",
              "      <td>96145.0</td>\n",
              "      <td>96145</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>21</th>\n",
              "      <td>SPLiT-seq</td>\n",
              "      <td>81936.0</td>\n",
              "      <td>81936</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>17</th>\n",
              "      <td>MARS-seq</td>\n",
              "      <td>70146.0</td>\n",
              "      <td>70146</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>27</th>\n",
              "      <td>Smart-seq v4</td>\n",
              "      <td>63920.0</td>\n",
              "      <td>63920</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>35</th>\n",
              "      <td>particle-templated instant partition sequencing</td>\n",
              "      <td>NaN</td>\n",
              "      <td>46500</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>25</th>\n",
              "      <td>Seq-Well S3</td>\n",
              "      <td>41560.0</td>\n",
              "      <td>41560</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>32</th>\n",
              "      <td>mCT-seq</td>\n",
              "      <td>NaN</td>\n",
              "      <td>37121</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>16</th>\n",
              "      <td>HIVE CLX Single-Cell RNAseq Solution</td>\n",
              "      <td>NaN</td>\n",
              "      <td>31000</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>13</th>\n",
              "      <td>CEL-seq2</td>\n",
              "      <td>25416.0</td>\n",
              "      <td>25416</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>9</th>\n",
              "      <td>Asteria scRNA-seq kit</td>\n",
              "      <td>NaN</td>\n",
              "      <td>15764</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>18</th>\n",
              "      <td>Parse Evercode Whole Transcriptome v2</td>\n",
              "      <td>NaN</td>\n",
              "      <td>15500</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>20</th>\n",
              "      <td>SORT-seq</td>\n",
              "      <td>11640.0</td>\n",
              "      <td>11640</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>22</th>\n",
              "      <td>STRT-seq</td>\n",
              "      <td>11357.0</td>\n",
              "      <td>9386</td>\n",
              "      <td>-1971.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>29</th>\n",
              "      <td>Smart-seq3</td>\n",
              "      <td>5005.0</td>\n",
              "      <td>5005</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>19</th>\n",
              "      <td>Quartz-seq</td>\n",
              "      <td>2882.0</td>\n",
              "      <td>2882</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>34</th>\n",
              "      <td>modified STRT-seq</td>\n",
              "      <td>NaN</td>\n",
              "      <td>1971</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>12</th>\n",
              "      <td>CEL-seq</td>\n",
              "      <td>NaN</td>\n",
              "      <td>796</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>26</th>\n",
              "      <td>Smart-seq</td>\n",
              "      <td>172.0</td>\n",
              "      <td>172</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "                                              assay     count_x   count_y  \\\n",
              "3                                         10x 3' v3  59668147.0  96655665   \n",
              "2                                         10x 3' v2  22750589.0  24507159   \n",
              "6                                         10x 5' v2   4363694.0   8748986   \n",
              "5                                         10x 5' v1   7448617.0   7683010   \n",
              "36                                     sci-RNA-seq3   5064268.0   5230353   \n",
              "8                                      10x multiome         NaN   3033426   \n",
              "4                    10x 5' transcription profiling   1968545.0   2125643   \n",
              "14                                         Drop-seq   1048377.0   1048377   \n",
              "23              ScaleBio single cell RNA sequencing    700524.0    700524   \n",
              "0                    10x 3' transcription profiling    665642.0    665642   \n",
              "33                                    microwell-seq    642559.0    642559   \n",
              "11         BD Rhapsody Whole Transcriptome Analysis    269298.0    269298   \n",
              "28                                       Smart-seq2    260350.0    267948   \n",
              "24                                         Seq-Well    255232.0    255232   \n",
              "1                                         10x 3' v1    213017.0    232475   \n",
              "31                                           inDrop     51304.0    136019   \n",
              "30                                          TruDrop    133507.0    133507   \n",
              "15                              GEXSCOPE technology    117962.0    117962   \n",
              "7                          10x gene expression flex     86356.0    117356   \n",
              "10                        BD Rhapsody Targeted mRNA     96145.0     96145   \n",
              "21                                        SPLiT-seq     81936.0     81936   \n",
              "17                                         MARS-seq     70146.0     70146   \n",
              "27                                     Smart-seq v4     63920.0     63920   \n",
              "35  particle-templated instant partition sequencing         NaN     46500   \n",
              "25                                      Seq-Well S3     41560.0     41560   \n",
              "32                                          mCT-seq         NaN     37121   \n",
              "16             HIVE CLX Single-Cell RNAseq Solution         NaN     31000   \n",
              "13                                         CEL-seq2     25416.0     25416   \n",
              "9                             Asteria scRNA-seq kit         NaN     15764   \n",
              "18            Parse Evercode Whole Transcriptome v2         NaN     15500   \n",
              "20                                         SORT-seq     11640.0     11640   \n",
              "22                                         STRT-seq     11357.0      9386   \n",
              "29                                       Smart-seq3      5005.0      5005   \n",
              "19                                       Quartz-seq      2882.0      2882   \n",
              "34                                modified STRT-seq         NaN      1971   \n",
              "12                                          CEL-seq         NaN       796   \n",
              "26                                        Smart-seq       172.0       172   \n",
              "\n",
              "         delta  \n",
              "3   36987518.0  \n",
              "2    1756570.0  \n",
              "6    4385292.0  \n",
              "5     234393.0  \n",
              "36    166085.0  \n",
              "8          NaN  \n",
              "4     157098.0  \n",
              "14         0.0  \n",
              "23         0.0  \n",
              "0          0.0  \n",
              "33         0.0  \n",
              "11         0.0  \n",
              "28      7598.0  \n",
              "24         0.0  \n",
              "1      19458.0  \n",
              "31     84715.0  \n",
              "30         0.0  \n",
              "15         0.0  \n",
              "7      31000.0  \n",
              "10         0.0  \n",
              "21         0.0  \n",
              "17         0.0  \n",
              "27         0.0  \n",
              "35         NaN  \n",
              "25         0.0  \n",
              "32         NaN  \n",
              "16         NaN  \n",
              "13         0.0  \n",
              "9          NaN  \n",
              "18         NaN  \n",
              "20         0.0  \n",
              "22     -1971.0  \n",
              "29         0.0  \n",
              "19         0.0  \n",
              "34         NaN  \n",
              "12         NaN  \n",
              "26         0.0  "
            ]
          },
          "execution_count": 10,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "assays_merged = pd.merge(assays_old, assays_new, how=\"outer\", on=\"assay\")\n",
        "assays_merged[\"delta\"] = assays_merged[\"count_y\"] - assays_merged[\"count_x\"]\n",
        "assays_merged = assays_merged.sort_values(\"count_y\", ascending=False)\n",
        "assays_merged"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "6cbffa71-54d4-4d4f-b2d9-b5044e52ebb8",
      "metadata": {
        "id": "6cbffa71-54d4-4d4f-b2d9-b5044e52ebb8"
      },
      "source": [
        "## Normalized layer for SMART-like data"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 11,
      "id": "4c38ae9f-6772-4bb0-a73d-adf80fd8113a",
      "metadata": {
        "id": "4c38ae9f-6772-4bb0-a73d-adf80fd8113a"
      },
      "outputs": [],
      "source": [
        "# From https://github.com/chanzuckerberg/cellxgene-census/blob/7d3ab1bbaa30dec8b09a96ec38c21374b0073333/tools/cellxgene_census_builder/src/cellxgene_census_builder/build_soma/globals.py#L336 \n",
        "FULL_GENE_ASSAY = {\n",
        "    \"EFO:0003755\": \"FL-cDNA\",\n",
        "    \"EFO:0008747\": \"FRISCR\",\n",
        "    \"EFO:0008763\": \"Hi-SCL\",\n",
        "    \"EFO:0008797\": \"MATQ-seq\",\n",
        "    \"EFO:0008877\": \"Quartz-seq\",\n",
        "    \"EFO:0008930\": \"Smart-seq\",\n",
        "    \"EFO:0008931\": \"Smart-seq2\",\n",
        "    \"EFO:0008956\": \"SUPeR-seq\",\n",
        "    \"EFO:0009999\": \"G&T-Seq\",\n",
        "    \"EFO:0010004\": \"SCRB-seq\",\n",
        "    \"EFO:0010006\": \"scM&T-seq\",\n",
        "    \"EFO:0010022\": \"Smart-3Seq\",\n",
        "    \"EFO:0010058\": \"Fluidigm C1-based SMARTer library preparation\",\n",
        "    \"EFO:0010184\": \"Smart-like\",\n",
        "    \"EFO:0022396\": \"TruSeq\",\n",
        "    \"EFO:0022488\": \"Smart-seq3\",\n",
        "    \"EFO:0022839\": \"STORM-seq\",\n",
        "    \"EFO:0030031\": \"SCOPE-chip\",\n",
        "    \"EFO:0030060\": \"mCT-seq\",\n",
        "    \"EFO:0030061\": \"mcSCRB-seq\",\n",
        "    \"EFO:0700016\": \"Smart-seq v4\",\n",
        "}"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 12,
      "id": "e2bff18f-af73-4844-8835-7b4fd54f339a",
      "metadata": {
        "id": "e2bff18f-af73-4844-8835-7b4fd54f339a",
        "outputId": "6feac678-beda-43fd-d616-6f0d1cc6d6f6"
      },
      "outputs": [
        {
          "data": {
            "application/vnd.microsoft.datawrangler.viewer.v0+json": {
              "columns": [
                {
                  "name": "index",
                  "rawType": "int64",
                  "type": "integer"
                },
                {
                  "name": "assay",
                  "rawType": "object",
                  "type": "string"
                },
                {
                  "name": "count_x",
                  "rawType": "float64",
                  "type": "float"
                },
                {
                  "name": "count_y",
                  "rawType": "int64",
                  "type": "integer"
                },
                {
                  "name": "delta",
                  "rawType": "float64",
                  "type": "float"
                }
              ],
              "ref": "cb23717c-cb76-48cd-8b74-7a13c8d3668a",
              "rows": [
                [
                  "28",
                  "Smart-seq2",
                  "260350.0",
                  "267948",
                  "7598.0"
                ],
                [
                  "27",
                  "Smart-seq v4",
                  "63920.0",
                  "63920",
                  "0.0"
                ],
                [
                  "32",
                  "mCT-seq",
                  null,
                  "37121",
                  null
                ],
                [
                  "29",
                  "Smart-seq3",
                  "5005.0",
                  "5005",
                  "0.0"
                ],
                [
                  "19",
                  "Quartz-seq",
                  "2882.0",
                  "2882",
                  "0.0"
                ],
                [
                  "26",
                  "Smart-seq",
                  "172.0",
                  "172",
                  "0.0"
                ]
              ],
              "shape": {
                "columns": 4,
                "rows": 6
              }
            },
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>assay</th>\n",
              "      <th>count_x</th>\n",
              "      <th>count_y</th>\n",
              "      <th>delta</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>28</th>\n",
              "      <td>Smart-seq2</td>\n",
              "      <td>260350.0</td>\n",
              "      <td>267948</td>\n",
              "      <td>7598.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>27</th>\n",
              "      <td>Smart-seq v4</td>\n",
              "      <td>63920.0</td>\n",
              "      <td>63920</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>32</th>\n",
              "      <td>mCT-seq</td>\n",
              "      <td>NaN</td>\n",
              "      <td>37121</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>29</th>\n",
              "      <td>Smart-seq3</td>\n",
              "      <td>5005.0</td>\n",
              "      <td>5005</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>19</th>\n",
              "      <td>Quartz-seq</td>\n",
              "      <td>2882.0</td>\n",
              "      <td>2882</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>26</th>\n",
              "      <td>Smart-seq</td>\n",
              "      <td>172.0</td>\n",
              "      <td>172</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "           assay   count_x  count_y   delta\n",
              "28    Smart-seq2  260350.0   267948  7598.0\n",
              "27  Smart-seq v4   63920.0    63920     0.0\n",
              "32       mCT-seq       NaN    37121     NaN\n",
              "29    Smart-seq3    5005.0     5005     0.0\n",
              "19    Quartz-seq    2882.0     2882     0.0\n",
              "26     Smart-seq     172.0      172     0.0"
            ]
          },
          "execution_count": 12,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "assays_new_full_gene = assays_merged[assays_merged[\"assay\"].isin(FULL_GENE_ASSAY.values())]\n",
        "assays_new_full_gene"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 13,
      "id": "552c1c43-b4a5-42ef-846f-2a2ce9e62495",
      "metadata": {
        "id": "552c1c43-b4a5-42ef-846f-2a2ce9e62495"
      },
      "outputs": [
        {
          "name": "stderr",
          "output_type": "stream",
          "text": [
            "/home/ubuntu/miniconda3/envs/sc/lib/python3.11/functools.py:909: ImplicitModificationWarning: Transforming to str index.\n",
            "  return dispatch(args[0].__class__)(*args, **kw)\n",
            "/home/ubuntu/miniconda3/envs/sc/lib/python3.11/functools.py:909: ImplicitModificationWarning: Transforming to str index.\n",
            "  return dispatch(args[0].__class__)(*args, **kw)\n"
          ]
        }
      ],
      "source": [
        "# get a few cells from each assay present\n",
        "smartseq1_ids = census_new[\"census_data\"][\"homo_sapiens\"].obs.read(column_names=[\"soma_joinid\"], value_filter = f\"assay == 'Quartz-seq'\").concat().to_pandas()\n",
        "smartseq4_ids = census_new[\"census_data\"][\"homo_sapiens\"].obs.read(column_names=[\"soma_joinid\"], value_filter = f\"assay == 'Smart-seq v4'\").concat().to_pandas()\n",
        "\n",
        "smartseq = smartseq1_ids[\"soma_joinid\"].tolist()[:1000] + smartseq4_ids[\"soma_joinid\"].tolist()[:1000]\n",
        "\n",
        "adata = cellxgene_census.get_anndata(\n",
        "    census_new,\n",
        "    organism=\"homo_sapiens\",\n",
        "    obs_coords=smartseq,\n",
        "    X_layers=[\"normalized\"]\n",
        ")\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 14,
      "id": "87c0ca70-e0aa-4d7e-9bb9-6869d13ef3ff",
      "metadata": {
        "id": "87c0ca70-e0aa-4d7e-9bb9-6869d13ef3ff",
        "outputId": "5a61655f-7058-4246-c1a3-27f7102059fd"
      },
      "outputs": [
        {
          "data": {
            "text/plain": [
              "<Compressed Sparse Row sparse matrix of dtype 'float32'\n",
              "\twith 13036267 stored elements and shape (2000, 61479)>"
            ]
          },
          "execution_count": 14,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "adata.layers[\"normalized\"]"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 15,
      "id": "f4e94760-15f5-40e6-a779-3f317c40c201",
      "metadata": {
        "id": "f4e94760-15f5-40e6-a779-3f317c40c201",
        "outputId": "3009d87c-b98c-42fa-831d-31020ccf94fe"
      },
      "outputs": [
        {
          "data": {
            "application/vnd.microsoft.datawrangler.viewer.v0+json": {
              "columns": [
                {
                  "name": "index",
                  "rawType": "object",
                  "type": "string"
                },
                {
                  "name": "assay",
                  "rawType": "category",
                  "type": "unknown"
                }
              ],
              "ref": "59b7db16-8b36-4a7c-a114-2224ad226d9a",
              "rows": [
                [
                  "0",
                  "Smart-seq v4"
                ],
                [
                  "1000",
                  "Quartz-seq"
                ]
              ],
              "shape": {
                "columns": 1,
                "rows": 2
              }
            },
            "text/plain": [
              "0       Smart-seq v4\n",
              "1000      Quartz-seq\n",
              "Name: assay, dtype: category\n",
              "Categories (37, object): ['10x 3' transcription profiling', '10x 3' v1', '10x 3' v2', '10x 3' v3', ..., 'microwell-seq', 'modified STRT-seq', 'particle-templated instant partition sequencing', 'sci-RNA-seq3']"
            ]
          },
          "execution_count": 15,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "adata.obs[\"assay\"].drop_duplicates()"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 16,
      "id": "d8b14a0d-680a-41b8-b95f-83b142adbab3",
      "metadata": {
        "id": "d8b14a0d-680a-41b8-b95f-83b142adbab3",
        "outputId": "0855a0ca-912c-454a-e02d-183fa39bda7a"
      },
      "outputs": [
        {
          "data": {
            "text/plain": [
              "np.float32(5.2735913e-09)"
            ]
          },
          "execution_count": 16,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "min(adata.layers[\"normalized\"].data)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 17,
      "id": "b21bf46e-59a9-4f11-a21c-713e9b0da7c1",
      "metadata": {
        "id": "b21bf46e-59a9-4f11-a21c-713e9b0da7c1",
        "outputId": "821be60f-b623-49ed-baff-5188b5d21c96"
      },
      "outputs": [
        {
          "data": {
            "text/plain": [
              "np.float32(0.063144684)"
            ]
          },
          "execution_count": 17,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "max(adata.layers[\"normalized\"].data)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 18,
      "id": "ccacd915-536b-4cbd-a280-75d4e393d1e0",
      "metadata": {
        "id": "ccacd915-536b-4cbd-a280-75d4e393d1e0",
        "outputId": "45b8b9a2-9533-490d-943f-4c842ec93e12"
      },
      "outputs": [
        {
          "data": {
            "text/plain": [
              "np.float32(0.99999875)"
            ]
          },
          "execution_count": 18,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "adata.layers[\"normalized\"].sum(axis=1).min()"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 19,
      "id": "66d3f0e1-ff1b-42f1-a970-cfb6f334a28b",
      "metadata": {
        "id": "66d3f0e1-ff1b-42f1-a970-cfb6f334a28b",
        "outputId": "4051b4e0-9f61-4a6b-e632-e9be9eeeb9fa"
      },
      "outputs": [
        {
          "data": {
            "text/plain": [
              "np.float32(1.0000011)"
            ]
          },
          "execution_count": 19,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "adata.layers[\"normalized\"].sum(axis=1).max()"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 20,
      "id": "86fc9f7b-e24a-4446-add2-7485a1c7af1b",
      "metadata": {
        "id": "86fc9f7b-e24a-4446-add2-7485a1c7af1b"
      },
      "outputs": [],
      "source": [
        "adata.layers[\"normalized_local\"] = adata.X.copy()\n",
        "adata.layers[\"normalized_local\"] = csr_matrix(adata.layers[\"normalized_local\"].multiply(1 / adata.var[\"feature_length\"].values[None,:]))\n",
        "adata.layers[\"normalized_local\"] = csr_matrix(adata.layers[\"normalized_local\"].multiply( 1 / adata.layers[\"normalized_local\"].sum(1).A))"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 21,
      "id": "6f833d7e-760c-4e15-8f24-b74860cb6493",
      "metadata": {
        "id": "6f833d7e-760c-4e15-8f24-b74860cb6493",
        "outputId": "4493f1f3-e42c-478b-88cd-b13e63b0d122"
      },
      "outputs": [
        {
          "data": {
            "text/plain": [
              "array([2.45865267e-05, 2.83067595e-06, 2.67409696e-05, ...,\n",
              "       8.28286526e-05, 7.54188263e-05, 1.28428137e-04], shape=(13036267,))"
            ]
          },
          "execution_count": 21,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "adata.layers[\"normalized_local\"].data"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 22,
      "id": "322af3b0-da87-41f3-b09c-fd010ecd1599",
      "metadata": {
        "id": "322af3b0-da87-41f3-b09c-fd010ecd1599",
        "outputId": "218216e2-f283-48f0-e882-3a8d9bb6141a"
      },
      "outputs": [
        {
          "data": {
            "text/plain": [
              "array([2.4586450e-05, 2.8306968e-06, 2.6741065e-05, ..., 8.2828104e-05,\n",
              "       7.5418502e-05, 1.2842938e-04], shape=(13036267,), dtype=float32)"
            ]
          },
          "execution_count": 22,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "adata.layers[\"normalized\"].data"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 23,
      "id": "f7673f43-7b87-4d88-8f14-43afdf0f08c6",
      "metadata": {
        "id": "f7673f43-7b87-4d88-8f14-43afdf0f08c6",
        "outputId": "2340fcf2-b068-417a-bcbb-03171d1ba193"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Decimal points:  10  values differing: n =  9389378 , fracion =  0.7202505134330249\n",
            "Decimal points:  9  values differing: n =  4750525 , fracion =  0.364408384700927\n",
            "Decimal points:  8  values differing: n =  924218 , fracion =  0.07089590908194808\n",
            "Decimal points:  7  values differing: n =  108095 , fracion =  0.008291867602895828\n",
            "Decimal points:  6  values differing: n =  11142 , fracion =  0.0008546925281600937\n",
            "Decimal points:  5  values differing: n =  1096 , fracion =  8.407314762730772e-05\n",
            "Decimal points:  4  values differing: n =  136 , fracion =  1.0432434377111177e-05\n",
            "Decimal points:  3  values differing: n =  2 , fracion =  1.5341815260457615e-07\n",
            "Decimal points:  2  values differing: n =  0 , fracion =  0.0\n",
            "Decimal points:  1  values differing: n =  0 , fracion =  0.0\n",
            "Decimal points:  0  values differing: n =  0 , fracion =  0.0\n"
          ]
        }
      ],
      "source": [
        "pre_calc = adata.layers[\"normalized\"].data.copy()\n",
        "fly_calc = adata.layers[\"normalized_local\"].data.copy()\n",
        "for r in reversed(range(11)):\n",
        "    n_differing = np.nonzero(np.round(pre_calc.copy(), r).astype(np.float32) != np.round(fly_calc.copy(), r).astype(np.float32))[0].shape[0]\n",
        "    percent_differing = n_differing / fly_calc.shape[0]\n",
        "    print(\"Decimal points: \", r, \" values differing: n = \", n_differing, \", fracion = \", percent_differing)"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "3ea4f96c-0fa5-4a8e-83f7-133c5820c032",
      "metadata": {
        "id": "3ea4f96c-0fa5-4a8e-83f7-133c5820c032"
      },
      "source": [
        "## Checking discover schema 7 changes\n",
        "\n",
        "###  Field diff in obs/var"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 24,
      "id": "4e4de7a7",
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "\n",
            "## homo_sapiens\n",
            "Shared obs columns: ['assay', 'assay_ontology_term_id', 'cell_type', 'cell_type_ontology_term_id', 'dataset_id', 'development_stage', 'development_stage_ontology_term_id', 'disease', 'disease_ontology_term_id', 'donor_id', 'is_primary_data', 'n_measured_vars', 'nnz', 'observation_joinid', 'raw_mean_nnz', 'raw_sum', 'raw_variance_nnz', 'self_reported_ethnicity', 'self_reported_ethnicity_ontology_term_id', 'sex', 'sex_ontology_term_id', 'soma_joinid', 'suspension_type', 'tissue', 'tissue_general', 'tissue_general_ontology_term_id', 'tissue_ontology_term_id', 'tissue_type']\n",
            "New obs columns: []\n",
            "Removed obs columns: []\n",
            "\n",
            "## mus_musculus\n"
          ]
        },
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Shared obs columns: ['assay', 'assay_ontology_term_id', 'cell_type', 'cell_type_ontology_term_id', 'dataset_id', 'development_stage', 'development_stage_ontology_term_id', 'disease', 'disease_ontology_term_id', 'donor_id', 'is_primary_data', 'n_measured_vars', 'nnz', 'observation_joinid', 'raw_mean_nnz', 'raw_sum', 'raw_variance_nnz', 'self_reported_ethnicity', 'self_reported_ethnicity_ontology_term_id', 'sex', 'sex_ontology_term_id', 'soma_joinid', 'suspension_type', 'tissue', 'tissue_general', 'tissue_general_ontology_term_id', 'tissue_ontology_term_id', 'tissue_type']\n",
            "New obs columns: []\n",
            "Removed obs columns: []\n"
          ]
        }
      ],
      "source": [
        "for organism in census_old[\"census_data\"].keys():\n",
        "    print(f\"\\n## {organism}\")\n",
        "    old_columns = list(census_old[\"census_data\"][organism].obs.keys())\n",
        "    new_columns = list(census_new[\"census_data\"][organism].obs.keys())\n",
        "\n",
        "    print(f\"Shared obs columns: {sorted(set(old_columns) & set(new_columns))}\")\n",
        "    print(f\"New obs columns: {sorted(set(new_columns) - set(old_columns))}\")\n",
        "    print(f\"Removed obs columns: {sorted(set(old_columns) - set(new_columns))}\")"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 25,
      "id": "92812f04",
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "\n",
            "## homo_sapiens\n",
            "Shared obs columns: ['feature_id', 'feature_length', 'feature_name', 'feature_type', 'n_measured_obs', 'nnz', 'soma_joinid']\n",
            "New obs columns: []\n",
            "Removed obs columns: []\n",
            "\n",
            "## mus_musculus\n",
            "Shared obs columns: ['feature_id', 'feature_length', 'feature_name', 'feature_type', 'n_measured_obs', 'nnz', 'soma_joinid']\n",
            "New obs columns: []\n",
            "Removed obs columns: []\n"
          ]
        }
      ],
      "source": [
        "for organism in census_old[\"census_data\"].keys():\n",
        "    print(f\"\\n## {organism}\")\n",
        "    old_columns = list(census_old[\"census_data\"][organism].ms[\"RNA\"].var.keys())\n",
        "    new_columns = list(census_new[\"census_data\"][organism].ms[\"RNA\"].var.keys())\n",
        "\n",
        "    print(f\"Shared obs columns: {sorted(set(old_columns) & set(new_columns))}\")\n",
        "    print(f\"New obs columns: {sorted(set(new_columns) - set(old_columns))}\")\n",
        "    print(f\"Removed obs columns: {sorted(set(old_columns) - set(new_columns))}\")"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "42f38b03",
      "metadata": {},
      "source": [
        "### New organisms"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 26,
      "id": "bf2c871c",
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/plain": [
              "(43653561, 53384)"
            ]
          },
          "execution_count": 26,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "n_cells = census_new[\"census_data\"][organism].ms[\"RNA\"].X[\"raw\"].shape\n",
        "n_cells"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 30,
      "id": "1092e75e-1d2c-46b1-9e26-84fef485dadb",
      "metadata": {
        "id": "1092e75e-1d2c-46b1-9e26-84fef485dadb"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "\n",
            "## callithrix_jacchus\n",
            "Number of obs cells: 2275451\n",
            "Number of X cells: 2275451\n",
            "Number of var genes: 28348\n",
            "Number of X genes: 28348\n"
          ]
        },
        {
          "name": "stderr",
          "output_type": "stream",
          "text": [
            "/home/ubuntu/miniconda3/envs/sc/lib/python3.11/functools.py:909: ImplicitModificationWarning: Transforming to str index.\n",
            "  return dispatch(args[0].__class__)(*args, **kw)\n",
            "/home/ubuntu/miniconda3/envs/sc/lib/python3.11/functools.py:909: ImplicitModificationWarning: Transforming to str index.\n",
            "  return dispatch(args[0].__class__)(*args, **kw)\n"
          ]
        },
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Min raw X value:  1.0\n",
            "Max raw X value:  301.0\n",
            "Min normalized X value:  1.9187108e-05\n",
            "Max normalized X value:  0.043024063\n",
            "N obs cols:  28\n",
            "N var cols:  7\n",
            "Obs columns:  soma_joinid, dataset_id, assay, assay_ontology_term_id, cell_type, cell_type_ontology_term_id, development_stage, development_stage_ontology_term_id, disease, disease_ontology_term_id, donor_id, is_primary_data, observation_joinid, self_reported_ethnicity, self_reported_ethnicity_ontology_term_id, sex, sex_ontology_term_id, suspension_type, tissue, tissue_ontology_term_id, tissue_type, tissue_general, tissue_general_ontology_term_id, raw_sum, nnz, raw_mean_nnz, raw_variance_nnz, n_measured_vars\n",
            "  First 5 elements of obs column soma_joinid: 0,1,2,3,4\n",
            "  First 5 elements of obs column dataset_id: 787138a6-a421-444e-8ae5-537c1842e70c,787138a6-a421-444e-8ae5-537c1842e70c,787138a6-a421-444e-8ae5-537c1842e70c,787138a6-a421-444e-8ae5-537c1842e70c,787138a6-a421-444e-8ae5-537c1842e70c\n",
            "  First 5 elements of obs column assay: 10x 3' v3,10x 3' v3,10x 3' v3,10x 3' v3,10x 3' v3\n",
            "  First 5 elements of obs column assay_ontology_term_id: EFO:0009922,EFO:0009922,EFO:0009922,EFO:0009922,EFO:0009922\n",
            "  First 5 elements of obs column cell_type: leptomeningeal cell,leptomeningeal cell,endothelial cell,endothelial cell,oligodendrocyte\n",
            "  First 5 elements of obs column cell_type_ontology_term_id: CL:0000708,CL:0000708,CL:0000115,CL:0000115,CL:0000128\n",
            "  First 5 elements of obs column development_stage: post-juvenile adult stage,post-juvenile adult stage,post-juvenile adult stage,post-juvenile adult stage,post-juvenile adult stage\n",
            "  First 5 elements of obs column development_stage_ontology_term_id: UBERON:0000113,UBERON:0000113,UBERON:0000113,UBERON:0000113,UBERON:0000113\n",
            "  First 5 elements of obs column disease: normal,normal,normal,normal,normal\n",
            "  First 5 elements of obs column disease_ontology_term_id: PATO:0000461,PATO:0000461,PATO:0000461,PATO:0000461,PATO:0000461\n",
            "  First 5 elements of obs column donor_id: bi005,bi005,bi005,bi005,bi005\n",
            "  First 5 elements of obs column is_primary_data: False,False,False,False,False\n",
            "  First 5 elements of obs column observation_joinid: hGj_LOlX@R,y_Pdgpv;qS,jLe2(pDnD2,JEPb(Z{+8u,b@`Y%gX}@X\n",
            "  First 5 elements of obs column self_reported_ethnicity: na,na,na,na,na\n",
            "  First 5 elements of obs column self_reported_ethnicity_ontology_term_id: na,na,na,na,na\n",
            "  First 5 elements of obs column sex: male,male,male,male,male\n",
            "  First 5 elements of obs column sex_ontology_term_id: PATO:0000384,PATO:0000384,PATO:0000384,PATO:0000384,PATO:0000384\n",
            "  First 5 elements of obs column suspension_type: nucleus,nucleus,nucleus,nucleus,nucleus\n",
            "  First 5 elements of obs column tissue: primary motor cortex,primary motor cortex,primary motor cortex,primary motor cortex,primary motor cortex\n",
            "  First 5 elements of obs column tissue_ontology_term_id: UBERON:0001384,UBERON:0001384,UBERON:0001384,UBERON:0001384,UBERON:0001384\n",
            "  First 5 elements of obs column tissue_type: tissue,tissue,tissue,tissue,tissue\n",
            "  First 5 elements of obs column tissue_general: brain,brain,brain,brain,brain\n",
            "  First 5 elements of obs column tissue_general_ontology_term_id: UBERON:0000955,UBERON:0000955,UBERON:0000955,UBERON:0000955,UBERON:0000955\n",
            "  First 5 elements of obs column raw_sum: 18692.0,14319.0,12628.0,12443.0,10260.0\n",
            "  First 5 elements of obs column nnz: 5262,4584,4582,4578,3842\n",
            "  First 5 elements of obs column raw_mean_nnz: 3.5522614975294564,3.12369109947644,2.756001745962462,2.717999126256007,2.670484122852681\n",
            "  First 5 elements of obs column raw_variance_nnz: 57.42009657635039,42.99167949958474,17.186904595500117,12.299964359186475,26.22672035915946\n",
            "  First 5 elements of obs column n_measured_vars: 14409,14409,14409,14409,14409\n",
            "  First 5 elements of var column soma_joinid: 0,1,2,3,4\n",
            "  First 5 elements of var column feature_id: ENSCJAG00000000314,ENSCJAG00000005045,ENSCJAG00000018756,ENSCJAG00000001190,ENSCJAG00000002444\n",
            "  First 5 elements of var column feature_name: A1BG,A1CF,A2ML1,A3GALT2,A4GNT\n",
            "  First 5 elements of var column feature_type: protein_coding,protein_coding,protein_coding,protein_coding,protein_coding\n",
            "  First 5 elements of var column feature_length: 2286,1994,4299,1020,1023\n",
            "  First 5 elements of var column nnz: 5510,1537,27463,24380,232\n",
            "  First 5 elements of var column n_measured_obs: 2086650,247040,2271925,2271925,247040\n",
            "\n",
            "## homo_sapiens\n",
            "Number of obs cells: 153128001\n",
            "Number of X cells: 153128001\n",
            "Number of var genes: 61479\n",
            "Number of X genes: 61479\n"
          ]
        },
        {
          "name": "stderr",
          "output_type": "stream",
          "text": [
            "/home/ubuntu/miniconda3/envs/sc/lib/python3.11/functools.py:909: ImplicitModificationWarning: Transforming to str index.\n",
            "  return dispatch(args[0].__class__)(*args, **kw)\n",
            "/home/ubuntu/miniconda3/envs/sc/lib/python3.11/functools.py:909: ImplicitModificationWarning: Transforming to str index.\n",
            "  return dispatch(args[0].__class__)(*args, **kw)\n"
          ]
        },
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Min raw X value:  1.0\n",
            "Max raw X value:  11551.0\n",
            "Min normalized X value:  1.8775929e-05\n",
            "Max normalized X value:  0.91412354\n",
            "N obs cols:  28\n",
            "N var cols:  7\n",
            "Obs columns:  soma_joinid, dataset_id, assay, assay_ontology_term_id, cell_type, cell_type_ontology_term_id, development_stage, development_stage_ontology_term_id, disease, disease_ontology_term_id, donor_id, is_primary_data, observation_joinid, self_reported_ethnicity, self_reported_ethnicity_ontology_term_id, sex, sex_ontology_term_id, suspension_type, tissue, tissue_ontology_term_id, tissue_type, tissue_general, tissue_general_ontology_term_id, raw_sum, nnz, raw_mean_nnz, raw_variance_nnz, n_measured_vars\n",
            "  First 5 elements of obs column soma_joinid: 0,1,2,3,4\n",
            "  First 5 elements of obs column dataset_id: d7476ae2-e320-4703-8304-da5c42627e71,d7476ae2-e320-4703-8304-da5c42627e71,d7476ae2-e320-4703-8304-da5c42627e71,d7476ae2-e320-4703-8304-da5c42627e71,d7476ae2-e320-4703-8304-da5c42627e71\n",
            "  First 5 elements of obs column assay: 10x 3' v3,10x 3' v3,10x 3' v3,10x 3' v3,10x 3' v3\n",
            "  First 5 elements of obs column assay_ontology_term_id: EFO:0009922,EFO:0009922,EFO:0009922,EFO:0009922,EFO:0009922\n",
            "  First 5 elements of obs column cell_type: endothelial cell,malignant cell,fibroblast,fibroblast,macrophage\n",
            "  First 5 elements of obs column cell_type_ontology_term_id: CL:0000115,CL:0001064,CL:0000057,CL:0000057,CL:0000235\n",
            "  First 5 elements of obs column development_stage: 29-year-old stage,29-year-old stage,29-year-old stage,29-year-old stage,29-year-old stage\n",
            "  First 5 elements of obs column development_stage_ontology_term_id: HsapDv:0000123,HsapDv:0000123,HsapDv:0000123,HsapDv:0000123,HsapDv:0000123\n",
            "  First 5 elements of obs column disease: breast cancer,breast cancer,breast cancer,breast cancer,breast cancer\n",
            "  First 5 elements of obs column disease_ontology_term_id: MONDO:0007254,MONDO:0007254,MONDO:0007254,MONDO:0007254,MONDO:0007254\n",
            "  First 5 elements of obs column donor_id: HTAPP-330,HTAPP-330,HTAPP-330,HTAPP-330,HTAPP-330\n",
            "  First 5 elements of obs column is_primary_data: False,False,False,False,False\n",
            "  First 5 elements of obs column observation_joinid: qsW0>t$X%X,>SS{_^o99U,na+Cy?0^77,i%wMdi?~p+,E0Ifjd2j@h\n",
            "  First 5 elements of obs column self_reported_ethnicity: unknown,unknown,unknown,unknown,unknown\n",
            "  First 5 elements of obs column self_reported_ethnicity_ontology_term_id: unknown,unknown,unknown,unknown,unknown\n",
            "  First 5 elements of obs column sex: female,female,female,female,female\n",
            "  First 5 elements of obs column sex_ontology_term_id: PATO:0000383,PATO:0000383,PATO:0000383,PATO:0000383,PATO:0000383\n",
            "  First 5 elements of obs column suspension_type: cell,cell,cell,cell,cell\n",
            "  First 5 elements of obs column tissue: liver,liver,liver,liver,liver\n",
            "  First 5 elements of obs column tissue_ontology_term_id: UBERON:0002107,UBERON:0002107,UBERON:0002107,UBERON:0002107,UBERON:0002107\n",
            "  First 5 elements of obs column tissue_type: tissue,tissue,tissue,tissue,tissue\n",
            "  First 5 elements of obs column tissue_general: liver,liver,liver,liver,liver\n",
            "  First 5 elements of obs column tissue_general_ontology_term_id: UBERON:0002107,UBERON:0002107,UBERON:0002107,UBERON:0002107,UBERON:0002107\n",
            "  First 5 elements of obs column raw_sum: 19636.0,17245.0,14629.0,13740.0,12701.0\n",
            "  First 5 elements of obs column nnz: 7152,5382,3940,5010,4420\n",
            "  First 5 elements of obs column raw_mean_nnz: 2.745525727069351,3.204199182460052,3.712944162436548,2.7425149700598803,2.873529411764706\n",
            "  First 5 elements of obs column raw_variance_nnz: 696.6162574217965,394.56914794619524,1603.1810966863964,347.04808231411005,422.6350538450275\n",
            "  First 5 elements of obs column n_measured_vars: 12621,12621,12621,12621,12621\n",
            "  First 5 elements of var column soma_joinid: 0,1,2,3,4\n",
            "  First 5 elements of var column feature_id: ENSG00000237491,ENSG00000188976,ENSG00000187642,ENSG00000272512,ENSG00000188290\n",
            "  First 5 elements of var column feature_name: LINC01409,NOC2L,PERM1,ENSG00000272512,HES4\n",
            "  First 5 elements of var column feature_type: lncRNA,protein_coding,protein_coding,lncRNA,protein_coding\n",
            "  First 5 elements of var column feature_length: 1089,1244,2765,2086,961\n",
            "  First 5 elements of var column nnz: 12451983,25164122,820116,1171001,23964448\n",
            "  First 5 elements of var column n_measured_obs: 138403531,152794359,142875264,137527561,152776134\n",
            "\n",
            "## macaca_mulatta\n",
            "Number of obs cells: 7010229\n",
            "Number of X cells: 7010229\n",
            "Number of var genes: 34676\n",
            "Number of X genes: 34676\n"
          ]
        },
        {
          "name": "stderr",
          "output_type": "stream",
          "text": [
            "/home/ubuntu/miniconda3/envs/sc/lib/python3.11/functools.py:909: ImplicitModificationWarning: Transforming to str index.\n",
            "  return dispatch(args[0].__class__)(*args, **kw)\n",
            "/home/ubuntu/miniconda3/envs/sc/lib/python3.11/functools.py:909: ImplicitModificationWarning: Transforming to str index.\n",
            "  return dispatch(args[0].__class__)(*args, **kw)\n"
          ]
        },
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Min raw X value:  1.0\n",
            "Max raw X value:  231.0\n",
            "Min normalized X value:  0.00018556416\n",
            "Max normalized X value:  0.5714264\n",
            "N obs cols:  28\n",
            "N var cols:  7\n",
            "Obs columns:  soma_joinid, dataset_id, assay, assay_ontology_term_id, cell_type, cell_type_ontology_term_id, development_stage, development_stage_ontology_term_id, disease, disease_ontology_term_id, donor_id, is_primary_data, observation_joinid, self_reported_ethnicity, self_reported_ethnicity_ontology_term_id, sex, sex_ontology_term_id, suspension_type, tissue, tissue_ontology_term_id, tissue_type, tissue_general, tissue_general_ontology_term_id, raw_sum, nnz, raw_mean_nnz, raw_variance_nnz, n_measured_vars\n",
            "  First 5 elements of obs column soma_joinid: 0,1,2,3,4\n",
            "  First 5 elements of obs column dataset_id: 885d325f-b478-408b-b184-510b4db4f85f,885d325f-b478-408b-b184-510b4db4f85f,885d325f-b478-408b-b184-510b4db4f85f,885d325f-b478-408b-b184-510b4db4f85f,885d325f-b478-408b-b184-510b4db4f85f\n",
            "  First 5 elements of obs column assay: sci-RNA-seq3,sci-RNA-seq3,sci-RNA-seq3,sci-RNA-seq3,sci-RNA-seq3\n",
            "  First 5 elements of obs column assay_ontology_term_id: EFO:0030028,EFO:0030028,EFO:0030028,EFO:0030028,EFO:0030028\n",
            "  First 5 elements of obs column cell_type: serotonergic neuron,serotonergic neuron,serotonergic neuron,serotonergic neuron,serotonergic neuron\n",
            "  First 5 elements of obs column cell_type_ontology_term_id: CL:0000850,CL:0000850,CL:0000850,CL:0000850,CL:0000850\n",
            "  First 5 elements of obs column development_stage: prime adult stage,prime adult stage,prime adult stage,prime adult stage,prime adult stage\n",
            "  First 5 elements of obs column development_stage_ontology_term_id: UBERON:0018241,UBERON:0018241,UBERON:0018241,UBERON:0018241,UBERON:0018241\n",
            "  First 5 elements of obs column disease: normal,normal,normal,normal,normal\n",
            "  First 5 elements of obs column disease_ontology_term_id: PATO:0000461,PATO:0000461,PATO:0000461,PATO:0000461,PATO:0000461\n",
            "  First 5 elements of obs column donor_id: CPRC_animal_id:6J2,CPRC_animal_id:6J2,CPRC_animal_id:6J2,CPRC_animal_id:6J2,CPRC_animal_id:6J2\n",
            "  First 5 elements of obs column is_primary_data: False,False,False,False,False\n",
            "  First 5 elements of obs column observation_joinid: A~B^dDGk3U,;l;z-H`{lx,;1hQlR!gEx,Yv?Msd%OYj,&7I5!1n<}J\n",
            "  First 5 elements of obs column self_reported_ethnicity: na,na,na,na,na\n",
            "  First 5 elements of obs column self_reported_ethnicity_ontology_term_id: na,na,na,na,na\n",
            "  First 5 elements of obs column sex: male,male,male,male,male\n",
            "  First 5 elements of obs column sex_ontology_term_id: PATO:0000384,PATO:0000384,PATO:0000384,PATO:0000384,PATO:0000384\n",
            "  First 5 elements of obs column suspension_type: nucleus,nucleus,nucleus,nucleus,nucleus\n",
            "  First 5 elements of obs column tissue: nucleus accumbens,nucleus accumbens,nucleus accumbens,nucleus accumbens,nucleus accumbens\n",
            "  First 5 elements of obs column tissue_ontology_term_id: UBERON:0001882,UBERON:0001882,UBERON:0001882,UBERON:0001882,UBERON:0001882\n",
            "  First 5 elements of obs column tissue_type: tissue,tissue,tissue,tissue,tissue\n",
            "  First 5 elements of obs column tissue_general: brain,brain,brain,brain,brain\n",
            "  First 5 elements of obs column tissue_general_ontology_term_id: UBERON:0000955,UBERON:0000955,UBERON:0000955,UBERON:0000955,UBERON:0000955\n",
            "  First 5 elements of obs column raw_sum: 180.0,217.0,136.0,238.0,157.0\n",
            "  First 5 elements of obs column nnz: 132,181,112,202,131\n",
            "  First 5 elements of obs column raw_mean_nnz: 1.3636363636363635,1.1988950276243093,1.2142857142857142,1.1782178217821782,1.1984732824427482\n",
            "  First 5 elements of obs column raw_variance_nnz: 1.7751561415683554,1.7157765500306936,0.47619047619047616,0.47554307669572926,0.498766881972989\n",
            "  First 5 elements of obs column n_measured_vars: 29060,29060,29060,29060,29060\n",
            "  First 5 elements of var column soma_joinid: 0,1,2,3,4\n",
            "  First 5 elements of var column feature_id: ENSMMUG00000000001,ENSMMUG00000000002,ENSMMUG00000000005,ENSMMUG00000000006,ENSMMUG00000000007\n",
            "  First 5 elements of var column feature_name: CCNF,TEDC2,NAGA,NEURL3,BET1\n",
            "  First 5 elements of var column feature_type: protein_coding,protein_coding,protein_coding,protein_coding,protein_coding\n",
            "  First 5 elements of var column feature_length: 3240,1749,1942,2872,369\n",
            "  First 5 elements of var column nnz: 81816,9256,23006,1316,39185\n",
            "  First 5 elements of var column n_measured_obs: 7010229,7010229,7010229,7010229,7010229\n",
            "\n",
            "## mus_musculus\n",
            "Number of obs cells: 43653561\n",
            "Number of X cells: 43653561\n",
            "Number of var genes: 53384\n",
            "Number of X genes: 53384\n"
          ]
        },
        {
          "name": "stderr",
          "output_type": "stream",
          "text": [
            "/home/ubuntu/miniconda3/envs/sc/lib/python3.11/functools.py:909: ImplicitModificationWarning: Transforming to str index.\n",
            "  return dispatch(args[0].__class__)(*args, **kw)\n",
            "/home/ubuntu/miniconda3/envs/sc/lib/python3.11/functools.py:909: ImplicitModificationWarning: Transforming to str index.\n",
            "  return dispatch(args[0].__class__)(*args, **kw)\n"
          ]
        },
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Min raw X value:  1.0\n",
            "Max raw X value:  1499.0\n",
            "Min normalized X value:  8.5236505e-05\n",
            "Max normalized X value:  0.31488037\n",
            "N obs cols:  28\n",
            "N var cols:  7\n",
            "Obs columns:  soma_joinid, dataset_id, assay, assay_ontology_term_id, cell_type, cell_type_ontology_term_id, development_stage, development_stage_ontology_term_id, disease, disease_ontology_term_id, donor_id, is_primary_data, observation_joinid, self_reported_ethnicity, self_reported_ethnicity_ontology_term_id, sex, sex_ontology_term_id, suspension_type, tissue, tissue_ontology_term_id, tissue_type, tissue_general, tissue_general_ontology_term_id, raw_sum, nnz, raw_mean_nnz, raw_variance_nnz, n_measured_vars\n",
            "  First 5 elements of obs column soma_joinid: 0,1,2,3,4\n",
            "  First 5 elements of obs column dataset_id: a810e511-c18b-4b2a-8fdf-98a6a0d433a7,a810e511-c18b-4b2a-8fdf-98a6a0d433a7,a810e511-c18b-4b2a-8fdf-98a6a0d433a7,a810e511-c18b-4b2a-8fdf-98a6a0d433a7,a810e511-c18b-4b2a-8fdf-98a6a0d433a7\n",
            "  First 5 elements of obs column assay: 10x 3' v2,10x 3' v2,10x 3' v2,10x 3' v2,10x 3' v2\n",
            "  First 5 elements of obs column assay_ontology_term_id: EFO:0009899,EFO:0009899,EFO:0009899,EFO:0009899,EFO:0009899\n",
            "  First 5 elements of obs column cell_type: fibroblast of connective tissue of prostate,fibroblast of connective tissue of prostate,fibroblast of connective tissue of prostate,fibroblast of connective tissue of prostate,fibroblast of connective tissue of prostate\n",
            "  First 5 elements of obs column cell_type_ontology_term_id: CL:1000299,CL:1000299,CL:1000299,CL:1000299,CL:1000299\n",
            "  First 5 elements of obs column development_stage: 2-month-old stage,2-month-old stage,2-month-old stage,2-month-old stage,2-month-old stage\n",
            "  First 5 elements of obs column development_stage_ontology_term_id: MmusDv:0000062,MmusDv:0000062,MmusDv:0000062,MmusDv:0000062,MmusDv:0000062\n",
            "  First 5 elements of obs column disease: normal,normal,normal,normal,normal\n",
            "  First 5 elements of obs column disease_ontology_term_id: PATO:0000461,PATO:0000461,PATO:0000461,PATO:0000461,PATO:0000461\n",
            "  First 5 elements of obs column donor_id: musAd001,musAd001,musAd001,musAd001,musAd001\n",
            "  First 5 elements of obs column is_primary_data: False,False,False,False,False\n",
            "  First 5 elements of obs column observation_joinid: @nmg?SvW93,3~(zUB2cU3,h}yb5usIXV,g$;axA2`2T,6<Szc0%b+%\n",
            "  First 5 elements of obs column self_reported_ethnicity: na,na,na,na,na\n",
            "  First 5 elements of obs column self_reported_ethnicity_ontology_term_id: na,na,na,na,na\n",
            "  First 5 elements of obs column sex: male,male,male,male,male\n",
            "  First 5 elements of obs column sex_ontology_term_id: PATO:0000384,PATO:0000384,PATO:0000384,PATO:0000384,PATO:0000384\n",
            "  First 5 elements of obs column suspension_type: cell,cell,cell,cell,cell\n",
            "  First 5 elements of obs column tissue: prostate gland,prostate gland,prostate gland,prostate gland,prostate gland\n",
            "  First 5 elements of obs column tissue_ontology_term_id: UBERON:0002367,UBERON:0002367,UBERON:0002367,UBERON:0002367,UBERON:0002367\n",
            "  First 5 elements of obs column tissue_type: tissue,tissue,tissue,tissue,tissue\n",
            "  First 5 elements of obs column tissue_general: prostate gland,prostate gland,prostate gland,prostate gland,prostate gland\n",
            "  First 5 elements of obs column tissue_general_ontology_term_id: UBERON:0002367,UBERON:0002367,UBERON:0002367,UBERON:0002367,UBERON:0002367\n",
            "  First 5 elements of obs column raw_sum: 2019.0,2789.0,1996.0,2238.0,2561.0\n",
            "  First 5 elements of obs column nnz: 1018,1123,966,981,1194\n",
            "  First 5 elements of obs column raw_mean_nnz: 1.9833005893909628,2.483526268922529,2.0662525879917184,2.2813455657492354,2.144891122278057\n",
            "  First 5 elements of obs column raw_variance_nnz: 12.698835899724333,32.763320174665836,61.015295165148736,18.481988391686947,67.07286993784233\n",
            "  First 5 elements of obs column n_measured_vars: 19344,19344,19344,19344,19344\n",
            "  First 5 elements of var column soma_joinid: 0,1,2,3,4\n",
            "  First 5 elements of var column feature_id: ENSMUSG00000021124,ENSMUSG00000039377,ENSMUSG00000085604,ENSMUSG00000085125,ENSMUSG00000029439\n",
            "  First 5 elements of var column feature_name: Vti1b,Hlx,Dhx58os,Gm16070,Sfswap\n",
            "  First 5 elements of var column feature_type: protein_coding,protein_coding,lncRNA,lncRNA,protein_coding\n",
            "  First 5 elements of var column feature_length: 841,1495,897,2024,692\n",
            "  First 5 elements of var column nnz: 7232717,521246,130164,2290557,17837941\n",
            "  First 5 elements of var column n_measured_obs: 43653561,43518432,41430375,42056246,43653561\n",
            "\n",
            "## pan_troglodytes\n",
            "Number of obs cells: 158099\n",
            "Number of X cells: 158099\n",
            "Number of var genes: 23534\n",
            "Number of X genes: 23534\n",
            "Min raw X value:  1.0\n",
            "Max raw X value:  502.0\n",
            "Min normalized X value:  2.9166229e-05\n",
            "Max normalized X value:  0.039647102\n",
            "N obs cols:  28\n",
            "N var cols:  7\n",
            "Obs columns:  soma_joinid, dataset_id, assay, assay_ontology_term_id, cell_type, cell_type_ontology_term_id, development_stage, development_stage_ontology_term_id, disease, disease_ontology_term_id, donor_id, is_primary_data, observation_joinid, self_reported_ethnicity, self_reported_ethnicity_ontology_term_id, sex, sex_ontology_term_id, suspension_type, tissue, tissue_ontology_term_id, tissue_type, tissue_general, tissue_general_ontology_term_id, raw_sum, nnz, raw_mean_nnz, raw_variance_nnz, n_measured_vars\n",
            "  First 5 elements of obs column soma_joinid: 0,1,2,3,4\n",
            "  First 5 elements of obs column dataset_id: 070b36c0-773a-45d7-ac44-92f690aee9ca,070b36c0-773a-45d7-ac44-92f690aee9ca,070b36c0-773a-45d7-ac44-92f690aee9ca,070b36c0-773a-45d7-ac44-92f690aee9ca,070b36c0-773a-45d7-ac44-92f690aee9ca\n",
            "  First 5 elements of obs column assay: 10x 3' v3,10x 3' v3,10x 3' v3,10x 3' v3,10x 3' v3\n",
            "  First 5 elements of obs column assay_ontology_term_id: EFO:0009922,EFO:0009922,EFO:0009922,EFO:0009922,EFO:0009922\n",
            "  First 5 elements of obs column cell_type: astrocyte,L2/3 intratelencephalic projecting glutamatergic neuron,astrocyte,sst GABAergic cortical interneuron,astrocyte\n",
            "  First 5 elements of obs column cell_type_ontology_term_id: CL:0000127,CL:4030059,CL:0000127,CL:4023017,CL:0000127\n",
            "  First 5 elements of obs column development_stage: prime adult stage,prime adult stage,prime adult stage,prime adult stage,prime adult stage\n",
            "  First 5 elements of obs column development_stage_ontology_term_id: UBERON:0018241,UBERON:0018241,UBERON:0018241,UBERON:0018241,UBERON:0018241\n",
            "  First 5 elements of obs column disease: normal,normal,normal,normal,normal\n",
            "  First 5 elements of obs column disease_ontology_term_id: PATO:0000461,PATO:0000461,PATO:0000461,PATO:0000461,PATO:0000461\n",
            "  First 5 elements of obs column donor_id: PTB165,PTB165,PTB165,PTB165,PTB165\n",
            "  First 5 elements of obs column is_primary_data: True,True,True,True,True\n",
            "  First 5 elements of obs column observation_joinid: <|^L79Wk0G,WEfB+lf$7k,E%H#yyY|{0,LT1*ocmvdK,Xv6QW9>HG_\n",
            "  First 5 elements of obs column self_reported_ethnicity: na,na,na,na,na\n",
            "  First 5 elements of obs column self_reported_ethnicity_ontology_term_id: na,na,na,na,na\n",
            "  First 5 elements of obs column sex: male,male,male,male,male\n",
            "  First 5 elements of obs column sex_ontology_term_id: PATO:0000384,PATO:0000384,PATO:0000384,PATO:0000384,PATO:0000384\n",
            "  First 5 elements of obs column suspension_type: nucleus,nucleus,nucleus,nucleus,nucleus\n",
            "  First 5 elements of obs column tissue: dorsolateral prefrontal cortex,dorsolateral prefrontal cortex,dorsolateral prefrontal cortex,dorsolateral prefrontal cortex,dorsolateral prefrontal cortex\n",
            "  First 5 elements of obs column tissue_ontology_term_id: UBERON:0009834,UBERON:0009834,UBERON:0009834,UBERON:0009834,UBERON:0009834\n",
            "  First 5 elements of obs column tissue_type: tissue,tissue,tissue,tissue,tissue\n",
            "  First 5 elements of obs column tissue_general: brain,brain,brain,brain,brain\n",
            "  First 5 elements of obs column tissue_general_ontology_term_id: UBERON:0000955,UBERON:0000955,UBERON:0000955,UBERON:0000955,UBERON:0000955\n",
            "  First 5 elements of obs column raw_sum: 3234.0,15374.0,3477.0,5919.0,2882.0\n",
            "  First 5 elements of obs column nnz: 1807,5297,1966,2932,1838\n",
            "  First 5 elements of obs column raw_mean_nnz: 1.7897066961815162,2.9023975835378515,1.7685656154628688,2.0187585266030013,1.5680087051142546\n",
            "  First 5 elements of obs column raw_variance_nnz: 6.361067854124572,32.20365172784915,8.65735234352957,11.58170190414085,3.0337500881114416\n",
            "  First 5 elements of obs column n_measured_vars: 23534,23534,23534,23534,23534\n",
            "  First 5 elements of var column soma_joinid: 0,1,2,3,4\n",
            "  First 5 elements of var column feature_id: ENSPTRG00000052463,ENSPTRG00000042737,ENSPTRG00000000799,ENSPTRG00000049634,ENSPTRG00000049400\n",
            "  First 5 elements of var column feature_name: ENSPTRG00000052463,ENSPTRG00000042737,HOOK1,RNF223,PRR9\n",
            "  First 5 elements of var column feature_type: protein_coding,protein_coding,protein_coding,protein_coding,protein_coding\n",
            "  First 5 elements of var column feature_length: 213,816,5839,723,351\n",
            "  First 5 elements of var column nnz: 615,38,48407,10,0\n",
            "  First 5 elements of var column n_measured_obs: 158099,158099,158099,158099,158099\n"
          ]
        },
        {
          "name": "stderr",
          "output_type": "stream",
          "text": [
            "/home/ubuntu/miniconda3/envs/sc/lib/python3.11/functools.py:909: ImplicitModificationWarning: Transforming to str index.\n",
            "  return dispatch(args[0].__class__)(*args, **kw)\n",
            "/home/ubuntu/miniconda3/envs/sc/lib/python3.11/functools.py:909: ImplicitModificationWarning: Transforming to str index.\n",
            "  return dispatch(args[0].__class__)(*args, **kw)\n"
          ]
        }
      ],
      "source": [
        "for organism in census_new[\"census_data\"].keys():\n",
        "    print(f\"\\n## {organism}\")\n",
        "    n_cells = census_new[\"census_data\"][organism].obs.domain[0][1]+1\n",
        "    n_cell_x = census_new[\"census_data\"][organism].ms[\"RNA\"].X[\"raw\"].shape[0]\n",
        "    print(f\"Number of obs cells: {n_cells}\")\n",
        "    print(f\"Number of X cells: {n_cell_x}\")\n",
        "\n",
        "    n_genes = census_new[\"census_data\"][organism].ms[\"RNA\"].var.domain[0][1]+1\n",
        "    n_gene_x = census_new[\"census_data\"][organism].ms[\"RNA\"].X[\"raw\"].shape[1]\n",
        "    print(f\"Number of var genes: {n_genes}\")\n",
        "    print(f\"Number of X genes: {n_gene_x}\")\n",
        "\n",
        "    adata = cellxgene_census.get_anndata(\n",
        "        census_new,\n",
        "        organism=organism,\n",
        "        measurement_name=\"RNA\",\n",
        "        X_name=\"raw\",\n",
        "        X_layers=[\"normalized\"],\n",
        "        obs_coords=list(range(1000))\n",
        "    )\n",
        "\n",
        "    print(\"Min raw X value: \", adata.X.data.min())\n",
        "    print(\"Max raw X value: \", adata.X.data.max())\n",
        "    print(\"Min normalized X value: \", adata.layers[\"normalized\"].data.min())\n",
        "    print(\"Max normalized X value: \", adata.layers[\"normalized\"].data.max())\n",
        "    print(\"N obs cols: \", adata.obs.shape[1])\n",
        "    print(\"N var cols: \", adata.var.shape[1])\n",
        "    print(\"Obs columns: \", \", \".join(adata.obs.columns.astype(str).tolist()))\n",
        "    \n",
        "    for obs in adata.obs.columns:\n",
        "        first_elements = \",\".join(adata.obs[obs].astype(str).tolist()[:5])\n",
        "        print(f\"  First 5 elements of obs column {obs}: {first_elements}\")\n",
        "\n",
        "    for var in adata.var.columns:\n",
        "        first_elements = \",\".join(adata.var[var].astype(str).tolist()[:5])\n",
        "        print(f\"  First 5 elements of var column {var}: {first_elements}\")"
      ]
    }
  ],
  "metadata": {
    "colab": {
      "provenance": []
    },
    "kernelspec": {
      "display_name": "sc",
      "language": "python",
      "name": "python3"
    },
    "language_info": {
      "codemirror_mode": {
        "name": "ipython",
        "version": 3
      },
      "file_extension": ".py",
      "mimetype": "text/x-python",
      "name": "python",
      "nbconvert_exporter": "python",
      "pygments_lexer": "ipython3",
      "version": "3.11.14"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 5
}
